diff -upk.orig gdb-6.3.orig/gdb/doc/gdb.texinfo gdb-6.3/gdb/doc/gdb.texinfo
--- gdb-6.3.orig/gdb/doc/gdb.texinfo	2004-10-08 19:19:03 +0000
+++ gdb-6.3/gdb/doc/gdb.texinfo	2004-12-20 18:02:41 +0000
@@ -40,7 +40,7 @@
 @c manuals to an info tree.
 @dircategory Software development
 @direntry
-* Gdb: (gdb).                     The GNU debugger.
+* Gdb: (gdb).                                   The GNU debugger.
 @end direntry
 
 @ifinfo
diff -upk.orig gdb-6.3.orig/gdb/doc/gdbint.texinfo gdb-6.3/gdb/doc/gdbint.texinfo
--- gdb-6.3.orig/gdb/doc/gdbint.texinfo	2004-10-12 19:14:31 +0000
+++ gdb-6.3/gdb/doc/gdbint.texinfo	2004-12-20 18:02:41 +0000
@@ -3,7 +3,7 @@
 @include gdb-cfg.texi
 @dircategory Software development
 @direntry
-* Gdb-Internals: (gdbint).	The GNU debugger's internals.
+* Gdb-Internals: (gdbint).                      The GNU debugger's internals.
 @end direntry
 
 @ifinfo
diff -upk.orig gdb-6.3.orig/gdb/doc/stabs.texinfo gdb-6.3/gdb/doc/stabs.texinfo
--- gdb-6.3.orig/gdb/doc/stabs.texinfo	2004-09-21 21:06:37 +0000
+++ gdb-6.3/gdb/doc/stabs.texinfo	2004-12-20 18:02:41 +0000
@@ -7,7 +7,8 @@
 @c manuals to an info tree.
 @dircategory Software development
 @direntry
-* Stabs: (stabs).                 The "stabs" debugging information format.   
+* Stabs: (stabs).                               The "stabs" debugging
+                                                information format.
 @end direntry
 
 @ifinfo
