--- gdb-6.3.orig/gdb/ada-lang.c	2004-10-15 09:12:24 +0000
+++ gdb-6.3/gdb/ada-lang.c	2014-06-28 15:13:02 +0000
@@ -1206,7 +1206,7 @@ desc_bounds (struct value *arr)
         desc_bounds_type (thin_descriptor_type (type));
       LONGEST addr;
 
-      if (desc_bounds_type == NULL)
+      if (bounds_type == NULL)
         error ("Bad GNAT array descriptor");
 
       /* NOTE: The following calculation is not really kosher, but
