--- gdb-6.1.1/gdb/Makefile.in.orig	Sat Feb 28 18:04:36 2004
+++ gdb-6.1.1/gdb/Makefile.in	Wed Jul  7 23:01:42 2004
@@ -113,10 +113,10 @@
 BFD_CFLAGS = -I$(BFD_DIR) -I$(BFD_SRC)
 
 # Where is the READLINE library?  Typically in ../readline.
-READLINE_DIR = ../readline
-READLINE = $(READLINE_DIR)/libreadline.a
-READLINE_SRC = $(srcdir)/$(READLINE_DIR)
-READLINE_CFLAGS = -I$(READLINE_SRC)/..
+READLINE_SRC=/usr/include/readline
+READLINE=-lreadline
+READLINE_CFLAGS=
+READLINE_DIR=
 
 WARN_CFLAGS = @WARN_CFLAGS@
 WERROR_CFLAGS = @WERROR_CFLAGS@
--- gdb-6.1.1/Makefile.in.orig	Mon Jun 14 22:40:55 2004
+++ gdb-6.1.1/Makefile.in	Wed Jul  7 23:01:42 2004
@@ -571,7 +571,6 @@
     maybe-configure-perl \
     maybe-configure-prms \
     maybe-configure-rcs \
-    maybe-configure-readline \
     maybe-configure-release \
     maybe-configure-recode \
     maybe-configure-sed \
@@ -661,7 +660,6 @@
     maybe-all-perl \
     maybe-all-prms \
     maybe-all-rcs \
-    maybe-all-readline \
     maybe-all-release \
     maybe-all-recode \
     maybe-all-sed \
@@ -755,7 +753,6 @@
     maybe-info-perl \
     maybe-info-prms \
     maybe-info-rcs \
-    maybe-info-readline \
     maybe-info-release \
     maybe-info-recode \
     maybe-info-sed \
@@ -2654,7 +2651,6 @@
     maybe-dvi-perl \
     maybe-dvi-prms \
     maybe-dvi-rcs \
-    maybe-dvi-readline \
     maybe-dvi-release \
     maybe-dvi-recode \
     maybe-dvi-sed \
@@ -4553,7 +4549,6 @@
     maybe-TAGS-perl \
     maybe-TAGS-prms \
     maybe-TAGS-rcs \
-    maybe-TAGS-readline \
     maybe-TAGS-release \
     maybe-TAGS-recode \
     maybe-TAGS-sed \
@@ -6452,7 +6447,6 @@
     maybe-install-info-perl \
     maybe-install-info-prms \
     maybe-install-info-rcs \
-    maybe-install-info-readline \
     maybe-install-info-release \
     maybe-install-info-recode \
     maybe-install-info-sed \
@@ -8433,7 +8427,6 @@
     maybe-installcheck-perl \
     maybe-installcheck-prms \
     maybe-installcheck-rcs \
-    maybe-installcheck-readline \
     maybe-installcheck-release \
     maybe-installcheck-recode \
     maybe-installcheck-sed \
@@ -10332,7 +10325,6 @@
     maybe-mostlyclean-perl \
     maybe-mostlyclean-prms \
     maybe-mostlyclean-rcs \
-    maybe-mostlyclean-readline \
     maybe-mostlyclean-release \
     maybe-mostlyclean-recode \
     maybe-mostlyclean-sed \
@@ -12107,7 +12099,6 @@
     maybe-clean-perl \
     maybe-clean-prms \
     maybe-clean-rcs \
-    maybe-clean-readline \
     maybe-clean-release \
     maybe-clean-recode \
     maybe-clean-sed \
@@ -13896,7 +13887,6 @@
     maybe-distclean-perl \
     maybe-distclean-prms \
     maybe-distclean-rcs \
-    maybe-distclean-readline \
     maybe-distclean-release \
     maybe-distclean-recode \
     maybe-distclean-sed \
@@ -15685,7 +15675,6 @@
     maybe-maintainer-clean-perl \
     maybe-maintainer-clean-prms \
     maybe-maintainer-clean-rcs \
-    maybe-maintainer-clean-readline \
     maybe-maintainer-clean-release \
     maybe-maintainer-clean-recode \
     maybe-maintainer-clean-sed \
@@ -17531,7 +17520,6 @@
     maybe-check-perl \
     maybe-check-prms \
     maybe-check-rcs \
-    maybe-check-readline \
     maybe-check-release \
     maybe-check-recode \
     maybe-check-sed \
@@ -17641,7 +17629,6 @@
     maybe-install-perl \
     maybe-install-prms \
     maybe-install-rcs \
-    maybe-install-readline \
     maybe-install-release \
     maybe-install-recode \
     maybe-install-sed \
@@ -23771,7 +23758,7 @@
 # GDB needs to know that the simulator is being built.
 configure-gdb: maybe-configure-itcl maybe-configure-tcl maybe-configure-tk maybe-configure-sim
 GDB_TK = @GDB_TK@
-all-gdb: maybe-all-libiberty maybe-all-opcodes maybe-all-bfd maybe-all-mmalloc maybe-all-readline maybe-all-bison maybe-all-byacc maybe-all-sim $(gdbnlmrequirements) $(GDB_TK)
+all-gdb: maybe-all-libiberty maybe-all-opcodes maybe-all-bfd maybe-all-mmalloc maybe-all-bison maybe-all-byacc maybe-all-sim $(gdbnlmrequirements) $(GDB_TK)
 install-gdb: maybe-install-tcl maybe-install-tk maybe-install-itcl maybe-install-tix maybe-install-libgui
 configure-libgui: maybe-configure-tcl maybe-configure-tk
 all-libgui: maybe-all-tcl maybe-all-tk maybe-all-itcl
