diff -ur pwdb-0.61.orig/libpwdb/pwdb/pwdb_module.c pwdb-0.61/libpwdb/pwdb/pwdb_module.c
--- pwdb-0.61.orig/libpwdb/pwdb/pwdb_module.c	Sat Apr 10 01:22:02 1999
+++ pwdb-0.61/libpwdb/pwdb/pwdb_module.c	Sun Apr  8 18:54:32 2001
@@ -307,12 +307,14 @@
     while (_pwdb_policy && *_pwdb_policy) {
 	free(*_pwdb_policy++);
     }
+    if (pwdb_policy) free(pwdb_policy);
     pwdb_policy = NULL;
     _pwdb_policy = NULL;
 
     while (_pwdb_group_policy && *_pwdb_group_policy) {
 	free(*_pwdb_group_policy++);
     }
+    if (pwdb_group_policy) free(pwdb_group_policy);
     pwdb_group_policy = NULL;
     _pwdb_group_policy = NULL;
 }
diff -ur pwdb-0.61.orig/libpwdb/pwdb/pwdb_start.c pwdb-0.61/libpwdb/pwdb/pwdb_start.c
--- pwdb-0.61.orig/libpwdb/pwdb/pwdb_start.c	Wed Oct  7 00:58:43 1998
+++ pwdb-0.61/libpwdb/pwdb/pwdb_start.c	Sun Apr  8 18:46:33 2001
@@ -74,7 +74,6 @@
     *e = NULL;
 
     D(("called."));
-    entry->name = NULL;
     if (entry->malloced && entry->value) {  /* name & value were malloc()'d */
         entry->name = _pwdb_delete_string(entry->name);
         memset(entry->value, 0, entry->length);
