diff -ur openssl-0.9.6e.orig/crypto/des/des.h openssl-0.9.6e/crypto/des/des.h
--- openssl-0.9.6e.orig/crypto/des/des.h	Fri Jun 14 01:41:46 2002
+++ openssl-0.9.6e/crypto/des/des.h	Wed Jul 31 05:30:10 2002
@@ -190,7 +190,8 @@
 char *des_fcrypt(const char *buf,const char *salt, char *ret);
 char *des_crypt(const char *buf,const char *salt);
 #if !defined(PERL5) && !defined(__FreeBSD__) && !defined(NeXT) && !defined(_UWIN)
-char *crypt(const char *buf,const char *salt);
+char *_openssl_crypt(const char *buf,const char *salt);
+#define crypt _openssl_crypt
 #endif
 void des_ofb_encrypt(const unsigned char *in,unsigned char *out,int numbits,
 		     long length,des_key_schedule schedule,des_cblock *ivec);
diff -ur openssl-0.9.6e.orig/crypto/des/fcrypt.c openssl-0.9.6e/crypto/des/fcrypt.c
--- openssl-0.9.6e.orig/crypto/des/fcrypt.c	Tue Jan  8 12:19:55 2002
+++ openssl-0.9.6e/crypto/des/fcrypt.c	Wed Jul 31 05:28:39 2002
@@ -62,7 +62,7 @@
 	DES_LONG Eswap0, DES_LONG Eswap1);
 
 #if !defined(PERL5) && !defined(__FreeBSD__) && !defined(NeXT) && !defined(_DARWIN)
-char *crypt(const char *buf, const char *salt)
+char *_openssl_crypt(const char *buf, const char *salt)
 	{
 	return(des_crypt(buf, salt));
 	}
