diff -ur openssl-0.9.6a.orig/crypto/uid.c openssl-0.9.6a/crypto/uid.c
--- openssl-0.9.6a.orig/crypto/uid.c	Fri Apr  6 00:09:37 2001
+++ openssl-0.9.6a/crypto/uid.c	Sun Apr 22 05:54:00 2001
@@ -53,6 +53,7 @@
  *
  */
 
+#define _GNU_SOURCE
 #include <openssl/crypto.h>
 
 #if defined(__OpenBSD__) || (defined(__FreeBSD__) && __FreeBSD__ > 2)
@@ -80,7 +81,7 @@
 	{
 	if (getuid() != geteuid()) return 1;
 	if (getgid() != getegid()) return 1;
-	return 0;
+	return __libc_enable_secure;
 	}
 #endif
 
