diff -ur openssl-0.9.6e.orig/FAQ openssl-0.9.6e/FAQ
--- openssl-0.9.6e.orig/FAQ	Tue Jul 30 14:33:22 2002
+++ openssl-0.9.6e/FAQ	Wed Jul 31 05:38:46 2002
@@ -61,7 +61,7 @@
 * Which is the current version of OpenSSL?
 
 The current version is available from <URL: http://www.openssl.org>.
-OpenSSL 0.9.6e was released on 30 May, 2002.
+OpenSSL 0.9.6e was released on 30 July, 2002.
 
 In addition to the current stable release, you can also access daily
 snapshots of the OpenSSL development version at <URL:
