--- openssl-0.9.7c/crypto/uid.c.orig	Wed Oct  9 13:41:11 2002
+++ openssl-0.9.7c/crypto/uid.c	Sat Feb 28 23:03:10 2004
@@ -53,6 +53,7 @@
  *
  */
 
+#define _GNU_SOURCE
 #include <openssl/crypto.h>
 #include <openssl/opensslconf.h>
 
@@ -77,11 +78,13 @@
 #include OPENSSL_UNISTD
 #include <sys/types.h>
 
+extern int __libc_enable_secure;
+
 int OPENSSL_issetugid(void)
 	{
 	if (getuid() != geteuid()) return 1;
 	if (getgid() != getegid()) return 1;
-	return 0;
+	return __libc_enable_secure;
 	}
 #endif
 
