diff -urp openssl-1.0.0m.orig/engines/e_padlock.c openssl-1.0.0m/engines/e_padlock.c
--- openssl-1.0.0m.orig/engines/e_padlock.c	2014-06-05 09:41:31 +0000
+++ openssl-1.0.0m/engines/e_padlock.c	2014-06-08 13:20:55 +0000
@@ -104,7 +104,9 @@
 # if (defined(__GNUC__) && (defined(__i386__) || defined(__i386))) || \
      (defined(_MSC_VER) && defined(_M_IX86))
 #  define COMPILE_HW_PADLOCK
+#  ifdef OPENSSL_NO_DYNAMIC_ENGINE
 static ENGINE *ENGINE_padlock (void);
+#  endif
 # endif
 #endif
 
@@ -197,6 +199,8 @@ padlock_bind_helper(ENGINE *e)
 	return 1;
 }
 
+#ifdef OPENSSL_NO_DYNAMIC_ENGINE
+
 /* Constructor */
 static ENGINE *
 ENGINE_padlock(void)
@@ -215,6 +219,8 @@ ENGINE_padlock(void)
 	return eng;
 }
 
+#endif
+
 /* Check availability of the engine */
 static int
 padlock_init(ENGINE *e)
