diff -urN logrotate-3.5.9/logrotate.c logrotate/logrotate.c
--- logrotate-3.5.9/logrotate.c	Thu Aug  2 20:08:23 2001
+++ logrotate/logrotate.c	Thu Nov  8 17:44:32 2001
@@ -615,6 +615,16 @@
     int i;
 
     f = fopen(stateFilename, "w");
+
+    /* If file is zero length, we have nothing to lose. pm@debian.org */
+
+    if (feof(f)) {
+	fclose(f);
+	f = NULL;
+	message(MESS_ERROR, "state file %s is zero length, starting fresh\n",
+		stateFilename);
+    }
+    
     if (!f) {
 	message(MESS_ERROR, "error creating state file %s: %s\n", 
 		    stateFilename, strerror(errno));
diff -urN logrotate-3.5.9/logrotate.spec logrotate/logrotate.spec
--- logrotate-3.5.9/logrotate.spec	Tue Sep  4 23:23:23 2001
+++ logrotate/logrotate.spec	Thu Nov  8 17:45:19 2001
@@ -1,6 +1,6 @@
 Summary: Rotates, compresses, removes and mails system log files.
 Name: logrotate
-Version: 3.5.9
+Version: 3.5.10
 Release: 1
 License: GPL
 Group: System Environment/Base
@@ -45,6 +45,9 @@
 %attr(0755, root, root) %dir /etc/logrotate.d
 
 %changelog
+* Thu Nov  8 2001 Preston Brown <pbrown@redhat.com> 3.5.10-1
+- fix from paul martin <pm@debian.org> for zero-length state files
+
 * Tue Sep  4 2001 Preston Brown <pbrown@redhat.com>
 - fix segfault when logfile is in current directory.
 
