--- logrotate-3.6.2/logrotate.c.orig	Tue Jan 29 22:27:17 2002
+++ logrotate-3.6.2/logrotate.c	Mon Mar 11 00:15:35 2002
@@ -65,8 +65,7 @@
 
 static int runScript(char * logfn, char * script) {
     int fd;
-    char filespec[32];
-    char * cmd;
+    char *filespec, *cmd;
     int rc;
 
     if (debug) {
@@ -75,22 +74,26 @@
 	return 0;
     }
 
-    strcpy(filespec, "/tmp/logrotate.XXXXXX");
-    if ((fd = mkstemp(filespec)) < 0 || fchmod(fd, 0700)) {
+    asprintf(&filespec, "%s/logrotate.XXXXXX", getenv("TMPDIR") ?: "/tmp");
+    fd = -1;
+    if (!filespec || (fd = mkstemp(filespec)) < 0 || fchmod(fd, 0700)) {
 	message(MESS_DEBUG, "error creating %s: %s\n", filespec,
 		strerror(errno));
-	if (fd >= 0) close(fd);
+	if (fd >= 0) {
+	    close(fd);
+	    unlink(filespec);
+	}
+	if (filespec)
+	    free(filespec);
 	return -1;
     }
 
-    if (write(fd, "#!/bin/sh\n\n", 11) != 11) {
-	close(fd);
-	unlink(filespec);
-	return -1;
-    }
-    if (write(fd, script, strlen(script)) != strlen(script)) {
+    if (write(fd, "#!/bin/sh\n\n", 11) != 11 ||
+	write(fd, script, strlen(script)) != strlen(script)) {
+	message(MESS_DEBUG, "error writing %s\n", filespec);
 	close(fd);
 	unlink(filespec);
+	free(filespec);
 	return -1;
     }
 
@@ -101,6 +104,7 @@
     rc = system(cmd);
 
     unlink(filespec);
+    free(filespec);
 
     return rc;
 }
