--- logrotate-3.5.9/logrotate.c.orig	Fri Nov 23 00:17:09 2001
+++ logrotate-3.5.9/logrotate.c	Fri Nov 23 00:25:36 2001
@@ -122,7 +122,7 @@
 	    close(fdcurr);
 	    return 1;
 	}
-	if (fchmod(fdsave, sb->st_mode)) {
+	if (fchmod(fdsave, (S_IRUSR | S_IWUSR) & sb->st_mode)) {
 	    message(MESS_ERROR, "error setting mode of %s: %s\n",
 		saveLog, strerror(errno));
 	    close(fdcurr);
@@ -136,6 +136,13 @@
 	    close(fdsave);
 	    return 1;
 	}
+	if (fchmod(fdsave, sb->st_mode)) {
+	    message(MESS_ERROR, "error setting mode of %s: %s\n",
+		saveLog, strerror(errno));
+	    close(fdcurr);
+	    close(fdsave);
+	    return 1;
+	}
 	while ((cnt = read(fdcurr, buf, sizeof(buf))) > 0) {
 	    if (write(fdsave, buf, cnt) != cnt) {
 		message(MESS_ERROR, "error writing to %s: %s\n", 
@@ -426,7 +433,7 @@
 				log->files[logNum], strerror(errno));
 			hasErrors = 1;
 		    } else {
-			if (fchmod(fd, createMode)) {
+			if (fchmod(fd, (S_IRUSR | S_IWUSR) & createMode)) {
 			    message(MESS_ERROR, "error setting mode of "
 				    "%s: %s\n", log->files[logNum], 
 				     strerror(errno));
@@ -434,6 +441,12 @@
 			}
 			if (fchown(fd, createUid, createGid)) {
 			    message(MESS_ERROR, "error setting owner of "
+				    "%s: %s\n", log->files[logNum], 
+				     strerror(errno));
+			    hasErrors = 1;
+			}
+			if (fchmod(fd, createMode)) {
+			    message(MESS_ERROR, "error setting mode of "
 				    "%s: %s\n", log->files[logNum], 
 				     strerror(errno));
 			    hasErrors = 1;
