diff -ur logrotate-3.5.2.orig/config.c logrotate-3.5.2/config.c
--- logrotate-3.5.2.orig/config.c	Fri Aug 11 19:48:28 2000
+++ logrotate-3.5.2/config.c	Sat Nov  4 05:15:58 2000
@@ -403,7 +403,7 @@
 				createOwner, createGroup, &foo);
 		if (rc == 4) {
 		    message(MESS_ERROR, "%s:%d extra arguments for "
-			    "create\n", configFile, lineNum, start[length]);
+			    "create\n", configFile, lineNum);
 		    return 1;
 		}
 
@@ -636,8 +636,8 @@
 
 		if (!S_ISDIR(sb.st_mode)) {
 		    message(MESS_ERROR, "%s:%d olddir path %s is not a "
-				"directory %s: %s\n", configFile, lineNum, 
-				newlog->oldDir, strerror(errno));
+				"directory\n", configFile, lineNum, 
+				newlog->oldDir);
 		    return 1;
 		}
 
diff -ur logrotate-3.5.2.orig/log.c logrotate-3.5.2/log.c
--- logrotate-3.5.2.orig/log.c	Wed Jan 13 02:01:02 1999
+++ logrotate-3.5.2/log.c	Sat Nov  4 04:55:53 2000
@@ -31,11 +31,12 @@
     flags &= ~newFlags;
 }
 
+#if 0
 void log(int fd, char * format, ...) {
     int i = 0;
     char * buf = NULL;
     va_list args;
-    int size = -1;
+    int size;
 
     va_start(args, format);
 
@@ -43,8 +44,8 @@
 	i += 1000;
 	if (buf) free(buf);
 	buf = malloc(i);
-	size = vsnprintf(buf, size, format, args);
-    } while (i == size);
+	size = vsnprintf(buf, i, format, args);
+    } while (size >= i);
 
     write(fd, buf, size);
 
@@ -52,6 +53,7 @@
 
     va_end(args);
 }
+#endif
 
 void message(int level, char * format, ...) {
     va_list args;
diff -ur logrotate-3.5.2.orig/log.h logrotate-3.5.2/log.h
--- logrotate-3.5.2.orig/log.h	Wed Nov 13 03:29:20 1996
+++ logrotate-3.5.2/log.h	Sat Nov  4 04:56:56 2000
@@ -12,8 +12,15 @@
 
 #define LOG_TIMES	(1 << 0)
 
-void message(int level, char * format, ...);
+void message(int level, char * format, ...)
+#ifdef __GNUC__
+	__attribute__ ((format (printf, 2, 3)));
+#else
+	;
+#endif
+#if 0
 void log(int fd, char * format, ...);
+#endif
 void logSetErrorFile(FILE * f);
 void logSetMessageFile(FILE * f);
 void logSetFlags(int flags);
diff -ur logrotate-3.5.2.orig/logrotate.c logrotate-3.5.2/logrotate.c
--- logrotate-3.5.2.orig/logrotate.c	Fri Aug  4 22:27:38 2000
+++ logrotate-3.5.2/logrotate.c	Sat Nov  4 05:14:29 2000
@@ -63,10 +63,9 @@
 
 static int runScript(char * logfn, char * script) {
     int fd;
-    char filespec[20];
+    char filespec[32];
     char * cmd;
     int rc;
-    struct stat sb;
 
     if (debug) {
 	message(MESS_DEBUG, "running script with arg %s: \"%s\"\n", 
@@ -74,25 +73,11 @@
 	return 0;
     }
 
-    strcpy(filespec, "/tmp/logrotXXXXXX");
-
-    mktemp(filespec);
-    if ((fd = open(filespec, O_CREAT | O_EXCL | O_WRONLY, 0700)) < 0) {
+    strcpy(filespec, "/tmp/logrotate.XXXXXX");
+    if ((fd = mkstemp(filespec)) < 0 || fchmod(fd, 0700)) {
 	message(MESS_DEBUG, "error creating %s: %s\n", filespec,
 		strerror(errno));
-	return -1;
-    }
-
-    /* make sure we didn't just follow a symlink as part of a DOS */
-    if (stat(filespec, &sb)) {
-	/* huh? */
-	close(fd);
-	return -1;
-    }
-
-    if (S_ISLNK(sb.st_mode)) {
-	message(MESS_DEBUG, "%s is a symlink! probable DOS attack!!\n");
-	close(fd);
+	if (fd >= 0) close(fd);
 	return -1;
     }
 
@@ -313,8 +298,8 @@
 		log->extension, strlen(log->extension)) == 0) {
 	    char *tempstr;
 	    fileext = log->extension;
-	    tempstr = calloc(strlen(baseName)-strlen(log->extension), sizeof(char));
-	    strncpy(tempstr, baseName,
+	    tempstr = calloc(strlen(baseName)-strlen(log->extension)+1, sizeof(char));
+	    strncat(tempstr, baseName,
 		    strlen(baseName)-strlen(log->extension));
 	    free(baseName);
 	    baseName = tempstr;
@@ -554,7 +539,7 @@
 int rotateLogSet(logInfo * log, logState ** statesPtr, int * numStatesPtr, 
 		 int force) {
     FILE * errorFile;
-    char * errorFileName = NULL;
+    char errorFileName[32];
     int oldstderr = 0, newerr;
     int i;
     int hasErrors = 0;
@@ -602,11 +587,9 @@
 
     if (log->errAddress) {
 	message(MESS_DEBUG, "errors will be mailed to %s\n", log->errAddress);
-	errorFileName = strdup(tmpnam(NULL));
-
-	newerr = open(errorFileName, O_WRONLY | O_CREAT | O_EXCL, 0600);
 
-	if (newerr < 0) {
+	strcpy(errorFileName, "/tmp/logrotate.XXXXXX");
+	if ((newerr = mkstemp(errorFileName)) < 0) {
 	    message(MESS_ERROR, "error creating temporary file %s\n",
 			errorFileName);
 	    return 1;
