diff -ur tcsh-6.14.00-orig/config_f.h tcsh-6.14.00/config_f.h
--- tcsh-6.14.00-orig/config_f.h	2005-03-04 16:46:04 +0300
+++ tcsh-6.14.00/config_f.h	2005-09-18 13:03:01 +0400
@@ -108,7 +108,7 @@
  *		on the name of the tty, and environment.
  *		Does not make sense in the modern window systems!
  */
-#define AUTOLOGOUT
+#undef AUTOLOGOUT
 
 /*
  * SUSPENDED	Newer shells say 'Suspended' instead of 'Stopped'.
