diff -ur tcsh-6.09.00.orig/sh.c tcsh-6.09.00/sh.c
--- tcsh-6.09.00.orig/sh.c	Fri Aug 13 20:32:51 1999
+++ tcsh-6.09.00/sh.c	Sat Nov  4 02:41:13 2000
@@ -35,6 +35,7 @@
  * SUCH DAMAGE.
  */
 #define EXTERN	/* Intern */
+#include <features.h>
 #include "sh.h"
 
 #ifndef lint
@@ -716,7 +717,16 @@
 	shtemp = Strspl(SAVE(strtmp2), doldol);	/* For << */
     }
 #else /* !WINNT */
+#if defined(__GLIBC__) && __GLIBC__ >= 2
+    {
+	char *tmpdir = getenv ("TMPDIR");
+	if (!tmpdir)
+	    tmpdir = "/tmp";
+	shtemp = Strspl(SAVE(tmpdir), SAVE("/sh.XXXXXX")); /* For << */
+    }
+#else
     shtemp = Strspl(STRtmpsh, doldol);	/* For << */
+#endif /* __GLIBC__ */
 #endif /* WINNT */
 
     /*
diff -ur tcsh-6.09.00.orig/sh.dol.c tcsh-6.09.00/sh.dol.c
--- tcsh-6.09.00.orig/sh.dol.c	Sun Oct 25 18:10:05 1998
+++ tcsh-6.09.00/sh.dol.c	Sat Nov  4 02:50:04 2000
@@ -34,6 +34,7 @@
  * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
  * SUCH DAMAGE.
  */
+#include <features.h>
 #include "sh.h"
 
 RCSID("$Id: sh.dol.c,v 3.38 1998/10/25 15:10:05 christos Exp $")
@@ -993,18 +994,30 @@
     register int c;
     Char   *Dv[2];
     Char    obuf[BUFSIZE], lbuf[BUFSIZE], mbuf[BUFSIZE];
-    int     ocnt, lcnt, mcnt;
+    int     ocnt, lcnt, mcnt, fd;
     register Char *lbp, *obp, *mbp;
     Char  **vp;
     bool    quoted;
     char   *tmp;
 
     tmp = short2str(shtemp);
+#if defined(__GLIBC__) && __GLIBC__ >= 2
+    {
+	char *dot = strrchr(tmp, '.');
+	if (!dot)
+	    stderror(ERR_NAME | ERR_NOMATCH);
+	strcpy(dot, ".XXXXXX");
+    }
+
+    if ((fd = mkstemp(tmp)) < 0)
+	stderror(ERR_SYSTEM, tmp, strerror(errno));
+#else
 #ifndef O_CREAT
 # define O_CREAT 0
-    if (creat(tmp, 0600) < 0)
+    if ((fd = creat(tmp, 0600)) < 0)
 	stderror(ERR_SYSTEM, tmp, strerror(errno));
 #endif
+#endif /* __GLIBC__ */
     (void) close(0);
 #ifndef O_TEMPORARY
 # define O_TEMPORARY 0
@@ -1147,4 +1160,5 @@
 	if (pargv)
 	    blkfree(pargv), pargv = 0;
     }
+    close(fd);
 }
