diff -ur tcsh-6.14.00-orig/config_f.h tcsh-6.14.00/config_f.h
--- tcsh-6.14.00-orig/config_f.h	2005-03-04 16:46:04 +0300
+++ tcsh-6.14.00/config_f.h	2006-01-04 13:08:44 +0300
@@ -108,7 +108,7 @@
  *		on the name of the tty, and environment.
  *		Does not make sense in the modern window systems!
  */
-#define AUTOLOGOUT
+#undef AUTOLOGOUT
 
 /*
  * SUSPENDED	Newer shells say 'Suspended' instead of 'Stopped'.
@@ -174,7 +174,7 @@
  * COLOR_LS_F Do you want to use builtin color ls-F ?
  *
  */
-#define COLOR_LS_F
+#undef COLOR_LS_F
 
 /*
  * COLORCAT Do you want to colorful message ?
