diff -pruN tcsh-6.17.00.orig/config_f.h tcsh-6.17.00/config_f.h
--- tcsh-6.17.00.orig/config_f.h	2009-06-25 12:10:56 +0000
+++ tcsh-6.17.00/config_f.h	2010-03-03 11:31:39 +0000
@@ -102,7 +102,7 @@
  *		on the name of the tty, and environment.
  *		Does not make sense in the modern window systems!
  */
-#define AUTOLOGOUT
+#undef AUTOLOGOUT
 
 /*
  * SUSPENDED	Newer shells say 'Suspended' instead of 'Stopped'.
@@ -168,7 +168,7 @@
  * COLOR_LS_F Do you want to use builtin color ls-F ?
  *
  */
-#define COLOR_LS_F
+#undef COLOR_LS_F
 
 /*
  * COLORCAT Do you want to colorful message ?
