--- tcsh-6.10.01/tcsh.man.orig	Thu Apr 26 23:22:36 2001
+++ tcsh-6.10.01/tcsh.man	Fri Jul  6 17:00:32 2001
@@ -2550,15 +2550,15 @@
 .IP "" 4
 > complete find \\
 .br
-\'n/\-name/f/' 'n/\-newer/f/' 'n/\-{,n}cpio/f/' \e
+\&'n/\-name/f/' 'n/\-newer/f/' 'n/\-{,n}cpio/f/' \e
 .br
-\'n/\-exec/c/' 'n/\-ok/c/' 'n/\-user/u/' \e
+\&'n/\-exec/c/' 'n/\-ok/c/' 'n/\-user/u/' \e
 .br
-\'n/\-group/g/' 'n/\-fstype/(nfs 4.2)/' \e
+\&'n/\-group/g/' 'n/\-fstype/(nfs 4.2)/' \e
 .br
-\'n/\-type/(b c d f l p s)/' \e
+\&'n/\-type/(b c d f l p s)/' \e
 .br
-\'c/\-/(name newer cpio ncpio exec ok user \e
+\&'c/\-/(name newer cpio ncpio exec ok user \e
 .br
 group fstype type atime ctime depth inum \e
 .br
@@ -2566,7 +2566,7 @@
 .br
 size xdev)/' \e
 .br
-\'p/*/d/'
+\&'p/*/d/'
 .PP
 This completes words following `\-name', `\-newer', `\-cpio' or `ncpio'
 (note the pattern which matches both) to files,
