--- tcsh-6.10.01/config/linux.orig	Fri Jul  6 12:46:58 2001
+++ tcsh-6.10.01/config/linux	Fri Jul  6 12:48:16 2001
@@ -138,7 +138,7 @@
 #define _FILE_OFFSET_BITS 64
 
 /****************** local defines *********************/
-#undef PW_SHADOW
+#define PW_SHADOW
 #ifndef _PATH_TCSHELL
 #define _PATH_TCSHELL "/bin/tcsh"
 #endif
--- tcsh-6.10.01/config_f.h.orig	Fri Jul  6 13:01:34 2001
+++ tcsh-6.10.01/config_f.h	Fri Jul  6 13:02:59 2001
@@ -97,7 +97,7 @@
 /*
  * NODOT	Don't put "." in the default path, for security reasons
  */
-#undef NODOT
+#define NODOT
 
 /*
  * AUTOLOGOUT	tries to determine if it should set autologout depending
