--- tcsh-6.10.00/sh.c.orig	Sun Nov 12 01:03:35 2000
+++ tcsh-6.10.00/sh.c	Sun Dec  8 16:35:13 2000
@@ -35,6 +35,7 @@
  * SUCH DAMAGE.
  */
 #define EXTERN	/* Intern */
+#include <features.h>
 #include "sh.h"
 
 #ifndef lint
@@ -717,7 +718,16 @@
 	shtemp = Strspl(SAVE(strtmp2), doldol);	/* For << */
     }
 #else /* !WINNT_NATIVE */
+#if defined(__GLIBC__) && __GLIBC__ >= 2
+    {
+       char *tmpdir = getenv ("TMPDIR");
+       if (!tmpdir)
+           tmpdir = "/tmp";
+       shtemp = Strspl(SAVE(tmpdir), SAVE("/sh.XXXXXX")); /* For << */
+    }
+#else
     shtemp = Strspl(STRtmpsh, doldol);	/* For << */
+#endif /* __GLIBC__ */
 #endif /* WINNT_NATIVE */
 
     /*
--- tcsh-6.10.00/sh.dol.c.orig	Sun Nov 19 22:50:43 2000
+++ tcsh-6.10.00/sh.dol.c	Sun Dec  8 16:37:37 2000
@@ -34,6 +34,7 @@
  * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
  * SUCH DAMAGE.
  */
+#include <features.h>
 #include "sh.h"
 
 RCSID("$Id: sh.dol.c,v 3.45 2000/11/19 20:50:43 christos Exp $")
@@ -1020,7 +1021,7 @@
     int c;
     Char   *Dv[2];
     Char    obuf[BUFSIZE], lbuf[BUFSIZE], mbuf[BUFSIZE];
-    int     ocnt, lcnt, mcnt;
+    int     ocnt, lcnt, mcnt, fd;
     register Char *lbp, *obp, *mbp;
     Char  **vp;
     bool    quoted;
@@ -1031,11 +1032,23 @@
 again:
 #endif /* WINNT_NATIVE */
     tmp = short2str(shtemp);
+#if defined(__GLIBC__) && __GLIBC__ >= 2
+    {
+       char *dot = strrchr(tmp, '.');
+       if (!dot)
+           stderror(ERR_NAME | ERR_NOMATCH);
+       strcpy(dot, ".XXXXXX");
+    }
+
+    if ((fd = mkstemp(tmp)) < 0)
+       stderror(ERR_SYSTEM, tmp, strerror(errno));
+#else
 #ifndef O_CREAT
 # define O_CREAT 0
-    if (creat(tmp, 0600) < 0)
+    if ((fd = creat(tmp, 0600)) < 0)
 	stderror(ERR_SYSTEM, tmp, strerror(errno));
 #endif
+#endif /* __GLIBC__ */
     (void) close(0);
 #ifndef O_TEMPORARY
 # define O_TEMPORARY 0
@@ -1190,4 +1203,5 @@
 	if (pargv)
 	    blkfree(pargv), pargv = 0;
     }
+    close(fd);
 }
