diff -puNr run-parts-1.15.orig/run-parts.c run-parts-1.15/run-parts.c
--- run-parts-1.15.orig/run-parts.c	Fri Jan  7 03:00:14 2005
+++ run-parts-1.15/run-parts.c	Fri Jan  7 03:04:04 2005
@@ -55,26 +55,26 @@ void error (char *format,...)
 void version ()
 {
   fprintf (stderr, "Debian GNU/Linux run-parts program, version " VERSION
-"\nCopyright (C) 1994 Ian Jackson, Copyright (C) 1996 Jeff Noxon.
-Copyright (C) 1996,1997,1998,1999 Guy Maor
-This is free software; see the GNU General Public License version 2
-or later for copying conditions.  There is NO warranty.
-");
+"\nCopyright (C) 1994 Ian Jackson, Copyright (C) 1996 Jeff Noxon."
+"\nCopyright (C) 1996,1997,1998,1999 Guy Maor"
+"\nThis is free software; see the GNU General Public License version 2"
+"\nor later for copying conditions.  There is NO warranty."
+"\n");
   exit (0);
 }
 
 
 void usage ()
 {
-  fprintf (stderr, "Usage: run-parts [OPTION]... DIRECTORY
-  --test          print script names which would run, but don't run them.
-  --verbose       print script names before running them.
-  --report        print script names if they produce output.
-  --umask=UMASK   sets umask to UMASK (octal), default is 077.
-  --arg=ARGUMENT  pass ARGUMENT to scripts, use once for each argument.
-  --version       output version information and exit.
-  --help          display this help and exit.
-");
+  fprintf (stderr, "Usage: run-parts [OPTION]... DIRECTORY\n"
+"  --test          print script names which would run, but don't run them.\n"
+"  --verbose       print script names before running them.\n"
+"  --report        print script names if they produce output.\n"
+"  --umask=UMASK   sets umask to UMASK (octal), default is 077.\n"
+"  --arg=ARGUMENT  pass ARGUMENT to scripts, use once for each argument.\n"
+"  --version       output version information and exit.\n"
+"  --help          display this help and exit.\n"
+);
   exit(0);
 }
 
