diff -uNrp ncurses-5.7.old/ncurses/tinfo/access.c ncurses-5.7/ncurses/tinfo/access.c
--- ncurses-5.7.old/ncurses/tinfo/access.c	2010-10-11 14:52:29 +0000
+++ ncurses-5.7/ncurses/tinfo/access.c	2010-10-12 16:37:01 +0000
@@ -30,6 +30,7 @@
  *  Author: Thomas E. Dickey                                                *
  ****************************************************************************/
 
+#include <unistd.h>
 #include <curses.priv.h>
 
 #include <ctype.h>
@@ -156,6 +157,9 @@ _nc_is_file_path(const char *path)
 }
 
 #ifndef USE_ROOT_ENVIRON
+
+extern int __libc_enable_secure;
+
 /*
  * Returns true if we allow application to use environment variables that are
  * used for searching lists of directories, etc.
@@ -163,6 +167,8 @@ _nc_is_file_path(const char *path)
 NCURSES_EXPORT(int)
 _nc_env_access(void)
 {
+    if (__libc_enable_secure)
+	return FALSE;
 #if HAVE_ISSETUGID
     if (issetugid())
 	return FALSE;
