diff -ur ncurses-5.2.orig/ncurses/tinfo/comp_error.c ncurses-5.2/ncurses/tinfo/comp_error.c
--- ncurses-5.2.orig/ncurses/tinfo/comp_error.c	Sun Oct 31 02:00:16 1999
+++ ncurses-5.2/ncurses/tinfo/comp_error.c	Thu Oct 26 20:56:39 2000
@@ -57,10 +57,9 @@
 
 void _nc_set_type(const char *const name)
 {
+	termtype[0] = '\0';
 	if (name)
-		strncpy( termtype, name, MAX_NAME_SIZE );
-	else
-		termtype[0] = '\0';
+		strncat(termtype, name, sizeof(termtype) - 1);
 }
 
 void _nc_get_type(char *name)
diff -ur ncurses-5.2.orig/ncurses/tinfo/lib_termname.c ncurses-5.2/ncurses/tinfo/lib_termname.c
--- ncurses-5.2.orig/ncurses/tinfo/lib_termname.c	Sun Jul 25 01:02:40 1999
+++ ncurses-5.2/ncurses/tinfo/lib_termname.c	Thu Oct 26 20:55:12 2000
@@ -39,7 +39,8 @@
 	T(("termname() called"));
 
 	if (name != 0) {
-		(void) strncpy(ret, name, sizeof(ret) - 1);
+		ret[0] = '\0';
+		(void) strncat(ret, name, sizeof(ret) - 1);
 		name = ret;
 	}
 	return name;
diff -ur ncurses-5.2.orig/ncurses/tinfo/read_termcap.c ncurses-5.2/ncurses/tinfo/read_termcap.c
--- ncurses-5.2.orig/ncurses/tinfo/read_termcap.c	Sat Oct 21 04:34:11 2000
+++ ncurses-5.2/ncurses/tinfo/read_termcap.c	Thu Oct 26 21:21:18 2000
@@ -1013,7 +1013,7 @@
 
 #define PRIVATE_CAP "%s/.termcap"
 
-	if ((h = getenv("HOME")) != NULL && *h != '\0'
+	if (use_terminfo_vars() && (h = getenv("HOME")) != NULL && *h != '\0'
 	    && (strlen(h) + sizeof(PRIVATE_CAP)) < PATH_MAX) {
 	    /* user's .termcap, if any, should override it */
 	    (void) strcpy(envhome, h);
