diff -uNp -r gcc-4.6.2.old/gcc/gcc.c gcc-4.6.2/gcc/gcc.c
--- gcc-4.6.2.old/gcc/gcc.c	2012-02-08 15:25:55.447089603 +0000
+++ gcc-4.6.2/gcc/gcc.c	2012-02-08 15:23:22.278330079 +0000
@@ -658,6 +658,7 @@ proper position among the other output f
     %{flto|flto=*:%<fcompare-debug*} \
     %{flto} %{flto=*} %l " LINK_PIE_SPEC \
    " -z relro "\
+   " -z now "\
    "%X %{o*} %{e*} %{N} %{n} %{r}\
     %{s} %{t} %{u*} %{z} %{Z} %{!nostdlib:%{!nostartfiles:%S}}\
     %{static:} %{L*} %(mfwrap) %(link_libgcc) %o\
