diff -ur gcc-2.95.2.orig/gcc/c-decl.c gcc-2.95.2/gcc/c-decl.c
--- gcc-2.95.2.orig/gcc/c-decl.c	Mon Apr 12 18:05:29 1999
+++ gcc-2.95.2/gcc/c-decl.c	Wed Nov  8 10:10:51 2000
@@ -1951,6 +1951,7 @@
 	{
 	  /* Since the type is OLDDECL's, make OLDDECL's size go with.  */
 	  DECL_SIZE (newdecl) = DECL_SIZE (olddecl);
+	  DECL_MODE (newdecl) = DECL_MODE (olddecl);
 	  if (TREE_CODE (olddecl) != FUNCTION_DECL)
 	    if (DECL_ALIGN (olddecl) > DECL_ALIGN (newdecl))
 	      DECL_ALIGN (newdecl) = DECL_ALIGN (olddecl);
