--- gcc-2.95.2/gcc/glimits.h.orig	Fri Nov 17 10:44:44 2000
+++ gcc-2.95.2/gcc/glimits.h	Fri Nov 17 10:46:41 2000
@@ -64,7 +64,7 @@
 /* Minimum and maximum values a `signed long int' can hold.
    (Same as `int').  */
 #ifndef __LONG_MAX__
-#if defined (__alpha__) || defined (__sparc_v9__) || defined (__sparcv9)
+#if defined (__alpha__) || defined (__arch64__)
 #define __LONG_MAX__ 9223372036854775807L
 #else
 #define __LONG_MAX__ 2147483647L
