--- gcc-2.95.3/gcc/glimits.h.orig	Mon Mar 27 21:22:10 2000
+++ gcc-2.95.3/gcc/glimits.h	Sun Mar 18 04:56:57 2001
@@ -64,7 +64,7 @@
 /* Minimum and maximum values a `signed long int' can hold.
    (Same as `int').  */
 #ifndef __LONG_MAX__
-#if defined (__alpha__) || (defined (_ARCH_PPC) && defined (__64BIT__)) || defined (__sparc_v9__) || defined (__sparcv9)
+#if defined (__alpha__) || (defined (_ARCH_PPC) && defined (__64BIT__)) || defined (__arch64__)
 #define __LONG_MAX__ 9223372036854775807L
 #else
 #define __LONG_MAX__ 2147483647L
