diff -ur xinetd-2.3.10.orig/libs/src/sio/siosup.c xinetd-2.3.10/libs/src/sio/siosup.c
--- xinetd-2.3.10.orig/libs/src/sio/siosup.c	Tue Dec  3 22:08:42 2002
+++ xinetd-2.3.10/libs/src/sio/siosup.c	Sat Jan 18 11:02:05 2003
@@ -764,9 +764,16 @@
  */
 int Sdone( int fd )
 {
-   __sio_descriptor_t *dp = &__sio_descriptors[ fd ] ;
+   __sio_descriptor_t *dp ;
 
-   if ( fd >= __sio_n_descriptors || fd < __sio_n_descriptors || ! DESCRIPTOR_INITIALIZED( dp ) )
+   if ( fd < 0 || fd >= __sio_n_descriptors )
+   {
+      errno = EBADF ;
+      return( SIO_ERR ) ;
+   }
+
+   dp = &__sio_descriptors[ fd ] ;
+   if ( ! DESCRIPTOR_INITIALIZED( dp ) )
    {
       errno = EBADF ;
       return( SIO_ERR ) ;
diff -ur xinetd-2.3.10.orig/xinetd/service.c xinetd-2.3.10/xinetd/service.c
--- xinetd-2.3.10.orig/xinetd/service.c	Tue Dec 10 00:48:51 2002
+++ xinetd-2.3.10/xinetd/service.c	Sat Jan 18 11:06:49 2003
@@ -353,7 +353,11 @@
 #ifdef HAVE_DNSREGISTRATION
    char *srvname;
 
-   asprintf(&srvname, "_%s._%s", scp->sc_name, scp->sc_protocol.name);
+   if (asprintf(&srvname, "_%s._%s", scp->sc_name, scp->sc_protocol.name) < 0)
+   {
+      deactivate( sp ) ;
+      return( FAILED ) ;
+   }
    scp->sc_mdnscon = DNSServiceRegistrationCreate("", srvname, "", htons(scp->sc_port), "", mdns_callback, NULL);
 #endif
 
