diff -uNr xinetd-2.1.8.9pre10.orig/xinetd/includedir.c xinetd-2.1.8.9pre10/xinetd/includedir.c
--- xinetd-2.1.8.9pre10.orig/xinetd/includedir.c	Tue Aug 15 13:32:49 2000
+++ xinetd-2.1.8.9pre10/xinetd/includedir.c	Wed Aug 30 10:38:34 2000
@@ -25,6 +25,21 @@
 	return strcmp(a[0], b[0]);
 }
 
+int skipfunc(char *filename)
+{
+    int filenamelen;
+
+    filenamelen=strlen(filename);
+    if( (strncmp(".rpmsave",&(filename[filenamelen-8]),9)==0)
+        || (strncmp(".rpmorig",&(filename[filenamelen-8]),9)==0)
+        || (strncmp(".rpmnew",&(filename[filenamelen-7]),8)==0)
+        || (strncmp("~",&(filename[filenamelen-1]),2)==0)){
+        return 1;
+    }
+    return 0;
+}
+
+
 void handle_includedir(char *service_name, struct configuration *confp)
 {
 	char *filename;
@@ -97,7 +112,14 @@
 			parsemsg( LOG_ERR, func, "Unable to stat includedir file %s\n", filename);
 			continue;
 		}
-	
+
+                /* Skip files with suffixes .rpmsave, .rpmorig, .rpmnew */
+
+                if(skipfunc(filename)){
+                    parsemsg( LOG_DEBUG,func,"Skipping configuration file: %s",filename);
+                    continue;
+                }
+                
 		/* Only open it if it's a regular file. */
 		if( !S_ISREG(sb.st_mode) )
 			continue;
