diff -ur xinetd-2.3.11.orig/config.h.in xinetd-2.3.11/config.h.in
--- xinetd-2.3.11.orig/config.h.in	Wed Feb 19 18:02:05 2003
+++ xinetd-2.3.11/config.h.in	Wed Apr 16 06:13:34 2003
@@ -123,7 +123,7 @@
 #undef socklen_t
 
 /* Versioning Information */
-#define XINETD_VERSION "xinetd Version 20030122"
+#define XINETD_VERSION "xinetd Version 2.3.11"
 #define MISC_VERSION   "MISC Version 1.2.2"
 #define PSET_VERSION   "PSET Version 1.2.0"
 #define SIO_VERSION    "SIO Version 1.6.2"
