--- xinetd-2.1.8.9pre15/xinetd/main.c.orig	Mon May 14 23:06:01 2001
+++ xinetd-2.1.8.9pre15/xinetd/main.c	Tue Jun 12 04:50:03 2001
@@ -59,14 +59,14 @@
    "loadavg "
 #endif
 #if !defined(INET6) && !defined(LIBWRAP) && !defined(HAVE_LOADAVG)
-   "no "
+   "no options "
 #endif
-   "options compiled in."
+   "compiled in."
    , XINETD_VERSION );
 
    msg( LOG_NOTICE, func, "Started working: %d available service%s",
       ps.rws.available_services,
-         ( ps.rws.available_services > 1 ) ? "s" : "" ) ;
+         ( ps.rws.available_services != 1 ) ? "s" : "" ) ;
 
    if ( cc_option )
 #ifndef NO_TIMERS
