diff -urp xinetd-2.3.14/xinetd/options.c xinetd/xinetd/options.c
--- xinetd-2.3.14/xinetd/options.c	2005-10-05 21:45:41 +0000
+++ xinetd/xinetd/options.c	2007-09-20 17:01:52 +0000
@@ -18,7 +18,7 @@
 #include "internals.h" /* for enable_periodic_check() */
 
 /*
- * $Id: options.c,v 1.7 2005/10/05 21:45:41 bbraun Exp $
+ * $Id: options.c,v 1.8 2007/09/20 17:01:52 bbraun Exp $
  */
 
 
@@ -78,7 +78,7 @@ int opt_recognize( int argc, char *argv[
          {
             if ( ++arg == argc )
                usage() ;
-            if ( parse_ull( argv[ arg ], 10, NUL, &ullarg_1 ) || ullarg_1 < 0 )
+            if ( parse_ull( argv[ arg ], 10, NUL, &ullarg_1 ) < 0 )
                usage() ;
             ps.ros.process_limit = (rlim_t)ullarg_1 ;
             if( ps.ros.process_limit != ullarg_1 )
@@ -98,7 +98,7 @@ int opt_recognize( int argc, char *argv[
          else if ( strcmp( &argv[ arg ][ 1 ], "logprocs" ) == 0 ) {
             if ( ++arg == argc )
                usage() ;
-            if ( parse_uint( argv[ arg ], 10, NUL, &uarg_1 ) || uarg_1 < 0 )
+            if ( parse_uint( argv[ arg ], 10, NUL, &uarg_1 ) < 0 )
                usage() ;
             logprocs_option_arg = uarg_1 ;
             logprocs_option = 1 ;
diff -urp xinetd-2.3.14/xinetd/parsers.c xinetd/xinetd/parsers.c
--- xinetd-2.3.14/xinetd/parsers.c	2005-10-05 21:45:41 +0000
+++ xinetd/xinetd/parsers.c	2007-09-20 17:01:52 +0000
@@ -1375,7 +1375,7 @@ status_e rlim_cpu_parser( pset_h values,
       SC_RLIM_CPU(scp) = (rlim_t)RLIM_INFINITY ;
    else
    {
-      if ( parse_ull(cpu_str, 10, -1, &cpu_int) || cpu_int < 0 )
+      if ( parse_ull(cpu_str, 10, -1, &cpu_int) < 0 )
       {
          parsemsg( LOG_ERR, func,
             "CPU limit is invalid: %s", cpu_str ) ;
