diff -urp procps-ng-3.3.14-qualys/procio.c procps-ng-3.3.14-owl/procio.c
--- procps-ng-3.3.14-qualys/procio.c	2018-04-10 11:26:34 +0000
+++ procps-ng-3.3.14-owl/procio.c	2018-05-23 13:39:01 +0000
@@ -32,6 +32,10 @@
 #include <sys/types.h>
 #include <unistd.h>
 
+#ifndef O_CLOEXEC
+#define O_CLOEXEC 02000000
+#endif
+
 typedef struct pcookie {
 	char	*buf;
 	size_t	count;
@@ -124,7 +128,11 @@ FILE *fprocopen(const char *path, const
 	cookie->length = 0;
 	cookie->delim = delim;
 
+#ifdef AT_FDCWD
 	cookie->fd = openat(AT_FDCWD, path, flags);
+#else
+	cookie->fd = open(path, flags);
+#endif
 	if (cookie->fd < 0) {
 		int errsv = errno;
 		free(cookie->buf);
