--- pcre/pcre-config.in
+++ pcre/pcre-config.in
@@ -12,16 +12,6 @@ if test $# -eq 0; then
       exit 1
 fi
 
-libR=
-case `uname -s` in
-  *SunOS*)
-  libR=" -R@libdir@"
-  ;;
-  *BSD*)
-  libR=" -Wl,-R@libdir@"
-  ;;
-esac
-
 while test $# -gt 0; do
   case "$1" in
   -*=*) optarg=`echo "$1" | sed 's/[-_a-zA-Z0-9]*=//'` ;;
@@ -55,10 +45,10 @@ while test $# -gt 0; do
       echo $includes
       ;;
     --libs-posix)
-      echo -L@libdir@$libR -lpcreposix -lpcre
+      echo -lpcreposix -lpcre
       ;;
     --libs)
-      echo -L@libdir@$libR -lpcre
+      echo -lpcre
       ;;
     *)
       echo "${usage}" 1>&2
