diff -urp vzctl-3.0.23.orig/etc/init.d/vz-redhat.in vzctl-3.0.23/etc/init.d/vz-redhat.in
--- vzctl-3.0.23.orig/etc/init.d/vz-redhat.in	2008-10-30 13:24:43 +0000
+++ vzctl-3.0.23/etc/init.d/vz-redhat.in	2012-02-27 06:39:33 +0000
@@ -19,7 +19,7 @@
 # OpenVZ startup script, used for redhat and debian related distributions.
 
 ###
-# chkconfig: 2345 96 88
+# chkconfig: - 96 88
 # description: OpenVZ startup script.
 ###
 
