diff -urp vzctl-3.0.23.orig/etc/init.d/vz-redhat.in vzctl-3.0.23/etc/init.d/vz-redhat.in
--- vzctl-3.0.23.orig/etc/init.d/vz-redhat.in	2008-10-30 13:24:43 +0000
+++ vzctl-3.0.23/etc/init.d/vz-redhat.in	2012-02-27 06:43:17 +0000
@@ -19,7 +19,7 @@
 # OpenVZ startup script, used for redhat and debian related distributions.
 
 ###
-# chkconfig: 2345 96 88
+# chkconfig: - 96 88
 # description: OpenVZ startup script.
 ###
 
@@ -57,22 +57,30 @@ MODULES_OTHER=
 NET_MODULES=
 IPT_MODULES=
 
+clean_module_list()
+{
+	local mod
+	for mod in $*; do
+		echo "${MODULES_DISABLED}" | fgrep -qw "${mod}" || echo "${mod}"
+	done
+}
+
 if [ "${MODULES_DISABLED}" != "yes" ]; then
-	PRELOAD_MODULES="af_packet"
-	MODULES="vzmon vzdquota vzdev"
+	PRELOAD_MODULES=`clean_module_list af_packet`
+	MODULES=`clean_module_list vzmon vzdquota vzdev`
 	CPT_MODULES="vzcpt vzrst"
-	MODULES_OTHER="vzcompat ${CPT_MODULES}"
+	MODULES_OTHER=`clean_module_list vzcompat ${CPT_MODULES}`
 	VNET_MODULES="vznetdev vznet"
 	VETH_MODULES="vzethdev"
-	NET_MODULES="${VNET_MODULES} ${VETH_MODULES}"
+	NET_MODULES=`clean_module_list ${VNET_MODULES} ${VETH_MODULES}`
 	if [ "${VZWDOG}" = "yes" ]; then
-		MODULES="${MODULES} vzwdog"
+		MODULES=`clean_module_list ${MODULES} vzwdog`
 	fi
-	IPT_MODULES="ip_tables ${IPTABLES} xt_tcpudp"
+	IPT_MODULES=`clean_module_list ip_tables ${IPTABLES} xt_tcpudp`
 	if [ "${IPV6}" = "yes" ]; then
-    		IPT_MODULES="${IPT_MODULES} ${IP6TABLES}"
+		IPT_MODULES=`clean_module_list ${IPT_MODULES} ${IP6TABLES}`
 	fi
-	VZFS_MODULES="simfs"
+	VZFS_MODULES=`clean_module_list simfs`
 fi
 
 
@@ -374,13 +382,13 @@ start_ves()
 	for veid in $velist; do
 		[ "${veid}" = "0" ] && continue
 		__echo "Starting CT ${veid}: "
-		if [ "x${VZFASTBOOT}" = "xyes" -a "x${DISK_QUOTA}" = "xyes" ];
+		if [ \( "x${VZFASTBOOT}" = "xyes" -o "x${VZFASTBOOT}" = "xextreme" \) -a "x${DISK_QUOTA}" = "xyes" ];
 		then
 			$VZQUOTA stat ${veid} >/dev/null 2>&1
 			if [ $? -eq 6 ]; then
 				if $VZQUOTA show ${veid} 2>&1 | grep "vzquota : (warning) Quota is running" >/dev/null 2>&1; then
 					$VZQUOTA on ${veid} --nocheck >/dev/null 2>&1
-					need_restart="${need_restart} ${veid}"
+					test "x${VZFASTBOOT}" = "xextreme" || need_restart="${need_restart} ${veid}"
 				fi
 			fi
 		fi
