--- tinycdb-0.75.orig/cdb.c	2005-04-18 00:26:40 +0000
+++ tinycdb-0.75/cdb.c	2005-11-04 20:27:53 +0000
@@ -299,7 +299,7 @@ dofile_cdb(struct cdb_make *cdbmp, FILE 
 }
 
 static void
-dofile_ln(struct cdb_make *cdbmp, FILE *f, const char *fn, int flags)
+dofile_ln(struct cdb_make *cdbmp, FILE *f, int flags)
 {
   char *k, *v;
   while(fgets(buf, blen, f) != NULL) {
@@ -332,7 +332,7 @@ static void
 dofile(struct cdb_make *cdbmp, FILE *f, const char *fn, int flags)
 {
   if (flags & F_MAP)
-    dofile_ln(cdbmp, f, fn, flags);
+    dofile_ln(cdbmp, f, flags);
   else
     dofile_cdb(cdbmp, f, fn, flags);
   if (ferror(f))
