--- chkconfig-1.3.9/Makefile.orig	Thu Dec 12 19:34:37 2002
+++ chkconfig-1.3.9/Makefile	Fri Feb  6 11:54:39 2004
@@ -16,7 +16,7 @@
 OBJS=chkconfig.o leveldb.o
 NTOBJS=ntsysv.o leveldb.o
 
-all: subdirs $(PROG) ntsysv alternatives
+all: subdirs $(PROG) alternatives
 
 subdirs:
 	for d in $(SUBDIRS); do \
@@ -59,7 +59,6 @@
 	[ -d $(instroot)/$(ALTDATADIR) ] || mkdir -p -m 755 $(instroot)/$(ALTDATADIR)
 
 	install -m 755 $(PROG) $(instroot)/$(BINDIR)/$(PROG)
-	install -m 755 ntsysv $(instroot)/$(USRSBINDIR)/ntsysv
 	install -m 755 alternatives $(instroot)/$(USRSBINDIR)/alternatives
 	ln -s alternatives $(instroot)/$(USRSBINDIR)/update-alternatives
 	
