diff -urN chkconfig-1.2.16.org/Makefile chkconfig-1.2.16/Makefile
--- chkconfig-1.2.16.org/Makefile	Wed Jul 19 04:39:06 2000
+++ chkconfig-1.2.16/Makefile	Sat Sep 23 23:42:32 2000
@@ -13,7 +13,7 @@
 OBJS=chkconfig.o leveldb.o
 NTOBJS=ntsysv.o leveldb.o
 
-all: subdirs $(PROG) ntsysv
+all: subdirs $(PROG) 
 
 subdirs:
 	for d in $(SUBDIRS); do \
@@ -48,7 +48,6 @@
 	[ -d $(instroot)/$(MANDIR)/man5 ] || mkdir -p $(instroot)/$(MANDIR)/man5
 
 	install -s -m 755 $(PROG) $(instroot)/$(BINDIR)/$(PROG)
-	install -s -m 755 ntsysv $(instroot)/$(USRSBINDIR)/ntsysv
 	for i in $(MAN); do \
 		install -m 644 $$i $(instroot)/$(MANDIR)/man`echo $$i | sed "s/.*\.//"`/$$i ; \
 	done
