diff -urp chkconfig-1.3.9.orig/chkconfig.c chkconfig-1.3.9/chkconfig.c
--- chkconfig-1.3.9.orig/chkconfig.c	Mon May 19 03:37:32 2003
+++ chkconfig-1.3.9/chkconfig.c	Sun Feb  8 19:37:47 2004
@@ -287,12 +287,14 @@ static int listService(char * item) {
     closedir(dir);
 	
     if (isXinetdEnabled()) {
-	    printf(_("xinetd based services:\n"));
 	    if (!(dir = opendir(XINETDDIR))) {
+#if 0
 		    fprintf(stderr, _("failed to open directory %s: %s\n"),
 			    XINETDDIR, strerror(err));
-		    return 1;
+#endif
+		    return 0;
 	    }
+	    printf(_("xinetd based services:\n"));
 	    while ((ent = readdir(dir))) {
 		    const char *dn;
 		    struct service s;
