--- e2fsprogs-1.41.5/lib/ext2fs/tst_bitops.c.orig	2009-02-28 03:56:41.000000000 +0000
+++ e2fsprogs-1.41.5/lib/ext2fs/tst_bitops.c	2009-04-29 21:21:00.000000000 +0000
@@ -102,8 +102,8 @@
 	/* Do bigarray test */
 	bigarray = malloc(1 << 29);
 	if (!bigarray) {
-		fprintf(stderr, "Failed to allocate scratch memory!\n");
-		exit(1);
+		printf("bigarray tests skipped.\n");
+		exit(0);
 	}
 
         bigarray[BIG_TEST_BIT >> 3] = 0;
