--- e2fsprogs-1.27/debugfs/debugfs.c.orig	Tue Mar  5 10:31:15 2002
+++ e2fsprogs-1.27/debugfs/debugfs.c	Mon Apr 15 22:58:33 2002
@@ -1415,8 +1415,6 @@
 	int		catastrophic = 0;
 	
 	initialize_ext2_error_table();
-	fprintf (stderr, "debugfs %s (%s)\n", E2FSPROGS_VERSION,
-		 E2FSPROGS_DATE);
 
 	while ((c = getopt (argc, argv, "iwcR:f:b:s:V")) != EOF) {
 		switch (c) {
@@ -1445,10 +1443,15 @@
 			break;
 		case 'V':
 			/* Print version number and exit */
+			fprintf (stderr, "debugfs %s, %s for EXT2 FS %s, %s\n",
+				E2FSPROGS_VERSION, E2FSPROGS_DATE,
+				EXT2FS_VERSION, EXT2FS_DATE);
 			fprintf(stderr, "\tUsing %s\n",
 				error_message(EXT2_ET_BASE));
 			exit(0);
 		default:
+			fprintf (stderr, "debugfs %s (%s)\n", E2FSPROGS_VERSION,
+				 E2FSPROGS_DATE);
 			com_err(argv[0], 0, usage);
 			return 1;
 		}
--- e2fsprogs-1.27/e2fsck/unix.c.orig	Fri Mar  8 08:05:02 2002
+++ e2fsprogs-1.27/e2fsck/unix.c	Mon Apr 15 22:58:33 2002
@@ -706,7 +706,7 @@
 	init_resource_track(&ctx->global_rtrack);
 #endif
 
-	if (!(ctx->options & E2F_OPT_PREEN) || show_version_only)
+	if (!(ctx->options & (E2F_OPT_PREEN|E2F_OPT_NO|E2F_OPT_YES)) || show_version_only)
 		fprintf (stderr, "e2fsck %s (%s)\n", my_ver_string,
 			 my_ver_date);
 
@@ -718,9 +718,7 @@
 	
 	check_mount(ctx);
 	
-	if (!(ctx->options & E2F_OPT_PREEN) &&
-	    !(ctx->options & E2F_OPT_NO) &&
-	    !(ctx->options & E2F_OPT_YES)) {
+	if (!(ctx->options & (E2F_OPT_PREEN|E2F_OPT_NO|E2F_OPT_YES))) {
 		if (!isatty (0) || !isatty (1))
 			fatal_error(ctx,
 				    _("need terminal for interactive repairs"));
--- e2fsprogs-1.27/misc/fsck.c.orig	Tue Mar  5 10:31:15 2002
+++ e2fsprogs-1.27/misc/fsck.c	Mon Apr 15 22:58:33 2002
@@ -959,8 +959,11 @@
 	return status;
 }
 
+__attribute__ ((noreturn))
 static void usage(NOARGS)
 {
+	fprintf (stderr, "fsck %s (%s)\n",
+		 E2FSPROGS_VERSION, E2FSPROGS_DATE);
 	fprintf(stderr,
 		_("Usage: fsck [-ACNPRTV] [-t fstype] [fs-options] filesys\n"));
 	exit(EXIT_USAGE);
--- e2fsprogs-1.27/misc/mke2fs.c.orig	Fri Mar  8 10:02:37 2002
+++ e2fsprogs-1.27/misc/mke2fs.c	Mon Apr 15 22:58:33 2002
@@ -84,8 +84,11 @@
 char *journal_device;
 int sync_kludge;	/* Set using the MKE2FS_SYNC env. option */
 
+__attribute__ ((noreturn))
 static void usage(void)
 {
+	fprintf (stderr, "mke2fs %s (%s)\n",
+		 E2FSPROGS_VERSION, E2FSPROGS_DATE);
 	fprintf(stderr, _("Usage: %s [-c|-t|-l filename] [-b block-size] "
 	"[-f fragment-size]\n\t[-i bytes-per-inode] [-j] [-J journal-options]"
 	" [-N number-of-inodes]\n\t[-m reserved-blocks-percentage] "
@@ -838,8 +841,6 @@
 		param.s_feature_ro_compat = 0;
 	}
 #endif
-	fprintf (stderr, "mke2fs %s (%s)\n",
-		 E2FSPROGS_VERSION, E2FSPROGS_DATE);
 
 	if (argc && *argv) {
 		program_name = get_progname(*argv);
@@ -1002,6 +1003,9 @@
 			break;
 		case 'V':
 			/* Print version number and exit */
+			fprintf (stderr, _("mke2fs %s, %s for EXT2 FS %s, %s\n"),
+				E2FSPROGS_VERSION, E2FSPROGS_DATE,
+				EXT2FS_VERSION, EXT2FS_DATE);
 			fprintf(stderr, _("\tUsing %s\n"),
 				error_message(EXT2_ET_BASE));
 			exit(0);
