diff -puNr e2fsprogs-1.36/e2fsck/pass3.c e2fsprogs-1.36.mode/e2fsck/pass3.c
--- e2fsprogs-1.36/e2fsck/pass3.c	Fri Dec 12 08:00:55 2003
+++ e2fsprogs-1.36.mode/e2fsck/pass3.c	Mon Feb 28 22:29:15 2005
@@ -220,7 +220,7 @@ static void check_root(e2fsck_t ctx)
 	 * Set up the inode structure
 	 */
 	memset(&inode, 0, sizeof(inode));
-	inode.i_mode = 040755;
+	inode.i_mode = 040700;
 	inode.i_size = fs->blocksize;
 	inode.i_atime = inode.i_ctime = inode.i_mtime = time(0);
 	inode.i_links_count = 2;
diff -puNr e2fsprogs-1.36/lib/ext2fs/mkdir.c e2fsprogs-1.36.mode/lib/ext2fs/mkdir.c
--- e2fsprogs-1.36/lib/ext2fs/mkdir.c	Sun Dec  7 17:11:38 2003
+++ e2fsprogs-1.36.mode/lib/ext2fs/mkdir.c	Mon Feb 28 22:29:55 2005
@@ -46,7 +46,7 @@ errcode_t ext2fs_mkdir(ext2_filsys fs, e
 	 * Allocate an inode, if necessary
 	 */
 	if (!ino) {
-		retval = ext2fs_new_inode(fs, parent, LINUX_S_IFDIR | 0755,
+		retval = ext2fs_new_inode(fs, parent, LINUX_S_IFDIR | 0700,
 					  0, &ino);
 		if (retval)
 			goto cleanup;
