--- e2fsprogs-1.27/e2fsck/unix.c.orig	Tue Apr 16 01:14:57 2002
+++ e2fsprogs-1.27/e2fsck/unix.c	Tue Apr 16 01:15:22 2002
@@ -122,7 +122,7 @@
 	 */
 #ifdef ENABLE_NLS
 	printf (_("\n%8d inodes used (%d%%)\n"), inodes_used,
-		(inodes_used != 1), 100 * inodes_used / inodes);
+		100 * inodes_used / inodes);
 	printf (_("%8d non-contiguous inodes (%0d.%d%%)\n"),
 		ctx->fs_fragmented, frag_percent / 10, frag_percent % 10);
 	printf (_("         # of inodes with ind/dind/tind blocks: %d/%d/%d\n"),
