--- e2fsprogs-1.41.9.orig/lib/ss/invocation.c	2009-08-13 01:39:57 +0000
+++ e2fsprogs-1.41.9/lib/ss/invocation.c	2009-10-25 06:28:12 +0000
@@ -71,6 +71,7 @@ int ss_create_invocation(subsystem_name,
 	*(new_table->rqt_tables+1) = (ss_request_table *) NULL;
 
 	new_table->readline_handle = 0;
+	new_table->readline_shutdown = 0;
 	new_table->readline = 0;
 	new_table->add_history = 0;
 	new_table->redisplay = 0;
