 resize/resize2fs.c |    4 ++--
 1 files changed, 2 insertions(+), 2 deletions(-)

diff --git a/resize/resize2fs.c b/resize/resize2fs.c
index 024f1cc..60a4344 100644
--- a/resize/resize2fs.c
+++ b/resize/resize2fs.c
@@ -298,7 +298,7 @@ errcode_t adjust_fs_info(ext2_filsys fs, ext2_filsys old_fs,
 	int		overhead = 0;
 	int		rem;
 	blk_t		blk, group_block;
-	ext2_ino_t	real_end;
+	blk64_t 	real_end;
 	int		adj, old_numblocks, numblocks, adjblocks;
 	unsigned long	i, j, old_desc_blocks, max_group;
 	unsigned int	meta_bg, meta_bg_size;
@@ -381,7 +381,7 @@ retry:
 					    fs->inode_map);
 	if (retval) goto errout;
 
-	real_end = ((EXT2_BLOCKS_PER_GROUP(fs->super)
+	real_end = (((blk64_t)EXT2_BLOCKS_PER_GROUP(fs->super)
 		     * fs->group_desc_count)) - 1 +
 			     fs->super->s_first_data_block;
 	retval = ext2fs_resize_block_bitmap(fs->super->s_blocks_count-1,
