diff -ur sysvinit-2.85.orig/src/init.c sysvinit-2.85/src/init.c
--- sysvinit-2.85.orig/src/init.c	Tue Apr 15 11:16:41 2003
+++ sysvinit-2.85/src/init.c	Tue Apr 22 20:25:58 2003
@@ -420,7 +420,7 @@
 #ifdef __GNUC__
 __attribute__ ((format (printf, 1, 2)))
 #endif
-int setproctitle(char *fmt, ...)
+static int setproctitle(char *fmt, ...)
 {
 	va_list ap;
 	int len;
diff -ur sysvinit-2.85.orig/src/init.h sysvinit-2.85/src/init.h
--- sysvinit-2.85.orig/src/init.h	Thu Jun  3 19:22:59 1999
+++ sysvinit-2.85/src/init.h	Tue Apr 22 20:24:08 2003
@@ -31,6 +31,9 @@
 /* Prototypes. */
 void write_utmp_wtmp(char *user, char *id, int pid, int type, char *line);
 void write_wtmp(char *user, char *id, int pid, int type, char *line);
+#ifdef __GNUC__
+__attribute__ ((format (printf, 2, 3)))
+#endif
 void log(int loglevel, char *fmt, ...);
 void set_term(int how);
 void print(char *fmt);
