diff -urp sysvinit-2.85.orig/src/shutdown.c sysvinit-2.85/src/shutdown.c
--- sysvinit-2.85.orig/src/shutdown.c	Mon Apr 14 11:35:51 2003
+++ sysvinit-2.85/src/shutdown.c	Tue Apr 22 21:45:44 2003
@@ -36,6 +36,7 @@
 #include <fcntl.h>
 #include <stdarg.h>
 #include <utmp.h>
+#include <syslog.h>
 #include "paths.h"
 #include "reboot.h"
 
@@ -295,6 +296,7 @@ void shutdown()
 {
   char *args[8];
   int argp = 0;
+  int do_halt = (down_level[0] == '0');
 
   /* Warn for the last time (hehe) */
   warn(0);
@@ -302,6 +304,12 @@ void shutdown()
 	hardsleep(1);
 	stopit();
   }
+  openlog("shutdown", LOG_PID, LOG_USER);
+  if (do_halt)
+  	syslog(LOG_NOTICE, "shutting down for system halt");
+  else
+	syslog(LOG_NOTICE, "shutting down for system reboot");
+  closelog();
 
   /* See if we have to do it ourself. */
   if (doself) fastdown();
@@ -326,6 +334,9 @@ void shutdown()
   fprintf(stderr, "\rshutdown: cannot execute %s\r\n", INIT);
   unlink(FASTBOOT);
   unlink(FORCEFSCK);
+  openlog("shutdown", LOG_PID, LOG_USER);
+  syslog(LOG_NOTICE, "shutdown failed");
+  closelog();
   exit(1);
 }
 
