--- coreutils-200408100654/src/ls.c.orig	2004-08-03 10:44:08 +0400
+++ coreutils-200408100654/src/ls.c	2004-08-10 19:07:18 +0400
@@ -1760,7 +1760,7 @@ decode_switches (int argc, char **argv)
 
       if (! style)
 	if (! (style = getenv ("TIME_STYLE")))
-	  style = "posix-long-iso";
+	  style = getenv("POSIXLY_CORRECT") ? "posix-long-iso" : "locale";
 
       while (strncmp (style, posix_prefix, sizeof posix_prefix - 1) == 0)
 	{
