--- coreutils-4.5.3/src/date.c.langinfo	2002-11-28 18:02:47 +0000
+++ coreutils-4.5.3/src/date.c	2002-11-28 18:04:04 +0000
@@ -509,14 +509,7 @@
       else
 	{
 	  char *date_fmt = DATE_FMT_LANGINFO ();
-	  /* Do not wrap the following literal format string with _(...).
-	     For example, suppose LC_ALL is unset, LC_TIME="POSIX",
-	     and LANG="ko_KR".	In that case, POSIX says that LC_TIME
-	     determines the format and contents of date and time strings
-	     written by date, which means "date" must generate output
-	     using the POSIX locale; but adding _() would cause "date"
-	     to use a Korean translation of the format.  */
-	  format = *date_fmt ? date_fmt : "%a %b %e %H:%M:%S %Z %Y";
+	  format = *date_fmt ? date_fmt : dcgettext(NULL, N_("%a %b %e %H:%M:%S %Z %Y"), LC_TIME);
 	}
     }
   else if (*format == '\0')
