--- coreutils-5.91.orig/src/dircolors.hin	2005-09-05 14:57:24 +0000
+++ coreutils-5.91/src/dircolors.hin	2005-10-20 15:08:02 +0000
@@ -2,7 +2,18 @@
 # LS_COLORS environment variable used by GNU ls with the --color option.
 
 # The keywords COLOR, OPTIONS, and EIGHTBIT (honored by the
-# slackware version of dircolors) are recognized but ignored.
+# slackware version of dircolors) are recognized but ignored by GNU dircolors.
+
+# This file goes in the /etc directory, and must be world readable.
+# You can copy this file to .dircolors in your $HOME directory to override
+# the system defaults.  This is also may be handled directly by the
+# /etc/profile.d/color_ls.sh and /etc/profile.d/color_ls.csh scripts.
+
+# COLOR needs one of these arguments:
+# 'tty' colorizes output to ttys, but not pipes.
+# 'all' adds color characters to all output.
+# 'none' shuts colorization off.
+COLOR tty
 
 # Below, there should be one TERM entry for each termtype that is colorizable
 TERM linux
@@ -19,11 +30,13 @@ TERM con80x30
 TERM con80x43
 TERM con80x50
 TERM con80x60
+TERM cons25
 TERM cygwin
 TERM dtterm
 TERM putty
 TERM xterm
 TERM xterm-color
+TERM color-xterm
 TERM xterm-debian
 TERM rxvt
 TERM screen
@@ -31,6 +44,7 @@ TERM screen-bce
 TERM screen-w
 TERM vt100
 TERM Eterm
+TERM ansi
 
 # Below are the color init strings for the basic file types. A color init
 # string consists of one or more of the following numeric codes:
@@ -70,12 +84,16 @@ EXEC 01;32
 #.com 01;32
 #.btm 01;32
 #.bat 01;32
+.sh 01;32
+.csh 01;32
 
 .tar 01;31 # archives or compressed (bright red)
 .tgz 01;31
+.arc 01;31
 .arj 01;31
 .taz 01;31
 .lzh 01;31
+.lha 01;31
 .zip 01;31
 .z   01;31
 .Z   01;31
@@ -84,6 +102,7 @@ EXEC 01;32
 .deb 01;31
 .rpm 01;31
 .jar 01;31
+.cpio 01;31
 
 # image formats
 .jpg 01;35
