diff -ur libpcap-0.6.2.orig/bpf/net/bpf_filter.c libpcap-0.6.2/bpf/net/bpf_filter.c
--- libpcap-0.6.2.orig/bpf/net/bpf_filter.c	Mon Oct 23 19:32:21 2000
+++ libpcap-0.6.2/bpf/net/bpf_filter.c	Mon Dec 15 14:29:00 2003
@@ -65,14 +65,16 @@
 #include <stdlib.h>
 #endif
 
+#ifdef HAVE_CONFIG_H
+#include "config.h"
+#endif
+
 #define int32 bpf_int32
 #define u_int32 bpf_u_int32
 
-#ifndef LBL_ALIGN
-#if defined(sparc) || defined(mips) || defined(ibm032) || \
-    defined(__alpha) || defined(__hpux)
+#if !defined(LBL_ALIGN) && \
+	!defined(__i386__) && !defined(__x86_64__) && !defined(__vax__)
 #define LBL_ALIGN
-#endif
 #endif
 
 #ifndef LBL_ALIGN
