--- libpcap-0.9.4/bpf/net/bpf_filter.c.orig	2003-11-15 23:24:07 +0000
+++ libpcap-0.9.4/bpf/net/bpf_filter.c	2005-11-17 11:17:04 +0000
@@ -80,19 +80,10 @@ static const char rcsid[] _U_ =
 #define int32 bpf_int32
 #define u_int32 bpf_u_int32
 
-#ifndef LBL_ALIGN
-/*
- * XXX - IA-64?  If not, this probably won't work on Win64 IA-64
- * systems, unless LBL_ALIGN is defined elsewhere for them.
- * XXX - SuperH?  If not, this probably won't work on WinCE SuperH
- * systems, unless LBL_ALIGN is defined elsewhere for them.
- */
-#if defined(sparc) || defined(__sparc__) || defined(mips) || \
-    defined(ibm032) || defined(__alpha) || defined(__hpux) || \
-    defined(__arm__)
+#if !defined(LBL_ALIGN) && \
+	!defined(__i386__) && !defined(__x86_64__) && !defined(__vax__)
 #define LBL_ALIGN
 #endif
-#endif
 
 #ifndef LBL_ALIGN
 #ifndef WIN32
