--- nmap/output.cc
+++ nmap/output.cc
@@ -566,7 +566,7 @@ void printportoutput(Target *currenths, PortList *plist) {
 		currenths->NameIP(hostname, sizeof(hostname)), 
 		(numignoredports == 1)? "is" : "are");
       if (plist->numIgnoredStates() == 1) {
-	log_write(LOG_PLAIN, statenum2str(plist->nextIgnoredState(PORT_UNKNOWN)));
+	log_write(LOG_PLAIN, "%s", statenum2str(plist->nextIgnoredState(PORT_UNKNOWN)));
       } else {
 	prevstate = PORT_UNKNOWN;
 	while ((istate = plist->nextIgnoredState(prevstate)) != PORT_UNKNOWN) {
--- nmap/scan_engine.cc
+++ nmap/scan_engine.cc
@@ -3597,7 +3597,6 @@ static bool do_one_select_round(UltraScanInfo *USI, struct timeval *stime) {
   int newhoststate = HOST_UNKNOWN;
   int optval;
   recvfrom6_t optlen = sizeof(int);
-  char buf[128];
   int numGoodSD = 0;
   int err = 0;
   u16 pport = 0;
@@ -3788,12 +3787,10 @@ static bool do_one_select_round(UltraScanInfo *USI, struct timeval *stime) {
 	case ENETDOWN:
 	case ENETRESET:
 	case ECONNABORTED:
-	  Snprintf(buf, sizeof(buf), "Strange SO_ERROR from connection to %s (%d - '%s') -- bailing scan", host->target->targetipstr(), optval, strerror(optval) );
-	  pfatal(buf);
+	  fatal("Strange SO_ERROR from connection to %s (%d - '%s') -- bailing scan", host->target->targetipstr(), optval, strerror(optval));
 	  break;
 	default:
-	  Snprintf(buf, sizeof(buf), "Strange read error from %s (%d - '%s')", host->target->targetipstr(), optval, strerror(optval));
-	  perror(buf);
+	  error("Strange read error from %s (%d - '%s')", host->target->targetipstr(), optval, strerror(optval));
 	  break;
 	}
 
