--- nmap-5.00/configure.ac.orig	2009-06-24 21:35:54.000000000 +0000
+++ nmap-5.00/configure.ac	2009-07-16 08:50:20.000000000 +0000
@@ -107,6 +107,7 @@
 needs_cpp_precomp=no
 
 LUA_CFLAGS="-DLUA_USE_POSIX -DLUA_USE_DLOPEN"
+NSE_LDFLAGS=""
 
 case "$host" in
   *alpha-dec-osf*)
@@ -114,15 +115,15 @@
     ;;
   *-netbsd* | *-knetbsd*-gnu)
     AC_DEFINE(NETBSD)
-    LDFLAGS="-Wl,-E $LDFLAGS" # needed for nse-C-module support
+    NSE_LDFLAGS="-Wl,-E"
     ;;
   *-openbsd*)
     AC_DEFINE(OPENBSD)
-    LDFLAGS="-Wl,-E $LDFLAGS" # needed for nse-C-module support
+    NSE_LDFLAGS="-Wl,-E"
     ;;
   *-freebsd* | *-kfreebsd*-gnu | *-dragonfly*)
     AC_DEFINE(FREEBSD)
-    LDFLAGS="-Wl,-E $LDFLAGS" # needed for nse-C-module support
+    NSE_LDFLAGS="-Wl,-E"
     ;;
   *-bsdi*)
     AC_DEFINE(BSDI)
@@ -177,7 +178,7 @@
     ;;
   *-linux*)
     AC_DEFINE(LINUX)
-    LDFLAGS="-Wl,-E $LDFLAGS" # needed for nse-C-module support
+    NSE_LDFLAGS="-Wl,-E"
     ;;
   *-apple-darwin*)
     AC_DEFINE(MACOSX)
@@ -612,6 +613,7 @@
   fi
 
   INSTALLNSE="install-nse"
+  LDFLAGS="$NSE_LDFLAGS $LDFLAGS"
 fi
 
 AC_SUBST(LIBLUA_LIBS)
