diff -ur nmap-3.27.orig/configure nmap-3.27/configure
--- nmap-3.27.orig/configure	Fri Apr 25 04:46:45 2003
+++ nmap-3.27/configure	Mon Jun  2 04:23:46 2003
@@ -1274,13 +1274,6 @@
 
 
 
-if test -d /usr/local/lib; then
-  LDFLAGS="$LDFLAGS -L/usr/local/lib"
-fi
-if test -d /usr/local/include; then
-  CXXFLAGS="$CXXFLAGS -I/usr/local/include"
-fi
-
 libpcapdir=libpcap-possiblymodified
 
 
diff -ur nmap-3.27.orig/configure.ac nmap-3.27/configure.ac
--- nmap-3.27.orig/configure.ac	Fri Apr 25 03:49:02 2003
+++ nmap-3.27/configure.ac	Mon Jun  2 04:22:15 2003
@@ -5,13 +5,6 @@
 dnl Process this file with autoconf to produce a configure script.
 AC_INIT(nmap.cc)
 
-if test -d /usr/local/lib; then
-  LDFLAGS="$LDFLAGS -L/usr/local/lib"
-fi
-if test -d /usr/local/include; then
-  CXXFLAGS="$CXXFLAGS -I/usr/local/include"
-fi
-
 libpcapdir=libpcap-possiblymodified
 AC_SUBST(libpcapdir)
 
diff -ur nmap-3.27.orig/nbase/configure nmap-3.27/nbase/configure
--- nmap-3.27.orig/nbase/configure	Fri Apr 18 04:19:41 2003
+++ nmap-3.27/nbase/configure	Mon Jun  2 04:22:15 2003
@@ -1261,12 +1261,6 @@
 
 
 
-if test -d /usr/local/lib; then
-  LDFLAGS="$LDFLAGS -L/usr/local/lib"
-fi
-if test -d /usr/local/include; then
-  CFLAGS="$CFLAGS -I/usr/local/include"
-fi
 
           ac_config_headers="$ac_config_headers nbase_config.h"
 
diff -ur nmap-3.27.orig/nbase/configure.ac nmap-3.27/nbase/configure.ac
--- nmap-3.27.orig/nbase/configure.ac	Fri Apr 18 04:40:33 2003
+++ nmap-3.27/nbase/configure.ac	Mon Jun  2 04:23:04 2003
@@ -12,13 +12,6 @@
 
 AC_INIT(nbase.h)
 
-if test -d /usr/local/lib; then
-  LDFLAGS="$LDFLAGS -L/usr/local/lib"
-fi
-if test -d /usr/local/include; then
-  CFLAGS="$CFLAGS -I/usr/local/include"
-fi
-
 dnl use config.h instad of -D macros
 AC_CONFIG_HEADER(nbase_config.h)
 
