diff -urp nmap-3.27.orig/nmap_rpc.cc nmap-3.27/nmap_rpc.cc
--- nmap-3.27.orig/nmap_rpc.cc	Fri Apr 25 05:35:39 2003
+++ nmap-3.27/nmap_rpc.cc	Mon Jun  2 04:30:51 2003
@@ -315,7 +315,7 @@ int rpc_are_we_done(char *msg, int msg_l
     return 1;
   }
   if (ntohl(rpc_pack->type_msg) != RPC_MSG_REPLY) {
-    error("Strange -- RPC type is %li shoulb be RPC_MSG_REPLY (1)", ntohl(rpc_pack->type_msg));
+    error("Strange -- RPC type is %lu should be RPC_MSG_REPLY (1)", (unsigned long)ntohl(rpc_pack->type_msg));
     return 0;
   }
   if (ntohl(rpc_pack->auth_flavor) != 0 /* AUTH_NULL */ ||
diff -urp nmap-3.27.orig/services.cc nmap-3.27/services.cc
--- nmap-3.27.orig/services.cc	Tue Apr  8 03:05:24 2003
+++ nmap-3.27/services.cc	Mon Jun  2 04:31:27 2003
@@ -191,7 +191,7 @@ struct scan_lists *getdefaultports(int t
 
   if (!services_initialized)
     if (nmap_services_init() == -1)
-      fatal("Getfastports: Coudn't get port numbers");
+      fatal("Getfastports: Couldn't get port numbers");
   
   bzero(usedports, sizeof(usedports));
   for(bucket = 1; bucket < 1025; bucket++) {  
@@ -253,7 +253,7 @@ struct scan_lists *getfastports(int tcps
 
   if (!services_initialized)
     if (nmap_services_init() == -1)
-      fatal("Getfastports: Coudn't get port numbers");
+      fatal("Getfastports: Couldn't get port numbers");
   
   bzero(usedports, sizeof(usedports));
 
diff -urp nmap-3.27.orig/targets.cc nmap-3.27/targets.cc
--- nmap-3.27.orig/targets.cc	Fri Apr 25 03:14:34 2003
+++ nmap-3.27/targets.cc	Mon Jun  2 04:29:47 2003
@@ -1250,7 +1250,7 @@ int get_ping_results(int sd, pcap_t *pd,
 	  hostnum = sequence / pt->max_tries;
 	  trynum = sequence % pt->max_tries;
 	} else {
-	  // Didn't get it back in either field -- we'll brute forc it ...
+	  // Didn't get it back in either field -- we'll brute force it ...
 	  for(hostnum = pt->group_end; hostnum >= 0; hostnum--) {
 	    if (hostbatch[hostnum]->v4host().s_addr == ip->ip_src.s_addr)
 	      break;
@@ -1328,7 +1328,7 @@ int get_ping_results(int sd, pcap_t *pd,
 	sequence = hostnum * pt->max_tries + trynum;
 
 	if (o.debugging) 
-	  log_write(LOG_STDOUT, "In response to UDP-ping, we got UDP packet back from %s port %hi (hostnum = %d trynum = %d\n", inet_ntoa(ip->ip_src), htons(tcp->th_sport), hostnum, trynum);
+	  log_write(LOG_STDOUT, "In response to UDP-ping, we got UDP packet back from %s port %hi (hostnum = %d trynum = %d\n", inet_ntoa(ip->ip_src), htons(udp->uh_sport), hostnum, trynum);
 	pingstyle = pingstyle_rawudp;
 	foundsomething = 1;
 	dotimeout = 1;
