diff -upk.orig nmap-4.01.orig/nmap.cc nmap-4.01/nmap.cc
--- nmap-4.01.orig/nmap.cc	2006-03-04 00:51:56 +0000
+++ nmap-4.01/nmap.cc	2006-03-04 00:52:53 +0000
@@ -2152,8 +2152,6 @@ int nmap_fetchfile(char *filename_return
   int res;
   int foundsomething = 0;
   struct passwd *pw;
-  char dot_buffer[512];
-  static int warningcount = 0;
 
   /* First we try [--datadir]/file, then $NMAPDIR/file
      next we try ~user/nmap/file
@@ -2225,6 +2223,9 @@ int nmap_fetchfile(char *filename_return
 	foundsomething = 1;
     }
   }
+#ifdef ENABLE_DOT_DIR_OPEN
+  char dot_buffer[512];
+  static int warningcount = 0;
   if (foundsomething && (*filename_returned != '.')) {    
     res = snprintf(dot_buffer, sizeof(dot_buffer), "./%s", file);
     if (res > 0 && res < bufferlen) {
@@ -2246,6 +2247,7 @@ int nmap_fetchfile(char *filename_return
 	foundsomething = 1;
     }
   }
+#endif /* ENABLE_DOT_DIR_OPEN */
 
   if (!foundsomething) {
     filename_returned[0] = '\0';
