--- nmap-4.76/configure.ac.orig	2008-09-08 23:37:03.000000000 +0000
+++ nmap-4.76/configure.ac	2009-05-18 19:21:43.000000000 +0000
@@ -107,6 +107,7 @@
 needs_cpp_precomp=no
 
 LUA_CFLAGS="-DLUA_USE_POSIX -DLUA_USE_DLOPEN"
+NSE_LDFLAGS=""
 
 case "$host" in
   *alpha-dec-osf*)
@@ -114,15 +115,15 @@
     ;;
   *-netbsd* | *-knetbsd*-gnu)
     AC_DEFINE(NETBSD)
-    LDFLAGS="-Wl,-E $LDFLAGS" # needed for nse-C-module support
+    NSE_LDFLAGS="-Wl,-E"
     ;;
   *-openbsd*)
     AC_DEFINE(OPENBSD)
-    LDFLAGS="-Wl,-E $LDFLAGS" # needed for nse-C-module support
+    NSE_LDFLAGS="-Wl,-E"
     ;;
   *-freebsd* | *-kfreebsd*-gnu | *-dragonfly*)
     AC_DEFINE(FREEBSD)
-    LDFLAGS="-Wl,-E $LDFLAGS" # needed for nse-C-module support
+    NSE_LDFLAGS="-Wl,-E"
     ;;
   *-bsdi*)
     AC_DEFINE(BSDI)
@@ -178,7 +179,7 @@
   *-linux*)
     AC_DEFINE(LINUX)
     AC_DEFINE(PCAP_TIMEOUT_IGNORED)  # libpcap doesn't even LOOK at the timeout you give it under Linux
-    LDFLAGS="-Wl,-E $LDFLAGS" # needed for nse-C-module support
+    NSE_LDFLAGS="-Wl,-E"
     ;;
   *-apple-darwin*)
     AC_DEFINE(MACOSX)
@@ -583,6 +584,7 @@
   fi
 
   INSTALLNSE="install-nse"
+  LDFLAGS="$NSE_LDFLAGS $LDFLAGS"
 fi
 
 AC_SUBST(LIBLUA_LIBS)
