--- nmap.cc.orig	2011-01-29 15:58:38.000000000 +0000
+++ nmap.cc	2011-01-29 15:59:23.000000000 +0000
@@ -2684,8 +2684,6 @@ int nmap_fetchfile(char *filename_return
   int res;
   int foundsomething = 0;
   struct passwd *pw;
-  char dot_buffer[512];
-  static int warningcount = 0;
   std::map<std::string, std::string>::iterator iter;
 
   /* First, check the map of requested data file names. If there's an entry for
@@ -2766,6 +2764,9 @@ int nmap_fetchfile(char *filename_return
     }
   }
 
+#ifdef ENABLE_DOT_DIR_OPEN
+  char dot_buffer[512];
+  static int warningcount = 0;
   if (foundsomething && (*filename_returned != '.')) {    
     res = Snprintf(dot_buffer, sizeof(dot_buffer), "./%s", file);
     if (res > 0 && res < bufferlen) {
@@ -2786,6 +2787,7 @@ int nmap_fetchfile(char *filename_return
       foundsomething = fileexistsandisreadable(filename_returned);
     }
   }
+#endif
 
   if (!foundsomething) {
     filename_returned[0] = '\0';
