--- nmap-4.76/nmap.cc.orig	2008-09-10 18:32:35.000000000 +0000
+++ nmap-4.76/nmap.cc	2009-05-18 13:29:46.000000000 +0000
@@ -2710,8 +2710,6 @@
   int res;
   int foundsomething = 0;
   struct passwd *pw;
-  char dot_buffer[512];
-  static int warningcount = 0;
   std::map<std::string, std::string>::iterator iter;
 
   /* First, check the map of requested data file names. If there's an entry for
@@ -2803,6 +2801,9 @@
     }
   }
 
+#ifdef ENABLE_DOT_DIR_OPEN
+  char dot_buffer[512];
+  static int warningcount = 0;
   if (foundsomething && (*filename_returned != '.')) {    
     res = Snprintf(dot_buffer, sizeof(dot_buffer), "./%s", file);
     if (res > 0 && res < bufferlen) {
@@ -2823,6 +2824,7 @@
       foundsomething = fileexistsandisreadable(filename_returned);
     }
   }
+#endif /* ENABLE_DOT_DIR_OPEN */
 
   if (!foundsomething) {
     filename_returned[0] = '\0';
