--- nmap-5.00/nmap.cc.orig	2009-07-17 14:09:10.000000000 +0000
+++ nmap-5.00/nmap.cc	2009-07-17 14:16:59.000000000 +0000
@@ -1682,7 +1682,8 @@
   }
   if (o.servicescan)
     o.scriptversion = 1;
-  open_nse();
+  if(o.script)
+    open_nse();
 #endif
   
   /* Time to create a hostgroup state object filled with all the requested
@@ -2018,8 +2019,10 @@
   if (o.extra_payload) free(o.extra_payload);
   if (o.ipoptions) free(o.ipoptions);
 #ifndef NOLUA
-  close_nse();
-  free(o.scriptargs);
+  if(o.script) {
+    close_nse();
+    free(o.scriptargs);
+  }
 #endif
 }
 
