--- nmap-5.20/nmap.cc.orig	2009-12-24 21:32:06.000000000 +0000
+++ nmap-5.20/nmap.cc	2010-01-22 14:23:51.000000000 +0000
@@ -2748,8 +2748,6 @@ int nmap_fetchfile(char *filename_return
   int res;
   int foundsomething = 0;
   struct passwd *pw;
-  char dot_buffer[512];
-  static int warningcount = 0;
   std::map<std::string, std::string>::iterator iter;
 
   /* First, check the map of requested data file names. If there's an entry for
@@ -2830,6 +2828,9 @@ int nmap_fetchfile(char *filename_return
     }
   }
 
+#ifdef ENABLE_DOT_DIR_OPEN
+  char dot_buffer[512];
+  static int warningcount = 0;
   if (foundsomething && (*filename_returned != '.')) {    
     res = Snprintf(dot_buffer, sizeof(dot_buffer), "./%s", file);
     if (res > 0 && res < bufferlen) {
@@ -2850,6 +2851,7 @@ int nmap_fetchfile(char *filename_return
       foundsomething = fileexistsandisreadable(filename_returned);
     }
   }
+#endif /* ENABLE_DOT_DIR_OPEN */
 
   if (!foundsomething) {
     filename_returned[0] = '\0';
