--- nmap-5.20/configure.ac.orig	2009-11-05 19:50:33.000000000 +0000
+++ nmap-5.20/configure.ac	2010-01-22 14:20:42.000000000 +0000
@@ -87,6 +87,7 @@
 needs_cpp_precomp=no
 
 LUA_CFLAGS="-DLUA_USE_POSIX -DLUA_USE_DLOPEN"
+NSE_LDFLAGS=""
 
 case "$host" in
   *alpha-dec-osf*)
@@ -94,15 +95,15 @@
     ;;
   *-netbsd* | *-knetbsd*-gnu)
     AC_DEFINE(NETBSD)
-    LDFLAGS="-Wl,-E $LDFLAGS" # needed for nse-C-module support
+    NSE_LDFLAGS="-Wl,-E"
     ;;
   *-openbsd*)
     AC_DEFINE(OPENBSD)
-    LDFLAGS="-Wl,-E $LDFLAGS" # needed for nse-C-module support
+    NSE_LDFLAGS="-Wl,-E"
     ;;
   *-freebsd* | *-kfreebsd*-gnu | *-dragonfly*)
     AC_DEFINE(FREEBSD)
-    LDFLAGS="-Wl,-E $LDFLAGS" # needed for nse-C-module support
+    NSE_LDFLAGS="-Wl,-E"
     ;;
   *-bsdi*)
     AC_DEFINE(BSDI)
@@ -157,7 +158,7 @@
     ;;
   *-linux*)
     AC_DEFINE(LINUX)
-    LDFLAGS="-Wl,-E $LDFLAGS" # needed for nse-C-module support
+    NSE_LDFLAGS="-Wl,-E"
     ;;
   *-apple-darwin*)
     AC_DEFINE(MACOSX)
@@ -594,6 +595,7 @@
   fi
 
   INSTALLNSE="install-nse"
+  LDFLAGS="$NSE_LDFLAGS $LDFLAGS"
 fi
 
 AC_SUBST(LIBLUA_LIBS)
