diff -uNrp nmap-5.50.orig/configure.ac nmap-5.50/configure.ac
--- nmap-5.50.orig/configure.ac	2011-01-29 11:22:53.000000000 +0000
+++ nmap-5.50/configure.ac	2011-01-29 11:23:57.000000000 +0000
@@ -8,6 +8,27 @@ dnl Give us the --enable-nls option, to 
 dnl installed. This gets replaced with AM_GNU_GETTEXT for full gettext support.
 AM_NLS
 
+AH_TOP([
+#ifndef CONFIG_H
+#define CONFIG_H
+])
+
+AH_BOTTOM([
+#ifdef NEED_USLEEP_PROTO
+#ifdef __cplusplus
+extern "C" int usleep (unsigned int);
+#endif
+#endif
+
+#ifdef NEED_GETHOSTNAME_PROTO
+#ifdef __cplusplus
+extern "C" int gethostname (char *, unsigned int);
+#endif
+#endif
+
+#endif /* CONFIG_H */
+])
+
 AC_ARG_WITH(localdirs,
 AC_HELP_STRING([--with-localdirs], [Explicitly ask compiler to use /usr/local/{include,libs} if they exist ]),
   [ case "$with_localdirs" in
@@ -57,6 +78,7 @@ if test -n "$GXX"; then
 fi
 
 # Remember that all following tests will run with this CXXFLAGS by default
+AH_TEMPLATE([__func__], [])
 AC_MSG_CHECKING(for __func__)
 AC_TRY_COMPILE([
 #include <stdio.h>
@@ -93,76 +115,77 @@ NSE_LDFLAGS=""
 case "$host" in
   *alpha-dec-osf*)
     AC_DEFINE(DEC)
+    AC_DEFINE(DEC, 1, [Define to 1 for DEC])
     ;;
   *-netbsd* | *-knetbsd*-gnu)
-    AC_DEFINE(NETBSD)
+    AC_DEFINE(NETBSD, 1, [Define to 1 for NetBSD])
     NSE_LDFLAGS="-Wl,-E"
     ;;
   *-openbsd*)
-    AC_DEFINE(OPENBSD)
+    AC_DEFINE(OPENBSD, 1, [Define to 1 for OpenBSD])
     NSE_LDFLAGS="-Wl,-E"
     ;;
   *-freebsd* | *-kfreebsd*-gnu | *-dragonfly*)
-    AC_DEFINE(FREEBSD)
+    AC_DEFINE(FREEBSD, 1, [Define to 1 for FreeBSD])
     NSE_LDFLAGS="-Wl,-E"
     ;;
   *-bsdi*)
-    AC_DEFINE(BSDI)
+    AC_DEFINE(BSDI, 1, [Define to 1 for BSDI])
     ;;
   *-sgi-irix5* | *-sgi-irix6*)
-    AC_DEFINE(IRIX)
+    AC_DEFINE(IRIX, 1, [Define to 1 for IRIX])
     ;;
   *-hpux*)
-    AC_DEFINE(HPUX)
+    AC_DEFINE(HPUX, 1, [Define to 1 for HPUX])
     # To link with libnet and NM (/usr/lib/libnm.sl) library
     # on HP-UX 11.11 (other versions?) Mikhail Zakharov (zmey20000@yahoo.com)
     AC_CHECK_LIB(nm, open_mib)
     ;;
   *-aix*)
     # use some AIX specific libraries
-    AC_CHECK_LIB(odm, odm_initialize)	
+    AC_CHECK_LIB(odm, odm_initialize)
     AC_CHECK_LIB(cfg, _system_configuration)
     ;;
-  *-solaris2.0*)  
-    AC_DEFINE(STUPID_SOLARIS_CHECKSUM_BUG)
-    AC_DEFINE(SOLARIS)
+  *-solaris2.0*)
+    AC_DEFINE(STUPID_SOLARIS_CHECKSUM_BUG, 1, [Define to 1 for Solaris checksum bug])
+    AC_DEFINE(SOLARIS, 1, [Define to 1 for Solaris])
     ;;
   *-solaris2.[[1-9]][[0-9]]*)
-    AC_DEFINE(SOLARIS)
+    AC_DEFINE(SOLARIS, 1, [Define to 1 for Solaris])
     ;;
   *-solaris2.1*)
-    AC_DEFINE(STUPID_SOLARIS_CHECKSUM_BUG)
-    AC_DEFINE(SOLARIS)
+    AC_DEFINE(STUPID_SOLARIS_CHECKSUM_BUG, 1, [Define to 1 for Solaris checksum bug])
+    AC_DEFINE(SOLARIS, 1, [Define to 1 for Solaris])
     ;;
   *-solaris2.2*)
-    AC_DEFINE(STUPID_SOLARIS_CHECKSUM_BUG)
-    AC_DEFINE(SOLARIS)
+    AC_DEFINE(STUPID_SOLARIS_CHECKSUM_BUG, 1, [Define to 1 for Solaris checksum bug])
+    AC_DEFINE(SOLARIS, 1, [Define to 1 for Solaris])
     ;;
   *-solaris2.3*)
-    AC_DEFINE(STUPID_SOLARIS_CHECKSUM_BUG)
-    AC_DEFINE(SOLARIS)
+    AC_DEFINE(STUPID_SOLARIS_CHECKSUM_BUG, 1, [Define to 1 for Solaris checksum bug])
+    AC_DEFINE(SOLARIS, 1, [Define to 1 for Solaris])
     ;;
   *-solaris2.4*)
-    AC_DEFINE(STUPID_SOLARIS_CHECKSUM_BUG)
-    AC_DEFINE(SOLARIS)
+    AC_DEFINE(STUPID_SOLARIS_CHECKSUM_BUG, 1, [Define to 1 for Solaris checksum bug])
+    AC_DEFINE(SOLARIS, 1, [Define to 1 for Solaris])
     ;;
   *-solaris2.5.1)
-    AC_DEFINE(STUPID_SOLARIS_CHECKSUM_BUG)
-    AC_DEFINE(SOLARIS)
+    AC_DEFINE(STUPID_SOLARIS_CHECKSUM_BUG, 1, [Define to 1 for Solaris checksum bug])
+    AC_DEFINE(SOLARIS, 1, [Define to 1 for Solaris])
     ;;
   *-solaris*)
-    AC_DEFINE(SOLARIS)
+    AC_DEFINE(SOLARIS, 1, [Define to 1 for Solaris])
     ;;
   *-sunos4*)
-    AC_DEFINE(SUNOS)
-    AC_DEFINE(SPRINTF_RETURNS_STRING)
+    AC_DEFINE(SUNOS, 1, [Define to 1 for SunOS])
+    AC_DEFINE(SPRINTF_RETURNS_STRING, 1, [Define to 1 if sprintf returns string])
     ;;
   *-linux*)
-    AC_DEFINE(LINUX)
+    AC_DEFINE(LINUX, 1, [Define to 1 for Linux])
     NSE_LDFLAGS="-Wl,-E"
     ;;
   *-apple-darwin*)
-    AC_DEFINE(MACOSX)
+    AC_DEFINE(MACOSX, 1, [Define to 1 for MacOSX])
     dnl on Mac OSX the math library seems to contain unwanted getopt cruft
     AC_CHECK_LIB(m, main)
     LUA_CFLAGS="-DLUA_USE_MACOSX"
@@ -342,7 +365,7 @@ fi
 
 OPENSSL_LIBS=
 if test "$use_openssl" = "yes"; then
-  AC_DEFINE(HAVE_OPENSSL)
+  AC_DEFINE(HAVE_OPENSSL, 1, [Define to 1 for OpenSSL support])
   OPENSSL_LIBS="-lssl -lcrypto"
 fi
 
@@ -418,7 +441,7 @@ if test $have_libpcap != yes; then
   PCAP_CLEAN="pcap_clean"
   PCAP_DIST_CLEAN="pcap_dist_clean"
 else
-  AC_DEFINE(HAVE_LIBPCAP)
+  AC_DEFINE(HAVE_LIBPCAP, 1, [Define to 1 for libpcap support])
   LIBPCAP_LIBS="-lpcap"
   PCAP_DEPENDS=""
   PCAP_BUILD=""
@@ -652,7 +675,7 @@ AC_CACHE_CHECK(if sockaddr{} has sa_len 
         ac_cv_sockaddr_has_sa_len=yes,
         ac_cv_sockaddr_has_sa_len=no))
 if test $ac_cv_sockaddr_has_sa_len = yes ; then
-        AC_DEFINE(HAVE_SOCKADDR_SA_LEN)
+        AC_DEFINE(HAVE_SOCKADDR_SA_LEN, 1, [Define to 1 for sa_len member in sockaddr])
 fi
 
 #dnl check endedness
@@ -661,11 +684,11 @@ AC_C_BIGENDIAN
 AC_MSG_CHECKING([if struct in_addr is a wacky huge structure (some Sun boxes)])
 
 AC_TRY_COMPILE([#include <netinet/in.h>], struct in_addr i; i._S_un._S_addr;, \
-             AC_DEFINE(IN_ADDR_DEEPSTRUCT) \
+             AC_DEFINE(IN_ADDR_DEEPSTRUCT, 1, [Define to 1 for huge struct in_addr]) \
              AC_MSG_RESULT(yes) , \
              AC_TRY_COMPILE([#include <sys/types.h>
 #include <netinet/in.h>], struct in_addr i; i.S_un.S_addr;, \
-                             AC_DEFINE(IN_ADDR_DEEPSTRUCT) \
+                             AC_DEFINE(IN_ADDR_DEEPSTRUCT, 1, [Define to 1 for huge struct in_addr]) \
                              AC_MSG_RESULT(yes) , \
                              AC_MSG_RESULT(no);))
 
@@ -684,7 +707,7 @@ AC_CACHE_CHECK(if struct icmp exists, ac
         ac_cv_struct_icmp_exists=yes,
         ac_cv_struct_icmp_exists=no))
 if test $ac_cv_struct_icmp_exists = yes ; then
-        AC_DEFINE(HAVE_STRUCT_ICMP)
+        AC_DEFINE(HAVE_STRUCT_ICMP, 1, [Define to 1 for struct icmp])
 fi
 
 AC_CACHE_CHECK(if struct ip exists, ac_cv_struct_ip_exists,
@@ -701,7 +724,7 @@ AC_CACHE_CHECK(if struct ip exists, ac_c
         ac_cv_struct_ip_exists=yes,
         ac_cv_struct_ip_exists=no))
 if test $ac_cv_struct_ip_exists = yes ; then
-        AC_DEFINE(HAVE_STRUCT_IP)
+        AC_DEFINE(HAVE_STRUCT_IP, 1, [Define to 1 for struct ip])
 fi
 
 AC_CACHE_CHECK(if struct ip has ip_sum member, ac_cv_ip_has_ip_sum,
@@ -719,7 +742,7 @@ AC_CACHE_CHECK(if struct ip has ip_sum m
         ac_cv_ip_has_ip_sum=yes,
         ac_cv_ip_has_ip_sum=no))
 if test $ac_cv_ip_has_ip_sum = yes ; then
-        AC_DEFINE(HAVE_IP_IP_SUM)
+        AC_DEFINE(HAVE_IP_IP_SUM, 1, [Define to 1 for ip_sum member])
 fi
 
 dnl Checks for library functions.
diff -uNrp nmap-5.50.orig/configure.ac.orig nmap-5.50/configure.ac.orig
--- nmap-5.50.orig/configure.ac.orig	2011-01-29 11:22:53.000000000 +0000
+++ nmap-5.50/configure.ac.orig	2011-01-29 11:22:40.000000000 +0000
@@ -88,6 +88,7 @@ AC_PATH_TOOL([STRIP], [strip], [/bin/tru
 needs_cpp_precomp=no
 
 LUA_CFLAGS="-DLUA_USE_POSIX -DLUA_USE_DLOPEN"
+NSE_LDFLAGS=""
 
 case "$host" in
   *alpha-dec-osf*)
@@ -95,15 +96,15 @@ case "$host" in
     ;;
   *-netbsd* | *-knetbsd*-gnu)
     AC_DEFINE(NETBSD)
-    LDFLAGS="-Wl,-E $LDFLAGS" # needed for nse-C-module support
+    NSE_LDFLAGS="-Wl,-E"
     ;;
   *-openbsd*)
     AC_DEFINE(OPENBSD)
-    LDFLAGS="-Wl,-E $LDFLAGS" # needed for nse-C-module support
+    NSE_LDFLAGS="-Wl,-E"
     ;;
   *-freebsd* | *-kfreebsd*-gnu | *-dragonfly*)
     AC_DEFINE(FREEBSD)
-    LDFLAGS="-Wl,-E $LDFLAGS" # needed for nse-C-module support
+    NSE_LDFLAGS="-Wl,-E"
     ;;
   *-bsdi*)
     AC_DEFINE(BSDI)
@@ -158,7 +159,7 @@ case "$host" in
     ;;
   *-linux*)
     AC_DEFINE(LINUX)
-    LDFLAGS="-Wl,-E $LDFLAGS" # needed for nse-C-module support
+    NSE_LDFLAGS="-Wl,-E"
     ;;
   *-apple-darwin*)
     AC_DEFINE(MACOSX)
@@ -626,6 +627,7 @@ else 
   fi
 
   INSTALLNSE="install-nse"
+  LDFLAGS="$NSE_LDFLAGS $LDFLAGS"
 fi
 
 AC_SUBST(LIBLUA_LIBS)
