--- nmap-4.76/tcpip.cc.orig	2008-09-04 14:41:59.000000000 +0000
+++ nmap-4.76/tcpip.cc	2009-05-18 19:23:27.000000000 +0000
@@ -3068,8 +3068,8 @@
     ifaces = getinterfaces(&numifaces);
     /* First let us try Linux-style /proc/net/route */
     routefp = fopen("/proc/net/route", "r");
-    if (routefp) {
-      (void) fgets(buf, sizeof(buf), routefp); /* Kill the first line (column headers) */
+    /* Kill the first line (column headers) */
+    if (routefp && fgets(buf, sizeof(buf), routefp)) {
       while(fgets(buf,sizeof(buf), routefp)) {
 	p = strtok(buf, " \t\n");
 	if (!p) {
