diff -ur nmap-3.27.orig/HACKING nmap-3.27/HACKING
--- nmap-3.27.orig/HACKING	Sun Oct 15 07:16:21 2000
+++ nmap-3.27/HACKING	Mon Jun  2 04:38:08 2003
@@ -90,10 +90,10 @@
     mv nmap-2.54BETA4 nmap-2.54BETA4-snazzy-feature
 
 3. Unpack the original Nmap source alongside it:
-    tar xzf /tmp/nmap-2.54BETA4.tgz
+    tar xzf nmap-2.54BETA4.tgz
 
 4. Generate the diffs:
-   diff -urNb nmap-2.54BETA4 nmap-2.54BETA4-snazzy-feature > /tmp/nmap.patch
+   diff -urNb nmap-2.54BETA4 nmap-2.54BETA4-snazzy-feature > nmap.patch
 
 5. Check the patch and remove any unnecessary patches from the file.
 
@@ -105,9 +105,9 @@
 
   cp file.c file.c.orig
   [Make changes to file.c ...]
-  diff -u file.c.orig file.c > /tmp/file.c.patch
+  diff -u file.c.orig file.c > file.c.patch
 
-and just send us the patch: /tmp/file.c.patch.
+and just send us the patch: file.c.patch.
 
 
 Credits
diff -ur nmap-3.27.orig/Makefile.in nmap-3.27/Makefile.in
--- nmap-3.27.orig/Makefile.in	Mon Apr 28 18:34:48 2003
+++ nmap-3.27/Makefile.in	Mon Jun  2 04:42:28 2003
@@ -74,6 +74,7 @@
 #	@echo "END OF SECTION WHERE FAILURES ARE OK"
 
 # This target is just for me ... don't you get any ideas --Fyodor
+# This is unsafe on shared systems, should use mktemp
 distro: 
 	autoconf
 	rm -f config.cache
diff -ur nmap-3.27.orig/docs/nmap_lithuanian.1 nmap-3.27/docs/nmap_lithuanian.1
--- nmap-3.27.orig/docs/nmap_lithuanian.1	Tue Mar 27 03:07:57 2001
+++ nmap-3.27/docs/nmap_lithuanian.1	Mon Jun  2 04:37:28 2003
@@ -381,10 +381,10 @@
 suranda visus *.company.com hostus ir atiduoda juos nmap'ui,
 kuris savo ruoþtu ásijungæs verbose mode visus juos nuskanuoja.
 .Sp
-.B nmap -sN -D microsoft.com,mail.takas.lt,ME -oN /tmp/crazy -p 1-1024 -O crazy.com
+.B nmap -sN -D microsoft.com,mail.takas.lt,ME -oN /root/crazy -p 1-1024 -O crazy.com
 .Sp
 skanauoja Null skanavimo reþimu, panaudoja du decoy adresus,
-viskà logina á /tmp/crazy failà, skanuoja nuo 1 iki 1024 crazy.com
+viskà logina á /root/crazy failà, skanuoja nuo 1 iki 1024 crazy.com
 portus bei stengiasi atspëti crazy.com serverio operacinæ sistemà
 .SH BUGAI
 Vabalëliai? Kokie dar vabalëliai? Na.. jei rasit kokiø, bûtinai
diff -ur nmap-3.27.orig/docs/nmap_manpage-lt.html nmap-3.27/docs/nmap_manpage-lt.html
--- nmap-3.27.orig/docs/nmap_manpage-lt.html	Mon Apr 28 19:35:32 2003
+++ nmap-3.27/docs/nmap_manpage-lt.html	Mon Jun  2 04:37:05 2003
@@ -408,11 +408,11 @@
        nmap'ui,  kuris  savo  ruoþtu ásijungæs verbose mode visus
        juos nuskanuoja.
 
-       <B>nmap</B> <B>-sN</B> <B>-D</B> <B>microsoft.com,mail.takas.lt,ME</B> <B>-oN</B>  <B>/tmp/crazy</B>
+       <B>nmap</B> <B>-sN</B> <B>-D</B> <B>microsoft.com,mail.takas.lt,ME</B> <B>-oN</B>  <B>/root/crazy</B>
        <B>-p</B> <B>1-1024</B> <B>-O</B> <B>crazy.com</B>
 
        skanauoja   Null  skanavimo  reþimu,  panaudoja  du  decoy
-       adresus, viskà logina á /tmp/crazy failà, skanuoja  nuo  1
+       adresus, viskà logina á /root/crazy failà, skanuoja  nuo  1
        iki  1024 crazy.com portus bei stengiasi atspëti crazy.com
        serverio operacinæ sistemà
 
diff -ur nmap-3.27.orig/nmap.cc nmap-3.27/nmap.cc
--- nmap-3.27.orig/nmap.cc	Mon Apr 28 18:37:14 2003
+++ nmap-3.27/nmap.cc	Mon Jun  2 04:42:47 2003
@@ -1320,7 +1320,7 @@
 h             -- Prints this help screen.\n\
 Examples:\n\
 n -sS -O -v example.com/24\n\
-f --spoof \"/usr/local/bin/pico -z hello.c\" -sS -oN /tmp/e.log example.com/24\n\n");
+f --spoof \"/usr/local/bin/pico -z hello.c\" -sS -oN e.log example.com/24\n\n");
 }
 
 char *seqreport(struct seq_info *seq) {
diff -ur nmap-3.27.orig/shtool nmap-3.27/shtool
--- nmap-3.27.orig/shtool	Wed Dec 22 05:04:11 1999
+++ nmap-3.27/shtool	Mon Jun  2 04:43:16 2003
@@ -1721,7 +1721,7 @@
     
     #   create the version file according the the selected language  
     echo "new version: $version.$revision$bptype$bplevel ($date)"
-    tmpfile="/tmp/version.tmp.$$"
+    tmpfile="version.tmp.$$"
     rm -f $tmpfile >/dev/null 2>&1
     case $LANGUAGE in
         txt )
@@ -1815,7 +1815,7 @@
     basename=no
     pathlist="$PATH"
     namelist=""
-    tmpfile="/tmp/path.tmp.$$"
+    tmpfile="path.tmp.$$"
     while [ ".$1" != . ]; do
         case $1 in
             -s ) silent=yes   ; shift; continue ;;
