--- nmap-4.62/configure.ac.orig	2008-05-18 09:22:24 +0000
+++ nmap-4.62/configure.ac	2008-05-18 09:22:42 +0000
@@ -108,6 +108,7 @@
 
 needs_cpp_precomp=no
 LUAFLAGS="MYCFLAGS=\"$CFLAGS -DLUA_USE_POSIX -DLUA_USE_DLOPEN\""
+NSE_LDFLAGS=""
 
 case "$host" in
   *alpha-dec-osf*)
@@ -115,15 +116,15 @@
     ;;
   *-netbsd* | *-knetbsd*-gnu)
     AC_DEFINE(NETBSD)
-    LDFLAGS="-Wl,-E $LDFLAGS" # needed for nse-C-module support
+    NSE_LDFLAGS="-Wl,-E"
     ;;
   *-openbsd*)
     AC_DEFINE(OPENBSD)
-    LDFLAGS="-Wl,-E $LDFLAGS" # needed for nse-C-module support
+    NSE_LDFLAGS="-Wl,-E"
     ;;
   *-freebsd* | *-kfreebsd*-gnu | *-dragonfly*)
     AC_DEFINE(FREEBSD)
-    LDFLAGS="-Wl,-E $LDFLAGS" # needed for nse-C-module support
+    NSE_LDFLAGS="-Wl,-E"
     ;;
   *-bsdi*)
     AC_DEFINE(BSDI)
@@ -152,7 +153,7 @@
   *-linux*)
     AC_DEFINE(LINUX)
     AC_DEFINE(PCAP_TIMEOUT_IGNORED)  # libpcap doesn't even LOOK at the timeout you give it under Linux
-    LDFLAGS="-Wl,-E $LDFLAGS" # needed for nse-C-module support
+    NSE_LDFLAGS="-Wl,-E"
     ;;
   *-apple-darwin*)
     AC_DEFINE(MACOSX)
@@ -546,6 +547,7 @@
   fi
 
   INSTALLNSE="install-nse"
+  LDFLAGS="$NSE_LDFLAGS $LDFLAGS"
 fi
 
 if test $no_lua != yes; then
