diff -urpN patchutils-0.2.30.orig/src/rediff.c patchutils-0.2.30/src/rediff.c
--- patchutils-0.2.30.orig/src/rediff.c	Mon Jun  7 14:07:46 2004
+++ patchutils-0.2.30/src/rediff.c	Sun Nov 28 17:01:35 2004
@@ -1039,7 +1039,7 @@ int main (int argc, char *argv[])
 		syntax (1);
 
 	if (argc - optind == 1) {
-		char *p = xmalloc (PATH_MAX);
+		char *p = xmalloc (strlen (argv[0]) + strlen ("recountdiff") + 1);
 		char *f;
 		char **const new_argv = xmalloc (sizeof (char *) * argc);
 		memcpy (new_argv, argv, sizeof (char *) * argc);
