diff -ur shadow-4.0.0.orig/libmisc/chkname.c shadow-4.0.0/libmisc/chkname.c
--- shadow-4.0.0.orig/libmisc/chkname.c	Thu Apr 16 23:57:43 1998
+++ shadow-4.0.0/libmisc/chkname.c	Fri Nov  2 02:22:13 2001
@@ -22,19 +22,16 @@
 good_name(const char *name)
 {
 	/*
-	 * User/group names must start with a letter, and may not
-	 * contain colons, commas, newlines (used in passwd/group
-	 * files...) or any non-printable characters.
+	 * User/group names must match [a-z_][a-z0-9_-]*
 	 */
-	if (!*name || !isalpha(*name))
+	if (!*name || !((*name >= 'a' && *name <= 'z') || *name == '_'))
 		return 0;
 
-	while (*name) {
-		if (*name == ':' || *name == ',' ||
-		    *name == '\n' || !isprint(*name))
+	while (*++name) {
+		if (!((*name >= 'a' && *name <= 'z') ||
+		    (*name >= '0' && *name <= '9') ||
+		    *name == '_' || *name == '-'))
 			return 0;
-
-		name++;
 	}
 
 	return 1;
@@ -43,6 +40,7 @@
 int
 check_user_name(const char *name)
 {
+#if 0
 #if HAVE_UTMPX_H
 	struct utmpx ut;
 #else
@@ -54,6 +52,9 @@
 	 * handle (usually max 8 characters).
 	 */
 	if (strlen(name) > sizeof(ut.ut_user))
+#else
+	if (strlen(name) > 8)
+#endif
 		return 0;
 
 	return good_name(name);
@@ -63,10 +64,9 @@
 check_group_name(const char *name)
 {
 	/*
-	 * Arbitrary limit for group names - max 16
-	 * characters (same as on HP-UX 10).
+	 * Arbitrary limit for group names - max 8 characters.
 	 */
-	if (strlen(name) > 16)
+	if (strlen(name) > 8)
 		return 0;
 
 	return good_name(name);
