diff -uNrp shadow-4.0.4.1/src/chage.c shadow-4.0.4.1.chage-ro-no-lock/src/chage.c
--- shadow-4.0.4.1/src/chage.c	2003-06-19 17:57:15 +0000
+++ shadow-4.0.4.1.chage-ro-no-lock/src/chage.c	2004-02-28 23:00:36 +0000
@@ -505,8 +505,7 @@ int main (int argc, char **argv)
 	 * file entries into memory. Then we get a pointer to the password
 	 * file entry for the requested user.
 	 */
-	pwrw = 0;
-	if (!pw_open (pwrw ? O_RDWR : O_RDONLY)) {
+	if (!pw_open (O_RDONLY)) {
 		fprintf (stderr, _("%s: can't open password file\n"),
 			 Prog);
 		cleanup (1);
