--- shadow-19990827/src/chage.c.orig	Sat Feb 10 01:16:09 2001
+++ shadow-19990827/src/chage.c	Sat Feb 10 01:33:12 2001
@@ -38,6 +38,7 @@
 #include <signal.h>
 #include <ctype.h>
 #include <time.h>
+#include <errno.h>
 #include "prototypes.h"
 #include "defines.h"
 
@@ -488,7 +489,7 @@
 	 * to the password file entry for the requested user.
 	 */
 
-	if (!pw_lock()) {
+	if (!lflg && !pw_lock()) {
 		fprintf(stderr, _("%s: can't lock password file\n"), Prog);
 		SYSLOG((LOG_ERR, LOCK_FAIL, PASSWD_FILE));
 		closelog();
@@ -520,7 +521,7 @@
 	 * not exist already.
 	 */
 
-	if (!spw_lock()) {
+	if (!lflg && !spw_lock()) {
 		fprintf(stderr, _("%s: can't lock shadow password file\n"), Prog);
 		cleanup(1);
 		SYSLOG((LOG_ERR, LOCK_FAIL, SHADOW_FILE));
@@ -533,6 +534,14 @@
 		SYSLOG((LOG_ERR, OPEN_FAIL, SHADOW_FILE));
 		closelog();
 		exit(1);
+	}
+
+	if (lflg) {
+		if (setgid(getgid()) || setuid(ruid)) {
+			fprintf(stderr, "%s: failed to drop privileges (%s)",
+				Prog, strerror(errno));
+			exit(1);
+		}
 	}
 
 	sp = spw_locate(argv[optind]);
