diff -puNr shadow-4.0.4.1.orig/configure.in shadow-4.0.4.1/configure.in
--- shadow-4.0.4.1.orig/configure.in	2004-01-14 14:11:25 +0000
+++ shadow-4.0.4.1/configure.in	2014-06-19 22:01:14 +0000
@@ -1,7 +1,8 @@
 dnl Process this file with autoconf to produce a configure script.
-AC_INIT
-AM_INIT_AUTOMAKE(shadow, 4.0.4.1)
+AC_INIT([shadow], [4.0.4.1])
+AC_CONFIG_SRCDIR([src/su.c])
 AC_CONFIG_HEADERS([config.h])
+AM_INIT_AUTOMAKE
 
 dnl Some hacks...
 test "$prefix" = "NONE" && prefix="/usr"
@@ -17,7 +18,7 @@ AC_PROG_CC
 AC_ISC_POSIX
 AC_PROG_LN_S
 AC_PROG_YACC
-AM_C_PROTOTYPES
+AC_C_PROTOTYPES
 AM_PROG_LIBTOOL
 
 dnl Checks for libraries.
