diff -ur shadow-4.0.0.orig/src/newgrp.c shadow-4.0.0/src/newgrp.c
--- shadow-4.0.0.orig/src/newgrp.c	Sat Sep  1 08:19:16 2001
+++ shadow-4.0.0/src/newgrp.c	Sun Nov 11 10:05:33 2001
@@ -40,8 +40,6 @@
 #include "prototypes.h"
 #include "defines.h"
 
-#include "getdef.h"
-
 extern char **environ;
 
 #ifdef HAVE_SETGROUPS
@@ -65,7 +63,7 @@
 	if (is_newgrp)
 		fprintf (stderr, _("usage: newgrp [ - ] [ group ]\n"));
 	else
-		fprintf (stderr, _("usage: sg group [[-c] command ]\n"));
+		fprintf (stderr, _("usage: sg group [[-c] command]\n"));
 }
 
 /*
@@ -315,20 +313,11 @@
 		cpasswd = pw_encrypt (cp, grp->gr_passwd);
 		strzero(cp);
 
-		if (grp->gr_passwd[0] == '\0') {
-		/*
-		 * there is no password, print out "Sorry" and give up
-		 */
-			sleep(1);
-			fputs (_("Sorry.\n"), stderr);
-			goto failure;
-		}
-
-		if (strcmp (cpasswd, grp->gr_passwd) != 0) {
+		if (grp->gr_passwd[0] == '\0' ||
+		    strcmp (cpasswd, grp->gr_passwd) != 0) {
 			SYSLOG((LOG_INFO,
 				"Invalid password for group `%s' from `%s'\n",
 				group, name));
-			sleep(1);
 			fputs (_("Sorry.\n"), stderr);
 			goto failure;
 		}
@@ -341,9 +330,8 @@
 	 */
 
 #ifdef	USE_SYSLOG
-	if (getdef_bool ("SYSLOG_SG_ENAB"))
-		SYSLOG((LOG_INFO, "user `%s' switched to group `%s'\n",
-			name, group));
+	SYSLOG((LOG_INFO, "user `%s' switched to group `%s'\n",
+		name, group));
 #endif
 	gid = grp->gr_gid;
 
