diff -ur shadow-4.0.0.orig/libmisc/fields.c shadow-4.0.0/libmisc/fields.c
--- shadow-4.0.0.orig/libmisc/fields.c	Mon Dec  8 02:27:04 1997
+++ shadow-4.0.0/libmisc/fields.c	Fri Nov  2 01:51:55 2001
@@ -37,6 +37,11 @@
 #include <stdio.h>
 #include "prototypes.h"
 
+#define fixed_iscntrl(c) \
+	(((c) & 0x7f) < 0x20 || (c) == 0x7f)
+
+#define MAX_FIELD_SIZE			0x80
+
 /*
  * valid_field - insure that a field contains all legal characters
  *
@@ -50,13 +55,14 @@
 {
 	const char *cp;
 
-	for (cp = field; *cp && isprint(*cp & 0x7F) && !strchr(illegal, *cp); cp++)
-		;
+	for (cp = field; *cp; cp++)
+	if (iscntrl(*cp) || fixed_iscntrl(*cp) || strchr(illegal, *cp))
+		return -1;
 
-	if (*cp)
+	if (cp - field > MAX_FIELD_SIZE)
 		return -1;
-	else
-		return 0;
+
+	return 0;
 }
 
 /*
diff -ur shadow-4.0.0.orig/src/chage.c shadow-4.0.0/src/chage.c
--- shadow-4.0.0.orig/src/chage.c	Mon Oct  9 22:39:50 2000
+++ shadow-4.0.0/src/chage.c	Fri Nov  2 01:51:55 2001
@@ -410,9 +410,11 @@
 #endif
 
 	sanitize_env();
-	setlocale(LC_ALL, "");
-	bindtextdomain(PACKAGE, LOCALEDIR);
-	textdomain(PACKAGE);
+	if (getuid() == 0) {
+		setlocale(LC_ALL, "");
+		bindtextdomain(PACKAGE, LOCALEDIR);
+		textdomain(PACKAGE);
+	}
 
 	ruid = getuid();
 	amroot = (ruid == 0);
@@ -898,9 +900,12 @@
 
 	Prog = Basename(argv[0]);
 
-	setlocale(LC_ALL, "");
-	bindtextdomain(PACKAGE, LOCALEDIR);
-	textdomain(PACKAGE);
+	sanitize_env();
+	if (getuid() == 0) {
+		setlocale(LC_ALL, "");
+		bindtextdomain(PACKAGE, LOCALEDIR);
+		textdomain(PACKAGE);
+	}
 
 	fprintf (stderr, _("%s: no aging information present\n"), Prog);
 	exit(1);
diff -ur shadow-4.0.0.orig/src/chfn.c shadow-4.0.0/src/chfn.c
--- shadow-4.0.0.orig/src/chfn.c	Sat Sep  2 22:40:43 2000
+++ shadow-4.0.0/src/chfn.c	Fri Nov  2 01:51:55 2001
@@ -259,9 +259,11 @@
 	char *user;
 
 	sanitize_env();
-	setlocale(LC_ALL, "");
-	bindtextdomain(PACKAGE, LOCALEDIR);
-	textdomain(PACKAGE);
+	if (getuid() == 0) {
+		setlocale(LC_ALL, "");
+		bindtextdomain(PACKAGE, LOCALEDIR);
+		textdomain(PACKAGE);
+	}
 
 	/*
 	 * This command behaves different for root and non-root
diff -ur shadow-4.0.0.orig/src/chsh.c shadow-4.0.0/src/chsh.c
--- shadow-4.0.0.orig/src/chsh.c	Sat Sep  2 22:40:43 2000
+++ shadow-4.0.0/src/chsh.c	Fri Nov  2 01:51:55 2001
@@ -162,10 +162,11 @@
 	struct	passwd	pwent;		/* New password entry                */
 
 	sanitize_env();
-
-	setlocale(LC_ALL, "");
-	bindtextdomain(PACKAGE, LOCALEDIR);
-	textdomain(PACKAGE);
+	if (getuid() == 0) {
+		setlocale(LC_ALL, "");
+		bindtextdomain(PACKAGE, LOCALEDIR);
+		textdomain(PACKAGE);
+	}
 
 	/*
 	 * This command behaves different for root and non-root
diff -ur shadow-4.0.0.orig/src/expiry.c shadow-4.0.0/src/expiry.c
--- shadow-4.0.0.orig/src/expiry.c	Sat Aug 26 22:27:18 2000
+++ shadow-4.0.0/src/expiry.c	Fri Nov  2 02:01:04 2001
@@ -103,6 +103,11 @@
 	char *Prog = argv[0];
 
 	sanitize_env();
+	if (getuid() == 0) {
+		setlocale(LC_ALL, "");
+		bindtextdomain(PACKAGE, LOCALEDIR);
+		textdomain(PACKAGE);
+	}
 
 	/* 
 	 * Start by disabling all of the keyboard signals.
@@ -119,10 +124,6 @@
 	 * expiry takes one of two arguments.  The default action
 	 * is to give the usage message.
 	 */
-
-	setlocale(LC_ALL, "");
-	bindtextdomain(PACKAGE, LOCALEDIR);
-	textdomain(PACKAGE);
 
 	if (argc != 2 || (strcmp (argv[1], "-f")  && strcmp (argv[1], "-c")))
 		usage ();
diff -ur shadow-4.0.0.orig/src/gpasswd.c shadow-4.0.0/src/gpasswd.c
--- shadow-4.0.0.orig/src/gpasswd.c	Sat Sep  2 22:40:43 2000
+++ shadow-4.0.0/src/gpasswd.c	Fri Nov  2 01:51:55 2001
@@ -208,9 +208,11 @@
 	char *members = NULL;
 
 	sanitize_env();
-	setlocale(LC_ALL, "");
-	bindtextdomain(PACKAGE, LOCALEDIR);
-	textdomain(PACKAGE);
+	if (getuid() == 0) {
+		setlocale(LC_ALL, "");
+		bindtextdomain(PACKAGE, LOCALEDIR);
+		textdomain(PACKAGE);
+	}
 
 	/*
 	 * Make a note of whether or not this command was invoked
diff -ur shadow-4.0.0.orig/src/login.c shadow-4.0.0/src/login.c
--- shadow-4.0.0.orig/src/login.c	Fri Jun 29 00:47:06 2001
+++ shadow-4.0.0/src/login.c	Fri Nov  2 01:51:55 2001
@@ -467,10 +467,11 @@
 	 */
 
 	sanitize_env();
-
-	setlocale(LC_ALL, "");
-	bindtextdomain(PACKAGE, LOCALEDIR);
-	textdomain(PACKAGE);
+	if (getuid() == 0) {
+		setlocale(LC_ALL, "");
+		bindtextdomain(PACKAGE, LOCALEDIR);
+		textdomain(PACKAGE);
+	}
 
 	initenv();
 
diff -ur shadow-4.0.0.orig/src/newgrp.c shadow-4.0.0/src/newgrp.c
--- shadow-4.0.0.orig/src/newgrp.c	Sat Sep  1 08:19:16 2001
+++ shadow-4.0.0/src/newgrp.c	Fri Nov  2 01:51:55 2001
@@ -100,10 +100,12 @@
 	sanitize_env();
 #endif
 
-	setlocale(LC_ALL, "");
-	bindtextdomain(PACKAGE, LOCALEDIR);
-	textdomain(PACKAGE);
-                  
+	if (getuid()==0) {
+		setlocale(LC_ALL, "");
+		bindtextdomain(PACKAGE, LOCALEDIR);
+		textdomain(PACKAGE);
+	}
+	           
 	/*
 	 * save my name for error messages and save my real gid incase
 	 * of errors.  if there is an error i have to exec a new login
diff -ur shadow-4.0.0.orig/src/passwd.c shadow-4.0.0/src/passwd.c
--- shadow-4.0.0.orig/src/passwd.c	Sat Sep  2 22:40:44 2000
+++ shadow-4.0.0/src/passwd.c	Fri Nov  2 01:51:55 2001
@@ -1020,9 +1020,12 @@
 #endif
 #endif
 
-	setlocale(LC_ALL, "");
-	bindtextdomain(PACKAGE, LOCALEDIR);
-	textdomain(PACKAGE);
+	sanitize_env();
+	if (getuid() == 0) {
+		setlocale(LC_ALL, "");
+		bindtextdomain(PACKAGE, LOCALEDIR);
+		textdomain(PACKAGE);
+	}
 
 	/*
 	 * The program behaves differently when executed by root
@@ -1035,8 +1038,6 @@
 	 * prefix to most error messages.
 	 */
 	Prog = Basename(argv[0]);
-
-	sanitize_env();
 
 	OPENLOG("passwd");
 
diff -ur shadow-4.0.0.orig/src/su.c shadow-4.0.0/src/su.c
--- shadow-4.0.0.orig/src/su.c	Fri Aug 24 03:09:31 2001
+++ shadow-4.0.0/src/su.c	Fri Nov  2 01:51:55 2001
@@ -181,10 +181,11 @@
 #endif /* !USE_PAM */
 
 	sanitize_env();
-
-	setlocale(LC_ALL, "");
-	bindtextdomain(PACKAGE, LOCALEDIR);
-	textdomain(PACKAGE);
+	if (getuid() == 0) {
+		setlocale(LC_ALL, "");
+		bindtextdomain(PACKAGE, LOCALEDIR);
+		textdomain(PACKAGE);
+	}
 
 	/*
 	 * Get the program name.  The program name is used as a
