--- shadow-4.0.0/src/usermod.c.orig	Mon Jul 28 23:39:45 2003
+++ shadow-4.0.0/src/usermod.c	Mon Jul 28 23:40:49 2003
@@ -1658,6 +1658,7 @@
 
 	sys_ngroups = sysconf(_SC_NGROUPS_MAX);
 	user_groups = malloc((1 + sys_ngroups) * sizeof(char *));
+	user_groups[0] = (char *) 0;
 
 	OPENLOG("usermod");
 
--- shadow-4.0.0/src/useradd.c.orig	Mon Jul 28 23:39:45 2003
+++ shadow-4.0.0/src/useradd.c	Mon Jul 28 23:41:08 2003
@@ -1930,6 +1930,7 @@
 
 	sys_ngroups = sysconf(_SC_NGROUPS_MAX);
 	user_groups = malloc((1 + sys_ngroups) * sizeof(char *));
+	user_groups[0] = (char *) 0;
 
 #ifdef SHADOWPWD
 	is_shadow_pwd = spw_file_present();
