diff -ur shadow-4.0.0.orig/src/usermod.c shadow-4.0.0/src/usermod.c
--- shadow-4.0.0.orig/src/usermod.c	Sat Sep  1 08:19:16 2001
+++ shadow-4.0.0/src/usermod.c	Sun Nov 11 04:55:45 2001
@@ -346,7 +346,7 @@
 		strcpy(buf, "!");
 		strcat(buf, pw_pass);
 		pw_pass = buf;
-	} else if (Uflg && pw_pass[0] == '!') {
+	} else if (Uflg && pw_pass[0] == '!' && pw_pass[1] != '!') {
 		char *s;
 
 		SYSLOG((LOG_INFO, "unlock user `%s' password\n",
