--- iproute2/etc/iproute2/rt_dsfield.orig	Sun Apr 16 20:42:48 2000
+++ iproute2/etc/iproute2/rt_dsfield	Thu Dec 27 15:40:31 2001
@@ -1,13 +1,15 @@
-0x10	lowdelay
-0x08	throughput
-0x04	reliability
-# This value overlap with ECT, do not use it!
-0x02	mincost
-# These values seems do not want to die, Cisco likes them by a strange reason.
-0x20	priority
-0x40	immediate
-0x60	flash
-0x80	flash-override
-0xa0	critical
-0xc0	internet
-0xe0	network
+#0x10	lowdelay
+#0x08	throughput
+#0x04	reliability
+
+# This value overlaps with ECT, do not use it!
+#0x02	mincost
+
+# These values seems to not want to die, Cisco likes them for a strange reason.
+#0x20	priority
+#0x40	immediate
+#0x60	flash
+#0x80	flash-override
+#0xa0	critical
+#0xc0	internet
+#0xe0	network
--- iproute2/etc/iproute2/rt_protos.orig	Sun Apr 16 20:42:48 2000
+++ iproute2/etc/iproute2/rt_protos	Thu Dec 27 15:36:10 2001
@@ -1,25 +1,26 @@
 #
 # Reserved protocols.
 #
-0	unspec
-1	redirect
-2	kernel
-3	boot
-4	static
-8	gated
-9	ra
-10	mrt
-11	zebra
-12	bird
+#0	unspec
+#1	redirect
+#2	kernel
+#3	boot
+#4	static
+#8	gated
+#9	ra
+#10	mrt
+#11	zebra
+#12	bird
+
 #
 #	Used by me for gated
 #
-254	gated/aggr
-253	gated/bgp
-252	gated/ospf
-251	gated/ospfase
-250	gated/rip
-249	gated/static
-248	gated/conn
-247	gated/inet
-246	gated/default
+#254	gated/aggr
+#253	gated/bgp
+#252	gated/ospf
+#251	gated/ospfase
+#250	gated/rip
+#249	gated/static
+#248	gated/conn
+#247	gated/inet
+#246	gated/default
--- iproute2/etc/iproute2/rt_realms.orig	Sun Apr 16 20:42:48 2000
+++ iproute2/etc/iproute2/rt_realms	Thu Dec 27 15:36:10 2001
@@ -1,13 +1,14 @@
 #
 # reserved values
 #
-0	cosmos
+#0	cosmos
+
 #
 # local
 #
-1	inr.ac
-2	inr.ruhep
-3	freenet
-4	radio-msu
-5	russia
-6	internet
+#1	inr.ac
+#2	inr.ruhep
+#3	freenet
+#4	radio-msu
+#5	russia
+#6	internet
--- iproute2/etc/iproute2/rt_scopes.orig	Sun Apr 16 20:42:48 2000
+++ iproute2/etc/iproute2/rt_scopes	Thu Dec 27 15:36:10 2001
@@ -1,11 +1,12 @@
 #
 # reserved values
 #
-0	global
-255	nowhere
-254	host
-253	link
+#0	global
+#255	nowhere
+#254	host
+#253	link
+
 #
 # pseudo-reserved
 #
-200	site
+#200	site
--- iproute2/etc/iproute2/rt_tables.orig	Sun Apr 16 20:42:48 2000
+++ iproute2/etc/iproute2/rt_tables	Thu Dec 27 15:36:10 2001
@@ -1,11 +1,12 @@
 #
 # reserved values
 #
-255	local
-254	main
-253	default
-0	unspec
+#255	local
+#254	main
+#253	default
+#0	unspec
+
 #
 # local
 #
-1	inr.ruhep
+#1	inr.ruhep
