--- ltrace-0.3.36/Makefile.in.orig	2004-11-10 02:12:03 +0300
+++ ltrace-0.3.36/Makefile.in	2005-01-08 07:02:07 +0300
@@ -20,10 +20,10 @@
 LIBS		=	@LIBS@
 
 INSTALL		=	@INSTALL@
-INSTALL_FILE	=	$(INSTALL) -p    -o root -g root  -m  644
-INSTALL_PROGRAM	=	$(INSTALL) -p    -o root -g root  -m  755
-INSTALL_SCRIPT	=	$(INSTALL) -p    -o root -g root  -m  755
-INSTALL_DIR	=	$(INSTALL) -p -d -o root -g root  -m  755
+INSTALL_FILE	=	$(INSTALL) -p    -m  644
+INSTALL_PROGRAM	=	$(INSTALL) -p    -m  755
+INSTALL_SCRIPT	=	$(INSTALL) -p    -m  755
+INSTALL_DIR	=	$(INSTALL) -p -d -m  755
 
 OBJ	=	ltrace.o options.o elf.o output.o read_config_file.o	\
 		execute_program.o wait_for_something.o process_event.o	\
