diff -ur ltrace-0.3.10.orig/Makefile.in ltrace-0.3.10/Makefile.in
--- ltrace-0.3.10.orig/Makefile.in	Mon Aug 30 00:49:23 1999
+++ ltrace-0.3.10/Makefile.in	Sun Jan 14 00:03:51 2001
@@ -17,6 +17,8 @@
 		execute_program.o wait_for_something.o process_event.o	\
 		display_args.o breakpoints.o proc.o demangle.o
 
+mandir = /usr/man
+
 all:		ltrace
 
 ltrace:		sysdeps/sysdep.o $(OBJ)
@@ -40,12 +42,11 @@
 		( cd .. ; tar zcvf ltrace-`date +%y%m%d`.tgz ltrace )
 
 install:	ltrace
-		@INSTALL@ -d $(DESTDIR)/usr/bin $(DESTDIR)/usr/doc/ltrace $(DESTDIR)/usr/share/man/man1
+		@INSTALL@ -d $(DESTDIR)/usr/bin $(DESTDIR)/$(mandir)/man1
 		@INSTALL@ -d $(DESTDIR)/etc
 		@INSTALL@ -s ltrace $(DESTDIR)/usr/bin
 		@INSTALL@ -m 644 etc/ltrace.conf $(DESTDIR)/etc
-		@INSTALL@ -m 644 COPYING README TODO BUGS ChangeLog $(DESTDIR)/usr/doc/ltrace
-		@INSTALL@ -m 644 ltrace.1 $(DESTDIR)/usr/share/man/man1
+		@INSTALL@ -m 644 ltrace.1 $(DESTDIR)/$(mandir)/man1
 
 dummy:
 
