--- procps-2.0.6/proc/sysinfo.c.msw	Thu Feb  3 23:22:22 2000
+++ procps-2.0.6/proc/sysinfo.c	Thu Feb  3 23:32:44 2000
@@ -275,6 +275,6 @@
 /**************************************************************************
  * shorthand for read_table("/proc/meminfo")[meminfo_main][meminfo_total] */
 unsigned read_total_main(void) {
-    unsigned** mem;
+    unsigned long long ** mem;
     return (mem = meminfo()) ? mem[meminfo_main][meminfo_total] : -1;
 }
