diff -ur procps-2.0.7.orig/free.c procps-2.0.7/free.c
--- procps-2.0.7.orig/free.c	Tue Nov  2 18:44:58 1999
+++ procps-2.0.7/free.c	Sat Dec  2 05:50:31 2000
@@ -20,7 +20,7 @@
     char *titles[6], name[32];
     int i, n = 0, col[6] = {0}, rtime = 0, old_fmt = 0, first_line;
     int old_meminfo = 0;
-    long long **mem;
+    unsigned long long **mem;
 
     static int sum[6]; /* statics get initialized to zero */
 
diff -ur procps-2.0.7.orig/proc/sysinfo.c procps-2.0.7/proc/sysinfo.c
--- procps-2.0.7.orig/proc/sysinfo.c	Mon Jul 10 23:36:13 2000
+++ procps-2.0.7/proc/sysinfo.c	Sat Dec  2 05:47:36 2000
@@ -217,7 +234,7 @@
     	for (i=0; i < MAX_ROW && *p; i++) {                /* loop over rows */
 		while(*p && !isdigit(*p)) p++;          /* skip chars until a digit */
 		for (j=0; j < MAX_COL && *p; j++) {     /* scanf column-by-column */
-		    l = sscanf(p, "%u%n", row[i] + j, &k);
+		    l = sscanf(p, "%Lu%n", row[i] + j, &k);
 		    p += k;                             /* step over used buffer */
 		    if (*p == '\n' || l < 1)            /* end of line/buffer */
 			break;
