--- procps-3.2.5/vmstat.c.parg	2005-06-28 14:57:15.819982176 +0200
+++ procps-3.2.5/vmstat.c	2005-06-28 14:57:38.801488456 +0200
@@ -608,7 +608,7 @@
         statMode |= PARTITIONSTAT;
 	if (argv[1]){
 	           ++argv;
-		   sprintf(partition, "%s", *argv);
+		   snprintf(partition, sizeof(partition), "%s", *argv);
 	 }else{fprintf(stderr, "-p requires an argument\n");
                exit(EXIT_FAILURE);
 	}
