diff -uNp -r procps-3.2.5.old/ps/common.h procps-3.2.5/ps/common.h
--- procps-3.2.5.old/ps/common.h	2004-10-09 03:55:50 +0000
+++ procps-3.2.5/ps/common.h	2012-06-28 15:12:31 +0000
@@ -14,7 +14,7 @@
 
 #include "../proc/procps.h"
 #include "../proc/readproc.h"
-#include <asm/page.h>  /* looks safe for glibc, we need PAGE_SIZE */
+#include <sys/user.h>  /* looks safe for glibc, we need PAGE_SIZE */
 
 #if 0
 #define trace(args...) printf(## args)
