--- libtool-1.5.18/libtool.m4.orig	2005-05-15 13:41:23 +0000
+++ libtool-1.5.18/libtool.m4	2005-05-26 11:22:33 +0000
@@ -4108,7 +4108,7 @@ available_tags=
 # ### BEGIN LIBTOOL CONFIG],
 [# ### BEGIN LIBTOOL TAG CONFIG: $tagname])
 
-# Libtool was configured on host `(hostname || uname -n) 2>/dev/null | sed 1q`:
+# Libtool was configured as follows:
 
 # Shell to use when invoking shell scripts.
 SHELL=$lt_SHELL
