diff -puNr libtool-2.4.2.orig/libtoolize.in libtool-2.4.2/libtoolize.in
--- libtool-2.4.2.orig/libtoolize.in	2011-10-17 10:19:37 +0000
+++ libtool-2.4.2/libtoolize.in	2014-06-15 11:14:03 +0000
@@ -758,31 +758,20 @@ func_mkdir_p ()
 # given, STRING is the basename for that directory.
 func_mktempdir ()
 {
-    my_template="${TMPDIR-/tmp}/${1-$progname}"
+    my_template="${1-$progname}"
 
     if test "$opt_dry_run" = ":"; then
       # Return a directory name, but don't create it in dry-run mode
-      my_tmpdir="${my_template}-$$"
+      my_mktemp='mktemp -dtu'
     else
-
-      # If mktemp works, use that first and foremost
-      my_tmpdir=`mktemp -d "${my_template}-XXXXXXXX" 2>/dev/null`
-
-      if test ! -d "$my_tmpdir"; then
-        # Failing that, at least try and use $RANDOM to avoid a race
-        my_tmpdir="${my_template}-${RANDOM-0}$$"
-
-        save_mktempdir_umask=`umask`
-        umask 0077
-        $MKDIR "$my_tmpdir"
-        umask $save_mktempdir_umask
-      fi
-
-      # If we're not in dry-run mode, bomb out on failure
-      test -d "$my_tmpdir" || \
-        func_fatal_error "cannot create temporary directory \`$my_tmpdir'"
+      my_mktemp='mktemp -dt'
     fi
 
+    my_tmpdir=`$my_mktemp "${my_template}-XXXXXXXX"` || {
+      $echo "cannot create temporary directory" 1>&2
+      exit $EXIT_FAILURE
+    }
+
     $ECHO "$my_tmpdir"
 }
 
