--- libtool-1.5.18/ltmain.in.orig	2005-05-04 19:13:00 +0000
+++ libtool-1.5.18/ltmain.in	2005-05-26 11:07:26 +0000
@@ -5767,16 +5767,8 @@ relink_command=\"$relink_command\""
 	  outputname=
 	  if test "$fast_install" = no && test -n "$relink_command"; then
 	    if test "$finalize" = yes && test -z "$run"; then
-	      tmpdir="/tmp"
-	      test -n "$TMPDIR" && tmpdir="$TMPDIR"
-	      tmpdir="$tmpdir/libtool-$$"
-	      save_umask=`umask`
-	      umask 0077
-	      if $mkdir "$tmpdir"; then
-	        umask $save_umask
-	      else
-	        umask $save_umask
-		$echo "$modename: error: cannot create temporary directory \`$tmpdir'" 1>&2
+	      if ! tmpdir=`mktemp -dt libtool.XXXXXXXXXX`; then
+		$echo "$modename: error: cannot create temporary directory" 1>&2
 		continue
 	      fi
 	      file=`$echo "X$file$stripped_ext" | $Xsed -e 's%^.*/%%'`
