diff -uNp -r libtool-1.5.22.old/tagdemo/foo.cpp libtool-1.5.22/tagdemo/foo.cpp
--- libtool-1.5.22.old/tagdemo/foo.cpp	2005-04-23 05:39:20 +0000
+++ libtool-1.5.22/tagdemo/foo.cpp	2011-10-11 10:13:43.799680651 +0000
@@ -20,7 +20,9 @@
 // USA.
 
 #include "foo.h"
-#include <iostream.h>
+#include <iostream>
+
+using namespace std;
 
 #ifdef HAVE_MATH_H
 #include <math.h>
diff -uNp -r libtool-1.5.22.old/tagdemo/main.cpp libtool-1.5.22/tagdemo/main.cpp
--- libtool-1.5.22.old/tagdemo/main.cpp	2005-04-23 05:39:20 +0000
+++ libtool-1.5.22/tagdemo/main.cpp	2011-10-11 10:13:50.219690052 +0000
@@ -22,8 +22,9 @@
 
 #include "foo.h"
 #include "baz.h"
-#include <iostream.h>
+#include <iostream>
 
+using namespace std;
 
 int
 main (int, char *[])
