diff -puNr glibc-2.3.3-200406160000.orig/misc/syslog.c glibc-2.3.3-200406160000/misc/syslog.c
--- glibc-2.3.3-200406160000.orig/misc/syslog.c	Thu Apr 22 08:55:01 2004
+++ glibc-2.3.3-200406160000/misc/syslog.c	Sat Jan  8 04:18:16 2005
@@ -66,6 +66,7 @@ static int	LogFile = -1;		/* fd for log 
 static int	connected;		/* have done connect */
 static int	LogStat;		/* status bits, set by openlog() */
 static const char *LogTag;		/* string to tag the entry with */
+static char	*LogTagDynamic;		/* same as LogTag if malloc()'ed */
 static int	LogFacility = LOG_USER;	/* default facility code */
 static int	LogMask = 0xff;		/* mask of priorities to be logged */
 extern char	*__progname;		/* Program name, from crt0. */
@@ -124,6 +125,51 @@ syslog(pri, fmt, va_alist)
 }
 libc_hidden_def (syslog)
 
+static void
+init_syslog_ident(void)
+{
+	char *ident, *safe_progname, *p;
+	uid_t uid, euid;
+	gid_t gid, egid;
+
+	LogTagDynamic = NULL;
+
+	if (!__libc_enable_secure) {
+		LogTag = __progname;
+		return;
+	}
+
+	safe_progname = strdup(__progname);
+	if (!safe_progname) {
+		LogTag = "NO MEMORY";
+		return;
+	}
+	for (p = safe_progname; *p; p++)
+		if ((*p & 0x7f) < 0x20 || *p == 0x7f || *p == '"')
+			*p = '?';
+
+	uid = getuid();
+	euid = geteuid();
+	gid = getgid();
+	egid = getegid();
+	if (uid != euid || gid == egid)
+		asprintf(&ident,
+		    "UNSPECIFIED (__progname=\"%s\" uid=%u euid=%u)",
+		    safe_progname, uid, euid);
+	else
+		asprintf(&ident,
+		    "UNSPECIFIED (__progname=\"%s\" uid=%u gid=%u egid=%u)",
+		    safe_progname, uid, gid, egid);
+	free(safe_progname);
+	if (!ident) {
+		LogTag = "NO MEMORY";
+		return;
+	}
+
+	LogTag = ident;
+	LogTagDynamic = ident;
+}
+
 void
 vsyslog(pri, fmt, ap)
 	int pri;
@@ -198,8 +244,17 @@ vsyslog(pri, fmt, ap)
 				   "%h %e %T ", __localtime_r (&now, &now_tm));
 #endif
 	    msgoff = ftell (f);
+
+	    /* Protect against multiple users and cancellation.  */
+	    __libc_cleanup_push (cancel_handler, NULL);
+	    __libc_lock_lock (syslog_lock);
+
 	    if (LogTag == NULL)
-	      LogTag = __progname;
+	      init_syslog_ident();
+
+	    /* Free the lock.  */
+	    __libc_cleanup_pop (1);
+
 	    if (LogTag != NULL)
 	      fputs_unlocked (LogTag, f);
 	    if (LogStat & LOG_PID)
@@ -263,7 +318,7 @@ vsyslog(pri, fmt, ap)
 
 	/* Get connected, output the message to the local logger. */
 	if (!connected)
-		openlog_internal(LogTag, LogStat | LOG_NDELAY, 0);
+		openlog_internal(NULL, LogStat | LOG_NDELAY, 0);
 
 	/* If we have a SOCK_STREAM connection, also send ASCII NUL as
 	   a record terminator.  */
@@ -317,8 +372,12 @@ static void
 internal_function
 openlog_internal(const char *ident, int logstat, int logfac)
 {
-	if (ident != NULL)
+	if (ident != NULL) {
+		if (LogTagDynamic)
+			free(LogTagDynamic);
+		LogTagDynamic = NULL;
 		LogTag = ident;
+	}
 	LogStat = logstat;
 	if (logfac != 0 && (logfac &~ LOG_FACMASK) == 0)
 		LogFacility = logfac;
@@ -397,6 +456,9 @@ closelog ()
   __libc_lock_lock (syslog_lock);
 
   closelog_internal ();
+  if (LogTagDynamic)
+    free(LogTagDynamic);
+  LogTagDynamic = NULL;
   LogTag = NULL;
   LogType = SOCK_DGRAM; /* this is the default */
 
