diff -ur glibc-2.1.3.orig/stdlib/stdlib.h glibc-2.1.3/stdlib/stdlib.h
--- glibc-2.1.3.orig/stdlib/stdlib.h	Thu Feb 24 19:19:03 2000
+++ glibc-2.1.3/stdlib/stdlib.h	Sat Jun 28 15:02:29 2003
@@ -580,18 +580,15 @@
 
 #ifdef	__USE_GNU
 /* Return a malloc'd string containing the canonical absolute name of the
-   named file.  The last file name component need not exist, and may be a
-   symlink to a nonexistent file.  */
+   named file.  */
 extern char *canonicalize_file_name __P ((__const char *__name));
 #endif
 
 #if defined __USE_BSD || defined __USE_XOPEN_EXTENDED
-/* Return the canonical absolute name of file NAME.  The last file name
-   component need not exist, and may be a symlink to a nonexistent file.
-   If RESOLVED is null, the result is malloc'd; otherwise, if the canonical
-   name is PATH_MAX chars or more, returns null with `errno' set to
-   ENAMETOOLONG; if the name fits in fewer than PATH_MAX chars, returns the
-   name in RESOLVED.  */
+/* Return the canonical absolute name of file NAME.  If the canonical name
+   is PATH_MAX chars or more, returns NULL with `errno' set to
+   ENAMETOOLONG; if the name fits in fewer than PATH_MAX chars, returns
+   the name in RESOLVED.  */
 extern char *realpath __P ((__const char *__restrict __name,
 			    char *__restrict __resolved));
 #endif
