--- glibc-2.3.3-200404190147.orig/linuxthreads/man/pthread_create.man	1998-03-11 15:42:23 +0300
+++ glibc-2.3.3-200404190147/linuxthreads/man/pthread_create.man	2004-04-21 16:13:09 +0400
@@ -31,10 +31,8 @@
 .SH ERRORS
 .TP
 !EAGAIN!
-not enough system resources to create a process for the new thread.
-.TP
-!EAGAIN!
-more than !PTHREAD_THREADS_MAX! threads are already active.
+not enough system resources to create a process for the new thread,
+or more than !PTHREAD_THREADS_MAX! threads are already active.
 
 .SH AUTHOR
 Xavier Leroy <Xavier.Leroy@inria.fr>
