diff -uNrp glibc-2.3.2.orig/glibcbug.in glibc-2.3.2/glibcbug.in
--- glibc-2.3.2.orig/glibcbug.in	2004-03-08 08:18:25 +0000
+++ glibc-2.3.2/glibcbug.in	2004-03-08 08:25:21 +0000
@@ -34,8 +34,6 @@ trap 'rm -f "$TEMP" "$TEMPx"' EXIT
 BUGGLIBC="glibc-bug-reports-${RELEASE}@gnu.org"
 BUGADDR=${1-$BUGGLIBC}
 
-: ${EDITOR=emacs}
-
 : ${USER=${LOGNAME-`whoami`}}
 
 # How to read the passwd database.
@@ -81,14 +79,21 @@ else
 fi
 
 # If they don't have a preferred editor set, then use
-if [ -z "$VISUAL" ]; then
-  if [ -z "$EDITOR" ]; then
+if [ -z "$EDITOR" ]; then
+  if [ -n "$VISUAL" ]; then
+    EDIT="$VISUAL"
+  elif [ -x /bin/vitmp ]; then
+    EDIT=vitmp
+  elif [ -x /bin/vi ]; then
     EDIT=vi
+  elif [ -x /usr/bin/vim ]; then
+    EDIT=vim
   else
-    EDIT="$EDITOR"
+    echo "$0: No default editor found: attempting to use vi" >&2
+    EDIT=vi
   fi
 else
-  EDIT="$VISUAL"
+  EDIT="$EDITOR"
 fi
 
 # Find out some information.
