diff -ur glibc-2.1.3.orig/resolv/res_send.c glibc-2.1.3/resolv/res_send.c
--- glibc-2.1.3.orig/resolv/res_send.c	Sat Aug  7 07:16:16 1999
+++ glibc-2.1.3/resolv/res_send.c	Tue Oct  1 02:49:43 2002
@@ -457,7 +457,9 @@
 				len = anssiz;
 			} else
 				len = resplen;
-			if (len < HFIXEDSZ) {
+			if (len < HFIXEDSZ ||
+			    (resplen < HFIXEDSZ + QFIXEDSZ &&
+			    anssiz >= HFIXEDSZ + QFIXEDSZ)) {
 				/*
 				 * Undersized message.
 				 */
@@ -669,7 +671,9 @@
 				goto next_ns;
 			}
 			gotsomewhere = 1;
-			if (resplen < HFIXEDSZ) {
+			if (resplen < HFIXEDSZ ||
+			    (resplen < HFIXEDSZ + QFIXEDSZ &&
+			    anssiz >= HFIXEDSZ + QFIXEDSZ)) {
 				/*
 				 * Undersized message.
 				 */
