--- glibc-2.1.3/elf/ldd.bash.in.chmou	Tue Jan  5 21:04:34 1999
+++ glibc-2.1.3/elf/ldd.bash.in	Sat Jun  3 19:06:02 2000
@@ -32,6 +32,7 @@
 warn=
 bind_now=
 verbose=
+file_magic_regex="ELF [0-9][0-9]*-bit [LM]SB (shared object|dynamic lib)"
 
 while test $# -gt 0; do
   case "$1" in
@@ -123,12 +124,20 @@
     echo "ldd: ${file}:" $"No such file or directory" >&2
     result=1
   elif test -r "$file"; then
+    if eval file -L "$file" 2>/dev/null \
+    | sed 10q | egrep -v "$file_magic_regex" > /dev/null; then
     test -x "$file" || echo 'ldd:' $"\
 warning: you do not have execution permission for" "\`$file'" >&2
+    fi
     verify_out=`${RTLD} --verify "$file"`
     case $? in
     0)
-      eval $add_env '"$file"' || result=1
+      if [ ! -x "$file" ] && eval file -L "$file" 2>/dev/null \
+      | sed 10q | egrep "$file_magic_regex" > /dev/null; then
+        eval $add_env ${RTLD} '"$file"' || result=1
+      else
+        eval $add_env '"$file"' || result=1
+      fi
       ;;
     1)
       # This can be a non-ELF binary or no binary at all.
