--- glibc-2.1.3/glibcbug.in.orig	Wed Jun 12 10:48:56 2002
+++ glibc-2.1.3/glibcbug.in	Wed Jun 12 10:57:10 2002
@@ -40,8 +40,6 @@
 
 BUGADDR=${1-$BUGGLIBC}
 
-: ${EDITOR=emacs}
-
 : ${USER=${LOGNAME-`whoami`}}
 
 # How to read the passwd database.
@@ -87,14 +85,21 @@
 fi
 
 # If they don't have a preferred editor set, then use
-if [ -z "$VISUAL" ]; then
-  if [ -z "$EDITOR" ]; then
+if [ -z "$EDITOR" ]; then
+  if [ -n "$VISUAL" ]; then
+    EDIT="$VISUAL"
+  elif [ -x /bin/vitmp ]; then
+    EDIT=vitmp
+  elif [ -x /bin/vi ]; then
     EDIT=vi
+  elif [ -x /usr/bin/vim ]; then
+    EDIT=vim
   else
-    EDIT="$EDITOR"
+    echo "$0: No default editor found: attempting to use vi" >&2
+    EDIT=vi
   fi
 else
-  EDIT="$VISUAL"
+  EDIT="$EDITOR"
 fi
 
 # Find out some information.
