diff -ur glibc-2.3.6.orig/crypt/crypt_util.c glibc-2.3.6/crypt/crypt_util.c
--- glibc-2.3.6.orig/crypt/crypt_util.c	2002-11-03 00:58:59 +0000
+++ glibc-2.3.6/crypt/crypt_util.c	2010-03-29 19:28:44 +0000
@@ -262,6 +262,8 @@
 struct crypt_data _ufc_foobar;
 
 #ifdef __GNU_LIBRARY__
+#include <signal.h>
+#include <atomic.h>
 #include <bits/libc-lock.h>
 
 __libc_lock_define_initialized (static, _ufc_tables_lock)
@@ -349,7 +351,7 @@
   ufc_long j;
   ufc_long mask1, mask2;
   int e_inverse[64];
-  static volatile int small_tables_initialized = 0;
+  static volatile sig_atomic_t small_tables_initialized = 0;
 
 #ifdef _UFC_32_
   long32 *sb[4];
@@ -468,6 +470,7 @@
 	  efp[comes_from_word][word_value][o_long] |= mask2;
       }
     }
+    atomic_write_barrier ();
     small_tables_initialized = 1;
 #ifdef __GNU_LIBRARY__
 small_tables_done:
