# Fix sys/mount.h for gcc -pedantic support.

--- glibc-2.3.5/sysdeps/unix/sysv/linux/sys/mount.h.orig	2004-09-23 14:06:51 +0000
+++ glibc-2.3.5/sysdeps/unix/sysv/linux/sys/mount.h	2005-05-08 21:54:40 +0000
@@ -57,7 +57,7 @@ enum
 #define MS_NOATIME	MS_NOATIME
   MS_NODIRATIME = 2048,		/* Do not update directory access times.  */
 #define MS_NODIRATIME	MS_NODIRATIME
-  MS_BIND = 4096,		/* Bind directory at different place.  */
+  MS_BIND = 4096		/* Bind directory at different place.  */
 #define MS_BIND		MS_BIND
 };
 
