# Minor io/fts.c cleanup.

--- glibc-2.3.5.orig/io/fts.c	2004-09-14 04:24:44 +0000
+++ glibc-2.3.5/io/fts.c	2005-05-08 22:12:28 +0000
@@ -1031,10 +1031,7 @@ fts_palloc(sp, more)
 	 * We limit fts_pathlen to USHRT_MAX to be safe in both cases.
 	 */
 	if (sp->fts_pathlen < 0 || sp->fts_pathlen >= USHRT_MAX) {
-		if (sp->fts_path) {
-			free(sp->fts_path);
-			sp->fts_path = NULL;
-		}
+		free(sp->fts_path);
 		sp->fts_path = NULL;
 		__set_errno (ENAMETOOLONG);
 		return (1);
