diff -urN glibc-2.1.3.orig/time/getdate.c glibc-2.1.3.rh/time/getdate.c
--- glibc-2.1.3.orig/time/getdate.c	Thu Sep 23 21:16:22 1999
+++ glibc-2.1.3.rh/time/getdate.c	Wed Mar 15 18:57:15 2000
@@ -1,5 +1,5 @@
 /* Convert a string representation of time to a time value.
-   Copyright (C) 1997, 1998, 1999 Free Software Foundation, Inc.
+   Copyright (C) 1997, 1998, 1999, 2000 Free Software Foundation, Inc.
    This file is part of the GNU C Library.
    Contributed by Mark Kettenis <kettenis@phys.uva.nl>, 1997.
 
@@ -91,7 +91,7 @@
 	return 1;
       break;
     case 1:
-      if (mday >= 1 && mday < (__isleap (year) ? 29 : 28))
+      if (mday >= 1 && mday <= (__isleap (year) ? 29 : 28))
 	return 1;
       break;
     }
