diff -urN glibc-2.1.3.crypt_blowfish/crypt/sysdeps/unix/Makefile glibc-2.1.3/crypt/sysdeps/unix/Makefile
--- glibc-2.1.3.crypt_blowfish/crypt/sysdeps/unix/Makefile	Fri May  4 17:52:25 2001
+++ glibc-2.1.3/crypt/sysdeps/unix/Makefile	Fri May  4 17:56:06 2001
@@ -1,4 +1,4 @@
 ifeq ($(subdir),md5-crypt)
-libcrypt-routines += crypt crypt_util crypt_blowfish x86 crypt_gensalt wrapper
-dont_distribute += crypt.c crypt_util.c crypt_blowfish.c x86.S crypt_gensalt.c wrapper.c
+libcrypt-routines += crypt crypt_util crypt_blowfish x86 crypt_gensalt crypt_freesec wrapper
+dont_distribute += crypt.c crypt_util.c crypt_blowfish.c x86.S crypt_gensalt.c crypt_freesec.c wrapper.c
 endif
diff -urN glibc-2.1.3.crypt_blowfish/crypt/sysdeps/unix/wrapper.c glibc-2.1.3/crypt/sysdeps/unix/wrapper.c
--- glibc-2.1.3.crypt_blowfish/crypt/sysdeps/unix/wrapper.c	Fri May  4 17:52:53 2001
+++ glibc-2.1.3/crypt/sysdeps/unix/wrapper.c	Fri May  4 18:03:24 2001
@@ -39,6 +39,8 @@
 extern char *_crypt_gensalt_md5_rn(unsigned long count,
 	__CONST char *input, int size, char *output, int output_size);
 
+extern char *_crypt_extended(__CONST char *key, __CONST char *setting);
+
 #if defined(__GLIBC__) && defined(_LIBC)
 /* crypt.h from glibc-crypt-2.1 will define struct crypt_data for us */
 #include "crypt.h"
@@ -155,6 +157,10 @@
 
 char *__crypt(__const char *key, __const char *setting)
 {
+	if (setting[0] == '_') {
+		return _crypt_retval_magic(_crypt_extended(key, setting),
+			setting, (char *)&_ufc_foobar);
+	}
 	return _crypt_retval_magic(
 		__crypt_rn(key, setting, &_ufc_foobar, sizeof(_ufc_foobar)),
 		setting, (char *)&_ufc_foobar);
