diff -uNrp glibc-2.3.2.orig/stdlib/stdlib.h glibc-2.3.2/stdlib/stdlib.h
--- glibc-2.3.2.orig/stdlib/stdlib.h	2004-03-09 05:54:11 +0000
+++ glibc-2.3.2/stdlib/stdlib.h	2004-03-09 10:37:33 +0000
@@ -699,18 +699,15 @@ __END_NAMESPACE_STD
 
 #ifdef	__USE_GNU
 /* Return a malloc'd string containing the canonical absolute name of the
-   named file.  The last file name component need not exist, and may be a
-   symlink to a nonexistent file.  */
+   named file.  */
 extern char *canonicalize_file_name (__const char *__name) __THROW;
 #endif
 
 #if defined __USE_BSD || defined __USE_XOPEN_EXTENDED
-/* Return the canonical absolute name of file NAME.  The last file name
-   component need not exist, and may be a symlink to a nonexistent file.
-   If RESOLVED is null, the result is malloc'd; otherwise, if the canonical
-   name is PATH_MAX chars or more, returns null with `errno' set to
-   ENAMETOOLONG; if the name fits in fewer than PATH_MAX chars, returns the
-   name in RESOLVED.  */
+/* Return the canonical absolute name of file NAME.  If the canonical name
+   is PATH_MAX chars or more, returns null with `errno' set to
+   ENAMETOOLONG; if the name fits in fewer than PATH_MAX chars, returns
+   the name in RESOLVED.  */
 extern char *realpath (__const char *__restrict __name,
 		       char *__restrict __resolved) __THROW;
 #endif
