diff -urN glibc-2.1.3.orig/sysdeps/generic/paths.h glibc-2.1.3.rh/sysdeps/generic/paths.h
--- glibc-2.1.3.orig/sysdeps/generic/paths.h	Mon Jul 20 20:19:37 1998
+++ glibc-2.1.3.rh/sysdeps/generic/paths.h	Mon Dec 27 18:34:12 1999
@@ -49,7 +49,7 @@
 #define	_PATH_DEVNULL	"/dev/null"
 #define	_PATH_DRUM	"/dev/drum"
 #define	_PATH_KMEM	"/dev/kmem"
-#define	_PATH_MAILDIR	"/var/mail"
+#define	_PATH_MAILDIR	"/var/spool/mail"
 #define	_PATH_LASTLOG	"/var/log/lastlog"
 #define	_PATH_MAN	"/usr/man"
 #define	_PATH_MEM	"/dev/mem"
@@ -62,10 +62,10 @@
 #define	_PATH_SHADOW	"/etc/shadow"
 #define	_PATH_SHELLS	"/etc/shells"
 #define	_PATH_TTY	"/dev/tty"
-#define	_PATH_UNIX	"/vmunix"
+#define	_PATH_UNIX	"/boot/vmunix"
 #define	_PATH_UTMP	"/var/run/utmp"
 #define	_PATH_UTMP_DB	"/var/run/utmp.db"
-#define	_PATH_VI	"/usr/bin/vi"
+#define	_PATH_VI	"/bin/vi"
 #define	_PATH_WTMP	"/var/log/wtmp"
 
 /* Provide trailing slash, since mostly used for building pathnames. */
diff -urN glibc-2.1.3.orig/sysdeps/unix/sysv/linux/paths.h glibc-2.1.3.rh/sysdeps/unix/sysv/linux/paths.h
--- glibc-2.1.3.orig/sysdeps/unix/sysv/linux/paths.h	Wed Aug 12 20:05:51 1998
+++ glibc-2.1.3.rh/sysdeps/unix/sysv/linux/paths.h	Mon Dec 27 18:34:12 1999
@@ -51,7 +51,7 @@
 #define	_PATH_KLOG	"/proc/kmsg"
 #define	_PATH_KMEM	"/dev/kmem"
 #define	_PATH_LASTLOG	"/var/log/lastlog"
-#define	_PATH_MAILDIR	"/var/mail"
+#define	_PATH_MAILDIR	"/var/spool/mail"
 #define	_PATH_MAN	"/usr/man"
 #define	_PATH_MEM	"/dev/mem"
 #define	_PATH_MNTTAB	"/etc/fstab"
@@ -63,9 +63,9 @@
 #define	_PATH_SHADOW	"/etc/shadow"
 #define	_PATH_SHELLS	"/etc/shells"
 #define	_PATH_TTY	"/dev/tty"
-#define	_PATH_UNIX	"/vmlinux"
+#define	_PATH_UNIX	"/boot/vmlinux"
 #define _PATH_UTMP	"/var/run/utmp"
-#define	_PATH_VI	"/usr/bin/vi"
+#define	_PATH_VI	"/bin/vi"
 #define _PATH_WTMP	"/var/log/wtmp"
 
 /* Provide trailing slash, since mostly used for building pathnames. */
