diff -uNrp glibc-2.3.2.orig/manual/libc.texinfo glibc-2.3.2/manual/libc.texinfo
--- glibc-2.3.2.orig/manual/libc.texinfo	2002-09-15 02:18:55 +0000
+++ glibc-2.3.2/manual/libc.texinfo	2004-03-08 08:31:47 +0000
@@ -7,7 +7,7 @@
 @comment Tell install-info what to do.
 @dircategory GNU libraries
 @direntry
-* Libc: (libc).                 C library.
+* Libc: (libc).                                 C library.
 @end direntry
 
 @c This tells texinfo.tex to use the real section titles in xrefs in
