# Fix linuxthreads documentation.

--- glibc-2.3.5/linuxthreads/man/pthread_create.man.orig	1998-03-11 12:42:23 +0000
+++ glibc-2.3.5/linuxthreads/man/pthread_create.man	2004-04-21 12:13:09 +0000
@@ -31,10 +31,8 @@ error, a non-zero error code is returned
 .SH ERRORS
 .TP
 !EAGAIN!
-not enough system resources to create a process for the new thread.
-.TP
-!EAGAIN!
-more than !PTHREAD_THREADS_MAX! threads are already active.
+not enough system resources to create a process for the new thread,
+or more than !PTHREAD_THREADS_MAX! threads are already active.
 
 .SH AUTHOR
 Xavier Leroy <Xavier.Leroy@inria.fr>
