diff -ur glibc-2.1.3.orig/glibc-compat/nss_dns/dns-host.c glibc-2.1.3/glibc-compat/nss_dns/dns-host.c
--- glibc-2.1.3.orig/glibc-compat/nss_dns/dns-host.c	Tue Jan  4 04:07:08 2000
+++ glibc-2.1.3/glibc-compat/nss_dns/dns-host.c	Fri Jul  5 17:57:55 2002
@@ -424,7 +424,7 @@
 	  linebuflen -= n;
 	  /* Get canonical name.  */
 	  n = strlen (tbuf) + 1;	/* For the \0.  */
-	  if ((size_t) n > buflen || n >= MAXHOSTNAMELEN)
+	  if ((size_t) n > linebuflen || n >= MAXHOSTNAMELEN)
 	    {
 	      ++had_error;
 	      continue;
@@ -447,7 +447,7 @@
 	  cp += n;
 	  /* Get canonical name. */
 	  n = strlen (tbuf) + 1;   /* For the \0.  */
-	  if ((size_t) n > buflen || n >= MAXHOSTNAMELEN)
+	  if ((size_t) n > linebuflen || n >= MAXHOSTNAMELEN)
 	    {
 	      ++had_error;
 	      continue;
@@ -542,7 +542,12 @@
 	      linebuflen -= nn;
 	    }
 
-	  bp += sizeof (align) - ((u_long) bp % sizeof (align));
+	  {
+	    register int nn;
+	    nn = sizeof (align) - ((u_long) bp % sizeof (align));
+	    bp += nn;
+	    linebuflen -= nn;
+	  }
 
 	  if (n >= linebuflen)
 	    {
diff -ur glibc-2.1.3.orig/glibc-compat/nss_dns/dns-network.c glibc-2.1.3/glibc-compat/nss_dns/dns-network.c
--- glibc-2.1.3.orig/glibc-compat/nss_dns/dns-network.c	Tue Jan  4 04:07:08 2000
+++ glibc-2.1.3/glibc-compat/nss_dns/dns-network.c	Fri Jul  5 17:59:42 2002
@@ -283,7 +283,9 @@
 	    }
 	  cp += n;
 	  *alias_pointer++ = bp;
-	  bp += strlen (bp) + 1;
+	  n = strlen (bp) + 1;
+	  bp += n;
+	  linebuflen -= n;
 	  result->n_addrtype = class == C_IN ? AF_INET : AF_UNSPEC;
 	  ++have_answer;
 	}
