--- glibc-2.3.6/elf/ldd.bash.in.orig	2005-02-16 10:29:34 +0000
+++ glibc-2.3.6/elf/ldd.bash.in	2006-05-16 16:33:45 +0000
@@ -74,6 +74,7 @@ For bug reporting instructions, please s
     ;;
   -u | --u | --un | --unu | --unus | --unuse | --unused)
     unused=yes
+    bind_now=yes
     shift
     ;;
   --v | --ve | --ver)
@@ -148,8 +149,7 @@ for file do
     echo "ldd: ${file}:" $"No such file or directory" >&2
     result=1
   elif test -r "$file"; then
-    test -x "$file" || echo 'ldd:' $"\
-warning: you do not have execution permission for" "\`$file'" >&2
+    ret=
     RTLD=
     for rtld in ${RTLDLIST}; do
       if test -x $rtld; then
@@ -160,24 +160,15 @@ warning: you do not have execution permi
 	esac
       fi
     done
-    if test -z "${RTLD}"; then
+    if test -z "${ret}"; then
       set ${RTLDLIST}
       RTLD=$1
       verify_out=`${RTLD} --verify "$file"`
       ret=$?
     fi
     case $ret in
-    0)
-      # If the program exits with exit code 5, it means the process has been
-      # invoked with __libc_enable_secure.  Fall back to running it through
-      # the dynamic linker.
-      try_trace "$file"
-      rc=$?
-      if [ $rc = 5 ]; then
-	try_trace "$RTLD" "$file"
-	rc=$?
-      fi
-      [ $rc = 0 ] || result=1
+    [02])
+      try_trace "$RTLD" "$file" || result=1
       ;;
     1)
       # This can be a non-ELF binary or no binary at all.
@@ -186,9 +177,6 @@ warning: you do not have execution permi
 	result=1
       }
       ;;
-    2)
-      try_trace "$RTLD" "$file" || result=1
-      ;;
     *)
       echo 'ldd:' ${RTLD} $"exited with unknown exit code" "($ret)" >&2
       exit 1
