diff -ur glibc-2.1.3.orig/malloc/malloc.c glibc-2.1.3/malloc/malloc.c
--- glibc-2.1.3.orig/malloc/malloc.c	Wed Feb 23 07:02:55 2000
+++ glibc-2.1.3/malloc/malloc.c	Mon Dec  8 03:07:36 2003
@@ -366,6 +366,11 @@
 #define assert(x) ((void)0)
 #endif
 
+#include <abort-instr.h>
+#ifndef ABORT_INSTRUCTION
+#define ABORT_INSTRUCTION
+#endif
+
 
 /*
   INTERNAL_SIZE_T is the word-size used for internal bookkeeping
@@ -2405,6 +2410,13 @@
 {                                                                             \
   BK = P->bk;                                                                 \
   FD = P->fd;                                                                 \
+  if (FD->bk != P || BK->fd != P)                                             \
+  {                                                                           \
+    ABORT_INSTRUCTION;                                                        \
+    _exit(127);                                                               \
+    while (1)                                                                 \
+      ABORT_INSTRUCTION;                                                      \
+  }                                                                           \
   FD->bk = BK;                                                                \
   BK->fd = FD;                                                                \
 }                                                                             \
