--- cvs-1.11.23/FAQ
+++ cvs-1.11.23/FAQ
@@ -7167,7 +7167,7 @@ kingdon@cyclic.com
 
         EDITOR          Used if CVSEDITOR doesn't exist.  If EDITOR
                         doesn't exist, CVS uses a configured constant,
-                        usually, "vi".
+                        usually, "vitmp".
 
         CVSREAD         Sets files to read-only on "checkout".
 
--- cvs-1.11.23/configure.in
+++ cvs-1.11.23/configure.in
@@ -443,7 +443,7 @@ fi
 # Set the default when --with-editor wasn't supplied or when it was supplied
 # without an argument.
 if test yes = $with_editor; then
-  with_editor="vim vi emacs nano pico edit"
+  with_editor="vitmp"
 fi
 
 if echo $with_editor |grep ^/ >/dev/null; then
--- cvs-1.11.23/contrib/rcs-to-cvs.sh
+++ cvs-1.11.23/contrib/rcs-to-cvs.sh
@@ -87,7 +87,7 @@ fi
 if [ $got_one -eq 0 ]; then
     echo "Please Edit this file to contain the RCS log information" >$message_file
     echo "to be associated with this directory (please remove these lines)">>$message_file
-    ${EDITOR-vi} $message_file
+    ${EDITOR-vitmp} $message_file
     got_one=1
 fi
 
--- cvs-1.11.23/man/cvsbug.8
+++ cvs-1.11.23/man/cvsbug.8
@@ -163,7 +163,7 @@ The environment variable
 specifies the editor to invoke on the template.
 .br
 default:
-.B vi
+.B vitmp
 .sp
 If the environment variable 
 .B PR_FORM
--- cvs-1.11.23/src/cvsbug.in
+++ cvs-1.11.23/src/cvsbug.in
@@ -136,7 +136,7 @@ fi
 # If they don't have a preferred editor set, then use
 if [ -z "$VISUAL" ]; then
   if [ -z "$EDITOR" ]; then
-    EDIT=vi
+    EDIT=vitmp
   else
     EDIT="$EDITOR"
   fi
