--- cvs-1.11.23/src/login.c
+++ cvs-1.11.23/src/login.c
@@ -201,11 +201,21 @@ password_entry_parseline (cvsroot_canonical, warn, linenumber, linebuf)
 	    return NULL;
 	}
 	*p = ' ';
-	tmp_root_canonical = normalize_cvsroot (tmp_root);
-	if (strcmp (cvsroot_canonical, tmp_root_canonical) == 0)
-	    password = p + 1;
-
-	free (tmp_root_canonical);
+        switch (tmp_root->method)
+        {
+	    case gserver_method:
+	    case pserver_method:
+#ifdef HAVE_KERBEROS
+	    case kserver_method:
+#endif
+	        tmp_root_canonical = normalize_cvsroot (tmp_root);
+		if (strcmp (cvsroot_canonical, tmp_root_canonical) == 0)
+		    password = p + 1;
+		free (tmp_root_canonical);
+		break;
+	    default:
+	        break;
+        }
 	free_cvsroot_t (tmp_root);
     }
 
