--- cvs-1.11.23/src/client.c
+++ cvs-1.11.23/src/client.c
@@ -3631,7 +3631,7 @@ get_port_number (envname, portname, defaultport)
 	if (port <= 0)
 	{
 	    error (0, 0, "%s must be a positive integer!  If you", envname);
-	    error (0, 0, "are trying to force a connection via rsh, please");
+	    error (0, 0, "are trying to force a connection via ssh, please");
 	    error (0, 0, "put \":server:\" at the beginning of your CVSROOT");
 	    error (1, 0, "variable.");
 	}
@@ -4785,7 +4785,7 @@ start_rsh_server (root, to_server, from_server)
     /* Do the deed. */
     child_pid = popenRW (rsh_argv, pipes);
     if (child_pid < 0)
-	error (1, errno, "cannot start server via rsh");
+	error (1, errno, "cannot start server via %s", cvs_rsh);
 
     /* Give caller the file descriptors in a form it can deal with. */
     make_bufs_from_fds (pipes[0], pipes[1], child_pid, to_server, from_server, 0);
@@ -4861,7 +4861,7 @@ start_rsh_server (root, to_server, from_server)
 	child_pid = piped_child (argv, &tofd, &fromfd, 1);
 
 	if (child_pid < 0)
-	    error (1, errno, "cannot start server via rsh");
+	    error (1, errno, "cannot start server via %s", cvs_rsh);
     }
     free (command);
 
