diff -urp cvs-1.11.5.orig/src/root.c cvs-1.11.5/src/root.c
--- cvs-1.11.5.orig/src/root.c	Fri Dec  6 19:10:28 2002
+++ cvs-1.11.5/src/root.c	Mon Jun 23 14:16:27 2003
@@ -571,6 +571,14 @@ parse_cvsroot (root_in)
 	    error (0, 0, "when using local access method.");
 	    goto error_exit;
 	}
+	{
+	    char *dir = realpath (newroot->directory, NULL); /* Assume glibc */
+	    if (dir)
+	    {
+		free (newroot->directory);
+		newroot->directory = dir;
+	    }
+	}
 	no_port = 1;
 	no_password = 1;
 	break;
@@ -724,7 +732,8 @@ local_cvsroot (dir)
 
     newroot->original = xstrdup(dir);
     newroot->method = local_method;
-    newroot->directory = xstrdup(dir);
+    /* Assume GNU C and glibc */
+    newroot->directory = realpath (dir, NULL) ?: xstrdup(dir);
 
     return newroot;
 }
diff -urp cvs-1.11.5.orig/src/server.c cvs-1.11.5/src/server.c
--- cvs-1.11.5.orig/src/server.c	Thu Jan 16 15:10:55 2003
+++ cvs-1.11.5/src/server.c	Mon Jun 23 14:16:39 2003
@@ -1151,6 +1151,14 @@ serve_directory (arg)
     status = buf_read_line (buf_from_net, &repos, (int *) NULL);
     if (status == 0)
     {
+	char *real_repos = realpath (repos, NULL); /* Assume glibc */
+
+	if (real_repos)
+	{
+	    free (repos);
+	    repos = real_repos;
+	}
+
 	if (!outside_root (repos))
 	    dirswitch (arg, repos);
 	free (repos);
