diff -upk.orig cvs-1.11.20.orig/doc/cvs.texinfo cvs-1.11.20/doc/cvs.texinfo
--- cvs-1.11.20.orig/doc/cvs.texinfo	2005-09-29 12:53:55 +0000
+++ cvs-1.11.20/doc/cvs.texinfo	2005-09-29 12:54:21 +0000
@@ -2214,8 +2214,9 @@ There is no need to edit @file{inetd.con
 There are two access methods that you use in @code{CVSROOT}
 for rsh.  @code{:server:} specifies an internal rsh
 client, which is supported only by some @sc{cvs} ports.
+This is not supported on most Unix-style systems, including GNU/Linux.
 @code{:ext:} specifies an external rsh program.  By
-default this is @code{rsh} but you may set the
+default this is @code{ssh} but you may set the
 @code{CVS_RSH} environment variable to invoke another
 program which can access the remote server (for
 example, @code{remsh} on HP-UX 9 because @code{rsh} is
@@ -3176,6 +3177,9 @@ $ cvs import -m "Created directory struc
 This will add yoyodyne/@var{dir} as a directory under
 @code{$CVSROOT}.
 
+This will add yoyodyne/@var{dir} as a directory under
+@code{$CVSROOT}.
+
 Then, use @code{add} to add files (and new directories)
 as they appear.
 
@@ -11968,7 +11972,7 @@ U first-dir/sdir/sfile
 @item @var{mname} [ options ] @var{dir} [ @var{files}@dots{} ]
 In the simplest case, this form of module definition
 reduces to @samp{@var{mname} @var{dir}}.  This defines
-all the files in directory @var{dir} as module mname.
+all the files in directory @var{dir} as module @var{mname}.
 @var{dir} is a relative path (from @code{$CVSROOT}) to a
 directory of source in the source repository.  In this
 case, on checkout, a single directory called
