--- cvs-1.11.23/src/mkmodules.c
+++ cvs-1.11.23/src/mkmodules.c
@@ -754,10 +754,6 @@ rename_dbmfile (temp)
     (void) sprintf (newpag, "%s.pag", temp);
     (void) sprintf (newdb, "%s.db", temp);
 
-    (void) chmod (newdir, 0666);
-    (void) chmod (newpag, 0666);
-    (void) chmod (newdb, 0666);
-
     /* don't mess with me */
     SIG_beginCrSect ();
 
@@ -1024,11 +1020,6 @@ init (argc, argv)
 	fp = open_file (info, "w");
 	if (fclose (fp) < 0)
 	    error (1, errno, "cannot close %s", info);
- 
-        /* Make the new history file world-writeable, since every CVS
-           user will need to be able to write to it.  We use chmod()
-           because xchmod() is too shy. */
-        chmod (info, 0666);
     }
 
     /* Make an empty val-tags file to prevent problems creating it later.  */
@@ -1042,11 +1033,6 @@ init (argc, argv)
 	fp = open_file (info, "w");
 	if (fclose (fp) < 0)
 	    error (1, errno, "cannot close %s", info);
- 
-        /* Make the new val-tags file world-writeable, since every CVS
-           user will need to be able to write to it.  We use chmod()
-           because xchmod() is too shy. */
-        chmod (info, 0666);
     }
 
     free (info);
