diff -ur cvs-1.11.5.orig/configure cvs-1.11.5/configure
--- cvs-1.11.5.orig/configure	Thu Jan 16 21:36:48 2003
+++ cvs-1.11.5/configure	Mon Apr 28 19:55:48 2003
@@ -9052,7 +9052,7 @@
   withval="$with_editor"
 
 else
-  with_editor="vi emacs pico edit"
+  with_editor="vitmp"
 fi;
 
 
diff -ur cvs-1.11.5.orig/contrib/rcs-to-cvs.sh cvs-1.11.5/contrib/rcs-to-cvs.sh
--- cvs-1.11.5.orig/contrib/rcs-to-cvs.sh	Fri Feb 21 03:44:46 1997
+++ cvs-1.11.5/contrib/rcs-to-cvs.sh	Mon Apr 28 19:57:18 2003
@@ -78,7 +78,7 @@
 if [ $got_one -eq 0 ]; then
     echo "Please Edit this file to contain the RCS log information" >$message_file
     echo "to be associated with this directory (please remove these lines)">>$message_file
-    ${EDITOR-vi} $message_file
+    ${EDITOR-vitmp} $message_file
     got_one=1
 fi
 
diff -ur cvs-1.11.5.orig/src/cvsbug.in cvs-1.11.5/src/cvsbug.in
--- cvs-1.11.5.orig/src/cvsbug.in	Mon Apr 29 04:45:47 2002
+++ cvs-1.11.5/src/cvsbug.in	Mon Apr 28 19:57:38 2003
@@ -135,7 +135,7 @@
 # If they don't have a preferred editor set, then use
 if [ -z "$VISUAL" ]; then
   if [ -z "$EDITOR" ]; then
-    EDIT=vi
+    EDIT=vitmp
   else
     EDIT="$EDITOR"
   fi
