--- cvs-1.11.23/src/sanity.sh
+++ cvs-1.11.23/src/sanity.sh
@@ -290,20 +290,25 @@ fi
 #
 CVS="${testcvs} -Q"
 
-LOGFILE=`pwd`/check.log
+if $remote; then
+	CHECK_LOG=check-r.log
+else
+	CHECK_LOG=check.log
+fi
+LOGFILE=`pwd`/$CHECK_LOG
 
 # Save the previous log in case the person running the tests decides
 # they want to look at it.  The extension ".plog" is chosen for consistency
 # with dejagnu.
-if test -f check.log; then
-	mv check.log check.plog
+if test -f $CHECK_LOG; then
+	mv $CHECK_LOG $CHECK_LOG.orig
 fi
 
-# Create the log file so check.log can be tailed almost immediately after
+# Create the log file so $CHECK_LOG can be tailed almost immediately after
 # this script is started.  Otherwise it can take up to a minute or two before
 # the log file gets created when $remotehost is specified on some systems,
 # which makes for a lot of failed `tail -f' attempts.
-touch check.log
+touch $CHECK_LOG
 
 # Workaround any X11Forwarding by ssh. Otherwise this text:
 #   Warning: No xauth data; using fake authentication data for X11 forwarding.
@@ -863,7 +868,7 @@ localonly ()
 fail ()
 {
   echo "FAIL: $1" | tee -a ${LOGFILE}
-  echo "*** Please see the \`TESTS' and \`check.log' files for more information." >&2
+  echo "*** Please see the \`TESTS' and \`$CHECK_LOG' files for more information." >&2
   # This way the tester can go and see what remnants were left
   exit 1
 }
@@ -11759,6 +11764,9 @@ EOF
 	    dotest_fail conflicts4-10r "$testcvs -q up" "C file1"
 	    dotest conflicts4-11r "cat $TESTDIR/conflicts4/client.out" \
 "$DOTSTAR
+$DOTSTAR
+$DOTSTAR
+$DOTSTAR
 Argument --
 Directory .
 $CVSROOT_DIRNAME/first-dir
