diff -upk.orig cvs-1.11.20.orig/configure.in cvs-1.11.20/configure.in
--- cvs-1.11.20.orig/configure.in	2005-09-28 22:57:03 +0000
+++ cvs-1.11.20/configure.in	2005-09-28 22:57:30 +0000
@@ -1,5 +1,5 @@
 dnl configure.in for cvs
-AC_INIT([Concurrent Versions System (CVS)],[1.11.20],[bug-cvs@gnu.org],[cvs])
+AC_INIT([Concurrent Versions System (CVS)],[1.11.20-alt],[bug-cvs@gnu.org],[cvs])
 AC_CONFIG_SRCDIR(src/cvs.h)
 AM_INIT_AUTOMAKE([gnu 1.7.9 dist-bzip2 no-define])
 AC_PREREQ(2.58)
diff -upk.orig cvs-1.11.20.orig/src/sanity.sh cvs-1.11.20/src/sanity.sh
--- cvs-1.11.20.orig/src/sanity.sh	2005-03-04 19:05:12 +0000
+++ cvs-1.11.20/src/sanity.sh	2005-09-28 22:57:30 +0000
@@ -2084,8 +2084,8 @@ Specify the --help option for further in
 
 	  if $remote; then
 		dotest version-2r "${testcvs} version" \
-'Client: Concurrent Versions System (CVS) [0-9p.]* (client/server)
-Server: Concurrent Versions System (CVS) [0-9p.]* (client/server)'
+'Client: Concurrent Versions System (CVS) [0-9p.]*-alt (client/server)
+Server: Concurrent Versions System (CVS) [0-9p.]*-alt (client/server)'
 	  else
 		dotest version-2 "${testcvs} version" \
 'Concurrent Versions System (CVS) [0-9.]*.*'
