diff -ur cvs-1.11.5/src/server.c cvs-1.11.5-owl-no-checkin-update-prog/src/server.c
--- cvs-1.11.5/src/server.c	Thu Jan 16 15:10:55 2003
+++ cvs-1.11.5-owl-no-checkin-update-prog/src/server.c	Tue Apr 22 13:26:28 2003
@@ -4638,6 +4638,7 @@
     buf_send_counted (protocol);
 }
 
+#if 0
 static void
 serve_checkin_prog (arg)
     char *arg;
@@ -4715,7 +4716,9 @@
 	return;
     }
 }
-
+
+#endif
+
 static void serve_valid_requests PROTO((char *arg));
 
 #endif /* SERVER_SUPPORT */
@@ -4745,8 +4748,8 @@
   REQ_LINE("Max-dotdot", serve_max_dotdot, 0),
   REQ_LINE("Static-directory", serve_static_directory, 0),
   REQ_LINE("Sticky", serve_sticky, 0),
-  REQ_LINE("Checkin-prog", serve_checkin_prog, 0),
-  REQ_LINE("Update-prog", serve_update_prog, 0),
+  REQ_LINE("Checkin-prog", serve_noop, 0),
+  REQ_LINE("Update-prog", serve_noop, 0),
   REQ_LINE("Entry", serve_entry, RQ_ESSENTIAL),
   REQ_LINE("Kopt", serve_kopt, 0),
   REQ_LINE("Checkin-time", serve_checkin_time, 0),
