diff -ur cvs-1.11.5.orig/configure cvs-1.11.5/configure
--- cvs-1.11.5.orig/configure	Thu Jan 16 21:36:48 2003
+++ cvs-1.11.5/configure	Sun Apr 27 13:31:05 2003
@@ -10184,24 +10184,13 @@
 # Create a temporary directory, and hook for its removal unless debugging.
 $debug ||
 {
-  trap 'exit_status=$?; rm -rf $tmp && exit $exit_status' 0
-  trap '{ (exit 1); exit 1; }' 1 2 13 15
+  tmp=
+  trap 'exit_status=$?; rm -rf -- "$tmp"; exit $exit_status' EXIT
+  trap 'trap - EXIT; rm -rf -- "$tmp"; exit 1' HUP INT PIPE TERM
 }
 
 # Create a (secure) tmp directory for tmp files.
-: ${TMPDIR=/tmp}
-{
-  tmp=`(umask 077 && mktemp -d -q "$TMPDIR/csXXXXXX") 2>/dev/null` &&
-  test -n "$tmp" && test -d "$tmp"
-}  ||
-{
-  tmp=$TMPDIR/cs$$-$RANDOM
-  (umask 077 && mkdir $tmp)
-} ||
-{
-   echo "$me: cannot create a temporary directory in $TMPDIR" >&2
-   { (exit 1); exit 1; }
-}
+tmp="`mktemp -dt cs.XXXXXXXXXX`" || exit 1
 
 _ACEOF
 
diff -ur cvs-1.11.5.orig/contrib/cvs2vendor.sh cvs-1.11.5/contrib/cvs2vendor.sh
--- cvs-1.11.5.orig/contrib/cvs2vendor.sh	Wed Feb 12 15:34:55 1997
+++ cvs-1.11.5/contrib/cvs2vendor.sh	Sun Apr 27 12:49:20 2003
@@ -36,11 +36,12 @@
 tsrcdir=$1
 tdstdir=$2
 
-revfile=/tmp/cvs2vendor_$$_rev
-rm -f $revfile
-
-commentfile=/tmp/cvs2vendor_$$_comment
-rm -f $commentfile
+revfile=
+commentfile=
+trap 'rm -f -- "$revfile" "$commentfile"' EXIT
+trap 'trap - EXIT ; rm -f -- "$revfile" "$commentfile"; exit 1' HUP INT QUIT TERM
+revfile="`mktemp -t cvs2vendor_rev.XXXXXXXXXX`" || exit
+commentfile="`mktemp -t cvs2vendor_comment.XXXXXXXXXX`" || exit
 
 srcdirs=`cd $tsrcdir && find . -type d -print | sed 's~^\.[/]*~~'`
 
diff -ur cvs-1.11.5.orig/contrib/debug_check_log.sh cvs-1.11.5/contrib/debug_check_log.sh
--- cvs-1.11.5.orig/contrib/debug_check_log.sh	Fri Nov  8 17:27:40 2002
+++ cvs-1.11.5/contrib/debug_check_log.sh	Sun Apr 27 12:34:36 2003
@@ -19,7 +19,10 @@
 # Contributed by Derek R. Price <derek.price@openavenue.com>
 #
 
-
+dcldir=
+trap 'rm -rf -- "$dcldir"' EXIT
+trap 'trap - EXIT; rm -rf -- "$dcldir"; exit 1' HUP INT QUIT TERM
+dcldir="`mktemp -td dcl.XXXXXXXXXX`" || exit
 
 usage ()
 {
@@ -96,13 +99,13 @@
 		# file contains a second regex
 		if test $dcl_dofirst -eq 1; then
 			# get the first pattern
-			sed -ne '/^\*\* expected: $/,/^\*\* or: $/p' <$1 >/tmp/dcle$$
-			dcl_exprfiles="$dcl_exprfiles /tmp/dcle$$"
+			sed -ne '/^\*\* expected: $/,/^\*\* or: $/p' <$1 >$dcldir/e
+			dcl_exprfiles="$dcl_exprfiles $dcldir/e"
 		fi
 		if test $dcl_doalternate -eq 1; then
 			# get the alternate pattern
-			sed -ne '/^\*\* or: $/,/^\*\* got: $/p' <$1 >/tmp/dclo$$
-			dcl_exprfiles="$dcl_exprfiles /tmp/dclo$$"
+			sed -ne '/^\*\* or: $/,/^\*\* got: $/p' <$1 >$dcldir/o
+			dcl_exprfiles="$dcl_exprfiles $dcldir/o"
 		else
 			echo "WARNING:  Ignoring alternate pattern in file: $1" >&2
 		fi
@@ -110,8 +113,8 @@
 		# file doesn't contain a second regex
 		if test $dcl_dofirst = 1; then
 			# get the only pattern
-			sed -ne '/^\*\* expected: $/,/^\*\* got: $/p' <$1 >/tmp/dcle$$
-			dcl_exprfiles="$dcl_exprfiles /tmp/dcle$$"
+			sed -ne '/^\*\* expected: $/,/^\*\* got: $/p' <$1 >$dcldir/e
+			dcl_exprfiles="$dcl_exprfiles $dcldir/e"
 		fi
 		if test $dcl_doalternate -eq 1; then
 			echo "WARNING:  No alternate pattern in file:  $1" >&2
@@ -119,18 +122,18 @@
 	fi
 
 	# and get the actual output
-	sed -ne '/^\*\* got: $/,$p' <$1 >/tmp/dclg$$
+	sed -ne '/^\*\* got: $/,$p' <$1 >$dcldir/g
 	sed -ne '1D
 $D
-p' </tmp/dclg$$ >/tmp/dclh$$
-	mv /tmp/dclh$$ /tmp/dclg$$
+p' <$dcldir/g >$dcldir/h
+	mv $dcldir/h $dcldir/g
 
 	# compare the output against each pattern requested
 	for dcl_f in $dcl_exprfiles; do
 		sed -ne '1D
 $D
-p' <$dcl_f >/tmp/dclp$$
-		mv /tmp/dclp$$ $dcl_f
+p' <$dcl_f >$dcldir/p
+		mv $dcldir/p $dcl_f
 
 		case $dcl_f in
 			/tmp/dcle*)
@@ -141,12 +144,12 @@
 				;;
 		esac
 
-		expr_line_by_line /tmp/dclg$$ $dcl_f
+		expr_line_by_line $dcldir/g $dcl_f
 
 		rm $dcl_f
 	done
 
-	rm /tmp/dclg$$
+	rm $dcldir/g
 }
 
 ###
diff -ur cvs-1.11.5.orig/contrib/rcs-to-cvs.sh cvs-1.11.5/contrib/rcs-to-cvs.sh
--- cvs-1.11.5.orig/contrib/rcs-to-cvs.sh	Fri Feb 21 03:44:46 1997
+++ cvs-1.11.5/contrib/rcs-to-cvs.sh	Sun Apr 27 12:36:26 2003
@@ -32,7 +32,10 @@
 usage="Usage: rcs-to-cvs [-v] [-m message] [-f message_file] repository"
 vbose=0
 message=""
-if [ -d /var/tmp ]; then message_file=/var/tmp/checkin.$$; else message_file=/usr/tmp/checkin.$$; fi
+message_file=
+trap 'rm -f -- "$message_file"' EXIT
+trap 'trap - EXIT; rm -f -- "$message_file"; exit 1' HUP INT QUIT TERM
+message_file="`mktemp -t checkin.XXXXXXXXXX`" || exit
 got_one=0
 
 if [ $# -lt 1 ]; then
diff -ur cvs-1.11.5.orig/contrib/rcs2log.sh cvs-1.11.5/contrib/rcs2log.sh
--- cvs-1.11.5.orig/contrib/rcs2log.sh	Thu Sep  6 15:55:51 2001
+++ cvs-1.11.5/contrib/rcs2log.sh	Sun Apr 27 12:37:20 2003
@@ -182,10 +182,12 @@
 	m[9]="Oct"; m[10]="Nov"; m[11]="Dec"
 '
 
-logdir=$TMPDIR/rcs2log$$
+logdir=
+trap 'rm -rf -- "$logdir"' EXIT
+trap 'trap - EXIT; rm -rf -- "$logdir"; exit 1' HUP INT QUIT TERM
+logdir="`mktemp -td rcs2log.XXXXXXXXXX`" || exit
+
 llogout=$logdir/l
-trap exit 1 2 13 15
-trap "rm -fr $logdir 2>/dev/null" 0
 (umask 077 && exec mkdir $logdir) || exit
 
 # If no rlog-format log file is given, generate one into $rlogfile.
diff -ur cvs-1.11.5.orig/contrib/rcs2sccs.sh cvs-1.11.5/contrib/rcs2sccs.sh
--- cvs-1.11.5.orig/contrib/rcs2sccs.sh	Tue Oct 16 19:06:24 2001
+++ cvs-1.11.5/contrib/rcs2sccs.sh	Sun Apr 27 12:38:59 2003
@@ -9,20 +9,19 @@
     mkdir SCCS
 fi
 
-logfile=/tmp/rcs2sccs_$$_log
-rm -f $logfile
-tmpfile=/tmp/rcs2sccs_$$_tmp
-rm -f $tmpfile
-emptyfile=/tmp/rcs2sccs_$$_empty
-echo -n "" > $emptyfile
-initialfile=/tmp/rcs2sccs_$$_init
+rcs2sccsdir=
+trap 'rm -rf -- "$rcs2sccsdir"' EXIT
+trap 'trap - EXIT; rm -rf -- "$rcs2sccsdir"; exit 1' HUP INT QUIT TERM
+rcs2sccsdir="`mktemp -td rcs2sccs.XXXXXXXXXX`" || exit
+logfile="$rcs2sccsdir/log"
+tmpfile="$rcs2sccsdir/tmp"
+emptyfile="$rcs2sccsdir/empty"
+> $emptyfile
+initialfile="$rcs2sccsdir/init"
 echo "Initial revision" > $initialfile
-sedfile=/tmp/rcs2sccs_$$_sed
-rm -f $sedfile
-revfile=/tmp/rcs2sccs_$$_rev
-rm -f $revfile
-commentfile=/tmp/rcs2sccs_$$_comment
-rm -f $commentfile
+sedfile="$rcs2sccsdir/sed"
+revfile="$rcs2sccsdir/rev"
+commentfile="$rcs2sccsdir/comment"
 
 # create the sed script
 cat > $sedfile << EOF
diff -ur cvs-1.11.5.orig/man/cvsbug.8 cvs-1.11.5/man/cvsbug.8
--- cvs-1.11.5.orig/man/cvsbug.8	Tue Nov 27 16:24:58 2001
+++ cvs-1.11.5/man/cvsbug.8	Tue Apr 22 12:48:37 2003
@@ -201,12 +201,11 @@
 instructions on submitting larger test cases and problematic source
 code.
 .SH FILES
-.ta \w'/tmp/pbad$$  'u
-/tmp/p$$	copy of PR used in editing session
+$TMPDIR/p.XXXXXXXXXX	copy of PR used in editing session
 .br
-/tmp/pf$$	copy of empty PR form, for testing purposes
+$TMPDIR/pf.XXXXXXXXXX	copy of empty PR form, for testing purposes
 .br
-/tmp/pbad$$	file for rejected PRs
+$TMPDIR/pbad.XXXXXXXXXX	file for rejected PRs
 .SH INSTALLATION AND CONFIGURATION
 See 
 .B INSTALL
diff -ur cvs-1.11.5.orig/src/cvsbug.in cvs-1.11.5/src/cvsbug.in
--- cvs-1.11.5.orig/src/cvsbug.in	Mon Apr 29 04:45:47 2002
+++ cvs-1.11.5/src/cvsbug.in	Sun Apr 27 12:48:05 2003
@@ -83,11 +83,18 @@
 
 #
 
-[ -z "$TMPDIR" ] && TMPDIR=/tmp
+# $xs kludge apparently needed by Sun /bin/sh (and is relied upon by the
+# rest of this script, on all platforms).
+xs=0
 
-TEMP=$TMPDIR/p$$
-BAD=$TMPDIR/pbad$$
-REF=$TMPDIR/pf$$
+TEMP=
+BAD=
+REF=
+trap 'rm -f -- "$TEMP" "$BAD" "$REF"; exit $xs' EXIT
+trap 'trap - EXIT; rm -f -- "$TEMP" "$BAD" "$REF"; exit 1' HUP INT QUIT TERM
+TEMP="`mktemp -t p.XXXXXXXXXX`"
+BAD="`mktemp -t pbad.XXXXXXXXXX`"
+REF="`mktemp -t pf.XXXXXXXXXX`"
 
 if [ -z "$LOGNAME" -a -n "$USER" ]; then
   LOGNAME=$USER
@@ -263,11 +270,6 @@
 DESCRIPTION_C='<precise description of the problem (multiple lines)>'
 HOW_TO_REPEAT_C='<code/input/activities to reproduce the problem (multiple lines)>'
 FIX_C='<how to correct or work around the problem, if known (multiple lines)>'
-
-# Catch some signals. ($xs kludge needed by Sun /bin/sh)
-xs=0
-trap 'rm -f $REF $TEMP; exit $xs' 0
-trap 'echo "$COMMAND: Aborting ..."; rm -f $REF $TEMP; xs=1; exit' 1 2 3 13 15
 
 # If they told us to use a specific file, then do so.
 if [ -n "$IN_FILE" ]; then
diff -ur cvs-1.11.5.orig/src/filesubr.c cvs-1.11.5/src/filesubr.c
--- cvs-1.11.5.orig/src/filesubr.c	Tue Sep 24 20:47:09 2002
+++ cvs-1.11.5/src/filesubr.c	Sun Apr 27 13:10:43 2003
@@ -703,7 +703,7 @@
     if (fp == NULL)
 	error (1, errno, "Failed to create temporary file");
     if (fclose (fp) == EOF)
-	error (0, errno, "Failed to close temporary file %s", fn);
+	error (1, errno, "Failed to close temporary file %s", fn);
     return fn;
 }
 
@@ -755,9 +755,10 @@
 
     {
     int fd;
+    int save_errno;
 
-    fn = xmalloc (strlen (Tmpdir) + 11);
-    sprintf (fn, "%s/%s", Tmpdir, "cvsXXXXXX" );
+    fn = xmalloc (strlen (Tmpdir) + 12);
+    sprintf (fn, "%s/%s", Tmpdir, "cvs.XXXXXX" );
     fd = mkstemp (fn);
 
     /* a NULL return will be interpreted by callers as an error and
@@ -766,23 +767,25 @@
     if (fd == -1) fp = NULL;
     else if ((fp = CVS_FDOPEN (fd, "w+")) == NULL)
     {
-	/* attempt to close and unlink the file since mkstemp returned sucessfully and
+	/* attempt to unlink the file since mkstemp returned sucessfully and
 	 * we believe it's been created and opened
 	 */
- 	int save_errno = errno;
+	save_errno = errno;
 	if (close (fd))
-	    error (0, errno, "Failed to close temporary file %s", fn);
+	    error (1, errno, "Failed to close temporary file %s", fn);
 	if (CVS_UNLINK (fn))
-	    error (0, errno, "Failed to unlink temporary file %s", fn);
+	    error (1, errno, "Failed to unlink temporary file %s", fn);
 	errno = save_errno;
     }
 
+    save_errno = errno;
     if (fp == NULL) free (fn);
     /* mkstemp is defined to open mode 0600 using glibc 2.0.7+ */
     /* FIXME - configure can probably tell us which version of glibc we are
      * linking to and not chmod for 2.0.7+
      */
     else chmod (fn, 0600);
+    errno = save_errno;
 
     }
 
diff -ur cvs-1.11.5.orig/src/logmsg.c cvs-1.11.5/src/logmsg.c
--- cvs-1.11.5.orig/src/logmsg.c	Fri Sep 20 02:48:59 2002
+++ cvs-1.11.5/src/logmsg.c	Sun Apr 27 12:08:20 2003
@@ -208,13 +208,8 @@
     if (strcmp (Editor, "") == 0 && !editinfo_editor)
 	error(1, 0, "no editor defined, must use -e or -m");
 
-    /* Create a temporary file */
-    /* FIXME - It's possible we should be relying on cvs_temp_file to open
-     * the file here - we get race conditions otherwise.
-     */
-    fname = cvs_temp_name ();
   again:
-    if ((fp = CVS_FOPEN (fname, "w+")) == NULL)
+    if ((fp = cvs_temp_file (&fname)) == NULL)
 	error (1, 0, "cannot create temporary file %s", fname);
 
     if (*messagep)
diff -ur cvs-1.11.5.orig/src/sanity.sh cvs-1.11.5/src/sanity.sh
--- cvs-1.11.5.orig/src/sanity.sh	Thu Dec 19 15:38:32 2002
+++ cvs-1.11.5/src/sanity.sh	Sun Apr 27 13:25:26 2003
@@ -176,8 +176,8 @@
 #     So this would be lost if everything was `pwd`-based.  I suppose
 #     if we wanted to get baroque we could start making symlinks
 #     to ensure the two are different.
-tmp=`(cd /tmp; /bin/pwd || pwd) 2>/dev/null`
-: ${TMPDIR=$tmp}
+# tmp=`(cd /tmp; /bin/pwd || pwd) 2>/dev/null`
+# : ${TMPDIR=$tmp}
 
 # Now:
 #	1) Set TESTDIR if it's not set already
@@ -185,25 +185,31 @@
 #	3) Create $TESTDIR
 #	4) Normalize TESTDIR with `cd && (/bin/pwd || pwd)`
 #	   (This will match CVS output later)
-: ${TESTDIR=$tmp/cvs-sanity}
+if [ -n "$TESTDIR" ]; then
 # clean any old remnants (we need the chmod because some tests make
 # directories read-only)
-if test -d ${TESTDIR}; then
-    chmod -R a+wx ${TESTDIR}
-    rm -rf ${TESTDIR}
-fi
+  if test -d $TESTDIR; then
+    chmod -R u+wx $TESTDIR
+    rm -rf -- $TESTDIR
+  fi
 # These exits are important.  The first time I tried this, if the `mkdir && cd`
 # failed then the build directory would get blown away.  Some people probably
 # wouldn't appreciate that.
-mkdir ${TESTDIR} || exit 1
-cd ${TESTDIR} || exit 1
-TESTDIR=`(/bin/pwd || pwd) 2>/dev/null`
+  mkdir $TESTDIR || exit 1
+  cd $TESTDIR || exit 1
+  TESTDIR=`(/bin/pwd || pwd) 2>/dev/null`
+else
+  TESTDIR=
+  trap 'cd "$TESTDIR/.."; rm -rf -- "$TESTDIR"' EXIT
+  trap 'trap - EXIT; cd "$TESTDIR/.."; rm -rf -- "$TESTDIR"; exit 1' HUP INT QUIT TERM
+  TESTDIR="`mktemp -td cvs-sanity.XXXXXXXXXX`" || exit
+fi
 # Ensure $TESTDIR is absolute
-if test -z "${TESTDIR}" || echo "${TESTDIR}" |grep '^[^/]'; then
+if test -z "$TESTDIR" || echo "$TESTDIR" |grep '^[^/]'; then
     echo "Unable to resolve TESTDIR to an absolute directory." >&2
     exit 1
 fi
-cd ${TESTDIR}
+cd $TESTDIR || exit 1
 
 # Make sure various tools work the way we expect, or try to find
 # versions that do.
@@ -732,14 +738,16 @@
 	DIR_2=$2
 
 	cd $DIR_1
-	find . -print | fgrep -v /CVS | sort > /tmp/dc$$d1
+	dcd1="$TESTDIR/dcd1"
+	find . -print | fgrep -v /CVS | sort > "$dcd1"
 
 	# go back where we were to avoid symlink hell...
 	cd $OLDPWD
 	cd $DIR_2
-	find . -print | fgrep -v /CVS | sort > /tmp/dc$$d2
+        dcd2="$TESTDIR/dcd2"
+	find . -print | fgrep -v /CVS | sort > "$dcd2"
 
-	if diff /tmp/dc$$d1 /tmp/dc$$d2 >/dev/null 2>&1
+	if diff "$dcd1" "$dcd2" >/dev/null 2>&1
 	then
 		:
 	else
@@ -754,8 +762,8 @@
 				return 1
 			fi
 		fi
-	done < /tmp/dc$$d1
-	rm -f /tmp/dc$$*
+	done < "$dcd1"
+	rm -f "$dcd1" "$dcd2"
 	return 0
 }
 
@@ -24057,7 +24065,7 @@
 fi
 
 # Remove the test directory, but first change out of it.
-cd `dirname ${TESTDIR}`
-rm -rf ${TESTDIR}
+cd "$TESTDIR/.."
+rm -rf $TESTDIR
 
 # end of sanity.sh
