--- cvs-1.11.23/doc/cvs.texinfo
+++ cvs-1.11.23/doc/cvs.texinfo
@@ -99,11 +99,11 @@ approved by the Free Software Foundation.
 
 @dircategory GNU Packages
 @direntry
-* CVS: (cvs).                   Concurrent Versions System
+* CVS: (cvs).                                   Concurrent Versions System
 @end direntry
 @dircategory Individual utilities
 @direntry
-* cvs: (cvs)CVS commands.       Concurrent Versions System
+* cvs: (cvs)CVS commands.                       Concurrent Versions System
 @end direntry
 
 @comment The titlepage section does not appear in the Info file.
--- cvs-1.11.23/doc/cvsclient.texi
+++ cvs-1.11.23/doc/cvsclient.texi
@@ -5,7 +5,7 @@
 
 @dircategory Programming
 @direntry
-* cvsclient: (cvsclient).      The CVS client/server protocol.
+* cvsclient: (cvsclient).                       The CVS client/server protocol
 @end direntry
 
 @node Top
