--- cvs-1.11.23/Makefile.am
+++ cvs-1.11.23/Makefile.am
@@ -17,10 +17,9 @@
 ## Subdirectories to run make in for the primary targets.
 # Unix source subdirs, where we'll want to run lint and etags:
 # This is a legacy variable from b4 Automake
-USOURCE_SUBDIRS = lib zlib diff src
+USOURCE_SUBDIRS = lib diff src
 # All other subdirs:
-SUBDIRS = $(USOURCE_SUBDIRS) man doc contrib tools \
-	windows-NT os2 emx vms
+SUBDIRS = $(USOURCE_SUBDIRS) man doc contrib tools
 
 EXTRA_DIST = \
 	.cvsignore \
--- cvs-1.11.23/configure.in
+++ cvs-1.11.23/configure.in
@@ -1135,8 +1135,6 @@ AC_CONFIG_FILES(contrib/rcslock, [chmod +x contrib/rcslock])
 AC_CONFIG_FILES(contrib/sccs2rcs, [chmod +x contrib/sccs2rcs])
 AC_CONFIG_FILES(doc/mkman:doc/mkman.pl, [chmod +x doc/mkman])
 AC_CONFIG_FILES(src/cvsbug, [chmod +x src/cvsbug])
-AC_CONFIG_FILES(windows-NT/fix-msvc-mak:windows-NT/fix-msvc-mak-head.pl:windows-NT/fix-msvc-mak.pl,
-                [chmod +x windows-NT/fix-msvc-mak])
 
 dnl the bulk files
 AC_CONFIG_FILES([Makefile \
@@ -1144,16 +1142,11 @@ AC_CONFIG_FILES([Makefile \
 	  cvs.spec \
 	  diff/Makefile \
 	  doc/Makefile \
-	  emx/Makefile \
 	  lib/Makefile \
 	  man/Makefile \
-	  os2/Makefile \
 	  src/Makefile \
 	  tools/Makefile \
-	  vms/Makefile \
-	  windows-NT/Makefile \
-	  windows-NT/SCC/Makefile \
-	  zlib/Makefile])
+	  ])
 
 dnl and we're done
 AC_OUTPUT
--- cvs-1.11.23/lib/Makefile.am
+++ cvs-1.11.23/lib/Makefile.am
@@ -46,21 +46,16 @@ libcvs_a_SOURCES = \
 	argmatch.c \
 	getdate.y \
 	getline.c \
-	getopt.c \
-	getopt1.c \
 	getpass.c \
 	md5.c \
-	regex.c \
 	savecwd.c \
 	sighandle.c \
 	stripslash.c \
 	xgetwd.c \
 	yesno.c \
 	getline.h \
-	getopt.h \
 	getpagesize.h \
 	md5.h \
-	regex.h \
 	savecwd.h \
 	system.h \
 	wait.h \
--- cvs-1.11.23/src/Makefile.am
+++ cvs-1.11.23/src/Makefile.am
@@ -24,7 +24,7 @@ SHELL = /bin/sh
 # try and remove naming ocnflicts and fix Automake to allow particular includes
 # to be attached only to particular object files.  Short term fix is either or.
 ##INCLUDES = -I. -I.. -I$(srcdir) -I$(top_srcdir)/lib
-INCLUDES = -I$(top_srcdir)/lib -I$(top_srcdir)/diff -I$(top_srcdir)/zlib $(includeopt)
+INCLUDES = -I$(top_srcdir)/lib -I$(top_srcdir)/diff $(includeopt)
 
 bin_PROGRAMS = cvs
 bin_SCRIPTS = cvsbug
@@ -105,7 +105,7 @@ cvs_SOURCES = \
 cvs_LDADD = \
 	../diff/libdiff.a \
 	../lib/libcvs.a \
-	../zlib/libz.a
+	-lz
 
 # extra clean targets
 # wish this could be distclean-hdr-local but it's not part of automake
--- cvs-1.11.23/src/zlib.c
+++ cvs-1.11.23/src/zlib.c
@@ -22,7 +22,7 @@
 
 #if defined (SERVER_SUPPORT) || defined (CLIENT_SUPPORT)
 
-#include "zlib.h"
+#include <zlib.h>
 
 /* OS/2 doesn't have EIO.  FIXME: this whole notion of turning
    a different error into EIO strikes me as pretty dubious.  */
