diff -ur procmail-3.22~/config.h procmail-3.22/config.h
--- procmail-3.22~/config.h	Sun Oct 14 01:24:08 2001
+++ procmail-3.22/config.h	Sun Oct 14 01:26:45 2001
@@ -33,8 +33,8 @@
  * override those settings you should uncomment and possibly change the
  * DEFSPATH and DEFPATH defines below
  */
-/*#define DEFSPATH	"PATH=/bin:/usr/bin"			/* */
-/*#define DEFPATH	"PATH=$HOME/bin:/bin:/usr/bin"		/* */
+#define defSPATH	"PATH=/bin:/usr/bin:/usr/local/bin"
+#define defPATH	"PATH=/bin:/usr/bin:/usr/local/bin:$HOME/bin"
 
 /* every environment variable appearing in PRESTENV will be set or wiped
  * out of the environment (variables without an '=' sign will be thrown
@@ -74,12 +74,11 @@
    list of strings;  if unauthorised users use the -f option, an extra
    >From_ field will be added in the header */
 
-#define TRUSTED_IDS	{"root","daemon","uucp","mail","x400","network",\
-			 "list","slist","lists","news",0}
+#define TRUSTED_IDS	{"root","mail",0}
 
 /*#define NO_fcntl_LOCK		/* uncomment any of these three if you	     */
-/*#define NO_lockf_LOCK		/* definitely do not want procmail to make   */
-/*#define NO_flock_LOCK		/* use of those kernel-locking methods	     */
+#define NO_lockf_LOCK		/* definitely do not want procmail to make   */
+#define NO_flock_LOCK		/* use of those kernel-locking methods	     */
 				/* If you set LOCKINGTEST to a binary number
 	than there's no need to set these.  These #defines are only useful
 	if you want to disable particular locking styles but are unsure which
@@ -91,14 +90,14 @@
 	restriction does not apply to the /etc/procmailrc and
 	/etc/procmailrcs files) */
 
-/*#define NO_NFS_ATIME_HACK	/* uncomment if you're definitely not using
+#define NO_NFS_ATIME_HACK	/* uncomment if you're definitely not using
 				   NFS mounted filesystems and can't afford
 	procmail to sleep for 1 sec. before writing to an empty regular
 	mailbox.  This lets programs correctly judge whether there is unread
 	mail present.  procmail automatically suppresses this when it isn't
 	needed or under heavy load. */
 
-/*#define DEFsendmail	"/usr/sbin/sendmail"	/* uncomment and/or change if
+#define DEFsendmail	"/usr/sbin/sendmail"	/* uncomment and/or change if
 						   the autoconfigured default
 	SENDMAIL is not suitable.  This program should quack like a sendmail:
 	it should accept the -oi flag (to tell it to _not_ treat a line
@@ -222,7 +221,7 @@
 
 #define BinSh		"/bin/sh"
 #define ROOT_DIR	"/"
-#define DEAD_LETTER	"/tmp/dead.letter"    /* $ORGMAIL if no passwd entry */
+#define DEAD_LETTER	"/var/run/dead.letter"    /* $ORGMAIL if no passwd entry */
 #define DevNull		"/dev/null"
 #define NICE_RANGE	39			  /* maximal nice difference */
 #define chCURDIR	'.'			    /* the current directory */
