diff -ur procmail-3.15.2.orig/man/mansed procmail-3.15.2/man/mansed
--- procmail-3.15.2.orig/man/mansed	Tue May  9 10:36:35 2000
+++ procmail-3.15.2/man/mansed	Fri Oct  4 11:45:14 2002
@@ -19,13 +19,19 @@
 RM="$4"
 DEVNULL=$5
 export SHELL SRC DEST RM DEVNULL
-TMPF0=/tmp/_mansed.0.$$
-TMPF1=/tmp/_mansed.1.$$
+TMPF0="`mktemp -t mansed.XXXXXXXXXX`" || exit 1
+TMPF1="`mktemp -t mansed.XXXXXXXXXX`" || {
+ $RM "$TMPF0"
+ exit 1
+}
 
 if test ! -f "$DEST"
 then
- trap "$RM \"$DEST\" $TMPF0 $TMPF1;exit 1" 1 2 3 15
+ trap '$RM "$DEST" "$TMPF0" "$TMPF1"; exit 1' HUP INT PIPE TERM
+else
+ trap '$RM "$TMPF0" "$TMPF1"; exit 1' HUP INT PIPE TERM
 fi
+trap '$RM "$TMPF0" "$TMPF1"' EXIT
 
 (cat <<\HERE
 .\"if n .pl +(135i-\n(.pu)
@@ -120,11 +126,10 @@
 .rm RE
 .rn Re RE
 HERE
- )| sed -f man.sed.0000 >$TMPF0
+ )| sed -f man.sed.0000 > "$TMPF0"
 
 if test $? != 0
 then
-  $RM $TMPF0
   exit 1
 fi
 
@@ -133,19 +138,16 @@
   case $a in
      man.sed.0000) ;;
      man.sed.????)
-	mv $TMPF0 $TMPF1
-	if sed -f $a <$TMPF1 >$TMPF0
+	if sed -f $a < "$TMPF0" > "$TMPF1"
 	then
-	:
+	   TMPF="$TMPF0"
+	   TMPF0="$TMPF1"
+	   TMPF1="$TMPF"
 	else
-	   $RM $TMPF1 $TMPF0
 	   exit 1
 	fi
   esac
 done
 
-cat $TMPF0 >"$DEST"
-result=$?
-
-$RM $TMPF0 $TMPF1
-exit $result
+cat "$TMPF0" > "$DEST"
+touch -r "$SRC" "$DEST"
diff -ur procmail-3.15.2.orig/src/autoconf procmail-3.15.2/src/autoconf
--- procmail-3.15.2.orig/src/autoconf	Mon Jul 16 12:08:13 2001
+++ procmail-3.15.2/src/autoconf	Tue Aug 13 01:53:37 2002
@@ -195,7 +195,7 @@
 ==============================================================================
 HERE
 
-testdirs="/tmp ."
+testdirs="."
 
 case "$LOCKINGTEST" in
   ""|*[a-zA-Z/]*) b=dummy ;;
@@ -1458,7 +1458,7 @@
 grep '^#define SENDMAIL ".*/sendmail"' $ACONF >$DEVNULL ||
  echo "#define DEFflagsendmail \"\"" >>$ACONF
 
-a=/tmp/_chowntst.$$
+a=_chowntst.$$
 $RM -r $a
 OLDTESTDIRS="$a $OLDTESTDIRS"
 mkdir $a
@@ -1466,13 +1466,18 @@
 
 _autotst $a/__ $a/__/__ 4 >>$ACONF
 
-cat /usr/lib/sendmail.cf /etc/sendmail.cf /etc/mail/sendmail.cf 2>$DEVNULL |
- grep 'Mlocal.*procmail' >$DEVNULL ||
- echo '#define CF_no_procmail_yet' >>$ACONF
-
-cat /usr/lib/sendmail.cf /etc/sendmail.cf /etc/mail/sendmail.cf 2>$DEVNULL |
- grep '^V' >$DEVNULL ||
- echo '#define buggy_SENDMAIL' >>$ACONF
+# Skip this check on systems without sendmail (possibly with another MTA)
+if test -r /usr/lib/sendmail.cf -o \
+	-r /etc/sendmail.cf -o \
+	-r /etc/mail/sendmail.cf; then
+  cat /usr/lib/sendmail.cf /etc/sendmail.cf /etc/mail/sendmail.cf 2>$DEVNULL |
+   grep 'Mlocal.*procmail' >$DEVNULL ||
+   echo '#define CF_no_procmail_yet' >>$ACONF
+
+  cat /usr/lib/sendmail.cf /etc/sendmail.cf /etc/mail/sendmail.cf 2>$DEVNULL |
+   grep '^V' >$DEVNULL ||
+   echo '#define buggy_SENDMAIL' >>$ACONF
+fi
 
 lpath='/bin'
 bins="/bin"
diff -ur procmail-3.15.2.orig/src/exopen.c procmail-3.15.2/src/exopen.c
--- procmail-3.15.2.orig/src/exopen.c	Sun Jul 15 13:55:22 2001
+++ procmail-3.15.2/src/exopen.c	Tue Aug 13 01:33:34 2002
@@ -22,13 +22,14 @@
  const size_t len;const mode_t mode;const int verbos,chownit;
 { static const char s2c[]=".,+%";static int serial=STRLEN(s2c);
   static time_t t;char*dot,*end=full+len,*op,*ldp;struct stat filebuf;
-  int nicediff,i,didnice,retry=RETRYunique;
+  int i,didnice,retry=RETRYunique;
+  didnice=0;
   if(chownit&doCHOWN)		  /* semi-critical, try raising the priority */
-   { nicediff=nice(0);SETerrno(0);nicediff-=nice(-NICE_RANGE);
+   { nice(0);SETerrno(0);nice(-NICE_RANGE);
      didnice=!errno;
    }
   *(end=len?full+len-1:p+UNIQnamelen-1)='\0';
-  *(op=p)=UNIQ_PREFIX;dot=ultoan((long)thepid,p+1);
+  *(ldp=op=p)=UNIQ_PREFIX;dot=ultoan((long)thepid,p+1);
   if(serial<STRLEN(s2c))
      goto in;
   do
@@ -65,8 +66,9 @@
   while((!i||errno!=ENOENT||	      /* casually check if it already exists */
 	 (0>(i=ropen(full,O_WRONLY|O_CREAT|O_EXCL,mode))&&errno==EEXIST))&&
 	(i= -1,retry--));
-  if(chownit&doCHOWN&&didnice)
-     nice(nicediff);		   /* put back the priority to the old level */
+  if(didnice) {
+     nice(NICE_RANGE/2);nice(0);   /* put back the priority to the old level */
+  }
   if(i<0)
    { if(verbos)			      /* this error message can be confusing */
 	writeerr(full);					 /* for casual users */
diff -ur procmail-3.15.2.orig/src/fields.c procmail-3.15.2/src/fields.c
--- procmail-3.15.2.orig/src/fields.c	Sat Jan 29 09:51:36 2000
+++ procmail-3.15.2/src/fields.c	Tue Aug 13 00:51:27 2002
@@ -30,7 +30,7 @@
 }
 
 void cleanheader P((void))		  /* zorch whitespace before the ':' */
-{ struct field**pp,*p;char*cp;int idlen;
+{ struct field**pp,*p;char*cp;
   for(pp=&rdheader;p= *pp;pp= &(*pp)->fld_next)
      if((cp=p->fld_text+p->id_len-1,*cp==HEAD_DELIMITER)&&	    /* has : */
 	(*--cp==' '||*cp=='\t'))				   /* has ws */
diff -ur procmail-3.15.2.orig/src/foldinfo.c procmail-3.15.2/src/foldinfo.c
--- procmail-3.15.2.orig/src/foldinfo.c	Sun Jul 15 13:55:24 2001
+++ procmail-3.15.2/src/foldinfo.c	Tue Aug 13 01:02:56 2002
@@ -11,10 +11,13 @@
  "$Id: foldinfo.c,v 1.1.2.3 2001/07/15 09:27:15 guenther Exp $";
 #endif
 #include "procmail.h"
+#include "acommon.h"
 #include "misc.h"
 #include "lastdirsep.h"
 #include "robust.h"
 #include "exopen.h"
+#include "goodies.h"
+#include "locking.h"
 #include "foldinfo.h"
 
 static const char
@@ -98,8 +101,9 @@
 static int mkmaildir(buffer,chp,paranoid)char*const buffer,*const chp;
  const int paranoid;
 { mode_t mode;int i;
+  i=chp-buffer+1;
   if(paranoid)
-     memcpy(buf2,buffer,i=chp-buffer+1),buf2[i-1]= *MCDIRSEP_,buf2[i]='\0';
+     memcpy(buf2,buffer,i),buf2[i-1]= *MCDIRSEP_,buf2[i]='\0';
   return
    (strcpy(chp,maildirnew),mode=trymkdir(buffer,paranoid,i),S_ISDIR(mode))&&
    (strcpy(chp,maildircur),mode=trymkdir(buffer,paranoid,i),S_ISDIR(mode))&&
@@ -109,6 +113,7 @@
 int foldertype(type,forcedir,modep,paranoid)int type,forcedir;
  mode_t*const modep;struct stat*const paranoid;
 { struct stat stbuf;mode_t mode;int i;char*chp;
+  mode=0; /* XXX: this appears to be actually used (type==ft_FILE&&forcedir) */
   if(!type)
      type=folderparse();
   switch(type)
diff -ur procmail-3.15.2.orig/src/formail.c procmail-3.15.2/src/formail.c
--- procmail-3.15.2.orig/src/formail.c	Sun Jul 15 13:55:25 2001
+++ procmail-3.15.2/src/formail.c	Tue Aug 13 01:04:41 2002
@@ -11,8 +11,8 @@
 #ifdef RCS
 static /*const*/char rcsid[]=
  "$Id: formail.c,v 1.97.2.5 2001/07/15 09:27:16 guenther Exp $";
-#endif
 static /*const*/char rcsdate[]="$Date: 2001/07/15 09:27:16 $";
+#endif
 #include "includes.h"
 #include <ctype.h>		/* iscntrl() */
 #include "formail.h"
@@ -307,6 +307,7 @@
  FILE*idcache;const off_t maxlen;const int split;
 { int dupid=0;char*key,*oldnewl;
   key=(char*)namep;		  /* not to worry, no change will be noticed */
+  oldnewl=NULL;
   if(!areply)
    { key=0;
      if(msid->rexl)					/* any Message-ID: ? */
@@ -364,9 +365,9 @@
 
 int main(lastm,argv)int lastm;const char*const argv[];
 { int i,split=0,force=0,bogus=1,every=0,headreply=0,digest=0,nowait=0,keepb=0,
-   minfields=(char*)progid-(char*)progid,conctenate=0,babyl=0,babylstart,
+   minfields=(char*)progid-(char*)progid,conctenate=0,babyl=0,babylstart=0,
    berkeley=0,forgetclen;
-  off_t maxlen,ctlength;FILE*idcache=0;pid_t thepid;
+  off_t maxlen=0,ctlength;FILE*idcache=NULL;pid_t thepid;
   size_t j,lnl,escaplen;char*chp,*namep,*escap=ESCAP;
   struct field*fldp,*fp2,**afldp,*fdate,*fcntlength,*fsubject,*fFrom_;
   if(lastm)			       /* sanity check, any argument at all? */
@@ -785,7 +786,10 @@
 	   else if(split&&digest&&(lnl||every)&&digheadr())	  /* digest? */
 accuhdr:    { for(i=minfields;--i&&readhead()&&digheadr();); /* found enough */
 	      if(!i)					   /* then split it! */
-splitit:       { if(!lnl)   /* did the previous mail end with an empty line? */
+#ifdef MAILBOX_SEPARATOR
+splitit:
+#endif
+	       { if(!lnl)   /* did the previous mail end with an empty line? */
 		    lputcs('\n');		      /* but now it does :-) */
 		 logfolder();
 		 if(fclose(mystdout)==EOF||errout==EOF)
diff -ur procmail-3.15.2.orig/src/lockfile.c procmail-3.15.2/src/lockfile.c
--- procmail-3.15.2.orig/src/lockfile.c	Sun Jul 15 13:55:29 2001
+++ procmail-3.15.2/src/lockfile.c	Tue Aug 13 01:06:28 2002
@@ -16,8 +16,8 @@
 #ifdef RCS
 static /*const*/char rcsid[]=
  "$Id: lockfile.c,v 1.43.2.2 2001/07/15 09:27:20 guenther Exp $";
-#endif
 static /*const*/char rcsdate[]="$Date: 2001/07/15 09:27:20 $";
+#endif
 #include "includes.h"
 #include "sublib.h"
 #include "exopen.h"
diff -ur procmail-3.15.2.orig/src/misc.c procmail-3.15.2/src/misc.c
--- procmail-3.15.2.orig/src/misc.c	Mon Jul 16 12:08:14 2001
+++ procmail-3.15.2/src/misc.c	Tue Aug 13 00:51:27 2002
@@ -672,6 +672,7 @@
       if(i)					 /* check out all conditions */
        { int negate,scoreany;double weight,xponent,lscore;
 	 char*lstartchar=startchar;long ltobesent=tobesent,sizecheck=filled;
+	 weight=xponent=0;
 	 for(chp=strchr(buf2,'\0');--chp>=buf2;)
 	  { switch(*chp)		  /* strip off whitespace at the end */
 	     { case ' ':case '\t':*chp='\0';
diff -ur procmail-3.15.2.orig/src/pipes.c procmail-3.15.2/src/pipes.c
--- procmail-3.15.2.orig/src/pipes.c	Sun Jul 15 13:55:36 2001
+++ procmail-3.15.2/src/pipes.c	Tue Aug 13 00:51:27 2002
@@ -137,6 +137,7 @@
 
 int pipthrough(line,source,len)char*line,*source;const long len;
 { int pinfd[2],poutfd[2];char*eq;
+  eq=NULL;
   if(Stdout&&(*(eq=strchr(Stdout,'\0')-1)='\0',		     /* chop the '=' */
    !(backblock=getenv(Stdout))))			/* no current value? */
      PRDB=PWRB= -1;
@@ -177,7 +178,7 @@
   rclose(PWRB);rclose(PWRI);getstdin(PRDI);
   if(forkerr(pidchild,procmailn))
      return -1;
-  if(Stdout)
+  if(eq)
    { char*name;
      *eq='=';name=Stdout;Stdout=0;primeStdout(name);free(name);
      Stdout=readdyn(Stdout,&Stdfilled);
@@ -238,7 +239,9 @@
       }
    }
   pidchild=0;
+#if 0
 builtin:
+#endif
   if(!sh)
      concatenate(line);
   if(asgnlastf)
diff -ur procmail-3.15.2.orig/src/procmail.c procmail-3.15.2/src/procmail.c
--- procmail-3.15.2.orig/src/procmail.c	Mon Jul 16 08:31:04 2001
+++ procmail-3.15.2/src/procmail.c	Tue Aug 13 02:13:36 2002
@@ -101,6 +101,7 @@
   ;{ int presenviron,Deliverymode,override;char*fromwhom=0;
      const char*idhint=0;gid_t egid=getegid();
      presenviron=Deliverymode=mailfilter=override=0;
+     chp=chp2=NULL;
      Openlog(procmailn,LOG_PID,LOG_MAIL);		  /* for the syslogd */
      if(argc)			       /* sanity check, any argument at all? */
       { Deliverymode=strncmp(lastdirsep(argv0=argv[0]),procmailn,
@@ -504,6 +505,7 @@
       }
    }
   ;{ int lastsucc,lastcond,prevcond;
+     lastsucc=lastcond=prevcond=0;
      if(etcrc)		  /* do we start with an /etc/procmailrc file first? */
       { if(0<=bopen(etcrc))
 	 { yell(drcfile,etcrc);
@@ -594,7 +596,10 @@
 	      if(((stbuf.st_uid!=uid&&stbuf.st_uid!=ROOT_uid||	/* check uid */
 		   stbuf.st_mode&S_IWOTH||	       /* writable by others */
 		   i&&stbuf.st_mode&S_IWGRP		/* writable by group */
-		    &&(NO_CHECK_stgid||stbuf.st_gid!=gid)
+		    &&(NO_CHECK_stgid||stbuf.st_gid!=gid)||
+		   !S_ISREG(stbuf.st_mode)||
+		   stbuf.st_size>stbuf.st_blocks*stbuf.st_blksize
+		    &&stbuf.st_blksize>=512
 		  )&&strcmp(devnull,buf)||    /* /dev/null is a special case */
 		 (*chp='\0',stat(buf,&stbuf))||	      /* check the directory */
 #ifndef CAN_chown				   /* sticky and can't chown */
diff -ur procmail-3.15.2.orig/src/regexp.c procmail-3.15.2/src/regexp.c
--- procmail-3.15.2.orig/src/regexp.c	Mon Apr  5 21:35:06 1999
+++ procmail-3.15.2/src/regexp.c	Tue Aug 13 00:51:27 2002
@@ -126,6 +126,7 @@
 	return;
      case R_BEG_CLASS:					   /* a simple class */
       { unsigned i,j=R_NOT_CLASS==*++p;
+	i=0;
 	if(e)
 	 { r->opc=OPC_CLASS;r->next=Ceps e;Cc(r,pos1.st_)=Cc(r,pos2.st_)=0;
 	   i=maxindex(rAc);
@@ -269,6 +270,7 @@
 
 static int por(e)const struct eps*const e;
 { uchar*pvold;struct eps*rvold;
+  pvold=NULL;rvold=NULL;
   if(!e)
    { rvold=r;
      if(cachea==(pvold=p))
@@ -509,7 +511,7 @@
    }
   while(--len);					     /* still text to search */
   goto wrapup;
-  ;{ const char*start,*bom;
+  ;{ const char*start,*bom=NULL;
      do
       { i= *++str;			 /* get the next real-text character */
 	if(i-'A'<='Z'-'A')
diff -ur procmail-3.15.2.orig/src/robust.c procmail-3.15.2/src/robust.c
--- procmail-3.15.2.orig/src/robust.c	Sun Jul 15 13:55:40 2001
+++ procmail-3.15.2/src/robust.c	Tue Aug 13 01:16:48 2002
@@ -20,7 +20,9 @@
 #define nomemretry	noresretry
 #define noforkretry	noresretry
 		       /* set nextexit to prevent elog() from using malloc() */
-void nomemerr(len)const size_t len;
+void
+__attribute__ ((noreturn))
+nomemerr(len)const size_t len;
 { static const char outofmem[]="Out of memory";
   nextexit=2;nlog(outofmem);elog("\n");
   syslog(LOG_NOTICE,"%s as I tried to allocate %ld bytes\n",outofmem,
@@ -130,7 +132,8 @@
 int ropen(name,mode,mask)const char*const name;const int mode;
  const mode_t mask;
 { int i,r;					       /* a SysV secure open */
-  for(r=noresretry,lcking|=lck_FILDES;0>(i=open(name,mode,mask));)
+  for(r=noresretry,lcking|=lck_FILDES;
+   0>(i=open(name,mode|O_NOCTTY|O_NONBLOCK,mask));)
      if(errno!=EINTR&&!(errno==ENFILE&&(r<0||r--)))
 	break;		 /* survives a temporary "file table full" condition */
   lcking&=~lck_FILDES;
