--- kbd-1.08/src/setfont.c.orig	Fri Oct 11 14:09:00 2002
+++ kbd-1.08/src/setfont.c	Fri Mar 28 00:16:33 2003
@@ -101,7 +101,7 @@
 int
 main(int argc, char *argv[]) {
 	char *ifiles[MAXIFILES];
-	char *mfil, *ufil, *Ofil, *ofil, *omfil, *oufil;
+	char *mfil, *ufil, *Ofil, *ofil, *omfil, *oufil, *tty;
 	int ifilct = 0, fd, i, iunit, hwunit, no_m, no_u;
 	int restore = 0;
 
@@ -111,9 +111,7 @@
 	bindtextdomain(PACKAGE, LOCALEDIR);
 	textdomain(PACKAGE);
 
-	fd = getfd();
-
-	ifiles[0] = mfil = ufil = Ofil = ofil = omfil = oufil = 0;
+	ifiles[0] = mfil = ufil = Ofil = ofil = omfil = oufil = tty = NULL;
 	iunit = hwunit = 0;
 	no_m = no_u = 0;
 
@@ -160,6 +158,8 @@
 		hwunit = atoi(argv[i]+2);
 		if (hwunit <= 0 || hwunit > 32)
 		  usage();
+	    } else if (!strncmp(argv[i], "--tty", 5)) {
+		tty = argv[i] + 6;
 	    } else if (argv[i][0] == '-') {
 		iunit = atoi(argv[i]+1);
 		if(iunit <= 0 || iunit > 32)
@@ -171,6 +171,15 @@
 		}
 		ifiles[ifilct++] = argv[i];
 	    }
+	}
+
+	if (tty) {
+	    if ((fd = open(tty, O_RDWR)) == -1) {
+		perror("open");
+		exit(EX_OSERR);
+	    }
+	} else {
+	    fd = getfd();
 	}
 
 	if (ifilct && restore) {
