--- kbd-1.08/src/Makefile.in.orig	Sun Oct 13 04:34:10 2002
+++ kbd-1.08/src/Makefile.in	Fri Mar 28 00:00:47 2003
@@ -63,14 +63,14 @@
 
 install:	all
 	install -d -m 755 $(BINDIR) $(LOADKEYS_BINDIR)
-	install -s -m 0755 -o root $(PROGS) $(OLDPROGS) $(BINDIR)
-#	install -s -m 0755 -o root $(MISC) $(BINDIR)
-	install -c -m 0755 -o root $(SHCMDS) $(BINDIR)
+	install -m 0755 $(PROGS) $(OLDPROGS) $(BINDIR)
+#	install -m 0755 $(MISC) $(BINDIR)
+	install -m 0755 $(SHCMDS) $(BINDIR)
 	for i in psfaddtable psfgettable psfstriptable; do \
 		rm -f $(BINDIR)/$$i; ln -s psfxtable $(BINDIR)/$$i; \
 	done
 	rm -f $(BINDIR)/loadkeys
-	install -s -m 0755 -o root loadkeys $(LOADKEYS_BINDIR)
+	install -m 0755 loadkeys $(LOADKEYS_BINDIR)
 	@echo "You may also want to add psf.magic to /usr/lib/magic"
 
 
