diff -urp cdrkit-1.1.9.orig/genisoimage/eltorito.c cdrkit-1.1.9/genisoimage/eltorito.c
--- cdrkit-1.1.9.orig/genisoimage/eltorito.c	2008-02-26 07:59:05 +0000
+++ cdrkit-1.1.9/genisoimage/eltorito.c	2009-05-09 03:52:45 +0000
@@ -286,7 +286,7 @@ get_torito_desc(struct eltorito_boot_des
 
 	memset(boot_desc, 0, sizeof (*boot_desc));
 	boot_desc->type[0] = 0;
-	memcpy(boot_desc->id, ISO_STANDARD_ID, sizeof (ISO_STANDARD_ID));
+	memcpy(boot_desc->id, ISO_STANDARD_ID, sizeof (ISO_STANDARD_ID) - 1);
 	boot_desc->version[0] = 1;
 
 	memcpy(boot_desc->system_id, EL_TORITO_ID, sizeof (EL_TORITO_ID));
diff -urp cdrkit-1.1.9.orig/genisoimage/write.c cdrkit-1.1.9/genisoimage/write.c
--- cdrkit-1.1.9.orig/genisoimage/write.c	2008-02-27 08:54:08 +0000
+++ cdrkit-1.1.9/genisoimage/write.c	2009-05-09 03:51:26 +0000
@@ -1912,7 +1912,7 @@ pvd_write(FILE *outfile)
 	/* Next we write out the primary descriptor for the disc */
 	memset(&vol_desc, 0, sizeof (vol_desc));
 	vol_desc.type[0] = ISO_VD_PRIMARY;
-	memcpy(vol_desc.id, ISO_STANDARD_ID, sizeof (ISO_STANDARD_ID));
+	memcpy(vol_desc.id, ISO_STANDARD_ID, sizeof (ISO_STANDARD_ID) - 1);
 	vol_desc.version[0] = 1;
 
 	memset(vol_desc.system_id, ' ', sizeof (vol_desc.system_id));
@@ -2035,7 +2035,7 @@ evd_write(FILE *outfile)
 	 */
 	memset(&evol_desc, 0, sizeof (evol_desc));
 	evol_desc.type[0] = (unsigned char) ISO_VD_END;
-	memcpy(evol_desc.id, ISO_STANDARD_ID, sizeof (ISO_STANDARD_ID));
+	memcpy(evol_desc.id, ISO_STANDARD_ID, sizeof (ISO_STANDARD_ID) - 1);
 	evol_desc.version[0] = 1;
 	jtwrite(&evol_desc, SECTOR_SIZE, 1, 0, TRUE);
 	xfwrite(&evol_desc, SECTOR_SIZE, 1, outfile, 0, TRUE);
