diff -urp cdrkit-1.1.9.orig/genisoimage/genisoimage.c cdrkit-1.1.9/genisoimage/genisoimage.c
--- cdrkit-1.1.9.orig/genisoimage/genisoimage.c	2008-05-25 22:12:37 +0000
+++ cdrkit-1.1.9/genisoimage/genisoimage.c	2009-05-09 04:01:49 +0000
@@ -987,10 +987,12 @@ read_rcfile(char *appname)
 		return;
 	if (read_one_rcfile(getenv("MKISOFSRC")))
 		return;
+#if 0
 	if (read_one_rcfile("." RCFILENAME))
 		return;
 	if (read_one_rcfile("." OLD_RCFILENAME))
 		return;
+#endif
 
 	p = getenv("HOME");
 	if (p && strlen(p) + 1 + sizeof(RCFILENAME) < sizeof(filename)) {
@@ -1000,6 +1002,7 @@ read_rcfile(char *appname)
 		strcpy(p, "." RCFILENAME);
 		if (read_one_rcfile(filename))
 			return;
+/* XXX: This assumes that OLD_RCFILENAME is not longer than RCFILENAME */
 		strcpy(p, "." OLD_RCFILENAME);
 		if (read_one_rcfile(filename))
 			return;
@@ -1008,6 +1011,7 @@ read_rcfile(char *appname)
 	if (read_one_rcfile(ETCDIR SPATH_SEPARATOR RCFILENAME))
 		return;
 
+#if 0
 	if (appname &&
 	    strlen(appname) + 1 + sizeof(RCFILENAME) < sizeof(filename)) {
 		strcpy(filename, appname);
@@ -1018,6 +1022,7 @@ read_rcfile(char *appname)
 				return;
 		}
 	}
+#endif
 }
 
 char	*path_table_l = NULL;
diff -urp cdrkit-1.1.9.orig/genisoimage/genisoimagerc.5 cdrkit-1.1.9/genisoimage/genisoimagerc.5
--- cdrkit-1.1.9.orig/genisoimage/genisoimagerc.5	2006-12-14 09:35:35 +0000
+++ cdrkit-1.1.9/genisoimage/genisoimagerc.5	2009-05-09 04:04:40 +0000
@@ -23,15 +23,9 @@ likewise for the
 environment variable.  Next,
 .B genisoimage
 looks for files named
-.IR .genisoimagerc " or " .mkisofsrc ,
-first in the current working directory, then in the user's home
-directory.  Next, it looks for
+.IR .genisoimagerc " or " .mkisofsrc
+in the user's home directory.  Finally, it looks for
 .IR /etc/genisoimagerc .
-Finally, it looks for a
-.I .genisoimagerc
-in the same directory as
-.B genisoimage
-itself is stored.
 .PP
 The
 .I .genisoimagerc
