diff -ur SimplePAMApps-0.60.orig/pamapps/passwd/passwd.c SimplePAMApps-0.60/pamapps/passwd/passwd.c
--- SimplePAMApps-0.60.orig/pamapps/passwd/passwd.c	Thu Feb 18 15:29:43 1999
+++ SimplePAMApps-0.60/pamapps/passwd/passwd.c	Sun Apr  1 19:03:58 2001
@@ -81,31 +81,27 @@
     }
 
     if (user == NULL) {
-#ifndef HAVE_PWDB
 	struct passwd *pwent;
-#endif /* HAVE_PWDB */
-	if ((user = getlogin()) == NULL) {
-	    fprintf(stderr, "passwd: cannot retrieve user's name\n");
-	    exit(1);
-	}
-#ifndef HAVE_PWDB
-	/* attempt to patch over libc's inability to handle longer that
-	   fixed length login names from the utmp file */
 
-	if ((pwent = getpwnam(user)) == NULL
-	    || (pwent = getpwuid(getuid())) == NULL
-	    || !(pwent->pw_name
-		 && !strncmp(pwent->pw_name, user, strlen(user)))) {
+	if ((user = getlogin())) {
+	    pwent = getpwnam(user);
+	    if (!pwent || pwent->pw_uid != getuid())
+		pwent = getpwuid(getuid());
+	} else
+	    pwent = getpwuid(getuid());
+
+	if (!pwent || !pwent->pw_name) {
 	    fprintf(stderr, "passwd: cannot retrieve user's name\n");
 	    exit(1);
-	} else {
-	    user = pwent->pw_name;
 	}
-#endif /* HAVE_PWDB */
+
+	user = pwent->pw_name;
     }
 
     /* here we know whose passwords are to be changed and whether
        we'll change everything or just the expired ones */
+
+    setlinebuf(stdout);
 
     D(("service=%s, user=%s\n", service, user));
     retval = pam_start(service, user, &conv, &pamh);
