diff -ur SimplePAMApps-0.60.orig/pamapps/su/su.c SimplePAMApps-0.60/pamapps/su/su.c
--- SimplePAMApps-0.60.orig/pamapps/su/su.c	Thu Feb 18 15:29:50 1999
+++ SimplePAMApps-0.60/pamapps/su/su.c	Sun Oct 29 00:07:35 2000
@@ -389,8 +389,6 @@
 	const char *ruser = getlogin();      /* Who is running this program? */
 	if (ruser) {
 	    retval = pam_set_item(pamh, PAM_RUSER, (const void *)ruser);
-	} else {
-	    retval = PAM_PERM_DENIED;             /* must be known to system */
 	}
 	ruser = NULL;
     }
