diff -ur SimplePAMApps-0.60.orig/pamapps/lib/wtmp-gate.c SimplePAMApps-0.60/pamapps/lib/wtmp-gate.c
--- SimplePAMApps-0.60.orig/pamapps/lib/wtmp-gate.c	Wed Nov 25 15:33:32 1998
+++ SimplePAMApps-0.60/pamapps/lib/wtmp-gate.c	Mon Nov 19 07:47:15 2001
@@ -35,6 +35,7 @@
         *err_descr = pam_strerror(pamh, retval);
         return -1;
     }
+    if (!terminal) terminal = "???";
     retval = pam_get_item(pamh, PAM_RHOST, (const void **)&rhost);
     if (retval != PAM_SUCCESS)
         rhost = NULL;
@@ -55,6 +56,7 @@
         *err_descr = pam_strerror(pamh, retval);
         return -1;
     }
+    if (!terminal) terminal = "???";
 
     return
         utmp_do_close_session(terminal, callname, err_descr);
diff -ur SimplePAMApps-0.60.orig/pamapps/su/su.c SimplePAMApps-0.60/pamapps/su/su.c
--- SimplePAMApps-0.60.orig/pamapps/su/su.c	Thu Feb 18 15:29:50 1999
+++ SimplePAMApps-0.60/pamapps/su/su.c	Mon Nov 19 07:48:08 2001
@@ -18,9 +18,10 @@
 #define DEFAULT_SHELL             "/bin/sh"
 #define SLEEP_TO_KILL_CHILDREN    3  /* seconds to wait after SIGTERM before
 					SIGKILL */
-#define SU_FAIL_DELAY     2000000    /* usec on authentication failure */
+#define SU_FAIL_DELAY             0  /* usec on authentication failure */
 
 #include <stdlib.h>
+#include <stdarg.h>
 #include <signal.h>
 #include <stdio.h>
 #include <sys/stat.h>
@@ -75,7 +76,7 @@
 /* ------ the application itself -------------- */
 /* -------------------------------------------- */
 
-void main(int argc, char *argv[])
+int main(int argc, char *argv[])
 {
     int retcode, is_login, status;
     int retval, final_retval; /* PAM_xxx return values */
@@ -293,7 +294,7 @@
     if (reset_terminal_modes() != 0 && !status)
 	status = 1;
 
-    exit(status);                 /* transparent exit */
+    return status;
 }
 
 /* -------------------------------------------- */
@@ -380,7 +381,7 @@
 	if (terminal) {
 	    retval = pam_set_item(pamh, PAM_TTY, (const void *)terminal);
 	} else {
-	    retval = PAM_PERM_DENIED;                /* how did we get here? */
+	    if (getuid() != 0) retval = PAM_PERM_DENIED;
 	}
 	terminal = NULL;
     }
@@ -389,8 +390,6 @@
 	const char *ruser = getlogin();      /* Who is running this program? */
 	if (ruser) {
 	    retval = pam_set_item(pamh, PAM_RUSER, (const void *)ruser);
-	} else {
-	    retval = PAM_PERM_DENIED;             /* must be known to system */
 	}
 	ruser = NULL;
     }
@@ -403,7 +402,7 @@
 	exit_now(1, "su: problem establishing environment\n");
     }
 
-#ifdef HAVE_PAM_FAIL_DELAY
+#if defined(HAVE_PAM_FAIL_DELAY) && SU_FAIL_DELAY
     /* have to pause on failure. At least this long (doubles..) */
     retval = pam_fail_delay(pamh, SU_FAIL_DELAY);
     if (retval != PAM_SUCCESS) {
