diff -uNrp lftp-4.0.10.orig/lib/glob.c lftp-4.0.10/lib/glob.c
--- lftp-4.0.10.orig/lib/glob.c	2010-09-04 12:47:58 +0000
+++ lftp-4.0.10/lib/glob.c	2010-09-04 13:18:57 +0000
@@ -20,6 +20,7 @@
 
 /* Don't use __attribute__ __nonnull__ in this compilation unit.  Otherwise gcc
    optimizes away the pattern == NULL || pglob == NULL tests below.  */
+#undef _GL_ARG_NONNULL
 #define _GL_ARG_NONNULL(params)
 
 #include <glob.h>
diff -uNrp lftp-4.0.10.orig/src/Torrent.cc lftp-4.0.10/src/Torrent.cc
--- lftp-4.0.10.orig/src/Torrent.cc	2010-09-04 12:47:58 +0000
+++ lftp-4.0.10/src/Torrent.cc	2010-09-04 13:18:31 +0000
@@ -998,7 +998,7 @@ void Torrent::ReducePeers()
 	    max_idle.toString(TimeInterval::TO_STR_TERSE+TimeInterval::TO_STR_TRANSLATE));
 	 peers.chop();
 	 if(max_idle<60)
-	    decline_timer.Set(60-max_idle);
+	    decline_timer.Set(60-(int)max_idle);
       }
    }
    peers.qsort(complete ? PeersCompareSendRate : PeersCompareRecvRate);
@@ -1104,8 +1104,8 @@ int Torrent::PeerBytesAllowed(const Torr
    int min_rate = 1000;
    // the more is the opposite rate the more rate allowed, with a minimum
    int bytes = rate_limit.BytesAllowed(dir);
-   bytes *= (peer_rate + min_rate);
-   bytes /= (rate + active_peers_count*min_rate);
+   bytes *= (int)(peer_rate + min_rate);
+   bytes /= (int)(rate + active_peers_count*min_rate);
    return bytes;
 }
 void Torrent::PeerBytesUsed(int b,RateLimit::dir_t dir)
diff -uNrp lftp-4.0.10.orig/src/lftp_rl.c lftp-4.0.10/src/lftp_rl.c
--- lftp-4.0.10.orig/src/lftp_rl.c	2010-09-04 12:47:58 +0000
+++ lftp-4.0.10/src/lftp_rl.c	2010-09-04 13:33:46 +0000
@@ -157,11 +157,11 @@ void lftp_rl_init(
    rl_attempted_completion_function  =attempted_completion_function;
    rl_getc_function                  =getc_function;
    rl_completer_quote_characters     =completer_quote_characters;
-   rl_completer_word_break_characters=completer_word_break_characters;
+   rl_completer_word_break_characters=(char*)completer_word_break_characters;
    rl_filename_quote_characters      =filename_quote_characters;
    rl_filename_quoting_function      =filename_quoting_function;
-   rl_filename_dequoting_function    =filename_dequoting_function;
-   rl_char_is_quoted_p               =char_is_quoted_p;
+   rl_filename_dequoting_function    =(rl_dequote_func_t*)filename_dequoting_function;
+   rl_char_is_quoted_p               =(rl_linebuf_func_t*)char_is_quoted_p;
 
    rl_completion_display_matches_hook = lftp_rl_display_match_list;
 }
diff -uNrp lftp-4.0.10.orig/src/xstring.cc lftp-4.0.10/src/xstring.cc
--- lftp-4.0.10.orig/src/xstring.cc	2010-09-04 12:47:58 +0000
+++ lftp-4.0.10/src/xstring.cc	2010-09-04 14:32:30 +0000
@@ -379,7 +379,7 @@ const char *xstring::dump_to(xstring& bu
 	    buf.appendf("%02X",(unsigned char)*s++);
 	 buf.append('>');
       } else {
-	 buf.appendf("<long binary, %d bytes>",length());
+	 buf.appendf("<long binary, %d bytes>",(int)length());
       }
    } else {
       while(len>0) {
