diff -ur sysklogd-1.3-31.orig/Makefile sysklogd-1.3-31/Makefile
--- sysklogd-1.3-31.orig/Makefile	Tue Oct 13 00:25:15 1998
+++ sysklogd-1.3-31/Makefile	Tue Sep 12 23:23:33 2000
@@ -8,8 +8,8 @@
 
 # Look where your install program is.
 INSTALL = /usr/bin/install
-BINDIR = /usr/sbin
-MANDIR = /usr/man
+BINDIR = $(TOPDIR)/sbin
+MANDIR = $(TOPDIR)/usr/man
 
 # There is one report that under an all ELF system there may be a need to
 # explicilty link with libresolv.a.  If linking syslogd fails you may wish
@@ -116,7 +116,7 @@
 	${INSTALL} -m 500 -s klogd ${BINDIR}/klogd
 
 install_man:
-	${INSTALL} -o ${MAN_OWNER} -g ${MAN_OWNER} -m 644 sysklogd.8 ${MANDIR}/man8/sysklogd.8
-	${INSTALL} -o ${MAN_OWNER} -g ${MAN_OWNER} -m 644 syslogd.8 ${MANDIR}/man8/syslogd.8
-	${INSTALL} -o ${MAN_OWNER} -g ${MAN_OWNER} -m 644 syslog.conf.5 ${MANDIR}/man5/syslog.conf.5
-	${INSTALL} -o ${MAN_OWNER} -g ${MAN_OWNER} -m 644 klogd.8 ${MANDIR}/man8/klogd.8
+	${INSTALL} -m 644 sysklogd.8 ${MANDIR}/man8/sysklogd.8
+	${INSTALL} -m 644 syslogd.8 ${MANDIR}/man8/syslogd.8
+	${INSTALL} -m 644 syslog.conf.5 ${MANDIR}/man5/syslog.conf.5
+	${INSTALL} -m 644 klogd.8 ${MANDIR}/man8/klogd.8
