diff -ur sysklogd-1.3-31.orig/syslogd.c sysklogd-1.3-31/syslogd.c
--- sysklogd-1.3-31.orig/syslogd.c	Tue Jan 19 14:31:58 1999
+++ sysklogd-1.3-31/syslogd.c	Tue Sep 12 21:38:17 2000
@@ -1389,7 +1389,8 @@
 		msg[len] = '\0';
 		for(p= msg+len-1; *p != '\0' && p > msg; )
 			--p;
-		ptlngth = strlen(++p);
+		if(*p == '\0') p++;
+		ptlngth = strlen(p);
 		if ( (parts[fd] = malloc(ptlngth + 1)) == (char *) 0 )
 			logerror("Cannot allocate memory for message part.");
 		else
