diff -puNr automake-1.14.1.orig/t/lex-clean-cxx.sh automake-1.14.1/t/lex-clean-cxx.sh
--- automake-1.14.1.orig/t/lex-clean-cxx.sh	2013-07-21 22:25:28 +0000
+++ automake-1.14.1/t/lex-clean-cxx.sh	2014-06-29 21:32:02 +0000
@@ -72,7 +72,7 @@ cat > mainfoo.cc << 'END'
 using namespace std;
 int main (int argc, char **argv)
 {
-  extern int yylex (void);
+  extern "C" { int yylex (void); }
   return yylex ();
 }
 END
