diff -puNr openssh-3.6.1p2.orig/auth-pam.c openssh-3.6.1p2/auth-pam.c
--- openssh-3.6.1p2.orig/auth-pam.c	2014-06-28 17:56:00 +0000
+++ openssh-3.6.1p2/auth-pam.c	2014-06-28 17:56:48 +0000
@@ -82,6 +82,7 @@ void do_pam_set_conv(struct pam_conv *co
 int do_pam_authenticate(int flags)
 {
 	int retval = pam_authenticate(__pamh, flags);
+	log_reinit();
 	was_authenticated = (retval == PAM_SUCCESS);
 	return retval;
 }
@@ -212,6 +213,7 @@ void do_pam_cleanup_proc(void *context)
 
 	if (__pamh && getpid() == session_opened) {
 		pam_retval = pam_close_session(__pamh, 0);
+		log_reinit();
 		if (pam_retval != PAM_SUCCESS)
 			log("Cannot close PAM session[%d]: %.200s",
 			    pam_retval, PAM_STRERROR(__pamh, pam_retval));
@@ -219,6 +221,7 @@ void do_pam_cleanup_proc(void *context)
 
 	if (__pamh && creds_set) {
 		pam_retval = pam_setcred(__pamh, PAM_DELETE_CRED);
+		log_reinit();
 		if (pam_retval != PAM_SUCCESS)
 			debug("Cannot delete credentials[%d]: %.200s", 
 			    pam_retval, PAM_STRERROR(__pamh, pam_retval));
@@ -275,6 +278,7 @@ int do_pam_account(char *username, char
 	}
 
 	pam_retval = pam_acct_mgmt(__pamh, 0);
+	log_reinit();
 	debug2("pam_acct_mgmt() = %d", pam_retval);
 	switch (pam_retval) {
 		case PAM_SUCCESS:
@@ -318,6 +322,7 @@ void do_pam_session(char *username, cons
 	}
 
 	pam_retval = pam_open_session(__pamh, 0);
+	log_reinit();
 	if (pam_retval != PAM_SUCCESS)
 		fatal("PAM session setup failed[%d]: %.200s",
 		    pam_retval, PAM_STRERROR(__pamh, pam_retval));
@@ -338,6 +343,7 @@ void do_pam_setcred(int init)
 	debug("PAM establishing creds");
 	pam_retval = pam_setcred(__pamh, 
 	    init ? PAM_ESTABLISH_CRED : PAM_REINITIALIZE_CRED);
+	log_reinit();
 	if (pam_retval != PAM_SUCCESS) {
 		if (was_authenticated)
 			fatal("PAM setcred failed[%d]: %.200s",
@@ -373,6 +379,7 @@ void do_pam_chauthtok(void)
 			    " with privilege separation");
 		pamstate = OTHER;
 		pam_retval = pam_chauthtok(__pamh, PAM_CHANGE_EXPIRED_AUTHTOK);
+		log_reinit();
 		if (pam_retval != PAM_SUCCESS)
 			fatal("PAM pam_chauthtok failed[%d]: %.200s",
 			    pam_retval, PAM_STRERROR(__pamh, pam_retval));
diff -puNr openssh-3.6.1p2.orig/entropy.c openssh-3.6.1p2/entropy.c
--- openssh-3.6.1p2.orig/entropy.c	2002-06-09 19:41:48 +0000
+++ openssh-3.6.1p2/entropy.c	2002-06-29 16:36:17 +0000
@@ -93,7 +93,8 @@ seed_rng(void)
 			    original_uid, strerror(errno));
 			_exit(1);
 		}
-		
+
+		log_done();
 		execl(SSH_RAND_HELPER, "ssh-rand-helper", NULL);
 		fprintf(stderr, "(rand child) Couldn't exec '%s': %s\n", 
 		    SSH_RAND_HELPER, strerror(errno));
diff -puNr openssh-3.6.1p2.orig/log.c openssh-3.6.1p2/log.c
--- openssh-3.6.1p2.orig/log.c	2003-01-14 11:22:43 +0000
+++ openssh-3.6.1p2/log.c	2003-04-07 21:20:02 +0000
@@ -262,6 +262,15 @@ fatal_cleanup(void)
  */
 
 void
+log_reinit(void)
+{
+	if (log_on_stderr)
+		return;
+
+	openlog(argv0 ? argv0 : __progname, LOG_PID|LOG_NDELAY, log_facility);
+}
+
+void
 log_init(char *av0, LogLevel level, SyslogFacility facility, int on_stderr)
 {
 	argv0 = av0;
@@ -332,6 +341,15 @@ log_init(char *av0, LogLevel level, Sysl
 		    (int) facility);
 		exit(1);
 	}
+
+	tzset();
+	log_reinit();
+}
+
+void
+log_done(void)
+{
+	closelog();
 }
 
 #define MSGBUFSIZ 1024
@@ -393,8 +411,6 @@ do_log(LogLevel level, const char *fmt,
 	if (log_on_stderr) {
 		fprintf(stderr, "%s\r\n", fmtbuf);
 	} else {
-		openlog(argv0 ? argv0 : __progname, LOG_PID, log_facility);
 		syslog(pri, "%.500s", fmtbuf);
-		closelog();
 	}
 }
diff -puNr openssh-3.6.1p2.orig/log.h openssh-3.6.1p2/log.h
--- openssh-3.6.1p2.orig/log.h	2002-07-23 21:01:57 +0000
+++ openssh-3.6.1p2/log.h	2003-04-07 21:18:21 +0000
@@ -49,6 +49,8 @@ typedef enum {
 }       LogLevel;
 
 void     log_init(char *, LogLevel, SyslogFacility, int);
+void     log_done(void);
+void     log_reinit(void);
 
 SyslogFacility	log_facility_number(char *);
 LogLevel log_level_number(char *);
diff -puNr openssh-3.6.1p2.orig/monitor.c openssh-3.6.1p2/monitor.c
--- openssh-3.6.1p2.orig/monitor.c	2014-06-28 17:56:00 +0000
+++ openssh-3.6.1p2/monitor.c	2014-06-28 17:56:48 +0000
@@ -292,8 +292,6 @@ monitor_child_preauth(struct monitor *pm
 		}
 
 		if (ent->flags & MON_AUTHDECIDE) {
-			auth_log(authctxt, authenticated, auth_method,
-			    compat20 ? " ssh2" : "");
 			if (!authenticated)
 				authctxt->failures++;
 		}
diff -puNr openssh-3.6.1p2.orig/openbsd-compat/bsd-cray.c openssh-3.6.1p2/openbsd-compat/bsd-cray.c
--- openssh-3.6.1p2.orig/openbsd-compat/bsd-cray.c	2002-09-26 00:38:51 +0000
+++ openssh-3.6.1p2/openbsd-compat/bsd-cray.c	2003-04-07 21:26:05 +0000
@@ -287,6 +287,7 @@ cray_setup (uid_t uid, char *username, c
 		     if (ttyname(0) != NULL) {
 		     /* Force a password change */
 		         printf("Your password has expired; Choose a new one.\n");
+		         log_done();
 		         execl("/bin/passwd", "passwd", username, 0);
 		         exit(9);
 		     }
@@ -724,7 +725,9 @@ cray_delete_tmpdir(char *login, int jid,
 		return;
 
 	if ((child = fork()) == 0) {
+		log_done();
 		execl(CLEANTMPCMD, CLEANTMPCMD, login, jtmp, (char *)NULL);
+		log_reinit();
 		fatal("cray_delete_tmpdir: execl of CLEANTMPCMD failed");
 	}
 
diff -puNr openssh-3.6.1p2.orig/readpass.c openssh-3.6.1p2/readpass.c
--- openssh-3.6.1p2.orig/readpass.c	2003-01-24 00:36:23 +0000
+++ openssh-3.6.1p2/readpass.c	2003-04-07 21:18:21 +0000
@@ -58,7 +58,9 @@ ssh_askpass(char *askpass, const char *m
 		close(p[0]);
 		if (dup2(p[1], STDOUT_FILENO) < 0)
 			fatal("ssh_askpass: dup2: %s", strerror(errno));
+		log_done();
 		execlp(askpass, askpass, msg, (char *) 0);
+		log_reinit();
 		fatal("ssh_askpass: exec(%s): %s", askpass, strerror(errno));
 	}
 	close(p[1]);
diff -puNr openssh-3.6.1p2.orig/scp.c openssh-3.6.1p2/scp.c
--- openssh-3.6.1p2.orig/scp.c	2003-03-21 00:55:32 +0000
+++ openssh-3.6.1p2/scp.c	2003-04-07 21:18:21 +0000
@@ -163,6 +163,7 @@ do_cmd(char *host, char *remuser, char *
 		addargs(&args, "%s", host);
 		addargs(&args, "%s", cmd);
 
+		log_done();
 		execvp(ssh_program, args.list);
 		perror(ssh_program);
 		exit(1);
diff -puNr openssh-3.6.1p2.orig/session.c openssh-3.6.1p2/session.c
--- openssh-3.6.1p2.orig/session.c	2014-06-28 17:56:00 +0000
+++ openssh-3.6.1p2/session.c	2014-06-28 17:56:48 +0000
@@ -468,7 +468,7 @@ do_exec_no_pty(Session *s, const char *c
 		fatal_remove_all_cleanups();
 
 		/* Child.  Reinitialize the log since the pid has changed. */
-		log_init(__progname, options.log_level, options.log_facility, log_stderr);
+		log_reinit();
 
 		/*
 		 * Create a new session and process group since the 4.4BSD
@@ -592,7 +592,7 @@ do_exec_pty(Session *s, const char *comm
 		fatal_remove_all_cleanups();
 
 		/* Child.  Reinitialize the log because the pid has changed. */
-		log_init(__progname, options.log_level, options.log_facility, log_stderr);
+		log_reinit();
 		/* Close the master side of the pseudo tty. */
 		close(ptyfd);
 
@@ -1273,6 +1273,7 @@ launch_login(struct passwd *pw, const ch
 {
 	/* Launch login(1). */
 
+	log_done();
 	execl(LOGIN_PROGRAM, "login", "-h", hostname,
 #ifdef xxxLOGIN_NEEDS_TERM
 		    (s->term ? s->term : "unknown"),
@@ -1448,6 +1449,8 @@ do_child(Session *s, const char *command
 			exit(1);
 		}
 
+		log_done();
+
 		/* Execute the shell. */
 		argv[0] = argv0;
 		argv[1] = NULL;
@@ -1457,6 +1460,9 @@ do_child(Session *s, const char *command
 		perror(shell);
 		exit(1);
 	}
+
+	log_done();
+
 	/*
 	 * Execute the command using the user's shell.  This uses the -c
 	 * option to execute the command.
diff -puNr openssh-3.6.1p2.orig/sftp-int.c openssh-3.6.1p2/sftp-int.c
--- openssh-3.6.1p2.orig/sftp-int.c	2003-03-10 00:21:18 +0000
+++ openssh-3.6.1p2/sftp-int.c	2003-04-07 21:18:21 +0000
@@ -172,9 +172,11 @@ local_do_shell(const char *args)
 		/* XXX: child has pipe fds to ssh subproc open - issue? */
 		if (args) {
 			debug3("Executing %s -c \"%s\"", shell, args);
+			log_done();
 			execl(shell, shell, "-c", args, (char *)NULL);
 		} else {
 			debug3("Executing %s", shell);
+			log_done();
 			execl(shell, shell, (char *)NULL);
 		}
 		fprintf(stderr, "Couldn't execute \"%s\": %s\n", shell,
diff -puNr openssh-3.6.1p2.orig/sftp.c openssh-3.6.1p2/sftp.c
--- openssh-3.6.1p2.orig/sftp.c	2003-01-10 10:43:25 +0000
+++ openssh-3.6.1p2/sftp.c	2003-04-07 21:18:21 +0000
@@ -86,6 +86,7 @@ connect_to_server(char *path, char **arg
 		close(*out);
 		close(c_in);
 		close(c_out);
+		log_done();
 		execv(path, args);
 		fprintf(stderr, "exec: %s: %s\n", path, strerror(errno));
 		exit(1);
diff -puNr openssh-3.6.1p2.orig/ssh-agent.c openssh-3.6.1p2/ssh-agent.c
--- openssh-3.6.1p2.orig/ssh-agent.c	2003-03-15 00:37:09 +0000
+++ openssh-3.6.1p2/ssh-agent.c	2003-04-07 21:18:21 +0000
@@ -1159,6 +1159,7 @@ main(int ac, char **av)
 			perror("setenv");
 			exit(1);
 		}
+		log_done();
 		execvp(av[0], av);
 		perror(av[0]);
 		exit(1);
diff -puNr openssh-3.6.1p2.orig/ssh-rand-helper.c openssh-3.6.1p2/ssh-rand-helper.c
--- openssh-3.6.1p2.orig/ssh-rand-helper.c	2003-03-17 05:13:53 +0000
+++ openssh-3.6.1p2/ssh-rand-helper.c	2003-04-07 21:18:21 +0000
@@ -312,7 +312,9 @@ hash_command_output(entropy_cmd_t *src,
 			close(p[1]);
 			close(devnull);
 
+			log_done();
 			execv(src->path, (char**)(src->args));
+			log_reinit();
 
 			debug("(child) Couldn't exec '%s': %s", 
 			    src->cmdstring, strerror(errno));
diff -puNr openssh-3.6.1p2.orig/sshconnect.c openssh-3.6.1p2/sshconnect.c
--- openssh-3.6.1p2.orig/sshconnect.c	2002-12-23 02:06:20 +0000
+++ openssh-3.6.1p2/sshconnect.c	2003-04-07 21:18:21 +0000
@@ -126,6 +126,8 @@ ssh_proxy_connect(const char *host, u_sh
 		/* Cannot be 1 because pin allocated two descriptors. */
 		close(pout[1]);
 
+		log_done();
+
 		/* Stderr is left as it is so that error messages get
 		   printed on the user's terminal. */
 		argv[0] = _PATH_BSHELL;
diff -puNr openssh-3.6.1p2.orig/sshconnect2.c openssh-3.6.1p2/sshconnect2.c
--- openssh-3.6.1p2.orig/sshconnect2.c	2003-04-01 11:43:40 +0000
+++ openssh-3.6.1p2/sshconnect2.c	2003-04-07 21:18:21 +0000
@@ -937,7 +937,9 @@ ssh_keysign(Key *key, u_char **sigp, u_i
 			fatal("ssh_keysign: dup2: %s", strerror(errno));
 		close(from[1]);
 		close(to[0]);
+		log_done();
 		execl(_PATH_SSH_KEY_SIGN, _PATH_SSH_KEY_SIGN, (char *) 0);
+		log_reinit();
 		fatal("ssh_keysign: exec(%s): %s", _PATH_SSH_KEY_SIGN,
 		    strerror(errno));
 	}
diff -puNr openssh-3.6.1p2.orig/sshd.c openssh-3.6.1p2/sshd.c
--- openssh-3.6.1p2.orig/sshd.c	2014-06-28 17:56:00 +0000
+++ openssh-3.6.1p2/sshd.c	2014-06-28 17:56:48 +0000
@@ -261,7 +261,9 @@ sighup_restart(void)
 	log("Received SIGHUP; restarting.");
 	close_listen_socks();
 	close_startup_pipes();
+	log_done();
 	execv(saved_argv[0], saved_argv);
+	log_reinit();
 	log("RESTART FAILED: av[0]='%.100s', error: %.100s.", saved_argv[0],
 	    strerror(errno));
 	exit(1);
@@ -1118,7 +1120,7 @@ main(int ac, char **av)
 #endif /* TIOCNOTTY */
 	}
 	/* Reinitialize the log (because of the fork above). */
-	log_init(__progname, options.log_level, options.log_facility, log_stderr);
+	log_reinit();
 
 	/* Initialize the random number generator. */
 	arc4random_stir();
@@ -1367,7 +1369,7 @@ main(int ac, char **av)
 						close_listen_socks();
 						sock_in = newsock;
 						sock_out = newsock;
-						log_init(__progname, options.log_level, options.log_facility, log_stderr);
+						log_reinit();
 						break;
 					}
 				}
