diff -ur openssh-2.1.1p2.orig/Makefile.in openssh-2.1.1p2/Makefile.in
--- openssh-2.1.1p2.orig/Makefile.in	Wed Jun 28 02:24:49 2000
+++ openssh-2.1.1p2/Makefile.in	Sun Jul  9 04:54:55 2000
@@ -150,20 +150,20 @@
 	fi
 
 host-key: ssh-keygen
-	if [ -f "$(sysconfdir)/ssh_host_key" ] ; then \
-		echo "$(sysconfdir)/ssh_host_key already exists, skipping." ; \
+	if [ -f "$(DESTDIR)$(sysconfdir)/ssh_host_key" ] ; then \
+		echo "$(DESTDIR)$(sysconfdir)/ssh_host_key already exists, skipping." ; \
 	else \
-		./ssh-keygen -b 1024 -f $(sysconfdir)/ssh_host_key -N "" ; \
+		./ssh-keygen -b 1024 -f $(DESTDIR)$(sysconfdir)/ssh_host_key -N "" ; \
 	fi ; \
-	if [ -f $(sysconfdir)/ssh_host_dsa_key ] ; then \
-		echo "$(sysconfdir)/ssh_host_dsa_key already exists, skipping." ; \
+	if [ -f $(DESTDIR)$(sysconfdir)/ssh_host_dsa_key ] ; then \
+		echo "$(DESTDIR)$(sysconfdir)/ssh_host_dsa_key already exists, skipping." ; \
 	else \
-		./ssh-keygen -d -f $(sysconfdir)/ssh_host_dsa_key -N "" ; \
+		./ssh-keygen -d -f $(DESTDIR)$(sysconfdir)/ssh_host_dsa_key -N "" ; \
 	fi ;
 
 host-key-force: ssh-keygen
-	./ssh-keygen -b 1024 -f $(sysconfdir)/ssh_host_key -N ""
-	./ssh-keygen -d -f $(sysconfdir)/ssh_host_dsa_key -N ""
+	./ssh-keygen -b 1024 -f $(DESTDIR)$(sysconfdir)/ssh_host_key -N ""
+	./ssh-keygen -d -f $(DESTDIR)$(sysconfdir)/ssh_host_dsa_key -N ""
 
 uninstallall:	uninstall
 	-rm -f $(DESTDIR)$(sysconfdir)/ssh_config
