diff -ur openssh-2.5.2p1.orig/auth-pam.c openssh-2.5.2p1/auth-pam.c
--- openssh-2.5.2p1.orig/auth-pam.c	Thu Mar  1 01:18:58 2001
+++ openssh-2.5.2p1/auth-pam.c	Wed Mar 21 04:02:20 2001
@@ -172,7 +172,7 @@
 /* Called at exit to cleanly shutdown PAM */
 void do_pam_cleanup_proc(void *context)
 {
-	int pam_retval;
+	int pam_retval = PAM_SUCCESS;
 
 	if (__pamh && session_opened) {
 		pam_retval = pam_close_session(__pamh, 0);
