diff -ur openssh-3.1p1.orig/scp.c openssh-3.1p1/scp.c
--- openssh-3.1p1.orig/scp.c	Tue Mar  5 21:59:45 2002
+++ openssh-3.1p1/scp.c	Sun Mar 17 00:23:34 2002
@@ -90,7 +90,7 @@
 #endif
 
 /* For progressmeter() -- number of seconds before xfer considered "stalled" */
-#define STALLTIME	5
+#define STALLTIME	60
 /* alarm() interval for updating progress meter */
 #define PROGRESSTIME	1
 
@@ -1158,10 +1158,12 @@
 	if (cursize > lastsize) {
 		lastupdate = now;
 		lastsize = cursize;
+#if 0
 		if (wait.tv_sec >= STALLTIME) {
 			start.tv_sec += wait.tv_sec;
 			start.tv_usec += wait.tv_usec;
 		}
+#endif
 		wait.tv_sec = 0;
 	}
 	timersub(&now, &start, &td);
