diff -ur openssh-3.4p1-owl-hide-unknown/auth-pam.c openssh-3.4p1/auth-pam.c
--- openssh-3.4p1-owl-hide-unknown/auth-pam.c	Wed May  8 06:27:56 2002
+++ openssh-3.4p1/auth-pam.c	Sat Jun 29 21:13:08 2002
@@ -204,15 +204,20 @@
 {
 	extern ServerOptions options;
 	int pam_retval;
+	int fail;
 	struct passwd *pw = authctxt->pw;
 
 	do_pam_set_conv(&conv);
 
 	/* deny if no user. */
-	if (pw == NULL)
-		return 0;
-	if (pw->pw_uid == 0 && options.permit_root_login == PERMIT_NO_PASSWD)
-		return 0;
+	fail = 0;
+	if (pw == NULL ||
+	    (pw->pw_uid == 0 && options.permit_root_login == PERMIT_NO_PASSWD)) {
+		/* Attempt to reduce information leaks via timings. */
+		fail = 1;
+		if (*password) password = "INCORRECT";
+	}
+
 	if (*password == '\0' && options.permit_empty_passwd == 0)
 		return 0;
 
@@ -221,14 +226,14 @@
 	pamstate = INITIAL_LOGIN;
 	pam_retval = do_pam_authenticate(
 	    options.permit_empty_passwd == 0 ? PAM_DISALLOW_NULL_AUTHTOK : 0);
-	if (pam_retval == PAM_SUCCESS) {
-		debug("PAM Password authentication accepted for "
-		    "user \"%.100s\"", pw->pw_name);
+	if (!fail && pam_retval == PAM_SUCCESS) {
+		debug("PAM password authentication accepted for "
+		    "%.100s", pw->pw_name);
 		return 1;
 	} else {
-		debug("PAM Password authentication for \"%.100s\" "
-		    "failed[%d]: %s", pw->pw_name, pam_retval, 
-		    PAM_STRERROR(__pamh, pam_retval));
+		debug("PAM password authentication failed for "
+		    "%.100s: %s", pw ? pw->pw_name : "an illegal user",
+		    fail ? "Forced" : PAM_STRERROR(__pamh, pam_retval));
 		return 0;
 	}
 }
diff -ur openssh-3.4p1-owl-hide-unknown/auth1.c openssh-3.4p1/auth1.c
--- openssh-3.4p1-owl-hide-unknown/auth1.c	Sat Jun 29 21:02:35 2002
+++ openssh-3.4p1/auth1.c	Sat Jun 29 21:13:08 2002
@@ -383,7 +383,7 @@
 	    use_privsep ? " [net]" : "");
 
 #ifdef USE_PAM
-	PRIVSEP(start_pam(authctxt->pw == NULL ? "NOUSER" : user));
+	PRIVSEP(start_pam(authctxt->pw ? user : "ILLEGAL USER"));
 #endif
 
 	/*
diff -ur openssh-3.4p1-owl-hide-unknown/auth2.c openssh-3.4p1/auth2.c
--- openssh-3.4p1-owl-hide-unknown/auth2.c	Sat Jun 29 21:03:37 2002
+++ openssh-3.4p1/auth2.c	Sat Jun 29 21:13:08 2002
@@ -164,7 +164,7 @@
 			authctxt->pw = NULL;
 			authctxt->valid = 0;
 #ifdef USE_PAM
-			PRIVSEP(start_pam("NOUSER"));
+			PRIVSEP(start_pam("ILLEGAL USER"));
 #endif
 		}
 		setproctitle("%s%s", authctxt->pw ? user : "UNKNOWN USER",
