diff -ur openssh-3.0.2p1.orig/auth.c openssh-3.0.2p1-owl-hide-unknown/auth.c
--- openssh-3.0.2p1.orig/auth.c	Mon Nov 12 03:06:07 2001
+++ openssh-3.0.2p1-owl-hide-unknown/auth.c	Wed Dec 12 19:02:38 2001
@@ -207,8 +207,10 @@
 	authlog("%s %s for %s%.100s from %.200s port %d%s",
 	    authmsg,
 	    method,
-	    authctxt->valid ? "" : "illegal user ",
-	    authctxt->user,
+	    !authctxt->known ? "UNKNOWN USER" :
+	    (!authctxt->valid ? "ILLEGAL USER " :
+	    (authctxt->pw->pw_uid == 0 ? "ROOT USER " : "")),
+	    authctxt->known ? authctxt->user : "",
 	    get_remote_ipaddr(),
 	    get_remote_port(),
 	    info);
diff -ur openssh-3.0.2p1.orig/auth.h openssh-3.0.2p1-owl-hide-unknown/auth.h
--- openssh-3.0.2p1.orig/auth.h	Wed Jul  4 08:46:57 2001
+++ openssh-3.0.2p1-owl-hide-unknown/auth.h	Wed Dec 12 18:57:34 2001
@@ -47,6 +47,7 @@
 	int		 success;
 	int		 postponed;
 	int		 valid;
+	int		 known;
 	int		 attempt;
 	int		 failures;
 	char		*user;
diff -ur openssh-3.0.2p1.orig/auth1.c openssh-3.0.2p1-owl-hide-unknown/auth1.c
--- openssh-3.0.2p1.orig/auth1.c	Tue Nov 13 15:46:19 2001
+++ openssh-3.0.2p1-owl-hide-unknown/auth1.c	Wed Dec 12 18:57:34 2001
@@ -388,6 +388,7 @@
 
 	/* Verify that the user is a valid user. */
 	pw = getpwnam(user);
+	authctxt->known = (pw != NULL);
 	if (pw && allowed_user(pw)) {
 		authctxt->valid = 1;
 		pw = pwcopy(pw);
@@ -397,7 +398,7 @@
 	}
 	authctxt->pw = pw;
 
-	setproctitle("%s", pw ? user : "unknown");
+	setproctitle("%s", authctxt->known ? user : "UNKNOWN USER");
 
 #ifdef USE_PAM
 	if (pw)
diff -ur openssh-3.0.2p1.orig/auth2.c openssh-3.0.2p1-owl-hide-unknown/auth2.c
--- openssh-3.0.2p1.orig/auth2.c	Tue Nov 13 15:46:19 2001
+++ openssh-3.0.2p1-owl-hide-unknown/auth2.c	Wed Dec 12 18:57:34 2001
@@ -196,6 +196,7 @@
 		/* setup auth context */
 		struct passwd *pw = NULL;
 		pw = getpwnam(user);
+		authctxt->known = (pw != NULL);
 		if (pw && allowed_user(pw) && strcmp(service, "ssh-connection")==0) {
 			authctxt->pw = pwcopy(pw);
 			authctxt->valid = 1;
@@ -204,12 +205,13 @@
 			start_pam(pw->pw_name);
 #endif
 		} else {
-			log("input_userauth_request: illegal user %s", user);
+			authctxt->pw = NULL;
+			authctxt->valid = 0;
 #ifdef USE_PAM
 			start_pam("NOUSER");
 #endif
 		}
-		setproctitle("%s", pw ? user : "unknown");
+		setproctitle("%s", authctxt->known ? user : "UNKNOWN USER");
 		authctxt->user = xstrdup(user);
 		authctxt->service = xstrdup(service);
 		authctxt->style = style ? xstrdup(style) : NULL;
