diff -ur openssh-3.6.1p1.orig/entropy.c openssh-3.6.1p1/entropy.c
--- openssh-3.6.1p1.orig/entropy.c	Sun Jun  9 19:41:48 2002
+++ openssh-3.6.1p1/entropy.c	Mon Apr  7 19:29:13 2003
@@ -138,9 +138,9 @@
 {
 	/*
 	 * OpenSSL version numbers: MNNFFPPS: major minor fix patch status
-	 * We match major, minor, fix and status (not patch)
+	 * We match major, minor and fix (but not patch and status)
 	 */
-	if ((SSLeay() ^ OPENSSL_VERSION_NUMBER) & ~0xff0L)
+	if ((SSLeay() ^ OPENSSL_VERSION_NUMBER) & ~0xfffL)
 		fatal("OpenSSL version mismatch. Built against %lx, you "
 		    "have %lx", OPENSSL_VERSION_NUMBER, SSLeay());
 
