diff -ur openssh-3.6.1p1.orig/cipher.c openssh-3.6.1p1/cipher.c
--- openssh-3.6.1p1.orig/cipher.c	Mon Dec 23 05:04:22 2002
+++ openssh-3.6.1p1/cipher.c	Wed Apr  2 18:43:34 2003
@@ -50,17 +50,17 @@
 
 #if OPENSSL_VERSION_NUMBER < 0x00907000L
 #include "rijndael.h"
-static const EVP_CIPHER *evp_rijndael(void);
+static EVP_CIPHER *evp_rijndael(void);
 #endif
-static const EVP_CIPHER *evp_ssh1_3des(void);
-static const EVP_CIPHER *evp_ssh1_bf(void);
+static EVP_CIPHER *evp_ssh1_3des(void);
+static EVP_CIPHER *evp_ssh1_bf(void);
 
 struct Cipher {
 	char	*name;
 	int	number;		/* for ssh1 only */
 	u_int	block_size;
 	u_int	key_len;
-	const EVP_CIPHER	*(*evptype)(void);
+	EVP_CIPHER	*(*evptype)(void);
 } ciphers[] = {
 	{ "none", 		SSH_CIPHER_NONE, 8, 0, EVP_enc_null },
 	{ "des", 		SSH_CIPHER_DES, 8, 8, EVP_des_cbc },
@@ -395,7 +395,7 @@
 	return (1);
 }
 
-static const EVP_CIPHER *
+static EVP_CIPHER *
 evp_ssh1_3des(void)
 {
 	static EVP_CIPHER ssh1_3des;
@@ -462,7 +462,7 @@
 	return (ret);
 }
 
-static const EVP_CIPHER *
+static EVP_CIPHER *
 evp_ssh1_bf(void)
 {
 	static EVP_CIPHER ssh1_bf;
@@ -567,7 +567,7 @@
 	return (1);
 }
 
-static const EVP_CIPHER *
+static EVP_CIPHER *
 evp_rijndael(void)
 {
 	static EVP_CIPHER rijndal_cbc;
diff -ur openssh-3.6.1p1.orig/mac.c openssh-3.6.1p1/mac.c
--- openssh-3.6.1p1.orig/mac.c	Thu Jun  6 23:48:16 2002
+++ openssh-3.6.1p1/mac.c	Wed Apr  2 18:43:34 2003
@@ -36,7 +36,7 @@
 
 struct {
 	char		*name;
-	const EVP_MD *	(*mdfunc)(void);
+	EVP_MD *	(*mdfunc)(void);
 	int		truncatebits;	/* truncate digest if != 0 */
 } macs[] = {
 	{ "hmac-sha1",			EVP_sha1, 0, },
