diff -ur openssh-2.9.9p2.orig/scp.c openssh-2.9.9p2/scp.c
--- openssh-2.9.9p2.orig/scp.c	Thu Sep 20 04:57:56 2001
+++ openssh-2.9.9p2/scp.c	Thu Sep 27 04:04:00 2001
@@ -90,7 +90,7 @@
 #endif
 
 /* For progressmeter() -- number of seconds before xfer considered "stalled" */
-#define STALLTIME	5
+#define STALLTIME	60
 /* alarm() interval for updating progress meter */
 #define PROGRESSTIME	1
 
@@ -1154,10 +1154,12 @@
 	if (cursize > lastsize) {
 		lastupdate = now;
 		lastsize = cursize;
+#if 0
 		if (wait.tv_sec >= STALLTIME) {
 			start.tv_sec += wait.tv_sec;
 			start.tv_usec += wait.tv_usec;
 		}
+#endif
 		wait.tv_sec = 0;
 	}
 	timersub(&now, &start, &td);
