diff -ur openssh-3.0.2p1.orig/sshd.c openssh-3.0.2p1/sshd.c
--- openssh-3.0.2p1.orig/sshd.c	Mon Nov 12 03:07:12 2001
+++ openssh-3.0.2p1/sshd.c	Wed Dec 12 19:37:30 2001
@@ -784,6 +784,9 @@
 		log_stderr = 1;
 	log_init(__progname, options.log_level, options.log_facility, log_stderr);
 
+	if (geteuid() == 0 && setgroups(0, NULL) != 0)
+		fatal("setgroups: %.100s", strerror(errno));
+
 	/*
 	 * If not in debugging mode, and not started from inetd, disconnect
 	 * from the controlling terminal, and fork.  The original process
