diff -ur openssh-2.1.1p2.orig/auth1.c openssh-2.1.1p2/auth1.c
--- openssh-2.1.1p2.orig/auth1.c	Sat Jul  1 06:56:09 2000
+++ openssh-2.1.1p2/auth1.c	Mon Jul 17 02:06:52 2000
@@ -65,8 +65,7 @@
 {
 	int attempt = 0;
 
-	log("Faking authloop for illegal user %.200s from %.200s port %d",
-	    user,
+	log("Faking authloop for an unknown user from %.200s port %d",
 	    get_remote_ipaddr(),
 	    get_remote_port());
 
@@ -457,8 +456,6 @@
 	/* Get the user name. */
 	user = packet_get_string(&ulen);
 	packet_integrity_check(plen, (4 + ulen), SSH_CMSG_USER);
-
-	setproctitle("%s", user);
 
 #ifdef AFS
 	/* If machine has AFS, set process authentication group. */
diff -ur openssh-2.1.1p2.orig/auth2.c openssh-2.1.1p2/auth2.c
--- openssh-2.1.1p2.orig/auth2.c	Wed Jun 28 09:22:42 2000
+++ openssh-2.1.1p2/auth2.c	Mon Jul 17 02:08:34 2000
@@ -396,8 +396,9 @@
 		authctxt->valid = 0;
 		authctxt->user = xstrdup(u);
 		authctxt->service = xstrdup(s);
-		setproctitle("%s", u);
 		pw = getpwnam(u);
+		if (pw)
+			setproctitle("%s", u);
 		if (!pw || !allowed_user(pw)) {
 			log("auth_set_user: illegal user %s", u);
 			return NULL;
