diff -ur openssh-3.1p1.orig/sshd.c openssh-3.1p1/sshd.c
--- openssh-3.1p1.orig/sshd.c	Tue Mar  5 04:31:30 2002
+++ openssh-3.1p1/sshd.c	Sun Mar 17 00:23:58 2002
@@ -818,6 +818,9 @@
 		log_stderr = 1;
 	log_init(__progname, options.log_level, options.log_facility, log_stderr);
 
+	if (geteuid() == 0 && setgroups(0, NULL) != 0)
+		fatal("setgroups: %.100s", strerror(errno));
+
 	/*
 	 * If not in debugging mode, and not started from inetd, disconnect
 	 * from the controlling terminal, and fork.  The original process
