diff -ur openssh-3.1p1.orig/auth.c openssh-3.1p1/auth.c
--- openssh-3.1p1.orig/auth.c	Tue Mar  5 04:42:43 2002
+++ openssh-3.1p1/auth.c	Sat Mar 16 23:48:45 2002
@@ -226,8 +226,10 @@
 	authlog("%s %s for %s%.100s from %.200s port %d%s",
 	    authmsg,
 	    method,
-	    authctxt->valid ? "" : "illegal user ",
-	    authctxt->user,
+	    !authctxt->known ? "UNKNOWN USER" :
+	    (!authctxt->valid ? "ILLEGAL USER " :
+	    (authctxt->pw->pw_uid == 0 ? "ROOT USER " : "")),
+	    authctxt->known ? authctxt->user : "",
 	    get_remote_ipaddr(),
 	    get_remote_port(),
 	    info);
diff -ur openssh-3.1p1.orig/auth.h openssh-3.1p1/auth.h
--- openssh-3.1p1.orig/auth.h	Tue Mar  5 04:53:04 2002
+++ openssh-3.1p1/auth.h	Sat Mar 16 23:48:45 2002
@@ -49,6 +49,7 @@
 	int		 success;
 	int		 postponed;
 	int		 valid;
+	int		 known;
 	int		 attempt;
 	int		 failures;
 	char		*user;
diff -ur openssh-3.1p1.orig/auth1.c openssh-3.1p1/auth1.c
--- openssh-3.1p1.orig/auth1.c	Thu Feb 14 12:39:50 2002
+++ openssh-3.1p1/auth1.c	Sat Mar 16 23:48:45 2002
@@ -383,6 +383,7 @@
 
 	/* Verify that the user is a valid user. */
 	pw = getpwnam(user);
+	authctxt->known = (pw != NULL);
 	if (pw && allowed_user(pw)) {
 		authctxt->valid = 1;
 		pw = pwcopy(pw);
@@ -392,7 +393,7 @@
 	}
 	authctxt->pw = pw;
 
-	setproctitle("%s", pw ? user : "unknown");
+	setproctitle("%s", authctxt->known ? user : "UNKNOWN USER");
 
 #ifdef USE_PAM
 	start_pam(pw == NULL ? "NOUSER" : user);
diff -ur openssh-3.1p1.orig/auth2.c openssh-3.1p1/auth2.c
--- openssh-3.1p1.orig/auth2.c	Tue Feb 26 21:09:43 2002
+++ openssh-3.1p1/auth2.c	Sat Mar 16 23:48:45 2002
@@ -185,6 +185,7 @@
 		/* setup auth context */
 		struct passwd *pw = NULL;
 		pw = getpwnam(user);
+		authctxt->known = (pw != NULL);
 		if (pw && allowed_user(pw) && strcmp(service, "ssh-connection")==0) {
 			authctxt->pw = pwcopy(pw);
 			authctxt->valid = 1;
@@ -193,12 +194,13 @@
 			start_pam(pw->pw_name);
 #endif
 		} else {
-			log("input_userauth_request: illegal user %s", user);
+			authctxt->pw = NULL;
+			authctxt->valid = 0;
 #ifdef USE_PAM
 			start_pam("NOUSER");
 #endif
 		}
-		setproctitle("%s", pw ? user : "unknown");
+		setproctitle("%s", authctxt->known ? user : "UNKNOWN USER");
 		authctxt->user = xstrdup(user);
 		authctxt->service = xstrdup(service);
 		authctxt->style = style ? xstrdup(style) : NULL;
