diff -urp strace-20020608.orig/defs.h strace-20020608/defs.h
--- strace-20020608.orig/defs.h	Thu May 30 00:48:52 2002
+++ strace-20020608/defs.h	Sat Jun  8 17:06:47 2002
@@ -395,6 +395,9 @@ extern void printsiginfo P((siginfo_t *,
 extern void printsock P((struct tcb *, long, int));
 extern void print_sock_optmgmt P((struct tcb *, long, int));
 extern void printrusage P((struct tcb *, long));
+#ifdef LINUX_64BIT
+extern void printrusage32 P((struct tcb *, long));
+#endif
 extern int clearbpt P((struct tcb *));
 extern int setbpt P((struct tcb *));
 extern int sigishandled P((struct tcb *, int));
diff -urp strace-20020608.orig/desc.c strace-20020608/desc.c
--- strace-20020608.orig/desc.c	Thu May 30 00:48:52 2002
+++ strace-20020608/desc.c	Sat Jun  8 17:03:11 2002
@@ -540,6 +540,7 @@ struct tcb *tcp;
 }
 
 #ifdef ALPHA
+int
 sys_osf_select(tcp)
 struct tcb *tcp;
 {
diff -urp strace-20020608.orig/file.c strace-20020608/file.c
--- strace-20020608.orig/file.c	Thu May 30 00:48:52 2002
+++ strace-20020608/file.c	Sat Jun  8 17:04:01 2002
@@ -40,6 +40,7 @@
 #include <linux/types.h>
 #include <linux/dirent.h>
 #undef dirent
+#undef dirent64
 #else
 #define kernel_dirent dirent
 #endif
@@ -630,6 +631,7 @@ long addr;
 }
 #endif /* LINUXSPARC */
 
+#if defined(FREEBSD) || HAVE_ST_FLAGS
 static struct xlat fileflags[] = {
 #ifdef FREEBSD
 	{ UF_NODUMP,	"UF_NODUMP"	},
@@ -651,6 +653,7 @@ static struct xlat fileflags[] = {
 #endif
 	{ 0,		NULL		},
 };
+#endif
 
 #ifdef FREEBSD
 int
@@ -721,7 +724,7 @@ struct stat *statbuf;
 #endif /* !HAVE_ST_RDEV */
 	    break;
     default:
-	    tprintf("st_size=%lu, ", statbuf->st_size);
+	    tprintf("st_size=%lu, ", (unsigned long)statbuf->st_size);
 	    break;
     }
     if (!abbrev(tcp)) {
@@ -1921,8 +1924,9 @@ struct tcb *tcp;
 		struct dirent64 *d = (struct dirent64 *) &buf[i];
 #ifdef linux
 		if (!abbrev(tcp)) {
-			tprintf("%s{d_ino=%lu, d_off=%lu, ",
-				i ? " " : "", d->d_ino, d->d_off);
+			tprintf("%s{d_ino=%lu, d_off=%llu, ",
+				i ? " " : "",
+				d->d_ino, (unsigned long long)d->d_off);
 			tprintf("d_reclen=%u, d_name=\"%s\"}",
 				d->d_reclen, d->d_name);
 		}
diff -urp strace-20020608.orig/linux/syscall.h strace-20020608/linux/syscall.h
--- strace-20020608.orig/linux/syscall.h	Thu May 30 00:48:52 2002
+++ strace-20020608/linux/syscall.h	Sat Jun  8 17:02:26 2002
@@ -83,7 +83,7 @@ int sys_rt_sigpending(), sys_rt_sigsuspe
 int sys_rt_sigtimedwait(), sys_prctl(), sys_poll(), sys_vfork();
 int sys_sendfile(), sys_old_mmap(), sys_stat64(), sys_lstat64(), sys_fstat64();
 int sys_truncate64(), sys_ftruncate64(), sys_pivotroot();
-int sys_getdents64, sys_clone2();
+int sys_getdents64(), sys_clone2();
 
 
 /* sys_socketcall subcalls */
diff -urp strace-20020608.orig/net.c strace-20020608/net.c
--- strace-20020608.orig/net.c	Thu May 30 00:48:52 2002
+++ strace-20020608/net.c	Thu May 15 23:55:20 2003
@@ -40,6 +40,8 @@
 #define in6_addr in6_addr_libc
 #define ipv6_mreq ipv6_mreq_libc
 #define sockaddr_in6 sockaddr_in6_libc
+#define in6addr_any in6addr_any_libc
+#define in6addr_loopback in6addr_loopback_libc
 #endif
 
 #include <netinet/in.h>
@@ -66,15 +68,21 @@
 #undef in6_addr
 #undef ipv6_mreq
 #undef sockaddr_in6
+#undef in6addr_any
+#undef in6addr_loopback
 #define in6_addr in6_addr_kernel
 #define ipv6_mreq ipv6_mreq_kernel
 #define sockaddr_in6 sockaddr_in6_kernel
+#define in6addr_any in6addr_any_kernel
+#define in6addr_loopback in6addr_loopback_kernel
 #endif
 #include <linux/in6.h>
 #if defined(HAVE_SIN6_SCOPE_ID_LINUX)
 #undef in6_addr
 #undef ipv6_mreq
 #undef sockaddr_in6
+#undef in6addr_any
+#undef in6addr_loopback
 #define in6_addr in6_addr_libc
 #define ipv6_mreq ipv6_mreq_libc
 #define sockaddr_in6 sockaddr_in6_kernel
diff -urp strace-20020608.orig/process.c strace-20020608/process.c
--- strace-20020608.orig/process.c	Thu May 30 00:48:52 2002
+++ strace-20020608/process.c	Sat Jun  8 17:02:26 2002
@@ -484,6 +484,7 @@ struct tcb *tcp;
 	return 0;
 }
 
+#ifdef SYS_vfork
 int
 change_syscall(tcp, new)
 struct tcb *tcp;
@@ -543,7 +544,9 @@ int new;
 #endif /* LINUX */
 	return -1;
 }
+#endif
 
+#if 0
 int
 setarg(tcp, argnum)
 	struct tcb *tcp;
@@ -592,6 +595,7 @@ setarg(tcp, argnum)
 #endif
 	return 0;
 }
+#endif
 
 #ifdef SYS_clone
 int
diff -urp strace-20020608.orig/syscall.c strace-20020608/syscall.c
--- strace-20020608.orig/syscall.c	Thu May 30 00:48:52 2002
+++ strace-20020608/syscall.c	Sat Jun  8 17:02:26 2002
@@ -455,8 +455,7 @@ const struct subcall subcalls_table[] = 
 };
 #endif /* FREEBSD */
 
-#if !(defined(LINUX) && ( defined(ALPHA) || defined(MIPS) ))
-
+#if defined(LINUX) && defined(SPARC)
 const int socket_map [] = {
 	       /* SYS_SOCKET      */ 97,
 	       /* SYS_BIND        */ 104,
@@ -477,7 +476,7 @@ const int socket_map [] = {
 	       /* SYS_RECVMSG     */ 113
 };
 
-void
+static void
 sparc_socket_decode (tcp)
 struct tcb *tcp;
 {
@@ -498,7 +497,11 @@ struct tcb *tcp;
 		addr += sizeof (arg);
 	}
 }
+#endif
 
+#if defined(LINUX) && \
+    !defined (ALPHA) && !defined(SPARC) && !defined(MIPS) && !defined(HPPA) || \
+    defined(SVR4) || defined(FREEBSD) || defined(SUNOS4)
 static void
 decode_subcall(tcp, subcall, nsubcalls, style)
 struct tcb *tcp;
@@ -832,7 +835,7 @@ struct tcb *tcp;
 			return -1;
 		case 0x91d02000:
 			/* SunOS syscall trap. (pers 1) */
-			fprintf(stderr,"syscall: SunOS no support\n");
+			fprintf(stderr,"syscall: SunOS not supported\n");
 			return -1;
 		case 0x91d02008:
 			/* Solaris 2.x syscall trap. (per 2) */
@@ -852,7 +855,7 @@ struct tcb *tcp;
 				tcp->flags &= ~TCB_WAITEXECVE;
 				return 0;
 			}
-			fprintf(stderr,"syscall: unknown syscall trap %08x %08x\n", trap, regs.r_pc);
+			fprintf(stderr,"syscall: unknown syscall trap %08lx %08x\n", trap, regs.r_pc);
 			return -1;
 		}
 
@@ -949,7 +952,10 @@ syscall_fixup(tcp)
 struct tcb *tcp;
 {
 #ifndef USE_PROCFS
+#if defined(LINUX) && (defined(I386) || defined(S390) || defined(POWERPC) || \
+    defined(M68K) || defined(ARM) || defined(HPPA) || defined(IA64))
 	int pid = tcp->pid;
+#endif
 #else /* USE_PROCFS */	
 	int scno = tcp->scno;
 
@@ -1278,7 +1284,8 @@ struct tcb *tcp;
 int syscall_enter(tcp)
 struct tcb *tcp;
 {
-#ifndef USE_PROCFS
+#if !defined(USE_PROCFS) && \
+    ((defined(LINUX) && !defined(SPARC)) || defined(SUNOS4))
 	int pid = tcp->pid;
 #endif /* !USE_PROCFS */	
 #ifdef LINUX
@@ -1738,7 +1745,7 @@ struct tcb *tcp;
 		decode_subcall(tcp, SYS_ipc_subcall,
 			SYS_ipc_nsubcalls, shift_style);
 		break;
-#endif /* !ALPHA && !MIPS && !SPARC */
+#endif /* !ALPHA && !SPARC && !MIPS && !HPPA */
 #ifdef SPARC
 	case SYS_socketcall:
 		sparc_socket_decode (tcp);
diff -urp strace-20020608.orig/util.c strace-20020608/util.c
--- strace-20020608.orig/util.c	Thu May 30 00:48:52 2002
+++ strace-20020608/util.c	Sat Jun  8 17:02:26 2002
@@ -69,7 +69,19 @@
 
 #if defined(LINUX) && defined(SPARC)
 
+#if HAVE_ASM_REG_H
+#ifdef SPARC
+#  define fpq kernel_fpq
+#  define fq kernel_fq
+#  define fpu kernel_fpu
+#endif
 #include <asm/reg.h>
+#ifdef SPARC
+#  undef fpq
+#  undef fq
+#  undef fpu
+#endif
+#endif /* HAVE_ASM_REG_H */
 
 #if !defined(__GLIBC__)
 
@@ -1040,7 +1052,7 @@ struct tcb *tcp;
 		tprintf("[????????] ");
 		return;
 	}
-	tprintf("[%08lx] ", regs.r_pc);
+	tprintf("[%08x] ", regs.r_pc);
 #elif defined(HPPA)
 	long pc;
 
