diff -ur utempter-0.5.2.orig/Makefile utempter-0.5.2/Makefile
--- utempter-0.5.2.orig/Makefile	Thu Mar 25 23:35:43 1999
+++ utempter-0.5.2/Makefile	Sun May 19 05:25:47 2002
@@ -23,11 +23,11 @@
 	$(CC) -c $(CFLAGS) -fPIC $< -o $@
 
 install:
-	mkdir -p $(RPM_BUILD_ROOT)/usr/sbin
+	mkdir -p $(RPM_BUILD_ROOT)/usr/libexec/utempter
 	mkdir -p $(RPM_BUILD_ROOT)/usr/lib
 	mkdir -p $(RPM_BUILD_ROOT)/usr/include
-	install -m 4755 utempter $(RPM_BUILD_ROOT)/usr/sbin
-	install -m 644 utempter.h $(RPM_BUILD_ROOT)/usr/include
+	install -m 700 utempter $(RPM_BUILD_ROOT)/usr/libexec/utempter/
+	install -m 644 utempter.h $(RPM_BUILD_ROOT)/usr/include/
 	install -m 644 $(SHAREDLIB) $(RPM_BUILD_ROOT)/usr/lib/$(SHAREDLIB).$(VERSION)
 	ln -sf $(SHAREDLIB).$(VERSION) $(RPM_BUILD_ROOT)/usr/lib/$(SHAREDLIB)
 
diff -ur utempter-0.5.2.orig/utmpintf.c utempter-0.5.2/utmpintf.c
--- utempter-0.5.2.orig/utmpintf.c	Fri Jun  4 15:29:37 1999
+++ utempter-0.5.2/utmpintf.c	Sun May 19 05:25:59 2002
@@ -8,12 +8,15 @@
 
 #include "utempter.h"
 
+#define UTEMPTER_PATH_1			"/usr/libexec/utempter/utempter"
+#define UTEMPTER_PATH_2			"/usr/libexec/screen/utempter"
+
 static char * ptyName = NULL;
 static int masterFd;
 
 void addToUtmp(const char *pty, const char *hostname, int cmdfd)
 {
-    char * argv[] = { "/usr/sbin/utempter", "-a", pty, hostname, NULL};
+    char * argv[] = { UTEMPTER_PATH_1, "-a", pty, hostname, NULL};
     pid_t child;
     int status;
     void (*osig) (int) = signal(SIGCHLD, SIG_IGN);
@@ -25,6 +28,8 @@
 	dup2(cmdfd, 1);
 	dup2(cmdfd, 3);
 	execv(argv[0], argv);
+	argv[0] = UTEMPTER_PATH_2;
+	execv(argv[0], argv);
 	exit(1);
     }
 
@@ -46,7 +51,7 @@
 }
 
 void removeLineFromUtmp(const char * pty, int fd) {
-    char * argv[] = { "/usr/sbin/utempter", "-d", pty, NULL};
+    char * argv[] = { UTEMPTER_PATH_1, "-d", pty, NULL};
     pid_t child;
     void (*osig) (int) = signal(SIGCHLD, SIG_IGN);
 
@@ -56,6 +61,8 @@
 	dup2(fd, 0);
 	dup2(fd, 1);
 	dup2(fd, 3);
+	execv(argv[0], argv);
+	argv[0] = UTEMPTER_PATH_2;
 	execv(argv[0], argv);
 	fprintf(stderr, "EXEC FAILED!\n");
     }
