--- ed-0.2/buf.c.orig   Thu Nov 23 13:51:39 2000
+++ ed-0.2/buf.c        Thu Nov 23 15:06:06 2000
@@ -201,12 +201,12 @@
 open_sbuf ()
 {
   char *mktemp ();
-  int u;
+  int u, fd;
 
   isbinary = newline_added = 0;
   u = umask(077);
   strcpy (sfn, "/tmp/ed.XXXXXX");
-  if (mktemp (sfn) == NULL || (sfp = fopen (sfn, "w+")) == NULL)
+  if ((fd=mkstemp (sfn)) == -1 || (sfp = fdopen (fd, "w+")) == NULL)
     {
       fprintf (stderr, "%s: %s\n", sfn, strerror (errno));
       sprintf (errmsg, "Cannot open temp file");

