diff -uNrp ed-1.5.old/main.c ed-1.5/main.c
--- ed-1.5.old/main.c	2010-10-06 15:41:04 +0000
+++ ed-1.5/main.c	2010-10-06 15:42:18 +0000
@@ -33,17 +33,18 @@
  *
  */
 
+#define _GNU_SOURCE
 #include <stdio.h>
 #include <stdlib.h>
 #include <string.h>
 #include <sys/stat.h>
 #include <locale.h>
+#include <errno.h>
 
 #include "carg_parser.h"
 #include "ed.h"
 
 
-static const char * invocation_name = 0;
 static const char * const Program_name = "GNU Ed";
 static const char * const program_name = "ed";
 static const char * const program_year = "2010";
@@ -61,7 +62,7 @@ bool traditional( void ) { return tradit
 static void show_help( void )
   {
   printf( "%s - The GNU line editor.\n", Program_name );
-  printf( "\nUsage: %s [options] [file]\n", invocation_name );
+  printf( "\nUsage: %s [options] [file]\n", program_invocation_short_name );
   printf( "\nOptions:\n" );
   printf( "  -h, --help                 display this help and exit\n" );
   printf( "  -V, --version              output version information and exit\n" );
@@ -108,8 +109,8 @@ static void show_error( const char * con
     if( errcode > 0 ) fprintf( stderr, ": %s", strerror( errcode ) );
     fprintf( stderr, "\n" );
     }
-  if( help && invocation_name && invocation_name[0] )
-    fprintf( stderr, "Try `%s --help' for more information.\n", invocation_name );
+  if( help )
+    fprintf( stderr, "Try `%s --help' for more information.\n", program_invocation_short_name );
   }
 
 
@@ -156,7 +157,6 @@ int main( const int argc, const char * c
     { show_error( "Memory exhausted.", 0, false ); return 1; }
   if( ap_error( &parser ) )				/* bad option */
     { show_error( ap_error( &parser ), 0, true ); return 1; }
-  invocation_name = argv[0];
 
   for( argind = 0; argind < ap_arguments( &parser ); ++argind )
     {
