diff -puNr elinks-0.9.1.orig/src/viewer/text/view.c elinks-0.9.1/src/viewer/text/view.c
--- elinks-0.9.1.orig/src/viewer/text/view.c	2006-09-23 17:09:58 +0000
+++ elinks-0.9.1/src/viewer/text/view.c	2006-09-23 21:35:42 +0000
@@ -327,7 +327,7 @@ page_down(struct session *ses, struct do
 	assert(ses && doc_view && doc_view->vs);
 	if_assert_failed return;
 
-	newpos = doc_view->vs->y + doc_view->document->options.height;
+	newpos = doc_view->vs->y + doc_view->height;
 	if (newpos < doc_view->document->height)
 		doc_view->vs->y = newpos;
 
