--- elinks-0.9.1/src/terminal/terminal.c.orig	Wed Jan  7 19:18:01 2004
+++ elinks-0.9.1/src/terminal/terminal.c	Tue Jan 27 23:54:22 2004
@@ -226,6 +226,7 @@
 #if defined(HAVE_SETPGID) && !defined(BEOS) && !defined(HAVE_BEGINTHREAD)
 	if (path[0] == 2) setpgid(0, 0);
 #endif
+	/* XXX: This interface should be re-arranged to use execvp(3). */
 	exe(path + 1);
 	if (path[plen]) unlink(path + plen);
 }
--- elinks-0.9.1/src/util/conv.c.orig	Sat Jan  3 15:07:43 2004
+++ elinks-0.9.1/src/util/conv.c	Tue Jan 27 23:54:22 2004
@@ -294,10 +294,16 @@
 struct string *
 add_shell_safe_to_string(struct string *string, unsigned char *cmd, int cmdlen)
 {
+	int prev_safe = 0;
+
 	for (; cmdlen; cmdlen--, cmd++) {
-		if (is_safe_in_shell(*cmd)) {
+		if ((*cmd == '-' && prev_safe) ||
+		    (prev_safe = is_safe_in_shell(*cmd))) {
 			add_char_to_string(string, *cmd);
 		} else {
+			/* XXX: Not all programs we might exec are capable of
+			 * decoding these.  For some, we should just report
+			 * an error rather than exec with an encoded string. */
 			add_char_to_string(string, '%');
 			add_char_to_string(string, hx((*cmd & 0xf0) >> 4));
 			add_char_to_string(string, hx(*cmd & 0x0f));
--- elinks-0.9.1/src/util/conv.h.orig	Sat Jan  3 15:07:43 2004
+++ elinks-0.9.1/src/util/conv.h	Tue Jan 27 23:54:22 2004
@@ -16,7 +16,13 @@
 static inline int
 is_safe_in_shell(unsigned char c)
 {
-	return c == '@' || c == '+' || c == '.' || c == '/' || c == ':' || isA(c);
+	/* Note: '-' is often used to indicate a command-line option and thus
+	 * is not always safe. */
+	return c == '@' || c == '+' || c == '.' || c == '/' || c == ':'
+		|| c == '_'
+		|| (c >= 'A' && c <= 'Z')
+		|| (c >= 'a' && c <= 'z')
+		|| (c >= '0' && c <= '9');
 }
 
 
--- elinks-0.9.1/src/setup.h.orig	Mon Dec  8 00:39:59 2003
+++ elinks-0.9.1/src/setup.h	Sun Jan 25 14:33:30 2004
@@ -96,9 +96,9 @@
 /* Default external programs for protocol.user.* autocreated options */
 #define DEFAULT_AC_OPT_MAILTO		"mutt %h -s \"%s\""
 #define DEFAULT_AC_OPT_TELNET		"telnet %h %p"
-#define DEFAULT_AC_OPT_TN3270		"tn3270 %h %p"
-#define DEFAULT_AC_OPT_GOPHER		"lynx %u"
-#define DEFAULT_AC_OPT_NEWS		"lynx %u"
-#define DEFAULT_AC_OPT_IRC		"irc %u"
+#define DEFAULT_AC_OPT_TN3270		""
+#define DEFAULT_AC_OPT_GOPHER		""
+#define DEFAULT_AC_OPT_NEWS		""
+#define DEFAULT_AC_OPT_IRC		""
 
 #endif
