diff -ur elinks-0.9.1.orig/src/config/options.inc elinks-0.9.1/src/config/options.inc
--- elinks-0.9.1.orig/src/config/options.inc	Thu Jan  8 22:00:01 2004
+++ elinks-0.9.1/src/config/options.inc	Sun Feb  1 05:31:58 2004
@@ -685,11 +685,11 @@
 		"not be enabled on all servers.")),
 
 	INIT_OPT_STRING("protocol.http", N_("User-agent identification"),
-		"user_agent", 0, "ELinks/%v (textmode; %s; %t)",
+		"user_agent", 0, "ELinks/%v (textmode; Linux; %t)",
 		N_("Change the User Agent ID. That means identification string, which\n"
 		"is sent to HTTP server when a document is requested.\n"
 		"%v in the string means ELinks version\n"
-		"%s in the string means system identification\n"
+		"%s in the string means system identification (uname -srm)\n"
 		"%t in the string means size of the terminal\n"
 		"Use \" \" if you don't want any User-Agent header to be sent at all.")),
 
