diff -urp elinks-0.9.0.orig/src/sched/download.c elinks-0.9.0/src/sched/download.c
--- elinks-0.9.0.orig/src/sched/download.c	Mon Dec 22 02:15:02 2003
+++ elinks-0.9.0/src/sched/download.c	Sun Jan 25 11:03:21 2004
@@ -536,23 +536,22 @@ create_download_file_do(struct terminal 
 	unsigned char *wd;
 	int h = -1;
 	int saved_errno;
-#ifdef NO_FILE_SECURITY
-	int sf = 0;
-#else
-	int sf = cdf_hop->safe;
-#endif
 
 	if (!file) goto finish;
 
 	wd = get_cwd();
 	set_cwd(term->cwd);
 
-	/* O_APPEND means repositioning at the end of file before each write(),
-	 * thus ignoring seek()s and that can hide mysterious bugs. IMHO.
-	 * --pasky */
-	h = open(file, O_CREAT | O_WRONLY | (resume ? 0 : O_TRUNC)
-			| (sf && !resume ? O_EXCL : 0),
-		 sf ? 0600 : 0666);
+	if (cdf_hop->safe) {
+		h = mkstemp(file);
+	} else {
+		/* O_APPEND means repositioning at the end of file before each
+		 * write(), thus ignoring seek()s and that can hide mysterious
+		 * bugs. IMHO.
+		 * --pasky */
+		h = open(file,
+			O_WRONLY | (resume ? 0 : (O_CREAT | O_TRUNC)), 0666);
+	}
 	saved_errno = errno; /* Saved in case of ... --Zas */
 
 	if (wd) {
@@ -600,38 +599,6 @@ finish:
 }
 
 
-static unsigned char *
-get_temp_name(unsigned char *url)
-{
-	struct string name;
-	unsigned char *extension;
-	/* FIXME
-	 * We use tempnam() here, which is unsafe (race condition), for now.
-	 * This should be changed at some time, but it needs an in-depth work
-	 * of whole download code. --Zas */
-	unsigned char *nm = tempnam(NULL, ELINKS_TEMPNAME_PREFIX);
-
-	if (!nm) return NULL;
-
-	if (!init_string(&name)) {
-		mem_free(nm);
-		return NULL;
-	}
-
-	add_to_string(&name, nm);
-	free(nm);
-
-	extension = get_extension_from_url(url);
-	if (extension) {
-		add_char_to_string(&name, '.');
-		add_shell_safe_to_string(&name, extension, strlen(extension));
-		mem_free(extension);
-	}
-
-	return name.source;
-}
-
-
 unsigned char *
 subst_file(unsigned char *prog, unsigned char *file)
 {
@@ -788,9 +755,15 @@ continue_download(void *data, unsigned c
 	}
 
 	if (tq->prog) {
-		/* FIXME: get_temp_name() calls tempnam(). --Zas */
-		file = get_temp_name(tq->url);
-		if (!file) {
+		unsigned char *tmpdir;
+
+		tmpdir = getenv("TMPDIR");
+		if (!tmpdir || !*tmpdir) tmpdir = "/tmp";
+
+		file = straconcat(tmpdir,
+			"/" ELINKS_TEMPNAME_PREFIX "-XXXXXX", NULL);
+
+		if(!file) {
 			mem_free(codw_hop);
 			tp_cancel(tq);
 			return;
@@ -832,9 +805,9 @@ continue_download_do(struct terminal *te
 	file_download->ses = codw_hop->tq->ses;
 
 	if (codw_hop->tq->prog) {
-		file_download->prog = subst_file(codw_hop->tq->prog, codw_hop->file);
+		file_download->prog = subst_file(codw_hop->tq->prog, codw_hop->real_file);
 		file_download->delete = 1;
-		mem_free(codw_hop->file);
+		mem_free(codw_hop->real_file);
 		mem_free(codw_hop->tq->prog);
 		codw_hop->tq->prog = NULL;
 	}
@@ -854,7 +827,7 @@ continue_download_do(struct terminal *te
 
 cancel:
 	tp_cancel(codw_hop->tq);
-	if (codw_hop->tq->prog && codw_hop->file) mem_free(codw_hop->file);
+	if (codw_hop->tq->prog && codw_hop->real_file) mem_free(codw_hop->real_file);
 	if (file_download) {
 		if (file_download->url) mem_free(file_download->url);
 		mem_free(file_download);
diff -urp elinks-0.9.0.orig/src/viewer/text/textarea.c elinks-0.9.0/src/viewer/text/textarea.c
--- elinks-0.9.0.orig/src/viewer/text/textarea.c	Mon Dec 22 09:53:29 2003
+++ elinks-0.9.0/src/viewer/text/textarea.c	Sun Jan 25 10:50:38 2004
@@ -287,12 +287,10 @@ textarea_edit(int op, struct terminal *t
 		int h;
 
 		tmpdir = getenv("TMPDIR");
-		if (!tmpdir || !*tmpdir) tmpdir = getenv("TMP");
-		if (!tmpdir || !*tmpdir) tmpdir = getenv("TEMPDIR");
-		if (!tmpdir || !*tmpdir) tmpdir = getenv("TEMP");
 		if (!tmpdir || !*tmpdir) tmpdir = "/tmp";
 
-		fn = straconcat(tmpdir, "/linksarea-XXXXXX", NULL);
+		fn = straconcat(tmpdir,
+			"/" ELINKS_TEMPNAME_PREFIX "area-XXXXXX", NULL);
 		if (!fn) goto free_and_return;
 
 		h = mkstemp(fn);
