diff -ur elinks-0.9.0.orig/src/osdep/osdep.c elinks-0.9.0/src/osdep/osdep.c
--- elinks-0.9.0.orig/src/osdep/osdep.c	Sat Nov  8 05:51:27 2003
+++ elinks-0.9.0/src/osdep/osdep.c	Sun Jan 25 12:30:11 2004
@@ -274,6 +274,7 @@
 void
 set_window_title(unsigned char *title)
 {
+#if 0
 	unsigned char *s;
 	int xsize, ysize;
 	register int j = 0;
@@ -319,6 +320,7 @@
 	fflush(stdout);
 
 	mem_free(s);
+#endif
 }
 
 #ifdef HAVE_X11
