--- elinks-0.9.1/src/sched/download.c.orig	Sat Jan 17 09:57:11 2004
+++ elinks-0.9.1/src/sched/download.c	Sun Feb  1 00:52:35 2004
@@ -536,23 +536,22 @@
 	unsigned char *wd;
 	int h = -1;
 	int saved_errno;
-#ifdef NO_FILE_SECURITY
-	int sf = 0;
-#else
-	int sf = cdf_hop->safe;
-#endif
 
 	if (!file) goto finish;
 
 	wd = get_cwd();
 	set_cwd(term->cwd);
 
-	/* O_APPEND means repositioning at the end of file before each write(),
-	 * thus ignoring seek()s and that can hide mysterious bugs. IMHO.
-	 * --pasky */
-	h = open(file, O_CREAT | O_WRONLY | (resume ? 0 : O_TRUNC)
-			| (sf && !resume ? O_EXCL : 0),
-		 sf ? 0600 : 0666);
+	if (cdf_hop->safe) {
+		h = mkstemp(file);
+	} else {
+		/* O_APPEND means repositioning at the end of file before each
+		 * write(), thus ignoring seek()s and that can hide mysterious
+		 * bugs. IMHO.
+		 * --pasky */
+		h = open(file,
+			O_WRONLY | (resume ? 0 : (O_CREAT | O_TRUNC)), 0666);
+	}
 	saved_errno = errno; /* Saved in case of ... --Zas */
 
 	if (wd) {
@@ -600,38 +599,6 @@
 }
 
 
-static unsigned char *
-get_temp_name(unsigned char *url)
-{
-	struct string name;
-	unsigned char *extension;
-	/* FIXME
-	 * We use tempnam() here, which is unsafe (race condition), for now.
-	 * This should be changed at some time, but it needs an in-depth work
-	 * of whole download code. --Zas */
-	unsigned char *nm = tempnam(NULL, ELINKS_TEMPNAME_PREFIX);
-
-	if (!nm) return NULL;
-
-	if (!init_string(&name)) {
-		mem_free(nm);
-		return NULL;
-	}
-
-	add_to_string(&name, nm);
-	free(nm);
-
-	extension = get_extension_from_url(url);
-	if (extension) {
-		add_char_to_string(&name, '.');
-		add_shell_safe_to_string(&name, extension, strlen(extension));
-		mem_free(extension);
-	}
-
-	return name.source;
-}
-
-
 unsigned char *
 subst_file(unsigned char *prog, unsigned char *file)
 {
@@ -774,7 +741,6 @@
 struct codw_hop {
 	struct tq *tq;
 	unsigned char *real_file;
-	unsigned char *file;
 };
 
 static void
@@ -792,9 +758,15 @@
 	}
 
 	if (tq->prog) {
-		/* FIXME: get_temp_name() calls tempnam(). --Zas */
-		file = get_temp_name(tq->url);
-		if (!file) {
+		unsigned char *tmpdir;
+
+		tmpdir = getenv("TMPDIR");
+		if (!tmpdir || !*tmpdir) tmpdir = "/tmp";
+
+		file = straconcat(tmpdir,
+			"/" ELINKS_TEMPNAME_PREFIX "-XXXXXX", NULL);
+
+		if(!file) {
 			mem_free(codw_hop);
 			tp_cancel(tq);
 			return;
@@ -802,12 +774,13 @@
 	}
 
 	codw_hop->tq = tq;
-	codw_hop->file = file;
 
 	kill_downloads_to_file(file);
 
 	create_download_file(tq->ses->tab->term, file, &codw_hop->real_file,
 			     !!tq->prog, 0, continue_download_do, codw_hop);
+
+	if(file) mem_free(file);
 }
 
 static void
@@ -836,9 +809,8 @@
 	file_download->ses = codw_hop->tq->ses;
 
 	if (codw_hop->tq->prog) {
-		file_download->prog = subst_file(codw_hop->tq->prog, codw_hop->file);
+		file_download->prog = subst_file(codw_hop->tq->prog, codw_hop->real_file);
 		file_download->delete = 1;
-		mem_free(codw_hop->file);
 		mem_free(codw_hop->tq->prog);
 		codw_hop->tq->prog = NULL;
 	}
@@ -861,7 +833,6 @@
 
 cancel:
 	tp_cancel(codw_hop->tq);
-	if (codw_hop->tq->prog && codw_hop->file) mem_free(codw_hop->file);
 	if (file_download) {
 		if (file_download->url) mem_free(file_download->url);
 		mem_free(file_download);
--- elinks-0.9.1/src/viewer/text/textarea.c.orig	Thu Jan  1 09:16:49 2004
+++ elinks-0.9.1/src/viewer/text/textarea.c	Wed Jan 28 01:59:51 2004
@@ -287,12 +287,10 @@
 		int h;
 
 		tmpdir = getenv("TMPDIR");
-		if (!tmpdir || !*tmpdir) tmpdir = getenv("TMP");
-		if (!tmpdir || !*tmpdir) tmpdir = getenv("TEMPDIR");
-		if (!tmpdir || !*tmpdir) tmpdir = getenv("TEMP");
 		if (!tmpdir || !*tmpdir) tmpdir = "/tmp";
 
-		fn = straconcat(tmpdir, "/linksarea-XXXXXX", NULL);
+		fn = straconcat(tmpdir,
+			"/" ELINKS_TEMPNAME_PREFIX "area-XXXXXX", NULL);
 		if (!fn) goto free_and_return;
 
 		h = mkstemp(fn);
