diff -ur elinks-0.9.1.orig/config.h.in elinks-0.9.1/config.h.in
--- elinks-0.9.1.orig/config.h.in	Thu Jan 22 22:28:19 2004
+++ elinks-0.9.1/config.h.in	Wed Jan 28 02:54:56 2004
@@ -1,5 +1,8 @@
 /* config.h.in.  Generated automatically from configure.in by autoheader.  */
 
+#ifndef EL__DOTDOT_CONFIG_H
+#define EL__DOTDOT_CONFIG_H
+
 /* Define if using alloca.c.  */
 #undef C_ALLOCA
 
@@ -476,3 +479,5 @@
 
 
 #include "../feature.h"
+
+#endif
