--- elinks-0.9.1/src/osdep/osdep.c.orig	Thu Jan  1 15:54:38 2004
+++ elinks-0.9.1/src/osdep/osdep.c	Tue Jan 27 23:56:23 2004
@@ -284,6 +284,7 @@
 void
 set_window_title(unsigned char *title)
 {
+#if 0
 	unsigned char *s;
 	int xsize, ysize;
 	register int j = 0;
@@ -329,6 +330,7 @@
 	fflush(stdout);
 
 	mem_free(s);
+#endif
 }
 
 #ifdef HAVE_X11
--- elinks-0.9.1.orig/src/config/options.inc	Thu Jan  8 22:00:01 2004
+++ elinks-0.9.1/src/config/options.inc	Sun Feb  1 05:31:58 2004
@@ -1255,10 +1255,11 @@
 
 
 	INIT_OPT_BOOL("ui", N_("Set window title"),
-		"window_title", 0, 1,
+		"window_title", 0, 0,
 		N_("Set the window title when running in a windowing environment\n"
 		"in an xterm-like terminal. This way the document's title is\n"
-		"shown on the window titlebar.")),
+		"shown on the window titlebar. This functionality is currently\n"
+		"disabled at compile time and may be re-enabled in the future.")),
 
 
 
