diff -urp elinks-0.9.0.orig/src/viewer/text/textarea.c elinks-0.9.0/src/viewer/text/textarea.c
--- elinks-0.9.0.orig/src/viewer/text/textarea.c	Mon Dec 22 09:53:29 2003
+++ elinks-0.9.0/src/viewer/text/textarea.c	Sun Jan 25 11:12:06 2004
@@ -304,7 +304,7 @@ textarea_edit(int op, struct terminal *t
 		fwrite(fs->value, strlen(fs->value), 1, taf);
 		fclose(taf);
 
-		if (!ed || !*ed) ed = "vi";
+		if (!ed || !*ed) ed = "vitmp";
 
 		ex = straconcat(ed, " ", fn, NULL);
 		if (!ex) {
