diff -puNr elinks-0.9.1.orig/src/dialogs/status.c elinks-0.9.1/src/dialogs/status.c
--- elinks-0.9.1.orig/src/dialogs/status.c	2004-01-06 06:28:15 +0000
+++ elinks-0.9.1/src/dialogs/status.c	2006-09-23 21:57:15 +0000
@@ -169,10 +170,6 @@ update_status(void)
 
 		if (show_tabs(show_tabs_bar, tabs) != status->show_tabs_bar) {
 			status->show_tabs_bar = show_tabs(show_tabs_bar, tabs);
-			/* Force the current document to be rerendered so the
-			 * document view and document height is updated to fit
-			 * into the new dimensions. Related to bug 87. */
-			render_document_frames(ses);
 			dirty = 1;
 		}
 #ifdef CONFIG_LEDS
@@ -186,6 +183,11 @@ update_status(void)
 
 		if (!dirty) continue;
 
+		/* Force the current document to be rerendered so the
+		 * document view and document height is updated to fit
+		 * into the new dimensions. Related to bug 87. */
+		render_document_frames(ses);
+
 		set_screen_dirty(term->screen, 0, term->height);
 	}
 }
