--- grep-2.5.1a/src/grep.c
+++ grep-2.5.1a/src/grep.c
@@ -88,6 +88,12 @@ static struct exclude *included_patterns
 static char const short_options[] =
 "0123456789A:B:C:D:EFGHIPUVX:abcd:e:f:hiKLlm:noqRrsuvwxyZz";
 
+/* Default for `file_list' if no files are given on the command line. */
+static char *stdin_argv[] =
+{
+  "-", NULL
+};
+
 /* Non-boolean long options that have no corresponding short equivalents.  */
 enum
 {
@@ -319,7 +325,10 @@ fillbuf (size_t save, struct stats const
       /* Grow newsize until it is at least as great as minsize.  */
       for (newsize = bufalloc - pagesize - 1; newsize < minsize; newsize *= 2)
 	if (newsize * 2 < newsize || newsize * 2 + pagesize + 1 < newsize * 2)
-	  xalloc_die ();
+	  {
+	    errno = ENOMEM;
+	    return 0;
+	  }
 
       /* Try not to allocate more memory than the file size indicates,
 	 as that might cause unnecessary memory exhaustion if the file
@@ -341,7 +350,9 @@ fillbuf (size_t save, struct stats const
 	 for byte sentinels fore and aft.  */
       newalloc = newsize + pagesize + 1;
 
-      newbuf = bufalloc < newalloc ? xmalloc (bufalloc = newalloc) : buffer;
+      newbuf = bufalloc < newalloc ? malloc (bufalloc = newalloc) : buffer;
+      if (!newbuf)
+        return 0;
       readbuf = ALIGN_TO (newbuf + 1 + save, pagesize);
       bufbeg = readbuf - save;
       memmove (bufbeg, buffer + saved_offset, save);
@@ -1305,6 +1316,7 @@ main (int argc, char **argv)
   FILE *fp;
   extern char *optarg;
   extern int optind;
+  char **file_list;
 
   initialize_main (&argc, &argv);
   program_name = argv[0];
@@ -1764,29 +1776,29 @@ warranty; not even for MERCHANTABILITY o
   if (max_count == 0)
     exit (1);
 
-  if (optind < argc)
+  file_list = (optind == argc ? stdin_argv : &argv[optind]);
+
+  status = 1;
+  while (1)
     {
-	status = 1;
-	do
+      char *file = *file_list++;
+
+      if (file == NULL)
+	break;
+
+      if ((included_patterns || excluded_patterns)
+	  && !isdir (file))
 	{
-	  char *file = argv[optind];
-	  if ((included_patterns || excluded_patterns)
-	      && !isdir (file))
-	    {
-	      if (included_patterns &&
-		  ! excluded_filename (included_patterns, file, 0))
-		continue;
-	      if (excluded_patterns &&
-		  excluded_filename (excluded_patterns, file, 0))
-		continue;
-	    }
-	  status &= grepfile (strcmp (file, "-") == 0 ? (char *) NULL : file,
-			      &stats_base);
+	  if (included_patterns &&
+	      ! excluded_filename (included_patterns, file, 0))
+	    continue;
+	  if (excluded_patterns &&
+	      excluded_filename (excluded_patterns, file, 0))
+	    continue;
 	}
-	while ( ++optind < argc);
+      status &= grepfile (strcmp (file, "-") == 0
+			  ? (char *) NULL : file, &stats_base);
     }
-  else
-    status = grepfile ((char *) NULL, &stats_base);
 
   /* We register via atexit() to test stdout.  */
   exit (errseen ? 2 : status);
