--- gnupg-1.4.2/g10/keygen.c.orig	2005-06-21 12:22:17.000000000 +0000
+++ gnupg-1.4.2/g10/keygen.c	2005-11-16 23:03:48.000000000 +0000
@@ -3243,15 +3243,21 @@
 static int
 write_keyblock( IOBUF out, KBNODE node )
 {
-    for( ; node ; node = node->next ) {
-	int rc = build_packet( out, node->pkt );
-	if( rc ) {
-	    log_error("build_packet(%d) failed: %s\n",
+  for( ; node ; node = node->next )
+    {
+      if(!is_deleted_kbnode(node))
+	{
+	  int rc = build_packet( out, node->pkt );
+	  if( rc )
+	    {
+	      log_error("build_packet(%d) failed: %s\n",
 			node->pkt->pkttype, g10_errstr(rc) );
-	    return G10ERR_WRITE_FILE;
+	      return G10ERR_WRITE_FILE;
+	    }
 	}
     }
-    return 0;
+
+  return 0;
 }
 
 
