--- gnupg/g10/getkey.c.orig	2006-04-03 08:13:20 +0000
+++ gnupg/g10/getkey.c	2006-06-11 15:35:25 +0000
@@ -173,7 +173,7 @@ static const char *
 get_primary_uid ( KBNODE keyblock, size_t *uidlen )
 {
     KBNODE k;
-    const char *s;
+    static const char *s = NULL;
 
     for (k=keyblock; k; k=k->next ) {
         if ( k->pkt->pkttype == PKT_USER_ID
@@ -183,9 +183,8 @@ get_primary_uid ( KBNODE keyblock, size_
             return k->pkt->pkt.user_id->name;
         }
     } 
-    /* fixme: returning translatable constants instead of a user ID is 
-     * not good because they are probably not utf-8 encoded. */
-    s = _("[User ID not found]");
+    if ( !s )
+        s = native_to_utf8 (_("[User ID not found]"));
     *uidlen = strlen (s);
     return s;
 }
@@ -2886,7 +2885,7 @@ get_user_id( u32 *keyid, size_t *rn )
             }
         }
     } while( ++pass < 2 && !get_pubkey( NULL, keyid ) );
-    p = xstrdup( _("[User ID not found]") );
+    p = native_to_utf8( _("[User ID not found]") );
     *rn = strlen(p);
     return p;
 }
