diff -upk.orig gnupg-1.4.3.orig/g10/gpg.c gnupg-1.4.3/g10/gpg.c
--- gnupg-1.4.3.orig/g10/gpg.c	2006-04-03 08:28:44 +0000
+++ gnupg-1.4.3/g10/gpg.c	2006-06-11 21:59:10 +0000
@@ -3106,7 +3106,7 @@ main (int argc, char **argv )
 	break;
       case aExportOwnerTrust: rc = setup_trustdb( 0, trustdb_name ); break;
       case aListTrustDB: rc = setup_trustdb( argc? 1:0, trustdb_name ); break;
-      default: rc = setup_trustdb(1, trustdb_name ); break;
+      default: rc = setup_trustdb( opt.trust_model != TM_ALWAYS, trustdb_name ); break;
     }
     if( rc )
 	log_error(_("failed to initialize the TrustDB: %s\n"), g10_errstr(rc));
diff -upk.orig gnupg-1.4.3.orig/g10/trustdb.c gnupg-1.4.3/g10/trustdb.c
--- gnupg-1.4.3.orig/g10/trustdb.c	2006-03-07 23:20:58 +0000
+++ gnupg-1.4.3/g10/trustdb.c	2006-06-11 21:59:10 +0000
@@ -1057,6 +1057,9 @@ get_validity (PKT_public_key *pk, PKT_us
   u32 kid[2];
   PKT_public_key *main_pk;
 
+  if (opt.trust_model == TM_ALWAYS)
+    return TRUST_ULTIMATE;
+
   if(uid)
     namehash_from_uid(uid);
 
