diff -urN rpm-3.0.5.org/configure.in rpm-3.0.5/configure.in
--- rpm-3.0.5.org/configure.in	Sat Jul 22 14:37:12 2000
+++ rpm-3.0.5/configure.in	Sun Sep  3 16:37:40 2000
@@ -844,7 +844,7 @@
 esac
 RPMCANONVENDOR="$build_vendor"
 case "${build_vendor}" in
-unknown|pc|redhat|pld)
+pc|redhat|pld)
 	test -f /etc/redhat-release && RPMCANONVENDOR=redhat
 	test -f /etc/pld-release && RPMCANONVENDOR=pld
 	;;
