diff -ur rpm-3.0.5.org/Makefile.am rpm-3.0.5/Makefile.am
--- rpm-3.0.5.org/Makefile.am	Sun Jul 16 22:00:30 2000
+++ rpm-3.0.5/Makefile.am	Sun Aug  6 18:36:18 2000
@@ -60,31 +60,31 @@
 	@@LN_S@ rpm/rpmrc $(DESTDIR)/$(libdir)/rpmrc
 	@rm -f $(DESTDIR)/$(libdir)/rpmpopt
 	@@LN_S@ rpm/rpmpopt $(DESTDIR)/$(libdir)/rpmpopt
-	@$(mkinstalldirs) $(DESTDIR)/$(prefix)/src/redhat/BUILD
+	@$(mkinstalldirs) $(DESTDIR)/$(prefix)/src/RPM/BUILD
 	@case "@build_cpu@" in \
-	*86) $(mkinstalldirs) $(DESTDIR)/$(prefix)/src/redhat/RPMS/i386 ;;\
-	*86)	$(mkinstalldirs) $(DESTDIR)/$(prefix)/src/redhat/RPMS/i386;\
-		$(mkinstalldirs) $(DESTDIR)/$(prefix)/src/redhat/RPMS/i486;\
-		$(mkinstalldirs) $(DESTDIR)/$(prefix)/src/redhat/RPMS/i586;\
-		$(mkinstalldirs) $(DESTDIR)/$(prefix)/src/redhat/RPMS/i686;\
-		$(mkinstalldirs) $(DESTDIR)/$(prefix)/src/redhat/RPMS/athlon ;;\
-	alpha*) $(mkinstalldirs) $(DESTDIR)/$(prefix)/src/redhat/RPMS/alpha ;;\
-	sparc*) $(mkinstalldirs) $(DESTDIR)/$(prefix)/src/redhat/RPMS/sparc;\
-		$(mkinstalldirs) $(DESTDIR)/$(prefix)/src/redhat/RPMS/sparcv9;\
-		$(mkinstalldirs) $(DESTDIR)/$(prefix)/src/redhat/RPMS/sparc64 ;;\
-	powerpc*) $(mkinstalldirs) $(DESTDIR)/$(prefix)/src/redhat/RPMS/ppc ;;\
+	*86) $(mkinstalldirs) $(DESTDIR)/$(prefix)/src/RPM/RPMS/i386 ;;\
+	*86)	$(mkinstalldirs) $(DESTDIR)/$(prefix)/src/RPM/RPMS/i386;\
+		$(mkinstalldirs) $(DESTDIR)/$(prefix)/src/RPM/RPMS/i486;\
+		$(mkinstalldirs) $(DESTDIR)/$(prefix)/src/RPM/RPMS/i586;\
+		$(mkinstalldirs) $(DESTDIR)/$(prefix)/src/RPM/RPMS/i686;\
+		$(mkinstalldirs) $(DESTDIR)/$(prefix)/src/RPM/RPMS/athlon ;;\
+	alpha*) $(mkinstalldirs) $(DESTDIR)/$(prefix)/src/RPM/RPMS/alpha ;;\
+	sparc*) $(mkinstalldirs) $(DESTDIR)/$(prefix)/src/RPM/RPMS/sparc;\
+		$(mkinstalldirs) $(DESTDIR)/$(prefix)/src/RPM/RPMS/sparcv9;\
+		$(mkinstalldirs) $(DESTDIR)/$(prefix)/src/RPM/RPMS/sparc64 ;;\
+	powerpc*) $(mkinstalldirs) $(DESTDIR)/$(prefix)/src/RPM/RPMS/ppc ;;\
 	esac
 	@case "@build_os@" in \
-	mint) $(mkinstalldirs) $(DESTDIR)/$(prefix)/src/redhat/RPMS/m68kmint ;;\
+	mint) $(mkinstalldirs) $(DESTDIR)/$(prefix)/src/RPM/RPMS/m68kmint ;;\
 	solaris*|linux*) \
 	   chmod u+x ./installplatform; DESTDIR="$(DESTDIR)" pkglibdir="$(pkglibdir)" ./installplatform rpmrc macros platform ; \
-	   $(mkinstalldirs) $(DESTDIR)/$(prefix)/src/redhat/RPMS/@build_cpu@ ;;\
-	*) $(mkinstalldirs) $(DESTDIR)/$(prefix)/src/redhat/RPMS/@build_cpu@ ;;\
+	   $(mkinstalldirs) $(DESTDIR)/$(prefix)/src/RPM/RPMS/@build_cpu@ ;;\
+	*) $(mkinstalldirs) $(DESTDIR)/$(prefix)/src/RPM/RPMS/@build_cpu@ ;;\
 	esac
-	@$(mkinstalldirs) $(DESTDIR)/$(prefix)/src/redhat/RPMS/noarch
-	@$(mkinstalldirs) $(DESTDIR)/$(prefix)/src/redhat/SOURCES
-	@$(mkinstalldirs) $(DESTDIR)/$(prefix)/src/redhat/SPECS
-	@$(mkinstalldirs) $(DESTDIR)/$(prefix)/src/redhat/SRPMS
+	@$(mkinstalldirs) $(DESTDIR)/$(prefix)/src/RPM/RPMS/noarch
+	@$(mkinstalldirs) $(DESTDIR)/$(prefix)/src/RPM/SOURCES
+	@$(mkinstalldirs) $(DESTDIR)/$(prefix)/src/RPM/SPECS
+	@$(mkinstalldirs) $(DESTDIR)/$(prefix)/src/RPM/SRPMS
 	@$(mkinstalldirs) $(DESTDIR)/var/tmp
 
 .PHONY: tar
diff -ur rpm-3.0.5.org/lib/rpmrc.c rpm-3.0.5/lib/rpmrc.c
--- rpm-3.0.5.org/lib/rpmrc.c	Mon Jul 17 02:53:27 2000
+++ rpm-3.0.5/lib/rpmrc.c	Sun Aug  6 18:51:45 2000
@@ -481,7 +481,7 @@
     addMacro(NULL, "_preScriptEnvironment",NULL, prescriptenviron,RMIL_DEFAULT);
 
     setVarDefault(-1,			"_topdir",
-		"/usr/src/redhat",	"%{_usr}/src/redhat");
+		"/usr/src/RPM",		"%{_usr}/src/RPM");
     setVarDefault(-1,			"_tmppath",
 		"/var/tmp",		"%{_var}/tmp");
     setVarDefault(-1,			"_dbpath",
diff -ur rpm-3.0.5.org/macros.in rpm-3.0.5/macros.in
--- rpm-3.0.5.org/macros.in	Sun Jul  9 18:35:56 2000
+++ rpm-3.0.5/macros.in	Sun Aug  6 18:35:41 2000
@@ -110,7 +110,7 @@
 %_specdir		%{_topdir}/SPECS
 %_srcrpmdir		%{_topdir}/SRPMS
 %_tmppath		%{_var}/tmp
-%_topdir		%{_usrsrc}/redhat
+%_topdir		%{_usrsrc}/RPM
 
 #==============================================================================
 # ---- Optional rpmrc macros.
diff -ur rpm-3.0.5.org/scripts/check-prereqs rpm-3.0.5/scripts/check-prereqs
--- rpm-3.0.5.org/scripts/check-prereqs	Thu Dec  2 23:58:04 1999
+++ rpm-3.0.5/scripts/check-prereqs	Sun Aug  6 18:51:05 2000
@@ -1,6 +1,6 @@
-#!/bin/bash2
+#!/bin/bash
 
-bashit="/bin/bash2 --rpm-requires"
+bashit="/bin/bash --rpm-requires"
 
 # Make sure that this bash has the rpm-requires hack
 $bashit < /dev/null 2>&1 > /dev/null || exit $?
