diff -ur rpm-3.0.6.orig/scripts/brp-strip rpm-3.0.6/scripts/brp-strip
--- rpm-3.0.6.orig/scripts/brp-strip	Sun Jul  9 15:36:20 2000
+++ rpm-3.0.6/scripts/brp-strip	Fri Dec 12 18:33:55 2003
@@ -7,6 +7,6 @@
 # Strip ELF binaries
 for f in `find $RPM_BUILD_ROOT -type f \( -perm -0100 -or -perm -0010 -or -perm -0001 \) -exec file {} \; | \
 	grep -v ' shared object,' | \
-	sed -n -e 's/^\(.*\):[ 	]*ELF.*, not stripped/\1/p'`; do
+	sed -n -e 's/^\(.*\):.*[[:space:]]ELF.*, not stripped/\1/p'`; do
 	strip $f || :
 done
diff -ur rpm-3.0.6.orig/scripts/brp-strip-comment-note rpm-3.0.6/scripts/brp-strip-comment-note
--- rpm-3.0.6.orig/scripts/brp-strip-comment-note	Sun Jul  9 15:36:20 2000
+++ rpm-3.0.6/scripts/brp-strip-comment-note	Fri Dec 12 18:34:41 2003
@@ -7,7 +7,7 @@
 # Strip .comment and .note sections (the latter only if it is not allocated)
 # for already stripped elf files in the build root
 for f in `find $RPM_BUILD_ROOT -type f \( -perm -0100 -or -perm -0010 -or -perm -0001 \) -exec file {} \; | \
-	sed -n -e 's/^\(.*\):[ 	]*ELF.*, stripped/\1/p'`; do
+	sed -n -e 's/^\(.*\):.*[[:space:]]ELF.*, stripped/\1/p'`; do
 	note="-R .note"
 	if objdump -h $f | grep '^[ 	]*[0-9]*[ 	]*.note[ 	]' -A 1 | \
 		grep ALLOC >/dev/null; then
