diff -puNr rpm-4.11.2.orig/macros.in rpm-4.11.2/macros.in
--- rpm-4.11.2.orig/macros.in	2014-06-30 01:32:55 +0000
+++ rpm-4.11.2/macros.in	2014-06-30 02:01:20 +0000
@@ -1099,9 +1099,13 @@ done \
 %{__patch} %{-p:-p%{-p*}} %{-q:-s}\
 %{__bzr} commit %{-q} -m %{-m*}
 
-# Single patch application
+# Single patch application with proper portable error detection
 %apply_patch(qp:m:)\
-%{uncompress:%{1}} | %{expand:%__scm_apply_%{__scm} %{-q} %{-p:-p%{-p*}} %{-m:-m%{-m*}}}
+exec 3>&1; rc=$(exec 4>&1 >&3; { \\\
+	%{uncompress:%{1}} || echo $? >&4 \\\
+	} | { \\\
+	%{expand:%__scm_apply_%{__scm} %{-q} %{-p:-p%{-p*}} %{-m:-m%{-m*}}} || echo $? >&4 ); \\\
+	test -z "$rc"
 
 # Automatically apply all patches
 %autopatch(vp:)\
diff -puNr rpm-4.11.2.orig/build/parsePrep.c rpm-4.11.2/build/parsePrep.c
--- rpm-4.11.2.orig/build/parsePrep.c	2013-11-22 10:31:31 +0000
+++ rpm-4.11.2/build/parsePrep.c	2014-06-30 06:06:19 +0000
@@ -102,7 +102,13 @@ static char *doPatch(rpmSpec spec, uint3
 		reverse ? " -R" : "", 
 		removeEmpties ? " -E" : "");
 
-    patchcmd = rpmExpand("%{uncompress: ", fn, "} | %{__patch} ", args, NULL);
+    /* XXX: Although the following works as expected it should be much more
+            flexible if the whole logic was outsourced to the macros file,
+            e.g. define a macro there, then pass the patch file and patch'es
+            arguments to this macro. -- (GM) */
+    patchcmd = rpmExpand("exec 3>&1;rc=$(exec 4>&1 >&3;{\n%{uncompress: ",
+		fn, "}||echo $? >&4\n}|{\n%{__patch} ", args,"||echo $? >&4\n});",
+		"test -z \"$rc\"",NULL);
 
     free(arg_fuzz);
     free(arg_dir);
