diff -uNp -r rpm-4.2.old/tools/Makefile.am rpm-4.2/tools/Makefile.am
--- rpm-4.2.old/tools/Makefile.am	2012-08-14 13:07:48 +0000
+++ rpm-4.2/tools/Makefile.am	2012-08-14 16:30:25 +0000
@@ -33,7 +33,25 @@ pkgbin_PROGRAMS = debugedit javadeps rpm
 
 bin_PROGRAMS =	rpmgraph
 
+rpmlead_LDADD =         $(top_builddir)/rpmio/.libs/librpmio.la \
+                        $(top_builddir)/lib/.libs/librpm.la
+rpmsignature_LDADD =    $(top_builddir)/rpmio/.libs/librpmio.la \
+                        $(top_builddir)/lib/.libs/librpm.la
+dump_LDADD =            $(top_builddir)/rpmio/.libs/librpmio.la \
+                        $(top_builddir)/lib/.libs/librpm.la \
+                        $(top_builddir)/rpmdb/.libs/librpmdb.la
+dumpdb_LDADD =          $(top_builddir)/rpmdb/.libs/librpmdb.la \
+                        $(top_builddir)/lib/.libs/librpm.la
+rpmarchive_LDADD =      $(top_builddir)/rpmio/.libs/librpmio.la \
+                        $(top_builddir)/rpmdb/.libs/librpmdb.la \
+                        $(top_builddir)/lib/.libs/librpm.la
+rpmheader_LDADD =       $(top_builddir)/rpmio/.libs/librpmio.la \
+                        $(top_builddir)/rpmdb/.libs/librpmdb.la \
+                        $(top_builddir)/lib/.libs/librpm.la
+
 convertdb1_SOURCES =	convertdb1.c
+convertdb1_LDADD =	$(top_builddir)/rpmdb/.libs/librpmdb.la \
+                        $(top_builddir)/rpmio/.libs/librpmio.la
 
 debugedit_SOURCES =	debugedit.c hashtab.c
 debugedit_LDADD =	@LDFLAGS_STATIC@ @WITH_LIBELF_LIB@ \
@@ -43,10 +61,18 @@ debugedit_LDADD =	@LDFLAGS_STATIC@ @WITH
 javadeps_SOURCES =	javadeps.c
 
 rpmcache_SOURCES =	rpmcache.c
+rpmcache_LDADD =	$(top_builddir)/popt/.libs/libpopt.la \
+                        $(top_builddir)/lib/.libs/librpm.la \
+                        $(top_builddir)/rpmdb/.libs/librpmdb.la \
+                        $(top_builddir)/rpmio/.libs/librpmio.la
 #rpmcache_LDFLAGS =	$(staticLDFLAGS)
 
 rpmdeps_SOURCES =	rpmdeps.c
-rpmdeps_LDADD =		$(top_builddir)/build/librpmbuild.la $(top_builddir)/build/.libs/librpmbuild.a
+rpmdeps_LDADD =		$(top_builddir)/build/librpmbuild.la \
+                        $(top_builddir)/build/.libs/librpmbuild.a \
+                        $(top_builddir)/lib/.libs/librpm.la \
+                        $(top_builddir)/popt/.libs/libpopt.la \
+                        $(top_builddir)/rpmio/.libs/librpmio.la
 #rpmdeps_LDFLAGS =	$(staticLDFLAGS)
 
 rpmfile_SOURCES =	rpmfile.c
@@ -54,6 +80,9 @@ rpmfile_LDADD =		$(top_builddir)/rpmio/l
 #rpmfile_LDFLAGS =	$(staticLDFLAGS)
 
 rpmgraph_SOURCES =	rpmgraph.c
+rpmgraph_LDADD =        $(top_builddir)/rpmio/librpmio.la \
+                        $(top_builddir)/lib/librpm.la \
+                        $(top_builddir)/popt/.libs/libpopt.la
 #rpmgraph_LDFLAGS =	$(staticLDFLAGS)
 
 rpminject_SOURCES =	rpminject.c
