--- rpm-4.11.2.orig/fileattrs/elf.attr	2012-11-07 12:55:24 +0000
+++ rpm-4.11.2/fileattrs/elf.attr	2014-07-05 01:23:44 +0000
@@ -1,4 +1,3 @@
 %__elf_provides		%{_rpmconfigdir}/elfdeps --provides %{?__filter_GLIBC_PRIVATE:--filter-private}
 %__elf_requires		%{_rpmconfigdir}/elfdeps --requires %{?__filter_GLIBC_PRIVATE:--filter-private}
 %__elf_magic		^(setuid )?(setgid )?(sticky )?ELF (32|64)-bit.*$
-%__elf_flags		exeonly
