diff -ur rpm-4.2.orig/beecrypt/configure rpm-4.2/beecrypt/configure
--- rpm-4.2.orig/beecrypt/configure	2003-03-10 19:08:52 +0000
+++ rpm-4.2/beecrypt/configure	2011-10-10 04:47:21 +0000
@@ -8693,7 +8693,7 @@
 
 
 # Initialize tool flags
-CFLAGS=""
+CFLAGS="-Wa,--noexecstack"
 CPPFLAGS=""
 LDFLAGS=""
 
diff -ur rpm-4.2.orig/beecrypt/configure.ac rpm-4.2/beecrypt/configure.ac
--- rpm-4.2.orig/beecrypt/configure.ac	2002-12-19 23:40:28 +0000
+++ rpm-4.2/beecrypt/configure.ac	2011-10-10 04:46:56 +0000
@@ -84,7 +84,7 @@
 AC_SYS_LARGEFILE
 
 # Initialize tool flags
-CFLAGS=""
+CFLAGS="-Wa,--noexecstack"
 CPPFLAGS=""
 LDFLAGS=""
 
