diff -urp rpm-4.2.orig/rpmdb/header.c rpm-4.2/rpmdb/header.c
--- rpm-4.2.orig/rpmdb/header.c	2003-01-08 19:27:44 +0000
+++ rpm-4.2/rpmdb/header.c	2011-10-10 01:54:09 +0000
@@ -486,6 +486,8 @@ static int regionSwab(/*@null@*/ indexEn
     int tdel, tl = dl;
     struct indexEntry_s ieprev;
 
+    if (regionid > 0)
+       return -1;
 /*@-boundswrite@*/
     memset(&ieprev, 0, sizeof(ieprev));
 /*@=boundswrite@*/
@@ -1090,7 +1092,10 @@ Header headerLoad(/*@kept@*/ void * uh)
 /*@-sizeoftype@*/
 		size_t nb = REGION_TAG_COUNT;
 /*@=sizeoftype@*/
-		int_32 * stei = memcpy(alloca(nb), dataStart + off, nb);
+		int_32 * stei;
+		if (dataStart + off + nb > dataEnd)
+		    goto errxit;
+		stei = memcpy(alloca(nb), dataStart + off, nb);
 		rdl = -ntohl(stei[2]);	/* negative offset */
 		ril = rdl/sizeof(*pe);
 		if (hdrchkTags(ril) || hdrchkData(rdl))
