diff -uNrp rpm-4.2.orig/gendiff rpm-4.2/gendiff
--- rpm-4.2.orig/gendiff	Sat Jul 13 19:53:59 2002
+++ rpm-4.2/gendiff	Wed Mar  3 17:05:28 2004
@@ -6,11 +6,10 @@
   exit 1
 }
 
-find $1 \( -name "*$2" -o -name ".*$2" \) -print |
-while read f; do
-    U=-u
+find "$1" \( -name "*$2" -o -name ".*$2" \) -print |
+while read f; do
+    U=-uNp
     [ "`basename $f`" = "ChangeLog$2" ] && U=-U0
-#    diff ${U} $f `echo $f | sed s/$2\$//`
     if [ -r "$f" ]; then
 	diff ${U} "${f}" "${f%$2}"
     else
