--- rpm-3.0.6/rpmrc.in.orig	Wed Sep 13 21:55:15 2000
+++ rpm-3.0.6/rpmrc.in	Tue Dec 17 18:29:05 2002
@@ -8,24 +8,35 @@
 #############################################################
 # Values for RPM_OPT_FLAGS for various platforms
 
-optflags: i386 -O2 -m486 -fno-strength-reduce
-optflags: i486 -O2 -march=i486
-optflags: i586 -O2 -march=i586
-optflags: i686 -O2 -march=i686
-optflags: athlon -O2 -march=athlon
-optflags: ia64 -O2
+# i386 and i486 assume that such packages will most of the time be run
+# on a i686+ and are only built for the older architecture flavors for
+# compatibility.  i586 is actually for the original Intel Pentium and
+# shouldn't be used on anything else (it should be slower than i386 or
+# i486 if wrongly used on the Pentium Pro to Pentium III).  If in doubt,
+# use i386 which is the safe choice and is reasonably good for all.
+optflags: i386 -pipe -march=i386 -mcpu=i686 -O2 -momit-leaf-frame-pointer -mpreferred-stack-boundary=3
+optflags: i486 -pipe -march=i486 -mcpu=i686 -O2 -momit-leaf-frame-pointer -mpreferred-stack-boundary=3
+optflags: i586 -pipe -march=i586 -O2 -momit-leaf-frame-pointer -mpreferred-stack-boundary=3
+optflags: i686 -pipe -march=i686 -O2 -momit-leaf-frame-pointer -mpreferred-stack-boundary=3
+optflags: athlon -pipe -march=athlon -O2 -momit-leaf-frame-pointer -mpreferred-stack-boundary=3
+
+# The official RPM starting with 3.0.5 uses -mieee on Alpha by default.
+# We don't as to not kill floating-point performance, but packages which
+# care might want to add that flag themselves.
+optflags: alpha -pipe -mcpu=ev4 -O2
+optflags: alphaev5 -pipe -mcpu=ev5 -O2
+optflags: alphaev56 -pipe -mcpu=ev56 -O2
+optflags: alphapca56 -pipe -mcpu=pca56 -O2
+optflags: alphaev6 -pipe -mcpu=ev6 -O2
+optflags: alphaev67 -pipe -mcpu=ev67 -O2
 
-# XXX Please note that -mieee has been added in rpm-3.0.5.
-optflags: alpha -O2 -mieee
-optflags: alphaev5 -O2 -mieee -mcpu=ev5
-optflags: alphaev56 -O2 -mieee -mcpu=ev56
-optflags: alphapca56 -O2 -mieee -mcpu=pca56
-optflags: alphaev6 -O2 -mieee -mcpu=ev6
-optflags: alphaev67 -O2 -mieee -mcpu=ev67
+# Note: we set plain "sparc" to require SPARC V8.  With the official RPM,
+# it would also work for V7.
+optflags: sparc -pipe -mcpu=v8 -mtune=ultrasparc -m32 -O2
+optflags: sparcv9 -pipe -mcpu=ultrasparc -Wa,-Av8plusa -m32 -O2
+optflags: sparc64 -pipe -mcpu=ultrasparc -m64 -O2
 
-optflags: sparc -O2 -m32 -mtune=ultrasparc
-optflags: sparcv9 -O2 -m32 -mcpu=ultrasparc
-optflags: sparc64 -O2 -m64 -mcpu=ultrasparc
+optflags: ia64 -O2
 optflags: m68k -O2 -fomit-frame-pointer
 optflags: ppc -O2 -fsigned-char
 optflags: parisc -O2 -mpa-risc-1-0
