--- rpm-3.0.6/macros.in.orig    Sun Feb 23 01:59:33 2003
+++ rpm-3.0.6/macros.in Sun Feb 23 02:01:38 2003
@@ -463,7 +463,7 @@
 CFLAGS="${CFLAGS:-%optflags}" ; export CFLAGS ; \
 CXXFLAGS="${CXXFLAGS:-%optflags}" ; export CXXFLAGS ; \
 FFLAGS="${FFLAGS:-%optflags}" ; export FFLAGS ; \
-%{?__libtoolize:[ -f configure.in ] && %{__libtoolize} --copy --force} ; \
+%{?__libtoolize:[ -f configure.in ] && %{__libtoolize} --copy --force ;} \
 ./configure %{_target_platform} \\\
        --prefix=%{_prefix} \\\
        --exec-prefix=%{_exec_prefix} \\\
