--- rpm-3.0.5/gendiff.owl	Wed Mar  4 18:05:15 1998
+++ rpm-3.0.5/gendiff	Sun Nov 19 12:42:11 2000
@@ -6,7 +6,10 @@
   exit 1
 }
 
-find $1 \( -name "*$2" -o -name ".*$2" \) -print |
-while read f; do
-	diff -u $f `echo $f | sed s/$2\$//`
+find "$1" \( -name "*$2" -o -name ".*$2" \) -print |
+while read fin; do
+	fou=`echo "$fin" | sed s/$2\$//`;
+	[ -r "$fin" ] || fin="/dev/null"
+	[ -r "$fou" ] || fou="/dev/null"
+	diff -u "$fin" "$fou"
 done
