--- rpm-4.11.2.orig/rpmrc.in	2014-06-17 16:01:31 +0000
+++ rpm-4.11.2/rpmrc.in	2014-06-17 19:26:25 +0000
@@ -12,32 +12,43 @@
 # "fat" binary with both archs, for Darwin
 optflags: fat -O2 -g -arch i386 -arch ppc
 
-optflags: i386 -O2 -g -march=i386 -mtune=i686
-optflags: i486 -O2 -g -march=i486
-optflags: i586 -O2 -g -march=i586
-optflags: i686 -O2 -g -march=i686
-optflags: pentium3 -O2 -g -march=pentium3
-optflags: pentium4 -O2 -g -march=pentium4
-optflags: athlon -O2 -g -march=athlon
-optflags: geode -Os -g -m32 -march=geode
-optflags: ia64 -O2 -g
-optflags: x86_64 -O2 -g
-optflags: amd64 -O2 -g
-optflags: ia32e -O2 -g
-
-optflags: alpha -O2 -g -mieee
-optflags: alphaev5 -O2 -g -mieee -mtune=ev5
-optflags: alphaev56 -O2 -g -mieee -mtune=ev56
-optflags: alphapca56 -O2 -g -mieee -mtune=pca56
-optflags: alphaev6 -O2 -g -mieee -mtune=ev6
-optflags: alphaev67 -O2 -g -mieee -mtune=ev67
-
-optflags: sparc -O2 -g -m32 -mtune=ultrasparc
-optflags: sparcv8 -O2 -g -m32 -mtune=ultrasparc -mv8
-optflags: sparcv9 -O2 -g -m32 -mtune=ultrasparc
-optflags: sparcv9v -O2 -g -m32 -mtune=niagara
-optflags: sparc64 -O2 -g -m64 -mtune=ultrasparc
-optflags: sparc64v -O2 -g -m64 -mtune=niagara
+# i386 and i486 assume that such packages will most of the time be run
+# on a i686+ and are only built for the older architecture flavors for
+# compatibility.  i586 is actually for the original Intel Pentium and
+# shouldn't be used on anything else (it should be slower than i386 or
+# i486 if wrongly used on the Pentium Pro to Pentium III).  If in doubt,
+# use i386 which is the safe choice and is reasonably good for all.
+optflags: i386 -g -pipe -march=i386 -mtune=i686 -O2 -momit-leaf-frame-pointer -mpreferred-stack-boundary=3
+optflags: i486 -g -pipe -march=i486 -mtune=i686 -O2 -momit-leaf-frame-pointer -mpreferred-stack-boundary=3
+optflags: i586 -g -pipe -march=i586 -O2 -momit-leaf-frame-pointer -mpreferred-stack-boundary=3
+optflags: i686 -g -pipe -march=i686 -O2 -momit-leaf-frame-pointer -mpreferred-stack-boundary=3
+optflags: pentium3 -g -pipe -march=pentium3 -O2 -momit-leaf-frame-pointer -mpreferred-stack-boundary=3
+optflags: pentium4 -g -pipe -march=pentium4 -O2 -momit-leaf-frame-pointer -mpreferred-stack-boundary=3
+optflags: athlon athlon -g -pipe -march=athlon -O2 -momit-leaf-frame-pointer -mpreferred-stack-boundary=3
+optflags: geode -g -pipe -march=geode -Os -m32
+optflags: ia64 -g -pipe -O2
+optflags: x86_64 -g -pipe -O2
+optflags: amd64 -g -pipe -O2
+optflags: ia32e -g -pipe -O2
+
+# The official RPM starting with 3.0.5 uses -mieee on Alpha by default.
+# We don't as to not kill floating-point performance, but packages which
+# care might want to add that flag themselves.
+optflags: alpha -g -pipe -O2
+optflags: alphaev5 -g -pipe -O2 -mtune=ev5
+optflags: alphaev56 -g -pipe -O2 -mtune=ev56
+optflags: alphapca56 -g -pipe -O2 -mtune=pca56
+optflags: alphaev6 -g -pipe -O2 -mtune=ev6
+optflags: alphaev67 -g -pipe -O2 -mtune=ev67
+
+# Note: we set plain "sparc" to require SPARC V8.  With the official RPM,
+# it would also work for V7.
+optflags: sparc -g -pipe -O2 -m32 -mcpu=v8 -mtune=ultrasparc
+optflags: sparcv8 -g -pipe -O2 -m32 -mtune=ultrasparc -mv8
+optflags: sparcv9 -g -pipe -O2 -m32 -mtune=ultrasparc
+optflags: sparcv9v -g -pipe -O2 -m32 -mtune=niagara
+optflags: sparc64 -g -pipe -O2 -m64 -mtune=ultrasparc
+optflags: sparc64v -g -pipe -O2 -m64 -mtune=niagara
 
 optflags: m68k -O2 -g -fomit-frame-pointer
 
@@ -253,13 +264,13 @@ buildarchtranslate: osfmach3_i586: i386
 buildarchtranslate: osfmach3_i486: i386
 buildarchtranslate: osfmach3_i386: i386
 
-buildarchtranslate: athlon: i386
-buildarchtranslate: geode: i386
-buildarchtranslate: pentium4: i386
-buildarchtranslate: pentium3: i386
-buildarchtranslate: i686: i386
-buildarchtranslate: i586: i386
-buildarchtranslate: i486: i386
+buildarchtranslate: athlon: i686
+buildarchtranslate: geode: i586
+buildarchtranslate: pentium4: i686
+buildarchtranslate: pentium3: i686
+buildarchtranslate: i686: i686
+buildarchtranslate: i586: i586
+buildarchtranslate: i486: i486
 buildarchtranslate: i386: i386
 
 buildarchtranslate: alphaev5: alpha
