diff -uNrp rpm-4.2.orig/build/files.c rpm-4.2/build/files.c
--- rpm-4.2.orig/build/files.c	Sat Mar  1 19:15:04 2003
+++ rpm-4.2/build/files.c	Wed Mar  3 18:04:00 2004
@@ -2182,6 +2182,7 @@ void initSourceHeader(Spec spec)
 	case RPMTAG_CHANGELOGNAME:
 	case RPMTAG_CHANGELOGTEXT:
 	case RPMTAG_URL:
+	case RPMTAG_BUILDHOST:
 	case HEADER_I18NTABLE:
 	    if (ptr)
 		(void)headerAddEntry(spec->sourceHeader, tag, type, ptr, count);
diff -uNrp rpm-4.2.orig/build/pack.c rpm-4.2/build/pack.c
--- rpm-4.2.orig/build/pack.c	Tue Dec 24 03:19:13 2002
+++ rpm-4.2/build/pack.c	Wed Mar  3 18:37:03 2004
@@ -473,7 +473,10 @@ int writeRPM(Header *hdrp, unsigned char
 
     /* Create and add the cookie */
     if (cookie) {
-	sprintf(buf, "%s %d", buildHost(), (int) (*getBuildTime()));
+	char *buildHost_ = NULL;
+	if (!headerGetEntry(h, RPMTAG_BUILDHOST, NULL, (void **)&buildHost_, NULL))
+		buildHost_ = buildHost();
+	sprintf(buf, "%s %d", buildHost_, (int) (*getBuildTime()));
 	*cookie = xstrdup(buf);
 	(void) headerAddEntry(h, RPMTAG_COOKIE, RPM_STRING_TYPE, *cookie, 1);
     }
@@ -766,8 +769,9 @@ int packageBinaries(Spec spec)
 	
 	(void) headerAddEntry(pkg->header, RPMTAG_RPMVERSION,
 		       RPM_STRING_TYPE, VERSION, 1);
-	(void) headerAddEntry(pkg->header, RPMTAG_BUILDHOST,
-		       RPM_STRING_TYPE, buildHost(), 1);
+	if (!headerIsEntry(pkg->header, RPMTAG_BUILDHOST))
+		(void) headerAddEntry(pkg->header, RPMTAG_BUILDHOST,
+			       RPM_STRING_TYPE, buildHost(), 1);
 	(void) headerAddEntry(pkg->header, RPMTAG_BUILDTIME,
 		       RPM_INT32_TYPE, getBuildTime(), 1);
 
@@ -855,8 +859,9 @@ int packageSources(Spec spec)
     /* Add some cruft */
     (void) headerAddEntry(spec->sourceHeader, RPMTAG_RPMVERSION,
 		   RPM_STRING_TYPE, VERSION, 1);
-    (void) headerAddEntry(spec->sourceHeader, RPMTAG_BUILDHOST,
-		   RPM_STRING_TYPE, buildHost(), 1);
+    if (!headerIsEntry(spec->sourceHeader, RPMTAG_BUILDHOST))
+        (void) headerAddEntry(spec->sourceHeader, RPMTAG_BUILDHOST,
+		       RPM_STRING_TYPE, buildHost(), 1);
     (void) headerAddEntry(spec->sourceHeader, RPMTAG_BUILDTIME,
 		   RPM_INT32_TYPE, getBuildTime(), 1);
 
diff -uNrp rpm-4.2.orig/build/parsePreamble.c rpm-4.2/build/parsePreamble.c
--- rpm-4.2.orig/build/parsePreamble.c	Sat Dec 21 17:26:43 2002
+++ rpm-4.2/build/parsePreamble.c	Wed Mar  3 18:38:17 2004
@@ -29,6 +29,7 @@ static rpmTag copyTagsDuringParse[] = {
     RPMTAG_CHANGELOGNAME,
     RPMTAG_CHANGELOGTEXT,
     RPMTAG_PREFIXES,
+    RPMTAG_BUILDHOST,
     RPMTAG_RHNPLATFORM,
     0
 };
@@ -328,6 +329,7 @@ static struct optionalTag {
     { RPMTAG_PACKAGER,		"%{packager}" },
     { RPMTAG_DISTRIBUTION,	"%{distribution}" },
     { RPMTAG_DISTURL,		"%{disturl}" },
+    { RPMTAG_BUILDHOST,		"%{buildhost}" },
     { -1, NULL }
 };
 
@@ -540,6 +542,7 @@ static int handlePreambleTag(Spec spec, 
     case RPMTAG_VENDOR:
     case RPMTAG_LICENSE:
     case RPMTAG_PACKAGER:
+    case RPMTAG_BUILDHOST:
 	if (!*lang)
 	    (void) headerAddEntry(pkg->header, tag, RPM_STRING_TYPE, field, 1);
 	else if (!(noLang && strcmp(lang, RPMBUILD_DEFAULT_LANG)))
@@ -740,6 +743,7 @@ static struct PreambleRec_s preambleList
     {RPMTAG_EPOCH,		0, 0, "epoch"},
     {RPMTAG_EPOCH,		0, 0, "serial"},
     {RPMTAG_SUMMARY,		0, 1, "summary"},
+    {RPMTAG_BUILDHOST,		0, 1, "buildhost"},
     {RPMTAG_LICENSE,		0, 0, "copyright"},
     {RPMTAG_LICENSE,		0, 0, "license"},
     {RPMTAG_DISTRIBUTION,	0, 0, "distribution"},
