diff -urN rpm-3.0.5.org/scripts/check-prereqs rpm-3.0.5/scripts/check-prereqs
--- rpm-3.0.5.org/scripts/check-prereqs	Thu Dec  2 23:58:04 1999
+++ rpm-3.0.5/scripts/check-prereqs	Sun Sep  3 16:37:40 2000
@@ -1,6 +1,6 @@
-#!/bin/bash2
+#!/bin/bash
 
-bashit="/bin/bash2 --rpm-requires"
+bashit="/bin/bash --rpm-requires"
 
 # Make sure that this bash has the rpm-requires hack
 $bashit < /dev/null 2>&1 > /dev/null || exit $?
