diff -upk.orig modutils-2.4.27.orig/depmod/depmod.c modutils-2.4.27/depmod/depmod.c
--- modutils-2.4.27.orig/depmod/depmod.c	2003-03-23 02:34:28 +0000
+++ modutils-2.4.27/depmod/depmod.c	2005-10-26 15:06:29 +0000
@@ -1549,7 +1549,9 @@ int DEPMOD_MAIN(int argc, char **argv)
 		{"errsyms", 0, 0, 'e'},
 		{"filesyms", 1, 0, 'F'},
 		{"help", 0, 0, 'h'},
+		{"ignore-versions", 0, 0, 'i'}, /* backward compatibility */
 		{"show", 0, 0, 'n'},
+		{"system-map", 1, 0, 'm'}, /* backward compatibility, same as 'F' */
 		{"quick", 0, 0, 'A'},
 		{"quiet", 0, 0, 'q'},
 		{"syslog", 0, 0, 's'},
@@ -1562,7 +1564,7 @@ int DEPMOD_MAIN(int argc, char **argv)
 
 	error_file = "depmod";
 
-	while ((o = getopt_long(argc, argv, "aAb:C:eF:hnqsvVru",
+	while ((o = getopt_long(argc, argv, "aAb:C:eF:him:nqsvVru",
 				&long_opts[0], NULL)) != EOF) {
 		switch (o) {
 		case 'A':
@@ -1585,10 +1587,17 @@ int DEPMOD_MAIN(int argc, char **argv)
 			return 0;
 			break;
 
+		case 'i':
+			fprintf(stderr, "Warning: -i is obsolete and deprecated.\n");
+			ignore_suffix = 1;
+			break;
+
 		case 'C':
 			conf_file = optarg;
 			break;
 
+		case 'm':
+			fprintf(stderr, "Warning: -m is deprecated; please use -F.\n");
 		case 'F':
 			file_syms = optarg;
 			break;
@@ -1635,7 +1644,7 @@ int DEPMOD_MAIN(int argc, char **argv)
 	argc -= optind;
 	argv += optind;
 
-	ignore_suffix = file_syms != NULL;
+	ignore_suffix += (file_syms != NULL);
 
 	if (stdmode || argc == 0) {
 		/* option -a is the default without arguments */
diff -upk.orig modutils-2.4.27.orig/man/depmod.8 modutils-2.4.27/man/depmod.8
--- modutils-2.4.27.orig/man/depmod.8	2003-02-12 03:51:53 +0000
+++ modutils-2.4.27/man/depmod.8	2005-10-26 15:06:29 +0000
@@ -8,7 +8,7 @@
 depmod \- handle dependency descriptions for loadable kernel modules
 .SH SYNOPSIS
 .B depmod
-[\-aA] [\-ehnqrsuvV]
+[\-aA] [\-eihnqrsuvV]
 .RI [\-C\  configfile "] [\-F\ " kernelsyms "] [\-b " basedirectory ]
 .RI [ forced_version ]
 .br
@@ -86,6 +86,19 @@ This option only updates the dependency 
 .BR \-e ", " \-\-errsyms
 Show all the unresolved symbols for each module.
 .TP
+.BR \-i ", " \-\-ignore-versions
+Ignore symbol versions when building dependency list.  This is useful
+especially if you use a versioned kernel and want to build dependencies
+against a System.map file (see the
+.B \-F
+option).  Use this with care, because if the symbol versions don't match,
+.B modprobe
+or
+.B insmod
+won't be able to load the modules eventhough
+.B depmod
+succeeds.
+.TP
 .BR \-h ", " \-\-help
 Display a summary of options and immediately exit.
 .TP
@@ -196,6 +209,10 @@ output, since that file contains the sym
 symbols.  However you can use a
 .I System.map
 even with versioned symbols.
+For compatibility,
+.B "\-m kernelsyms"
+is identical to
+.B "\-F kernelsyms"
 .SH CONFIGURATION
 The behavior of
 .B depmod
