diff -urN modutils-2.3.17.org/util/alias.h modutils-2.3.17/util/alias.h
--- modutils-2.3.17.org/util/alias.h	Thu Sep 21 14:00:44 2000
+++ modutils-2.3.17/util/alias.h	Fri Oct 13 22:27:01 2000
@@ -101,6 +101,7 @@
 	"char-major-10-144 nvram",	/* from Tigran Aivazian <tigran@sco.COM> */
 	"char-major-10-157 applicom",	/* from David Woodhouse <dwmw2@infradead.org> */
 	"char-major-10-175 agpgart",    /* /dev/agpgart GART AGP mapping access */
+	"char-major-10-181 toshiba",    /* /dev/toshiba - for Toshiba notebooks */
 
 	"char-major-14 soundcore",
 	"char-major-19 cyclades",
@@ -135,13 +136,13 @@
 	"net-pf-1 unix",	/* PF_UNIX	1  Unix domain sockets */
 	"net-pf-2 ipv4",	/* PF_INET	2  Internet IP Protocol */
 	"net-pf-3 off",		/* PF_AX25	3  Amateur Radio AX.25 */
-	"net-pf-4 off",		/* PF_IPX	4  Novell IPX */
-	"net-pf-5 off",		/* PF_APPLETALK	5  Appletalk DDP */
+	"net-pf-4 ipx",		/* PF_IPX	4  Novell IPX */
+	"net-pf-5 appletalk",	/* PF_APPLETALK	5  Appletalk DDP */
 	"net-pf-6 off",		/* PF_NETROM	6  Amateur radio NetROM */
 				/* PF_BRIDGE	7  Multiprotocol bridge */
 				/* PF_AAL5	8  Reserved for Werner's ATM */
 				/* PF_X25	9  Reserved for X.25 project */
-				/* PF_INET6	10 IP version 6 */
+	"net-pf-10 ipv6",	/* PF_INET6	10 IP version 6 */
 
 	/* next two from <dairiki@matthews.dairiki.org>  Thanks! */
 	"net-pf-17 af_packet",
@@ -150,6 +151,10 @@
 	"netalias-2 ip_alias",
 	"plip0 plip",
 	"plip1 plip",
+        "cipcb0 cipcb",
+        "cipcb1 cipcb",
+        "cipcb2 cipcb",
+        "cipcb3 cipcb",
 	"ppp0 ppp",
 	"ppp1 ppp",
 	"scsi_hostadapter off",	/* if not in config file */
