diff -upk.orig modutils-2.4.27.orig/util/alias.h modutils-2.4.27/util/alias.h
--- modutils-2.4.27.orig/util/alias.h	2004-02-22 07:33:30 +0000
+++ modutils-2.4.27/util/alias.h	2005-10-26 15:10:34 +0000
@@ -97,7 +97,7 @@ char *aliaslist[] =
 	"char-major-6 lp",
 	"char-major-9 st",
 	"char-major-10 off",		/* was: mouse, was: misc */
-	"char-major-10-0 busmouse",	/* /dev/logibm Logitech bus mouse */
+	"char-major-10-0 logibusmouse",	/* /dev/logibm Logitech bus mouse */
 	"char-major-10-1 off",		/* /dev/psaux PS/2-style mouse port */
 	"char-major-10-2 msbusmouse",	/* /dev/inportbm Microsoft Inport bus mouse */
 	"char-major-10-3 atixlmouse",	/* /dev/atibm ATI XL bus mouse */
@@ -116,6 +116,7 @@ char *aliaslist[] =
 	"char-major-10-144 nvram",	/* from Tigran Aivazian <tigran@sco.COM> */
 	"char-major-10-157 applicom",	/* from David Woodhouse <dwmw2@infradead.org> */
 	"char-major-10-175 agpgart",    /* /dev/agpgart GART AGP mapping access */
+	"char-major-10-181 toshiba",	/* /dev/toshiba - for Toshiba notebooks */
 	"char-major-10-184 microcode",	/* Tigran Aivazian <tigran@veritas.com> */
 	"char-major-10-200 tun",	/* Universal TUN/TAP device driver */
 	"char-major-10-250 hci_vhci",
@@ -148,9 +149,10 @@ char *aliaslist[] =
 	"char-major-109 lvm-mod",
 	"char-major-161 ircomm-tty",
 	"char-major-171 raw1394",
-	"char-major-195 NVdriver",
+	"char-major-195 nvidia",
 	"char-major-200 vxspec",
 	"char-major-206 osst",  /* OnStream SCSI tape */
+	"char-major-212 slamrmo", /* SLMmodem driver */
 	"char-major-216 rfcomm",
 
 	"dos msdos",
@@ -173,7 +175,7 @@ char *aliaslist[] =
 				/* PF_BRIDGE	7  Multiprotocol bridge */
 				/* PF_AAL5	8  Reserved for Werner's ATM */
 				/* PF_X25	9  Reserved for X.25 project */
-	"net-pf-10 off",	/* PF_INET6	10 IP version 6 */
+	"net-pf-10 ipv6",	/* PF_INET6	10 IP version 6 */
 	"net-pf-15 af_key",	/* PF_KEY       15 Key Management API */
 
 	/* next two from <dairiki@matthews.dairiki.org>  Thanks! */
@@ -229,12 +231,36 @@ char *aliaslist[] =
 	"ppp-compress-24 ppp_deflate",
 	"ppp-compress-26 ppp_deflate",
 
+	/* crypto patch support */
+	"loop-xfer-gen-0 loop_gen",
+	"loop-xfer-gen-10 loop_gen",
+	"cipher-2 des",
+	"cipher-4 blowfish",
+	"cipher-6 idea",
+	"cipher-7 serp6f",
+	"cipher-8 mars6",
+	"cipher-11 rc62",
+	"cipher-15 dfc2",
+	"cipher-16 rijndael",
+	"cipher-17 rc5",
+
 #ifndef __sparc__
 	"parport_lowlevel parport_pc",
 #else
 	"parport_lowlevel parport_ax",
 #endif
 
+	/* To be able to attach some dongles */
+	"irlan0 irlan",
+	"irda-dongle-0 tekram",
+	"irda-dongle-1 esi",
+	"irda-dongle-2 actisys",
+	"irda-dongle-3 actisys",
+	"irda-dongle-4 girbil",
+	"irda-dongle-5 litelink",
+	"irda-dongle-6 airport",
+	"irda-dongle-7 old_belkin",
+
 	"usbdevfs usbcore",
 
 	"xfrm-type-2-50 esp4",
@@ -248,6 +274,8 @@ char *aliaslist[] =
 	"digest_null crypto_null",
 	"compress_null crypto_null",
 	"sha384 sha512",
+	"hid keybdev mousedev",
+	"wacom evdev",
 
 	NULL			/* marks the end of the list! */
 };
@@ -295,6 +323,8 @@ char *below[] =
  */
 char *prune[] =
 {
+	".rhkmvtag",
+	".versions",
 	"modules.dep",
 	"modules.description",
 	"modules.generic_string",
