diff -uNrp lilo-23.1.orig/Makefile lilo-23.1/Makefile
--- lilo-23.1.orig/Makefile	2010-11-04 15:35:15 +0000
+++ lilo-23.1/Makefile	2011-01-16 15:15:44 +0000
@@ -30,21 +30,21 @@ help:
 # everything needed to run, just short of installation
 #
 all: test
-	make -C src all
-	make -C images all
+	$(MAKE) -C src all
+	#$(MAKE) -C images all
 
 #
 # everything above plus the statically linked version
 #
 alles: test
-	make -C src alles
-	make -C images all
+	$(MAKE) -C src alles
+	#$(MAKE) -C images all
 
 #
 # documentation files
 #
 docs:
-	make -C doc all
+	$(MAKE) -C doc all
 
 #
 # if you have the 'bcc' compiler, then you can make the diagnostics, too
@@ -61,23 +61,23 @@ floppy: test
 	@echo before you proceed from this point.
 	@echo "Press <Enter> to continue, <^C> to abort ..."
 	@read
-	@make -C src floppy1
+	@$(MAKE) -C src floppy1
 	@echo Done.
 	@echo
 	@echo Remove the floppy from the drive.  Label it "\"1.6\""
 	@echo "Press <Enter> to continue, <^C> to abort ..."
 	@read
-	@make -C src floppy2
+	@$(MAKE) -C src floppy2
 	@echo Done.
 	@echo
 	@echo Remove the floppy from the drive.  Label it "\"2.4\""
 	@echo
 
 diagnostic: test
-	make -C src diagnostic
+	$(MAKE) -C src diagnostic
 
 dosexe: test
-	make -C dos lilo
+	$(MAKE) -C dos lilo
 
 #
 # test for compilers & utilities
@@ -91,42 +91,42 @@ test.img:
 # shorthand install, if one knows that one has the 'bcc' compiler
 #
 ins:
-	make -C src ins
+	$(MAKE) -C src ins
 
 #
 #  normal install, but doesn't make the diagnostic binaries
 #
 install:  all
-	make -C src install
-	make -C images install
-	make -C hooks install
-	make -C sample install
-	make -C man install
-	make -C dos install
+	$(MAKE) -C src install
+	#$(MAKE) -C images install
+	$(MAKE) -C hooks install
+	$(MAKE) -C sample install
+	$(MAKE) -C man install
+	$(MAKE) -C dos install
 
 tidy:
-	make -C src tidy
-	make -C diagnose tidy
-	make -C dos tidy
-	make -C doc tidy
+	$(MAKE) -C src tidy
+	$(MAKE) -C diagnose tidy
+	$(MAKE) -C dos tidy
+	$(MAKE) -C doc tidy
 
 clean: tidy
 	rm -f test.img
-	make -C src clean
-	make -C images clean
-	make -C diagnose clean
-	make -C dos clean
-	make -C doc clean
+	$(MAKE) -C src clean
+	#$(MAKE) -C images clean
+	$(MAKE) -C diagnose clean
+	$(MAKE) -C dos clean
+	$(MAKE) -C doc clean
 
 spotless: distclean
 distclean: clean
-	make -C src distclean
-	make -C diagnose distclean
-	make -C dos distclean
+	$(MAKE) -C src distclean
+	$(MAKE) -C diagnose distclean
+	$(MAKE) -C dos distclean
 
 uninstall:
-	make -C src uninstall
-	make -C images uninstall
-	make -C hooks uninstall
-	make -C sample uninstall
-	make -C man uninstall
+	$(MAKE) -C src uninstall
+	#$(MAKE) -C images uninstall
+	$(MAKE) -C hooks uninstall
+	$(MAKE) -C sample uninstall
+	$(MAKE) -C man uninstall
diff -uNrp lilo-23.1.orig/src/Makefile lilo-23.1/src/Makefile
--- lilo-23.1.orig/src/Makefile	2010-11-04 15:35:52 +0000
+++ lilo-23.1/src/Makefile	2011-01-16 15:15:44 +0000
@@ -15,7 +15,7 @@ LD86=ld86 -0
 NASM=nasm
 G=`cat foo1 foo2 | grep version | cut -d " " -f 3`
 
-CFLAGS=$(OPT) -Wall -g $(PCONFIG)
+CFLAGS=$(OPT) -Wall $(PCONFIG)
 LDFLAGS=#-Xlinker -qmagic
 LIBS=$(DEVMAPPER)
 
@@ -63,8 +63,8 @@ floppy1: disk.com bootsect.b
 floppy2: disk.b
 	@echo
 	@echo Creating Diagnostic Floppy 2.4
-	@make -C ../diagnose check
-	@make -C ../diagnose floppy
+	@$(MAKE) -C ../diagnose check
+	@$(MAKE) -C ../diagnose floppy
 
 diagnostic: ../test.img diag1.img diag2.img diag3.img
 
@@ -84,7 +84,7 @@ manpath:
 		dd if=$*.img of=$*.b bs=32 skip=1
 
 edit:		$(EDIT)
-		$(CC) -Wall -s -O -DSTANDALONE -o edit $(EDIT) $(LDFLAGS)
+		$(CC) $(CFLAGS) -DSTANDALONE -o edit $(EDIT) $(LDFLAGS)
 
 loader.i:	mkloader first.b second.b third.b bitmap.b mbr.b \
 			chain.b os2_d.b mbr2.b
@@ -123,7 +123,7 @@ pseudo1.s: bootsect.S pseudo.S disk.com
 	rm -f $(TMP).S
 
 pseudo2.s: bootsect.S pseudo.S
-	make -C ../diagnose all
+	$(MAKE) -C ../diagnose all
 	cat bootsect.S pseudo.S >$(TMP).S
 	$(CPP) -traditional $(PCONFIG) \
 		-DSIZEKRNL=`wc -c <../diagnose/test4.com | sed "s/ //g"` \
@@ -131,7 +131,7 @@ pseudo2.s: bootsect.S pseudo.S
 	rm -f $(TMP).S
 
 pseudo3.s: bootsect.S pseudo.S
-	make -C ../diagnose all
+	$(MAKE) -C ../diagnose all
 	cat bootsect.S pseudo.S >$(TMP).S
 	$(CPP) -traditional $(PCONFIG) \
 		-DSIZEKRNL=`wc -c <../diagnose/test5.com | sed "s/ //g"` \
@@ -147,11 +147,11 @@ diag1.img:	disk.com pseudo1.b
 	cat pseudo1.b disk.com >diag1.img
 
 diag2.img: pseudo2.b
-	make -C ../diagnose all
+	$(MAKE) -C ../diagnose all
 	cat pseudo2.b ../diagnose/test4.com >diag2.img
 
 diag3.img: pseudo3.b
-	make -C ../diagnose all
+	$(MAKE) -C ../diagnose all
 	cat pseudo3.b ../diagnose/test5.com >diag3.img
 
 flags.i:	../test.img Makefile version $(CFILES) $(HFILES)
@@ -244,9 +244,7 @@ install:
 	if [ -f diag2.img ]; then cp -f diag2.img $$DESTDIR$(BOOT_DIR); fi
 
 	cp lilo $$DESTDIR$(SBIN_DIR)/lilo
-	strip $$DESTDIR$(SBIN_DIR)/lilo
 	cp ../mkrescue $$DESTDIR$(USRSBIN_DIR)/mkrescue
-	cp ../keytab-lilo.pl $$DESTDIR$(USRSBIN_DIR)/keytab-lilo
 
 dep:
 	sed '/\#\#\# Dependencies/q' <Makefile >tmp_make
