diff -ur lilo-22.7.1.orig/second.S lilo-22.7.1/second.S
--- lilo-22.7.1.orig/second.S	2005-05-04 20:40:31 +0000
+++ lilo-22.7.1/second.S	2005-10-21 11:01:48 +0000
@@ -64,9 +64,23 @@
 # undef WR_ENABLE
 #endif
 
+#undef USE_FS
+
+#ifdef USE_FS
 #define BEG_FS
 #define SEG_FS seg fs
 #define END_FS
+#else
+/* Use DS, not FS, since some BIOSes appear to clobber FS.
+   This implementation comes from LILO 22.1 where it would only be enabled
+   with the non-default setting M386. */
+#define BEG_FS db 0x1e,0x2e,0x8e,0x1e,0x02,0x00
+/*	push ds; \
+	seg cs; \
+	mov	ds,firstseg */
+#define SEG_FS
+#define END_FS pop ds
+#endif
 /* get rid of the following to revert to old int 0x15/fn 0x88 mem scheme */
 #define HIGHMEM_MAX 0x38000000
 
@@ -106,6 +120,9 @@
 
 _main:	jmp	start
 
+#ifndef USE_FS
+firstseg:	dw	0
+#endif
 
 	.org	6
 
@@ -154,8 +171,13 @@
 	.blkb	0x10
 
 start:	cld			! only CLD in the code; there is no STD
+#ifdef USE_FS
 	push	ds
 	pop	fs		! address parameters from here
+#else
+	seg	cs
+	mov	firstseg,ds	! save DS here
+#endif
 
 	seg	cs
 	mov	[init_dx],dx	! save DX passed in from first.S
