--- lilo-22.1/bsect.c.orig	Wed Oct 31 23:56:33 2001
+++ lilo-22.1/bsect.c	Tue Nov 13 17:34:43 2001
@@ -22,6 +22,7 @@
 #define	_I386_STATFS_H	/* two versions of statfs is not good ... */
 #endif
 
+#define	_LINUX_STRING_H_
 #include <linux/fs.h>
 #include <linux/hdreg.h>
 #include <linux/fd.h>
--- lilo-22.1/geometry.c.orig	Tue Oct 30 06:16:55 2001
+++ lilo-22.1/geometry.c	Tue Nov 13 17:34:03 2001
@@ -23,6 +23,8 @@
 #define	_I386_STATFS_H	/* two versions of statfs is not good ... */
 #endif
 #endif
+
+#define	_LINUX_STRING_H_
 #include <linux/fs.h>
 #include <linux/hdreg.h>
 #include <linux/fd.h>
--- lilo-22.1/partition.c.orig	Tue Oct 30 22:00:59 2001
+++ lilo-22.1/partition.c	Tue Nov 13 17:34:28 2001
@@ -21,6 +21,7 @@
 #include <sys/types.h>
 #include <linux/unistd.h>
 #include <limits.h>
+#define	_LINUX_STRING_H_
 #include <linux/fs.h>
 #include <time.h>
 #include "config.h"
