--- lilo-22.1/boot.c.orig	Thu Apr 26 00:48:50 2001
+++ lilo-22.1/boot.c	Sun Dec 23 23:13:35 2001
@@ -241,6 +241,7 @@
     BOOT_SECTOR buff[SETUPSECS-1];
     struct stat st;
     char *pos;
+    char *vanilla_part = part;
     int i;
     int letter = 0;
 
@@ -272,7 +273,7 @@
 
     if (cfg_get_flag(cf_other,"unsafe")) {
 	(void) geo_open_boot(&geo,boot);
-	if (part) die("TABLE and UNSAFE are mutually incompatible.");
+	if (vanilla_part) die("TABLE and UNSAFE are mutually incompatible.");
     }
     else {
 	b_fd = geo_open(&geo,boot,O_RDONLY);
