--- lilo-21.6/geometry.c.orig	Thu Sep 28 03:03:27 2000
+++ lilo-21.6/geometry.c	Mon Jul 23 14:19:35 2001
@@ -271,7 +271,7 @@
     struct floppy_struct fdprm;
     struct hd_geometry hdprm;
 
-    get_all = all || MAJOR(device) != MAJOR_FD;
+    get_all = (all || MAJOR(device) != MAJOR_FD) && MAJOR(device) != MAJOR_LOOP;
     if (!MAJOR(device))
 	die("Trying to map files from unnamed device 0x%04x (NFS ?)",device);
     if (device == MAJMIN_RAM)
@@ -280,6 +280,13 @@
     if (get_all) fd = dev_open(&dev,device,O_NOACCESS);
     else fd = -1; /* pacify GCC */
     switch (MAJOR(device)) {
+	case MAJOR_LOOP:
+	    geo->device = 0;
+	    geo->heads = 2;
+	    geo->cylinders = 80;
+	    geo->sectors = 18;
+	    geo->start = 0;
+	    break;
 	case MAJOR_FD:
 	    geo->device = device & 3;
 	    if (!get_all) {
--- lilo-21.6/lilo.h.orig	Fri Sep 22 03:18:50 2000
+++ lilo-21.6/lilo.h	Sat Jul 21 20:05:52 2001
@@ -32,6 +32,7 @@
 #define MAJMIN_RAM	0x101 /* RAM disk */
 #define MAJOR_FD	2 /* floppy disks */
 #define MAJOR_HD	3 /* IDE-type hard disks */
+#define MAJOR_LOOP	7 /* loopback block devices */
 #define MAJOR_SD	8 /* SCSI disks */
 #define MAJOR_XT	13 /* XT-type hard disks */
 #define MAJOR_IDE2	22 /* IDE on second interface */
