diff -urp fileutils-4.1.11.orig/src/ls.c fileutils-4.1.11/src/ls.c
--- fileutils-4.1.11.orig/src/ls.c	Sat Jul 20 06:50:44 2002
+++ fileutils-4.1.11/src/ls.c	Wed Oct 15 20:04:45 2003
@@ -194,6 +194,8 @@ int rpl_lstat PARAMS((const char *, stru
 #define ID_LENGTH_MAX \
   MAX (LOGIN_NAME_MAX - 1, LONGEST_HUMAN_READABLE)
 
+#define COLUMNS_MAX 1024
+
 enum filetype
   {
     unknown DT_INIT (DT_UNKNOWN),
@@ -1250,7 +1252,7 @@ decode_switches (int argc, char **argv)
       {
 	long int tmp_long;
 	if (xstrtol (p, NULL, 0, &tmp_long, NULL) == LONGINT_OK
-	    && 0 < tmp_long && tmp_long <= INT_MAX)
+	    && 0 < tmp_long && tmp_long <= COLUMNS_MAX)
 	  {
 	    line_length = (int) tmp_long;
 	  }
@@ -1267,8 +1269,8 @@ decode_switches (int argc, char **argv)
   {
     struct winsize ws;
 
-    if (ioctl (STDOUT_FILENO, TIOCGWINSZ, &ws) != -1 && ws.ws_col != 0)
-      line_length = ws.ws_col;
+    if (ioctl (STDOUT_FILENO, TIOCGWINSZ, &ws) != -1 && ws.ws_col > 0)
+      line_length = MIN(ws.ws_col, COLUMNS_MAX);
   }
 #endif
 
@@ -1281,7 +1283,7 @@ decode_switches (int argc, char **argv)
       {
 	long int tmp_long;
 	if (xstrtol (p, NULL, 0, &tmp_long, NULL) == LONGINT_OK
-	    && 0 <= tmp_long && tmp_long <= INT_MAX)
+	    && 0 <= tmp_long && tmp_long <= COLUMNS_MAX)
 	  {
 	    tabsize = (int) tmp_long;
 	  }
@@ -1402,7 +1404,7 @@ decode_switches (int argc, char **argv)
 	  {
 	    long int tmp_long;
 	    if (xstrtol (optarg, NULL, 0, &tmp_long, NULL) != LONGINT_OK
-		|| tmp_long <= 0 || tmp_long > INT_MAX)
+		|| tmp_long <= 0 || tmp_long > COLUMNS_MAX)
 	      error (EXIT_FAILURE, 0, _("invalid line width: %s"),
 		     quotearg (optarg));
 	    line_length = (int) tmp_long;
@@ -1472,7 +1474,7 @@ decode_switches (int argc, char **argv)
 	  {
 	    long int tmp_long;
 	    if (xstrtol (optarg, NULL, 0, &tmp_long, NULL) != LONGINT_OK
-		|| tmp_long < 0 || tmp_long > INT_MAX)
+		|| tmp_long < 0 || tmp_long > COLUMNS_MAX)
 	      error (EXIT_FAILURE, 0, _("invalid tab size: %s"),
 		     quotearg (optarg));
 	    tabsize = (int) tmp_long;
