diff -ur fileutils-4.1.11.orig/src/ls.c fileutils-4.1.11/src/ls.c
--- fileutils-4.1.11.orig/src/ls.c	Sat Jul 20 10:50:44 2002
+++ fileutils-4.1.11/src/ls.c	Sun Aug  4 17:45:16 2002
@@ -1613,7 +1613,7 @@
 
       if (! style)
 	if (! (style = getenv ("TIME_STYLE")))
-	  style = "posix-long-iso";
+	  style = getenv("POSIXLY_CORRECT") ? "posix-long-iso" : "locale";
 
       while (strncmp (style, posix_prefix, sizeof posix_prefix - 1) == 0)
 	{
