--- fileutils-4.0x/src/copy.c.overwrite	Sat Jul 15 11:47:05 2000
+++ fileutils-4.0x/src/copy.c	Thu Aug  3 13:10:31 2000
@@ -495,14 +495,16 @@
 			   _("%s: overwrite %s, overriding mode %04lo? "),
 			   program_name, quote (dst_path),
 			   (unsigned long) (dst_sb.st_mode & CHMOD_MODE_BITS));
+		  if(!yesno())
+			  return (move_mode ? 1 : 0);
 		}
-	      else
+	      else if(x->interactive)
 		{
 		  fprintf (stderr, _("%s: overwrite %s? "),
 			   program_name, quote (dst_path));
+		  if(!yesno())
+			  return (move_mode ? 1 : 0);
 		}
-	      if (!yesno ())
-		return (move_mode ? 1 : 0);
 	    }
 
 	  if (move_mode)
