diff -urpN dhcp-3.0pl2.orig/includes/cf/linux.h dhcp-3.0pl2/includes/cf/linux.h
--- dhcp-3.0pl2.orig/includes/cf/linux.h	Thu May 10 21:28:55 2001
+++ dhcp-3.0pl2/includes/cf/linux.h	Mon May  5 11:02:30 2003
@@ -92,11 +92,11 @@ extern int h_errno;
    directory. */
 
 #ifndef _PATH_DHCPD_DB
-#define _PATH_DHCPD_DB		"/var/state/dhcp/dhcpd.leases"
+#define _PATH_DHCPD_DB		"/var/lib/dhcp/var/state/dhcp/dhcpd.leases"
 #endif
 
 #ifndef _PATH_DHCLIENT_DB
-#define _PATH_DHCLIENT_DB	"/var/state/dhcp/dhclient.leases"
+#define _PATH_DHCLIENT_DB	"/var/lib/dhcp/var/state/dhcp/dhclient.leases"
 #endif
 
 /* Varargs stuff... */
diff -urpN dhcp-3.0pl2.orig/relay/dhcrelay.8 dhcp-3.0pl2/relay/dhcrelay.8
--- dhcp-3.0pl2.orig/relay/dhcrelay.8	Thu Oct  4 23:47:39 2001
+++ dhcp-3.0pl2/relay/dhcrelay.8	Mon May  5 10:53:18 2003
@@ -79,6 +79,14 @@ dhcrelay - Dynamic Host Configuration Pr
 |
 .I discard
 ]
+[
+.B -u
+.I user
+]
+[
+.B -j
+.I chroot-dir
+]
 .I server0
 [
 .I ...serverN
@@ -141,6 +149,12 @@ This can be unhelpful in a system startu
 behaviour, specify the
 .B -q
 flag.
+.PP
+To run Dhcrelay in a chrooted jail, specify the
+.B -u
+option for the dedicated pseudo-user and additional
+.B -j
+for the chroot path (e.g. /var/lib/dhcp).
 .SH RELAY AGENT INFORMATION OPTIONS
 If the
 .B -a
diff -urpN dhcp-3.0pl2.orig/relay/dhcrelay.c dhcp-3.0pl2/relay/dhcrelay.c
--- dhcp-3.0pl2.orig/relay/dhcrelay.c	Thu Apr 19 18:48:53 2001
+++ dhcp-3.0pl2/relay/dhcrelay.c	Mon May  5 10:53:45 2003
@@ -48,6 +48,9 @@ static char ocopyright[] =
 
 #include "dhcpd.h"
 #include "version.h"
+#include <sys/types.h>
+#include <pwd.h>
+#include <unistd.h>
 
 static void usage PROTO ((void));
 
@@ -112,6 +115,30 @@ static char arr [] = "All rights reserve
 static char message [] = "Internet Software Consortium DHCP Relay Agent";
 static char url [] = "For info, please visit http://www.isc.org/products/DHCP";
 
+static int drop_root(char *server_user, char *server_jail)
+{
+	struct passwd *pw;
+	
+	if (!(pw = getpwnam(server_user)))
+		return (-1);
+	if (!pw->pw_uid)
+		return (-1);
+
+	if (chroot(server_jail))
+		return (-1);
+	if (chdir("/"))
+		return (-1);
+
+	if (initgroups(server_user, pw->pw_gid))
+		return (-1);
+	if (setgid(pw->pw_gid))
+		return (-1);
+	if (setuid(pw->pw_uid))
+		return (-1);
+
+	return (0);
+}
+
 int main (argc, argv, envp)
 	int argc;
 	char **argv, **envp;
@@ -124,6 +151,9 @@ int main (argc, argv, envp)
 	isc_result_t status;
 	char *s;
 
+	char *server_user = NULL;
+	char *server_jail = NULL;
+
 	/* Make sure we have stdin, stdout and stderr. */
 	i = open ("/dev/null", O_RDWR);
 	if (i == 0)
@@ -186,6 +216,14 @@ int main (argc, argv, envp)
 			if (++i == argc)
 				usage ();
 			dhcp_max_agent_option_packet_length = atoi (argv [i]);
+		} else if (!strcmp (argv [i], "-u")) {
+			if (++i == argc)
+				usage ();
+			server_user = argv[i];
+		} else if (!strcmp (argv [i], "-j")) {
+			if (++i == argc)
+				usage ();
+			server_jail = argv[i];
 		} else if (!strcmp (argv [i], "-m")) {
 			if (++i == argc)
 				usage ();
@@ -317,6 +355,19 @@ int main (argc, argv, envp)
 		pid = setsid ();
 	}
 
+	/* chroot if possible */	
+	if (server_user && !server_jail) {
+		if (drop_root(server_user, "/var/lib/dhcp") < 0)
+			log_fatal("chroot failed.");
+	}
+	else if (server_jail && !server_user)
+		log_fatal("'-j' is only valid with '-u'");
+	else {
+		if (drop_root(server_user, server_jail) < 0)
+			log_fatal("chroot failed.");
+	}
+
+		
 	/* Start dispatching packets and timeouts... */
 	dispatch ();
 
@@ -449,7 +500,7 @@ static void usage ()
 	log_fatal ("Usage: dhcrelay [-p <port>] [-d] [-D] [-i %s%s%s",
 	       "interface]\n                ",
 	       "[-q] [-a] [-A length] [-m append|replace|forward|discard]\n",
-	       "                [server1 [... serverN]]");
+	       "                [-u user] [-j chroot-dir] [server1 [... serverN]]");
 }
 
 int write_lease (lease)
diff -urpN dhcp-3.0pl2.orig/server/dhcpd.8 dhcp-3.0pl2/server/dhcpd.8
--- dhcp-3.0pl2.orig/server/dhcpd.8	Thu Oct  4 23:49:13 2001
+++ dhcp-3.0pl2/server/dhcpd.8	Mon May  5 10:50:02 2003
@@ -74,6 +74,14 @@ dhcpd - Dynamic Host Configuration Proto
 .I trace-playback-file
 ]
 [
+.B -u
+.I user
+]
+[
+.B -j
+.I chroot-dir 
+]
+[
 .I if0
 [
 .I ...ifN
@@ -236,6 +244,11 @@ using the \fB-lf\fR switch, so that the 
 your existing lease file with its test data.  The DHCP server will
 refuse to operate in playback mode unless you specify an alternate
 lease file.
+.PP
+To run the DHCP server in a chrooted jail, specify the \fB-u\fR
+option with the dedicated pseudo-user and additional \fB-j\fR with the
+chroot path (e.g. /var/lib/dhcp).
+.PP
 .SH CONFIGURATION
 The syntax of the dhcpd.conf(5) file is discussed seperately.   This
 section should be used as an overview of the configuration process,
diff -urpN dhcp-3.0pl2.orig/server/dhcpd.c dhcp-3.0pl2/server/dhcpd.c
--- dhcp-3.0pl2.orig/server/dhcpd.c	Wed Jan 15 00:02:33 2003
+++ dhcp-3.0pl2/server/dhcpd.c	Mon May  5 10:49:40 2003
@@ -55,6 +55,9 @@ static char url [] = "For info, please v
 #include "dhcpd.h"
 #include "version.h"
 #include <omapip/omapip_p.h>
+#include <sys/types.h>
+#include <pwd.h>
+#include <unistd.h>
 
 static void usage PROTO ((void));
 
@@ -204,6 +207,29 @@ static void omapi_listener_start (void *
 	omapi_object_dereference (&listener, MDL);
 }
 
+static int drop_root(char *server_user, char *server_jail)
+{
+	struct passwd *pw;
+
+	if (!(pw = getpwnam(server_user)))
+		return (-1);
+	if (!pw->pw_uid)
+		return (-1);
+
+	if (chroot(server_jail))
+		return (-1);
+	if (chdir("/"))
+		return (-1);
+
+	if (initgroups(server_user, pw->pw_gid))
+		return (-1);
+	if (setgid(pw->pw_gid))
+		return (-1);
+	if (setuid(pw->pw_uid))
+		return (-1);
+	return (0);
+}
+
 int main (argc, argv, envp)
 	int argc;
 	char **argv, **envp;
@@ -237,6 +263,9 @@ int main (argc, argv, envp)
 	char *traceoutfile = (char *)0;
 #endif
 
+	char *server_user = NULL;
+	char *server_jail = NULL;
+	
 	/* Make sure we have stdin, stdout and stderr. */
 	status = open ("/dev/null", O_RDWR);
 	if (status == 0)
@@ -263,6 +292,7 @@ int main (argc, argv, envp)
 	dhcp_common_objects_setup ();
 
 	/* Initially, log errors to stderr as well as to syslogd. */
+	tzset();
 #ifdef SYSLOG_4_2
 	openlog ("dhcpd", LOG_NDELAY);
 	log_priority = DHCPD_LOG_FACILITY;
@@ -331,6 +361,14 @@ int main (argc, argv, envp)
 		} else if (!strcmp (argv [i], "-q")) {
 			quiet = 1;
 			quiet_interface_discovery = 1;
+		} else if (!strcmp (argv [i], "-u")) {
+			if (++i == argc)
+				usage();
+			server_user = argv[i];
+		} else if (!strcmp (argv [i], "-j")) {
+			if (++i == argc)
+				usage();
+			server_jail = argv[i];
 		} else if (!strcmp (argv [i], "--version")) {
 			log_info ("isc-dhcpd-%s", DHCP_VERSION);
 			exit (0);
@@ -565,6 +603,19 @@ int main (argc, argv, envp)
 			log_fatal ("There's already a DHCP server running.");
 	}
 
+	if (pidfilewritten) {
+		if (server_user && !server_jail) {
+			if (drop_root(server_user, "/var/lib/dhcp") < 0)
+				log_fatal("chroot failed.");
+		}
+		else if (server_jail && !server_user)
+			log_fatal("'-j' is only valid with '-u'");
+		else {
+			if (drop_root(server_user, server_jail) < 0)
+				log_fatal("chroot failed.");
+		}
+	}
+	
 	/* If we were requested to log to stdout on the command line,
 	   keep doing so; otherwise, stop. */
 	if (log_perror == -1)
@@ -593,6 +644,16 @@ int main (argc, argv, envp)
 			close (i);
 			pidfilewritten = 1;
 		}
+		if (server_user && !server_jail) {
+			if (drop_root(server_user, "/var/lib/dhcp") < 0)
+				log_fatal("chroot failed.");
+		}
+		else if (server_jail && !server_user)
+			log_fatal("'-j' is only valid with '-u'");
+		else {
+			if (drop_root(server_user, server_jail) < 0)
+				log_fatal("chroot failed.");
+		}
 	}
 #endif /* !DEBUG */
 
@@ -888,6 +949,7 @@ static void usage ()
 
 	log_fatal ("Usage: dhcpd [-p <UDP port #>] [-d] [-f]%s%s%s%s",
 		   "\n             [-cf config-file] [-lf lease-file]",
+		   "\n             [-u user] [-j chroot-dir]",
 #if defined (TRACING)
 		   "\n		   [-tf trace-output-file]",
 		   "\n		   [-play trace-input-file]",
