diff -puNr dhcp-3.0pl2.orig/includes/dhcpd.h dhcp-3.0pl2/includes/dhcpd.h
--- dhcp-3.0pl2.orig/includes/dhcpd.h	Thu Jun 21 23:57:40 2001
+++ dhcp-3.0pl2/includes/dhcpd.h	Fri Jan  7 03:51:29 2005
@@ -303,8 +303,8 @@ struct lease {
 #	define EPHEMERAL_FLAGS		(MS_NULL_TERMINATION | \
 					 UNICAST_BROADCAST_HACK)
 
-	binding_state_t __attribute__ ((mode (__byte__))) binding_state;
-	binding_state_t __attribute__ ((mode (__byte__))) next_binding_state;
+	binding_state_t /* __attribute__ ((mode (__byte__))) */ binding_state;
+	binding_state_t /* __attribute__ ((mode (__byte__))) */ next_binding_state;
 	
 	struct lease_state *state;
 
