diff -urpN dhcp-3.0.4.orig/includes/cf/linux.h dhcp-3.0.4/includes/cf/linux.h
--- dhcp-3.0.4.orig/includes/cf/linux.h	2005-10-14 15:36:27.000000000 +0000
+++ dhcp-3.0.4/includes/cf/linux.h	2006-09-26 15:05:23.000000000 +0000
@@ -83,11 +83,11 @@ extern int h_errno;
    directory. */
 
 #ifndef _PATH_DHCPD_DB
-#define _PATH_DHCPD_DB		"/var/state/dhcp/dhcpd.leases"
+#define _PATH_DHCPD_DB		"/state/dhcpd.leases"
 #endif
 
 #ifndef _PATH_DHCLIENT_DB
-#define _PATH_DHCLIENT_DB	"/var/state/dhcp/dhclient.leases"
+#define _PATH_DHCLIENT_DB	"/var/lib/dhcp/dhclient/state/dhclient.leases"
 #endif
 
 /* Varargs stuff... */
diff -urpN dhcp-3.0.4.orig/relay/dhcrelay.8 dhcp-3.0.4/relay/dhcrelay.8
--- dhcp-3.0.4.orig/relay/dhcrelay.8	2004-10-04 20:40:07.000000000 +0000
+++ dhcp-3.0.4/relay/dhcrelay.8	2006-09-26 15:07:00.000000000 +0000
@@ -77,6 +77,14 @@ dhcrelay - Dynamic Host Configuration Pr
 |
 .I discard
 ]
+[
+.B -u
+.I user
+]
+[
+.B -j
+.I chroot-dir
+]
 .I server0
 [
 .I ...serverN
@@ -139,6 +147,11 @@ This can be unhelpful in a system startu
 behaviour, specify the
 .B -q
 flag.
+.PP
+Upon startup, this version of dhcrelay will switch to a non-root
+pseudo-user and enter a chroot jail.  The default username (\fIdhcp\fR)
+and the default chroot jail directory path (\fI/var/empty\fR)
+may be overridden with the \fB-u\fR and \fB-j\fR options, respectively.
 .SH RELAY AGENT INFORMATION OPTIONS
 If the
 .B -a
@@ -239,7 +252,12 @@ has been written for Internet Systems Co
 by Ted Lemon in cooperation with Vixie
 Enterprises.  To learn more about Internet Systems Consortium,
 see
-.B http://www.isc.org/isc.
+.BR http://www.isc.org/isc .
 To learn more about Vixie
 Enterprises, see
-.B http://www.vix.com.
+.BR http://www.vix.com .
+.PP
+This version of dhcpd has been modified for Openwall GNU/*/Linux
+.RB ( http://www.openwall.com/Owl/ ).
+In particular, the privilege reduction functionality and the \fB-u\fR
+and \fB-j\fR options are Openwall extensions.
diff -urpN dhcp-3.0.4.orig/relay/dhcrelay.c dhcp-3.0.4/relay/dhcrelay.c
--- dhcp-3.0.4.orig/relay/dhcrelay.c	2006-04-27 21:38:30.000000000 +0000
+++ dhcp-3.0.4/relay/dhcrelay.c	2006-09-26 15:08:45.000000000 +0000
@@ -39,6 +39,12 @@ static char ocopyright[] =
 
 #include "dhcpd.h"
 #include "version.h"
+#include <sys/types.h>
+#include <pwd.h>
+#include <unistd.h>
+#define group real_group
+#include <grp.h>
+#undef group
 
 static void usage PROTO ((void));
 
@@ -102,6 +108,32 @@ static char arr [] = "All rights reserve
 static char message [] = "Internet Systems Consortium DHCP Relay Agent";
 static char url [] = "For info, please visit http://www.isc.org/sw/dhcp/";
 
+static int drop_root(char *server_user, char *server_jail)
+{
+	struct passwd *pw;
+
+	if (!(pw = getpwnam(server_user)))
+		return -1;
+	if (!pw->pw_uid)
+		return -1;
+
+	if (chroot(server_jail))
+		return -1;
+	if (chdir("/"))
+		return -1;
+
+#define group real_group
+	if (initgroups(server_user, pw->pw_gid))
+		return -1;
+	if (setgid(pw->pw_gid))
+		return -1;
+#undef group
+	if (setuid(pw->pw_uid))
+		return -1;
+
+	return 0;
+}
+
 int main (argc, argv, envp)
 	int argc;
 	char **argv, **envp;
@@ -114,6 +146,9 @@ int main (argc, argv, envp)
 	isc_result_t status;
 	char *s;
 
+	char *server_user = "dhcp";
+	char *server_jail = "/var/empty";
+
 	/* Make sure we have stdin, stdout and stderr. */
 	i = open ("/dev/null", O_RDWR);
 	if (i == 0)
@@ -185,6 +220,14 @@ int main (argc, argv, envp)
 			if (++i == argc)
 				usage ();
 			dhcp_max_agent_option_packet_length = atoi (argv [i]);
+		} else if (!strcmp (argv [i], "-u")) {
+			if (++i == argc)
+				usage ();
+			server_user = argv[i];
+		} else if (!strcmp (argv [i], "-j")) {
+			if (++i == argc)
+				usage ();
+			server_jail = argv[i];
 		} else if (!strcmp (argv [i], "-m")) {
 			if (++i == argc)
 				usage ();
@@ -316,6 +359,10 @@ int main (argc, argv, envp)
 		pid = setsid ();
 	}
 
+	/* Drop root and enter a chroot jail. */
+	if (drop_root(server_user, server_jail))
+		log_fatal("Failed to drop root.");
+
 	/* Start dispatching packets and timeouts... */
 	dispatch ();
 
@@ -459,7 +506,7 @@ static void usage ()
 		"interface] [-q] [-a]\n                ",
 		"[-c count] [-A length] ",
 		"[-m append|replace|forward|discard]\n",
-		"                [server1 [... serverN]]");
+		"                [-u user] [-j chroot-dir] [server1 [... serverN]]");
 }
 
 int write_lease (lease)
diff -urpN dhcp-3.0.4.orig/server/dhcpd.8 dhcp-3.0.4/server/dhcpd.8
--- dhcp-3.0.4.orig/server/dhcpd.8	2005-09-22 16:19:59.000000000 +0000
+++ dhcp-3.0.4/server/dhcpd.8	2006-09-26 15:09:53.000000000 +0000
@@ -74,6 +74,14 @@ dhcpd - Dynamic Host Configuration Proto
 .I trace-playback-file
 ]
 [
+.B -u
+.I user
+]
+[
+.B -j
+.I chroot-dir
+]
+[
 .I if0
 [
 .I ...ifN
@@ -238,6 +246,11 @@ using the \fB-lf\fR switch, so that the 
 your existing lease file with its test data.  The DHCP server will
 refuse to operate in playback mode unless you specify an alternate
 lease file.
+.PP
+Upon startup, this version of the DHCP server will switch to a non-root
+pseudo-user and enter a chroot jail.  The default username (\fIdhcp\fR)
+and the default chroot jail directory path (\fI/var/lib/dhcp/dhcpd\fR)
+may be overridden with the \fB-u\fR and \fB-j\fR options, respectively.
 .SH CONFIGURATION
 The syntax of the dhcpd.conf(5) file is discussed separately.   This
 section should be used as an overview of the configuration process,
@@ -745,3 +758,8 @@ Consortium.   Version 3 of the DHCP serv
 Information about Internet Systems Consortium is available at
 .B http://www.isc.org/\fR.
 Information about Nominum can be found at \fBhttp://www.nominum.com/\fR.
+.PP
+This version of dhcpd has been modified for Openwall GNU/*/Linux
+.RB ( http://www.openwall.com/Owl/ ).
+In particular, the privilege reduction functionality and the \fB-u\fR
+and \fB-j\fR options are Openwall extensions.
diff -urpN dhcp-3.0.4.orig/server/dhcpd.c dhcp-3.0.4/server/dhcpd.c
--- dhcp-3.0.4.orig/server/dhcpd.c	2006-04-27 21:38:30.000000000 +0000
+++ dhcp-3.0.4/server/dhcpd.c	2006-09-26 15:05:23.000000000 +0000
@@ -46,6 +46,12 @@ static char url [] = "For info, please v
 #include "dhcpd.h"
 #include "version.h"
 #include <omapip/omapip_p.h>
+#include <sys/types.h>
+#include <pwd.h>
+#include <unistd.h>
+#define group real_group
+#include <grp.h>
+#undef group
 
 static void usage PROTO ((void));
 
@@ -193,6 +199,30 @@ static void omapi_listener_start (void *
 	omapi_object_dereference (&listener, MDL);
 }
 
+static int drop_root(char *server_user, char *server_jail)
+{
+	struct passwd *pw;
+
+	if (!(pw = getpwnam(server_user)))
+		return -1;
+	if (!pw->pw_uid)
+		return -1;
+
+	if (chroot(server_jail))
+		return -1;
+	if (chdir("/"))
+		return -1;
+#define group real_group
+	if (initgroups(server_user, pw->pw_gid))
+		return -1;
+	if (setgid(pw->pw_gid))
+		return -1;
+#undef group
+	if (setuid(pw->pw_uid))
+		return -1;
+	return 0;
+}
+
 int main (argc, argv, envp)
 	int argc;
 	char **argv, **envp;
@@ -226,6 +256,9 @@ int main (argc, argv, envp)
 	char *traceoutfile = (char *)0;
 #endif
 
+	char *server_user = "dhcp";
+	char *server_jail = "/var/lib/dhcp/dhcpd";
+
 	/* Make sure we have stdin, stdout and stderr. */
 	status = open ("/dev/null", O_RDWR);
 	if (status == 0)
@@ -252,6 +285,7 @@ int main (argc, argv, envp)
 	dhcp_common_objects_setup ();
 
 	/* Initially, log errors to stderr as well as to syslogd. */
+	tzset();
 #ifdef SYSLOG_4_2
 	openlog ("dhcpd", LOG_NDELAY);
 	log_priority = DHCPD_LOG_FACILITY;
@@ -320,6 +354,14 @@ int main (argc, argv, envp)
 		} else if (!strcmp (argv [i], "-q")) {
 			quiet = 1;
 			quiet_interface_discovery = 1;
+		} else if (!strcmp (argv [i], "-u")) {
+			if (++i == argc)
+				usage();
+			server_user = argv[i];
+		} else if (!strcmp (argv [i], "-j")) {
+			if (++i == argc)
+				usage();
+			server_jail = argv[i];
 		} else if (!strcmp (argv [i], "--version")) {
 			log_info ("isc-dhcpd-%s", DHCP_VERSION);
 			exit (0);
@@ -499,12 +541,6 @@ int main (argc, argv, envp)
 
 	group_write_hook = group_writer;
 
-	/* Start up the database... */
-	db_startup (lftest);
-
-	if (lftest)
-		exit (0);
-
 	/* Discover all the network interfaces and initialize them. */
 	discover_interfaces (DISCOVER_SERVER);
 
@@ -560,6 +596,11 @@ int main (argc, argv, envp)
 		}
 	}
 
+	if (pidfilewritten) {
+		if (drop_root(server_user, server_jail))
+			log_fatal("Failed to drop root.");
+	}
+
 	/* If we were requested to log to stdout on the command line,
 	   keep doing so; otherwise, stop. */
 	if (log_perror == -1)
@@ -588,7 +629,16 @@ int main (argc, argv, envp)
 			close (i);
 			pidfilewritten = 1;
 		}
+		if (drop_root(server_user, server_jail))
+			log_fatal("Failed to drop root.");
 	}
+
+	/* Start up the database... */
+	db_startup (lftest);
+
+	if (lftest)
+		exit (0);
+
 #endif /* !DEBUG */
 
 #if defined (DEBUG_MEMORY_LEAKAGE) || defined (DEBUG_MALLOC_POOL) || \
@@ -881,8 +931,9 @@ static void usage ()
 	log_info (copyright);
 	log_info (arr);
 
-	log_fatal ("Usage: dhcpd [-p <UDP port #>] [-d] [-f]%s%s%s%s",
+	log_fatal ("Usage: dhcpd [-p <UDP port #>] [-d] [-f]%s%s%s%s%s",
 		   "\n             [-cf config-file] [-lf lease-file]",
+		   "\n             [-u user] [-j chroot-dir]",
 #if defined (TRACING)
 		   "\n		   [-tf trace-output-file]",
 		   "\n		   [-play trace-input-file]",
