diff -ur iputils.orig/ping.c iputils/ping.c
--- iputils.orig/ping.c	Thu Nov  2 22:53:23 2000
+++ iputils/ping.c	Sat Apr 14 05:40:54 2001
@@ -778,7 +778,9 @@
 			&iov, 1,
 			ans_data, sizeof(ans_data),
 		};
+#ifdef SO_TIMESTAMP
 		struct cmsghdr *c;
+#endif
 		int cc;
 
 		if (npackets && nreceived >= npackets)
diff -ur iputils.orig/tracepath.c iputils/tracepath.c
--- iputils.orig/tracepath.c	Fri Nov 10 16:52:48 2000
+++ iputils/tracepath.c	Sat Apr 14 05:40:54 2001
@@ -312,7 +312,7 @@
 	}
 
 	for (ttl=1; ttl<32; ttl++) {
-		int res;
+		int res = 0;
 		int i;
 
 		on = ttl;
diff -ur iputils.orig/tracepath6.c iputils/tracepath6.c
--- iputils.orig/tracepath6.c	Fri Nov 10 16:55:03 2000
+++ iputils/tracepath6.c	Sat Apr 14 05:40:54 2001
@@ -346,7 +346,7 @@
 	}
 
 	for (ttl=1; ttl<32; ttl++) {
-		int res;
+		int res = 0;
 		int i;
 
 		on = ttl;
diff -ur iputils.orig/traceroute6.c iputils/traceroute6.c
--- iputils.orig/traceroute6.c	Tue Oct 10 22:15:15 2000
+++ iputils/traceroute6.c	Sat Apr 14 06:02:02 2001
@@ -444,7 +444,7 @@
 	if (datalen < 0 || datalen >= MAXPACKET - sizeof(struct pkt_format)) {
 		Fprintf(stderr,
 		    "traceroute: packet size must be 0 <= s < %u.\n",
-		    MAXPACKET - sizeof(struct pkt_format));
+		    MAXPACKET - (unsigned int)sizeof(struct pkt_format));
 		exit(1);
 	}
 
