diff -ur iputils.orig/ping_common.c iputils/ping_common.c
--- iputils.orig/ping_common.c	Thu Jan 24 22:41:44 2002
+++ iputils/ping_common.c	Wed May 29 05:00:09 2002
@@ -473,7 +473,9 @@
 	char ans_data[4096];
 	struct iovec iov;
 	struct msghdr msg;
+#ifdef SO_TIMESTAMP
 	struct cmsghdr *c;
+#endif
 	int cc;
 	int next;
 	int polling;
diff -ur iputils.orig/tftpd.c iputils/tftpd.c
--- iputils.orig/tftpd.c	Thu Jan 24 03:31:41 2002
+++ iputils/tftpd.c	Wed May 29 05:01:19 2002
@@ -151,6 +151,7 @@
 		int pid;
 		int i, j;
 
+		pid = 0; /* never used */
 		for (i = 1; i < 20; i++) {
 		    pid = fork();
 		    if (pid < 0) {
diff -ur iputils.orig/tracepath.c iputils/tracepath.c
--- iputils.orig/tracepath.c	Sun Sep  2 06:04:17 2001
+++ iputils/tracepath.c	Wed May 29 04:59:25 2002
@@ -330,7 +330,7 @@
 	}
 
 	for (ttl=1; ttl<32; ttl++) {
-		int res;
+		int res = 0;
 		int i;
 
 		on = ttl;
diff -ur iputils.orig/tracepath6.c iputils/tracepath6.c
--- iputils.orig/tracepath6.c	Sun Sep  2 06:03:46 2001
+++ iputils/tracepath6.c	Wed May 29 04:59:25 2002
@@ -364,7 +364,7 @@
 	}
 
 	for (ttl=1; ttl<32; ttl++) {
-		int res;
+		int res = 0;
 		int i;
 
 		on = ttl;
