diff -urN -k.orig iputils/rarpd.c.orig iputils/rarpd.c
--- iputils/rarpd.c.orig	Thu Aug  2 09:47:09 2001
+++ iputils/rarpd.c	Tue Aug 14 20:32:35 2001
@@ -10,6 +10,7 @@
  */
 
 #include <stdio.h>
+#include <stdlib.h>
 #include <syslog.h>
 #include <dirent.h>
 #include <malloc.h>
diff -urN -k.orig iputils/Modules/pg3.c.orig iputils/Modules/pg3.c
--- iputils/Modules/pg3.c.orig	Fri Aug  3 09:52:33 2001
+++ iputils/Modules/pg3.c	Tue Aug 14 20:30:43 2001
@@ -77,7 +77,7 @@
 #include <linux/ptrace.h>
 #include <linux/errno.h>
 #include <linux/ioport.h>
-#include <linux/malloc.h>
+#include <linux/slab.h>
 #include <linux/interrupt.h>
 #include <linux/pci.h>
 #include <linux/delay.h>
diff -Naur iputils/rdisc.c.orig iputils/rdisc.c
--- iputils/rdisc.c.orig	Thu Aug  2 09:41:29 2001
+++ iputils/rdisc.c	Tue Aug 14 20:25:52 2001
@@ -31,6 +31,8 @@
 #include <unistd.h>
 #include <sys/types.h>
 #include <sys/time.h>
+#include <stdlib.h>
+
 /* Do not use "improved" glibc version! */
 #include <linux/limits.h>
 
--- iputils-s20071127/ping.c.orig	2008-09-15 12:00:38 +0400
+++ iputils-s20071127/ping.c	2008-09-15 12:32:32 +0400
@@ -1162,7 +1162,9 @@
 }
 
 
-/* Set Type of Service (TOS) and other Quality of Service relating bits */
+/* Set Type of Service (TOS) / Differentiated Services (DS) and other 
+ * Quality of Service relating bits
+ */
 int parsetos(char *str)
 {
 	const char *cp;
@@ -1183,7 +1185,7 @@
 	}
 
 	if (tos > TOS_MAX) {
-		fprintf(stderr, "ping: the decimal value of TOS bits must be 0-254 (or zero)\n");
+		fprintf(stderr, "ping: the decimal value of TOS bits must be 0-254\n");
 		exit(2);
 	}
 	return(tos);
