--- tar-1.13.19/configure.in.orig	Sat Jan 13 10:59:19 2001
+++ tar-1.13.19/configure.in	Tue Jul 10 15:23:19 2001
@@ -165,7 +165,7 @@
 # Solaris 7 prefers the library name -lrt to the obsolescent name -lposix4.
 tar_LDADD=
 tar_save_LIBS="$LIBS"
-AC_SEARCH_LIBS(clock_gettime, [rt posix4])
+#AC_SEARCH_LIBS(clock_gettime, [rt posix4])
 AC_CHECK_FUNCS(clock_gettime)
 if test " $LIBS" != " $tar_save_LIBS"; then
   tar_LDADD="$ac_cv_search_clock_gettime"
