--- tar-1.13.17/src/names.c.orig	Mon Feb  7 21:45:30 2000
+++ tar-1.13.17/src/names.c	Mon Feb  7 21:48:19 2000
@@ -957,13 +957,15 @@
   name += FILESYSTEM_PREFIX_LEN (name);
 
   if (excluded_filename (excluded_with_slash, name,
-			 FNM_FILE_NAME | FNM_LEADING_DIR))
+			 FNM_FILE_NAME | FNM_LEADING_DIR) ||
+      excluded_filename (excluded_with_slash, name, FNM_FILE_NAME))
     return 1;
 
   for (p = name; *p; p++)
     if (((p == name || ISSLASH (p[-1])) && !ISSLASH (p[0]))
-	&& excluded_filename (excluded_without_slash, p,
-			      FNM_FILE_NAME | FNM_LEADING_DIR))
+	&& (excluded_filename (excluded_without_slash, p,
+			       FNM_FILE_NAME | FNM_LEADING_DIR) ||
+	    excluded_filename (excluded_without_slash, p, FNM_FILE_NAME)))
       return 1;
 
   return 0;
