--- tar-1.15.1/src/create.c.orig	2005-11-13 03:21:37 +0000
+++ tar-1.15.1/src/create.c	2005-11-13 03:22:49 +0000
@@ -1410,6 +1410,16 @@ dump_file0 (struct tar_stat_info *st, ch
 	utime (p, &restore_times);
       return;
     }
+  else if (S_ISSOCK (st->stat.st_mode))
+    {
+      WARN ((0, 0, _("%s: socket ignored"), quotearg_colon (p)));
+      return;
+    }
+  else if (S_ISDOOR (st->stat.st_mode))
+    {
+      WARN ((0, 0, _("%s: door ignored"), quotearg_colon (p)));
+      return;
+    }
   else
     {
       /* Check for multiple links.  */
@@ -1522,16 +1532,6 @@ dump_file0 (struct tar_stat_info *st, ch
 	type = BLKTYPE;
       else if (S_ISFIFO (st->stat.st_mode))
 	type = FIFOTYPE;
-      else if (S_ISSOCK (st->stat.st_mode))
-	{
-	  WARN ((0, 0, _("%s: socket ignored"), quotearg_colon (p)));
-	  return;
-	}
-      else if (S_ISDOOR (st->stat.st_mode))
-	{
-	  WARN ((0, 0, _("%s: door ignored"), quotearg_colon (p)));
-	  return;
-	}
       else
 	{
 	  unknown_file_error (p);
