--- tar-1.15.1/configure.ac.orig	2004-12-21 13:29:02 +0000
+++ tar-1.15.1/configure.ac	2005-01-15 06:05:17 +0000
@@ -93,21 +93,6 @@
 #include <sys/types.h>
 #include <sys/stat.h>])
 
-# Save and restore LIBS so e.g., -lrt, isn't added to it.  Otherwise, *all*
-# programs in the package would end up linked with that potentially-shared
-# library, inducing unnecessary run-time overhead.
-
-# Solaris 2.5.1 needs -lposix4 to get the clock_gettime function.
-# Solaris 7 prefers the library name -lrt to the obsolescent name -lposix4.
-tar_save_LIBS=$LIBS
-  LIB_CLOCK_GETTIME=
-  AC_SEARCH_LIBS(clock_gettime, [rt posix4])
-  case "$ac_cv_search_clock_gettime" in
-    -l*) LIB_CLOCK_GETTIME=$ac_cv_search_clock_gettime;;
-  esac
-  AC_SUBST(LIB_CLOCK_GETTIME)
-  AC_CHECK_FUNCS(clock_gettime)
-LIBS=$tar_save_LIBS
 
 AC_CHECK_FUNCS(fsync lstat mkfifo readlink strerror symlink setlocale utimes)
 AC_CHECK_DECLS([getgrgid],,, [#include <grp.h>])
--- tar-1.15.1/m4/clock_time.m4.orig	2004-03-31 07:44:44 +0000
+++ tar-1.15.1/m4/clock_time.m4	2005-01-15 06:05:17 +0000
@@ -12,17 +12,7 @@ AC_DEFUN([gl_CLOCK_TIME],
   # dnl Persuade glibc <time.h> to declare these functions.
   AC_REQUIRE([AC_GNU_SOURCE])
 
-  # Solaris 2.5.1 needs -lposix4 to get the clock_gettime function.
-  # Solaris 7 prefers the library name -lrt to the obsolescent name -lposix4.
-
-  # Save and restore LIBS so e.g., -lrt, isn't added to it.  Otherwise, *all*
-  # programs in the package would end up linked with that potentially-shared
-  # library, inducing unnecessary run-time overhead.
-  fetish_saved_libs=$LIBS
-    AC_SEARCH_LIBS(clock_gettime, [rt posix4],
-                   [test "$ac_cv_search_clock_gettime" = "none required" ||
-                    LIB_CLOCK_GETTIME=$ac_cv_search_clock_gettime])
+    LIB_CLOCK_GETTIME=
     AC_SUBST(LIB_CLOCK_GETTIME)
     AC_CHECK_FUNCS(clock_gettime clock_settime)
-  LIBS=$fetish_saved_libs
 ])
