--- tar-1.15.1/src/tar.c.orig	2004-12-21 14:11:26 +0000
+++ tar-1.15.1/src/tar.c	2005-05-15 12:33:02 +0000
@@ -709,10 +709,11 @@ parse_opt(int key, char *arg, struct arg
       break;
       
     case 'I':
-      USAGE_ERROR ((0, 0,
-		    _("Warning: the -I option is not supported;"
-		      " perhaps you meant -j or -T?")));
-      break;
+    case 'y':
+      fprintf (stderr,
+	       "Warning: option '%c' is deprecated! Next time use -j instead.\n",
+	       key);
+      /* Fall through to using bzip2. */
       
     case 'j':
       set_use_compress_program_option ("bzip2");
@@ -900,12 +901,6 @@ parse_opt(int key, char *arg, struct arg
 	}
       break;
       
-    case 'y':
-      USAGE_ERROR ((0, 0,
-		    _("Warning: the -y option is not supported;"
-		      " perhaps you meant -j?")));
-      break;
-      
     case 'z':
       set_use_compress_program_option ("gzip");
       break;
