diff -upk.orig tar-1.15.1.orig/src/buffer.c tar-1.15.1/src/buffer.c
--- tar-1.15.1.orig/src/buffer.c	2004-12-21 15:09:24 +0000
+++ tar-1.15.1/src/buffer.c	2005-05-19 21:20:59 +0000
@@ -182,7 +182,6 @@ enum compress_type 
 check_compressed_archive ()
 {
   struct zip_magic *p;
-  size_t status;
   bool sfr, srp;
 
   /* Prepare global data needed for find_next_block: */
diff -upk.orig tar-1.15.1.orig/src/list.c tar-1.15.1/src/list.c
--- tar-1.15.1.orig/src/list.c	2005-05-15 21:29:43 +0000
+++ tar-1.15.1/src/list.c	2005-05-19 21:21:17 +0000
@@ -346,7 +346,6 @@ tar_checksum (union block *header, bool 
 enum read_header
 read_header (bool raw_extended_headers)
 {
-  char *p;
   union block *header;
   union block *header_copy;
   char *bp;
--- tar-1.15.1.orig/tests/genfile.c	2004-09-08 11:50:20 +0000
+++ tar-1.15.1/tests/genfile.c	2006-05-15 22:21:18 +0000
@@ -60,8 +60,8 @@
 /* Block buffer for sparse file */
 char *buffer;
 
-static const char *argp_program_version = "genfile (" PACKAGE ") " VERSION;
-static const char *argp_program_bug_address = "<" PACKAGE_BUGREPORT ">";
+const char *argp_program_version = "genfile (" PACKAGE ") " VERSION;
+const char *argp_program_bug_address = "<" PACKAGE_BUGREPORT ">";
 static char doc[] = N_("genfile generates data files for GNU paxutils test suite");
 
 static struct argp_option options[] = {
