--- net-tools-1.56/man/Makefile.fhs	Tue Jun  6 22:15:50 2000
+++ net-tools-1.56/man/Makefile	Tue Jun  6 22:16:02 2000
@@ -23,6 +23,7 @@
 #
 
 INSTALL=install
+mandir=/usr/man
 
 -include ../config.make
 ifeq ($(I18N),1)
@@ -45,9 +46,9 @@
 			cd $$LANG; \
 			for SECTION in 1 5 8; do \
 				if [ "$$LANG" = "en_US" ] ; then \
-					MAN_DIR=${BASEDIR}/usr/man/man$$SECTION; \
+					MAN_DIR=${BASEDIR}${mandir}/man$$SECTION; \
 				else \
-					MAN_DIR=${BASEDIR}/usr/man/$$LANG/man$$SECTION; \
+					MAN_DIR=${BASEDIR}${mandir}/$$LANG/man$$SECTION; \
 				fi ; \
 				CMDS=`ls *.$$SECTION`; \
 				[ -z "$$CMDS" ] && continue; \
