diff -urp net-tools-1.57.orig/hostname.c net-tools-1.57/hostname.c
--- net-tools-1.57.orig/hostname.c	Sun Feb 20 21:46:45 2000
+++ net-tools-1.57/hostname.c	Fri Jan  7 01:19:46 2005
@@ -97,6 +97,7 @@ static void sethname(char *hname)
 	    fprintf(stderr, _("%s: name too long\n"), program_name);
 	    break;
 	default:
+	    ;
 	}
 	exit(1);
     };
@@ -116,6 +117,7 @@ static void setdname(char *dname)
 	    fprintf(stderr, _("%s: name too long\n"), program_name);
 	    break;
 	default:
+	    ;
 	}
 	exit(1);
     };
@@ -173,6 +175,7 @@ static void showhname(char *hname, int c
 	printf("%s\n", hp->h_name);
 	break;
     default:
+	;
     }
 }
 
diff -urp net-tools-1.57.orig/lib/fddi.c net-tools-1.57/lib/fddi.c
--- net-tools-1.57.orig/lib/fddi.c	Sun Mar  5 11:26:02 2000
+++ net-tools-1.57/lib/fddi.c	Thu Jan 20 04:17:18 2005
@@ -26,6 +26,7 @@
 #error "No FDDI Support in your current Kernelsource Tree."
 #error "Disable HW Type FDDI"
 #endif
+#include <linux/netdevice.h>
 #if __GLIBC__ >= 2
 #include <netinet/if_fddi.h>
 #else
diff -urp net-tools-1.57.orig/lib/inet_sr.c net-tools-1.57/lib/inet_sr.c
--- net-tools-1.57.orig/lib/inet_sr.c	Sun Feb 20 21:46:45 2000
+++ net-tools-1.57/lib/inet_sr.c	Fri Jan  7 01:14:50 2005
@@ -105,6 +105,7 @@ static int INET_setroute(int action, int
     case 2:
        isnet = 0; break;
     default:
+       ;
     }
 
     /* Fill in the other fields. */
diff -urp net-tools-1.57.orig/mii-tool.c net-tools-1.57/mii-tool.c
--- net-tools-1.57.orig/mii-tool.c	Sun May 21 14:31:17 2000
+++ net-tools-1.57/mii-tool.c	Thu Jan 20 04:19:43 2005
@@ -379,17 +379,17 @@ static void watch_one_xcvr(int skfd, cha
 /*--------------------------------------------------------------------*/
 
 const char *usage =
-"usage: %s [-VvRrwl] [-A media,... | -F media] [interface ...]
-       -V, --version               display version information
-       -v, --verbose               more verbose output
-       -R, --reset                 reset MII to poweron state
-       -r, --restart               restart autonegotiation
-       -w, --watch                 monitor for link status changes
-       -l, --log                   with -w, write events to syslog
-       -A, --advertise=media,...   advertise only specified media
-       -F, --force=media           force specified media technology
-media: 100baseT4, 100baseTx-FD, 100baseTx-HD, 10baseT-FD, 10baseT-HD,
-       (to advertise both HD and FD) 100baseTx, 10baseT\n";
+"usage: %s [-VvRrwl] [-A media,... | -F media] [interface ...]\n"
+"       -V, --version               display version information\n"
+"       -v, --verbose               more verbose output\n"
+"       -R, --reset                 reset MII to poweron state\n"
+"       -r, --restart               restart autonegotiation\n"
+"       -w, --watch                 monitor for link status changes\n"
+"       -l, --log                   with -w, write events to syslog\n"
+"       -A, --advertise=media,...   advertise only specified media\n"
+"       -F, --force=media           force specified media technology\n"
+"media: 100baseT4, 100baseTx-FD, 100baseTx-HD, 10baseT-FD, 10baseT-HD,\n"
+"       (to advertise both HD and FD) 100baseTx, 10baseT\n";
 
 int main(int argc, char **argv)
 {
