--- vsftpd-2.0.6/ftpcmdio.h.orig	2008-02-02 01:30:41.000000000 +0000
+++ vsftpd-2.0.6/ftpcmdio.h	2008-06-09 20:40:29.000000000 +0000
@@ -50,6 +50,9 @@
  * The same as vsf_cmdio_write(), and then the calling process is exited. The
  * write is _guaranteed_ to not block (ditching output if neccessary).
  */
+#ifdef __GNUC__
+__attribute__ ((noreturn))
+#endif
 void vsf_cmdio_write_exit(struct vsf_session* p_sess, int status,
                           const char* p_text);
 
--- vsftpd-2.0.6/str.c.orig	2008-02-02 01:30:39.000000000 +0000
+++ vsftpd-2.0.6/str.c	2008-06-09 20:46:17.000000000 +0000
@@ -435,9 +435,7 @@
   struct str_locate_result retval;
   unsigned int num_chars = vsf_sysutil_strlen(p_chars);
   unsigned int i = 0;
-  retval.found = 0;
-  retval.char_found = 0;
-  retval.index = 0;
+  vsf_sysutil_memclr(&retval, sizeof(retval));
   for (; i < p_str->len; ++i)
   {
     unsigned int j = 0;
@@ -462,9 +460,7 @@
   struct str_locate_result retval;
   unsigned int i;
   unsigned int text_len = vsf_sysutil_strlen(p_text);
-  retval.found = 0;
-  retval.char_found = 0;
-  retval.index = 0;
+  vsf_sysutil_memclr(&retval, sizeof(retval));
   if (text_len == 0 || text_len > p_str->len)
   {
     /* Not found */
@@ -489,9 +485,7 @@
   struct str_locate_result retval;
   unsigned int i;
   unsigned int text_len = vsf_sysutil_strlen(p_text);
-  retval.found = 0;
-  retval.char_found = 0;
-  retval.index = 0;
+  vsf_sysutil_memclr(&retval, sizeof(retval));
   if (text_len == 0 || text_len > p_str->len)
   {
     return retval;
--- vsftpd-2.0.6/sysutil.h.orig	2008-02-02 01:30:39.000000000 +0000
+++ vsftpd-2.0.6/sysutil.h	2008-06-09 20:40:35.000000000 +0000
@@ -162,6 +162,9 @@
 unsigned int vsf_sysutil_getpid(void);
 int vsf_sysutil_fork(void);
 int vsf_sysutil_fork_failok(void);
+#ifdef __GNUC__
+__attribute__ ((noreturn))
+#endif
 void vsf_sysutil_exit(int exit_code);
 struct vsf_sysutil_wait_retval
 {
--- vsftpd-2.0.6/twoprocess.c.orig	2008-02-12 03:18:34.000000000 +0000
+++ vsftpd-2.0.6/twoprocess.c	2008-06-09 20:40:35.000000000 +0000
@@ -41,6 +41,9 @@
                                 const struct mystr* p_user_str,
                                 const struct mystr* p_orig_user_str);
 
+#ifdef __GNUC__
+__attribute__ ((noreturn))
+#endif
 static void
 handle_sigchld(int duff)
 {
--- vsftpd-2.0.6/utility.h.orig	2008-02-02 01:30:41.000000000 +0000
+++ vsftpd-2.0.6/utility.h	2008-06-09 20:40:35.000000000 +0000
@@ -10,6 +10,9 @@
  * PARAMETERS
  * p_text       - text string describing why the process is exiting
  */
+#ifdef __GNUC__
+__attribute__ ((noreturn))
+#endif
 void die(const char* p_text);
 
 /* die2()
@@ -20,6 +23,9 @@
  * p_text1      - text string describing why the process is exiting
  * p_text2      - text to safely concatenate to p_text1
  */
+#ifdef __GNUC__
+__attribute__ ((noreturn))
+#endif
 void die2(const char* p_text1, const char* p_text2);
 
 /* bug()
@@ -29,6 +35,9 @@
  * PARAMETERS
  * p_text       - text string describing what bug trap has triggered
  *       */
+#ifdef __GNUC__
+__attribute__ ((noreturn))
+#endif
 void bug(const char* p_text);
 
 /* vsf_exit()
@@ -38,6 +47,9 @@
  * PARAMETERS
  * p_text       - text string describing why the process is exiting
  */
+#ifdef __GNUC__
+__attribute__ ((noreturn))
+#endif
 void vsf_exit(const char* p_text);
 
 #endif
