--- vsftpd-2.0.3.orig/sysdeputil.c	2004-09-14 01:18:54 +0000
+++ vsftpd-2.0.3/sysdeputil.c	2005-11-27 16:40:02 +0000
@@ -139,7 +139,7 @@
 /* END config */
 
 /* PAM support - we include our own dummy version if the system lacks this */
-#include <security/pam_appl.h>
+#include <security/pam_userpass.h>
 
 /* No PAM? Try getspnam() with a getpwnam() fallback */
 #ifndef VSF_SYSDEP_HAVE_PAM
@@ -273,9 +273,7 @@ vsf_sysdep_check_auth(const struct mystr
 #else /* VSF_SYSDEP_HAVE_PAM */
 
 static pam_handle_t* s_pamh;
-static struct mystr s_pword_str;
-static int pam_conv_func(int nmsg, const struct pam_message** p_msg,
-                         struct pam_response** p_reply, void* p_addata);
+static pam_userpass_t userpass;
 static void vsf_auth_shutdown(void);
 
 int
@@ -286,14 +284,15 @@ vsf_sysdep_check_auth(const struct mystr
   int retval;
   struct pam_conv the_conv =
   {
-    &pam_conv_func,
-    0
+    pam_userpass_conv,
+    &userpass
   };
   if (s_pamh != 0)
   {
     bug("vsf_sysdep_check_auth");
   }
-  str_copy(&s_pword_str, p_pass_str);
+  userpass.user = str_getbuf(p_user_str);
+  userpass.pass = str_getbuf(p_pass_str);
   retval = pam_start(tunable_pam_service_name,
                      str_getbuf(p_user_str), &the_conv, &s_pamh);
   if (retval != PAM_SUCCESS)
@@ -371,6 +370,7 @@ vsf_auth_shutdown(void)
   vsf_remove_uwtmp();
 }
 
+#if 0
 static int
 pam_conv_func(int nmsg, const struct pam_message** p_msg,
               struct pam_response** p_reply, void* p_addata)
@@ -406,6 +406,7 @@ pam_conv_func(int nmsg, const struct pam
   *p_reply = p_resps;
   return PAM_SUCCESS;
 }
+#endif
 
 #endif /* VSF_SYSDEP_HAVE_PAM */
 
