--- vsftpd-2.1.0/tunables.c.orig	2008-12-18 01:42:45.000000000 +0000
+++ vsftpd-2.1.0/tunables.c	2009-05-25 14:48:39.000000000 +0000
@@ -173,7 +173,7 @@
   tunable_userlist_deny = 1;
   tunable_use_localtime = 0;
   tunable_check_shell = 1;
-  tunable_hide_ids = 0;
+  tunable_hide_ids = 1;
   tunable_listen = 1;
   tunable_port_promiscuous = 0;
   tunable_passwd_chroot_enable = 0;
@@ -242,19 +242,19 @@
   /* -rw------- */
   tunable_chown_upload_mode = 0600;
 
-  install_str_setting("/usr/share/empty", &tunable_secure_chroot_dir);
+  install_str_setting("/var/empty", &tunable_secure_chroot_dir);
   install_str_setting("ftp", &tunable_ftp_username);
   install_str_setting("root", &tunable_chown_username);
   install_str_setting("/var/log/xferlog", &tunable_xferlog_file);
-  install_str_setting("/var/log/vsftpd.log", &tunable_vsftpd_log_file);
+  install_str_setting("/var/log/vsftpd", &tunable_vsftpd_log_file);
   install_str_setting(".message", &tunable_message_file);
-  install_str_setting("nobody", &tunable_nopriv_user);
+  install_str_setting("vsftpd", &tunable_nopriv_user);
   install_str_setting(0, &tunable_ftpd_banner);
-  install_str_setting("/etc/vsftpd.banned_emails", &tunable_banned_email_file);
-  install_str_setting("/etc/vsftpd.chroot_list", &tunable_chroot_list_file);
-  install_str_setting("ftp", &tunable_pam_service_name);
+  install_str_setting("/etc/vsftpd/banned_emails", &tunable_banned_email_file);
+  install_str_setting("/etc/vsftpd/chroot_list", &tunable_chroot_list_file);
+  install_str_setting("vsftpd", &tunable_pam_service_name);
   install_str_setting("ftp", &tunable_guest_username);
-  install_str_setting("/etc/vsftpd.user_list", &tunable_userlist_file);
+  install_str_setting("/etc/vsftpd/user_list", &tunable_userlist_file);
   install_str_setting(0, &tunable_anon_root);
   install_str_setting(0, &tunable_local_root);
   install_str_setting(0, &tunable_banner_file);
@@ -267,7 +267,7 @@
   install_str_setting(0, &tunable_hide_file);
   install_str_setting(0, &tunable_deny_file);
   install_str_setting(0, &tunable_user_sub_token);
-  install_str_setting("/etc/vsftpd.email_passwords",
+  install_str_setting("/etc/vsftpd/email_passwords",
                       &tunable_email_password_file);
   install_str_setting("/usr/share/ssl/certs/vsftpd.pem",
                       &tunable_rsa_cert_file);
--- vsftpd-2.1.0/vsftpd.conf.orig	2008-12-17 20:31:40.000000000 +0000
+++ vsftpd-2.1.0/vsftpd.conf	2009-05-25 19:10:58.000000000 +0000
@@ -1,18 +1,16 @@
-# Example config file /etc/vsftpd.conf
+# The configuration file for vsftpd.
 #
-# The default compiled in settings are fairly paranoid. This sample file
-# loosens things up a bit, to make the ftp daemon more usable.
-# Please see vsftpd.conf.5 for all compiled in defaults.
+# Please see vsftpd.conf(5) for all compiled in defaults.
 #
 # READ THIS: This example file is NOT an exhaustive list of vsftpd options.
-# Please read the vsftpd.conf.5 manual page to get a full idea of vsftpd's
+# Please read the vsftpd.conf(5) manual page to get a full idea of vsftpd's
 # capabilities.
 #
 # Allow anonymous FTP? (Beware - allowed by default if you comment this out).
-anonymous_enable=YES
+anonymous_enable=NO
 #
-# Uncomment this to allow local users to log in.
-#local_enable=YES
+# Allow local users to log in?
+local_enable=NO
 #
 # Uncomment this to enable any form of FTP write command.
 #write_enable=YES
@@ -37,9 +35,20 @@
 # Activate logging of uploads/downloads.
 xferlog_enable=YES
 #
+# Uncomment this to disallow the PORT method of obtaining a data connection.
+#port_enable=NO
+#
+# Uncomment this to disallow the PASV method of obtaining a data connection.
+#pasv_enable=NO
+#
 # Make sure PORT transfer connections originate from port 20 (ftp-data).
 connect_from_port_20=YES
 #
+# The port range to allocate for PASV-style data connections.
+# Can be used to specify a narrow port range to assist firewalling.
+pasv_min_port=49152
+pasv_max_port=65535
+#
 # If you want, you can arrange for uploaded anonymous files to be owned by
 # a different user. Note! Using "root" for uploaded files is not
 # recommended!
@@ -62,7 +71,7 @@
 #
 # It is recommended that you define on your system a unique user which the
 # ftp server can use as a totally isolated and unprivileged user.
-#nopriv_user=ftpsecure
+#nopriv_user=vsftpd
 #
 # Enable this and the server will recognise asynchronous ABOR requests. Not
 # recommended for security (the code is non-trivial). Not enabling it,
@@ -77,8 +86,8 @@
 # predicted this attack and has always been safe, reporting the size of the
 # raw file.
 # ASCII mangling is a horrible feature of the protocol.
-#ascii_upload_enable=YES
-#ascii_download_enable=YES
+ascii_upload_enable=YES
+ascii_download_enable=YES
 #
 # You may fully customise the login banner string:
 #ftpd_banner=Welcome to blah FTP service.
@@ -87,14 +96,17 @@
 # useful for combatting certain DoS attacks.
 #deny_email_enable=YES
 # (default follows)
-#banned_email_file=/etc/vsftpd.banned_emails
+#banned_email_file=/etc/vsftpd/banned_emails
 #
 # You may specify an explicit list of local users to chroot() to their home
 # directory. If chroot_local_user is YES, then this list becomes a list of
 # users to NOT chroot().
+# Warning: these features have non-trivial security implications, especially
+# if the users also have shell access. Only enable if you know what you are
+# doing (and you probably don't).
 #chroot_list_enable=YES
 # (default follows)
-#chroot_list_file=/etc/vsftpd.chroot_list
+#chroot_list_file=/etc/vsftpd/chroot_list
 #
 # You may activate the "-R" option to the builtin ls. This is disabled by
 # default to avoid remote users being able to cause excessive I/O on large
@@ -105,9 +117,19 @@
 # When "listen" directive is enabled, vsftpd runs in standalone mode and
 # listens on IPv4 sockets. This directive cannot be used in conjunction
 # with the listen_ipv6 directive.
-listen=YES
+listen=NO
 #
 # This directive enables listening on IPv6 sockets. To listen on IPv4 and IPv6
 # sockets, you must run two copies of vsftpd whith two configuration files.
 # Make sure, that one of the listen options is commented !!
 #listen_ipv6=YES
+#
+# If enabled, all user and group information in directory listings will be
+# displayed as "ftp".
+# The default is to hide user and group information.
+#hide_ids=YES
+#
+# If enabled, vsftpd will display directory listings with the time in your
+# local time zone. The default is to display GMT. The times returned by the
+# MDTM FTP command are also affected by this option.
+#use_localtime=YES
--- vsftpd-2.1.0/vsftpd.conf.5.orig	2008-12-18 01:44:21.000000000 +0000
+++ vsftpd-2.1.0/vsftpd.conf.5	2009-05-25 14:28:43.000000000 +0000
@@ -138,7 +138,7 @@
 different if chroot_local_user is set to YES. In this case, the list becomes
 a list of users which are NOT to be placed in a chroot() jail.
 By default, the file containing this list is
-/etc/vsftpd.chroot_list, but you may override this with the
+/etc/vsftpd/chroot_list, but you may override this with the
 .BR chroot_list_file
 setting.
 
@@ -177,7 +177,7 @@
 .B deny_email_enable
 If activated, you may provide a list of anonymous password e-mail responses
 which cause login to be denied. By default, the file containing this list is
-/etc/vsftpd.banned_emails, but you may override this with the
+/etc/vsftpd/banned_emails, but you may override this with the
 .BR banned_email_file
 setting.
 
@@ -205,7 +205,7 @@
 If enabled, two log files are generated in parallel, going by default to
 .BR /var/log/xferlog
 and
-.BR /var/log/vsftpd.log .
+.BR /var/log/vsftpd .
 The former is a wu-ftpd style transfer log, parseable by standard tools. The
 latter is vsftpd's own style log.
 
@@ -262,7 +262,7 @@
 If enabled, all user and group information in directory listings will be
 displayed as "ftp".
 
-Default: NO
+Default: YES
 .TP
 .B implicit_ssl
 If enabled, an SSL handshake is the first thing expect on all connections
@@ -430,7 +430,7 @@
 file specified by the
 .BR email_password_file
 setting. The file format is one password per line, no extra whitespace. The
-default filename is /etc/vsftpd.email_passwords.
+default filename is /etc/vsftpd/email_passwords.
 
 Default: NO
 .TP
@@ -514,7 +514,7 @@
 Default: NO
 .TP
 .B syslog_enable
-If enabled, then any log output which would have gone to /var/log/vsftpd.log
+If enabled, then any log output which would have gone to /var/log/vsftpd
 goes to the system log instead. Logging is done under the FTPD facility.
 
 Default: NO
@@ -600,7 +600,7 @@
 .TP
 .B xferlog_enable
 If enabled, a log file will be maintained detailling uploads and downloads.
-By default, this file will be placed at /var/log/vsftpd.log, but this location
+By default, this file will be placed at /var/log/vsftpd, but this location
 may be overridden using the configuration setting
 .BR vsftpd_log_file .
 
@@ -761,7 +761,7 @@
 .BR deny_email_enable
 is enabled.
 
-Default: /etc/vsftpd.banned_emails
+Default: /etc/vsftpd/banned_emails
 .TP
 .B banner_file
 This option is the name of a file containing text to display when someone
@@ -798,7 +798,7 @@
 is enabled, then the list file becomes a list of users to NOT place in a
 chroot() jail.
 
-Default: /etc/vsftpd.chroot_list
+Default: /etc/vsftpd/chroot_list
 .TP
 .B cmds_allowed
 This options specifies a comma separated list of allowed FTP commands (post
@@ -859,7 +859,7 @@
 .BR secure_email_list_enable
 setting.
 
-Default: /etc/vsftpd.email_passwords
+Default: /etc/vsftpd/email_passwords
 .TP
 .B ftp_username
 This is the name of the user we use for handling anonymous FTP. The home
@@ -930,12 +930,12 @@
 than nobody. The user nobody tends to be used for rather a lot of important
 things on most machines.
 
-Default: nobody
+Default: vsftpd
 .TP
 .B pam_service_name
 This string is the name of the PAM service vsftpd will use.
 
-Default: ftp
+Default: vsftpd
 .TP
 .B pasv_address
 Use this option to override the IP address that vsftpd will advertise in
@@ -964,7 +964,7 @@
 directory should not be writable by the ftp user. This directory is used
 as a secure chroot() jail at times vsftpd does not require filesystem access.
 
-Default: /usr/share/empty
+Default: /var/empty
 .TP
 .B ssl_ciphers
 This option can be used to select which SSL ciphers vsftpd will allow for
@@ -982,10 +982,10 @@
 with an example. If you set
 .BR user_config_dir
 to be
-.BR /etc/vsftpd_user_conf
+.BR /etc/vsftpd/user_conf
 and then log on as the user "chris", then vsftpd will apply the settings in
 the file
-.BR /etc/vsftpd_user_conf/chris
+.BR /etc/vsftpd/user_conf/chris
 for the duration of the session. The format of this file is as detailed in
 this manual page! PLEASE NOTE that not all settings are effective on a
 per-user basis. For example, many settings only prior to the user's session
@@ -1021,7 +1021,7 @@
 .BR userlist_enable
 option is active.
 
-Default: /etc/vsftpd.user_list
+Default: /etc/vsftpd/user_list
 .TP
 .B vsftpd_log_file
 This option is the name of the file to which we write the vsftpd style
@@ -1035,7 +1035,7 @@
 .BR syslog_enable ,
 then this file is not written and output is sent to the system log instead.
 
-Default: /var/log/vsftpd.log
+Default: /var/log/vsftpd
 .TP
 .B xferlog_file
 This option is the name of the file to which we write the wu-ftpd style
