--- console-data-1999.08.29/consolefonts/Makefile.am.orig	Wed Aug 25 01:02:30 1999
+++ console-data-1999.08.29/consolefonts/Makefile.am	Fri Nov  5 20:43:56 1999
@@ -73,7 +73,8 @@
 lat4a-08.psf lat4a-10.psf lat4a-12.psf lat4a-14.psf lat4a-16+.psf lat4a-16.psf lat4a-19.psf \
 lat7-14.psf \
 \
-lat0-sun16.psf lat2-sun16.psf  \
+lat0-sun16.psf lat2-sun16.psf \
+ucw08.psf ucw11m.psf ucw11z.psf ucw16.psf lat5-12.psf lat5-14.psf lat5-16.psf \
 \
 lat0-08.psf lat0-10.psf lat0-12.psf lat0-14.psf lat0-16.psf \
 \
--- console-data-1999.08.29/consoletrans/Makefile.am.orig	Sun Aug 29 20:55:22 1999
+++ console-data-1999.08.29/consoletrans/Makefile.am	Fri Nov  5 20:45:13 1999
@@ -59,13 +59,14 @@
 iso07.sfm iso08.sfm iso09.sfm iso10.sfm iso14.sfm iso15.sfm \
 ECMA144.sfm cp437.sfm def.sfm ethiopic.sfm koi8-r.sfm \
 lat1.sfm lat1u.sfm lat2u.sfm lat4.sfm lat4u.sfm \
-armscii8.sfm cp850.sfm cp850b.sfm tcvn.sfm viscii.sfm \
+armscii8.sfm cp850.sfm cp850b.sfm tcvn.sfm viscii.sfm latin2-ucw.uni \
 koi8u.sfm ruscii.sfm
 
 OLD_ACM_FILES = null.trans space.trans trivial.trans zero.trans \
 vga2iso.trans koi2alt.trans koi8u2ruscii.trans \
 cp437_to_iso01.trans cp850_to_iso01.trans \
-viscii1.0_to_tcvn.trans viscii1.0_to_viscii1.1.trans
+viscii1.0_to_tcvn.trans viscii1.0_to_viscii1.1.trans latin2-ucw.scrnmap
+
 
 ACM_FILES = \
 cp437.acm iso01.acm iso02.acm iso03.acm iso04.acm iso05.acm iso06.acm \
