--- console-tools-0.3.3/console-data-1999.08.29/keymaps/i386/qwertz/de-latin1.kmap.keymaps	Sun Aug 27 17:33:50 2000
+++ console-tools-0.3.3/console-data-1999.08.29/keymaps/i386/qwertz/de-latin1.kmap	Sun Aug 27 17:43:30 2000
@@ -3,26 +3,27 @@
 # Some changes due to Olaf Flebbe (flebbe@pluto.tat.physik.uni-tuebingen.de)
 # More changes by Jochen Hein <jochen.hein@delphi.central.de>.
 
-keymaps 0-2,4-6,8-10,12
+keymaps 0-6,8-10,12
 alt_is_meta
 include "qwertz-layout"
 	altgr keycode 16 = at
+	altgr keycode 19 = registered
 	altgr keycode 50 = mu
 include "linux-with-alt-and-altgr"
 	plain keycode  83 = KP_Comma
 strings as usual
 
-#             Normal           Shift            AltGr           Strg
+#             Normal           Shift            AltGr           AltGr+Shift
 keycode   1 = Escape           Escape
-keycode   2 = one              exclam
+keycode   2 = one              exclam           onesuperior     exclamdown
 keycode   3 = two              quotedbl         twosuperior     nul
 keycode   4 = three            section          threesuperior   Escape
 keycode   5 = four             dollar
 keycode   6 = five             percent
 keycode   7 = six              ampersand
 keycode   8 = seven            slash            braceleft
-keycode   9 = eight            parenleft        bracketleft
-keycode  10 = nine             parenright       bracketright
+keycode   9 = eight            parenleft        bracketleft     guillemotleft
+keycode  10 = nine             parenright       bracketright    guillemotright
         control altgr keycode 10 = Control_bracketright
 keycode  11 = zero             equal            braceright
 keycode  12 = ssharp           question         backslash
@@ -43,6 +44,10 @@
 keycode  41 = dead_circumflex   degree   Meta_asciicircum        Control_asciicircum
 keycode  42 = Shift
 keycode  43 = numbersign       apostrophe
+keycode  46 = +c             +C             copyright
+        control keycode  46 = Control_c
+keycode  50 = +m             +M             mu
+        control keycode  50 = Control_m
 keycode  51 = comma            semicolon
 keycode  52 = period           colon
 keycode  53 = minus            underscore       Meta_minus
--- console-tools-0.3.3/console-data-1999.08.29/keymaps/i386/qwerty/se-latin1.kmap.keymaps	Sun Aug 29 21:15:28 1999
+++ console-tools-0.3.3/console-data-1999.08.29/keymaps/i386/qwerty/se-latin1.kmap	Sun Aug 27 17:37:48 2000
@@ -4,7 +4,7 @@
 # where no AltGr is used to get the three national characters.
 
 charset "iso-8859-1"
-keymaps 0-2,4-6,8,12
+keymaps 0-6,8,12
 include "qwerty-layout"
 include "linux-with-alt-and-altgr"
 	plain keycode  83 = KP_Comma
@@ -12,68 +12,52 @@
 
 keycode   1 = Escape
 	alt     keycode   1 = Meta_Escape
-keycode   2 = one              exclam
-	alt     keycode   2 = Meta_one
-keycode   3 = two              quotedbl         at
-	control keycode   3 = nul
-	alt     keycode   3 = Meta_two
-keycode   4 = three            numbersign       sterling
-	control keycode   4 = Escape
-	alt     keycode   4 = Meta_three
-keycode   5 = four             ccedilla
-	control keycode   5 = Control_backslash
-	alt     keycode   5 = Meta_four
-	altgr   keycode   5 = dollar
-keycode   6 = five             percent
-	control keycode   6 = Control_bracketright
-	alt     keycode   6 = Meta_five
-keycode   7 = six              ampersand        notsign
-	control keycode   7 = Control_asciicircum
-	alt     keycode   7 = Meta_six
-keycode   8 = seven            slash            braceleft
-	control keycode   8 = Control_underscore
-	alt     keycode   8 = Meta_seven
-keycode   9 = eight            parenleft        bracketleft
-	control keycode   9 = Delete
-	alt     keycode   9 = Meta_eight
-keycode  10 = nine             parenright    bracketright
-	alt     keycode  10 = Meta_nine
-keycode  11 = zero             equal         braceright
-	alt     keycode  11 = Meta_zero
-keycode  12 = plus       question         backslash
-	control keycode  12 = Control_underscore
-	alt     keycode  12 = Meta_minus
-keycode  13 = dead_acute  dead_grave
-	alt     keycode  13 = Meta_equal
+keycode   2 = one           exclam        exclamdown       onesuperior
+keycode   3 = two           quotedbl      at               twosuperior
+keycode   4 = three         numbersign    sterling         threesuperior
+keycode   5 = four          currency      dollar           onequarter
+keycode   6 = five          percent       cent             cent
+keycode   7 = six           ampersand     yen
+keycode   8 = seven         slash         braceleft        division
+keycode   9 = eight         parenleft     bracketleft      guillemotleft
+keycode  10 = nine          parenright    bracketright     guillemotright
+keycode  11 = zero          equal         braceright       degree
+keycode  12 = plus          question      backslash        questiondown
+keycode  13 = dead_acute    dead_grave    notsign          plusminus
 keycode  14 = Delete
 	alt     keycode  14 = Meta_Delete
 keycode  15 = Tab
 	alt     keycode  15 = Meta_Tab
+keycode  19 = +r             +R             registered
+keycode  20 = +t             +T             thorn            THORN
 keycode  26 = +aring          +Aring
 	control keycode  26 = Escape
 	alt     keycode  26 = Meta_bracketleft
-keycode  27 = dead_diaeresis   asciicircum           asciitilde
+keycode  27 = dead_diaeresis   dead_circumflex           dead_tilde
 	control keycode  27 = Control_bracketright
 	alt     keycode  27 = Meta_bracketright
 keycode  28 = Return
 	alt     keycode  28 = Meta_Control_m
 keycode  29 = Control
-keycode  39 = +odiaeresis       +Odiaeresis
-	alt     keycode  39 = Meta_semicolon
-keycode  40 = +adiaeresis       +Adiaeresis
-	control keycode  40 = Control_g
-	alt     keycode  40 = Meta_apostrophe
-keycode  41 = section          degree
-	control keycode  41 = nul
-	alt     keycode  41 = Meta_grave
+keycode  32 = +d             +D             +eth              +ETH
+	control keycode 32 = Control_d
+keycode  39 = +odiaeresis       +Odiaeresis  +oslash     +Oslash
+	control keycode 39 = Control_o
+keycode  40 = +adiaeresis       +Adiaeresis  +ae         +AE
+	control keycode 40 = Control_a
+
+keycode  41 = section          onehalf       paragraph     threequarters
 keycode  42 = Shift
 keycode  43 = apostrophe       asterisk
 	control keycode  43 = Control_backslash
 	alt     keycode  43 = Meta_backslash
+keycode  46 = +c             +C             copyright
+	control keycode  46 = Control_c
+keycode  50 = +m             +M             mu
+	control keycode  50 = Control_m
 keycode  51 = comma            semicolon
 	alt     keycode  51 = Meta_comma
-keycode  52 = period           colon
-	alt     keycode  52 = Meta_period
+keycode  52 = period           colon      periodcentered
 keycode  53 = minus            underscore
 	control keycode  53 = Delete
 	alt     keycode  53 = Meta_slash
--- console-tools-0.3.3/console-data-1999.08.29/keymaps/i386/qwerty/dk-latin1.kmap.keymaps	Sun Aug 29 21:15:26 1999
+++ console-tools-0.3.3/console-data-1999.08.29/keymaps/i386/qwerty/dk-latin1.kmap	Sun Aug 27 17:40:02 2000
@@ -2,7 +2,7 @@
 # Niels Kristian Bech Jensen <nkbj@kemi.aau.dk>:
 #  shift keycode 5 is currency, not dollar
 #
-keymaps 0-2,4-6,8,12
+keymaps 0-6,8,12
 include "qwerty-layout"
 include "linux-with-alt-and-altgr"
 	plain keycode  83 = KP_Comma
@@ -10,43 +10,24 @@
 
 keycode   1 = Escape           Escape
 	alt     keycode   1 = Meta_Escape
-keycode   2 = one              exclam
-	alt     keycode   2 = Meta_one
-keycode   3 = two              quotedbl         at
-	control keycode   3 = nul
-	alt     keycode   3 = Meta_two
-keycode   4 = three            numbersign       sterling
-	control keycode   4 = Escape
-	alt     keycode   4 = Meta_three
-# keycode   5 = four             dollar           dollar
-keycode   5 = four             currency           dollar
-	control keycode   5 = Control_backslash
-	alt     keycode   5 = Meta_four
-keycode   6 = five             percent
-	control keycode   6 = Control_bracketright
-	alt     keycode   6 = Meta_five
-keycode   7 = six              ampersand
-	control keycode   7 = Control_asciicircum
-	alt     keycode   7 = Meta_six
-keycode   8 = seven            slash            braceleft
-	control keycode   8 = Control_underscore
-	alt     keycode   8 = Meta_seven
-keycode   9 = eight            parenleft        bracketleft
-	control keycode   9 = Delete
-	alt     keycode   9 = Meta_eight
-keycode  10 = nine             parenright       bracketright
-	alt     keycode  10 = Meta_nine
-keycode  11 = zero             equal            braceright
-	alt     keycode  11 = Meta_zero
-keycode  12 = plus             question
-	control keycode  12 = Control_underscore
-	alt     keycode  12 = Meta_minus
-keycode  13 = dead_acute       dead_grave       bar
-	alt     keycode  13 = Meta_equal
+keycode   2 = one              exclam         exclamdown      onesuperior
+keycode   3 = two              quotedbl       at              twosuperior
+keycode   4 = three            numbersign     sterling        threesuperior
+keycode   5 = four             currency       dollar          onequarter
+keycode   6 = five             percent        cent            cent
+keycode   7 = six              ampersand      yen
+keycode   8 = seven            slash          braceleft       division
+keycode   9 = eight            parenleft      bracketleft     guillemotleft
+keycode  10 = nine             parenright     bracketright    guillemotright
+keycode  11 = zero             equal          braceright      degree
+keycode  12 = plus             question       plusminus       questiondown
+keycode  13 = dead_acute       dead_grave     bar
 keycode  14 = Delete           Delete
 	alt     keycode  14 = Meta_Delete
 keycode  15 = Tab              Tab
 	alt     keycode  15 = Meta_Tab
+keycode  19 = +r               +R             registered
+keycode  20 = +t               +T             +thorn          +THORN
 keycode  26 = +aring            +Aring
 	control keycode  26 = Escape
 	alt     keycode  26 = Meta_bracketleft
@@ -56,18 +37,22 @@
 keycode  28 = Return
 	alt     keycode  28 = Meta_Control_m
 keycode  29 = Control
-keycode  39 = +ae               +AE
+keycode  31 = +s               +S             ssharp
+	control keycode  31 = Control_s
+keycode  32 = +d               +D             +eth            +ETH
+	control keycode  32 = Control_d
+keycode  39 = +ae              +AE
 	alt     keycode  39 = Meta_semicolon
 keycode  40 = +oslash             +Oslash
 	control keycode  40 = Control_g
 	alt     keycode  40 = Meta_apostrophe
-keycode  41 = onehalf          section
-	control keycode  41 = nul
-	alt     keycode  41 = Meta_grave
+keycode  41 = onehalf          section      threequarters     paragraph
 keycode  42 = Shift
-keycode  43 = apostrophe       asterisk
-	control keycode  43 = Control_backslash
-	alt     keycode  43 = Meta_backslash
+keycode  43 = apostrophe       asterisk     onehalf           threequarters
+keycode  46 = +c               +C           copyright
+	control keycode 46 = Control_c
+keycode  50 = +m               +M           mu
+	control keycode 50 = Control_m
 keycode  51 = comma            semicolon
 	alt     keycode  51 = Meta_comma
 keycode  52 = period           colon
--- console-tools-0.3.3/console-data-1999.08.29/keymaps/i386/qwerty/no-latin1.kmap.keymaps	Sun Aug 27 17:33:50 2000
+++ console-tools-0.3.3/console-data-1999.08.29/keymaps/i386/qwerty/no-latin1.kmap	Sun Aug 27 17:33:51 2000
@@ -10,48 +10,20 @@
     alt	keycode   1 = Meta_Escape
 
 	keycode   2 = one	exclam		exclamdown	onesuperior
-    alt	keycode   2 = Meta_one
-
 	keycode   3 = two	quotedbl	at		twosuperior
-control	keycode   3 = nul
-    alt	keycode   3 = Meta_two
-
 	keycode   4 = three	numbersign	sterling	threesuperior
 control	keycode   4 = Escape
     alt	keycode   4 = Meta_three
-
 	keycode   5 = four	dollar		currency	onequarter
-control	keycode   5 = Control_backslash
-    alt	keycode   5 = Meta_four
-
 	keycode   6 = five	percent		onehalf		onehalf
-control	keycode   6 = Control_bracketright
-    alt	keycode   6 = Meta_five
-
 	keycode   7 = six	ampersand	threequarters	threequarters
-control	keycode   7 = Control_asciicircum
-    alt	keycode   7 = Meta_six
-
 	keycode   8 = seven	slash		braceleft	division
-control	keycode   8 = Control_underscore
-    alt	keycode   8 = Meta_seven
-
-	keycode   9 = eight	parenleft	bracketleft
-control	keycode   9 = Delete
-    alt	keycode   9 = Meta_eight
-
-	keycode  10 = nine	parenright	bracketright
-    alt	keycode  10 = Meta_nine
-
-	keycode  11 = zero	equal		braceright
-    alt	keycode  11 = Meta_zero
-
+	keycode   9 = eight	parenleft	bracketleft     guillemotleft
+	keycode  10 = nine	parenright	bracketright    guillemotright
+	keycode  11 = zero	equal		braceright      degree
 	keycode  12 = plus	question	plusminus	questiondown
-    alt	keycode  12 = Meta_plus
 
-	keycode  13 = backslash	grave		dead_acute	dead_grave
-control	keycode  13 = Control_backslash
-    alt	keycode  13 = Meta_backslash
+	keycode  13 = backslash	grave		dead_acute	notsign
 
 	keycode  14 = Delete
 control	keycode  14 = Control_underscore	# For Emacs' UNDO :-)
@@ -72,8 +44,6 @@
     alt	keycode  19 = Meta_r
 
 	keycode  20 = +t	+T		+thorn		+THORN
-control	keycode  20 = Control_t
-    alt	keycode  20 = Meta_t
 
 	keycode  21 = +y	+Y		ydiaeresis	yen
 control	keycode  21 = Control_y
@@ -99,7 +69,7 @@
 control	keycode  26 = Control_bracketright
     alt	keycode  26 = Meta_bracketright
 
-	keycode  27 = asciitilde asciicircum	dead_diaeresis	dead_tilde
+	keycode  27 = dead_diaeresis dead_circumflex dead_tilde
 control	keycode  27 = Control_asciicircum
     alt	keycode  27 = Meta_asciicircum
 
