--- console-tools-0.3.3/fontfiletools/psfgettable.c.jj	Wed Aug 25 22:58:14 1999
+++ console-tools-0.3.3/fontfiletools/psfgettable.c	Mon Mar  6 15:15:42 2000
@@ -100,6 +100,7 @@ int main(int argc, char *argv[])
   int glyph;
   unicode unichar;
   int fontlen;
+  unsigned char buf[2];
 
   miscsetup();
   
@@ -158,8 +159,8 @@ int main(int argc, char *argv[])
       else
 	fprintf(out, "0x%03x\t", glyph);
 
-      while ( fread(&unichar, sizeof(unicode), 1, in),
-	      unichar != PSF_SEPARATOR )
+      while ( fread(buf, sizeof(buf), 1, in),
+	      (unichar = buf[0] + 256 * buf[1]) != PSF_SEPARATOR )
 	fprintf(out, " U+%04x", unichar);
 
       putc('\n', out);
