--- console-tools-0.3.3/lib/cfont/findfile-font.c.foo	Wed Jan 12 16:56:40 2000
+++ console-tools-0.3.3/lib/cfont/findfile-font.c	Wed Jan 12 16:56:49 2000
@@ -5,7 +5,7 @@
 
 static const char *fontdirpath[] = { "", DATADIR "/" FONTDIR "/", 
 #ifdef LOCALDATADIR
-  LOCALDATADIR "/" FONTDIR "/", 
+  LOCALDATADIR "/",
 #endif
   0 };
 static const char *fontsuffixes[] = { 
--- console-tools-0.3.3/lib/ctutils/findfile-lct.c.foo	Wed Jan 12 16:55:27 2000
+++ console-tools-0.3.3/lib/ctutils/findfile-lct.c	Wed Jan 12 16:56:06 2000
@@ -9,7 +9,7 @@
   DATADIR "/" TRANSDIR "/", 
   CHARMAPDIR "/",
 #ifdef LOCALDATADIR
-  LOCALDATADIR "/" TRANSDIR "/",
+  LOCALDATADIR "/",
 #endif
   0 };
 static const char *mapsuffixes[] = {
@@ -27,7 +27,7 @@
 {
   "", DATADIR "/" KEYMAPDIR "/**/", 
 #ifdef LOCALDATADIR
-  LOCALDATADIR "/" KEYMAPDIR "/**/",
+  LOCALDATADIR "/",
 #endif
   KERNDIR "/", 
   0 };
