--- console-tools-0.3.3/console-data-1999.08.29/keymaps/i386/qwerty/jp106.kmap.jp106	Wed Jul 19 19:16:23 2000
+++ console-tools-0.3.3/console-data-1999.08.29/keymaps/i386/qwerty/jp106.kmap	Wed Jul 19 19:16:33 2000
@@ -3,7 +3,7 @@
 # Date: Tue, 20 Aug 1996 16:15:31 +0900
 # From: Hideya Hane <wing@aries.bekkoame.or.jp>
 #
-keymaps 0-1,4-5,8-9,12
+keymaps 0-2,4-5,8-9,12
 alt_is_meta
 include "qwerty-layout"
 include "linux-with-two-alt-keys"
