--- dev86-0.16.0/bcc/bcc.c.orig	Wed Sep 27 23:43:51 2000
+++ dev86-0.16.0/bcc/bcc.c	Thu Jan  3 17:29:43 2002
@@ -80,7 +80,7 @@
 #else
 #define STANDARD_EXEC_PREFIX_2	"/usr/bin/"
 #endif
-#define DEFAULT_INCLUDE         "-I~/include"
+#define DEFAULT_INCLUDE         "-I~/lib/bcc/include"
 #define DEFAULT_LIBDIR0         "-L~/lib/bcc/i86/"
 #define DEFAULT_LIBDIR3         "-L~/lib/bcc/i386/"
 #define OPTIM_RULES    		"-d~/lib/bcc/i86"
