--- dev86-0.16.0/ar/ar.c.orig	Tue Mar 26 02:24:53 2002
+++ dev86-0.16.0/ar/ar.c	Thu Nov 14 12:55:58 2002
@@ -44,7 +44,7 @@ extern int sys_nerr;
 #undef  HAVE_FSYNC
 #endif
 
-extern int errno;
+#include <errno.h>
 
 #ifdef	__GNUC__
 # ifndef alloca
@@ -66,9 +66,6 @@ char *alloca ();
 
 /* Locking is normally disabled because fcntl hangs on the Sun
    and it isn't supported properly across NFS anyway.  */
-#ifdef LOCKS
-#include <errno.h>
-#endif
 
 /* This structure is used internally to represent the info
    on a member of an archive.  This is to make it easier to change format.  */
