--- dev86-0.16.0/ar/ar.c.orig	Sun Dec  3 10:20:47 2000
+++ dev86-0.16.0/ar/ar.c	Wed Mar 20 21:38:38 2002
@@ -252,7 +252,7 @@
 extern void extract_members ();
 
 extern char *basename (char *path);
-extern char *concat (char *s1, char *s2, char *s3);
+extern char *concat (const char *s1, const char *s2, const char *s3);
 extern char *make_tempname (char *name);
 extern char *xmalloc (unsigned int size);
 extern char *xrealloc (char *ptr, unsigned int size);
@@ -277,7 +277,7 @@
 extern void lock_for_update (void);
 extern void make_new_symdefs (struct mapelt *mapelt, int archive_indesc);
 extern void move_members (void);
-extern void mywrite (int desc, char *buf, int bytes, char *file);
+extern void mywrite (int desc, void *buf, int bytes, char *file);
 extern void perror_with_name (char *name);
 extern void pfatal_with_name (char *name);
 extern void print_contents (struct member_desc member, FILE *istream);
@@ -300,7 +300,7 @@
 void
 mywrite (desc, buf, bytes, file)
      int desc;
-     char *buf;
+     void *buf;
      int bytes;
      char *file;
 {
@@ -311,7 +311,7 @@
       val = write (desc, buf, bytes);
       if (val <= 0)
 	perror_with_name (file);
-      buf += val;
+      (char *) buf += val;
       bytes -= val;
     }
 }
@@ -349,7 +349,7 @@
 
     if (*p == '-')
       p++;
-   while (c = *p++)
+   while ((c = *p++))
       {
 	switch (c)
 	  {
@@ -444,11 +444,12 @@
 
   i = 2;
 
-  if (postype != POS_DEFAULT)
+  if (postype != POS_DEFAULT) {
     if (i < argc)
       posname = argv[i++];
     else
       usage ("no position operand", 0);
+  }
 
   if (i >= argc)
     usage ("no archive specified", 0);
@@ -500,7 +501,7 @@
 	break;
 
     default:
-	usage ("invalid operation %d", operation);
+	usage ("invalid operation %d", (char *)operation);
     }
 
   exit (0);
@@ -674,10 +675,10 @@
   if (preserve_dates)
     {
 #if defined(USG) || defined(linux) || defined(__BCC__)
-      long tv[2];
-      tv[0] = member.date;
-      tv[1] = member.date;
-      utime (member.name, tv);
+      struct utimbuf tv;
+      tv.actime = member.date;
+      tv.modtime = member.date;
+      utime (member.name, &tv);
 #else
       struct timeval tv[2];
       tv[0].tv_sec = member.date;
@@ -2053,7 +2054,7 @@
 
 char *
 concat (s1, s2, s3)
-     char *s1, *s2, *s3;
+     const char *s1, *s2, *s3;
 {
   int len1 = strlen (s1), len2 = strlen (s2), len3 = strlen (s3);
   char *result = (char *) xmalloc (len1 + len2 + len3 + 1);
--- dev86-0.16.0/as/typeconv.c.orig	Sat Jun  7 10:38:46 1997
+++ dev86-0.16.0/as/typeconv.c	Wed Mar 20 16:19:23 2002
@@ -15,8 +15,8 @@
 
 static int no_swap   = 1;
 
-static long_off[4] = {0,1,2,3};
-static int_off[2] = {0,1};
+static int long_off[4] = {0,1,2,3};
+static int int_off[2] = {0,1};
 
 PUBLIC bool_pt typeconv_init(big_endian, long_big_endian)
 bool_pt big_endian;
--- dev86-0.16.0/bcc/bcc.c.orig	Wed Sep 27 23:43:51 2000
+++ dev86-0.16.0/bcc/bcc.c	Wed Mar 20 16:19:23 2002
@@ -2,7 +2,8 @@
 
 /* Copyright (C) 1992 Bruce Evans */
 
-#define _POSIX_SOURCE 1
+#define _XOPEN_SOURCE
+#define _XOPEN_SOURCE_EXTENDED
 
 #include <sys/types.h>
 #include <sys/stat.h>
--- dev86-0.16.0/copt/copt.c.orig	Fri Jul 23 10:47:43 1999
+++ dev86-0.16.0/copt/copt.c	Wed Mar 20 16:19:23 2002
@@ -352,7 +352,7 @@
  */
 static long eval(char *str, int len)
 {
-  char *oldcp, *cp, c;
+  char *cp, c;
   int state = 0;
   int i, varnum;
 
@@ -811,7 +811,7 @@
 /*
  * Main program
  */
-void main(int argc, char **argv)
+int main(int argc, char **argv)
 {
   char comment = NOCHAR;
   char *srcfile = NULL;
--- dev86-0.16.0/ld/typeconv.c.orig	Sat Jun  7 10:38:46 1997
+++ dev86-0.16.0/ld/typeconv.c	Wed Mar 20 16:19:23 2002
@@ -15,8 +15,8 @@
 
 static int no_swap   = 1;
 
-static long_off[4] = {0,1,2,3};
-static int_off[2] = {0,1};
+static int long_off[4] = {0,1,2,3};
+static int int_off[2] = {0,1};
 
 PUBLIC bool_pt typeconv_init(big_endian, long_big_endian)
 bool_pt big_endian;
