--- dev86-0.16.0/libc/kinclude/Makefile.orig	Sat Aug 16 11:26:04 1997
+++ dev86-0.16.0/libc/kinclude/Makefile	Thu Jan  3 17:32:10 2002
@@ -6,11 +6,11 @@
 	@:
 
 transfer:
-	-@rm -f ../include/linuxmt ../include/arch
-	ln -s ../kinclude/linuxmt ../include
-	ln -s ../kinclude/arch ../include
+	-@rm -rf ../include/linuxmt ../include/arch
+	cp -pr linuxmt ../include
+	cp -pr arch ../include
 	@touch Used
 
 clean:
-	-@rm -f ../include/linuxmt ../include/arch
+	-@rm -rf ../include/linuxmt ../include/arch
 	-@rm -f Used
--- dev86-0.16.0/makefile.in.orig	Sat Jan  6 16:20:47 2001
+++ dev86-0.16.0/makefile.in	Thu Jan  3 17:32:10 2002
@@ -123,9 +123,9 @@
 	@rm -f include
 	@ln -s libc/include include 2>/dev/null || true
 #ifndef GNUMAKE
-	@rm -f include/linuxmt include/arch || true
-	@ln -s ../kinclude/linuxmt include/linuxmt 2>/dev/null || true
-	@ln -s ../kinclude/arch    include/arch    2>/dev/null || true
+	@rm -rf include/linuxmt include/arch || true
+	@ cp -pr kinclude/linuxmt include/linuxmt 2>/dev/null || true
+	@ cp -pr kinclude/arch    include/arch    2>/dev/null || true
 #endif
 
 bcc: bindir
