--- dev86-0.16.17/libc/kinclude/Makefile.orig	2003-01-29 19:40:25 +0000
+++ dev86-0.16.17/libc/kinclude/Makefile	2005-06-30 11:26:36 +0000
@@ -6,9 +6,9 @@
 	@:
 
 transfer:
-	-@rm -f ../include/linuxmt ../include/arch
-	ln -s ../kinclude/linuxmt ../include
-	ln -s ../kinclude/arch ../include
+	-@rm -rf ../include/linuxmt ../include/arch
+	cp -pr ../kinclude/linuxmt ../include
+	cp -pr ../kinclude/arch ../include
 
 clean:
-	-rm -f ../include/linuxmt ../include/arch
+	-rm -rf ../include/linuxmt ../include/arch
--- dev86-0.16.17/makefile.in.orig	2004-10-02 13:00:53 +0000
+++ dev86-0.16.17/makefile.in	2005-06-30 11:24:47 +0000
@@ -144,9 +144,9 @@
 	@rm -f include
 	@ln -s libc/include include 2>/dev/null || true
 #ifndef GNUMAKE
-	@rm -f include/linuxmt include/arch || true
-	@ln -s ../kinclude/linuxmt include/linuxmt 2>/dev/null || true
-	@ln -s ../kinclude/arch    include/arch    2>/dev/null || true
+	@rm -rf include/linuxmt include/arch || true
+	@ cp -pr kinclude/linuxmt include/linuxmt 2>/dev/null || true
+	@ cp -pr kinclude/arch    include/arch    2>/dev/null || true
 #endif
 
 bcc86: bindir
