diff -upk.orig Linux-PAM-0.80.orig/modules/pam_filter/Makefile Linux-PAM-0.80/modules/pam_filter/Makefile
--- Linux-PAM-0.80.orig/modules/pam_filter/Makefile	2001-11-11 07:43:54 +0000
+++ Linux-PAM-0.80/modules/pam_filter/Makefile	2005-08-03 15:28:55 +0000
@@ -28,11 +28,14 @@ LIBOBJ = $(TITLE).o
 LIBOBJD = $(addprefix dynamic/,$(LIBOBJ))
 LIBOBJS = $(addprefix static/,$(LIBOBJ))
 
+LINK_PAMMODUTILS = -L../pammodutil -lpammodutil -L../../libpam -lpam
+INCLUDE_PAMMODUTILS = -I../pammodutil/include
+
 dynamic/%.o : %.c
-	$(CC) $(CFLAGS) $(DYNAMIC) $(CPPFLAGS) $(TARGET_ARCH) -c $< -o $@
+	$(CC) $(CFLAGS) $(INCLUDE_PAMMODUTILS) $(DYNAMIC) $(CPPFLAGS) $(TARGET_ARCH) -c $< -o $@
 
 static/%.o : %.c
-	$(CC) $(CFLAGS) $(STATIC) $(CPPFLAGS) $(TARGET_ARCH) -c $< -o $@
+	$(CC) $(CFLAGS) $(INCLUDE_PAMMODUTILS) $(STATIC) $(CPPFLAGS) $(TARGET_ARCH) -c $< -o $@
 
 ifdef DYNAMIC
 LIBSHARED = $(TITLE).so
@@ -78,7 +81,7 @@ endif
 
 ifdef DYNAMIC
 $(LIBSHARED):	$(LIBOBJD)
-		$(LD_D) -o $@ $(LIBOBJD)
+	$(LD_D) -o $@ $(LIBOBJD) $(LINK_PAMMODUTILS)
 endif
 
 ifdef STATIC
