diff -upk.orig Linux-PAM-0.80.orig/configure.in Linux-PAM-0.80/configure.in
--- Linux-PAM-0.80.orig/configure.in	2005-08-03 15:27:50 +0000
+++ Linux-PAM-0.80/configure.in	2005-08-03 15:28:07 +0000
@@ -58,6 +58,8 @@ dnl
 dnl options and defaults
 dnl
 
+AC_SYS_LARGEFILE
+
 dnl lots of debugging information goes to /var/run/pam-debug.log
 AC_ARG_ENABLE(debug,
 [  --enable-debug           specify you are building with debugging on],
