diff -upk.orig Linux-PAM-0.80.orig/configure.in Linux-PAM-0.80/configure.in
--- Linux-PAM-0.80.orig/configure.in	2005-08-03 15:31:38 +0000
+++ Linux-PAM-0.80/configure.in	2005-08-03 15:31:55 +0000
@@ -254,11 +254,6 @@ fi
 if test $HAVE_LIBNSL = yes ; then
 	pwdblibs="$pwdblibs -lnsl"
 fi
-AC_CHECK_LIB(pwdb, pwdb_db_name, HAVE_LIBPWDB=yes ;
-	AC_DEFINE([HAVE_LIBPWDB],,[Define to 1 if you have the `pwdb' library (-lpwdb).]),
-	HAVE_LIBPWDB=no,$pwdblibs)
-AC_SUBST(HAVE_LIBPWDB)
-unset pwdblibs
 
 AC_CHECK_LIB(fl, yywrap, HAVE_LIBFLEX=yes ;
 	AC_DEFINE([HAVE_LIBFLEX],,[Define to 1 if you have the `fl' library (-lfl).]),
