diff -upk.orig Linux-PAM-0.80.orig/configure.in Linux-PAM-0.80/configure.in
--- Linux-PAM-0.80.orig/configure.in	2005-08-03 15:29:36 +0000
+++ Linux-PAM-0.80/configure.in	2005-08-03 15:29:51 +0000
@@ -213,15 +213,6 @@ dnl Checks for the existence of libdl - 
 AC_CHECK_LIB(dl, dlopen, LIBDL=-ldl)
 AC_SUBST(LIBDL)
 
-dnl
-dnl At least on Solaris, the existing libcrack must be dynamic.
-dnl Ought to introduce a check for this.
-dnl
-AC_CHECK_LIB(crack, FascistCheck, HAVE_LIBCRACK=yes ;
-	AC_DEFINE([HAVE_LIBCRACK],,[Define to 1 if you have the `crack' library (-lcrack).]),
-	HAVE_LIBCRACK=no)
-AC_SUBST(HAVE_LIBCRACK)
-
 AC_CHECK_LIB(crypt, crypt, HAVE_LIBCRYPT=yes ;
 	AC_DEFINE([HAVE_LIBCRYPT],,[Define to 1 if you have the `crypt' library (-lcrypt).]),
 	HAVE_LIBCRYPT=no)
@@ -284,9 +275,6 @@ AC_HEADER_STDC
 AC_HEADER_SYS_WAIT
 AC_CHECK_HEADERS(fcntl.h limits.h malloc.h sys/file.h sys/ioctl.h sys/time.h syslog.h termio.h unistd.h sys/fsuid.h inittypes.h)
 
-dnl For module/pam_cracklib
-AC_CHECK_HEADERS(crypt.h)
-
 dnl For module/pam_userdb
 AC_CHECK_HEADERS(ndbm.h db.h)
 dnl I suspect the following two lines are a hack.
@@ -299,30 +287,6 @@ AC_CHECK_HEADERS(lastlog.h utmp.h utmpx.
 dnl This following rule should be made conditional upon HAVE_LIBCRYPT
 dnl being found.
 
-dnl Look for cracklib dictionary
-AC_MSG_CHECKING(path to cracklib dictionary)
-DICT_DIR_CANDIDATES="/usr/lib /usr/share/dict /usr/share/lib \
-  /usr/local/lib /usr/local/share/lib /usr/share/cracklib"
-DICT_FILE_CANDIDATES="pw_dict cracklib_dict"
-CRACKLIB_DICTPATH=""
-for d in $DICT_DIR_CANDIDATES ; do
-      for f in $DICT_FILE_CANDIDATES ; do
-              if test -r $d/$f.hwm ; then
-                      CRACKLIB_DICTPATH=$d/$f
-                      break 2
-              elif test -r $d/dict/$f.hwm ; then
-                      CRACKLIB_DICTPATH=$d/dict/$f
-                      break 2
-              fi
-      done
-done
-if test -z "$CRACKLIB_DICTPATH" ; then
-      AC_MSG_WARN([none found - pam_cracklib will not be built])
-else
-      AC_MSG_RESULT($CRACKLIB_DICTPATH)
-fi
-AC_SUBST(CRACKLIB_DICTPATH)
-
 dnl Set FLAGS, linker options etc. depending on C compiler.
 dnl gcc is tested and much preferred; others less so, if at all
 dnl
