Index: AUTHORS
===================================================================
RCS file: /cvsroot/pam/Linux-PAM/AUTHORS,v
retrieving revision 1.2
retrieving revision 1.3
diff -u -p -a -r1.2 -r1.3
--- AUTHORS	26 Sep 2005 19:35:39 -0000	1.2
+++ AUTHORS	16 Dec 2005 17:52:36 -0000	1.3
@@ -1,7 +1,7 @@
 Original authors and current maintainers of Linux-PAM:
 
-Andrew G. Morgan	<morgan@linux.kernel.org>
+Andrew G. Morgan	<morgan@kernel.org>
 Dmitry V. Levin		<ldv@altlinux.org>
-Thorsten Kukuk		<kukuk@suse.de>
+Thorsten Kukuk		<kukuk@thkukuk.de>
 Sebastien Tricaud	<toady@gscore.org>
 Tomas Mraz		<t8m@centrum.cz>
Index: ChangeLog
===================================================================
RCS file: /cvsroot/pam/Linux-PAM/ChangeLog,v
retrieving revision 1.44
retrieving revision 1.51
diff -u -p -a -r1.44 -r1.51
--- ChangeLog	12 Dec 2005 17:34:10 -0000	1.44
+++ ChangeLog	21 Dec 2005 10:04:10 -0000	1.51
@@ -1,3 +1,58 @@
+2005-12-21  Tomas Mraz <t8m@centrum.cz>
+
+	* modules/pam_succeed_if/pam_succeed_if.c (evaluate_ingroup),
+	(evaluate_notingroup): Simplified.
+	(evaluate_innetgr), (evaluate_notinnetgr): New functions.
+	(evaluate): Added calls to evaluate_(not)innetgr().
+	* modules/pam_succeed_if/README: Documented netgroup matching.
+	* NEWS: Mentioned the added netgroup matching support.
+
+2005-12-20  Thorsten Kukuk  <kukuk@thkukuk.de>
+
+	* modules/pam_lastlog/pam_lastlog.c (last_login_read): Use
+	strftime instead of ctime.
+
+	* po/de.po: Fix typo.
+
+2005-12-19  Thorsten Kukuk  <kukuk@thkukuk.de>
+
+	* libpam/pam_syslog.c: Define LOG_AUTHPRIV as LOG_AUTH on Solaris.
+	Reported by Charles_H_Bedford@nbc.gov.
+
+	* modules/pam_time/pam_time.c (check_account): Implement
+	support for netgroups.
+
+	* modules/pam_time/time.conf: Document usage of netgroups.
+
+2005-12-16  Thorsten Kukuk  <kukuk@thkukuk.de>
+
+	* modules/pam_group/pam_group.c (check_account): Implement
+	support for netgroups.
+
+	* modules/pam_group/group.conf: Add all documentation to this
+	example config file and don't reference to outdated configs.
+
+	* modules/pam_group/README: New.
+
+	* modules/pam_group/Makefile.am: Add README to EXTRADIST.
+
+2005-12-15  Thorsten Kukuk  <kukuk@suse.de>
+
+	* modules/pam_lastlog/pam_lastlog.c (last_login_read): Don't report an
+	error if user logins the first time.
+
+	* modules/pam_lastlog/README: New.
+
+	* modules/pam_lastlog/Makefile.am: Add README to EXTRADIST.
+
+2005-12-14  Thorsten Kukuk  <kukuk@suse.de>
+
+	* modules/pam_deny/pam_deny.c: Fix comment.
+
+	* doc/pam_appl.sgml: Fix typo.
+
+	Reported by Russell Bateman <russ@windofkeltia.com>
+
 2005-12-12  Thorsten Kukuk  <kukuk@thkukuk.de>
 
 	* release version 0.99.2.1
Index: NEWS
===================================================================
RCS file: /cvsroot/pam/Linux-PAM/NEWS,v
retrieving revision 1.9
retrieving revision 1.10
diff -u -p -a -r1.9 -r1.10
--- NEWS	12 Dec 2005 17:34:10 -0000	1.9
+++ NEWS	21 Dec 2005 10:04:09 -0000	1.10
@@ -1,5 +1,6 @@
 Linux-PAM NEWS -- history of user-visible changes.
 
+* pam_succeed_if, pam_group, pam_time: Support netgroup matching.
 
 Release 0.99.2.1
 
Index: doc/pam_appl.sgml
===================================================================
RCS file: /cvsroot/pam/Linux-PAM/doc/pam_appl.sgml,v
retrieving revision 1.10
retrieving revision 1.11
diff -u -p -a -r1.10 -r1.11
--- doc/pam_appl.sgml	22 Sep 2004 09:37:47 -0000	1.10
+++ doc/pam_appl.sgml	14 Dec 2005 09:53:06 -0000	1.11
@@ -2,7 +2,7 @@
 
 <!--
 
- $Id: pam_appl.sgml,v 1.10 2004/09/22 09:37:47 kukuk Exp $
+ $Id: pam_appl.sgml,v 1.11 2005/12/14 09:53:06 kukuk Exp $
 
     Copyright (C) Andrew G. Morgan 1996-2001.  All rights reserved.
 
@@ -223,7 +223,7 @@ is, is not well defined.  However, some 
 glossary below.
 
 <sect>The public interface to <bf>Linux-PAM</bf>
- 
+
 <p>
 Firstly, the relevant include file for the <bf>Linux-PAM</bf> library
 is <tt>&lt;security/pam_appl.h&gt;</tt>. It contains the definitions
@@ -321,7 +321,7 @@ extern int pam_set_item(pam_handle_t *pa
 
 <tag><tt/PAM_USER/</tag>
 
-	The username of the entity under who's identity service will
+	The username of the entity under whose identity service will
 	be given. That is, following authentication, <tt/PAM_USER/
 	identifies the local entity that gets to use the
 	service. Note, this value can be mapped from something (eg.,
@@ -723,14 +723,14 @@ error-returns are:
 <descrip>
 <tag><tt/PAM_AUTHTOK_ERR/</tag>
 	A module was unable to obtain the new authentication token.
-	
+
 <tag><tt/PAM_AUTHTOK_RECOVERY_ERR/</tag>
 	A module was unable to obtain the old authentication token.
 
 <tag><tt/PAM_AUTHTOK_LOCK_BUSY/</tag>
 	One or more of the modules was unable to change the
 authentication token since it is currently locked.
-	
+
 <tag><tt/PAM_AUTHTOK_DISABLE_AGING/</tag>
 	Authentication token aging has been disabled for at least one
 of the modules.
@@ -1559,7 +1559,7 @@ int main(int argc, char *argv[])
     }
 
     retval = pam_start("check_user", user, &ero;conv, &ero;pamh);
-	
+
     if (retval == PAM_SUCCESS)
         retval = pam_authenticate(pamh, 0);    /* is user really user? */
 
@@ -1659,7 +1659,7 @@ This document was written by Andrew G. M
 <!-- insert credits here -->
 <!--
  an sgml list of people to credit for their contributions to Linux-PAM
- $Id: pam_appl.sgml,v 1.10 2004/09/22 09:37:47 kukuk Exp $
+ $Id: pam_appl.sgml,v 1.11 2005/12/14 09:53:06 kukuk Exp $
   -->
 Chris Adams,
 Peter Allgeyer,
@@ -1772,6 +1772,6 @@ USE OF THIS SOFTWARE, EVEN IF ADVISED OF
 DAMAGE.
 
 <p>
-<tt>$Id: pam_appl.sgml,v 1.10 2004/09/22 09:37:47 kukuk Exp $</tt>
+<tt>$Id: pam_appl.sgml,v 1.11 2005/12/14 09:53:06 kukuk Exp $</tt>
 
 </article>
Index: libpam/pam_syslog.c
===================================================================
RCS file: /cvsroot/pam/Linux-PAM/libpam/pam_syslog.c,v
retrieving revision 1.4
retrieving revision 1.5
diff -u -p -a -r1.4 -r1.5
--- libpam/pam_syslog.c	14 Sep 2005 17:49:01 -0000	1.4
+++ libpam/pam_syslog.c	19 Dec 2005 07:41:52 -0000	1.5
@@ -45,6 +45,10 @@
 
 #include "pam_private.h"
 
+#ifndef LOG_AUTHPRIV
+#define LOG_AUTHPRIV LOG_AUTH
+#endif
+
 static const char *
 _pam_choice2str (int choice)
 {
Index: modules/pam_deny/pam_deny.c
===================================================================
RCS file: /cvsroot/pam/Linux-PAM/modules/pam_deny/pam_deny.c,v
retrieving revision 1.3
retrieving revision 1.4
diff -u -p -a -r1.3 -r1.4
--- modules/pam_deny/pam_deny.c	17 Sep 2005 09:02:06 -0000	1.3
+++ modules/pam_deny/pam_deny.c	14 Dec 2005 09:58:32 -0000	1.4
@@ -1,7 +1,7 @@
-/* pam_permit module */
+/* pam_deny module */
 
 /*
- * $Id: pam_deny.c,v 1.3 2005/09/17 09:02:06 t8m Exp $
+ * $Id: pam_deny.c,v 1.4 2005/12/14 09:58:32 kukuk Exp $
  *
  * Written by Andrew Morgan <morgan@parc.power.net> 1996/3/11
  *
Index: modules/pam_group/Makefile.am
===================================================================
RCS file: /cvsroot/pam/Linux-PAM/modules/pam_group/Makefile.am,v
retrieving revision 1.5
retrieving revision 1.6
diff -u -p -a -r1.5 -r1.6
--- modules/pam_group/Makefile.am	21 Sep 2005 13:27:11 -0000	1.5
+++ modules/pam_group/Makefile.am	16 Dec 2005 17:52:36 -0000	1.6
@@ -4,7 +4,7 @@
 
 CLEANFILES = *~
 
-EXTRA_DIST = group.conf
+EXTRA_DIST = README group.conf
 
 securelibdir = $(SECUREDIR)
 secureconfdir = $(SCONFIGDIR)
Index: modules/pam_group/README
===================================================================
RCS file: modules/pam_group/README
diff -N modules/pam_group/README
--- /dev/null	1 Jan 1970 00:00:00 -0000
+++ modules/pam_group/README	16 Dec 2005 17:52:36 -0000	1.1
@@ -0,0 +1,23 @@
+
+This is a help file for the pam_group module. It explains the need for
+pam_group and also the syntax of the /etc/security/group.conf file.
+
+1. Introduction
+===============
+
+It is desirable to give extra privileges to a user running a specific
+PAM aware application at various times of the day and on specific days
+or over various terminal lines by adding this user to extra groups.
+
+The pam_group module is intended to offer a configurable module that
+satisfies this purpose, within the context of Linux-PAM.
+
+2. the /etc/security/group.conf file
+===================================
+
+Its syntax is described in the sample group.conf file.
+
+unrecognised rules are ignored (but an error is logged to syslog(3))
+
+--------------------
+Bugs to the list <pam-list@redhat.com>
Index: modules/pam_group/group.conf
===================================================================
RCS file: /cvsroot/pam/Linux-PAM/modules/pam_group/group.conf,v
retrieving revision 1.2
retrieving revision 1.3
diff -u -p -a -r1.2 -r1.3
--- modules/pam_group/group.conf	26 Nov 2000 07:32:39 -0000	1.2
+++ modules/pam_group/group.conf	16 Dec 2005 17:52:36 -0000	1.3
@@ -1,10 +1,7 @@
-##
-## Note, to get this to work as it is currently typed you need
-##
-## 1. to run an application as root
-## 2. add the following groups to the /etc/group file:
-##		floppy, games, sound
-##
+#
+# This is the configuration file for the pam_group module. 
+#
+
 #
 # *** Please note that giving group membership on a session basis is
 # *** NOT inherently secure. If a user can create an executable that
@@ -16,11 +13,9 @@
 # *** "chgrp games toplay; chmod g+s toplay". They are basically able
 # *** to play games any time... You have been warned. AGM
 #
-# this is an example configuration file for the pam_group module. Its
-# syntax is based on that of the pam_time module and (at some point in
-# the distant past was inspired by the 'shadow' package)
+
 #
-# the syntax of the lines is as follows:
+# The syntax of the lines is as follows:
 #
 #       services;ttys;users;times;groups
 #
@@ -28,15 +23,59 @@
 # newlines). From reading these comments, it is clear that
 # text following a '#' is ignored to the end of the line.
 #
-# the first four fields are described in the pam_time directory.
-# The only difference for these is how the time field is interpretted:
-# it is used to indicate "when" these groups are to be given to the user.
+# the combination of individual users/terminals etc is a logic list
+# namely individual tokens that are optionally prefixed with '!' (logical
+# not) and separated with '&' (logical and) and '|' (logical or).
+#
+# services
+#       is a logic list of PAM service names that the rule applies to.
+#
+# ttys
+#       is a logic list of terminal names that this rule applies to.
+#
+# users
+#       is a logic list of users or a netgroup of users to whom this
+#       rule applies.
+#
+# NB. For these items the simple wildcard '*' may be used only once.
+#     With netgroups no wildcards or logic operators are allowed.
+#
+# times
+#       It is used to indicate "when" these groups are to be given to the
+#       user. The format here is a logic list of day/time-range
+#       entries the days are specified by a sequence of two character
+#       entries, MoTuSa for example is Monday Tuesday and Saturday. Note
+#       that repeated days are unset MoMo = no day, and MoWk = all weekdays
+#       bar Monday. The two character combinations accepted are
+#
+#               Mo Tu We Th Fr Sa Su Wk Wd Al
+#
+#       the last two being week-end days and all 7 days of the week
+#       respectively. As a final example, AlFr means all days except Friday.
+#
+#       Each day/time-range can be prefixed with a '!' to indicate "anything
+#       but"
+#
+#       The time-range part is two 24-hour times HHMM separated by a hyphen
+#       indicating the start and finish time (if the finish time is smaller
+#       than the start time it is deemed to apply on the following day).
 #
 # groups
-#	The (comma or space separated) list of groups that the user 
+#	The (comma or space separated) list of groups that the user
 #	inherits membership of. These groups are added if the previous
 #	fields are satisfied by the user's request
 #
+# For a rule to be active, ALL of service+ttys+users must be satisfied
+# by the applying process.
+#
+
+#
+# Note, to get this to work as it is currently typed you need
+#
+# 1. to run an application as root
+# 2. add the following groups to the /etc/group file:
+#		floppy, games, sound
+#
 
 #
 # Here is a simple example: running 'xsh' on tty* (any ttyXXX device),
Index: modules/pam_group/pam_group.c
===================================================================
RCS file: /cvsroot/pam/Linux-PAM/modules/pam_group/pam_group.c,v
retrieving revision 1.17
retrieving revision 1.18
diff -u -p -a -r1.17 -r1.18
--- modules/pam_group/pam_group.c	12 Dec 2005 14:45:01 -0000	1.17
+++ modules/pam_group/pam_group.c	16 Dec 2005 17:52:36 -0000	1.18
@@ -1,16 +1,11 @@
 /* pam_group module */
 
 /*
- * $Id: pam_group.c,v 1.17 2005/12/12 14:45:01 ldv Exp $
+ * $Id: pam_group.c,v 1.18 2005/12/16 17:52:36 kukuk Exp $
  *
  * Written by Andrew Morgan <morgan@linux.kernel.org> 1996/7/6
  */
 
-static const char rcsid[] =
-"$Id: pam_group.c,v 1.17 2005/12/12 14:45:01 ldv Exp $;\n"
-"Version 0.5 for Linux-PAM\n"
-"Copyright (c) Andrew G. Morgan 1996 <morgan@linux.kernel.org>\n";
-
 #include "config.h"
 
 #include <sys/file.h>
@@ -27,6 +22,7 @@ static const char rcsid[] =
 #include <sys/types.h>
 #include <sys/stat.h>
 #include <fcntl.h>
+#include <netdb.h>
 
 #define PAM_GROUP_BUFLEN        1000
 #define FIELD_SEPARATOR         ';'   /* this is new as of .02 */
@@ -686,7 +682,11 @@ static int check_account(pam_handle_t *p
 		       "%s: no user entry #%d", PAM_GROUP_CONF, count);
 	    continue;
 	}
-	good &= logic_field(pamh,user, buffer, count, is_same);
+	/* If buffer starts with @, we are using netgroups */
+	if (buffer[0] == '@')
+	  good &= innetgr (&buffer[1], NULL, user, NULL);
+	else
+	  good &= logic_field(pamh,user, buffer, count, is_same);
 	D(("with user: %s", good ? "passes":"fails" ));
 
 	/* here we get the time field */
Index: modules/pam_lastlog/Makefile.am
===================================================================
RCS file: /cvsroot/pam/Linux-PAM/modules/pam_lastlog/Makefile.am,v
retrieving revision 1.6
retrieving revision 1.7
diff -u -p -a -r1.6 -r1.7
--- modules/pam_lastlog/Makefile.am	21 Sep 2005 13:27:11 -0000	1.6
+++ modules/pam_lastlog/Makefile.am	16 Dec 2005 07:57:10 -0000	1.7
@@ -7,6 +7,8 @@ CLEANFILES = *~
 securelibdir = $(SECUREDIR)
 secureconfdir = $(SCONFIGDIR)
 
+EXTRADIST = README
+
 AM_CFLAGS = -I$(top_srcdir)/libpam/include -I$(top_srcdir)/libpamc/include
 AM_LDFLAGS = -no-undefined -avoid-version -module \
 	-L$(top_builddir)/libpam -lpam -lutil
Index: modules/pam_lastlog/README
===================================================================
RCS file: modules/pam_lastlog/README
diff -N modules/pam_lastlog/README
--- /dev/null	1 Jan 1970 00:00:00 -0000
+++ modules/pam_lastlog/README	16 Dec 2005 07:57:10 -0000	1.1
@@ -0,0 +1,30 @@
+pam_lastlogin: The last login module
+
+This module can be used as session entry to provide a `Last login on ...'
+message. When the user logs into the system from what ever application
+uses the PAM libraries.  In addition, the module maintains the
+/var/log/lastlog file and the wtmp entry.
+
+
+The behavior of this module can be modified with one of the following
+flags:
+
+ * debug - write more information to syslog(3).
+
+ * nodate - neglect to give the date of the last login when displaying
+            information about the last login on the system.
+
+ * noterm - neglect to diplay the terminal name on which the last
+            login was attempt.
+
+ * nohost - neglect to indicate from which host the last login was
+            attempted.
+
+ * silent - neglect to inform the user about any previous login:
+            just update the /var/log/lastlog file.
+
+ * never - if the /var/log/lastlog file does not contain any old
+           entries for the user, indicate that the user has never
+           previously logged in with a ``welcome..." message.
+
+ * nowtmp - Don't update the wtmp entry.
Index: modules/pam_lastlog/pam_lastlog.c
===================================================================
RCS file: /cvsroot/pam/Linux-PAM/modules/pam_lastlog/pam_lastlog.c,v
retrieving revision 1.14
retrieving revision 1.16
diff -u -p -a -r1.14 -r1.16
--- modules/pam_lastlog/pam_lastlog.c	21 Sep 2005 10:54:50 -0000	1.14
+++ modules/pam_lastlog/pam_lastlog.c	20 Dec 2005 21:28:07 -0000	1.16
@@ -1,8 +1,6 @@
 /* pam_lastlog module */
 
 /*
- * $Id: pam_lastlog.c,v 1.14 2005/09/21 10:54:50 t8m Exp $
- *
  * Written by Andrew Morgan <morgan@linux.kernel.org> 1996/3/11
  *
  * This module does the necessary work to display the last login
@@ -140,61 +138,54 @@ last_login_read(pam_handle_t *pamh, int 
 {
     struct flock last_lock;
     struct lastlog last_login;
-    int retval = PAM_SERVICE_ERR;
+    int retval = PAM_SUCCESS;
+    char the_time[256];
     char *date = NULL;
     char *host = NULL;
     char *line = NULL;
 
-    do {
-	memset(&last_lock, 0, sizeof(last_lock));
-	last_lock.l_type = F_RDLCK;
-	last_lock.l_whence = SEEK_SET;
-	last_lock.l_start = sizeof(last_login) * (off_t) uid;
-	last_lock.l_len = sizeof(last_login);
-
-	if (fcntl(last_fd, F_SETLK, &last_lock) < 0) {
-	    D(("locking %s failed..(waiting a little)", _PATH_LASTLOG));
-	    pam_syslog(pamh, LOG_WARNING,
-		       "file %s is locked/read", _PATH_LASTLOG);
-	    sleep(LASTLOG_IGNORE_LOCK_TIME);
-	}
+    memset(&last_lock, 0, sizeof(last_lock));
+    last_lock.l_type = F_RDLCK;
+    last_lock.l_whence = SEEK_SET;
+    last_lock.l_start = sizeof(last_login) * (off_t) uid;
+    last_lock.l_len = sizeof(last_login);
 
-	if (pam_modutil_read(last_fd, (char *) &last_login,
-			     sizeof(last_login)) != sizeof(last_login)) {
-	    memset(&last_login, 0, sizeof(last_login));
-	}
+    if (fcntl(last_fd, F_SETLK, &last_lock) < 0) {
+        D(("locking %s failed..(waiting a little)", _PATH_LASTLOG));
+	pam_syslog(pamh, LOG_WARNING,
+		   "file %s is locked/read", _PATH_LASTLOG);
+	sleep(LASTLOG_IGNORE_LOCK_TIME);
+    }
+
+    if (pam_modutil_read(last_fd, (char *) &last_login,
+			 sizeof(last_login)) != sizeof(last_login)) {
+        memset(&last_login, 0, sizeof(last_login));
+    }
 
-	last_lock.l_type = F_UNLCK;
-	(void) fcntl(last_fd, F_SETLK, &last_lock);        /* unlock */
+    last_lock.l_type = F_UNLCK;
+    (void) fcntl(last_fd, F_SETLK, &last_lock);        /* unlock */
 
-	if (!last_login.ll_time) {
-	    D(("First login for user uid=%d", uid));
-	    if (announce & LASTLOG_DEBUG) {
-		pam_syslog(pamh, LOG_DEBUG, "creating lastlog for uid %d", uid);
-	    }
+    if (!last_login.ll_time) {
+        if (announce & LASTLOG_DEBUG) {
+	    pam_syslog(pamh, LOG_DEBUG, "first login for user with uid %d", uid);
 	}
+    }
 
-	if ((announce & LASTLOG_QUIET)) {
-	    retval = PAM_SUCCESS;
-	    break;
-	}
+    if (!(announce & LASTLOG_QUIET)) {
 
 	if (last_login.ll_time) {
-	    retval = PAM_BUF_ERR;
 
 	    /* we want the date? */
 	    if (announce & LASTLOG_DATE) {
+	        struct tm *tm, tm_buf;
 		time_t ll_time;
-		char *the_time;
 
 		ll_time = last_login.ll_time;
-		the_time = ctime(&ll_time);
-		the_time[strlen(the_time)-1] = '\0';    /* strip trailing '\n' */
+		tm = localtime_r (&ll_time, &tm_buf);
+		strftime (the_time, sizeof (the_time),
+			  " %a %b %e %H:%M:%S %Z %Y", tm);
 
-		if (asprintf(&date, " %s", the_time) < 0) {
-		    pam_syslog(pamh, LOG_ERR, "out of memory");
-		    break;
-		}
+		date = the_time;
 	    }
 
 	    /* we want & have the host? */
@@ -204,18 +195,20 @@ last_login_read(pam_handle_t *pamh, int 
 		if (asprintf(&host, _(" from %.*s"), UT_HOSTSIZE,
 			     last_login.ll_host) < 0) {
 		    pam_syslog(pamh, LOG_ERR, "out of memory");
-		    break;
+		    retval = PAM_BUF_ERR;
+		    goto cleanup;
 		}
 	    }
 
 	    /* we want and have the terminal? */
 	    if ((announce & LASTLOG_LINE)
 		&& (last_login.ll_line[0] != '\0')) {
-		/* TRANSLATORS: " on <terminal>" */ 
+		/* TRANSLATORS: " on <terminal>" */
 		if (asprintf(&line, _(" on %.*s"), UT_LINESIZE,
 			     last_login.ll_line) < 0) {
 		    pam_syslog(pamh, LOG_ERR, "out of memory");
-		    break;
+		    retval = PAM_BUF_ERR;
+		    goto cleanup;
 		}
 	    }
 
@@ -228,12 +221,12 @@ last_login_read(pam_handle_t *pamh, int 
 		D(("this is the first time this user has logged in"));
 		retval = pam_info(pamh, "%s", _("Welcome to your new account!"));
 	}
-    } while (0);
+    }
 
     /* cleanup */
+ cleanup:
     memset(&last_login, 0, sizeof(last_login));
     _pam_overwrite(date);
-    _pam_drop(date);
     _pam_overwrite(host);
     _pam_drop(host);
     _pam_overwrite(line);
@@ -325,7 +318,7 @@ last_login_write(pam_handle_t *pamh, int
 static int
 last_login_date(pam_handle_t *pamh, int announce, uid_t uid, const char *user)
 {
-    int rc_read, rc_write;
+    int retval;
     int last_fd;
 
     /* obtain the last login date and all the relevant info */
@@ -342,16 +335,20 @@ last_login_date(pam_handle_t *pamh, int 
 	return PAM_SERVICE_ERR;
     }
 
-    rc_read = last_login_read(pamh, announce, last_fd, uid);
-    rc_write = last_login_write(pamh, announce, last_fd, uid, user);
+    retval = last_login_read(pamh, announce, last_fd, uid);
+    if (retval != PAM_SUCCESS)
+      {
+	close(last_fd);
+	D(("error while reading lastlog file"));
+	return retval;
+      }
+
+    retval = last_login_write(pamh, announce, last_fd, uid, user);
 
     close(last_fd);
     D(("all done with last login"));
 
-    if (rc_write != PAM_SUCCESS)
-	return rc_write;
-    else
-	return rc_read;
+    return retval;
 }
 
 /* --- authentication management functions (only) --- */
Index: modules/pam_succeed_if/README
===================================================================
RCS file: /cvsroot/pam/Linux-PAM/modules/pam_succeed_if/README,v
retrieving revision 1.1
retrieving revision 1.2
diff -u -p -a -r1.1 -r1.2
--- modules/pam_succeed_if/README	24 Sep 2004 11:42:39 -0000	1.1
+++ modules/pam_succeed_if/README	21 Dec 2005 10:04:10 -0000	1.2
@@ -34,10 +34,16 @@ pam_succeed_if:
 		!~		- Wildcard mismatch.
 		ingroup		- Group membership check. [*]
 		notingroup	- Group non-membership check. [*]
+		innetgr		- Netgroup membership check. [*][+]
+		notinnetgr	- Netgroup non-membership check. [*][+]
 
-		* The "ingroup" and "notingroup" operators should only be
-		  used with the USER attribute.
+		* The "ingroup", "notingroup", "innetgr" and "notinnetgr"
+		  operators should only be used with the USER attribute.
 
+		+ The "innetgr" and "notinnetgr" operators always match
+		  both remote host and USER against the netgroup. If a remote
+		  host is not set by the application it will be matched
+		  against any host in the netgroup triplet.
 	Examples:
 
 		Deny authentication to all users except those in the wheel
Index: modules/pam_succeed_if/pam_succeed_if.c
===================================================================
RCS file: /cvsroot/pam/Linux-PAM/modules/pam_succeed_if/pam_succeed_if.c,v
retrieving revision 1.8
retrieving revision 1.9
diff -u -p -a -r1.8 -r1.9
--- modules/pam_succeed_if/pam_succeed_if.c	12 Dec 2005 14:45:03 -0000	1.8
+++ modules/pam_succeed_if/pam_succeed_if.c	21 Dec 2005 10:04:10 -0000	1.9
@@ -52,6 +52,7 @@
 #include <unistd.h>
 #include <pwd.h>
 #include <grp.h>
+#include <netdb.h>
 #include <security/pam_modules.h>
 #include <security/pam_modutil.h>
 #include <security/pam_ext.h>
@@ -183,30 +184,32 @@ evaluate_noglob(const char *left, const 
 static int
 evaluate_ingroup(pam_handle_t *pamh, const char *user, const char *group)
 {
-	int ret;
-	ret = pam_modutil_user_in_group_nam_nam(pamh, user, group);
-	switch (ret) {
-	case 1:
+	if (pam_modutil_user_in_group_nam_nam(pamh, user, group) == 1)
 		return PAM_SUCCESS;
-		break;
-	default:
-		break;
-	}
 	return PAM_AUTH_ERR;
 }
 /* Return PAM_SUCCESS if the user is NOT in the group. */
 static int
 evaluate_notingroup(pam_handle_t *pamh, const char *user, const char *group)
 {
-	int ret;
-	ret = pam_modutil_user_in_group_nam_nam(pamh, user, group);
-	switch (ret) {
-	case 0:
+	if (pam_modutil_user_in_group_nam_nam(pamh, user, group) == 0)
+		return PAM_SUCCESS;
+	return PAM_AUTH_ERR;
+}
+/* Return PAM_SUCCESS if the (host,user) is in the netgroup. */
+static int
+evaluate_innetgr(const char *host, const char *user, const char *group)
+{
+	if (innetgr(group, host, user, NULL) == 1)
+		return PAM_SUCCESS;
+	return PAM_AUTH_ERR;
+}
+/* Return PAM_SUCCESS if the (host,user) is NOT in the netgroup. */
+static int
+evaluate_notinnetgr(const char *host, const char *user, const char *group)
+{
+	if (innetgr(group, host, user, NULL) == 0)
 		return PAM_SUCCESS;
-		break;
-	default:
-		break;
-	}
 	return PAM_AUTH_ERR;
 }
 
@@ -306,6 +309,20 @@ evaluate(pam_handle_t *pamh, int debug,
 	if (strcasecmp(qual, "notingroup") == 0) {
 		return evaluate_notingroup(pamh, pwd->pw_name, right);
 	}
+	/* (Rhost, user) is in this netgroup. */
+	if (strcasecmp(qual, "innetgr") == 0) {
+		const void *rhost;
+		if (pam_get_item(pamh, PAM_RHOST, &rhost) != PAM_SUCCESS)
+			rhost = NULL;		
+		return evaluate_innetgr(rhost, pwd->pw_name, right);
+	}
+	/* (Rhost, user) is not in this group. */
+	if (strcasecmp(qual, "notinnetgr") == 0) {
+		const void *rhost;
+		if (pam_get_item(pamh, PAM_RHOST, &rhost) != PAM_SUCCESS)
+			rhost = NULL;		
+		return evaluate_notinnetgr(rhost, pwd->pw_name, right);
+	}
 	/* Fail closed. */
 	return PAM_SERVICE_ERR;
 }
Index: modules/pam_time/pam_time.c
===================================================================
RCS file: /cvsroot/pam/Linux-PAM/modules/pam_time/pam_time.c,v
retrieving revision 1.10
retrieving revision 1.11
diff -u -p -a -r1.10 -r1.11
--- modules/pam_time/pam_time.c	9 Nov 2005 10:17:00 -0000	1.10
+++ modules/pam_time/pam_time.c	19 Dec 2005 07:31:17 -0000	1.11
@@ -1,18 +1,11 @@
 /* pam_time module */
 
 /*
- * $Id: pam_time.c,v 1.10 2005/11/09 10:17:00 t8m Exp $
- *
  * Written by Andrew Morgan <morgan@linux.kernel.org> 1996/6/22
  * (File syntax and much other inspiration from the shadow package
  * shadow-960129)
  */
 
-static const char rcsid[] =
-"$Id: pam_time.c,v 1.10 2005/11/09 10:17:00 t8m Exp $;\n"
-"\t\tVersion 0.22 for Linux-PAM\n"
-"Copyright (C) Andrew G. Morgan 1996 <morgan@linux.kernel.org>\n";
-
 #include "config.h"
 
 #include <sys/file.h>
@@ -27,6 +20,7 @@ static const char rcsid[] =
 #include <sys/types.h>
 #include <sys/stat.h>
 #include <fcntl.h>
+#include <netdb.h>
 
 #define PAM_TIME_BUFLEN        1000
 #define FIELD_SEPARATOR        ';'   /* this is new as of .02 */
@@ -522,7 +516,11 @@ check_account(pam_handle_t *pamh, const 
 			  "%s: no user entry #%d", PAM_TIME_CONF, count);
 	       continue;
 	  }
-	  good &= logic_field(pamh, user, buffer, count, is_same);
+	  /* If buffer starts with @, we are using netgroups */
+	  if (buffer[0] == '@')
+	    good &= innetgr (&buffer[1], NULL, user, NULL);
+	  else
+	    good &= logic_field(pamh, user, buffer, count, is_same);
 	  D(("with user: %s", good ? "passes":"fails" ));
 
 	  /* here we get the time field */
Index: modules/pam_time/time.conf
===================================================================
RCS file: /cvsroot/pam/Linux-PAM/modules/pam_time/time.conf,v
retrieving revision 1.1.1.1
retrieving revision 1.2
diff -u -p -a -r1.1.1.1 -r1.2
--- modules/pam_time/time.conf	20 Jun 2000 22:12:01 -0000	1.1.1.1
+++ modules/pam_time/time.conf	19 Dec 2005 07:31:17 -0000	1.2
@@ -20,7 +20,8 @@
 #	is a logic list of terminal names that this rule applies to.
 #
 # users
-#	is a logic list of users to whom this rule applies.
+#	is a logic list of users or a netgroup of users to whom this
+#	rule applies.
 #
 # NB. For these items the simple wildcard '*' may be used only once.
 #
@@ -61,4 +62,4 @@
 
 #
 # End of example file.
-#
\ No newline at end of file
+#
Index: po/Linux-PAM.pot
===================================================================
RCS file: /cvsroot/pam/Linux-PAM/po/Linux-PAM.pot,v
retrieving revision 1.19
retrieving revision 1.20
diff -u -p -a -r1.19 -r1.20
--- po/Linux-PAM.pot	12 Dec 2005 17:34:10 -0000	1.19
+++ po/Linux-PAM.pot	20 Dec 2005 21:28:07 -0000	1.20
@@ -8,7 +8,7 @@ msgid ""
 msgstr ""
 "Project-Id-Version: PACKAGE VERSION\n"
 "Report-Msgid-Bugs-To: http://sourceforge.net/projects/pam\n"
-"POT-Creation-Date: 2005-12-12 18:34+0100\n"
+"POT-Creation-Date: 2005-12-20 22:26+0100\n"
 "PO-Revision-Date: YEAR-MO-DA HO:MI+ZONE\n"
 "Last-Translator: FULL NAME <EMAIL@ADDRESS>\n"
 "Language-Team: LANGUAGE <LL@li.org>\n"
@@ -54,24 +54,24 @@ msgid "%s: Can't reset all users to non-
 msgstr ""
 
 #. TRANSLATORS: " from <host>"
-#: modules/pam_lastlog/pam_lastlog.c:204
+#: modules/pam_lastlog/pam_lastlog.c:195
 #, c-format
 msgid " from %.*s"
 msgstr ""
 
 #. TRANSLATORS: " on <terminal>"
-#: modules/pam_lastlog/pam_lastlog.c:215
+#: modules/pam_lastlog/pam_lastlog.c:207
 #, c-format
 msgid " on %.*s"
 msgstr ""
 
 #. TRANSLATORS: "Last login: <date> from <host> on <terminal>"
-#: modules/pam_lastlog/pam_lastlog.c:223
+#: modules/pam_lastlog/pam_lastlog.c:216
 #, c-format
 msgid "Last login:%s%s%s"
 msgstr ""
 
-#: modules/pam_lastlog/pam_lastlog.c:229
+#: modules/pam_lastlog/pam_lastlog.c:222
 msgid "Welcome to your new account!"
 msgstr ""
 
Index: po/cs.po
===================================================================
RCS file: /cvsroot/pam/Linux-PAM/po/cs.po,v
retrieving revision 1.20
retrieving revision 1.21
diff -u -p -a -r1.20 -r1.21
--- po/cs.po	12 Dec 2005 17:34:10 -0000	1.20
+++ po/cs.po	20 Dec 2005 21:28:07 -0000	1.21
@@ -7,7 +7,7 @@ msgid ""
 msgstr ""
 "Project-Id-Version: Linux-PAM\n"
 "Report-Msgid-Bugs-To: http://sourceforge.net/projects/pam\n"
-"POT-Creation-Date: 2005-12-12 18:34+0100\n"
+"POT-Creation-Date: 2005-12-20 22:26+0100\n"
 "PO-Revision-Date: 2005-12-12 13:14+0100\n"
 "Last-Translator: Tomas Mraz <t8m@centrum.cz>\n"
 "Language-Team: cs_CZ <cs@li.org>\n"
@@ -57,24 +57,24 @@ msgid "%s: Can't reset all users to non-
 msgstr "%s: Nelze resetovat všechny uživatele nenulově\n"
 
 #. TRANSLATORS: " from <host>"
-#: modules/pam_lastlog/pam_lastlog.c:204
+#: modules/pam_lastlog/pam_lastlog.c:195
 #, c-format
 msgid " from %.*s"
 msgstr " z %.*s"
 
 #. TRANSLATORS: " on <terminal>"
-#: modules/pam_lastlog/pam_lastlog.c:215
+#: modules/pam_lastlog/pam_lastlog.c:207
 #, c-format
 msgid " on %.*s"
 msgstr " na %.*s"
 
 #. TRANSLATORS: "Last login: <date> from <host> on <terminal>"
-#: modules/pam_lastlog/pam_lastlog.c:223
+#: modules/pam_lastlog/pam_lastlog.c:216
 #, c-format
 msgid "Last login:%s%s%s"
 msgstr "Poslední přihlášení:%s%s%s"
 
-#: modules/pam_lastlog/pam_lastlog.c:229
+#: modules/pam_lastlog/pam_lastlog.c:222
 msgid "Welcome to your new account!"
 msgstr "Vítejte na vašem novém účtu!"
 
Index: po/de.po
===================================================================
RCS file: /cvsroot/pam/Linux-PAM/po/de.po,v
retrieving revision 1.20
retrieving revision 1.22
diff -u -p -a -r1.20 -r1.22
--- po/de.po	12 Dec 2005 17:34:10 -0000	1.20
+++ po/de.po	20 Dec 2005 21:28:07 -0000	1.22
@@ -7,7 +7,7 @@ msgid ""
 msgstr ""
 "Project-Id-Version: Linux-PAM\n"
 "Report-Msgid-Bugs-To: http://sourceforge.net/projects/pam\n"
-"POT-Creation-Date: 2005-12-12 18:34+0100\n"
+"POT-Creation-Date: 2005-12-20 22:26+0100\n"
 "PO-Revision-Date: 2005-12-12 17:55+01:00\n"
 "Last-Translator: Novell Language <language@novell.com>\n"
 "Language-Team: Novell Language <language@novell.com>\n"
@@ -55,24 +55,24 @@ msgstr ""
 "%s: Es können nicht alle Benutzer auf Nicht-null zurückgesetzt werden\n"
 
 #. TRANSLATORS: " from <host>"
-#: modules/pam_lastlog/pam_lastlog.c:204
+#: modules/pam_lastlog/pam_lastlog.c:195
 #, c-format
 msgid " from %.*s"
 msgstr " von %.*s"
 
 #. TRANSLATORS: " on <terminal>"
-#: modules/pam_lastlog/pam_lastlog.c:215
+#: modules/pam_lastlog/pam_lastlog.c:207
 #, c-format
 msgid " on %.*s"
 msgstr " auf %.*s"
 
 #. TRANSLATORS: "Last login: <date> from <host> on <terminal>"
-#: modules/pam_lastlog/pam_lastlog.c:223
+#: modules/pam_lastlog/pam_lastlog.c:216
 #, c-format
 msgid "Last login:%s%s%s"
-msgstr "Letzte Anmeldung: %s%s%s"
+msgstr "Letzte Anmeldung:%s%s%s"
 
-#: modules/pam_lastlog/pam_lastlog.c:229
+#: modules/pam_lastlog/pam_lastlog.c:222
 msgid "Welcome to your new account!"
 msgstr "Willkommen in Ihrem neuen Account!"
 
Index: po/es.po
===================================================================
RCS file: /cvsroot/pam/Linux-PAM/po/es.po,v
retrieving revision 1.19
retrieving revision 1.20
diff -u -p -a -r1.19 -r1.20
--- po/es.po	12 Dec 2005 17:34:10 -0000	1.19
+++ po/es.po	20 Dec 2005 21:28:07 -0000	1.20
@@ -7,7 +7,7 @@ msgid ""
 msgstr ""
 "Project-Id-Version: Linux-PAM\n"
 "Report-Msgid-Bugs-To: http://sourceforge.net/projects/pam\n"
-"POT-Creation-Date: 2005-12-12 18:34+0100\n"
+"POT-Creation-Date: 2005-12-20 22:26+0100\n"
 "PO-Revision-Date: 2005-08-18 HO:MI+ZONE\n"
 "Last-Translator: Novell Language <language@novell.com>\n"
 "Language-Team: Novell Language <language@novell.com>\n"
@@ -57,24 +57,24 @@ msgstr ""
 "cero\n"
 
 #. TRANSLATORS: " from <host>"
-#: modules/pam_lastlog/pam_lastlog.c:204
+#: modules/pam_lastlog/pam_lastlog.c:195
 #, c-format
 msgid " from %.*s"
 msgstr ""
 
 #. TRANSLATORS: " on <terminal>"
-#: modules/pam_lastlog/pam_lastlog.c:215
+#: modules/pam_lastlog/pam_lastlog.c:207
 #, c-format
 msgid " on %.*s"
 msgstr ""
 
 #. TRANSLATORS: "Last login: <date> from <host> on <terminal>"
-#: modules/pam_lastlog/pam_lastlog.c:223
+#: modules/pam_lastlog/pam_lastlog.c:216
 #, c-format
 msgid "Last login:%s%s%s"
 msgstr ""
 
-#: modules/pam_lastlog/pam_lastlog.c:229
+#: modules/pam_lastlog/pam_lastlog.c:222
 msgid "Welcome to your new account!"
 msgstr ""
 
Index: po/fi.po
===================================================================
RCS file: /cvsroot/pam/Linux-PAM/po/fi.po,v
retrieving revision 1.12
retrieving revision 1.13
diff -u -p -a -r1.12 -r1.13
--- po/fi.po	12 Dec 2005 17:34:10 -0000	1.12
+++ po/fi.po	20 Dec 2005 21:28:07 -0000	1.13
@@ -7,7 +7,7 @@ msgid ""
 msgstr ""
 "Project-Id-Version: Linux-PAM\n"
 "Report-Msgid-Bugs-To: http://sourceforge.net/projects/pam\n"
-"POT-Creation-Date: 2005-12-12 18:34+0100\n"
+"POT-Creation-Date: 2005-12-20 22:26+0100\n"
 "PO-Revision-Date: 2005-09-17 12:27+0300\n"
 "Last-Translator: Ilkka Pirskanen <ilkka.pirskanen@kolumbus.fi>\n"
 "Language-Team: Finnish <fi@li.org>\n"
@@ -56,24 +56,24 @@ msgid "%s: Can't reset all users to non-
 msgstr "%s: Ei voida palauttaa kaikkia käyttäjiä ei-nolliksi\n"
 
 #. TRANSLATORS: " from <host>"
-#: modules/pam_lastlog/pam_lastlog.c:204
+#: modules/pam_lastlog/pam_lastlog.c:195
 #, c-format
 msgid " from %.*s"
 msgstr ""
 
 #. TRANSLATORS: " on <terminal>"
-#: modules/pam_lastlog/pam_lastlog.c:215
+#: modules/pam_lastlog/pam_lastlog.c:207
 #, c-format
 msgid " on %.*s"
 msgstr ""
 
 #. TRANSLATORS: "Last login: <date> from <host> on <terminal>"
-#: modules/pam_lastlog/pam_lastlog.c:223
+#: modules/pam_lastlog/pam_lastlog.c:216
 #, c-format
 msgid "Last login:%s%s%s"
 msgstr ""
 
-#: modules/pam_lastlog/pam_lastlog.c:229
+#: modules/pam_lastlog/pam_lastlog.c:222
 msgid "Welcome to your new account!"
 msgstr ""
 
Index: po/fr.po
===================================================================
RCS file: /cvsroot/pam/Linux-PAM/po/fr.po,v
retrieving revision 1.19
retrieving revision 1.20
diff -u -p -a -r1.19 -r1.20
--- po/fr.po	12 Dec 2005 17:34:10 -0000	1.19
+++ po/fr.po	20 Dec 2005 21:28:07 -0000	1.20
@@ -7,7 +7,7 @@ msgid ""
 msgstr ""
 "Project-Id-Version: Linux-PAM\n"
 "Report-Msgid-Bugs-To: http://sourceforge.net/projects/pam\n"
-"POT-Creation-Date: 2005-12-12 18:34+0100\n"
+"POT-Creation-Date: 2005-12-20 22:26+0100\n"
 "PO-Revision-Date: 2005-08-18 12:12+0000\n"
 "Last-Translator: Novell Language <language@novell.com>\n"
 "Language-Team: Novell Language <language@novell.com>\n"
@@ -54,24 +54,24 @@ msgid "%s: Can't reset all users to non-
 msgstr "%s: Impossible de réinitialiser tous les utilisateurs à non-zéro\n"
 
 #. TRANSLATORS: " from <host>"
-#: modules/pam_lastlog/pam_lastlog.c:204
+#: modules/pam_lastlog/pam_lastlog.c:195
 #, c-format
 msgid " from %.*s"
 msgstr ""
 
 #. TRANSLATORS: " on <terminal>"
-#: modules/pam_lastlog/pam_lastlog.c:215
+#: modules/pam_lastlog/pam_lastlog.c:207
 #, c-format
 msgid " on %.*s"
 msgstr ""
 
 #. TRANSLATORS: "Last login: <date> from <host> on <terminal>"
-#: modules/pam_lastlog/pam_lastlog.c:223
+#: modules/pam_lastlog/pam_lastlog.c:216
 #, c-format
 msgid "Last login:%s%s%s"
 msgstr ""
 
-#: modules/pam_lastlog/pam_lastlog.c:229
+#: modules/pam_lastlog/pam_lastlog.c:222
 msgid "Welcome to your new account!"
 msgstr ""
 
Index: po/hu.po
===================================================================
RCS file: /cvsroot/pam/Linux-PAM/po/hu.po,v
retrieving revision 1.19
retrieving revision 1.20
diff -u -p -a -r1.19 -r1.20
--- po/hu.po	12 Dec 2005 17:34:10 -0000	1.19
+++ po/hu.po	20 Dec 2005 21:28:07 -0000	1.20
@@ -7,7 +7,7 @@ msgid ""
 msgstr ""
 "Project-Id-Version: Linux-PAM 0.80\n"
 "Report-Msgid-Bugs-To: http://sourceforge.net/projects/pam\n"
-"POT-Creation-Date: 2005-12-12 18:34+0100\n"
+"POT-Creation-Date: 2005-12-20 22:26+0100\n"
 "PO-Revision-Date: 2005-08-07 02:59+0100\n"
 "Last-Translator: Szabolcs Varga <shirokuma@shirokuma.hu>\n"
 "Language-Team: hu\n"
@@ -54,24 +54,24 @@ msgid "%s: Can't reset all users to non-
 msgstr "%s: Nem állítható vissza minden felhasználó nem nullára\n"
 
 #. TRANSLATORS: " from <host>"
-#: modules/pam_lastlog/pam_lastlog.c:204
+#: modules/pam_lastlog/pam_lastlog.c:195
 #, c-format
 msgid " from %.*s"
 msgstr ""
 
 #. TRANSLATORS: " on <terminal>"
-#: modules/pam_lastlog/pam_lastlog.c:215
+#: modules/pam_lastlog/pam_lastlog.c:207
 #, c-format
 msgid " on %.*s"
 msgstr ""
 
 #. TRANSLATORS: "Last login: <date> from <host> on <terminal>"
-#: modules/pam_lastlog/pam_lastlog.c:223
+#: modules/pam_lastlog/pam_lastlog.c:216
 #, c-format
 msgid "Last login:%s%s%s"
 msgstr ""
 
-#: modules/pam_lastlog/pam_lastlog.c:229
+#: modules/pam_lastlog/pam_lastlog.c:222
 msgid "Welcome to your new account!"
 msgstr ""
 
Index: po/it.po
===================================================================
RCS file: /cvsroot/pam/Linux-PAM/po/it.po,v
retrieving revision 1.19
retrieving revision 1.20
diff -u -p -a -r1.19 -r1.20
--- po/it.po	12 Dec 2005 17:34:10 -0000	1.19
+++ po/it.po	20 Dec 2005 21:28:07 -0000	1.20
@@ -7,7 +7,7 @@ msgid ""
 msgstr ""
 "Project-Id-Version: Linux-PAM\n"
 "Report-Msgid-Bugs-To: http://sourceforge.net/projects/pam\n"
-"POT-Creation-Date: 2005-12-12 18:34+0100\n"
+"POT-Creation-Date: 2005-12-20 22:26+0100\n"
 "PO-Revision-Date: 2005-08-18 13:43-0000\n"
 "Last-Translator: Novell Language <language@novell.com>\n"
 "Language-Team: Novell Language <language@novell.com>\n"
@@ -55,24 +55,24 @@ msgstr ""
 "%s: Impossibile ripristinare tutti gli utenti a valori diversi da zero\n"
 
 #. TRANSLATORS: " from <host>"
-#: modules/pam_lastlog/pam_lastlog.c:204
+#: modules/pam_lastlog/pam_lastlog.c:195
 #, c-format
 msgid " from %.*s"
 msgstr ""
 
 #. TRANSLATORS: " on <terminal>"
-#: modules/pam_lastlog/pam_lastlog.c:215
+#: modules/pam_lastlog/pam_lastlog.c:207
 #, c-format
 msgid " on %.*s"
 msgstr ""
 
 #. TRANSLATORS: "Last login: <date> from <host> on <terminal>"
-#: modules/pam_lastlog/pam_lastlog.c:223
+#: modules/pam_lastlog/pam_lastlog.c:216
 #, c-format
 msgid "Last login:%s%s%s"
 msgstr ""
 
-#: modules/pam_lastlog/pam_lastlog.c:229
+#: modules/pam_lastlog/pam_lastlog.c:222
 msgid "Welcome to your new account!"
 msgstr ""
 
Index: po/ja.po
===================================================================
RCS file: /cvsroot/pam/Linux-PAM/po/ja.po,v
retrieving revision 1.19
retrieving revision 1.20
diff -u -p -a -r1.19 -r1.20
--- po/ja.po	12 Dec 2005 17:34:10 -0000	1.19
+++ po/ja.po	20 Dec 2005 21:28:07 -0000	1.20
@@ -7,7 +7,7 @@ msgid ""
 msgstr ""
 "Project-Id-Version: Linux-PAM\n"
 "Report-Msgid-Bugs-To: http://sourceforge.net/projects/pam\n"
-"POT-Creation-Date: 2005-12-12 18:34+0100\n"
+"POT-Creation-Date: 2005-12-20 22:26+0100\n"
 "PO-Revision-Date: 2005-08-18 11:52-0000\n"
 "Last-Translator: Novell Language <language@novell.com>\n"
 "Language-Team: Novell Language <language@novell.com>\n"
@@ -54,24 +54,24 @@ msgid "%s: Can't reset all users to non-
 msgstr "%s: すべてのユーザを非ゼロにリセットできません\n"
 
 #. TRANSLATORS: " from <host>"
-#: modules/pam_lastlog/pam_lastlog.c:204
+#: modules/pam_lastlog/pam_lastlog.c:195
 #, c-format
 msgid " from %.*s"
 msgstr ""
 
 #. TRANSLATORS: " on <terminal>"
-#: modules/pam_lastlog/pam_lastlog.c:215
+#: modules/pam_lastlog/pam_lastlog.c:207
 #, c-format
 msgid " on %.*s"
 msgstr ""
 
 #. TRANSLATORS: "Last login: <date> from <host> on <terminal>"
-#: modules/pam_lastlog/pam_lastlog.c:223
+#: modules/pam_lastlog/pam_lastlog.c:216
 #, c-format
 msgid "Last login:%s%s%s"
 msgstr ""
 
-#: modules/pam_lastlog/pam_lastlog.c:229
+#: modules/pam_lastlog/pam_lastlog.c:222
 msgid "Welcome to your new account!"
 msgstr ""
 
Index: po/nb.po
===================================================================
RCS file: /cvsroot/pam/Linux-PAM/po/nb.po,v
retrieving revision 1.19
retrieving revision 1.20
diff -u -p -a -r1.19 -r1.20
--- po/nb.po	12 Dec 2005 17:34:10 -0000	1.19
+++ po/nb.po	20 Dec 2005 21:28:07 -0000	1.20
@@ -7,7 +7,7 @@ msgid ""
 msgstr ""
 "Project-Id-Version: Linux-PAM\n"
 "Report-Msgid-Bugs-To: http://sourceforge.net/projects/pam\n"
-"POT-Creation-Date: 2005-12-12 18:34+0100\n"
+"POT-Creation-Date: 2005-12-20 22:26+0100\n"
 "PO-Revision-Date: 2005-08-13 22:44+0200\n"
 "Last-Translator: Olav Pettershagen <olav.pet@online.no>\n"
 "Language-Team:  <nb@li.org>\n"
@@ -55,24 +55,24 @@ msgid "%s: Can't reset all users to non-
 msgstr "%s: Kan ikke tilbakestille alle brukere til non-zero\n"
 
 #. TRANSLATORS: " from <host>"
-#: modules/pam_lastlog/pam_lastlog.c:204
+#: modules/pam_lastlog/pam_lastlog.c:195
 #, c-format
 msgid " from %.*s"
 msgstr ""
 
 #. TRANSLATORS: " on <terminal>"
-#: modules/pam_lastlog/pam_lastlog.c:215
+#: modules/pam_lastlog/pam_lastlog.c:207
 #, c-format
 msgid " on %.*s"
 msgstr ""
 
 #. TRANSLATORS: "Last login: <date> from <host> on <terminal>"
-#: modules/pam_lastlog/pam_lastlog.c:223
+#: modules/pam_lastlog/pam_lastlog.c:216
 #, c-format
 msgid "Last login:%s%s%s"
 msgstr ""
 
-#: modules/pam_lastlog/pam_lastlog.c:229
+#: modules/pam_lastlog/pam_lastlog.c:222
 msgid "Welcome to your new account!"
 msgstr ""
 
Index: po/pa.po
===================================================================
RCS file: /cvsroot/pam/Linux-PAM/po/pa.po,v
retrieving revision 1.19
retrieving revision 1.20
diff -u -p -a -r1.19 -r1.20
--- po/pa.po	12 Dec 2005 17:34:10 -0000	1.19
+++ po/pa.po	20 Dec 2005 21:28:07 -0000	1.20
@@ -8,7 +8,7 @@ msgid ""
 msgstr ""
 "Project-Id-Version: Linux-PAM.pa\n"
 "Report-Msgid-Bugs-To: http://sourceforge.net/projects/pam\n"
-"POT-Creation-Date: 2005-12-12 18:34+0100\n"
+"POT-Creation-Date: 2005-12-20 22:26+0100\n"
 "PO-Revision-Date: 2005-08-06 08:34+0530\n"
 "Last-Translator: Amanpreet Singh Alam[ਆਲਮ] <amanpreetalam@yahoo.com>\n"
 "Language-Team: Panjabi <pa@li.org>\n"
@@ -57,24 +57,24 @@ msgid "%s: Can't reset all users to non-
 msgstr ""
 
 #. TRANSLATORS: " from <host>"
-#: modules/pam_lastlog/pam_lastlog.c:204
+#: modules/pam_lastlog/pam_lastlog.c:195
 #, c-format
 msgid " from %.*s"
 msgstr ""
 
 #. TRANSLATORS: " on <terminal>"
-#: modules/pam_lastlog/pam_lastlog.c:215
+#: modules/pam_lastlog/pam_lastlog.c:207
 #, c-format
 msgid " on %.*s"
 msgstr ""
 
 #. TRANSLATORS: "Last login: <date> from <host> on <terminal>"
-#: modules/pam_lastlog/pam_lastlog.c:223
+#: modules/pam_lastlog/pam_lastlog.c:216
 #, c-format
 msgid "Last login:%s%s%s"
 msgstr ""
 
-#: modules/pam_lastlog/pam_lastlog.c:229
+#: modules/pam_lastlog/pam_lastlog.c:222
 msgid "Welcome to your new account!"
 msgstr ""
 
Index: po/pl.po
===================================================================
RCS file: /cvsroot/pam/Linux-PAM/po/pl.po,v
retrieving revision 1.18
retrieving revision 1.19
diff -u -p -a -r1.18 -r1.19
--- po/pl.po	12 Dec 2005 17:34:10 -0000	1.18
+++ po/pl.po	20 Dec 2005 21:28:07 -0000	1.19
@@ -5,7 +5,7 @@ msgid ""
 msgstr ""
 "Project-Id-Version: Linux-PAM TBD\n"
 "Report-Msgid-Bugs-To: http://sourceforge.net/projects/pam\n"
-"POT-Creation-Date: 2005-12-12 18:34+0100\n"
+"POT-Creation-Date: 2005-12-20 22:26+0100\n"
 "PO-Revision-Date: 2005-08-23 15:22+0200\n"
 "Last-Translator: Piotr Bolek <pb@7bulls.com>\n"
 "Language-Team: Polish <i18n@suse.de>\n"
@@ -54,24 +54,24 @@ msgid "%s: Can't reset all users to non-
 msgstr "%s: Nie można zresetować wszystkich użytkowników\n"
 
 #. TRANSLATORS: " from <host>"
-#: modules/pam_lastlog/pam_lastlog.c:204
+#: modules/pam_lastlog/pam_lastlog.c:195
 #, c-format
 msgid " from %.*s"
 msgstr ""
 
 #. TRANSLATORS: " on <terminal>"
-#: modules/pam_lastlog/pam_lastlog.c:215
+#: modules/pam_lastlog/pam_lastlog.c:207
 #, c-format
 msgid " on %.*s"
 msgstr ""
 
 #. TRANSLATORS: "Last login: <date> from <host> on <terminal>"
-#: modules/pam_lastlog/pam_lastlog.c:223
+#: modules/pam_lastlog/pam_lastlog.c:216
 #, c-format
 msgid "Last login:%s%s%s"
 msgstr ""
 
-#: modules/pam_lastlog/pam_lastlog.c:229
+#: modules/pam_lastlog/pam_lastlog.c:222
 msgid "Welcome to your new account!"
 msgstr ""
 
Index: po/pt.po
===================================================================
RCS file: /cvsroot/pam/Linux-PAM/po/pt.po,v
retrieving revision 1.19
retrieving revision 1.20
diff -u -p -a -r1.19 -r1.20
--- po/pt.po	12 Dec 2005 17:34:10 -0000	1.19
+++ po/pt.po	20 Dec 2005 21:28:07 -0000	1.20
@@ -7,7 +7,7 @@ msgid ""
 msgstr ""
 "Project-Id-Version: Linux-PAM-pt\n"
 "Report-Msgid-Bugs-To: http://sourceforge.net/projects/pam\n"
-"POT-Creation-Date: 2005-12-12 18:34+0100\n"
+"POT-Creation-Date: 2005-12-20 22:26+0100\n"
 "PO-Revision-Date: 2005-08-07 00:41+0100\n"
 "Last-Translator: Antonio Cardoso Martins <digiplan@netvisao.pt>\n"
 "Language-Team: portuguese\n"
@@ -55,24 +55,24 @@ msgid "%s: Can't reset all users to non-
 msgstr "%s: Não foi possível reiniciar todos os utilizadores para não zero\n"
 
 #. TRANSLATORS: " from <host>"
-#: modules/pam_lastlog/pam_lastlog.c:204
+#: modules/pam_lastlog/pam_lastlog.c:195
 #, c-format
 msgid " from %.*s"
 msgstr ""
 
 #. TRANSLATORS: " on <terminal>"
-#: modules/pam_lastlog/pam_lastlog.c:215
+#: modules/pam_lastlog/pam_lastlog.c:207
 #, c-format
 msgid " on %.*s"
 msgstr ""
 
 #. TRANSLATORS: "Last login: <date> from <host> on <terminal>"
-#: modules/pam_lastlog/pam_lastlog.c:223
+#: modules/pam_lastlog/pam_lastlog.c:216
 #, c-format
 msgid "Last login:%s%s%s"
 msgstr ""
 
-#: modules/pam_lastlog/pam_lastlog.c:229
+#: modules/pam_lastlog/pam_lastlog.c:222
 msgid "Welcome to your new account!"
 msgstr ""
 
Index: po/pt_BR.po
===================================================================
RCS file: /cvsroot/pam/Linux-PAM/po/pt_BR.po,v
retrieving revision 1.19
retrieving revision 1.20
diff -u -p -a -r1.19 -r1.20
--- po/pt_BR.po	12 Dec 2005 17:34:10 -0000	1.19
+++ po/pt_BR.po	20 Dec 2005 21:28:07 -0000	1.20
@@ -7,7 +7,7 @@ msgid ""
 msgstr ""
 "Project-Id-Version: Linux-PAM\n"
 "Report-Msgid-Bugs-To: http://sourceforge.net/projects/pam\n"
-"POT-Creation-Date: 2005-12-12 18:34+0100\n"
+"POT-Creation-Date: 2005-12-20 22:26+0100\n"
 "PO-Revision-Date: 2005-08-18 15:02+0000\n"
 "Last-Translator: Novell Language <language@novell.com>\n"
 "Language-Team: Novell Language <language@novell.com>\n"
@@ -54,24 +54,24 @@ msgid "%s: Can't reset all users to non-
 msgstr "%s: Impossível redefinir todos os usuários para não-zero\n"
 
 #. TRANSLATORS: " from <host>"
-#: modules/pam_lastlog/pam_lastlog.c:204
+#: modules/pam_lastlog/pam_lastlog.c:195
 #, c-format
 msgid " from %.*s"
 msgstr ""
 
 #. TRANSLATORS: " on <terminal>"
-#: modules/pam_lastlog/pam_lastlog.c:215
+#: modules/pam_lastlog/pam_lastlog.c:207
 #, c-format
 msgid " on %.*s"
 msgstr ""
 
 #. TRANSLATORS: "Last login: <date> from <host> on <terminal>"
-#: modules/pam_lastlog/pam_lastlog.c:223
+#: modules/pam_lastlog/pam_lastlog.c:216
 #, c-format
 msgid "Last login:%s%s%s"
 msgstr ""
 
-#: modules/pam_lastlog/pam_lastlog.c:229
+#: modules/pam_lastlog/pam_lastlog.c:222
 msgid "Welcome to your new account!"
 msgstr ""
 
Index: po/zh_CN.po
===================================================================
RCS file: /cvsroot/pam/Linux-PAM/po/zh_CN.po,v
retrieving revision 1.19
retrieving revision 1.20
diff -u -p -a -r1.19 -r1.20
--- po/zh_CN.po	12 Dec 2005 17:34:10 -0000	1.19
+++ po/zh_CN.po	20 Dec 2005 21:28:07 -0000	1.20
@@ -7,7 +7,7 @@ msgid ""
 msgstr ""
 "Project-Id-Version: Linux_PAM\n"
 "Report-Msgid-Bugs-To: http://sourceforge.net/projects/pam\n"
-"POT-Creation-Date: 2005-12-12 18:34+0100\n"
+"POT-Creation-Date: 2005-12-20 22:26+0100\n"
 "PO-Revision-Date: 2005-08-18 HO:MI+ZONE\n"
 "Last-Translator: Novell Language <language@novell.com>\n"
 "Language-Team: Novell Language <language@novell.com>\n"
@@ -53,24 +53,24 @@ msgid "%s: Can't reset all users to non-
 msgstr "%s: 无法将所有用户重设置为非零\n"
 
 #. TRANSLATORS: " from <host>"
-#: modules/pam_lastlog/pam_lastlog.c:204
+#: modules/pam_lastlog/pam_lastlog.c:195
 #, c-format
 msgid " from %.*s"
 msgstr ""
 
 #. TRANSLATORS: " on <terminal>"
-#: modules/pam_lastlog/pam_lastlog.c:215
+#: modules/pam_lastlog/pam_lastlog.c:207
 #, c-format
 msgid " on %.*s"
 msgstr ""
 
 #. TRANSLATORS: "Last login: <date> from <host> on <terminal>"
-#: modules/pam_lastlog/pam_lastlog.c:223
+#: modules/pam_lastlog/pam_lastlog.c:216
 #, c-format
 msgid "Last login:%s%s%s"
 msgstr ""
 
-#: modules/pam_lastlog/pam_lastlog.c:229
+#: modules/pam_lastlog/pam_lastlog.c:222
 msgid "Welcome to your new account!"
 msgstr ""
 
Index: po/zh_TW.po
===================================================================
RCS file: /cvsroot/pam/Linux-PAM/po/zh_TW.po,v
retrieving revision 1.19
retrieving revision 1.20
diff -u -p -a -r1.19 -r1.20
--- po/zh_TW.po	12 Dec 2005 17:34:10 -0000	1.19
+++ po/zh_TW.po	20 Dec 2005 21:28:07 -0000	1.20
@@ -7,7 +7,7 @@ msgid ""
 msgstr ""
 "Project-Id-Version: Linux_PAM\n"
 "Report-Msgid-Bugs-To: http://sourceforge.net/projects/pam\n"
-"POT-Creation-Date: 2005-12-12 18:34+0100\n"
+"POT-Creation-Date: 2005-12-20 22:26+0100\n"
 "PO-Revision-Date: 2005-08-18 12:12+0200\n"
 "Last-Translator: Novell Language <language@novell.com>\n"
 "Language-Team: Novell Language <language@novell.com>\n"
@@ -54,24 +54,24 @@ msgid "%s: Can't reset all users to non-
 msgstr "%s: 無法將所有使用者重新設定為非零\n"
 
 #. TRANSLATORS: " from <host>"
-#: modules/pam_lastlog/pam_lastlog.c:204
+#: modules/pam_lastlog/pam_lastlog.c:195
 #, c-format
 msgid " from %.*s"
 msgstr ""
 
 #. TRANSLATORS: " on <terminal>"
-#: modules/pam_lastlog/pam_lastlog.c:215
+#: modules/pam_lastlog/pam_lastlog.c:207
 #, c-format
 msgid " on %.*s"
 msgstr ""
 
 #. TRANSLATORS: "Last login: <date> from <host> on <terminal>"
-#: modules/pam_lastlog/pam_lastlog.c:223
+#: modules/pam_lastlog/pam_lastlog.c:216
 #, c-format
 msgid "Last login:%s%s%s"
 msgstr ""
 
-#: modules/pam_lastlog/pam_lastlog.c:229
+#: modules/pam_lastlog/pam_lastlog.c:222
 msgid "Welcome to your new account!"
 msgstr ""
 
