--- Linux-PAM/modules/pam_xauth/pam_xauth.c
+++ Linux-PAM/modules/pam_xauth/pam_xauth.c
@@ -37,6 +37,9 @@
 #include <sys/types.h>
 #include <sys/fsuid.h>
 #include <sys/wait.h>
+#include <sys/stat.h>
+#include <fcntl.h>
+#include <unistd.h>
 #include <errno.h>
 #include <fnmatch.h>
 #include <grp.h>
@@ -209,8 +212,8 @@ check_acl(pam_handle_t *pamh,
 {
 	char path[PATH_MAX];
 	struct passwd *pwd;
-	FILE *fp;
-	int i;
+	FILE *fp = 0;
+	int fd, i;
 	uid_t euid;
 	/* Check this user's <sense> file. */
 	pwd = pam_modutil_getpwnam(pamh, this_user);
@@ -229,8 +232,16 @@ check_acl(pam_handle_t *pamh,
 	}
 	euid = geteuid();
 	setfsuid(pwd->pw_uid);
-	fp = fopen(path, "r");
+	fd = open(path, O_RDONLY | O_NOFOLLOW | O_NOCTTY);
 	setfsuid(euid);
+	if (fd >= 0) {
+		struct stat st;
+
+		if (fstat(fd, &st) ||
+		    !S_ISREG(st.st_mode) ||
+		    !(fp = fdopen(fd, "r")))
+			close(fd);
+	}
 	if (fp != NULL) {
 		char buf[LINE_MAX], *tmp;
 		/* Scan the file for a list of specs of users to "trust". */
