diff -ur pam-0.75.orig/libpam_misc/misc_conv.c pam-0.75/libpam_misc/misc_conv.c
--- pam-0.75.orig/libpam_misc/misc_conv.c	Mon Jan 22 19:23:08 2001
+++ pam-0.75/libpam_misc/misc_conv.c	Fri Sep 21 03:53:09 2001
@@ -16,6 +16,7 @@
 #include <termios.h>
 #include <time.h>
 #include <unistd.h>
+#include <errno.h>
 
 #include "../libpam/include/security/pam_appl.h"
 #include "../libpamc/include/security/pam_client.h"
@@ -180,22 +181,34 @@
 	    }
 	    if (expired) {
 		delay = get_delay();
-	    } else if (nc > 0) {                 /* we got some user input */
-		char *input;
-
-		if (nc > 0 && line[nc-1] == '\n') {     /* <NUL> terminate */
-		    line[--nc] = '\0';
-		} else {
-		    line[nc] = '\0';
+	    } else {
+		if (nc > 0) {                    /* we got some user input */
+		    char *input;
+
+		    if (nc > 0 && line[nc-1] == '\n') {   /* NUL-terminate */
+			line[--nc] = '\0';
+		    } else {
+			line[nc] = '\0';
+		    }
+		    input = x_strdup(line);
+		    _pam_overwrite(line);
+
+		    return input;              /* return malloc()ed string */
+	        } else {
+		    int break_on_error = 0;
+
+		    if (nc == 0) {                               /* Ctrl-D */
+			D(("user did not want to type anything"));
+			break_on_error = 1;
+		    } else {
+			D(("error while reading"));
+			if (errno != EINTR && errno != EWOULDBLOCK)
+			    break_on_error = 1;
+		    }
+		    fprintf(stderr, "\n");
+		    if (break_on_error)
+			break;
 		}
-		input = x_strdup(line);
-		_pam_overwrite(line);
-
-		return input;                  /* return malloc()ed string */
-	    } else if (nc == 0) {                                /* Ctrl-D */
-		D(("user did not want to type anything"));
-		fprintf(stderr, "\n");
-		break;
 	    }
 	}
     }
