diff -ur pam-0.75.orig/configure.in pam-0.75/configure.in
--- pam-0.75.orig/configure.in	Mon Apr 23 22:14:19 2001
+++ pam-0.75/configure.in	Sun Oct  7 09:48:48 2001
@@ -206,9 +206,6 @@
 dnl Linux wants features.h in some of the source files.
 AC_CHECK_HEADERS(features.h)
 
-dnl For module/pam_cracklib
-AC_CHECK_HEADERS(crypt.h)
-
 dnl For module/pam_userdb
 AC_CHECK_HEADERS(ndbm.h db.h)
 dnl I suspect the following two lines are a hack.
@@ -220,29 +217,7 @@
 
 dnl This following rule should be made conditional upon HAVE_LIBCRYPT
 dnl being found.
-
-dnl Look for cracklib dictionary
-AC_MSG_CHECKING(path to cracklib dictionary)
-DICT_DIR_CANDIDATES="/usr/lib /usr/share/dict /usr/share/lib \
-  /usr/local/lib /usr/local/share/lib"
-DICT_FILE_CANDIDATES="pw_dict cracklib_dict"
 CRACKLIB_DICTPATH=""
-for d in $DICT_DIR_CANDIDATES ; do
-      for f in $DICT_FILE_CANDIDATES ; do
-              if test -r $d/$f.hwm ; then
-                      CRACKLIB_DICTPATH=$d/$f
-                      break 2
-              elif test -r $d/dict/$f.hwm ; then
-                      CRACKLIB_DICTPATH=$d/dict/$f
-                      break 2
-              fi
-      done
-done
-if test -z "$CRACKLIB_DICTPATH" ; then
-      AC_MSG_RESULT(none found)
-else
-      AC_MSG_RESULT($CRACKLIB_DICTPATH)
-fi
 AC_SUBST(CRACKLIB_DICTPATH)
 
 dnl Set FLAGS, linker options etc. depending on C compiler.
