diff -urp pam-0.75.orig/modules/pam_wheel/README pam-0.75/modules/pam_wheel/README
--- pam-0.75.orig/modules/pam_wheel/README	Fri Apr  6 19:49:49 2001
+++ pam-0.75/modules/pam_wheel/README	Tue Jul 22 10:34:57 2003
@@ -1,15 +1,14 @@
 
 pam_wheel:
-	only permit root authentication too members of wheel group
+	only permit root authentication to members of wheel group
 
 RECOGNIZED ARGUMENTS:
 	debug		write a message to syslog indicating success or
 			failure.
 
 	use_uid		the check for wheel membership will be done against
-			the current uid instead of the original one
-			(useful when jumping with su from one account to
-			another for example)
+			the current uid; this version of pam_wheel can only
+			work in this way and ignores the 'use_uid' argument.
 
 	trust		the pam_wheel module will return PAM_SUCCESS instead
 			of PAM_IGNORE if the user is a member of the wheel
@@ -26,7 +25,7 @@ RECOGNIZED ARGUMENTS:
 			which case we return PAM_SUCCESS).
 
 	group=xxxx	Instead of checking the GID 0 group, use the xxxx
-			group to perform the authentification.
+			group to perform the authentication.
 
 MODULE SERVICES PROVIDED:
 	auth		_authentication and _setcred (blank)
diff -urp pam-0.75.orig/modules/pam_wheel/pam_wheel.c pam-0.75/modules/pam_wheel/pam_wheel.c
--- pam-0.75.orig/modules/pam_wheel/pam_wheel.c	Mon Jul  9 23:38:13 2001
+++ pam-0.75/modules/pam_wheel/pam_wheel.c	Tue Jul 22 10:49:37 2003
@@ -120,7 +120,7 @@ int pam_sm_authenticate(pam_handle_t *pa
      char use_group[BUFSIZ];
     
      ctrl = _pam_parse(argc, argv, use_group, sizeof(use_group));
-     retval = pam_get_user(pamh,&username,"login: ");
+     retval = pam_get_user(pamh, &username, NULL);
      if ((retval != PAM_SUCCESS) || (!username)) {
         if (ctrl & PAM_DEBUG_ARG)
             _pam_log(LOG_DEBUG,"can not get the username");
@@ -131,7 +131,7 @@ int pam_sm_authenticate(pam_handle_t *pa
      pwd = getpwnam(username);
      if (!pwd) {
         if (ctrl & PAM_DEBUG_ARG)
-            _pam_log(LOG_NOTICE,"unknown user %s",username);
+            _pam_log(LOG_NOTICE,"unknown username");
         return PAM_USER_UNKNOWN;
      }
      
@@ -147,7 +147,9 @@ int pam_sm_authenticate(pam_handle_t *pa
         return PAM_IGNORE;
      }
      
+#if 0
      if (ctrl & PAM_USE_UID_ARG) {
+#endif
          tpwd = getpwuid(getuid());
          if (!tpwd) {
             if (ctrl & PAM_DEBUG_ARG)
@@ -155,6 +157,7 @@ int pam_sm_authenticate(pam_handle_t *pa
             return PAM_SERVICE_ERR;
          }
          fromsu = tpwd->pw_name;
+#if 0
      } else {
          fromsu = getlogin();
          tpwd = getpwnam(fromsu);
@@ -164,6 +167,7 @@ int pam_sm_authenticate(pam_handle_t *pa
              return PAM_SERVICE_ERR;
          }
      }
+#endif
      
      if (!use_group[0]) {
 	 if ((grp = getgrnam("wheel")) == NULL) {
