diff -purN vim61.orig/src/digraph.c vim61/src/digraph.c
--- vim61.orig/src/digraph.c	Sat Mar 16 12:44:47 2002
+++ vim61/src/digraph.c	Wed Jan 28 19:48:44 2004
@@ -2175,7 +2175,7 @@ putdigraph(str)
 	str = skipwhite(str);
 	if (!isdigit(*str))
 	{
-	    EMSG(_(e_number));
+	    EMSG(_(e_notnumber));
 	    return;
 	}
 	n = getdigits(&str);
diff -purN vim61.orig/src/globals.h vim61/src/globals.h
--- vim61.orig/src/globals.h	Wed Jan 28 19:47:46 2004
+++ vim61/src/globals.h	Wed Jan 28 19:48:25 2004
@@ -1114,7 +1114,7 @@ EXTERN char_u e_notread[]	INIT(=N_("Can'
 EXTERN char_u e_nowrtmsg[]	INIT(=N_("E37: No write since last change (add ! to override)"));
 EXTERN char_u e_null[]		INIT(=N_("E38: Null argument"));
 #ifdef FEAT_DIGRAPHS
-EXTERN char_u e_number[]	INIT(=N_("E39: Number expected"));
+EXTERN char_u e_notnumber[]	INIT(=N_("E39: Number expected"));
 #endif
 #ifdef FEAT_QUICKFIX
 EXTERN char_u e_openerrf[]	INIT(=N_("E40: Can't open errorfile %s"));
