--- vim72/src/option.c.orig	2008-07-18 13:05:33.000000000 +0000
+++ vim72/src/option.c	2009-10-15 20:07:21.000000000 +0000
@@ -10123,7 +10123,7 @@ langmap_set()
 {
     char_u  *p;
     char_u  *p2;
-    int	    from, to;
+    int	    from=NUL, to=NUL;
 
     langmap_init();			    /* back to one-to-one map first */
 
@@ -10141,6 +10141,11 @@ langmap_set()
 	    p2 = NULL;	    /* aAbBcCdD form, p2 is NULL */
 	while (p[0])
 	{
+	    if (p[0] == ',')
+	    {
+		++p;
+		break;
+	    }
 	    if (p[0] == '\\' && p[1] != NUL)
 		++p;
 #ifdef FEAT_MBYTE
@@ -10151,23 +10156,29 @@ langmap_set()
 	    if (p2 == NULL)
 	    {
 		mb_ptr_adv(p);
-		if (p[0] == '\\')
-		    ++p;
+		if (p[0] != ',')
+		{
+		    if (p[0] == '\\')
+			++p;
 #ifdef FEAT_MBYTE
-		to = (*mb_ptr2char)(p);
+		    to = (*mb_ptr2char)(p);
 #else
-		to = p[0];
+		    to = p[0];
 #endif
+		}
 	    }
 	    else
 	    {
-		if (p2[0] == '\\')
-		    ++p2;
+		if (p2[0] != ',')
+		{
+		    if (p2[0] == '\\')
+			++p2;
 #ifdef FEAT_MBYTE
-		to = (*mb_ptr2char)(p2);
+		    to = (*mb_ptr2char)(p2);
 #else
-		to = p2[0];
+		    to = p2[0];
 #endif
+		}
 	    }
 	    if (to == NUL)
 	    {
