--- vim64/src/xxd/xxd.c.orig	2005-01-22 14:38:33.000000000 +0000
+++ vim64/src/xxd/xxd.c	2005-12-16 12:37:27.000000000 +0000
@@ -83,12 +83,12 @@
 #if __MWERKS__ && !defined(BEBOX)
 # include <unix.h>	/* for fdopen() on MAC */
 #endif
-
 #if defined(__BORLANDC__) && __BORLANDC__ <= 0x0410 && !defined(fileno)
 /* Missing define and prototype grabbed from the BC 4.0 <stdio.h> */
 # define fileno(f)       ((f)->fd)
 FILE   _FAR *_Cdecl _FARFUNC fdopen(int __handle, char _FAR *__type);
 #endif
+#include <locale.h>
 
 
 /*  This corrects the problem of missing prototypes for certain functions
@@ -437,6 +437,11 @@
   long length = -1, n = 0, seekoff = 0;
   char l[LLEN+1];
   char *pname, *pp;
+  char *lang=getenv("LANG");
+  if(!lang) lang=getenv("LC_ALL");
+  if(!lang) lang=getenv("LC_CTYPE");
+  if(lang)
+	  setlocale(LC_ALL, lang);
 
 #ifdef AMIGA
   /* This program doesn't work when started from the Workbench */
@@ -656,7 +661,7 @@
       if (fp != stdin)
 	{
 	  fprintf(fpo, "unsigned char %s", isdigit((int)argv[1][0]) ? "__" : "");
-	  for (e = 0; (c = argv[1][e]) != 0; e++)
+	  for (e = 0; (c = (unsigned char)argv[1][e]) != 0; e++)
 	    putc(isalnum(c) ? c : '_', fpo);
 	  fputs("[] = {\n", fpo);
 	}
@@ -675,7 +680,7 @@
       if (fp != stdin)
 	{
 	  fprintf(fpo, "unsigned int %s", isdigit((int)argv[1][0]) ? "__" : "");
-	  for (e = 0; (c = argv[1][e]) != 0; e++)
+	  for (e = 0; (c = (unsigned char)argv[1][e]) != 0; e++)
 	    putc(isalnum(c) ? c : '_', fpo);
 	  fprintf(fpo, "_len = %d;\n", p);
 	}
@@ -739,7 +744,7 @@
 #ifdef __MVS__
 	  (e >= 64)
 #else
-	  (e > 31 && e < 127)
+	  ((e > 31 && e < 127) || isalnum(e))
 #endif
 	  ? e : '.';
       if (e)
