diff -ur vim73.orig/src/Makefile vim73/src/Makefile
--- vim73.orig/src/Makefile	2010-08-15 12:56:15 +0000
+++ vim73/src/Makefile	2010-10-16 21:19:54 +0000
@@ -2165,7 +2165,7 @@
 		$(GVIMNAME) $(GVIEWNAME) $(RGVIMNAME) $(RGVIEWNAME) \
 		$(GVIMDIFFNAME) $(EVIEWNAME)
 
-installmanlinks:
+installmanlinks: installvimbin
 	-$(SHELL) ./installml.sh install "$(GUI_MAN_TARGETS)" \
 		$(DEST_MAN) $(INSTALLMLARGS)
 
