diff -ur vim61.orig/src/term.c vim61/src/term.c
--- vim61.orig/src/term.c	Fri Mar 15 00:21:46 2002
+++ vim61/src/term.c	Sun Apr 21 03:13:36 2002
@@ -971,7 +971,6 @@
     {(int)KS_CWS,	IF_EB("\033[8;%d;%dt", ESC_STR "[8;%d;%dt")},
     {(int)KS_CWP,	IF_EB("\033[3;%d;%dt", ESC_STR "[3;%d;%dt")},
 #  endif
-    {(int)KS_CRV,	IF_EB("\033[>c", ESC_STR "[>c")},
     {K_UP,		IF_EB("\033OA", ESC_STR "OA")},
     {K_DOWN,		IF_EB("\033OB", ESC_STR "OB")},
     {K_RIGHT,		IF_EB("\033OC", ESC_STR "OC")},
@@ -983,14 +982,14 @@
     {K_S_LEFT,		IF_EB("\033O2D", ESC_STR "O2D")},
     {K_C_LEFT,		IF_EB("\033O5D", ESC_STR "O5D")},
     /* An extra set of function keys for vt100 mode */
-    {K_XF1,		IF_EB("\033OP", ESC_STR "OP")},
-    {K_XF2,		IF_EB("\033OQ", ESC_STR "OQ")},
-    {K_XF3,		IF_EB("\033OR", ESC_STR "OR")},
-    {K_XF4,		IF_EB("\033OS", ESC_STR "OS")},
-    {K_F1,		IF_EB("\033[11~", ESC_STR "[11~")},
-    {K_F2,		IF_EB("\033[12~", ESC_STR "[12~")},
-    {K_F3,		IF_EB("\033[13~", ESC_STR "[13~")},
-    {K_F4,		IF_EB("\033[14~", ESC_STR "[14~")},
+    {K_XF1,		IF_EB("\033[11~", ESC_STR "[11~")},
+    {K_XF2,		IF_EB("\033[12~", ESC_STR "[12~")},
+    {K_XF3,		IF_EB("\033[13~", ESC_STR "[13~")},
+    {K_XF4,		IF_EB("\033[14~", ESC_STR "[14~")},
+    {K_F1,		IF_EB("\033OP", ESC_STR "OP")},
+    {K_F2,		IF_EB("\033OQ", ESC_STR "OQ")},
+    {K_F3,		IF_EB("\033OR", ESC_STR "OR")},
+    {K_F4,		IF_EB("\033OS", ESC_STR "OS")},
     {K_F5,		IF_EB("\033[15~", ESC_STR "[15~")},
     {K_F6,		IF_EB("\033[17~", ESC_STR "[17~")},
     {K_F7,		IF_EB("\033[18~", ESC_STR "[18~")},
