--- vim-5.6a/runtime/doc/vim2html.pl.paths	Thu Oct  7 09:42:56 1999
+++ vim-5.6a/runtime/doc/vim2html.pl	Tue Jan 11 17:37:17 2000
@@ -1,4 +1,4 @@
-#!/usr/bin/env perl
+#!/usr/bin/perl
 
 # converts vim documentation to simple html
 # Sirtaj Singh Kang (taj@kde.org)
--- vim-5.6a/runtime/tools/shtags.pl.paths	Thu Oct  7 09:49:18 1999
+++ vim-5.6a/runtime/tools/shtags.pl	Tue Jan 11 17:37:36 2000
@@ -1,4 +1,4 @@
-#!/usr/bin/env perl
+#!/usr/bin/perl
 #
 # shtags: create a tags file for perl scripts
 #
--- vim-5.6a/runtime/tools/pltags.pl.paths	Sun Nov 28 13:01:57 1999
+++ vim-5.6a/runtime/tools/pltags.pl	Tue Jan 11 17:37:33 2000
@@ -1,4 +1,4 @@
-#!/usr/bin/env perl
+#!/usr/bin/perl
 
 # pltags - create a tags file for Perl code, for use by vi(m)
 #
--- vim-5.6a/runtime/tools/efm_filter.pl.paths	Thu Oct  7 09:48:42 1999
+++ vim-5.6a/runtime/tools/efm_filter.pl	Tue Jan 11 17:37:30 2000
@@ -1,4 +1,4 @@
-#!/usr/bin/env perl
+#!/usr/bin/perl
 #
 # This program works as a filter that reads from stdin, copies to
 # stdout *and* creates an error file that can be read by vim.
