--- vim60r/src/term.c.crv	Wed Jan  3 13:37:24 2001
+++ vim60r/src/term.c	Wed Jan  3 13:37:34 2001
@@ -940,7 +940,6 @@
     {(int)KS_CWS,	IF_EB("\033[8;%d;%dt", ESC_STR "[8;%d;%dt")},
     {(int)KS_CWP,	IF_EB("\033[3;%d;%dt", ESC_STR "[3;%d;%dt")},
 #  endif
-    {(int)KS_CRV,	IF_EB("\033[>c", ESC_STR "[>c")},
     {K_UP,		IF_EB("\033OA", ESC_STR "OA")},
     {K_DOWN,		IF_EB("\033OB", ESC_STR "OB")},
     {K_RIGHT,		IF_EB("\033OC", ESC_STR "OC")},
