--- vim60b/runtime/syntax/spec.vim.highlite	Sun Jul 16 16:57:23 2000
+++ vim60b/runtime/syntax/spec.vim	Sun Jul 16 22:20:24 2000
@@ -28,12 +28,12 @@
 
 
 "PreAmble
-syn region specPreAmble matchgroup=specCommand start='^\(Prereq\|Summary\|Name\|Version\|Packager\|Requires\|Icon\|URL\|Source\d*\|Patch\d*\|Prefix\|Packager\|Group\|Copyright\|License\|Release\|BuildRoot\|Distribution\|Vendor\|Provides\|ExclusiveArch\|ExclusiveOS\|Serial\|Obsoletes\|BuildArch\|BuildRequires\|BuildConflicts\|BuildPreReq\|AutoRequires\|AutoReq\)' end='$' contains=specEmail,specURL,specSpecial,specColon oneline
+syn region specPreAmble matchgroup=specCommand start='^\(Prereq\|Summary\|Name\|Version\|Packager\|Requires\|Icon\|URL\|Source\d*\|Patch\d*\|Prefix\|Packager\|Group\|Copyright\|License\|Release\|BuildRoot\|Distribution\|Vendor\|Provides\|ExclusiveArch\|ExclusiveOS\|Serial\|Obsoletes\|BuildArch\|BuildRequires\|BuildConflicts\|BuildPreReq\|AutoRequires\|AutoReq\|Epoch\|ExcludeArch\|ExcludeOS\)' end='$' contains=specEmail,specURL,specSpecial,specColon oneline
 
 syn region specPackage matchgroup=specMacro start='%package' end='^%'me=e-1 contains=specPreAmble
 " sections that contain a lot of scripts
 syn region specScriptMacro  matchgroup=specMacro start='%setup\|%install\|%postun\|%post\|%prep\|%preun\|%pre\($\|\s\)\|%build\|%clean' end='^%'me=e-1 contains=specSpecialVariables,@specCommands,specDeref,shDo,shFor,shCaseEsac,specNoNumberHilite,specCommandOpts,specComment,shIf,specSpecialChar,specBabyMacro
-syn match specBabyMacro '%else\|%configure\|%GNUconfigure'
+syn match specBabyMacro '%else\|%configure\|%GNUconfigure\|%makeinstall'
 
 syn case match
 
@@ -41,7 +41,7 @@
 syn match specWeekday 'Mon\|Tue\|Wed\|Thu\|Fri\|Sat\|\<Sun\>'
 syn match specMonth 'January\|February\|March\|April\|May\|June\|July\|August\|September\|October\|November\|December\|Jan\|Feb\|Mar\|Apr\|Jun\|Jul\|Aug\|Sep\|Oct\|Nov\|Dec'
 syn match specNumber +^-\=[0-9.-]*[0-9]\|[ \t]-\=[0-9.-]*[0-9]\|-[0-9.-]*[0-9]+
-syn match specListedFilesPrefix		contained '%{prefix}/\|/usr/\|/local/\|/opt/\|/X11R6/\|/X11/'me=e-1
+syn match specListedFilesPrefix		contained '%{prefix}/\|%{_prefix}/\|%{_mandir}/\|%{_infodir}/\|/usr/\|/local/\|/opt/\|/X11R6/\|/X11/'me=e-1
 syn match specListedFilesBin		contained '/bin/\|/sbin/'me=e-1
 syn match speclistedFilesLib		contained '/lib/\|/include/'me=e-1
 syn match specListedFilesDoc		contained '/man\d*\|/doc\|/info'
@@ -60,13 +60,13 @@
 syn match specCommandOpts '[ \t]--\=[a-zA-Z-]\+'
 
 syn match specConfigure '\./configure'
-syn match specTarCommand 'tar[ \t]*[xvzf]\{,5}[ \t]'
+syn match specTarCommand 'tar[ \t]*[xvzIf]\{,5}[ \t]'
 syn cluster specCommands contains=specCommand,specTarCommand,specConfigureCommand,specCommandSpecial
 
 syn keyword specSpecialVariables	RPM_BUILD_ROOT RPM_SOURCE_DIR RPM_OPT_FLAGS LDFLAGS CC CC_FLAGS CPPNAME CFLAGS CXX CXXFLAGS CPPFLAGS
 
-syn keyword specCommand	make xmkmf mkdir  chmod ln find sed rm strip moc echo grep ls rm mv mkdir install cp pwd cat tail then else elif cd gzip rmdir ln
-syn keyword specSpecial		GPL Makefile  LGPL
+syn keyword specCommand	make xmkmf mkdir chmod ln find sed rm strip moc echo grep ls rm mv mkdir install cp pwd cat tail then else elif cd gzip rmdir ln bzip2
+syn keyword specSpecial		GPL Makefile LGPL BSD
 syn keyword specCommandSpecial root
 
 syn region specComment start='#' end='$'
