diff -upk.orig m4-1.4.4.orig/src/builtin.c m4-1.4.4/src/builtin.c
--- m4-1.4.4.orig/src/builtin.c	2005-05-01 11:40:47 +0000
+++ m4-1.4.4/src/builtin.c	2005-10-22 00:07:25 +0000
@@ -231,6 +231,7 @@ builtin_init (void)
 
   for (bp = &builtin_tab[0]; bp->name != NULL; bp++)
     if (!no_gnu_extensions || !bp->gnu_extension)
+    {
       if (prefix_all_builtins)
 	{
 	  string = (char *) xmalloc (strlen (bp->name) + 4);
@@ -241,6 +242,7 @@ builtin_init (void)
 	}
       else
 	define_builtin (bp->name, bp, SYMBOL_INSERT, FALSE);
+    }
 
   for (pp = &predefined_tab[0]; pp->func != NULL; pp++)
     if (no_gnu_extensions)
diff -upk.orig m4-1.4.4.orig/src/path.c m4-1.4.4/src/path.c
--- m4-1.4.4.orig/src/path.c	2005-05-01 11:36:11 +0000
+++ m4-1.4.4/src/path.c	2005-10-22 00:05:54 +0000
@@ -134,7 +134,7 @@ path_search (const char *dir)
       if (fp != NULL)
 	{
 	  if (debug_level & DEBUG_TRACE_PATH)
-	    DEBUG_MESSAGE (("path search for `%s' found `%s'", dir, name));
+	    DEBUG_MESSAGE2 ("path search for `%s' found `%s'", dir, name);
 	  break;
 	}
     }
