diff -puNr perl-5.8.8.orig/lib/CPAN/FirstTime.pm perl-5.8.8/lib/CPAN/FirstTime.pm
--- perl-5.8.8.orig/lib/CPAN/FirstTime.pm	2006-01-30 21:08:57 +0000
+++ perl-5.8.8/lib/CPAN/FirstTime.pm	2006-05-05 07:09:45 +0000
@@ -292,14 +292,15 @@ by ENTER.
     my(@path) = split /$Config{'path_sep'}/, $ENV{'PATH'};
     local $^W = $old_warn;
     my $progname;
-    for $progname (qw/gzip tar unzip make lynx wget ncftpget ncftp ftp gpg/){
+    for $progname (qw/gzip tar unzip make links lynx wget lftpget lftp ncftpget ncftp ftp gpg/){
       if ($^O eq 'MacOS') {
           $CPAN::Config->{$progname} = 'not_here';
           next;
       }
       my $progcall = $progname;
-      # we don't need ncftp if we have ncftpget
-      next if $progname eq "ncftp" && $CPAN::Config->{ncftpget} gt " ";
+      # we don't need *ftp if we have *ftpget
+      next if (($progname eq "lftp") || ($progname eq "ncftp")) &&
+          (($CPAN::Config->{lftpget} gt " ") || ($CPAN::Config->{ncftpget} gt " "));
       my $path = $CPAN::Config->{$progname} 
 	  || $Config::Config{$progname}
 	      || "";
diff -puNr perl-5.8.8.orig/lib/CPAN.pm perl-5.8.8/lib/CPAN.pm
--- perl-5.8.8.orig/lib/CPAN.pm	2006-05-05 05:26:27 +0000
+++ perl-5.8.8/lib/CPAN.pm	2006-05-05 07:05:09 +0000
@@ -2604,7 +2604,7 @@ sub hosthard {
 
 	$self->debug("localizing funkyftpwise[$url]") if $CPAN::DEBUG;
 	my($f,$funkyftp);
-	for $f ('lynx','ncftpget','ncftp','wget') {
+	for $f ('links', 'lynx', 'lftpget', 'lftp', 'ncftpget', 'ncftp', 'wget') {
 	  next unless exists $CPAN::Config->{$f};
 	  $funkyftp = $CPAN::Config->{$f};
 	  next unless defined $funkyftp;
@@ -2613,16 +2613,16 @@ sub hosthard {
 	  ($asl_ungz = $aslocal) =~ s/\.gz//;
           $asl_gz = "$asl_ungz.gz";
 	  my($src_switch) = "";
-	  if ($f eq "lynx"){
+	  if (($f eq "links") || ($f eq "lynx")){
 	    $src_switch = " -source";
-	  } elsif ($f eq "ncftp"){
+	  } elsif (($f eq "lftp") || ($f eq "ncftp")){
 	    $src_switch = " -c";
           } elsif ($f eq "wget"){
               $src_switch = " -O -";
 	  }
 	  my($chdir) = "";
 	  my($stdout_redir) = " > $asl_ungz";
-	  if ($f eq "ncftpget"){
+	  if (($f eq "lftpget") || ($f eq "ncftpget")){
 	    $chdir = "cd $aslocal_dir && ";
 	    $stdout_redir = "";
 	  }
