--- perl-5.8.8/lib/Net/hostent.t.orig	2003-03-30 10:18:26 +0000
+++ perl-5.8.8/lib/Net/hostent.t	2007-03-21 15:27:50 +0000
@@ -17,6 +17,9 @@ BEGIN {
     if ($^O eq 'MacOS' || ($^O eq 'irix' && $Config{osvers} == 5)) {
 	plan skip_all => "Test relies on resolution of localhost, fails on $^O ($Config{osvers})";
     }
+    if (defined($ENV{"IN_RPM"})) {
+	plan skip_all => "Test disabled for in RPM build";
+    }
 }
 
 use Test::More tests => 7;
