diff -urp perl-5.6.0.orig/pp_sys.c perl-5.6.0/pp_sys.c
--- perl-5.6.0.orig/pp_sys.c	Sun Mar 19 07:18:10 2000
+++ perl-5.6.0/pp_sys.c	Fri Jul  4 15:40:05 2003
@@ -4772,7 +4772,7 @@ PP(pp_gpwent)
     register SV *sv;
     struct passwd *pwent;
     STRLEN n_a;
-#if defined(HAS_GETSPENT) || defined(HAS_GETSPNAM)
+#ifdef HAS_GETSPNAM
     struct spwd *spwent = NULL;
 #endif
 
@@ -4788,20 +4788,8 @@ PP(pp_gpwent)
 #endif
 
 #ifdef HAS_GETSPNAM
-    if (which == OP_GPWNAM) {
-	if (pwent)
-	    spwent = getspnam(pwent->pw_name);
-    }
-#  ifdef HAS_GETSPUID /* AFAIK there isn't any anywhere. --jhi */ 
-    else if (which == OP_GPWUID) {
-	if (pwent)
-	    spwent = getspnam(pwent->pw_name);
-    }
-#  endif
-#  ifdef HAS_GETSPENT
-    else
-	spwent = (struct spwd *)getspent();
-#  endif
+    if (pwent)
+	spwent = getspnam(pwent->pw_name);
 #endif
 
     EXTEND(SP, 10);
@@ -4826,7 +4814,7 @@ PP(pp_gpwent)
 
 	PUSHs(sv = sv_mortalcopy(&PL_sv_no));
 #ifdef PWPASSWD
-#   if defined(HAS_GETSPENT) || defined(HAS_GETSPNAM)
+#   ifdef HAS_GETSPNAM
       if (spwent)
               sv_setpv(sv, spwent->sp_pwdp);
       else
