--- perl-5.6.0/lib/ExtUtils/MM_Unix.pm.instman	Thu Mar  2 18:52:52 2000
+++ perl-5.6.0/lib/ExtUtils/MM_Unix.pm	Thu Mar 23 17:32:54 2000
@@ -1757,6 +1757,12 @@
     # you to build directly into, say $Config::Config{privlibexp}.
     unless ($self->{INST_LIB}){
 
+    for $install_variable (qw/
+			   INSTALLMAN1DIR
+			   INSTALLMAN3DIR
+			   /) {
+	$self->prefixify($install_variable,$configure_prefix,qq[\$\(PREFIX\)]);
+    }
 
 	##### XXXXX We have to change this nonsense
 
