--- dialog-0.9a-20010527/inputstr.c.orig	Sat Oct 28 04:04:54 2000
+++ dialog-0.9a-20010527/inputstr.c	Sat Jun  9 15:31:08 2001
@@ -87,28 +87,21 @@
 dlg_show_string(WINDOW *win, char *string, int offset, chtype attr,
 		int y_base, int x_base, int x_last, bool hidden, bool force)
 {
-    if (hidden) {
-	if (force) {
-	    (void) wmove(win, y_base, x_base);
-	    wrefresh_lock(win);
-	}
-    } else {
-	int i, input_x;
-	int len = strlen(string);
-	int scrollamt = (offset + 1 - x_last);
+     int i, input_x;
+     int len = strlen(string);
+     int scrollamt = (offset + 1 - x_last);
 
-	if (scrollamt < 0)
-	    scrollamt = 0;
-	input_x = offset - scrollamt;
+     if (scrollamt < 0)
+         scrollamt = 0;
+     input_x = offset - scrollamt;
 
-	wattrset(win, attr);
-	(void) wmove(win, y_base, x_base);
-	for (i = 0; i < x_last; i++)
-	    (void) waddch(win,
-			  (i + scrollamt) < len
-			  ? CharOf(string[scrollamt + i])
-			  : ' ');
-	(void) wmove(win, y_base, x_base + input_x);
-	wrefresh_lock(win);
-    }
+     wattrset(win, attr);
+     (void) wmove(win, y_base, x_base);
+     for (i = 0; i < x_last; i++)
+          (void) waddch(win,
+	                    (i + scrollamt) < len
+	                    ? ( hidden ? '*' : CharOf(string[scrollamt + i]))
+			    : ' ');
+     (void) wmove(win, y_base, x_base + input_x);
+     wrefresh_lock(win);
 }
