--- dialog-0.9b-20020814/inputstr.c.orig	Sat Jun 15 03:24:45 2002
+++ dialog-0.9b-20020814/inputstr.c	Fri Jan 10 00:42:56 2003
@@ -111,28 +111,21 @@
 dlg_show_string(WINDOW *win, char *string, int offset, chtype attr,
 		int y_base, int x_base, int x_last, bool hidden, bool force)
 {
-    if (hidden) {
-	if (force) {
-	    (void) wmove(win, y_base, x_base);
-	    wrefresh(win);
-	}
-    } else {
-	int i;
-	int input_x = dlg_edit_offset(offset, x_last);
-	int len = strlen(string);
-	int scrollamt = (offset + 1 - x_last);
+    int i;
+    int input_x = dlg_edit_offset(offset, x_last);
+    int len = strlen(string);
+    int scrollamt = (offset + 1 - x_last);
 
-	if (scrollamt < 0)
-	    scrollamt = 0;
+    if (scrollamt < 0)
+        scrollamt = 0;
 
-	wattrset(win, attr);
-	(void) wmove(win, y_base, x_base);
-	for (i = 0; i < x_last; i++)
-	    (void) waddch(win,
-			  (i + scrollamt) < len
-			  ? CharOf(string[scrollamt + i])
-			  : ' ');
-	(void) wmove(win, y_base, x_base + input_x);
-	wrefresh(win);
-    }
+    wattrset(win, attr);
+    (void) wmove(win, y_base, x_base);
+    for (i = 0; i < x_last; i++)
+        (void) waddch(win,
+		      (i + scrollamt) < len
+		      ? ( hidden ? '*' : CharOf(string[scrollamt + i]))
+		      : ' ');
+    (void) wmove(win, y_base, x_base + input_x);
+    wrefresh(win);
 }
