diff -ur newt-0.50.18.orig/form.c newt-0.50.18/form.c
--- newt-0.50.18.orig/form.c	Fri Apr 28 19:35:55 2000
+++ newt-0.50.18/form.c	Fri May 18 08:55:00 2001
@@ -147,12 +147,13 @@
 
 static int Gpm_Open(Gpm_Connect *conn, int flag)
 {
-  char tty[32];
   char *term;
   int i;
   struct sockaddr_un addr;
   Gpm_Stst *new;
+#ifndef SO_PEERCRED
   char* sock_name = 0;
+#endif
 
   /*....................................... First of all, check xterm */
 
@@ -194,21 +195,16 @@
       if (flag>0)
         {  /* forced vc number */
           conn->vc=flag;
-          sprintf(tty,"/dev/tty%i",flag);
         }
       else if (flag==0)  /* use your current vc */
         {
           char *t = ttyname(0); /* stdin */
           if (!t) t = ttyname(1); /* stdout */
           if (!t) goto err;
-          strcpy(tty,t);
-          if (strncmp(tty,"/dev/tty",8) || !isdigit(tty[8]))
+          if (strncmp(t,"/dev/tty",8) || !isdigit(t[8]))
             goto err;
-          conn->vc=atoi(tty+8);
+          conn->vc=atoi(t+8);
         }
-      else /* a default handler -- use console */
-        sprintf(tty,"/dev/tty0");
-
     }
 
   new->info=*conn;
@@ -223,6 +219,7 @@
           goto err;
         }
 
+#ifndef SO_PEERCRED
       bzero((char *)&addr,sizeof(addr));
       addr.sun_family=AF_UNIX;
       if (!(sock_name = tempnam (0, "gpm"))) {
@@ -233,6 +230,7 @@
                 sizeof (addr.sun_family) + strlen (addr.sun_path))==-1) {
         goto err;
       } /*if*/
+#endif
 
       bzero((char *)&addr,sizeof(addr));
       addr.sun_family=AF_UNIX;
@@ -299,11 +297,13 @@
     }
   while(gpm_stack);
   if (gpm_fd>=0) close(gpm_fd);
+#ifndef SO_PEERCRED
   if (sock_name) {
     unlink(sock_name);
     free(sock_name);
     sock_name = 0;
   } /*if*/
+#endif
   gpm_flag=0;
   return -1;
 }
