--- newt-0.50.18/form.c.strcpy	Fri Dec 15 21:06:40 2000
+++ newt-0.50.18/form.c	Fri Dec 15 21:07:09 2000
@@ -201,7 +201,8 @@
           char *t = ttyname(0); /* stdin */
           if (!t) t = ttyname(1); /* stdout */
           if (!t) goto err;
-          strcpy(tty,t);
+	  tty[0]='\0';
+          strncat(tty,t,sizeof(tty));
           if (strncmp(tty,"/dev/tty",8) || !isdigit(tty[8]))
             goto err;
           conn->vc=atoi(tty+8);
