--- quota-2.00/edquota.c.orig	Mon Sep 18 09:38:41 2000
+++ quota-2.00/edquota.c	Sun Jul  1 18:38:59 2001
@@ -56,6 +56,7 @@
 #include <grp.h>
 #include <ctype.h>
 #include <stdio.h>
+#include <stdlib.h>
 #include <string.h>
 #include <signal.h>
 #include <unistd.h>
@@ -65,7 +66,7 @@
 #include "pot.h"
 
 static char *qfextension[] = INITQFNAMES;
-static char tmpfil[] = _PATH_TMP "EdP.aXXXXXX";
+static char tmppat[] = "/EdP.aXXXXXX";
 
 #define   FOUND   0x01
 
@@ -99,7 +100,7 @@
    extern int optind;
    long id, protoid;
    int quotatype, tmpfd;
-   char *protoname, ch;
+   char *protoname, ch, *tmpfil;
    int tflag = 0, pflag = 0, nflag = 0;
     
    gettexton();
@@ -170,8 +171,16 @@
       exit(0);
    }
 
+
+   tmpfil = strdup(__secure_getenv("TMPDIR") ?: "/tmp");
+   tmpfil = realloc(tmpfil, strlen(tmpfil) + strlen(tmppat) + 1);
+   strcat(tmpfil, tmppat);
+   
    umask(077);
-   tmpfd = mkstemp(tmpfil);
+   if ((tmpfd = mkstemp(tmpfil)) < 0) {
+      perror("mkstemp");
+      exit(1);
+   }
    fchown(tmpfd, getuid(), getgid());
    if (tflag) {
       protoprivs = getprivs(0, quotatype, (nflag == 0));
