--- sh-utils-2.0/lib/getdate.c.cest	Wed Jan  5 08:18:31 2000
+++ sh-utils-2.0/lib/getdate.c	Wed Jan  5 08:19:34 2000
@@ -1612,6 +1612,7 @@
     { "nt",	tZONE,     HOUR (11) },	/* Nome */
     { "idlw",	tZONE,     HOUR (12) },	/* International Date Line West */
     { "cet",	tZONE,     -HOUR (1) },	/* Central European */
+    { "cest",	tDAYZONE,  -HOUR (1) },	/* Central European Summer */
     { "met",	tZONE,     -HOUR (1) },	/* Middle European */
     { "mewt",	tZONE,     -HOUR (1) },	/* Middle European Winter */
     { "mest",	tDAYZONE,  -HOUR (1) },	/* Middle European Summer */
