diff -upk.orig initscripts-5.00.orig/rc.d/init.d/netfs initscripts-5.00/rc.d/init.d/netfs
--- initscripts-5.00.orig/rc.d/init.d/netfs	1999-09-20 17:45:22 +0000
+++ initscripts-5.00/rc.d/init.d/netfs	2005-12-31 01:13:14 +0000
@@ -20,7 +20,7 @@ fi
 . /etc/sysconfig/network
 
 # Check that networking is up.
-[ ${NETWORKING} = "no" ] && exit 0
+[ "$NETWORKING" = "no" ] && exit 0
 
 NFSFSTAB=`grep -v '^#' /etc/fstab | awk '{ if ($3 ~ /^nfs$/ && $4 !~ /noauto/) print $2}'`
 SMBFSTAB=`grep -v '^#' /etc/fstab | awk '{ if ($3 ~ /^smbfs$/ && $4 !~ /noauto/) print $2}'`
diff -upk.orig initscripts-5.00.orig/rc.d/init.d/network initscripts-5.00/rc.d/init.d/network
--- initscripts-5.00.orig/rc.d/init.d/network	2005-12-31 01:11:28 +0000
+++ initscripts-5.00/rc.d/init.d/network	2005-12-31 01:13:14 +0000
@@ -22,7 +22,7 @@ fi
 
 
 # Check that networking is up.
-[ ${NETWORKING} = "no" ] && exit 0
+[ "$NETWORKING" = "no" ] && exit 0
 
 [ -x /sbin/ifconfig ] || exit 0
 
