--- db-4.0.14/hash/hash_dup.c.orig	Thu Aug 16 17:21:10 2001
+++ db-4.0.14/hash/hash_dup.c	Wed Mar  3 00:42:03 2004
@@ -613,6 +613,11 @@
 	 * __ham_del_pair decremented nelem.  This is incorrect;  we
 	 * manually copied the element elsewhere, so the total number
 	 * of elements hasn't changed.  Increment it again.
+	 *
+	 * !!!
+	 * Note that we still have the metadata page pinned, and
+	 * __ham_del_pair dirtied it, so we don't need to set the dirty
+	 * flag again.
 	 */
 	if (!STD_LOCKING(dbc))
 		hcp->hdr->nelem++;
--- db-4.0.14/hash/hash_page.c.orig	Thu Oct 25 19:01:12 2001
+++ db-4.0.14/hash/hash_page.c	Wed Mar  3 00:42:03 2004
@@ -643,8 +643,11 @@
 	 * XXX
 	 * Perhaps we can retain incremental numbers and apply them later.
 	 */
-	if (!STD_LOCKING(dbc))
+	if (!STD_LOCKING(dbc)) {
 		--hcp->hdr->nelem;
+		if ((ret = __ham_dirty_meta(dbc)) != 0)
+			return (ret);
+	}
 
 	/*
 	 * If we need to reclaim the page, then check if the page is empty.
@@ -1414,8 +1417,11 @@
 	 * XXX
 	 * Maybe keep incremental numbers here.
 	 */
-	if (!STD_LOCKING(dbc))
+	if (!STD_LOCKING(dbc)) {
 		hcp->hdr->nelem++;
+		if ((ret = __ham_dirty_meta(dbc)) != 0)
+			return (ret);
+	}
 
 	if (do_expand || (hcp->hdr->ffactor != 0 &&
 	    (u_int32_t)H_NUMPAIRS(hcp->page) > hcp->hdr->ffactor))
