--- db-4.0.14/mp/mp_bh.c.orig	Sat Jul 28 21:05:21 2001
+++ db-4.0.14/mp/mp_bh.c	Wed Mar  3 00:25:46 2004
@@ -347,7 +347,7 @@
 	 * already, unless it was modified by the application *after* the sync
 	 * was scheduled.
 	 */
-	if (LOGGING_ON(dbenv) && mfp->lsn_off != -1 &&
+	if (LOGGING_ON(dbenv) && !IS_RECOVERING(dbenv) && mfp->lsn_off != -1 &&
 	    (!F_ISSET(bhp, BH_SYNC) || F_ISSET(bhp, BH_SYNC_LOGFLSH))) {
 		memcpy(&lsn, bhp->buf + mfp->lsn_off, sizeof(DB_LSN));
 		if ((ret = dbenv->log_flush(dbenv, &lsn)) != 0)
