--- db-4.3.29.orig/dist/configure.ac	2004-12-23 02:58:22 +0000
+++ db-4.3.29/dist/configure.ac	2005-10-07 19:36:38 +0000
@@ -395,7 +395,9 @@ if test "$db_cv_java" = "yes"; then
 
 	for JNI_INCLUDE_DIR in $JNI_INCLUDE_DIRS
 	do
-		CPPFLAGS="$CPPFLAGS -I$JNI_INCLUDE_DIR"
+		if test "$JNI_INCLUDE_DIR" != "/usr/include" -a "$JNI_INCLUDE_DIR" != "/usr/include/linux"; then
+			CPPFLAGS="$CPPFLAGS -I$JNI_INCLUDE_DIR"
+		fi
 	done
 
 	ADDITIONAL_LANG="$ADDITIONAL_LANG java"
--- db-4.3.29.orig/dist/configure	2005-09-06 07:44:24 +0000
+++ db-4.3.29/dist/configure	2005-10-13 10:44:12 +0000
@@ -21653,7 +21653,9 @@ echo "$as_me: error: Java version 1.3 or
 
 	for JNI_INCLUDE_DIR in $JNI_INCLUDE_DIRS
 	do
-		CPPFLAGS="$CPPFLAGS -I$JNI_INCLUDE_DIR"
+		if test "$JNI_INCLUDE_DIR" != "/usr/include" -a "$JNI_INCLUDE_DIR" != "/usr/include/linux"; then
+			CPPFLAGS="$CPPFLAGS -I$JNI_INCLUDE_DIR"
+		fi
 	done
 
 	ADDITIONAL_LANG="$ADDITIONAL_LANG java"
