--- db-4.2.52/dist/aclocal_java/ac_prog_jar.ac.fastjar	2001-08-23 12:58:43.000000000 -0400
+++ db-4.2.52/dist/aclocal_java/ac_prog_jar.ac	2004-11-29 18:06:27.299424360 -0500
@@ -27,9 +27,9 @@
 AC_DEFUN([AC_PROG_JAR],[
 AC_REQUIRE([AC_EXEEXT])dnl
 if test "x$JAVAPREFIX" = x; then
-        test "x$JAR" = x && AC_CHECK_PROGS(JAR, jar$EXEEXT)
+        test "x$JAR" = x && AC_CHECK_PROGS(JAR, fastjar$EXEEXT jar$EXEEXT)
 else
-        test "x$JAR" = x && AC_CHECK_PROGS(JAR, jar, $JAVAPREFIX)
+        test "x$JAR" = x && AC_CHECK_PROGS(JAR, fastjar jar, $JAVAPREFIX)
 fi
 test "x$JAR" = x && AC_MSG_ERROR([no acceptable jar program found in \$PATH])
 AC_PROVIDE([$0])dnl
