--- db-4.2.52/dist/configure.orig	2003-12-04 00:10:02 +0300
+++ db-4.2.52/dist/configure	2004-02-12 16:04:06 +0300
@@ -20914,7 +20914,9 @@
 
 	for JNI_INCLUDE_DIR in $JNI_INCLUDE_DIRS
 	do
-		CPPFLAGS="$CPPFLAGS -I$JNI_INCLUDE_DIR"
+		if test "$JNI_INCLUDE_DIR" != "/usr/include" -a "$JNI_INCLUDE_DIR" != "/usr/include/linux"; then
+			CPPFLAGS="$CPPFLAGS -I$JNI_INCLUDE_DIR"
+		fi
 	done
 
 	ADDITIONAL_LANG="$ADDITIONAL_LANG java"
