diff -urp postfix-2.4.14.orig/src/global/mail_version.h postfix-2.4.14/src/global/mail_version.h
--- postfix-2.4.14.orig/src/global/mail_version.h	2010-03-22 23:47:21 +0000
+++ postfix-2.4.14/src/global/mail_version.h	2010-07-28 17:08:19 +0000
@@ -20,7 +20,7 @@
   * Patches change both the patchlevel and the release date. Snapshots have no
   * patchlevel; they change the release date only.
   */
-#define MAIL_RELEASE_DATE	"20090322"
+#define MAIL_RELEASE_DATE	"20100322"
 #define MAIL_VERSION_NUMBER	"2.4.14"
 
 #ifdef SNAPSHOT
