--- postfix/conf/postfix-script
+++ postfix/conf/postfix-script
@@ -223,11 +223,11 @@ check-warn)
 		-exec $WARN not owned by $mail_owner: {} \;
 
 	find $queue_directory/public $queue_directory/maildrop \
-	    $command_directory/postqueue $command_directory/postdrop \
+	    $daemon_directory/postqueuedir/postqueue $command_directory/postdrop \
 	    -prune ! -group $setgid_group \
 	    -exec $WARN not owned by group $setgid_group: {} \;
 
-	find $command_directory/postqueue $command_directory/postdrop \
+	find $daemon_directory/postqueuedir/postqueue $command_directory/postdrop \
 	    -prune ! -perm -02111 \
 	    -exec $WARN not set-gid or not owner+group+world executable: {} \;
 
@@ -248,10 +248,15 @@ check-warn)
 	do
 		test -d $dir && find $dir -type f -print | while read path
 		do
-			test -f /$path && {
-			    cmp -s $path /$path || 
-				$WARN $queue_directory/$path and /$path differ
-			}
+			if [ -e "/$path" ]; then
+				cmp -s "$path" "/$path" ||
+				    $WARN "$queue_directory/$path and /$path differ"
+			elif [ -e "/usr/$path" ]; then
+				cmp -s "$path" "/usr/$path" ||
+				    $WARN "$queue_directory/$path and /usr/$path differ"
+			else
+				$WARN "$queue_directory/$path has no original source"
+			fi
 		done
 	done
 
