--- postfix/proto/postconf.proto
+++ postfix/proto/postconf.proto
@@ -2438,6 +2438,11 @@ mailbox_command = /some/where/maildrop -d "$USER"
         -f "$SENDER" "$EXTENSION"
 </pre>
 
+%PARAM mailbox_unpriv_delivery yes
+
+<p> If enabled, then local(8) will not apply root privileges
+for mailbox delivery.  </p>
+
 %PARAM mailbox_size_limit 51200000
 
 <p> The maximal size of any local(8) individual mailbox or maildir
--- postfix/src/global/mail_params.h
+++ postfix/src/global/mail_params.h
@@ -486,6 +486,10 @@ extern char *var_mail_spool_dir;
 #define DEF_HOME_MAILBOX	""
 extern char *var_home_mailbox;
 
+#define VAR_MAILBOX_UNPRIV_DELIVERY	"mailbox_unpriv_delivery"
+#define DEF_MAILBOX_UNPRIV_DELIVERY	1
+bool	var_mailbox_unpriv_delivery;
+
 #define VAR_MAILBOX_COMMAND	"mailbox_command"
 #define DEF_MAILBOX_COMMAND	""
 extern char *var_mailbox_command;
--- postfix/src/local/local.c
+++ postfix/src/local/local.c
@@ -616,6 +616,7 @@ char   *var_allow_commands;
 char   *var_alias_maps;
 int     var_dup_filter_limit;
 char   *var_home_mailbox;
+bool    var_mailbox_unpriv_delivery;
 char   *var_mailbox_command;
 char   *var_mailbox_cmd_maps;
 char   *var_rcpt_fdelim;
@@ -886,6 +887,7 @@ int     main(int argc, char **argv)
 	VAR_STAT_HOME_DIR, DEF_STAT_HOME_DIR, &var_stat_home_dir,
 	VAR_MAILTOOL_COMPAT, DEF_MAILTOOL_COMPAT, &var_mailtool_compat,
 	VAR_FROZEN_DELIVERED, DEF_FROZEN_DELIVERED, &var_frozen_delivered,
+	VAR_MAILBOX_UNPRIV_DELIVERY, DEF_MAILBOX_UNPRIV_DELIVERY, &var_mailbox_unpriv_delivery,
 	0,
     };
 
--- postfix/src/local/mailbox.c
+++ postfix/src/local/mailbox.c
@@ -144,8 +144,9 @@ static int deliver_mailbox_file(LOCAL_STATE state, USER_ATTR usr_attr)
      * Laboratory, USA).
      * 
      * If delivering to the mail spool directory, and the spool directory is
-     * world-writable, deliver as the recipient; if the spool directory is
-     * group-writable, use the recipient user id and the mail spool group id.
+     * world-writable or var_mailbox_unpriv_delivery is set, then deliver as
+     * the recipient; if the spool directory is group-writable, use the
+     * recipient user id and the mail spool group id.
      * 
      * Otherwise, use root privileges and chown the mailbox.
      */
@@ -154,6 +155,10 @@ static int deliver_mailbox_file(LOCAL_STATE state, USER_ATTR usr_attr)
 	|| (st.st_mode & S_IWOTH) != 0) {
 	spool_uid = usr_attr.uid;
 	spool_gid = usr_attr.gid;
+    } else if (var_mailbox_unpriv_delivery) {
+	spool_uid = usr_attr.uid;
+	spool_gid = usr_attr.gid;
+	local_mbox_lock_mask |= MBOX_DOT_LOCK_MAY_FAIL;
     } else if ((st.st_mode & S_IWGRP) != 0) {
 	spool_uid = usr_attr.uid;
 	spool_gid = st.st_gid;
