diff -ur postfix-19991231-pl10.orig/conf/main.cf postfix-19991231-pl10/conf/main.cf
--- postfix-19991231-pl10.orig/conf/main.cf	Thu Jan  6 17:52:45 2000
+++ postfix-19991231-pl10/conf/main.cf	Wed Nov 22 14:05:50 2000
@@ -311,13 +311,11 @@
 # to distinguish local clients from strangers. See permit_mynetworks
 # and smtpd_recipient_restrictions in the file sample-smtpd.cf file.
 #
-# The default is a list of all networks attached to the machine:  a
-# complete class A network (X.0.0.0/8), a complete class B network
-# (X.X.0.0/16), and so on. If you want stricter control, specify a
-# list of network/mask patterns, where the mask specifies the number
-# of bits in the network part of a host address. You can also specify
-# the absolute pathname of a pattern file instead of listing the
-# patterns here.
+# The default is a list of all interface addresses on the machine.
+# If you want to allow relaying, specify a list of network/mask
+# patterns, where the mask specifies the number of bits in the
+# network part of a host address.  You can also specify the absolute
+# pathname of a pattern file instead of listing the patterns here.
 #
 #mynetworks = 168.100.189.0/28, 127.0.0.0/8
 #mynetworks = $config_directory/mynetworks
diff -ur postfix-19991231-pl10.orig/conf/sample-misc.cf postfix-19991231-pl10/conf/sample-misc.cf
--- postfix-19991231-pl10.orig/conf/sample-misc.cf	Thu Sep 23 22:10:52 1999
+++ postfix-19991231-pl10/conf/sample-misc.cf	Wed Nov 22 14:06:22 2000
@@ -180,13 +180,11 @@
 # to distinguish local clients from strangers. See permit_mynetworks
 # and smtpd_recipient_restrictions in the file sample-smtpd.cf file.
 #
-# The default is a list of all networks attached to the machine:  a
-# complete class A network (X.0.0.0/8), a complete class B network
-# (X.X.0.0/16), and so on. If you want stricter control, specify a
-# list of network/mask patterns, where the mask specifies the number
-# of bits in the network part of a host address. You can also specify
-# the absolute pathname of a pattern file instead of listing the
-# patterns here.
+# The default is a list of all interface addresses on the machine.
+# If you want to allow relaying, specify a list of network/mask
+# patterns, where the mask specifies the number of bits in the
+# network part of a host address.  You can also specify the absolute
+# pathname of a pattern file instead of listing the patterns here.
 #
 #mynetworks = 168.100.189.0/28, 127.0.0.0/8
 
diff -ur postfix-19991231-pl10.orig/global/mynetworks.c postfix-19991231-pl10/global/mynetworks.c
--- postfix-19991231-pl10.orig/global/mynetworks.c	Fri Dec 11 21:55:22 1998
+++ postfix-19991231-pl10/global/mynetworks.c	Wed Nov 22 13:27:34 2000
@@ -10,9 +10,7 @@
 /* DESCRIPTION
 /*	This routine uses the address list built by own_inet_addr()
 /*	to produce a list of patterns that match the corresponding
-/*	networks. The patterns are conservative: they match whole
-/*	class A, B, C or D networks. This is usually sufficient to
-/*	distinguish between organizations.
+/*	addresses.
 /* LICENSE
 /* .ad
 /* .fi
@@ -69,6 +67,7 @@
 
 	for (i = 0; i < my_addr_list->used; i++) {
 	    addr = ntohl(my_addr_list->addrs[i].s_addr);
+#if 0
 	    if (IN_CLASSA(addr)) {
 		mask = IN_CLASSA_NET;
 		shift = IN_CLASSA_NSHIFT;
@@ -85,6 +84,9 @@
 		msg_fatal("%s: bad address class: %s",
 			  myname, inet_ntoa(my_addr_list->addrs[i]));
 	    }
+#endif
+	    mask = 0xffffffff;
+	    shift = 0;
 	    net.s_addr = htonl(addr & mask);
 	    vstring_sprintf_append(result, "%s/%d ",
 				   inet_ntoa(net), BITS_PER_ADDR - shift);
diff -ur postfix-19991231-pl10.orig/html/basic.html postfix-19991231-pl10/html/basic.html
--- postfix-19991231-pl10.orig/html/basic.html	Thu Mar 18 01:55:22 1999
+++ postfix-19991231-pl10/html/basic.html	Wed Nov 22 14:04:39 2000
@@ -292,11 +292,10 @@
 
 <p>
 
-By default, <b>mynetworks</b> is set to the class A, B or C networks
-that the machine is attached to. For example, for my machines at
-home, the result is: <b>168.100.0.0/16 127.0.0.0/8.  </b> However,
-network <b>168.100</b> is owned by my ISP. Of course I do not want
-to consider all their customer systems as local, so I use instead:
+By default, <b>mynetworks</b> is a list of all interface addresses on
+the machine. If you want to allow relaying, specify a list of network/mask
+patterns, where the mask specifies the number of bits in the network part
+of a host address:
 
 <dl>
 
diff -ur postfix-19991231-pl10.orig/html/faq.html postfix-19991231-pl10/html/faq.html
--- postfix-19991231-pl10.orig/html/faq.html	Mon Oct 30 02:59:33 2000
+++ postfix-19991231-pl10/html/faq.html	Wed Nov 22 14:07:50 2000
@@ -240,9 +240,9 @@
 
 Postfix has sane defaults for all parameters, so the text shows
 only the overrides. In particular, Postfix will relay mail only
-for clients in its own domain (and subdomains) and in its class A,
-B or C networks. The master.cf file (somewhat like inetd.conf)
-needs tweaking only if you have a very slow or a very fast net/machine.
+for clients in its own domain (and subdomains). The master.cf file
+(somewhat like inetd.conf) needs tweaking only if you have a very
+slow or a very fast net/machine.
 
 <p>
 
