diff -ur postfix-19991231-pl10.orig/conf/postfix-script-sgid postfix-19991231-pl10/conf/postfix-script-sgid
--- postfix-19991231-pl10.orig/conf/postfix-script-sgid	Fri Dec 17 22:12:45 1999
+++ postfix-19991231-pl10/conf/postfix-script-sgid	Mon Dec  4 21:25:40 2000
@@ -177,36 +177,30 @@
 
 	test -d maildrop || {
 		$WARN creating missing Postfix maildrop directory
-		mkdir maildrop || exit 1
-		chmod 1730 maildrop
-		chown $mail_owner maildrop
-		chgrp maildrop maildrop
+		mkdir -m 1730 maildrop || exit 1
+		chown ${mail_owner}.maildrop maildrop || exit 1
 	}
 	test -d pid || {
 		$WARN creating missing Postfix pid directory
-		mkdir pid || exit 1
-		chmod 755 pid
-		chown $mail_owner pid
+		mkdir -m 700 pid || exit 1
 	}
 	for dir in incoming active bounce defer deferred saved corrupt; do
 		test -d $dir || {
 			$WARN creating missing Postfix $dir directory
-			mkdir $dir || exit 1
-			chmod 700 $dir; $CHATTR $dir
-			chown $mail_owner $dir
+			mkdir -m 700 $dir || exit 1
+			$CHATTR $dir || exit 1
+			chown $mail_owner $dir || exit 1
 		}
 	done
 	test -d public || {
 		$WARN creating missing Postfix public directory
-		mkdir public || exit 1
-		chmod 755 public
-		chown $mail_owner public
+		mkdir -m 755 public || exit 1
+		chown $mail_owner public || exit 1
 	}
 	test -d private || {
 		$WARN creating missing Postfix private directory
-		mkdir private || exit 1
-		chmod 700 private
-		chown $mail_owner private
+		mkdir -m 700 private || exit 1
+		chown $mail_owner private || exit 1
 	}
 	find `ls -d $queue_directory/* | \
 	    egrep '/(incoming|active|defer|deferred|bounce|saved|corrupt|public|private)$'` \
@@ -226,6 +220,18 @@
 			cmp -s $path /$path || 
 			    $WARN $queue_directory/$path and /$path differ
 		done
+	done
+
+	# Update the chroot jail
+	mkdir -m 755 -p ${queue_directory}/{etc,lib}
+	for src in /etc/{localtime,resolv.conf,host.conf,hosts,services} \
+	    /lib/libnss_dns{-2.*.so,.so.2}; do
+		test -e "$src" || continue
+		dst="${queue_directory}${src}"
+		cmp -s "$src" "$dst" && continue
+		$WARN "updating $dst"
+		cp -a "$src" "${dst}.tmp"
+		mv -f "${dst}.tmp" "$dst"
 	done
 
 	$command_directory/postsuper || exit 1
