--- postfix/conf/post-install
+++ postfix/conf/post-install
@@ -387,29 +387,11 @@ for name in daemon_directory command_directory queue_directory mail_owner \
     setgid_group sendmail_path newaliases_path mailq_path manpage_directory \
     sample_directory readme_directory html_directory
 do
-    eval test \"\$$name\" = \"`$POSTCONF -c $config_directory -h $name`\" || {
-	override=1
-	break
-    }
+    eval test \"\$$name\" = \"`$POSTCONF -c $config_directory -h $name`\" ||
+	eval \$POSTCONF -c \$config_directory -e \"\$name = \$$name\" ||
+	exit 1
 done
 
-test -n "$override" && {
-    $POSTCONF -c $config_directory -e \
-	"daemon_directory = $daemon_directory" \
-	"command_directory = $command_directory" \
-	"queue_directory = $queue_directory" \
-	"mail_owner = $mail_owner" \
-	"setgid_group = $setgid_group" \
-	"sendmail_path = $sendmail_path" \
-	"mailq_path = $mailq_path" \
-	"newaliases_path = $newaliases_path" \
-	"html_directory = $html_directory" \
-	"manpage_directory = $manpage_directory" \
-	"sample_directory = $sample_directory" \
-	"readme_directory = $readme_directory" \
-    || exit 1
-}
-
 # Use file/directory status information in $config_directory/postfix-files.
 
 test -n "$create" && {
@@ -459,7 +441,7 @@ test -n "$create" && {
 	# Create missing directories with proper owner/group/mode settings.
 	if [ -n "$create" -a "$type" = "d" -a -n "$create_flag" -a ! -d "$path" ]
 	then
-	    mkdir $path || exit 1
+	    mkdir -m700 -- $path || exit 1
 	    set_permission=1
 	# Update all owner/group/mode settings.
 	elif [ -n "$set_perms" ]
@@ -471,14 +453,14 @@ test -n "$create" && {
 	    set_permission=1
 	fi
 	test -n "$set_permission" && {
-	    chown $recursive $owner $path || exit 1
+	    chown $recursive $owner -- $path || exit 1
 	    test -z "$group" || chgrp $recursive $group $path || exit 1
 	    # Don't "chmod -R"; queue file status is encoded in mode bits.
 	    if [ "$type" = "d" -a -n "$recursive" ]
 	    then
-		find $path -type d -exec chmod $mode "{}" ";"
+		find $path -type d -exec chmod $mode -- "{}" ";"
 	    else
-		chmod $mode $path
+		chmod $mode -- $path
 	    fi || exit 1
 	}
     done
@@ -494,7 +476,7 @@ test -n "$upgrade_conf" && {
     grep '^relay' $config_directory/master.cf >/dev/null || {
 	echo Editing $config_directory/master.cf, adding missing entry for relay service
 	cat >>$config_directory/master.cf <<EOF || exit 1
-relay	  unix	-	-	n	-	-	smtp
+relay	  unix	-	-	-	-	-	smtp
 EOF
     }
 
@@ -503,7 +485,7 @@ EOF
     grep '^flush.*flush' $config_directory/master.cf >/dev/null || {
 	echo Editing $config_directory/master.cf, adding missing entry for flush service
 	cat >>$config_directory/master.cf <<EOF || exit 1
-flush     unix  -       -       n       1000?   0       flush
+flush     unix  -       -       -       1000?   0       flush
 EOF
     }
 
@@ -512,7 +494,7 @@ EOF
     grep 'trace.*bounce' $config_directory/master.cf >/dev/null || {
 	echo Editing $config_directory/master.cf, adding missing entry for trace service
 	cat >>$config_directory/master.cf <<EOF || exit 1
-trace	  unix	-	-	n	-	0	bounce
+trace	  unix	-	-	-	-	0	bounce
 EOF
     }
 
@@ -521,7 +503,7 @@ EOF
     grep '^verify.*verify' $config_directory/master.cf >/dev/null || {
 	echo Editing $config_directory/master.cf, adding missing entry for verify service
 	cat >>$config_directory/master.cf <<EOF || exit 1
-verify	  unix	-	-	n	-	1	verify
+verify	  unix	-	-	-	-	1	verify
 EOF
     }
 
@@ -604,7 +586,7 @@ EOF
     grep '^anvil.*anvil' $config_directory/master.cf >/dev/null || {
 	echo Editing $config_directory/master.cf, adding missing entry for anvil service
 	cat >>$config_directory/master.cf <<EOF || exit 1
-anvil	  unix	-	-	n	-	1	anvil
+anvil	  unix	-	-	-	-	1	anvil
 EOF
     }
 
@@ -613,7 +595,7 @@ EOF
     grep '^scache.*scache' $config_directory/master.cf >/dev/null || {
 	echo Editing $config_directory/master.cf, adding missing entry for scache service
 	cat >>$config_directory/master.cf <<EOF || exit 1
-scache	  unix	-	-	n	-	1	scache
+scache	  unix	-	-	-	-	1	scache
 EOF
     }
 
@@ -622,7 +604,7 @@ EOF
     grep '^discard.*discard' $config_directory/master.cf >/dev/null || {
 	echo Editing $config_directory/master.cf, adding missing entry for discard service
 	cat >>$config_directory/master.cf <<EOF || exit 1
-discard	  unix	-	-	n	-	-	discard
+discard	  unix	-	-	-	-	-	discard
 EOF
     }
 
@@ -646,7 +628,7 @@ EOF
     grep '^tlsmgr.*tlsmgr' $config_directory/master.cf >/dev/null || {
 	echo Editing $config_directory/master.cf, adding missing entry for tlsmgr service
 	cat >>$config_directory/master.cf <<EOF || exit 1
-tlsmgr    unix  -       -       n       1000?   1       tlsmgr
+tlsmgr    unix  -       -       -       1000?   1       tlsmgr
 EOF
     }
 
@@ -655,7 +637,7 @@ EOF
     grep '^retry.*error' $config_directory/master.cf >/dev/null || {
 	echo Editing $config_directory/master.cf, adding missing entry for retry service
 	cat >>$config_directory/master.cf <<EOF || exit 1
-retry     unix  -       -       n       -       -       error
+retry     unix  -       -       -       -       -       error
 EOF
     }
 
