--- file-3.41/file.man.orig	Thu Feb 27 22:47:46 2003
+++ file-3.41/file.man	Wed Mar  5 23:18:18 2003
@@ -20,7 +20,7 @@
 .I magicfiles
 ]
 .I file
-\*[Am]...
+&...
 .br
 .B file
 .B -C
@@ -55,10 +55,9 @@
 meaning anything else (data is usually `binary' or non-printable).
 Exceptions are well-known file formats (core files, tar archives)
 that are known to contain binary data.
-When modifying the file
-.I __MAGIC__
-or the program itself,
-.B "preserve these keywords" .
+When adding local definitions to
+.IR /etc/magic ,
+.BR "preserve these keywords" .
 People depend on knowing that all the readable files in a directory
 have the word ``text'' printed.
 Don't do as Berkeley did and change ``shell commands text''
@@ -80,7 +79,7 @@
 implement them)
 are intuited if they are defined in
 the system header file
-.IR \*[Lt]sys/stat.h\*[Gt] .
+.IR <sys/stat.h> .
 .PP
 The magic number tests are used to check for files with data in
 particular fixed formats.
@@ -97,7 +96,9 @@
 The concept of `magic number' has been applied by extension to data files.
 Any file with some invariant identifier at a small fixed
 offset into the file can usually be described in this way.
-The information identifying these files is read from the compiled
+The information identifying these files is read from
+.I /etc/magic
+and the compiled
 magic file
 .I __MAGIC__.mgc ,
 or
@@ -152,20 +153,20 @@
 in any of the character sets listed above is simply said to be ``data''.
 .SH OPTIONS
 .TP 8
-.B \-b
+.B "\-b, \-\-brief"
 Do not prepend filenames to output lines (brief mode).
 .TP 8
-.B \-c
+.B "\-c, \-\-checking\-printout"
 Cause a checking printout of the parsed form of the magic file.
 This is usually used in conjunction with
 .B \-m
 to debug a new magic file before installing it.
 .TP 8
-.B \-C
+.B "\-C, \-\-compile"
 Write a magic.mgc output file that contains a pre-parsed version of
 file.
 .TP 8
-.BI \-f " namefile"
+.BI "\-f, \-\-files\-from" " namefile"
 Read the names of the files to be examined from
 .I namefile
 (one per line)
@@ -175,10 +176,10 @@
 or at least one filename argument must be present;
 to test the standard input, use ``\-'' as a filename argument.
 .TP 8
-.BI \-F " separator"
+.BI "\-F, \-\-separator" " separator"
 Use the specified separator character instead of ``:''.
 .TP 8
-.B \-i
+.B "\-i, \-\-mime"
 Causes the file command to output mime type strings rather than the more
 traditional human readable ones.
 Thus it may say
@@ -191,33 +192,34 @@
 ``magic'' file.
 (See ``FILES'' section, below).
 .TP 8
-.B \-k
+.B "\-k, \-\-keep\-going"
 Don't stop at the first match, keep going.
 .TP 8
-.BI \-m " list"
+.BI "\-m, \-\-magic\-file" " list"
 Specify an alternate list of files containing magic numbers.
 This can be a single file, or a colon-separated list of files.
 .TP 8
-.B \-n
+.B "\-n, \-\-no\-buffer"
 Force stdout to be flushed after checking each file.
 This is only useful if checking a list of files.
 It is intended to be used by programs that want filetype output from a pipe.
 .TP 8
-.B \-N
+.B "\-N, \-\-no\-pad"
 Don't pad output to align filenames nicely.
 .TP 8
-.B \-v
+.B "\-v, \-\-version"
 Print the version of the program and exit.
 .TP 8
-.B \-z
+.B "\-z, \-\-uncompress"
 Try to look inside compressed files.
 .TP 8
-.B \-L
+.B "\-L, \-\-dereference"
+This
 option causes symlinks to be followed, as the like-named option in
 .BR ls (1).
 (on systems that support symbolic links).
 .TP 8
-.B \-s
+.B "\-s, \-\-special\-files"
 Normally,
 .B file
 only attempts to read and determine the type of argument files which
@@ -237,27 +239,33 @@
 to disregard the file size as reported by
 .BR stat (2)
 since on some systems it reports a zero size for raw disk partitions.
+.TP
+.B "\-\-help"
+Print a help message and exit.
 .SH FILES
+.TP
 .I __MAGIC__.mgc
-\- default compiled list of magic numbers
-.PP
+Default compiled list of magic numbers
+.TP
 .I __MAGIC__
-\- default list of magic numbers
-.PP
+Default list of magic numbers
+.TP
 .I __MAGIC__.mime
-\- default list of magic numbers, used to output mime types when the -i option
+Default list of magic numbers, used to output mime types when the -i option
 is specified.
+.TP
+.I /etc/magic
+Local additions to magic wisdom.
 
 .SH ENVIRONMENT
 The environment variable
 .B MAGIC
 can be used to set the default magic number files.
 .SH SEE ALSO
+.BR hexdump (1),
+.BR od (1),
+.BR strings (1),
 .BR magic (__FSECTION__)
-\- description of magic file format.
-.br
-.BR strings (1), " od" (1), " hexdump(1)"
-\- tools for examining non-textfiles.
 .SH STANDARDS CONFORMANCE
 This program is believed to exceed the System V Interface Definition
 of FILE(CMD), as near as one can determine from the vague language
@@ -272,11 +280,11 @@
 as a delimiter, so that spaces in pattern strings must be escaped.
 For example,
 .br
-\*[Gt]10	string	language impress\ 	(imPRESS data)
+>10	string	language impress\ 	(imPRESS data)
 .br
 in an existing magic file would have to be changed to
 .br
-\*[Gt]10	string	language\e impress	(imPRESS data)
+>10	string	language\e impress	(imPRESS data)
 .br
 In addition, in this version, if a pattern string contains a backslash,
 it must be escaped.  For example
@@ -292,10 +300,10 @@
 .BR file (1)
 command derived from the System V one, but with some extensions.
 My version differs from Sun's only in minor ways.
-It includes the extension of the `\*[Am]' operator, used as,
+It includes the extension of the `&' operator, used as,
 for example,
 .br
-\*[Gt]16	long\*[Am]0x7fffffff	\*[Gt]0		not stripped
+>16	long&0x7fffffff	>0		not stripped
 .SH MAGIC DIRECTORY
 The magic file entries have been collected from various sources,
 mainly USENET, and contributed by various authors.
@@ -307,12 +315,6 @@
 The order of entries in the magic file is significant.
 Depending on what system you are using, the order that
 they are put together may be incorrect.
-If your old
-.B file
-command uses a magic file,
-keep the old magic file around for comparison purposes
-(rename it to
-.IR __MAGIC__.orig ).
 .SH EXAMPLES
 .nf
 $ file file.c file /dev/{wd0a,hda}
@@ -371,14 +373,14 @@
 This slowed the program down slightly but made it a lot more flexible.
 .PP
 This program, based on the System V version,
-was written by Ian Darwin \*[Lt]ian@darwinsys.com\*[Gt]
+was written by Ian Darwin <ian@darwinsys.com>
 without looking at anybody else's source code.
 .PP
 John Gilmore revised the code extensively, making it better than
 the first version.
 Geoff Collyer found several inadequacies
 and provided some magic file entries.
-Contributions by the `\*[Am]' operator by Rob McMahon, cudcv@warwick.ac.uk, 1989.
+Contributions by the `&' operator by Rob McMahon, cudcv@warwick.ac.uk, 1989.
 .PP
 Guy Harris, guy@netapp.com, made many changes from 1993 to the present.
 .PP
--- file-3.41/magic.man.orig	Thu Feb 27 22:47:46 2003
+++ file-3.41/magic.man	Wed Mar  5 23:27:32 2003
@@ -41,12 +41,15 @@
 .IP string
 A string of bytes.
 The string type specification can be optionally followed by /[Bbc]*.
-The ``B'' flag compacts whitespace in the target, which must
+The `B' flag compacts whitespace in the target, which must
 contain at least one whitespace character.
-If the magic has "n" consecutive
-blanks, the target needs at least "n" consecutive blanks to match.
-The ``b'' flag treats every blank in the target as an optional blank.
-Finally the ``c''
+If the magic has
+.I n
+consecutive blanks, the target needs at least
+.I n
+consecutive blanks to match.
+The `b' flag treats every blank in the target as an optional blank.
+Finally the `c'
 flag, specifies case insensitive matching: lowercase characters in the magic
 match both lower and upper case characters in the targer, whereas upper case
 characters in the magic, only much uppercase characters in the target.
@@ -76,7 +79,7 @@
 .RE
 .PP
 The numeric types may optionally be followed by
-.B \*[Am]
+.B &
 and a numeric value,
 to specify that the value is to be AND'ed with the
 numeric value before any comparisons are done.
@@ -94,13 +97,13 @@
 It may be
 .BR = ,
 to specify that the value from the file must equal the specified value,
-.BR \*[Lt] ,
+.BR < ,
 to specify that the value from the file must be less than the specified
 value,
-.BR \*[Gt] ,
+.BR > ,
 to specify that the value from the file must be greater than the specified
 value,
-.BR \*[Am] ,
+.BR & ,
 to specify that the value from the file must have set all of the bits
 that are set in the specified value,
 .BR ^ ,
@@ -123,17 +126,17 @@
 file must match the specified byte string.
 The operators
 .BR = ,
-.B \*[Lt]
+.B <
 and
-.B \*[Gt]
+.B >
 (but not
-.BR \*[Am] )
+.BR & )
 can be applied to strings.
 The length used for matching is that of the string argument
 in the magic file.
 This means that a line can match any string, and
 then presumably print that string, by doing
-.B \*[Gt]\e0
+.B >\e0
 (because all strings are greater than the null string).
 .IP message
 The message to be printed if the comparison succeeds.
@@ -145,12 +148,12 @@
 Some file formats contain additional information which is to be printed
 along with the file type.
 A line which begins with the character
-.B \*[Gt]
+.B >
 indicates additional tests and messages to be printed.
 The number of
-.B \*[Gt]
+.B >
 on the line indicates the level of the test; a line with no
-.B \*[Gt]
+.B >
 at the beginning is considered to be at level 0.
 Each line at level
 .IB n \(pl1
@@ -166,7 +169,7 @@
 .I n
 terminates this.
 If the first character following the last
-.B \*[Gt]
+.B >
 is a
 .B (
 then the string after the parenthesis is interpreted as an indirect offset.
@@ -195,10 +198,10 @@
 You can specify an offset relative to the end of the
 last uplevel field (of course this may only be done for sublevel tests, i.e.
 test beginning with
-.B \*[Gt]
+.B >
 ).
 Such a relative offset is specified using
-.B \*[Am]
+.B &
 as a prefix to the offset.
 .SH BUGS
 The formats
@@ -221,12 +224,11 @@
 indirect offsets.
 .SH SEE ALSO
 .BR file (__CSECTION__)
-\- the command that reads this file.
 .\"
 .\" From: guy@sun.uucp (Guy Harris)
 .\" Newsgroups: net.bugs.usg
 .\" Subject: /etc/magic's format isn't well documented
-.\" Message-ID: \*[Lt]2752@sun.uucp\*[Gt]
+.\" Message-ID: <2752@sun.uucp>
 .\" Date: 3 Sep 85 08:19:07 GMT
 .\" Organization: Sun Microsystems, Inc.
 .\" Lines: 136
