diff -rup file-3.28/Magdir/elf file-3.28-/Magdir/elf
--- file-3.28/Magdir/elf	Sun Oct 31 21:10:42 1999
+++ file-3.28-/Magdir/elf	Thu Feb 17 00:28:01 2000
@@ -79,7 +79,7 @@
 >>18	leshort		47		Hitachi H8/300H,
 >>18	leshort		48		Hitachi H8S,
 >>18	leshort		49		Hitachi H8/500,
->>18	leshort		50		IA-64 (Intel 64 bit architecture)
+>>18	leshort		50		IA-64,
 >>18	leshort		51		Stanford MIPS-X,
 >>18	leshort		52		Motorola Coldfire,
 >>18	leshort		53		Motorola M68HC12,
@@ -154,7 +154,7 @@
 >>18	beshort		47		Hitachi H8/300H,
 >>18	beshort		48		Hitachi H8S,
 >>18	beshort		49		Hitachi H8/500,
->>18	beshort		50		Intel Merced Processor,
+>>18	beshort		50		IA-64,
 >>18	beshort		51		Stanford MIPS-X,
 >>18	beshort		52		Motorola Coldfire,
 >>18	beshort		53		Motorola M68HC12,
--- file-3.33/readelf.c.orig	Sat Aug  5 22:00:12 2000
+++ file-3.33/readelf.c	Wed Nov 29 10:26:08 2000
@@ -423,7 +423,7 @@
 	    return;
 
 
-	class = buf[4];
+	class = buf[EI_CLASS];
 
 	if (class == ELFCLASS32) {
 		Elf32_Ehdr elfhdr;
@@ -433,7 +433,7 @@
 
 		u.l = 1;
 		(void) memcpy(&elfhdr, buf, sizeof elfhdr);
-		swap = (u.c[sizeof(int32) - 1] + 1) != elfhdr.e_ident[5];
+		swap = (u.c[sizeof(u.l) - 1] + 1) != elfhdr.e_ident[EI_DATA];
 
 		if (getu16(swap, elfhdr.e_type) == ET_CORE) 
 #ifdef ELFCORE
@@ -470,7 +470,7 @@
 
 		u.l = 1;
 		(void) memcpy(&elfhdr, buf, sizeof elfhdr);
-		swap = (u.c[sizeof(int32) - 1] + 1) != elfhdr.e_ident[5];
+		swap = (u.c[sizeof(u.l) - 1] + 1) != elfhdr.e_ident[EI_DATA];
 
 		if (getu16(swap, elfhdr.e_type) == ET_CORE) 
 #ifdef ELFCORE
