--- file-3.41/magic.mime.orig	Thu Feb 27 22:47:46 2003
+++ file-3.41/magic.mime	Wed Mar  5 23:48:26 2003
@@ -304,6 +304,7 @@
 0	string		\<!doctype\ HTML	text/html
 0	string		\<!DOCTYPE\ HTML	text/html
 0	string		\<!doctype\ html	text/html
+0	string		\<!DOCTYPE\ html	text/html
 
 #------------------------------------------------------------------------------
 # images:  file(1) magic for image formats (see also "c-lang" for XPM bitmaps)
