--- file-3.41/Makefile.am.orig	Thu Feb 27 22:47:45 2003
+++ file-3.41/Makefile.am	Wed Mar  5 02:10:56 2003
@@ -3,10 +3,12 @@
 
 bin_PROGRAMS = file
 
+bin_SCRIPTS = magic2mime
+
 data_DATA = magic magic.mime magic.mgc magic.mime.mgc
 
 MAGIC	 = @datadir@/magic
-CPPFLAGS = -DMAGIC='"$(MAGIC)"'
+CPPFLAGS = -DMAGIC='"/etc/magic:$(MAGIC)"'
 
 if FSECT5
 man_MAGIC = magic.5
@@ -14,7 +16,7 @@
 man_MAGIC = magic.4
 endif
 fsect = @fsect@
-man_MANS = file.1 $(man_MAGIC)
+man_MANS = file.1 $(man_MAGIC) magic2mime.1
 
 file_SOURCES = file.c apprentice.c fsmagic.c softmagic.c ascmagic.c \
 	compress.c is_tar.c readelf.c print.c \
@@ -25,8 +27,8 @@
 
 CLEANFILES = $(man_MANS) magic magic.mgc magic.mime.mgc
 
-magic: Header Localstuff $(magic_FRAGMENTS)
-	cat $(srcdir)/Header $(srcdir)/Localstuff > $@
+magic: Header $(magic_FRAGMENTS)
+	cat $(srcdir)/Header > $@
 	for frag in $(magic_FRAGMENTS); do \
 	  if test -f $(srcdir)/$$frag; then \
 	    f=$(srcdir)/$$frag; \
@@ -49,6 +51,13 @@
 	    -e s@__VERSION__@${VERSION}@g \
 	    -e s@__MAGIC__@${MAGIC}@g $(srcdir)/file.man > $@
 
+magic2mime.1: Makefile magic2mime.man
+	@rm -f $@
+	sed -e s@__CSECTION__@1@g \
+	    -e s@__FSECTION__@${fsect}@g \
+	    -e s@__VERSION__@${VERSION}@g \
+	    -e s@__MAGIC__@${MAGIC}@g $(srcdir)/magic2mime.man > $@
+
 magic.${fsect}: Makefile magic.man
 	@rm -f $@
 	sed -e s@__CSECTION__@1@g \
@@ -57,6 +66,8 @@
 	    -e s@__MAGIC__@${MAGIC}@g $(srcdir)/magic.man > $@
 
 magic_FRAGMENTS = \
+Magdir/debian-additions \
+Magdir/linux \
 Magdir/acorn \
 Magdir/adi \
 Magdir/adventure \
@@ -127,7 +138,6 @@
 Magdir/lecter \
 Magdir/lex \
 Magdir/lif \
-Magdir/linux \
 Magdir/lisp \
 Magdir/mach \
 Magdir/macintosh \
