--- file-3.40.orig/apprentice.c
+++ file-3.40/apprentice.c
@@ -136,7 +136,8 @@
 			return rv;
 	}
 #ifndef COMPILE_ONLY
-	if ((rv = apprentice_map(&magic, &nmagic, fn, action)) != 0)
+	if ((rv = apprentice_map(&magic, &nmagic, fn, action)) != 0 && strcmp(fn, "/etc/magic"))
+	        /* Say we use a certain magic file, except when it's the standard host wide file */
 		(void)fprintf(stderr, "%s: Using regular magic file `%s'\n",
 		    progname, fn);
