--- file-3.41/compress.c.orig	Wed Jul  3 21:26:37 2002
+++ file-3.41/compress.c	Wed Mar  5 21:03:56 2003
@@ -302,9 +302,23 @@
 	default: /* parent */
 		(void) close(fdin[0]);
 		(void) close(fdout[1]);
-		if (swrite(fdin[1], old, n) != n) {
-			n = 0;
-			goto err;
+		/* fork again, to avoid blocking because both pipes filled */
+		switch (fork()) {
+		case 0: /* child */
+		    (void) close(fdout[0]);
+		    /*fprintf(stderr, "about to write %d bytes to pipe\n", n);*/
+		    if (swrite(fdin[1], old, n) != n) {
+			exit(1);
+		    }
+		    exit(0);
+		    /*NOTREACHED*/
+		case -1:
+		    error("could not fork (%s).\n", strerror(errno));
+		    /*NOTREACHED*/
+
+		/* default: // parent
+		     fall through */
+
 		}
 		(void) close(fdin[1]);
 		fdin[1] = -1;
@@ -312,6 +326,7 @@
 			n = 0;
 			goto err;
 		}
+		/*fprintf(stderr, "about to read %d bytes from pipe\n", HOWMANY);*/
 		if ((n = sread(fdout[0], *newch, HOWMANY)) <= 0) {
 			free(*newch);
 			n = 0;
@@ -323,7 +338,9 @@
 		if (fdin[1] != -1)
 			(void) close(fdin[1]);
 		(void) close(fdout[0]);
-		(void) wait(NULL);
+		/*(void) wait(NULL);*/
+		/* reap any terminated children, but don't wait for them. */
+		while (waitpid(-1, NULL, WNOHANG) > 0);
 		return n;
 	}
 }
