diff -uNr file-3.33.orig/Magdir/animation file-3.33/Magdir/animation
--- file-3.33.orig/Magdir/animation	Sat Aug  5 20:12:08 2000
+++ file-3.33/Magdir/animation	Tue Nov 28 14:42:45 2000
@@ -15,80 +15,80 @@
 # MPEG Audio (*.mpx)
 # from dreesen@math.fu-berlin.de
 
-# XXX
-# This conflicts with the FF FE signature for UTF-16-encoded Unicode
-# text, which will be identified as an MP3 file.  I don't have any MP3s
-# so I don't know how to (or even if it's possible to) change this to
-# tell the two apart.    enf@pobox.com
-
-0       beshort         &0xfff0         MP
+0       beshort         &0xffe0         MPEG
 # MPEG 1.0
->1      byte&0x08       =0x08           \b
+>1      byte&0x08       =0x08           \b 1.0
 # Layer 3
->>1     byte            &0x02           \b3
->>>2    byte&0xf0       =0x10           \b,  32 kBits
->>>2    byte&0xf0       =0x20           \b,  40 kBits
->>>2    byte&0xf0       =0x30           \b,  48 kBits
->>>2    byte&0xf0       =0x40           \b,  56 kBits
->>>2    byte&0xf0       =0x50           \b,  64 kBits
->>>2    byte&0xf0       =0x60           \b,  80 kBits
->>>2    byte&0xf0       =0x70           \b,  96 kBits
->>>2    byte&0xf0       =0x80           \b, 112 kBits
->>>2    byte&0xf0       =0x90           \b, 128 kBits
->>>2    byte&0xf0       =0xA0           \b, 160 kBits
->>>2    byte&0xf0       =0xB0           \b, 192 kBits
->>>2    byte&0xf0       =0xC0           \b, 224 kBits
->>>2    byte&0xf0       =0xD0           \b, 256 kBits
->>>2    byte&0xf0       =0xE0           \b, 320 kBits
+>>1     byte            &0x02           \b layer 3 audio stream data
+>>>2    byte&0xf0       =0x10           \b,  32 kBit/s
+>>>2    byte&0xf0       =0x20           \b,  40 kBit/s
+>>>2    byte&0xf0       =0x30           \b,  48 kBit/s
+>>>2    byte&0xf0       =0x40           \b,  56 kBit/s
+>>>2    byte&0xf0       =0x50           \b,  64 kBit/s
+>>>2    byte&0xf0       =0x60           \b,  80 kBit/s
+>>>2    byte&0xf0       =0x70           \b,  96 kBit/s
+>>>2    byte&0xf0       =0x80           \b, 112 kBit/s
+>>>2    byte&0xf0       =0x90           \b, 128 kBit/s
+>>>2    byte&0xf0       =0xA0           \b, 160 kBit/s
+>>>2    byte&0xf0       =0xB0           \b, 192 kBit/s
+>>>2    byte&0xf0       =0xC0           \b, 224 kBit/s
+>>>2    byte&0xf0       =0xD0           \b, 256 kBit/s
+>>>2    byte&0xf0       =0xE0           \b, 320 kBit/s
 # Layer 2
->>1     byte            &0x04           \b2
->>>2    byte&0xf0       =0x10           \b,  32 kBits
->>>2    byte&0xf0       =0x20           \b,  48 kBits
->>>2    byte&0xf0       =0x30           \b,  56 kBits
->>>2    byte&0xf0       =0x40           \b,  64 kBits
->>>2    byte&0xf0       =0x50           \b,  80 kBits
->>>2    byte&0xf0       =0x60           \b,  96 kBits
->>>2    byte&0xf0       =0x70           \b, 112 kBits
->>>2    byte&0xf0       =0x80           \b, 128 kBits
->>>2    byte&0xf0       =0x90           \b, 160 kBits
->>>2    byte&0xf0       =0xA0           \b, 192 kBits
->>>2    byte&0xf0       =0xB0           \b, 224 kBits
->>>2    byte&0xf0       =0xC0           \b, 256 kBits
->>>2    byte&0xf0       =0xD0           \b, 320 kBits
->>>2    byte&0xf0       =0xE0           \b, 384 kBits
+>>1     byte            &0x04           \b layer 2 audio stream data
+>>>2    byte&0xf0       =0x10           \b,  32 kBit/s
+>>>2    byte&0xf0       =0x20           \b,  48 kBit/s
+>>>2    byte&0xf0       =0x30           \b,  56 kBit/s
+>>>2    byte&0xf0       =0x40           \b,  64 kBit/s
+>>>2    byte&0xf0       =0x50           \b,  80 kBit/s
+>>>2    byte&0xf0       =0x60           \b,  96 kBit/s
+>>>2    byte&0xf0       =0x70           \b, 112 kBit/s
+>>>2    byte&0xf0       =0x80           \b, 128 kBit/s
+>>>2    byte&0xf0       =0x90           \b, 160 kBit/s
+>>>2    byte&0xf0       =0xA0           \b, 192 kBit/s
+>>>2    byte&0xf0       =0xB0           \b, 224 kBit/s
+>>>2    byte&0xf0       =0xC0           \b, 256 kBit/s
+>>>2    byte&0xf0       =0xD0           \b, 320 kBit/s
+>>>2    byte&0xf0       =0xE0           \b, 384 kBit/s
 # freq
 >>2     byte&0x0C       =0x00           \b, 44.1 kHz
 >>2     byte&0x0C       =0x04           \b, 48 kHz
 >>2     byte&0x0C       =0x08           \b, 32 kHz
+# MPEG 2.5
+>1     byte&0x18       =0x00            \b 2.5
+# Layer 3
+>>1    byte            &0x02            \b layer 3 audio stream data
+# Layer 2
+>>1    byte            &0x04            \b layer 2 audio stream data
 # MPEG 2.0
->1      byte&0x08       =0x00           \b
+>1      byte&0x18       =0x10           \b 2.0
 # Layer 3
->>1     byte            &0x02           \b3
+>>1     byte            &0x02           \b layer 3 audio stream data
 # Layer 2
->>1     byte            &0x04           \b2
->>2     byte&0xf0       =0x10           \b,   8 kBits
->>2     byte&0xf0       =0x20           \b,  16 kBits
->>2     byte&0xf0       =0x30           \b,  24 kBits
->>2     byte&0xf0       =0x40           \b,  32 kBits
->>2     byte&0xf0       =0x50           \b,  40 kBits
->>2     byte&0xf0       =0x60           \b,  48 kBits
->>2     byte&0xf0       =0x70           \b,  56 kBits
->>2     byte&0xf0       =0x80           \b,  64 kBits
->>2     byte&0xf0       =0x90           \b,  80 kBits
->>2     byte&0xf0       =0xA0           \b,  96 kBits
->>2     byte&0xf0       =0xB0           \b, 112 kBits
->>2     byte&0xf0       =0xC0           \b, 128 kBits
->>2     byte&0xf0       =0xD0           \b, 144 kBits
->>2     byte&0xf0       =0xE0           \b, 160 kBits
+>>1     byte            &0x04           \b layer 2 audio stream data
+>>2     byte&0xf0       =0x10           \b,   8 kBit/s
+>>2     byte&0xf0       =0x20           \b,  16 kBit/s
+>>2     byte&0xf0       =0x30           \b,  24 kBit/s
+>>2     byte&0xf0       =0x40           \b,  32 kBit/s
+>>2     byte&0xf0       =0x50           \b,  40 kBit/s
+>>2     byte&0xf0       =0x60           \b,  48 kBit/s
+>>2     byte&0xf0       =0x70           \b,  56 kBit/s
+>>2     byte&0xf0       =0x80           \b,  64 kBit/s
+>>2     byte&0xf0       =0x90           \b,  80 kBit/s
+>>2     byte&0xf0       =0xA0           \b,  96 kBit/s
+>>2     byte&0xf0       =0xB0           \b, 112 kBit/s
+>>2     byte&0xf0       =0xC0           \b, 128 kBit/s
+>>2     byte&0xf0       =0xD0           \b, 144 kBit/s
+>>2     byte&0xf0       =0xE0           \b, 160 kBit/s
 # freq
 >>2     byte&0x0C       =0x00           \b, 22.05 kHz
 >>2     byte&0x0C       =0x04           \b, 24 kHz
 >>2     byte&0x0C       =0x08           \b, 16 kHz
 # misc
->3      byte&0xC0       =0x00           \b, Stereo
->3      byte&0xC0       =0x40           \b, JStereo
->3      byte&0xC0       =0x80           \b, Dual-Ch
->3      byte&0xC0       =0xC0           \b, Mono
+>3      byte&0xC0       =0x00           \b, stereo
+>3      byte&0xC0       =0x40           \b, jstereo
+>3      byte&0xC0       =0x80           \b, dual-ch
+>3      byte&0xC0       =0xC0           \b, mono
 #>1     byte&0x01       =0x00           \b, Error Protection
 #>2     byte&0x02       =0x02           \b, Padding
 #>2     byte&0x01       =0x01           \b, Private
diff -uNr file-3.33.orig/Magdir/fonts file-3.33/Magdir/fonts
--- file-3.33.orig/Magdir/fonts	Sun Nov 28 22:02:30 1999
+++ file-3.33/Magdir/fonts	Tue Nov 28 14:46:14 2000
@@ -49,3 +49,6 @@
 # downloadable fonts for browser (prints type) anthon@mnt.org
 0	string		PFR1		PFR1 font
 >102	string		>0		\b: %s
+
+# Windows fonts
+0      belong          0x00010000      MS Windows true type font
\ No newline at end of file
diff -uNr file-3.33.orig/Magdir/images file-3.33/Magdir/images
--- file-3.33.orig/Magdir/images	Mon Nov 13 02:30:51 2000
+++ file-3.33/Magdir/images	Tue Nov 28 14:49:39 2000
@@ -6,7 +6,6 @@
 # additions by janl@ifi.uio.no as well as others. Jan also suggested
 # merging several one- and two-line files into here.
 #
-# little magic: PCX (first byte is 0x0a)
 
 # Targa - matches `povray', `ppmtotga' and `xv' outputs
 # by Philippe De Muyter <phdm@macqel.be>
@@ -43,6 +42,10 @@
 >21	byte		48		version 2
 >21	byte		51		version 3
 
+# NITF is defined by United States MIL-STD-2500A
+0      string  NITF    National Imagery Transmission Format
+>25    string  >\0     dated %.14s
+
 # NIFF (Navy Interchange File Format, a modification of TIFF) images
 0	string		IIN1		NIFF image data
 
@@ -79,7 +82,7 @@
 >4	string		7a		\b, version 8%s,
 >4	string		9a		\b, version 8%s,
 >6	leshort		>0		%hd x
->8	leshort		>0		%hd,
+>8	leshort		>0		%hd
 #>10	byte		&0x80		color mapped,
 #>10	byte&0x07	=0x00		2 colors
 #>10	byte&0x07	=0x01		4 colors
@@ -268,6 +271,15 @@
 
 # Visio drawings
 03000	string	Visio\ (TM)\ Drawing	%s
+
+# PSD (Adobe Photoshop bitmap) (Wolfram Kleff)
+0      string          8BPS    PSD image data (Adobe Photoshop bitmap)
+
+# PCX (should be reliable with a 2 byte header - Wolfram Kleff)
+0      beshort         0x0A00  PCX image data, version 2.5
+0      beshort         0x0A02  PCX image data, version 2.8 with palette
+0      beshort         0x0A03  PCX image data, version 2.8 without palette
+0      beshort         0x0A05  PCX image data, version 3.0
 
 # Tgif files
 0	string	\%TGIF\ x 		Tgif file version %s
--- file-3.33.orig/Magdir/c-lang
+++ file-3.33/Magdir/c-lang
@@ -1,4 +1,3 @@
-
 #------------------------------------------------------------------------------
 # c-lang:  file(1) magic for C programs (or REXX)
 #
@@ -10,4 +9,4 @@
 # this first will upset you if you're a PL/1 shop...
 # in which case rm it; ascmagic will catch real C programs
 #0	string		/*		C or REXX program text
-0	string		//		C++ program text
+#0	string		//		C++ program text
--- file-3.33.orig/Magdir/filesystems
+++ file-3.33/Magdir/filesystems
@@ -23,36 +23,38 @@
 >0770	long		x		%ld blocks
 # Is there a boot block written 1 sector in?
 >512    belong&077777777	0600407	\b, boot block present
-0x1FE	leshort	0xAA55			x86 boot sector
->2	string	OSBS			\b, OS/BS MBR
->0x8C	string	Invalid\ partition\ table	\b, MS-DOS MBR
->0	string	\0\0\0\0		\b, extended partition table
->0	leshort 0x3CEB			\b, system
->>3	string	>\0			%s
->>0x36	string	FAT			\b, %s
->>>0x39	string	12			(%s bit)
->>>0x39	string	16			(%s bit)
->0x52	string	FAT32			\b, FAT (32 bit)
->>>43	string		>NO\ NAME	label: %.11s,
->>>43	string		<NO\ NAME	label: %.11s,
->>>43	string		NO\ NAME	unlabeled,
->>>19	leshort		>0		%d sectors
->>>19	leshort		0		
->>>>32	lelong		x		%d sectors
->0x200	lelong	0x82564557		\b, BSD disklabel
 
-# Minix filesystems - Juan Cespedes <cespedes@debian.org>
+# Moved to linux.
+#0x1FE	leshort	0xAA55			x86 boot sector
+#>2	string	OSBS			\b, OS/BS MBR
+#>0x8C	string	Invalid\ partition\ table	\b, MS-DOS MBR
+#>0	string	\0\0\0\0		\b, extended partition table
+#>0	leshort 0x3CEB			\b, system
+#>>3	string	>\0			%s
+#>>0x36	string	FAT			\b, %s
+#>>>0x39	string	12			(%s bit)
+#>>>0x39	string	16			(%s bit)
+#>0x52	string	FAT32			\b, FAT (32 bit)
+#>>>43	string		>NO\ NAME	label: %.11s,
+#>>>43	string		<NO\ NAME	label: %.11s,
+#>>>43	string		NO\ NAME	unlabeled,
+#>>>19	leshort		>0		%d sectors
+#>>>19	leshort		0		
+#>>>>32	lelong		x		%d sectors
+#>0x200	lelong	0x82564557		\b, BSD disklabel
+
+# Minix filesystems - Juan Céspedes <cespedes@debian.org>
 0x410	leshort		0x137f		Minix filesystem
 0x410	leshort		0x138f		Minix filesystem, 30 char names
 0x410	leshort		0x2468		Minix filesystem, version 2
 0x410	leshort		0x2478		Minix filesystem, version 2, 30 char names
 
-# romfs filesystems - Juan Cespedes <cespedes@debian.org>
+# romfs filesystems - Juan Céspedes <cespedes@debian.org>
 0	string		-rom1fs-\0	romfs filesystem, version 1
 >8	belong	x			%d bytes,
 >16	string	x			named %s.
 
-# netboot image - Juan Cespedes <cespedes@debian.org>
+# netboot image - Juan Céspedes <cespedes@debian.org>
 0	lelong		0x1b031336L	Netboot image,
 >4	lelong&0xFFFFFF00	0
 >>4	lelong&0x100	0x000		mode 2
--- file-3.33.orig/Magdir/gnu
+++ file-3.33/Magdir/gnu
@@ -1,3 +1,4 @@
+
 #
 # GNU nlsutils message catalog file format
 #
--- file-3.33.orig/Magdir/java
+++ file-3.33/Magdir/java
@@ -1,3 +1,4 @@
+
 #------------------------------------------------------------
 # Java ByteCode
 # From Larry Schwimmer (schwim@cs.stanford.edu)
--- file-3.33.orig/Magdir/linux
+++ file-3.33/Magdir/linux
@@ -45,10 +45,10 @@
 2	string		LILO		Linux/i386 LILO boot/chain loader
 #
 # Debian Packages, from Peter Tobias <tobias@server.et-inf.fho-emden.de>
-0	string		0.9
->8	byte		0x0a		old Debian Binary Package
->>3	byte		>0		\b, created by dpkg 0.9%c
->>4	byte		>0		pl%c
+#0	string		0.9
+#>8	byte		0x0a		old Debian Binary Package
+#>>3	byte		>0		\b, created by dpkg 0.9%c
+#>>4	byte		>0		pl%c
 # PSF fonts, from H. Peter Anvin <hpa@yggdrasil.com>
 0	leshort		0x0436		Linux/i386 PC Screen Font data,
 >2	byte		0		256 characters, no directory,
@@ -72,30 +72,51 @@
 >8	long		>0		not stripped
 >8	long		0		stripped
 >23	leshort		>0		- version %ld.
-#
-# Linux kernel boot images, from Albert Cahalan <acahalan@cs.uml.edu>
-# and others such as Axel Kohlmeyer <akohlmey@rincewind.chemie.uni-ulm.de>
-# and Nicolás Lichtmaier <nick@debian.org>
-# All known start with: b8 c0 07 8e d8 b8 00 90 8e c0 b9 00 01 29 f6 29
-514		string	HdrS		Linux kernel
->518		leshort	>0
->>529		byte	0		zImage data,
->>529		byte	1		bzImage data,
->0x048c		byte	0x31
->>0x048c	string	x		version %s
->0x0493		byte	0x31
->>0x0493	string	x		version %s
->0x048c		byte	0x32
->>0x048c	string	x		version %s
->0x0493		byte	0x32
->>0x0493	string	x		version %s
->0x04df		byte	0x32
->>0x04df	string	x		version %s
->0x04fb		byte	0x32
->>0x04fb	string	x		version %s
+
+# Linux kernel boot images (i386 arch) (Wolfram Kleff)
+514	string		HdrS		Linux kernel
+>510	leshort		0xAA55		x86 boot executable
+>>518	leshort		0x201
+>>>529	byte		0		zImage,
+>>>529	byte		1		bzImage,
+>>>(526.s+0x200) string	>\0		version %s,
+>>498	leshort		1		RO-rootFS,
+>>498	leshort		0		RW-rootFS,
+>>508	leshort		>0		root_dev=0x%X,
+>>502	leshort		>0		swap_dev=0x%X,
+>>504	leshort		>0		RAMdisksize=%u KB,
+>>506	leshort		0xFFFF		Normal VGA
+>>506	leshort		0xFFFE		Extended VGA
+>>506	leshort		0xFFFD		Prompt for Videomode
+>>506	leshort		>0		Video mode=%d
+
 # This also matches new kernels, which were caught above by "HdrS".
 0		belong	0xb8c0078e	Linux kernel
 >0x1e3		string	Loading		version 1.3.79 or older
 >0x1e9		string	Loading		from prehistoric times
-# LSM entries - Nicolás Lichtmaier <nick@feedback.net.ar>
+
+# System.map files - Nicolás Lichtmaier <nick@debian.org>
+8	string	\ A\ _text	Linux kernel symbol map text
+
+# LSM entries - Nicolás Lichtmaier <nick@debian.org>
 0	string	Begin3	Linux Software Map entry text
+0	string	Begin4	Linux Software Map entry text (new format)
+
+# Moved from filesistems to be able to detect linux kernels.
+0x1FE	leshort	0xAA55			x86 boot sector
+>2	string	OSBS			\b, OS/BS MBR
+>0x8C	string	Invalid\ partition\ table	\b, MS-DOS MBR
+>0	string	\0\0\0\0		\b, extended partition table
+>0	leshort 0x3CEB			\b, system
+>>3	string	>\0			%s
+>>0x36	string	FAT			\b, %s
+>>>0x39	string	12			(%s bit)
+>>>0x39	string	16			(%s bit)
+>0x52	string	FAT32			\b, FAT (32 bit)
+>>>43	string		>NO\ NAME	label: %.11s,
+>>>43	string		<NO\ NAME	label: %.11s,
+>>>43	string		NO\ NAME	unlabeled,
+>>>19	leshort		>0		%d sectors
+>>>19	leshort		0		
+>>>>32	lelong		x		%d sectors
+>0x200	lelong	0x82564557		\b, BSD disklabel
--- file-3.33.orig/Magdir/motorola
+++ file-3.33/Magdir/motorola
@@ -2,6 +2,26 @@
 #------------------------------------------------------------------------------
 # motorola:  file(1) magic for Motorola 68K and 88K binaries
 #
+# Atari ST/TT... program format (sent by Wolfram Kleff <kleff@cs.uni-bonn.de>)
+0       beshort         0x601A          Atari 68xxx executable,
+>2      belong          x               text len %lu,
+>6      belong          x               data len %lu,
+>10     belong          x               BSS len %lu,
+>14     belong          x               symboltab len %lu,
+>18     belong          0
+>22     belong          &0x01           fastload flag,
+>22     belong          &0x02           may be loaded to alternate RAM,
+>22     belong          &0x04           malloc may be from alternate RAM,
+>22     belong          x               flags: 0x%lX,
+>26     beshort         0               no relocation tab
+>26     beshort         !0              + relocation tab
+>30     string          SFX             [Self-Extracting LZH SFX archive]
+>38     string          SFX             [Self-Extracting LZH SFX archive]
+>44     string          ZIP!            [Self-Extracting ZIP SFX archive]
+
+0       beshort         0x0064          Atari 68xxx CPX file
+>8      beshort         x               (version %04lx)
+
 # 68K
 #
 0	beshort		0520		mc68k COFF
--- file-3.33.orig/Magdir/msdos
+++ file-3.33/Magdir/msdos
@@ -5,6 +5,7 @@
 
 # .BAT files (Daniel Quinlan, quinlan@yggdrasil.com)
 0	string	@echo\ off	MS-DOS batch file text
+0	string	@ECHO\ OFF	MS-DOS batch file text
 
 # XXX - according to Microsoft's spec, at an offset of 0x3c in a
 # PE-format executable is the offset in the file of the PE header;
@@ -142,9 +143,10 @@
 #
 
 #
-# Windows NT Registry files.
+# Windows Registry files.
 #
-0	string		regf		Windows NT Registry file
+0	string		regf		Windows NT registry file
+0	string		CREG		Windows 95 registry file
 
 # Popular applications
 2080	string	Microsoft\ Word\ 6.0\ Document	%s
@@ -155,7 +157,11 @@
 0	belong	0x31be0000			Microsoft Word Document
 #
 0       string  PO^Q`				Microsoft Word 6.0 Document
-#
+
+# I guess this is the one that works:
+# MS-Word Document (Wolfram Kleff)
+0	string	\xD0\xCF\x11\xE0\xA1\xB1\x1A\xE1	Microsoft Word document data
+
 0	string	\376\067\0\043			Microsoft Office Document
 0	string	\320\317\021\340\241\261	Microsoft Office Document
 0	string	\333\245-\0\0\0			Microsoft Office Document
--- file-3.33.orig/Magdir/octave
+++ file-3.33/Magdir/octave
@@ -1,3 +1,4 @@
+
 #------------------------------------------------------------------------------
 # octave binary data file(1) magic, from Dirk Eddelbuettel <edd@debian.org>
 0	string		Octave-1-L	Octave binary data (little endian)
--- file-3.33.orig/Magdir/os2
+++ file-3.33/Magdir/os2
@@ -11,16 +11,16 @@
 # OS/2 URL objects
 # Provided 1998/08/22 by
 # David Mediavilla <davidme.news@REMOVEIFNOTSPAMusa.net>
-0	string	http:			OS/2 URL object text
->5	string	>\			(WWW) <http:%s>
-0	string	mailto:			OS/2 URL object text
->7	string	>\			(email) <%s>
-0	string	news:			OS/2 URL object text
->5	string	>\			(Usenet) <%s>
-0	string	ftp:			OS/2 URL object text
->4	string	>\			(FTP) <ftp:%s>
-0	string	file:			OS/2 URL object text
->5	string	>\			(Local file) <%s>
+#0	string	http:			OS/2 URL object text
+#>5	string	>\			(WWW) <http:%s>
+#0	string	mailto:			OS/2 URL object text
+#>7	string	>\			(email) <%s>
+#0	string	news:			OS/2 URL object text
+#>5	string	>\			(Usenet) <%s>
+#0	string	ftp:			OS/2 URL object text
+#>4	string	>\			(FTP) <ftp:%s>
+#0	string	file:			OS/2 URL object text
+#>5	string	>\			(Local file) <%s>
 
 # >>>>> OS/2 INF/HLP <<<<<  (source: Daniel Dissett ddissett@netcom.com)
 # Carl Hauser (chauser.parc@xerox.com) and 
--- file-3.33.orig/Magdir/ti-8x
+++ file-3.33/Magdir/ti-8x
@@ -1,3 +1,4 @@
+
 # ------------------------------------------------------------------------
 # ti-8x: file(1) magic for the TI-8x and TI-92 Graphing Calculators.
 #
--- file-3.33.orig/Magdir/kleff
+++ file-3.33/Magdir/kleff
@@ -0,0 +1,123 @@
+--- animation~  Fri May 28 05:40:12 1999
++++ animation   Fri May 28 05:40:12 1999
+@@ -12,65 +12,71 @@
+ 0      belong          0x000001ba              MPEG system stream data
+ # MPEG Audio (*.mpx)
+ # from dreesen@math.fu-berlin.de
+-0       beshort         &0xfff0         MP
++0       beshort         &0xffe0         MPEG
+ # MPEG 1.0
+->1      byte&0x08       =0x08           \b
++>1      byte&0x08       =0x08           \b 1.0
+ # Layer 3
+->>1     byte            &0x02           \b3
+->>>2    byte&0xf0       =0x10           \b,  32 kBits
+->>>2    byte&0xf0       =0x20           \b,  40 kBits
+->>>2    byte&0xf0       =0x30           \b,  48 kBits
+->>>2    byte&0xf0       =0x40           \b,  56 kBits
+->>>2    byte&0xf0       =0x50           \b,  64 kBits
+->>>2    byte&0xf0       =0x60           \b,  80 kBits
+->>>2    byte&0xf0       =0x70           \b,  96 kBits
+->>>2    byte&0xf0       =0x80           \b, 112 kBits
+->>>2    byte&0xf0       =0x90           \b, 128 kBits
+->>>2    byte&0xf0       =0xA0           \b, 160 kBits
+->>>2    byte&0xf0       =0xB0           \b, 192 kBits
+->>>2    byte&0xf0       =0xC0           \b, 224 kBits
+->>>2    byte&0xf0       =0xD0           \b, 256 kBits
+->>>2    byte&0xf0       =0xE0           \b, 320 kBits
++>>1     byte            &0x02           \b layer 3 audio stream data
++>>>2    byte&0xf0       =0x10           \b,  32 kBit/s
++>>>2    byte&0xf0       =0x20           \b,  40 kBit/s
++>>>2    byte&0xf0       =0x30           \b,  48 kBit/s
++>>>2    byte&0xf0       =0x40           \b,  56 kBit/s
++>>>2    byte&0xf0       =0x50           \b,  64 kBit/s
++>>>2    byte&0xf0       =0x60           \b,  80 kBit/s
++>>>2    byte&0xf0       =0x70           \b,  96 kBit/s
++>>>2    byte&0xf0       =0x80           \b, 112 kBit/s
++>>>2    byte&0xf0       =0x90           \b, 128 kBit/s
++>>>2    byte&0xf0       =0xA0           \b, 160 kBit/s
++>>>2    byte&0xf0       =0xB0           \b, 192 kBit/s
++>>>2    byte&0xf0       =0xC0           \b, 224 kBit/s
++>>>2    byte&0xf0       =0xD0           \b, 256 kBit/s
++>>>2    byte&0xf0       =0xE0           \b, 320 kBit/s
+ # Layer 2
+->>1     byte            &0x04           \b2
+->>>2    byte&0xf0       =0x10           \b,  32 kBits
+->>>2    byte&0xf0       =0x20           \b,  48 kBits
+->>>2    byte&0xf0       =0x30           \b,  56 kBits
+->>>2    byte&0xf0       =0x40           \b,  64 kBits
+->>>2    byte&0xf0       =0x50           \b,  80 kBits
+->>>2    byte&0xf0       =0x60           \b,  96 kBits
+->>>2    byte&0xf0       =0x70           \b, 112 kBits
+->>>2    byte&0xf0       =0x80           \b, 128 kBits
+->>>2    byte&0xf0       =0x90           \b, 160 kBits
+->>>2    byte&0xf0       =0xA0           \b, 192 kBits
+->>>2    byte&0xf0       =0xB0           \b, 224 kBits
+->>>2    byte&0xf0       =0xC0           \b, 256 kBits
+->>>2    byte&0xf0       =0xD0           \b, 320 kBits
+->>>2    byte&0xf0       =0xE0           \b, 384 kBits
++>>1     byte            &0x04           \b layer 2 audio stream data
++>>>2    byte&0xf0       =0x10           \b,  32 kBit/s
++>>>2    byte&0xf0       =0x20           \b,  48 kBit/s
++>>>2    byte&0xf0       =0x30           \b,  56 kBit/s
++>>>2    byte&0xf0       =0x40           \b,  64 kBit/s
++>>>2    byte&0xf0       =0x50           \b,  80 kBit/s
++>>>2    byte&0xf0       =0x60           \b,  96 kBit/s
++>>>2    byte&0xf0       =0x70           \b, 112 kBit/s
++>>>2    byte&0xf0       =0x80           \b, 128 kBit/s
++>>>2    byte&0xf0       =0x90           \b, 160 kBit/s
++>>>2    byte&0xf0       =0xA0           \b, 192 kBit/s
++>>>2    byte&0xf0       =0xB0           \b, 224 kBit/s
++>>>2    byte&0xf0       =0xC0           \b, 256 kBit/s
++>>>2    byte&0xf0       =0xD0           \b, 320 kBit/s
++>>>2    byte&0xf0       =0xE0           \b, 384 kBit/s
+ # freq
+ >>2     byte&0x0C       =0x00           \b, 44.1 kHz
+ >>2     byte&0x0C       =0x04           \b, 48 kHz
+ >>2     byte&0x0C       =0x08           \b, 32 kHz
++# MPEG 2.5
++>1      byte&0x18       =0x00           \b 2.5
++# Layer 3
++>>1     byte            &0x02           \b layer 3 audio stream data
++# Layer 2
++>>1     byte            &0x04           \b layer 2 audio stream data
+ # MPEG 2.0
+->1      byte&0x08       =0x00           \b
++>1      byte&0x18       =0x10           \b 2.0
+ # Layer 3
+->>1     byte            &0x02           \b3
++>>1     byte            &0x02           \b layer 3 audio stream data
+ # Layer 2
+->>1     byte            &0x04           \b2
+->>2     byte&0xf0       =0x10           \b,   8 kBits
+->>2     byte&0xf0       =0x20           \b,  16 kBits
+->>2     byte&0xf0       =0x30           \b,  24 kBits
+->>2     byte&0xf0       =0x40           \b,  32 kBits
+->>2     byte&0xf0       =0x50           \b,  40 kBits
+->>2     byte&0xf0       =0x60           \b,  48 kBits
+->>2     byte&0xf0       =0x70           \b,  56 kBits
+->>2     byte&0xf0       =0x80           \b,  64 kBits
+->>2     byte&0xf0       =0x90           \b,  80 kBits
+->>2     byte&0xf0       =0xA0           \b,  96 kBits
+->>2     byte&0xf0       =0xB0           \b, 112 kBits
+->>2     byte&0xf0       =0xC0           \b, 128 kBits
+->>2     byte&0xf0       =0xD0           \b, 144 kBits
+->>2     byte&0xf0       =0xE0           \b, 160 kBits
++>>1     byte            &0x04           \b layer 2 audio stream data
++>>2     byte&0xf0       =0x10           \b,   8 kBit/s
++>>2     byte&0xf0       =0x20           \b,  16 kBit/s
++>>2     byte&0xf0       =0x30           \b,  24 kBit/s
++>>2     byte&0xf0       =0x40           \b,  32 kBit/s
++>>2     byte&0xf0       =0x50           \b,  40 kBit/s
++>>2     byte&0xf0       =0x60           \b,  48 kBit/s
++>>2     byte&0xf0       =0x70           \b,  56 kBit/s
++>>2     byte&0xf0       =0x80           \b,  64 kBit/s
++>>2     byte&0xf0       =0x90           \b,  80 kBit/s
++>>2     byte&0xf0       =0xA0           \b,  96 kBit/s
++>>2     byte&0xf0       =0xB0           \b, 112 kBit/s
++>>2     byte&0xf0       =0xC0           \b, 128 kBit/s
++>>2     byte&0xf0       =0xD0           \b, 144 kBit/s
++>>2     byte&0xf0       =0xE0           \b, 160 kBit/s
+ # freq
+ >>2     byte&0x0C       =0x00           \b, 22.05 kHz
+ >>2     byte&0x0C       =0x04           \b, 24 kHz
--- file-3.33.orig/Magdir/wordprocessors
+++ file-3.33/Magdir/wordprocessors
@@ -0,0 +1,13 @@
+
+#------------------------------------------------------------------------------
+# wordprocessors:  file(1) magic fo word processors.
+#
+
+####### PWP file format used on Smith Corona Personal Word Processors:
+2	string	\040\040\040\040\040\040\040\040\040\040\040ML4D\040\'92	Smith Corona PWP
+>24	byte	2	\b, single spaced
+>24	byte	3	\b, 1.5 spaced
+>24	byte	4	\b, double spaced
+>25	byte	0x42	\b, letter
+>25	byte	0x54	\b, legal
+>26	byte	0x46	\b, A4
--- file-3.33.orig/Magdir/misctools
+++ file-3.33/Magdir/misctools
@@ -0,0 +1,6 @@
+
+#-----------------------------------------------------------------------------
+# misctools:  file(1) magic for miscelanous UNIX tools.
+#
+
+0	string		%%!!		X-Post-It-Note text
--- file-3.33.orig/Magdir/communications
+++ file-3.33/Magdir/communications
@@ -0,0 +1,18 @@
+
+# TTCN is the Tree and Tabular Combined Notation described in ISO 9646-3.
+# It is used for conformance testing of communication protocols.
+# Added by W. Borgert <debacle@debian.org>.
+0	string		$Suite			TTCN Abstract Test Suite
+>&1	string		$SuiteId
+>>&1	string		>\n			%s
+>&2	string		$SuiteId
+>>&1	string		>\n			%s
+>&3	string		$SuiteId
+>>&1	string		>\n			%s
+
+# MSC (message sequence charts) are a formal description technique,
+# described in ITU-T Z.120, mainly used for communication protocols.
+# Added by W. Borgert <debacle@debian.org>.
+0	string		mscdocument	Message Sequence Chart (document)
+0	string		msc		Message Sequence Chart (chart)
+0	string		submsc		Message Sequence Chart (subchart)
