diff -puNr file-5.19.orig/src/compress.c file-5.19/src/compress.c
--- file-5.19.orig/src/compress.c	2014-01-05 15:55:21 +0000
+++ file-5.19/src/compress.c	2014-06-15 17:16:01 +0000
@@ -378,7 +378,7 @@ uncompressbuf(struct magic_set *ms, int
 {
 	int fdin[2], fdout[2];
 	ssize_t r;
-	pid_t pid;
+	pid_t pid1 = -1, pid2 = -1;
 
 #ifdef BUILTIN_DECOMPRESS
         /* FIXME: This doesn't cope with bzip2 */
@@ -392,7 +392,7 @@ uncompressbuf(struct magic_set *ms, int
 		file_error(ms, errno, "cannot create pipe");	
 		return NODATA;
 	}
-	switch (pid = fork()) {
+	switch (pid1 = fork()) {
 	case 0:	/* child */
 		(void) close(0);
 		if (fd != -1) {
@@ -436,7 +436,7 @@ uncompressbuf(struct magic_set *ms, int
 			 * fork again, to avoid blocking because both
 			 * pipes filled
 			 */
-			switch (fork()) {
+			switch (pid2 = fork()) {
 			case 0: /* child */
 				(void)close(fdout[0]);
 				if (swrite(fdin[1], old, n) != (ssize_t)n) {
@@ -491,12 +491,8 @@ err:
 		if (fdin[1] != -1)
 			(void) close(fdin[1]);
 		(void) close(fdout[0]);
-#ifdef WNOHANG
-		while (waitpid(pid, NULL, WNOHANG) != -1)
-			continue;
-#else
-		(void)wait(NULL);
-#endif
+		waitpid(pid1, NULL, 0);
+		waitpid(pid2, NULL, 0);
 		(void) close(fdin[0]);
 	    
 		return n;
