--- file-3.41/magic2mime.orig	Mon Feb 16 01:18:53 1998
+++ file-3.41/magic2mime	Wed Mar  5 02:07:59 2003
@@ -1,4 +1,4 @@
-#! /usr/local/bin/perl
+#! /usr/bin/perl -w
 # -*- PERL -*-
 # $Id: magic2mime,v 1.1 1998/02/15 23:18:53 christos Exp $
 # Copyright (c) 1996, 1997 vax@linkdead.paranoia.com (VaX#n8)
@@ -19,50 +19,65 @@
 @mapping =
 (
 	# defaults
-    'data', 'application/octet-stream',
-    'text', 'text/plain',
+    'data' => 'application/octet-stream',
+    'text' => 'text/plain',
 	# more specific
-	'^Rich Text Format data', 'text/richtext',
-	'^HTML document text', 'text/html',
-	'^exported SGML document text', 'text/sgml',
-	'mail text', 'message/rfc822',
-	'news text', 'message/news',
-	'^PostScript document text', 'application/postscript',
-	'^BinHex binary text', 'application/mac-binhex40',
-	'^Zip archive data', 'application/zip',
-	'^Microsoft Word', 'application/msword',
-	'^PGP key', 'application/pgp-keys',
-	'^PGP encrypted', 'application/pgp-encrypted',
-	'^PGP armored data signature', 'application/pgp-signature',
-    '^JPEG image', 'image/jpeg',
-    '^GIF image', 'image/gif',
-	'^PNG image', 'image/png',
-    '^TIFF image', 'image/tiff',
-	'Computer Graphics Metafile', 'image/cgf',
-	'^Sun/NeXT audio data', 'audio/basic',
-    '^MPEG', 'video/mpeg',
-    '^Apple QuickTime movie', 'video/quicktime',
+	'^Rich Text Format data' => 'text/richtext',
+	'^(ASCII |ISO-8859 )?HTML document text' => 'text/html',
+	'^exported SGML document text' => 'text/sgml',
+	'mail text' => 'message/rfc822',
+	'news text' => 'message/news',
+	'^PostScript document text.*type EPS' => 'image/eps',
+	'^PostScript document text' => 'application/postscript',
+	'^PDF document' => 'application/pdf',
+	'^Rich Text Format' => 'application/rtf',
+	'^TeX DVI file' => 'application/x-dvi',
+	'^BinHex binary text' => 'application/mac-binhex40',
+	'^Zip archive data' => 'application/zip',
+	'Microsoft Word[ 0-9.]*document data' => 'application/msword',
+	'^PGP key' => 'application/pgp-keys',
+	'^PGP encrypted' => 'application/pgp-encrypted',
+	'^PGP armored data signature' => 'application/pgp-signature',
+    '^JPEG image' => 'image/jpeg',
+    '^GIF image' => 'image/gif',
+	'^PNG image' => 'image/png',
+    '^TIFF image' => 'image/tiff',
+	'Computer Graphics Metafile' => 'image/cgf',
+	'^Sun/NeXT audio data' => 'audio/basic',
+    '^MPEG.*layer 3 audio' => 'audio/mpeg',
+    '^MPEG' => 'video/mpeg',
+    '^Apple QuickTime movie' => 'video/quicktime',
+    '^X pixmap image' => 'image/x-xpixmap',
 	# made up by me
-    '^bitmap', 'image/x-bitmap',
-    '^PC bitmap data, Windows 3.x format', 'image/x-msw3bmp',
-    '^FLI', 'video/x-fli',
-    '^FLC', 'video/x-flc',
-    'AVI data', 'video/x-avi',
-    'WAVE', 'audio/x-wav',
-    'VOC', 'audio/x-voc',
+    '^bitmap' => 'image/x-bitmap',
+    '^PC bitmap data, Windows 3.x format' => 'image/x-msw3bmp',
+    '^FLI' => 'video/x-fli',
+    '^FLC' => 'video/x-flc',
+    'AVI data' => 'video/x-avi',
+    'WAVE' => 'audio/x-wav',
+    'VOC' => 'audio/x-voc',
+	'Debian binary package' => 'application/x-debian-package',
+	'compiled Java class data' => 'application/x-java',
+	'MPEG.*audio stream data' => 'audio/mpeg',
+	'Standard MIDI data' => 'audio/midi',
 );
 
-local($mimetype,$index,$regexp);
+my($mimetype,$index,$found);
 while (<>)
   {
     chop;
     $index = $#mapping - 1;
-    while ($index > -1 && !defined($mimetype))
+    $found = 0;
+    $mimetype = "application/octet-stream";
+    while ($index > -1 && !$found)
       {
-         $mimetype = $mapping[$index + 1] if (/$mapping[$index]/);
+         if(/$mapping[$index]/)
+         {
+		$mimetype = $mapping[$index + 1];
+		$found=1;
+         }
          $index -= 2;
       }
     print "$mimetype\n";
-	undef $mimetype; # hack
   }
 0;
--- file-3.41/magic2mime.man.orig	Wed Mar  5 02:07:59 2003
+++ file-3.41/magic2mime.man	Wed Mar  5 02:07:59 2003
@@ -0,0 +1,30 @@
+.\" Manpage written by Nicolás Lichtmaier <nick@debian.org>
+.TH FILE __CSECTION__ "January 1999" "Debian/GNU Linux"
+.SH NAME
+file \- determine file type
+.SH SYNOPSIS
+.B magic2mime
+.SH DESCRIPTION
+This manual page documents version __VERSION__ of the
+.B magic2mime
+command.
+.PP
+This command receives the output of the
+.BR file (__CSECTION__)
+command and outputs a posible mime type for the data described;
+e. g.:
+.nf
+.nr 5i
+file -b files... | magic2mime.
+.fi
+.SH SEE ALSO
+.BR file (__CSECTION__)
+.br
+.SH BUGS
+This program depends on exact output from the
+.BR file (__CSECTION__)
+command. Don't use this command, use
+.BR file (1)
+with the
+.B \-i
+option. This command might be removed in the future.
