diff -urp linux-2.6.18-194.8.1.el5.028stab070.4/drivers/message/fusion/mptsas.c linux-2.6.18-194.8.1.el5.028stab070.4-redhat/drivers/message/fusion/mptsas.c
--- linux-2.6.18-194.8.1.el5.028stab070.4/drivers/message/fusion/mptsas.c	2010-08-27 17:10:43 +0000
+++ linux-2.6.18-194.8.1.el5.028stab070.4-redhat/drivers/message/fusion/mptsas.c	2010-08-29 22:03:06 +0000
@@ -1098,10 +1098,8 @@ mptsas_target_reset_queue(MPT_ADAPTER *i
 	id = sas_event_data->TargetID;
 	channel = sas_event_data->Bus;
 
-	if (!(vtarget = mptsas_find_vtarget(ioc, channel, id)))
-		return;
-
-	vtarget->deleted = 1; /* block IO */
+	if ((vtarget = mptsas_find_vtarget(ioc, channel, id)))
+		vtarget->deleted = 1; /* block IO */
 
 	target_reset_list = kzalloc(sizeof(struct mptsas_target_reset_event),
 	    GFP_ATOMIC);
@@ -1207,7 +1205,7 @@ mptsas_taskmgmt_complete(MPT_ADAPTER *io
 	 * enable work queue to remove device from upper layers
 	 */
 	list_del(&target_reset_list->list);
-	if ((mptsas_find_vtarget(ioc, channel, id)) && !ioc->fw_events_off)
+	if (!ioc->fw_events_off)
 		mptsas_queue_device_delete(ioc,
 			&target_reset_list->sas_event_data);
 
diff -urp linux-2.6.18-194.8.1.el5.028stab070.4/drivers/net/cnic.c linux-2.6.18-194.8.1.el5.028stab070.4-redhat/drivers/net/cnic.c
--- linux-2.6.18-194.8.1.el5.028stab070.4/drivers/net/cnic.c	2010-08-27 17:10:52 +0000
+++ linux-2.6.18-194.8.1.el5.028stab070.4-redhat/drivers/net/cnic.c	2010-08-29 22:03:06 +0000
@@ -326,6 +326,12 @@ static int cnic_iscsi_nl_msg_recv(struct
 		if (l5_cid >= MAX_CM_SK_TBL_SZ)
 			break;
 
+		rcu_read_lock();
+		if (!rcu_dereference(cp->ulp_ops[CNIC_ULP_L4])) {
+			rc = -ENODEV;
+			rcu_read_unlock();
+			break;
+		}
 		csk = &cp->csk_tbl[l5_cid];
 		csk_hold(csk);
 		if (cnic_in_use(csk)) {
@@ -340,6 +346,7 @@ static int cnic_iscsi_nl_msg_recv(struct
 				cnic_cm_set_pg(csk);
 		}
 		csk_put(csk);
+		rcu_read_unlock();
 		rc = 0;
 		break;
 	}
diff -urp linux-2.6.18-194.8.1.el5.028stab070.4/drivers/pci/hotplug/acpiphp_core.c linux-2.6.18-194.8.1.el5.028stab070.4-redhat/drivers/pci/hotplug/acpiphp_core.c
--- linux-2.6.18-194.8.1.el5.028stab070.4/drivers/pci/hotplug/acpiphp_core.c	2010-08-27 17:10:10 +0000
+++ linux-2.6.18-194.8.1.el5.028stab070.4-redhat/drivers/pci/hotplug/acpiphp_core.c	2010-08-29 22:03:06 +0000
@@ -307,8 +307,10 @@ static int __init init_acpi(void)
 	/* read initial number of slots */
 	if (!retval) {
 		num_slots = acpiphp_get_num_slots();
-		if (num_slots == 0)
+		if (num_slots == 0) {
+			acpiphp_glue_exit();
 			retval = -ENODEV;
+		}
 	}
 
 	return retval;
diff -urp linux-2.6.18-194.8.1.el5.028stab070.4/fs/cifs/cifssmb.c linux-2.6.18-194.8.1.el5.028stab070.4-redhat/fs/cifs/cifssmb.c
--- linux-2.6.18-194.8.1.el5.028stab070.4/fs/cifs/cifssmb.c	2010-08-27 17:10:43 +0000
+++ linux-2.6.18-194.8.1.el5.028stab070.4-redhat/fs/cifs/cifssmb.c	2010-08-29 22:03:06 +0000
@@ -1588,6 +1588,14 @@ CIFSSMBWrite(const int xid, struct cifsT
 		*nbytes = le16_to_cpu(pSMBr->CountHigh);
 		*nbytes = (*nbytes) << 16;
 		*nbytes += le16_to_cpu(pSMBr->Count);
+
+		/*
+		 * Mask off high 16 bits when bytes written as returned by the
+		 * server is greater than bytes requested by the client. Some
+		 * OS/2 servers are known to set incorrect CountHigh values.
+		 */
+		if (*nbytes > count)
+			*nbytes &= 0xFFFF;
 	}
 
 	cifs_buf_release(pSMB);
@@ -1676,6 +1684,14 @@ CIFSSMBWrite2(const int xid, struct cifs
 		*nbytes = le16_to_cpu(pSMBr->CountHigh);
 		*nbytes = (*nbytes) << 16;
 		*nbytes += le16_to_cpu(pSMBr->Count);
+
+		/*
+		 * Mask off high 16 bits when bytes written as returned by the
+		 * server is greater than bytes requested by the client. OS/2
+		 * servers are known to set incorrect CountHigh values.
+		 */
+		if (*nbytes > count)
+			*nbytes &= 0xFFFF;
 	}
 
 /*	cifs_small_buf_release(pSMB); */ /* Freed earlier now in SendReceive2 */
diff -urp linux-2.6.18-194.8.1.el5.028stab070.4/fs/cifs/dns_resolve.c linux-2.6.18-194.8.1.el5.028stab070.4-redhat/fs/cifs/dns_resolve.c
--- linux-2.6.18-194.8.1.el5.028stab070.4/fs/cifs/dns_resolve.c	2010-08-27 17:10:32 +0000
+++ linux-2.6.18-194.8.1.el5.028stab070.4-redhat/fs/cifs/dns_resolve.c	2010-08-29 22:03:06 +0000
@@ -64,6 +64,9 @@ dns_resolver_instantiate(struct key *key
 	int rc = 0;
 	char *ip;
 
+	if (test_bit(KEY_FLAG_ADDED, &key->flags))
+		return -EACCES;
+
 	ip = kmalloc(datalen + 1, GFP_KERNEL);
 	if (!ip)
 		return -ENOMEM;
diff -urp linux-2.6.18-194.8.1.el5.028stab070.4/fs/ext4/move_extent.c linux-2.6.18-194.8.1.el5.028stab070.4-redhat/fs/ext4/move_extent.c
--- linux-2.6.18-194.8.1.el5.028stab070.4/fs/ext4/move_extent.c	2010-08-27 17:10:44 +0000
+++ linux-2.6.18-194.8.1.el5.028stab070.4-redhat/fs/ext4/move_extent.c	2010-08-29 22:03:06 +0000
@@ -993,6 +993,9 @@ mext_check_arguments(struct inode *orig_
 		return -EINVAL;
 	}
 
+	if (IS_IMMUTABLE(donor_inode) || IS_APPEND(donor_inode))
+		return -EPERM;
+
 	/* Ext4 move extent does not support swapfile */
 	if (IS_SWAPFILE(orig_inode) || IS_SWAPFILE(donor_inode)) {
 		ext4_debug("ext4 move extent: The argument files should "
diff -urp linux-2.6.18-194.8.1.el5.028stab070.4/fs/nfsd/nfs4xdr.c linux-2.6.18-194.8.1.el5.028stab070.4-redhat/fs/nfsd/nfs4xdr.c
--- linux-2.6.18-194.8.1.el5.028stab070.4/fs/nfsd/nfs4xdr.c	2010-08-27 17:10:42 +0000
+++ linux-2.6.18-194.8.1.el5.028stab070.4-redhat/fs/nfsd/nfs4xdr.c	2010-08-29 22:03:06 +0000
@@ -164,10 +164,10 @@ static u32 *read_buf(struct nfsd4_compou
 	argp->p = page_address(argp->pagelist[0]);
 	argp->pagelist++;
 	if (argp->pagelen < PAGE_SIZE) {
-		argp->end = p + (argp->pagelen>>2);
+		argp->end = argp->p + (argp->pagelen>>2);
 		argp->pagelen = 0;
 	} else {
-		argp->end = p + (PAGE_SIZE>>2);
+		argp->end = argp->p + (PAGE_SIZE>>2);
 		argp->pagelen -= PAGE_SIZE;
 	}
 	memcpy(((char*)p)+avail, argp->p, (nbytes - avail));
@@ -1033,10 +1033,10 @@ nfsd4_decode_compound(struct nfsd4_compo
 			argp->p = page_address(argp->pagelist[0]);
 			argp->pagelist++;
 			if (argp->pagelen < PAGE_SIZE) {
-				argp->end = p + (argp->pagelen>>2);
+				argp->end = argp->p + (argp->pagelen>>2);
 				argp->pagelen = 0;
 			} else {
-				argp->end = p + (PAGE_SIZE>>2);
+				argp->end = argp->p + (PAGE_SIZE>>2);
 				argp->pagelen -= PAGE_SIZE;
 			}
 		}
diff -urp linux-2.6.18-194.8.1.el5.028stab070.4/fs/xfs/xfs_dfrag.c linux-2.6.18-194.8.1.el5.028stab070.4-redhat/fs/xfs/xfs_dfrag.c
--- linux-2.6.18-194.8.1.el5.028stab070.4/fs/xfs/xfs_dfrag.c	2010-08-27 17:10:31 +0000
+++ linux-2.6.18-194.8.1.el5.028stab070.4-redhat/fs/xfs/xfs_dfrag.c	2010-08-29 22:03:06 +0000
@@ -62,7 +62,9 @@ xfs_swapext(
 		goto out;
 	}
 
-	if (!(file->f_mode & FMODE_WRITE) || (file->f_flags & O_APPEND)) {
+	if (!(file->f_mode & FMODE_WRITE) ||
+	    !(file->f_mode & FMODE_READ) ||
+	    (file->f_flags & O_APPEND)) {
 		error = XFS_ERROR(EBADF);
 		goto out_put_file;
 	}
@@ -74,6 +76,7 @@ xfs_swapext(
 	}
 
 	if (!(target_file->f_mode & FMODE_WRITE) ||
+	    !(target_file->f_mode & FMODE_READ) ||
 	    (target_file->f_flags & O_APPEND)) {
 		error = XFS_ERROR(EBADF);
 		goto out_put_target_file;
diff -urp linux-2.6.18-194.8.1.el5.028stab070.4/include/linux/key.h linux-2.6.18-194.8.1.el5.028stab070.4-redhat/include/linux/key.h
--- linux-2.6.18-194.8.1.el5.028stab070.4/include/linux/key.h	2010-08-27 17:10:52 +0000
+++ linux-2.6.18-194.8.1.el5.028stab070.4-redhat/include/linux/key.h	2010-08-29 22:03:06 +0000
@@ -147,6 +147,7 @@ struct key {
 #define KEY_FLAG_IN_QUOTA	3	/* set if key consumes quota */
 #define KEY_FLAG_USER_CONSTRUCT	4	/* set if key is being constructed in userspace */
 #define KEY_FLAG_NEGATIVE	5	/* set if key is negative */
+#define KEY_FLAG_ADDED		6	/* set if key is being added via userspace add_key */
 
 	/* the description string
 	 * - this is used to match a key against search criteria
diff -urp linux-2.6.18-194.8.1.el5.028stab070.4/net/bluetooth/l2cap.c linux-2.6.18-194.8.1.el5.028stab070.4-redhat/net/bluetooth/l2cap.c
--- linux-2.6.18-194.8.1.el5.028stab070.4/net/bluetooth/l2cap.c	2010-08-27 17:10:04 +0000
+++ linux-2.6.18-194.8.1.el5.028stab070.4-redhat/net/bluetooth/l2cap.c	2010-08-29 22:03:06 +0000
@@ -2145,16 +2145,24 @@ static ssize_t l2cap_sysfs_show(struct c
 	struct sock *sk;
 	struct hlist_node *node;
 	char *str = buf;
+	int size = PAGE_SIZE;
 
 	read_lock_bh(&l2cap_sk_list.lock);
 
 	sk_for_each(sk, node, &l2cap_sk_list.head) {
 		struct l2cap_pinfo *pi = l2cap_pi(sk);
+		int len;
 
-		str += sprintf(str, "%s %s %d %d 0x%4.4x 0x%4.4x %d %d 0x%x\n",
+		len = snprintf(str, size, "%s %s %d %d 0x%4.4x 0x%4.4x %d %d 0x%x\n",
 				batostr(&bt_sk(sk)->src), batostr(&bt_sk(sk)->dst),
 				sk->sk_state, pi->psm, pi->scid, pi->dcid, pi->imtu,
 				pi->omtu, pi->link_mode);
+
+		size -= len;
+		if (size <= 0)
+			break;
+
+		str += len;
 	}
 
 	read_unlock_bh(&l2cap_sk_list.lock);
diff -urp linux-2.6.18-194.8.1.el5.028stab070.4/net/bluetooth/rfcomm/core.c linux-2.6.18-194.8.1.el5.028stab070.4-redhat/net/bluetooth/rfcomm/core.c
--- linux-2.6.18-194.8.1.el5.028stab070.4/net/bluetooth/rfcomm/core.c	2006-09-20 03:42:06 +0000
+++ linux-2.6.18-194.8.1.el5.028stab070.4-redhat/net/bluetooth/rfcomm/core.c	2010-08-29 22:03:06 +0000
@@ -2027,6 +2027,7 @@ static ssize_t rfcomm_dlc_sysfs_show(str
 	struct rfcomm_session *s;
 	struct list_head *pp, *p;
 	char *str = buf;
+	int size = PAGE_SIZE;
 
 	rfcomm_lock();
 
@@ -2035,11 +2036,21 @@ static ssize_t rfcomm_dlc_sysfs_show(str
 		list_for_each(pp, &s->dlcs) {
 			struct sock *sk = s->sock->sk;
 			struct rfcomm_dlc *d = list_entry(pp, struct rfcomm_dlc, list);
+			int len;
 
-			str += sprintf(str, "%s %s %ld %d %d %d %d\n",
+			len = snprintf(str, size, "%s %s %ld %d %d %d %d\n",
 					batostr(&bt_sk(sk)->src), batostr(&bt_sk(sk)->dst),
 					d->state, d->dlci, d->mtu, d->rx_credits, d->tx_credits);
+
+			size -= len;
+			if (size <= 0)
+				break;
+
+			str += len;
 		}
+
+		if (size <= 0)
+			break;
 	}
 
 	rfcomm_unlock();
diff -urp linux-2.6.18-194.8.1.el5.028stab070.4/net/bluetooth/rfcomm/sock.c linux-2.6.18-194.8.1.el5.028stab070.4-redhat/net/bluetooth/rfcomm/sock.c
--- linux-2.6.18-194.8.1.el5.028stab070.4/net/bluetooth/rfcomm/sock.c	2006-09-20 03:42:06 +0000
+++ linux-2.6.18-194.8.1.el5.028stab070.4-redhat/net/bluetooth/rfcomm/sock.c	2010-08-29 22:03:06 +0000
@@ -890,13 +890,22 @@ static ssize_t rfcomm_sock_sysfs_show(st
 	struct sock *sk;
 	struct hlist_node *node;
 	char *str = buf;
+	int size = PAGE_SIZE;
 
 	read_lock_bh(&rfcomm_sk_list.lock);
 
 	sk_for_each(sk, node, &rfcomm_sk_list.head) {
-		str += sprintf(str, "%s %s %d %d\n",
+		int len;
+
+		len = snprintf(str, size, "%s %s %d %d\n",
 				batostr(&bt_sk(sk)->src), batostr(&bt_sk(sk)->dst),
 				sk->sk_state, rfcomm_pi(sk)->channel);
+
+		size -= len;
+		if (size <= 0)
+			break;
+
+		str += len;
 	}
 
 	read_unlock_bh(&rfcomm_sk_list.lock);
diff -urp linux-2.6.18-194.8.1.el5.028stab070.4/net/bluetooth/sco.c linux-2.6.18-194.8.1.el5.028stab070.4-redhat/net/bluetooth/sco.c
--- linux-2.6.18-194.8.1.el5.028stab070.4/net/bluetooth/sco.c	2006-09-20 03:42:06 +0000
+++ linux-2.6.18-194.8.1.el5.028stab070.4-redhat/net/bluetooth/sco.c	2010-08-29 22:03:06 +0000
@@ -895,13 +895,22 @@ static ssize_t sco_sysfs_show(struct cla
 	struct sock *sk;
 	struct hlist_node *node;
 	char *str = buf;
+	int size = PAGE_SIZE;
 
 	read_lock_bh(&sco_sk_list.lock);
 
 	sk_for_each(sk, node, &sco_sk_list.head) {
-		str += sprintf(str, "%s %s %d\n",
+		int len;
+
+		len = snprintf(str, size, "%s %s %d\n",
 				batostr(&bt_sk(sk)->src), batostr(&bt_sk(sk)->dst),
 				sk->sk_state);
+
+		size -= len;
+		if (size <= 0)
+			break;
+
+		str += len;
 	}
 
 	read_unlock_bh(&sco_sk_list.lock);
diff -urp linux-2.6.18-194.8.1.el5.028stab070.4/net/ipv4/tcp_input.c linux-2.6.18-194.8.1.el5.028stab070.4-redhat/net/ipv4/tcp_input.c
--- linux-2.6.18-194.8.1.el5.028stab070.4/net/ipv4/tcp_input.c	2010-08-27 17:10:52 +0000
+++ linux-2.6.18-194.8.1.el5.028stab070.4-redhat/net/ipv4/tcp_input.c	2010-08-29 22:03:06 +0000
@@ -129,7 +129,7 @@ static void tcp_measure_rcv_mss(struct s
 	/* skb->len may jitter because of SACKs, even if peer
 	 * sends good full-sized frames.
 	 */
-	len = skb->len;
+	len = skb_shinfo(skb)->gso_size ?: skb->len;
 	if (len >= icsk->icsk_ack.rcv_mss) {
 		icsk->icsk_ack.rcv_mss = len;
 	} else {
diff -urp linux-2.6.18-194.8.1.el5.028stab070.4/security/keys/key.c linux-2.6.18-194.8.1.el5.028stab070.4-redhat/security/keys/key.c
--- linux-2.6.18-194.8.1.el5.028stab070.4/security/keys/key.c	2010-08-27 17:10:27 +0000
+++ linux-2.6.18-194.8.1.el5.028stab070.4-redhat/security/keys/key.c	2010-08-29 22:03:06 +0000
@@ -826,6 +826,9 @@ key_ref_t key_create_or_update(key_ref_t
 		goto error_3;
 	}
 
+	/* this is an unsolicited add_key() call from userspace */
+	set_bit(KEY_FLAG_ADDED, &key->flags);
+
 	/* instantiate it and link it into the target keyring */
 	ret = __key_instantiate_and_link(key, payload, plen, keyring, NULL);
 	if (ret < 0) {
