diff -ur diffutils-2.7.orig/sdiff.c diffutils-2.7/sdiff.c
--- diffutils-2.7.orig/sdiff.c	Sat Oct  1 08:14:15 1994
+++ diffutils-2.7/sdiff.c	Thu Jan  4 09:22:51 2001
@@ -20,6 +20,7 @@
 /* GNU SDIFF was written by Thomas Lord. */
 
 #include "system.h"
+#define _GNU_SOURCE
 #include <stdio.h>
 #include <signal.h>
 #include "getopt.h"
@@ -935,13 +936,27 @@
 	case 'q':
 	  return 0;
 	case 'e':
-	  if (! tmpname && ! (tmpname = private_tempnam ()))
-	    perror_fatal ("temporary file name");
+	  {
+	    int tfd;
+	    FILE *tmp;
 
-	  tmpmade = 1;
+	    if (tmpmade)
+	      {
+	        unlink (tmpname);
+	        tmpmade = 0;
+		free (tmpname);
+	      }
 
-	  {
-	    FILE *tmp = ck_fopen (tmpname, "w+");
+	    asprintf (&tmpname, "%s/sdiff.XXXXXX",
+	      getenv("TMPDIR") ?: P_tmpdir);
+	    tfd = mkstemp(tmpname);
+	    if (tfd == -1)
+	      perror_fatal ("temporary file name");
+	    tmp = fdopen (tfd, "w+");
+	    if (tmp == NULL)
+	      perror_fatal ("temporary file name");
+ 
+	    tmpmade = 1;
 
 	    if (cmd1 == 'l' || cmd1 == 'b')
 	      lf_copy (left, lenl, tmp);
