--- diffutils-3.0.orig/src/diff.c	2010-08-25 06:05:36 +0000
+++ diffutils-3.0/src/diff.c	2010-08-25 06:06:06 +0000
@@ -636,7 +636,7 @@ main (int argc, char **argv)
   if (output_style != OUTPUT_CONTEXT || hard_locale (LC_TIME))
     {
 #if (defined STAT_TIMESPEC || defined STAT_TIMESPEC_NS \
-     || defined HAVE_STRUCT_STAT_ST_SPARE1)
+     || defined HAVE_STRUCT_STAT_ST_SPARE1) && 0
       time_format = "%Y-%m-%d %H:%M:%S.%N %z";
 #else
       time_format = "%Y-%m-%d %H:%M:%S %z";
