--- extensions/libipt_MASQUERADE.c
+++ extensions/libipt_MASQUERADE.c
@@ -56,7 +56,7 @@ parse_ports(const char *arg, struct nf_nat_multi_range *mr)
 		int maxport;
 
 		maxport = atoi(dash + 1);
-		if (maxport == 0 || maxport > 65535)
+		if (maxport <= 0 || maxport > 65535)
 			xtables_error(PARAMETER_PROBLEM,
 				   "Port `%s' not valid\n", dash+1);
 		if (maxport < port)
--- extensions/libipt_REDIRECT.c
+++ extensions/libipt_REDIRECT.c
@@ -51,7 +51,7 @@ parse_ports(const char *arg, struct nf_nat_multi_range *mr)
 	if (port == 0)
 		port = xtables_service_to_port(arg, NULL);
 
-	if (port == 0 || port > 65535)
+	if (port <= 0 || port > 65535)
 		xtables_error(PARAMETER_PROBLEM, "Port \"%s\" not valid\n", arg);
 
 	dash = strchr(arg, '-');
@@ -63,7 +63,7 @@ parse_ports(const char *arg, struct nf_nat_multi_range *mr)
 		int maxport;
 
 		maxport = atoi(dash + 1);
-		if (maxport == 0 || maxport > 65535)
+		if (maxport <= 0 || maxport > 65535)
 			xtables_error(PARAMETER_PROBLEM,
 				   "Port `%s' not valid\n", dash+1);
 		if (maxport < port)
