--- libiptc/libip4tc.c
+++ libiptc/libip4tc.c
@@ -124,7 +124,7 @@ typedef unsigned int socklen_t;
 
 #define IP_PARTS(n) IP_PARTS_NATIVE(ntohl(n))
 
-int
+static int
 dump_entry(STRUCT_ENTRY *e, struct iptc_handle *const handle)
 {
 	size_t i;
--- libiptc/libiptc.c
+++ libiptc/libiptc.c
@@ -1563,6 +1563,7 @@ TC_NEXT_RULE(const STRUCT_ENTRY *prev, struct xtc_handle *handle)
 	return r->entry;
 }
 
+#if 0
 /* How many rules in this chain? */
 static unsigned int
 TC_NUM_RULES(const char *chain, struct xtc_handle *handle)
@@ -1601,6 +1602,7 @@ TC_GET_RULE(const char *chain, unsigned int n, struct xtc_handle *handle)
 		return NULL;
 	return r->entry;
 }
+#endif
 
 /* Returns a pointer to the target name of this position. */
 static const char *standard_target_map(int verdict)
