--- gettext-0.14.5/gettext-tools/misc/gettextize.in.orig	2005-05-20 20:36:38 +0000
+++ gettext-0.14.5/gettext-tools/misc/gettextize.in	2006-04-06 14:03:34 +0000
@@ -155,6 +155,7 @@ Options:
       --version        print version information and exit
   -c, --copy           copy files instead of making symlinks
   -f, --force          force writing of new files even if old exist
+  -q, --quiet          don't ask acknowledgements
       --intl           install libintl in a subdirectory
       --no-changelog   don't update or create ChangeLog files
   -n, --dry-run        print modifications but don't perform them
@@ -194,6 +195,7 @@ func_fatal_error ()
 # - doit            false if --dry-run was given, : otherwise
 {
   force=0
+  quiet=0
   intldir=
   try_ln_s=:
   do_changelog=:
@@ -210,6 +212,9 @@ func_fatal_error ()
       -f | --force | --forc | --for | --fo | --f )
         shift
         force=1 ;;
+      -q | --quiet | --quie | --qui | --qu | --q )
+        shift
+        quiet=1 ;;
       --help | --hel | --he | --h )
         func_usage; exit 0 ;;
       --intl | --int | --in | --i )
@@ -1204,10 +1209,12 @@ if $doit; then
     5) count="five paragraphs";;
     *) count="$count paragraphs";;
   esac
+  if test $quiet -eq 0; then
   echo "Press Return to acknowledge the previous $count."
   # Read from /dev/tty, not stdin, so that gettextize cannot be abused by
   # non-interactive tools.
   read dummy < /dev/tty
+  fi
 fi
 
 exit 0
