diff -puNr gettext-0.19.1.orig/gettext-tools/gnulib-tests/test-term-ostream-xterm gettext-0.19.1/gettext-tools/gnulib-tests/test-term-ostream-xterm
--- gettext-0.19.1.orig/gettext-tools/gnulib-tests/test-term-ostream-xterm	2013-01-09 04:01:27 +0000
+++ gettext-0.19.1/gettext-tools/gnulib-tests/test-term-ostream-xterm	2014-06-14 19:25:13 +0000
@@ -1,5 +1,6 @@
 #!/bin/sh
 
+exit 77 # skip until our terminfo is updated
 tmpfiles=""
 trap 'rm -fr $tmpfiles' 1 2 3 15
 
