--- elfutils-0.108/src/findtextrel.c.orig	Tue May 24 23:41:40 2005
+++ elfutils-0.108/src/findtextrel.c	Tue May 24 23:41:54 2005
@@ -476,7 +476,7 @@ ptrcompare (const void *p1, const void *


 static void
-check_rel (size_t nsegments, struct segments segments[nsegments],
+check_rel (size_t nsegments, struct segments segments[],
 	   GElf_Addr addr, Elf *elf, Elf_Scn *symscn, Dwarf *dw,
 	   const char *fname, bool more_than_one, void **knownsrcs)
 {
