diff -uprk.orig vixie-cron-4.1.20040916.orig/usr.sbin/cron/cron.c vixie-cron-4.1.20040916/usr.sbin/cron/cron.c
--- vixie-cron-4.1.20040916.orig/usr.sbin/cron/cron.c	2004-11-03 01:37:36 +0300
+++ vixie-cron-4.1.20040916/usr.sbin/cron/cron.c	2004-11-03 01:38:00 +0300
@@ -76,7 +76,9 @@ main(int argc, char *argv[]) {
 	ProgramName = program_invocation_short_name;
 	FixedName = "crond";
 
+#ifdef HAVE_SETLOCALE
 	setlocale(LC_ALL, "");
+#endif
 
 #if defined(BSD)
 	setlinebuf(stdout);
diff -uprk.orig vixie-cron-4.1.20040916.orig/usr.sbin/cron/crontab.c vixie-cron-4.1.20040916/usr.sbin/cron/crontab.c
--- vixie-cron-4.1.20040916.orig/usr.sbin/cron/crontab.c	2004-11-03 01:37:36 +0300
+++ vixie-cron-4.1.20040916/usr.sbin/cron/crontab.c	2004-11-03 01:38:00 +0300
@@ -83,7 +83,9 @@ main(int argc, char *argv[]) {
 	ProgramName = argv[0];
 	FixedName = "crontab";
 
+#ifdef HAVE_SETLOCALE
 	setlocale(LC_ALL, "");
+#endif
 
 #if defined(BSD)
 	setlinebuf(stderr);
