--- vixie-cron-3.0.2.7/usr.sbin/cron/do_command.c.orig	Wed Jan 29 00:31:14 2003
+++ vixie-cron-3.0.2.7/usr.sbin/cron/do_command.c	Wed Jan 29 00:33:02 2003
@@ -103,20 +103,12 @@
 	usernm = env_get("LOGNAME", e->envp);
 	mailto = env_get("MAILTO", e->envp);
 
-#ifdef USE_SIGCHLD
 	/* our parent is watching for our death by catching SIGCHLD.  we
 	 * do not care to watch for our children's deaths this way -- we
-	 * use wait() explictly.  so we have to disable the signal (which
+	 * use wait() explictly.  so we have to reset the signal (which
 	 * was inherited from the parent).
 	 */
-	(void) signal(SIGCHLD, SIG_IGN);
-#else
-	/* on system-V systems, we are ignoring SIGCLD.  we have to stop
-	 * ignoring it now or the wait() in cron_pclose() won't work.
-	 * because of this, we have to wait() for our children here, as well.
-	 */
 	(void) signal(SIGCLD, SIG_DFL);
-#endif /*BSD*/
 
 	/* create some pipes to talk to our future child
 	 */
