diff -uprk.orig vixie-cron-4.1.20040916.orig/usr.sbin/cron/atrun.c vixie-cron-4.1.20040916/usr.sbin/cron/atrun.c
--- vixie-cron-4.1.20040916.orig/usr.sbin/cron/atrun.c	2004-11-03 01:27:46 +0300
+++ vixie-cron-4.1.20040916/usr.sbin/cron/atrun.c	2004-11-03 01:28:00 +0300
@@ -492,6 +492,9 @@ run_job(atjob *job, char *atfile)
 		}
 #endif /*DEBUGGING*/
 
+		(void) signal(SIGPIPE, SIG_DFL);
+		(void) signal(SIGUSR1, SIG_DFL);
+
 		/*
 		 * Exec /bin/sh with stdin connected to the at job file
 		 * and stdout/stderr hooked up to our parent.
diff -uprk.orig vixie-cron-4.1.20040916.orig/usr.sbin/cron/do_command.c vixie-cron-4.1.20040916/usr.sbin/cron/do_command.c
--- vixie-cron-4.1.20040916.orig/usr.sbin/cron/do_command.c	2004-11-03 01:27:46 +0300
+++ vixie-cron-4.1.20040916/usr.sbin/cron/do_command.c	2004-11-03 01:28:00 +0300
@@ -253,6 +253,9 @@ child_process(entry *e, user *u) {
 #endif /* LOGIN_CAP */
 		chdir(env_get("HOME", e->envp));
 
+		(void) signal(SIGPIPE, SIG_DFL);
+		(void) signal(SIGUSR1, SIG_DFL);
+
 		/*
 		 * Exec the command.
 		 */
