diff -ur vixie-cron-3.0.2.7.linux/usr.sbin/cron/pathnames.h vixie-cron-3.0.2.7/usr.sbin/cron/pathnames.h
--- vixie-cron-3.0.2.7.linux/usr.sbin/cron/pathnames.h	Sat Aug 19 11:24:52 2000
+++ vixie-cron-3.0.2.7/usr.sbin/cron/pathnames.h	Sun Apr 21 05:26:41 2002
@@ -66,11 +66,7 @@
 			/* what editor to use if no EDITOR or VISUAL
 			 * environment variable specified.
 			 */
-#if defined(_PATH_VI)
-# define EDITOR _PATH_VI
-#else
-# define EDITOR "/usr/ucb/vi"
-#endif
+#define EDITOR "/bin/vitmp"
 
 #ifndef _PATH_BSHELL
 # define _PATH_BSHELL "/bin/sh"
