diff -puNr vixie-cron-3.0.2.7.orig/usr.sbin/cron/crontab.c vixie-cron-3.0.2.7/usr.sbin/cron/crontab.c
--- vixie-cron-3.0.2.7.orig/usr.sbin/cron/crontab.c	Fri Jan  7 02:39:56 2005
+++ vixie-cron-3.0.2.7/usr.sbin/cron/crontab.c	Fri Jan  7 02:45:38 2005
@@ -127,6 +127,7 @@ main(argc, argv)
 				break;
 	case opt_unknown:
 		/* can't happen */
+		;
 	}
 	exit(0);
 	/*NOTREACHED*/
@@ -554,7 +555,7 @@ replace_cmd() {
 		return (-2);
 	}
 
-	TempFilename = strdup(__CONCAT(SPOOL_DIR,"/tmp.XXXXXX"));
+	TempFilename = strdup(SPOOL_DIR "/tmp.XXXXXX");
 	if (TempFilename == NULL) {
 		fprintf(stderr, "%s: Cannot allocate memory.\n", ProgramName);
 		return (-2);
