--- vixie-cron-4.1.20060426/usr.sbin/cron/database.c.orig
+++ vixie-cron-4.1.20060426/usr.sbin/cron/database.c
@@ -230,7 +230,7 @@ process_crontab(const char *uname, const
 		log_it(fname, getpid(), "BAD FILE MODE", tabname);
 		goto next_crontab;
 	}
-	if (lstatbuf.st_nlink != 1) {
+	if (pw && lstatbuf.st_nlink != 1) {
 		log_it(fname, getpid(), "BAD LINK COUNT", tabname);
 		goto next_crontab;
 	}
@@ -260,7 +260,7 @@ process_crontab(const char *uname, const
 		log_it(fname, getpid(), "WRONG FILE OWNER", tabname);
 		goto next_crontab;
 	}
-	if (statbuf->st_nlink != 1) {
+	if (pw && statbuf->st_nlink != 1) {
 		log_it(fname, getpid(), "BAD LINK COUNT", tabname);
 		goto next_crontab;
 	}
