diff -urN textutils-2.0e.org/lib/Makefile.am textutils-2.0e/lib/Makefile.am
--- textutils-2.0e.org/lib/Makefile.am	Sun Mar  5 00:09:51 2000
+++ textutils-2.0e/lib/Makefile.am	Sun Jul 30 01:38:37 2000
@@ -61,7 +61,7 @@
 
 install-exec-local: all-local
 	$(mkinstalldirs) $(DESTDIR)$(libdir)
-	$(INSTALL_DATA) charset.alias $(libdir)/charset.alias
+	$(INSTALL_DATA) charset.alias $(DESTDIR)$(libdir)/charset.alias
 
 # Probably not a good idea. Other packages need $(libdir)/charset.alias too.
 uninstall-local:
diff -urN textutils-2.0e.org/lib/Makefile.in textutils-2.0e/lib/Makefile.in
--- textutils-2.0e.org/lib/Makefile.in	Sun Mar  5 00:10:26 2000
+++ textutils-2.0e/lib/Makefile.in	Sun Jul 30 01:38:44 2000
@@ -785,7 +636,7 @@
 
 install-exec-local: all-local
 	$(mkinstalldirs) $(DESTDIR)$(libdir)
-	$(INSTALL_DATA) charset.alias $(libdir)/charset.alias
+	$(INSTALL_DATA) charset.alias $(DESTDIR)$(libdir)/charset.alias
 
 # Probably not a good idea. Other packages need $(libdir)/charset.alias too.
 uninstall-local:
