diff -ur man-1.5l.orig/src/makewhatis.sh man-1.5l/src/makewhatis.sh
--- man-1.5l.orig/src/makewhatis.sh	2003-03-10 20:14:51 +0000
+++ man-1.5l/src/makewhatis.sh	2005-10-24 07:12:37 +0000
@@ -3,6 +3,7 @@
 # Created: Sun Jun 14 10:49:37 1992
 # Revised: Sat Jan  8 14:12:37 1994 by faith@cs.unc.edu
 # Revised: Sat Mar 23 17:56:18 1996 by micheal@actrix.gen.nz
+# (please see below for further revisions)
 # Copyright 1992, 1993, 1994 Rickard E. Faith (faith@cs.unc.edu)
 # May be freely distributed and modified as long as copyright is retained.
 #
@@ -25,15 +26,19 @@
 # 971012 - replaced "test -z" - it doesnt work on SunOS 4.1.3_U1.
 # 980710 - be more careful with TMPFILE
 # 000323 - do not change PATH, better treatment of catpages - Bryan Henderson
+# 000810 - use mktemp, keep whatis files consistent while running - solar@owl
+# 010518 - /usr/share/man is optional - solar@owl
 # 011117 - avoid suspicious filenames
+# 020920 - use the new mktemp -t - solar@owl
 # 030310 - find files only; fix LAPACK cruft; no /usr/man default;
 #	use /dev/stderr instead of /dev/tty; handle files with strange names;
 #	add support for chinese, hungarian, indonesian, japanese, korean,
 #	polish, russian (Thierry Vignaud); 
+# 051024 - mktemp first, set the trap later to not trigger in-shell races
 #
 # Note for Slackware users: "makewhatis -v -w -c" will work.
 #
-# makewhatis aeb 030310 (from %version%)
+# makewhatis aeb 030310 / Owl 051024 (from %version%)
 
 program=`basename $0`
 
@@ -63,24 +68,10 @@
 # AWK=/usr/bin/gawk
 AWK=%awk%
 
-# Find a place for our temporary files. If security is not a concern, use
-#	TMPFILE=/tmp/whatis$$; TMPFILEDIR=none
-# Of course makewhatis should only have the required permissions
-# (for reading and writing directories like /usr/man).
-# We try here to be careful (and avoid preconstructed symlinks)
-# in case makewhatis is run as root, by creating a subdirectory of /tmp.
-
-TMPFILEDIR=/tmp/whatis.tmp.dir.$$
-rm -rf TMPFILEDIR
-if ! mkdir -m 0700 $TMPFILEDIR; then
-    echo Could not create $TMPFILEDIR
-    exit 1;
-fi
-TMPFILE=$TMPFILEDIR/w
-
-# make sure TMPFILEDIR is deleted if program is killed or terminates
-# (just delete this line if your shell doesnt know about trap)
-trap "rm -rf $TMPFILEDIR" 0 1 2 3 15
+# make sure TMPFILE is deleted if program is killed or terminates
+TMPFILE="`mktemp -t $program.XXXXXXXXXX`" || exit
+trap 'rm -f -- "$TMPFILE"' EXIT
+trap 'trap - EXIT; rm -f -- "$TMPFILE"; exit 1' HUP INT QUIT TERM
 
 # default find arg: no directories, no empty files
 findarg0="-type f -size +0"
@@ -146,7 +137,7 @@
 fi
 catpath=`echo ${catpath} | tr : ' '`
 
-# first truncate all the whatis files that will be created new,
+# first mark all the whatis files that will be created new,
 # then only update - we might visit the same directory twice
 if [ x$update = x ]; then
    for pages in man cat
@@ -154,7 +145,7 @@
       eval path="\$$pages"path
       for mandir in $path
       do
-	 cp /dev/null $mandir/whatis
+	 touch $mandir/whatis.update
       done
    done
 fi
@@ -168,7 +159,7 @@
      if [ x$verbose != x ]; then
 	echo "about to enter $mandir" > /dev/stderr
      fi
-     if [ -s ${mandir}/whatis -a $pages = man -a x$update = x ]; then
+     if [ ! -f ${mandir}/whatis.update -a $pages = man -a x$update = x ]; then
 	if [ x$verbose != x ]; then
 	   echo skipping $mandir - we did it already > /dev/stderr
 	fi
@@ -235,12 +226,12 @@
 	      sub(/\..*/, "", progname);
 	      if (use_zcat || use_bzcat) {
 		if (use_zcat) {
-		  pipe_cmd = "zcat \"" filename "\"";
+		  pipe_cmd = "zcat '"'"'" filename "'"'"'";
 		} else {
-		  pipe_cmd = "bzcat \"" filename "\"";
+		  pipe_cmd = "bzcat '"'"'" filename "'"'"'";
 		}
 		# try to avoid suspicious stuff
-		if (filename ~ /[;&|`$(]/) {
+		if (filename ~ /[;&|`'"'"'"$(]/) {
 		  print "ignored strange file name " filename " in " curdir > "/dev/stderr";
 		  return;
 		}
@@ -387,7 +378,7 @@
 	    ' pages=$pages section=$section verbose=$verbose curdir=$curdir
 	    cd ..
 	 fi
-       done > $TMPFILE
+       done > "$TMPFILE"
 
        cd $here
 
@@ -401,16 +392,14 @@
 
        if [ -f ${mandir1}/whatis ]
        then
-	 cat ${mandir1}/whatis >> $TMPFILE
+	 cat ${mandir1}/whatis >> "$TMPFILE"
        fi
-       sed '/^$/d' < $TMPFILE | sort | uniq > ${mandir1}/whatis
+       touch ${mandir1}/whatis.tmp
+       chmod 644 ${mandir1}/whatis.tmp
+       sed '/^$/d' < "$TMPFILE" | sort -u > ${mandir1}/whatis.tmp
 
-       chmod 644 ${mandir1}/whatis
-       rm $TMPFILE
+       mv -f ${mandir1}/whatis.tmp ${mandir1}/whatis
+       rm ${mandir1}/whatis.update
      fi
    done
 done
-
-# remove tempdir
-rm -rf $TMPFILEDIR
-
