--- man-1.6b/configure.orig	2005-08-20 23:26:06.000000000 +0000
+++ man-1.6b/configure	2005-12-24 00:18:31.000000000 +0000
@@ -474,12 +474,7 @@
     troff="troff -mandoc"
     echo "Warning: could not find groff"
   else
-    if test $Fnroff = "missing"
-    then
-      nroff="$Fgroff -Tlatin1 -mandoc"
-    else
-      nroff="$Fnroff -Tlatin1 -mandoc"
-    fi
+    nroff="$Fgroff -Tlatin1 -mandoc"
     troff="$Fgroff -Tps -mandoc"
     jnroff="$Fgroff -Tnippon -mandocj"
   fi
