diff -ur man-1.5l.orig/configure man-1.5l/configure
--- man-1.5l.orig/configure	Mon Mar 10 20:21:00 2003
+++ man-1.5l/configure	Sun Mar 16 19:43:08 2003
@@ -480,12 +480,7 @@
     troff="troff -mandoc"
     echo "Warning: could not find groff"
   else
-    if test $Fnroff = "missing"
-    then
-      nroff="$Fgroff -Tlatin1 -mandoc"
-    else
-      nroff="$Fnroff -Tlatin1 -mandoc"
-    fi
+    nroff="$Fgroff -Tlatin1 -mandoc"
     troff="$Fgroff -Tps -mandoc"
     jnroff="$Fgroff -Tnippon -mandocj"
   fi
