diff -ur man-1.5h1.orig/man/Makefile.in man-1.5h1/man/Makefile.in
--- man-1.5h1.orig/man/Makefile.in	Tue Apr  6 05:05:20 1999
+++ man-1.5h1/man/Makefile.in	Thu Aug 10 01:43:09 2000
@@ -26,13 +26,13 @@
 		do if test -f $$i.man; then make -f ../Makefile $$i.8; fi; done
 
 install: $(ALL)
-	mkdir -p $(mandir)/man1 $(mandir)/man5 $(mandir)/man8
+	mkdir -p $(PREFIX)$(mandir)/man1 $(PREFIX)$(mandir)/man5 $(PREFIX)$(mandir)/man8
 	for i in $(MAN1); \
-		do $(INSTALL) $$i.1 $(mandir)/man1/$$i.@man1ext@; done
+		do $(INSTALL) $$i.1 $(PREFIX)$(mandir)/man1/$$i.@man1ext@; done
 	for i in $(MAN5); \
-		do $(INSTALL) $$i.5 $(mandir)/man5/$$i.@man5ext@; done
+		do $(INSTALL) $$i.5 $(PREFIX)$(mandir)/man5/$$i.@man5ext@; done
 	for i in $(MAN8); \
-		do if test -f $$i.8; then $(INSTALL) $$i.8 $(mandir)/man8/$$i.@man8ext@; fi; done
+		do if test -f $$i.8; then $(INSTALL) $$i.8 $(PREFIX)$(mandir)/man8/$$i.@man8ext@; fi; done
 
 clean:
 	rm -f core *.in *.@man1ext@ *.@man5ext@ *~
diff -ur man-1.5h1.orig/man2html/Makefile.in man-1.5h1/man2html/Makefile.in
--- man-1.5h1.orig/man2html/Makefile.in	Tue Jun 29 03:30:33 1999
+++ man-1.5h1/man2html/Makefile.in	Thu Aug 10 01:43:09 2000
@@ -11,11 +11,12 @@
 
 # This installs the man2html utility
 install:	man2html
-	install -m 755 man2html /usr/bin
-	mkdir -p $(mandir)/man1
-	install -m 644 man2html.1 $(mandir)/man1/man2html.@man1ext@
+	mkdir -p $(PREFIX)/usr/bin
+	install -m 755 man2html $(PREFIX)/usr/bin
+	mkdir -p $(PREFIX)/$(mandir)/man1
+	install -m 644 man2html.1 $(PREFIX)/$(mandir)/man1/man2html.@man1ext@
 
-install-scripts: install-man-scripts install-glimpse-stuff install-hman
+install-scripts: install-hman
 
 # These are the scripts that allow pointing a browser at
 #   http://localhost/cgi-bin/man/man2html
@@ -45,8 +46,8 @@
 	sed -e 's,%version%,@version@,' hman.sh > hman
 
 install-hman: hman
-	install -m 555 hman /usr/bin/hman
-	install -m 644 hman.1 $(mandir)/man1/hman.@man1ext@
+	install -m 555 hman $(PREFIX)/usr/bin/hman
+	install -m 644 hman.1 $(PREFIX)/$(mandir)/man1/hman.@man1ext@
 
 clean:
 	rm -f core hman man2html $(OBJECTS) *~
diff -ur man-1.5h1.orig/src/Makefile.in man-1.5h1/src/Makefile.in
--- man-1.5h1.orig/src/Makefile.in	Tue Jun 29 03:27:46 1999
+++ man-1.5h1/src/Makefile.in	Thu Aug 10 01:43:09 2000
@@ -98,13 +98,12 @@
 MANCONFIG=$(BINROOTDIR)@man_config_file@
 
 install: all apropos whatis makewhatis
-	$(INSTALL) -c @man_install_flags@ man $(BINROOTDIR)@man@
-	$(INSTALL) -c -m 755 apropos $(BINROOTDIR)@apropos@
-	$(INSTALL) -c -m 755 whatis $(BINROOTDIR)@whatis@
-	$(INSTALL) -c -m 754 makewhatis $(BINROOTDIR)@makewhatis@
-	$(INSTALL) -d $(BINROOTDIR)@man_config_dir@
-	if [ -f $(MANCONFIG) ]; then mv $(MANCONFIG) $(MANCONFIG).orig; fi
-	$(INSTALL) -c -m 644 man.conf $(MANCONFIG)
+	$(INSTALL) -c @man_install_flags@ man $(PREFIX)@man@
+	$(INSTALL) -c -m 755 apropos $(PREFIX)@apropos@
+	$(INSTALL) -c -m 755 whatis $(PREFIX)@whatis@
+	$(INSTALL) -c -m 754 makewhatis $(PREFIX)@makewhatis@
+	$(INSTALL) -d $(PREFIX)@man_config_dir@
+	$(INSTALL) -c -m 644 man.conf $(PREFIX)$(MANCONFIG)
 
 clean:
 	rm -f *.o *~ core man apropos whatis makewhatis makemsg
