diff -ur gpm-1.19.6.orig/doc/Makefile.in gpm-1.19.6/doc/Makefile.in
--- gpm-1.19.6.orig/doc/Makefile.in	Thu Oct  4 00:51:46 2001
+++ gpm-1.19.6/doc/Makefile.in	Sat Oct  6 03:31:24 2001
@@ -68,7 +68,7 @@
 
 # Main portion
 
-all: $(srcdir)/gpmdoc.ps $(srcdir)/gpm.info $(srcdir)/gpmdoc.txt $(MANPAGES)
+all: $(srcdir)/gpm.info $(srcdir)/gpmdoc.txt $(MANPAGES)
 
 $(MANPAGES): doc.gpm manpager
 	gawk -f ./manpager doc.gpm
