--- gpm-1.20.1.orig/configure.in	2006-09-05 03:22:16 +0000
+++ gpm-1.20.1/configure.in	2006-09-05 03:29:59 +0000
@@ -59,7 +59,10 @@ else
         lispdir='${datadir}/emacs/site-lisp'
 fi
 
-AC_CHECK_HEADERS(syslog.h linux/input.h linux/joystick.h ncurses.h ncurses/curses.h curses.h)
+AC_CHECK_HEADERS(syslog.h ncurses.h ncurses/curses.h curses.h)
+dnl XXX: (GM): hack to check when linux/input.h gets fixed
+_AC_CHECK_HEADER_NEW(linux/input.h)
+_AC_CHECK_HEADER_NEW(linux/joystick.h)
 
 AC_ARG_WITH(curses,
 [  --without-curses        disable curses support even if curses found])
