--- links-0.95/configure.in.orig	Wed Jun  6 00:52:22 2001
+++ links-0.95/configure.in	Thu Jun  7 10:47:11 2001
@@ -140,12 +140,16 @@
 			AC_TRY_LINK([#include <openssl/ssl.h>], [OpenSSL_add_all_algorithms()], cf_result=yes, cf_result=no)
 		fi
 	done
-fi
-if test "$cf_result" != yes; then
-	CFLAGS="$CFLAGS_X"
-	LIBS="$LIBS_X"
-else
-	AC_DEFINE(HAVE_SSL)
+
+	if test "$cf_result" != yes; then
+		if test "$withval" = yes; then
+			AC_MSG_ERROR("OpenSSL not found")
+		fi
+	        CFLAGS="$CFLAGS_X"
+        	LIBS="$LIBS_X"
+	else
+		AC_DEFINE(HAVE_SSL)
+	fi
 fi
 AC_MSG_RESULT($cf_result)
 #endif
