--- links-0.96/codepage.inc.orig	Mon Jul  2 03:28:35 2001
+++ links-0.96/codepage.inc	Fri Jul 27 22:29:43 2001
@@ -3012,6 +3012,8 @@
 
 struct codepage_desc codepages [] = {
 	"7-bit ASCII", aliases_7bit, table_7bit,
+	"KOI8-R", aliases_koi8_r, table_koi8_r,
+	"KOI8-U", aliases_koi8_u, table_koi8_u,
 	"ISO 8859-1", aliases_8859_1, table_8859_1,
 	"ISO 8859-2", aliases_8859_2, table_8859_2,
 	"ISO 8859-4", aliases_8859_4, table_8859_4,
@@ -3031,8 +3033,6 @@
 	"MacRoman 2000", aliases_macroman, table_macroman,
 	"Mac latin 2", aliases_mac_lat2, table_mac_lat2,
 	"Kamenicky Brothers", aliases_kamen, table_kamen,
-	"KOI8-R", aliases_koi8_r, table_koi8_r,
-	"KOI8-U", aliases_koi8_u, table_koi8_u,
 	"TCVN-5712", aliases_tcvn5712, table_tcvn5712,
 	"VISCII", aliases_viscii, table_viscii,
 	"Unicode UTF-8", aliases_utf_8, table_utf_8,
