diff -urN links-0.95-orig links-0.95
--- links-0.95-orig/default.c	Mon Jan  1 05:26:52 2001
+++ links-0.95/default.c	Thu Mar 15 13:16:37 2001
@@ -906,7 +906,7 @@
 	1, gen_cmd, num_rd, num_wr, 0, 1, &http_bugs.bug_302_redirect, "http_bugs.bug_302_redirect", "http-bugs.bug-302-redirect",
 	1, gen_cmd, num_rd, num_wr, 0, 1, &http_bugs.bug_post_no_keepalive, "http_bugs.bug_post_no_keepalive", "http-bugs.bug_post-no-keepalive",
 	1, gen_cmd, str_rd, str_wr, 0, MAX_STR_LEN, default_anon_pass, "ftp.anonymous_password", "ftp.anonymous-password",
-	1, NULL, cp_rd, NULL, 0, 0, &dds.assume_cp, "assume_codepage", "assume-codepage",
+	1, gen_cmd, cp_rd, NULL, 0, 0, &dds.assume_cp, "assume_codepage", "assume-codepage",
 	1, NULL, term_rd, term_wr, 0, 0, NULL, "terminal", NULL,
 	1, NULL, term2_rd, NULL, 0, 0, NULL, "terminal2", NULL,
 	1, NULL, type_rd, type_wr, 0, 0, NULL, "association", NULL,
--- links-0.95.old/codepage.inc	Mon Jan  1 02:26:52 2001
+++ links-0.95/codepage.inc	Mon Mar 19 04:41:24 2001
@@ -2344,6 +2344,8 @@
 
 struct codepage_desc codepages [] = {
 	"7-bit ASCII", aliases_7bit, table_7bit,
+	"KOI8-R", aliases_koi8_r, table_koi8_r,
+	"KOI8-U", aliases_koi8_u, table_koi8_u,
 	"ISO 8859-1", aliases_8859_1, table_8859_1,
 	"ISO 8859-2", aliases_8859_2, table_8859_2,
 	"ISO 8859-4", aliases_8859_4, table_8859_4,
@@ -2360,8 +2362,6 @@
 	"CP 1125 (Ukrainian)", aliases_cp1125, table_cp1125,
 	"Mac latin 2", aliases_mac_lat2, table_mac_lat2,
 	"Kamenicky Brothers", aliases_kamen, table_kamen,
-	"KOI8-R", aliases_koi8_r, table_koi8_r,
-	"KOI8-U", aliases_koi8_u, table_koi8_u,
 	"Unicode UTF-8", aliases_utf_8, table_utf_8,
 	NULL, NULL, NULL
 };
