--- cpio-2.4.2.orig/userspec.c
+++ cpio-2.4.2/userspec.c
@@ -82,7 +82,12 @@
 
 #define isdigit(c) ((c) >= '0' && (c) <= '9')

-char *strdup ();
+/* Debian hack: required to compile in glibc_2.1 (provided by Juan
+   Cespedes) <cespedes@debian.org>.  This bug has been reported to
+   "bug-gnu-utils@prep.ai.mit.edu".  (98/5/19) -BEM */
+#ifndef __GLIBC__
+char *strdup (const char *s);
+#endif
 
 /* Return nonzero if STR represents an unsigned decimal integer,
    otherwise return 0. */
 
