--- cpio-2.10.90/src/main.c.orig	2009-07-31 09:16:48.000000000 +0000
+++ cpio-2.10.90/src/main.c	2009-09-08 20:23:17.000000000 +0000
@@ -238,6 +238,7 @@ static struct argp_option options[] = {
 
 static char *input_archive_name = 0;
 static char *output_archive_name = 0;
+static int abs_paths_flag_changed = false;
 
 static int
 warn_control (char *arg)
@@ -404,10 +405,12 @@ crc newc odc bin ustar tar (all-caps als
 
     case NO_ABSOLUTE_FILENAMES_OPTION:		/* --no-absolute-filenames */
       no_abs_paths_flag = true;
+      abs_paths_flag_changed = true;
       break;
 
     case ABSOLUTE_FILENAMES_OPTION:		/* --absolute-filenames */
       no_abs_paths_flag = false;
+      abs_paths_flag_changed = true;
       break;
       
     case NO_PRESERVE_OWNER_OPTION:		/* --no-preserve-owner */
@@ -599,6 +602,8 @@ process_args (int argc, char *argv[])
       save_patterns = &argv[index];
       if (input_archive_name)
 	archive_name = input_archive_name;
+      if (!abs_paths_flag_changed)
+	no_abs_paths_flag = true;
     }
   else if (copy_function == copy_out)
     {
@@ -634,6 +639,8 @@ process_args (int argc, char *argv[])
 	archive_format = arf_binary;
       if (output_archive_name)
 	archive_name = output_archive_name;
+      if (!abs_paths_flag_changed)
+	no_abs_paths_flag = false;
     }
   else
     {
@@ -654,10 +661,10 @@ process_args (int argc, char *argv[])
       CHECK_USAGE(rename_flag, "--rename", "--pass-through");
       CHECK_USAGE(append_flag, "--append", "--pass-through");
       CHECK_USAGE(rename_batch_file, "--rename-batch-file", "--pass-through");
-      CHECK_USAGE(no_abs_paths_flag, "--no-absolute-pathnames",
-		  "--pass-through");
-      CHECK_USAGE(no_abs_paths_flag, "--absolute-pathnames",
-		  "--pass-through");
+      CHECK_USAGE((abs_paths_flag_changed && no_abs_paths_flag),
+		  "--no-absolute-pathnames", "--pass-through");
+      CHECK_USAGE((abs_paths_flag_changed && !no_abs_paths_flag),
+		  "--absolute-pathnames", "--pass-through");
       CHECK_USAGE(to_stdout_option, "--to-stdout", "--pass-through");
       
       directory_name = argv[index];
