--- cpio-2.4.2/copyout.c.ewt	Wed Apr 23 12:52:01 1997
+++ cpio-2.4.2/copyout.c	Wed Apr 23 12:52:19 1997
@@ -54,7 +54,7 @@
       else
 	magic_string = "070701";
       sprintf (ascii_header,
-	       "%6s%08lx%08lx%08lx%08lx%08lx%08lx%08lx%08lx%08lx%08lx%08lx%08lx%08lx",
+	       "%6s%08lx%08lx%08lx%08lx%08lx%08x%08lx%08lx%08lx%08lx%08lx%08lx%08lx",
 	       magic_string,
 	       file_hdr->c_ino, file_hdr->c_mode, file_hdr->c_uid,
 	       file_hdr->c_gid, file_hdr->c_nlink, file_hdr->c_mtime,
