--- cpio-2.10.90/lib/rtapelib.c.orig	2009-06-18 16:07:51.000000000 +0000
+++ cpio-2.10.90/lib/rtapelib.c	2009-09-07 19:12:48.000000000 +0000
@@ -504,12 +504,14 @@ rmt_open__ (const char *file_name, int o
 	/* Child.  */
 
 	close (STDIN_FILENO);
-	dup (to_remote[remote_pipe_number][PREAD]);
+	if (dup (to_remote[remote_pipe_number][PREAD]) != STDIN_FILENO)
+	  error (EXIT_ON_EXEC_ERROR, errno, _("Cannot dup"));
 	close (to_remote[remote_pipe_number][PREAD]);
 	close (to_remote[remote_pipe_number][PWRITE]);
 
 	close (STDOUT_FILENO);
-	dup (from_remote[remote_pipe_number][PWRITE]);
+	if (dup (from_remote[remote_pipe_number][PWRITE]) != STDOUT_FILENO)
+	  error (EXIT_ON_EXEC_ERROR, errno, _("Cannot dup"));
 	close (from_remote[remote_pipe_number][PREAD]);
 	close (from_remote[remote_pipe_number][PWRITE]);
 
--- cpio-2.10.90/src/copyin.c.orig	2009-07-31 09:16:26.000000000 +0000
+++ cpio-2.10.90/src/copyin.c	2009-09-07 19:12:48.000000000 +0000
@@ -178,7 +178,7 @@ list_file(struct cpio_file_stat* file_hd
 #endif
       if (crc != file_hdr->c_chksum)
 	{
-	  error (0, 0, _("%s: checksum error (0x%lx, should be 0x%lx)"),
+	  error (0, 0, _("%s: checksum error (0x%x, should be 0x%lx)"),
 		 file_hdr->c_name, crc, file_hdr->c_chksum);
 	}
     }
@@ -525,7 +525,7 @@ copyin_regular_file (struct cpio_file_st
       if (archive_format == arf_crcascii)
 	{
 	  if (crc != file_hdr->c_chksum)
-	    error (0, 0, _("%s: checksum error (0x%lx, should be 0x%lx)"),
+	    error (0, 0, _("%s: checksum error (0x%x, should be 0x%lx)"),
 		   file_hdr->c_name, crc, file_hdr->c_chksum);
 	}
       tape_skip_padding (in_file_des, file_hdr->c_filesize);
@@ -537,8 +537,9 @@ copyin_regular_file (struct cpio_file_st
      "bug-gnu-utils@prep.ai.mit.edu".  (96/7/10) -BEM */
   if (delayed_seek_count > 0)
     {
-      lseek (out_file_des, delayed_seek_count-1, SEEK_CUR);
-      write (out_file_des, "", 1);
+      if (lseek (out_file_des, delayed_seek_count-1, SEEK_CUR) < 0 ||
+	  write (out_file_des, "", 1) != 1)
+	error (0, errno, _("cannot seek on output"));
       delayed_seek_count = 0;
     }
 
@@ -550,7 +551,7 @@ copyin_regular_file (struct cpio_file_st
   if (archive_format == arf_crcascii)
     {
       if (crc != file_hdr->c_chksum)
-	error (0, 0, _("%s: checksum error (0x%lx, should be 0x%lx)"),
+	error (0, 0, _("%s: checksum error (0x%x, should be 0x%lx)"),
 	       file_hdr->c_name, crc, file_hdr->c_chksum);
     }
 
@@ -991,14 +992,14 @@ from_ascii (char const *where, size_t di
       char *p = strchr (codetab, toupper (*buf));
       if (!p)
 	{
-	  error (0, 0, _("Malformed number %.*s"), digs, where);
+	  error (0, 0, _("Malformed number %.*s"), (unsigned) digs, where);
 	  break;
 	}
       
       d = p - codetab;
       if ((d >> logbase) > 1)
 	{
-	  error (0, 0, _("Malformed number %.*s"), digs, where);
+	  error (0, 0, _("Malformed number %.*s"), (unsigned) digs, where);
 	  break;
 	}
       value += d;
@@ -1009,7 +1010,7 @@ from_ascii (char const *where, size_t di
     }
   if (overflow)
     error (0, 0, _("Archive value %.*s is out of range"),
-	   digs, where);
+	   (unsigned) digs, where);
   return value;
 }
 
@@ -1521,7 +1522,7 @@ process_copy_in ()
 	    tape_skip_padding (in_file_des, file_hdr.c_filesize);
 	    if (crc != file_hdr.c_chksum)
 	      {
-		error (0, 0, _("%s: checksum error (0x%lx, should be 0x%lx)"),
+		error (0, 0, _("%s: checksum error (0x%x, should be 0x%lx)"),
 		       file_hdr.c_name, crc, file_hdr.c_chksum);
 	      }
          /* Debian hack: -v and -V now work with --only-verify-crc.
--- cpio-2.10.90/src/copypass.c.orig	2009-02-14 12:39:37.000000000 +0000
+++ cpio-2.10.90/src/copypass.c	2009-09-07 19:12:51.000000000 +0000
@@ -189,8 +189,9 @@ process_copy_pass ()
                  "bug-gnu-utils@prep.ai.mit.edu".  (96/7/10) -BEM */
 	      if (delayed_seek_count > 0)
 		{
-		  lseek (out_file_des, delayed_seek_count-1, SEEK_CUR);
-		  write (out_file_des, "", 1);
+		  if (lseek (out_file_des, delayed_seek_count-1, SEEK_CUR) < 0 ||
+		      write (out_file_des, "", 1) != 1)
+		      error (0, errno, _("cannot seek on output"));
 		  delayed_seek_count = 0;
 		}
 
--- cpio-2.10.90/src/util.c.orig	2009-02-14 12:39:37.000000000 +0000
+++ cpio-2.10.90/src/util.c	2009-09-07 19:12:51.000000000 +0000
@@ -484,9 +484,9 @@ copy_files_disk_to_tape (int in_des, int
   while (num_bytes > 0)
     {
       if (input_size == 0)
-	if (rc = disk_fill_input_buffer (in_des,
-					 num_bytes < DISK_IO_BLOCK_SIZE ?
-					 num_bytes : DISK_IO_BLOCK_SIZE))
+	if ((rc = disk_fill_input_buffer (in_des,
+					  num_bytes < DISK_IO_BLOCK_SIZE ?
+					  num_bytes : DISK_IO_BLOCK_SIZE)))
 	  {
 	    if (rc > 0)
 	      {
@@ -499,7 +499,7 @@ copy_files_disk_to_tape (int in_des, int
 	      }
 	    else
 	      error (0, 0, _("Read error at byte %lld in file %s, padding with zeros"),
-			original_num_bytes - num_bytes, filename);
+			(long long)(original_num_bytes - num_bytes), filename);
 	    write_nuls_to_file (num_bytes, out_des, tape_buffered_write);
 	    break;
 	  }
@@ -536,7 +536,7 @@ copy_files_disk_to_disk (int in_des, int
   while (num_bytes > 0)
     {
       if (input_size == 0)
-	if (rc = disk_fill_input_buffer (in_des, num_bytes))
+	if ((rc = disk_fill_input_buffer (in_des, num_bytes)))
 	  {
 	    if (rc > 0)
 	      {
@@ -549,7 +549,7 @@ copy_files_disk_to_disk (int in_des, int
 	      }
 	    else
 	      error (0, 0, _("Read error at byte %lld in file %s, padding with zeros"),
-			original_num_bytes - num_bytes, filename);
+			(long long)(original_num_bytes - num_bytes), filename);
 	    write_nuls_to_file (num_bytes, out_des, disk_buffered_write);
 	    break;
 	  }
--- cpio-2.10.90/tests/genfile.c.orig	2009-06-18 16:08:14.000000000 +0000
+++ cpio-2.10.90/tests/genfile.c	2009-09-07 19:12:51.000000000 +0000
@@ -478,7 +478,8 @@ mkhole (int fd, off_t displ)
 {
   if (lseek (fd, displ, SEEK_CUR) == -1)
     error (EXIT_FAILURE, errno, "lseek");
-  ftruncate (fd, lseek (fd, 0, SEEK_CUR));
+  if (ftruncate (fd, lseek (fd, 0, SEEK_CUR)))
+    error (EXIT_FAILURE, errno, "ftruncate");
 }
 
 static void
@@ -676,13 +677,15 @@ exec_checkpoint (struct action *p)
 	    error (0, errno, _("cannot open `%s'"), p->name);
 	    break;
 	  }
-	ftruncate (fd, p->size);
+	if (ftruncate (fd, p->size))
+	  error (0, errno, _("cannot truncate `%s'"), p->name);
 	close (fd);
       }
       break;
 
     case OPT_EXEC:
-      system (p->name);
+      if (system (p->name) == -1)
+	error (0, errno, _("cannot execute `%s'"), p->name);
       break;
 
     default:
@@ -741,7 +744,8 @@ exec_command (void)
   signal (SIGCHLD, SIG_DFL);
 #endif
 
-  pipe (fd);
+  if (pipe (fd))
+    error (EXIT_FAILURE, errno, "pipe");
 
   pid = fork ();
   if (pid == -1)
