--- cpio-2.6/src/mt.c.orig	2005-05-05 12:48:41 +0000
+++ cpio-2.6/src/mt.c	2005-05-05 12:58:29 +0000
@@ -91,6 +91,8 @@
 # include <stdlib.h>
 #endif
 
+#include <argmatch.h>
+
 #if !HAVE_DECL_ERRNO
 extern int errno;
 #endif
@@ -277,7 +279,7 @@ main (int argc, char **argv)
   if (optind == argc)
     usage (stderr, 1);
 
-  i = argmatch (argv[optind], opnames);
+  i = argmatch (argv[optind], opnames, NULL, 0);
   if (i < 0)
     {
       invalid_arg ("tape operation", argv[optind], i);
