--- findutils-4.2.25/find/find.c.orig	2005-08-30 07:40:03 +0000
+++ findutils-4.2.25/find/find.c	2005-09-04 14:25:03 +0000
@@ -357,7 +357,7 @@ get_info (const char *pathname,
 
 /* Determine if we can use O_NOFOLLOW.
  */
-#if defined(O_NOFOLLOW)
+#if defined(O_NOFOLLOW) && defined(CHECK_NOFOLLOW)
 static boolean 
 check_nofollow(void)
 {
@@ -405,8 +405,12 @@ main (int argc, char **argv)
    * if we call it after setlocale().
    */
 #ifdef O_NOFOLLOW
+#ifdef CHECK_NOFOLLOW
   options.open_nofollow_available = check_nofollow();
 #else
+  options.open_nofollow_available = true;
+#endif /* CHECK_NOFOLLOW */
+#else
   options.open_nofollow_available = false;
 #endif
 
