diff -ur findutils-4.1.7.orig/find/defs.h findutils-4.1.7/find/defs.h
--- findutils-4.1.7.orig/find/defs.h	Mon May 21 00:39:37 2001
+++ findutils-4.1.7/find/defs.h	Tue May 14 11:00:01 2002
@@ -334,6 +334,7 @@
 boolean pred_cnewer PARAMS((char *pathname, struct stat *stat_buf, struct predicate *pred_ptr));
 boolean pred_comma PARAMS((char *pathname, struct stat *stat_buf, struct predicate *pred_ptr));
 boolean pred_ctime PARAMS((char *pathname, struct stat *stat_buf, struct predicate *pred_ptr));
+boolean pred_delete PARAMS((char *pathname, struct stat *stat_buf, struct predicate *pred_ptr));
 boolean pred_empty PARAMS((char *pathname, struct stat *stat_buf, struct predicate *pred_ptr));
 boolean pred_exec PARAMS((char *pathname, struct stat *stat_buf, struct predicate *pred_ptr));
 boolean pred_false PARAMS((char *pathname, struct stat *stat_buf, struct predicate *pred_ptr));
diff -ur findutils-4.1.7.orig/find/parser.c findutils-4.1.7/find/parser.c
--- findutils-4.1.7.orig/find/parser.c	Mon May 21 00:39:37 2001
+++ findutils-4.1.7/find/parser.c	Tue May 14 11:00:15 2002
@@ -62,6 +62,7 @@
 static boolean parse_comma PARAMS((char *argv[], int *arg_ptr));
 static boolean parse_ctime PARAMS((char *argv[], int *arg_ptr));
 static boolean parse_daystart PARAMS((char *argv[], int *arg_ptr));
+static boolean parse_delete PARAMS((char *argv[], int *arg_ptr));
 static boolean parse_depth PARAMS((char *argv[], int *arg_ptr));
 static boolean parse_empty PARAMS((char *argv[], int *arg_ptr));
 static boolean parse_exec PARAMS((char *argv[], int *arg_ptr));
@@ -160,6 +161,7 @@
 #endif
   {"ctime", parse_ctime},
   {"daystart", parse_daystart},	/* GNU */
+  {"delete", parse_delete},
   {"depth", parse_depth},
   {"empty", parse_empty},	/* GNU */
   {"exec", parse_exec},
@@ -398,6 +400,20 @@
 }
 
 static boolean
+parse_delete (argv, arg_ptr)
+  char *argv[];
+  int *arg_ptr;
+{
+  struct predicate *our_pred;
+
+  our_pred = insert_primary (pred_delete);
+  our_pred->side_effects = true;
+  /* -delete implies -depth */
+  do_dir_first = false;
+  return (true);
+}
+
+static boolean
 parse_depth (char **argv, int *arg_ptr)
 {
   do_dir_first = false;
@@ -559,7 +575,7 @@
       ( EXPR ) ! EXPR -not EXPR EXPR1 -a EXPR2 EXPR1 -and EXPR2\n"));
   puts (_("\
       EXPR1 -o EXPR2 EXPR1 -or EXPR2 EXPR1 , EXPR2\n\
-options (always true): -daystart -depth -follow --help\n\
+options (always true): -daystart -delete -depth -follow --help\n\
       -maxdepth LEVELS -mindepth LEVELS -mount -noleaf --version -xdev\n\
 tests (N can be +N or -N or N): -amin N -anewer FILE -atime N -cmin N\n"));
   puts (_("\
diff -ur findutils-4.1.7.orig/find/pred.c findutils-4.1.7/find/pred.c
--- findutils-4.1.7.orig/find/pred.c	Mon May 21 00:39:37 2001
+++ findutils-4.1.7/find/pred.c	Tue May 14 11:00:15 2002
@@ -157,6 +157,7 @@
   {pred_cnewer, "cnewer  "},
   {pred_comma, ",       "},
   {pred_ctime, "ctime   "},
+  {pred_delete, "delete  "},
   {pred_empty, "empty   "},
   {pred_exec, "exec    "},
   {pred_false, "false   "},
@@ -398,6 +399,17 @@
       break;
     }
   return (false);
+}
+
+boolean
+pred_delete (pathname, stat_buf, pred_ptr)
+     char *pathname;
+     struct stat *stat_buf;
+     struct predicate *pred_ptr;
+{
+  if (strcmp (rel_pathname, "."))
+    return (remove (rel_pathname));
+  return (true);
 }
 
 boolean
