diff -ur termcap-2.0.8.orig/Makefile termcap-2.0.8/Makefile
--- termcap-2.0.8.orig/Makefile	Tue Apr 16 07:23:23 1996
+++ termcap-2.0.8/Makefile	Wed Aug  2 06:12:48 2000
@@ -41,7 +41,7 @@
 
 $(SHARED_LIB): $(OBJS)
 	cd pic; \
-	$(CC) -shared -o ../$@ -Wl,-soname,$(SONAME_SHARED_LIB) $(OBJS)
+	$(CC) -shared -o ../$@ -Wl,-soname,$(SONAME_SHARED_LIB) $(OBJS) -lc
 
 pic:
 	-if [ ! -d pic ]; then mkdir pic; fi
@@ -53,11 +53,10 @@
 		$(prefix)/lib/backup/$(SHARED_LIB).$$$$; \
 	fi
 	cp $(SHARED_LIB) $(prefix)/lib
-	chown $(OWNER) $(prefix)/lib/$(SHARED_LIB)
-	if [ -x /sbin/ldconfig -o -x /etc/ldconfig ]; then \
-	  ldconfig; \
-	fi
 	-if [ $(prefix) = "/" ]; then \
+	  if [ -x /sbin/ldconfig -o -x /etc/ldconfig ]; then \
+	    ldconfig; \
+	  fi; \
 	  cp -f $(STATIC_LIB) /usr/lib; \
 	  rm -f /lib/libtermcap.so; \
 	  rm -f /usr/lib/libtermcap.so; \
@@ -75,10 +74,6 @@
 	  ln -s $(prefix)/lib/$(SHARED_LIB) \
 		$(prefix)/lib/libtermcap.so; \
 	  cp -f termcap.h $(prefix)/include; \
-	  chown $(OWNER) \
-		$(prefix)/lib/$(STATIC_LIB) \
-		$(prefix)/lib/libtermcap.so \
-		$(prefix)/include/termcap.h; \
 	fi
 
 install-dirs:
@@ -89,7 +84,6 @@
 install-data:
 	-if [ $(prefix) = "/" ]; then \
 	  cp termcap.src /etc/termcap; \
-	  chown $(OWNER) /etc/termcap; \
 	fi
 
 info: termcap.info
