diff -ur termcap-2.0.8.orig/termcap.c termcap-2.0.8/termcap.c
--- termcap-2.0.8.orig/termcap.c	Tue Apr 16 07:23:23 1996
+++ termcap-2.0.8/termcap.c	Sun Sep  5 10:19:54 1999
@@ -451,7 +451,7 @@
 
   /* First copy the description to the buffer. */
   count = 0;
-  for(bp = s; *bp; bp++) {
+  for(bp = s; *bp && sp - term_entry < maxlen - 2; bp++) {
 	*sp++ = *bp;
 	count++;
   }
@@ -539,7 +539,7 @@
 {
   char *s;
   char *sp, *r, *ret;
-  int c, i;
+  int c, i, size;
 
   s = find_cap(term_entry, cap, '=');
   if (s == NULL) return(s);
@@ -550,12 +550,15 @@
 		if (*sp =='\\' && sp[1] == ':') sp++;
 	}
 	ret = xmalloc(sp - s + 1);
-  } else
+	size = sp - s + 1;
+  } else {
 	ret = *bufp;
+	size = tgetent_bufsize;
+  }
   r = ret;
 
   /* Translate escaped characters and hat-notation. */
-  while((c = *s++) && c != ':') {
+  while((c = *s++) && c != ':' && r - ret < size - 1) {
 	if (c == '\\') {
 
 		/* Escaped character. */
