diff -puNr tcp_wrappers_7.6.orig/fix_options.c tcp_wrappers_7.6/fix_options.c
--- tcp_wrappers_7.6.orig/fix_options.c	2014-06-28 18:45:23 +0000
+++ tcp_wrappers_7.6/fix_options.c	2014-06-28 18:46:04 +0000
@@ -28,6 +28,27 @@ static char sccsid[] = "@(#) fix_options
 
 #define BUFFER_SIZE	512		/* Was: BUFSIZ */
 
+#include <stdint.h>
+struct ip_options {
+  uint32_t     faddr;                          /* Saved first hop address */
+  unsigned char        optlen;
+  unsigned char srr;
+  unsigned char rr;
+  unsigned char ts;
+  unsigned char is_setbyuser:1,                        /* Set by setsockopt?                   */
+                is_data:1,                     /* Options in __data, rather than skb   */
+                is_strictroute:1,              /* Strict source route                  */
+                srr_is_hit:1,                  /* Packet destination addr was our one  */
+                is_changed:1,                  /* IP checksum more not valid           */      
+                rr_needaddr:1,                 /* Need to record addr of outgoing dev  */
+                ts_needtime:1,                 /* Need to record timestamp             */
+                ts_needaddr:1;                 /* Need to record addr of outgoing dev  */
+  unsigned char router_alert;
+  unsigned char __pad1;
+  unsigned char __pad2;
+  unsigned char __data[0];
+};
+
 /* fix_options - get rid of IP-level socket options */
 
 void
