--- groff-1.16.1/src/preproc/eqn/main.cc.orig	Tue Nov 21 15:46:09 2000
+++ groff-1.16.1/src/preproc/eqn/main.cc	Tue Nov 21 15:46:26 2000
@@ -329,7 +329,7 @@
 	 device, program_name);
   if (load_startup_file) {
     char *path;
-    FILE *fp = macro_path.open_file(STARTUP_FILE, &path);
+    FILE *fp = safer_macro_path.open_file(STARTUP_FILE, &path);
     if (fp) {
       do_file(fp, path);
       fclose(fp);
