--- groff-1.17/src/preproc/html/pre-html.cc.orig	Fri Apr 13 13:03:57 2001
+++ groff-1.17/src/preproc/html/pre-html.cc	Sat May  5 22:52:07 2001
@@ -578,12 +578,29 @@
  *  createAllPages - creates a set of images, one per page.
  */
 
-static void createAllPages (void)
+static int createAllPages (void)
 {
   char buffer[4096];
+  int retries;
+
+  imagePageStem = xtmptemplate("-page-");
+  strcpy(buffer, imagePageStem);
+  retries = 0x1000;
+  do {
+    if (mktemp(imagePageStem) == NULL) {
+      sys_fatal("mktemp");
+      return -1;
+    }
+    if (mkdir(imagePageStem, 0700) == 0) break;
+    if (errno != EEXIST || !--retries) {
+      sys_fatal("mkdir");
+      return -1;
+    }
+    strcpy(imagePageStem, buffer);
+  } while (1);
 
   sprintf(buffer,
-	  "echo showpage | gs -q -dSAFER -sDEVICE=%s -r%d -sOutputFile=%s%%d %s - > /dev/null 2>&1 \n",
+	  "echo showpage | gs -q -dSAFER -sDEVICE=%s -r%d -sOutputFile=%s/%%d %s - > /dev/null 2>&1 \n",
 	  image_device,
 	  image_res,
 	  imagePageStem,
@@ -593,6 +610,8 @@
   fflush(stderr);
 #endif
   system(buffer);
+
+  return 0;
 }
 
 /*
@@ -606,9 +625,10 @@
   int  i=1;
 
   do {
-    sprintf(buffer, "%s%d", imagePageStem, i);
+    sprintf(buffer, "%s/%d", imagePageStem, i);
     i++;
-  } while (remove(buffer) == 0);
+  } while (unlink(buffer) == 0);
+  rmdir(imagePageStem);
 #endif
 }
 
@@ -665,7 +685,7 @@
     int  y2 = (image_res*vertical_offset/72)+max(i->Y1, i->Y2)*image_res/POSTSCRIPTRES+1*IMAGE_BOARDER_PIXELS;
 
     sprintf(buffer,
-	    "pnmcut %d %d %d %d < %s%d | pnmtopng %s > %s \n",
+	    "pnmcut %d %d %d %d < %s/%d | pnmtopng %s > %s \n",
 	    x1, y1, x2-x1+1, y2-y1+1,
 	    imagePageStem,
 	    i->pageNo,
@@ -1053,7 +1073,7 @@
  *  makeTempFiles - name the temporary files
  */
 
-static void makeTempFiles (void)
+static int makeTempFiles (void)
 {
 #if defined(DEBUGGING)
   psFileName     = "/tmp/prehtml-ps";
@@ -1062,9 +1082,19 @@
   troffFileName  = "/tmp/prehtml-troff";
   htmlFileName   = "/tmp/prehtml-html";
 #else
-  psFileName     = mktemp(xtmptemplate("-ps-"));
-  regionFileName = mktemp(xtmptemplate("-regions-"));
-  imagePageStem  = mktemp(xtmptemplate("-page-"));
+  int fd;
+
+  if ((fd = mkstemp(psFileName = xtmptemplate("-ps-"))) == -1) {
+    sys_fatal("mkstemp");
+    return -1;
+  }
+  close(fd);
+  if ((fd = mkstemp(regionFileName = xtmptemplate("-regions-"))) == -1) {
+    sys_fatal("mkstemp");
+    unlink(psFileName);
+    return -1;
+  }
+  close(fd);
 #endif
 }
 
@@ -1075,8 +1105,8 @@
 static void removeTempFiles (void)
 {
 #if !defined(DEBUGGING)
-  remove(psFileName);
-  remove(regionFileName);
+  unlink(psFileName);
+  unlink(regionFileName);
 #endif
 }
 
@@ -1123,13 +1153,16 @@
   if (! found) {
     do_file("-");
   }
-  makeTempFiles();
+  if (makeTempFiles())
+    return 1;
   ok = inputFile.do_image(argc, argv);
   if (ok == 0) {
-    createAllPages();
-    generateImages(regionFileName);
-    ok = inputFile.do_html(argc, argv);
-    removeAllPages();
+    ok = createAllPages();
+    if (ok == 0) {
+      generateImages(regionFileName);
+      ok = inputFile.do_html(argc, argv);
+      removeAllPages();
+    }
   }
   removeTempFiles();
   return ok;
