--- popa3d-0.4.9/params.h.orig	Sun May 20 09:00:31 2001
+++ popa3d-0.4.9/params.h	Mon May 28 02:30:54 2001
@@ -8,7 +8,7 @@
 /*
  * Are we going to be a standalone server or start via an inetd clone?
  */
-#define POP_STANDALONE			0
+#define POP_STANDALONE			1
 
 #if POP_STANDALONE
 
@@ -108,9 +108,9 @@
  * Note that there's no built-in password aging support.
  */
 #define AUTH_PASSWD			0
-#define AUTH_SHADOW			1
+#define AUTH_SHADOW			0
 #define AUTH_PAM			0
-#define AUTH_PAM_USERPASS		0
+#define AUTH_PAM_USERPASS		1
 
 #if AUTH_PAM || AUTH_PAM_USERPASS
 #define AUTH_PAM_SERVICE		"popa3d"
@@ -125,7 +125,7 @@
  * it harder (but still far from impossible, on most systems) to check
  * for valid usernames. Adjust it for your crypt(3).
  */
-#define AUTH_DUMMY_SALT			"xx"
+#define AUTH_DUMMY_SALT			"$2a$08$xxxxxxxxxxxxxxxxxxxxxx"
 
 #endif
 
