--- raidtools-1.00.3/Makefile.in.orig	Thu Sep 11 16:23:08 2003
+++ raidtools-1.00.3/Makefile.in	Thu Sep 11 18:03:55 2003
@@ -53,7 +53,7 @@
 
 BINARIES = raidstart mkraid detect_multipath raidreconf arytst lsraid
 CC = @CC@
-CFLAGS = -O2 -Wall -DMD_VERSION=\""raidtools-@VERS@"\"
+CFLAGS = -Wall -DMD_VERSION=\""raidtools-@VERS@"\" @CFLAGS@
 LDFLAGS = @LDFLAGS@ -Wl,-Bstatic -lpopt -Wl,-Bdynamic
 RAIDSTARTLINKS = raidstop raidhotadd raidhotremove raidsetfaulty
 MKRAIDLINKS = raid0run
