--- util-linux-2.11z/login-utils/Makefile.orig	Sat Sep 29 21:11:24 2001
+++ util-linux-2.11z/login-utils/Makefile	Mon Apr  7 22:44:40 2003
@@ -13,7 +13,7 @@
 
 MAN1.MISC=	last.1 mesg.1 wall.1
 
-MAN1.PUTILS= 	chfn.1 chsh.1 login.1 newgrp.1
+MAN1.PUTILS= 	chfn.1 chsh.1 newgrp.1
 MAN1.PASSWD=    passwd.1
 
 MAN8.GETTY=	agetty.8
@@ -30,7 +30,7 @@
 
 SBIN.INIT=	simpleinit shutdown initctl
 
-BIN.PUTILS=	login
+BIN.PUTILS=
 
 USRBIN.MISC=	last mesg wall
 
@@ -166,7 +166,6 @@
 
 install-putils: $(BIN.PUTILS) $(USRBIN.PUTILS) $(USRSBIN.PUTILS)
 	$(INSTALLDIR) $(BINDIR) $(USRBINDIR) $(USRSBINDIR)
-	$(INSTALLBIN) $(BIN.PUTILS) $(BINDIR)
 	$(INSTALLSUID) $(USRBIN.PUTILS) $(USRBINDIR)
 	$(INSTALLBIN) $(USRSBIN.PUTILS) $(USRSBINDIR)
 	(cd $(USRSBINDIR); ln -sf vipw vigr)
--- util-linux-2.11z/misc-utils/Makefile.orig	Sat Nov  2 15:51:02 2002
+++ util-linux-2.11z/misc-utils/Makefile	Mon Apr  7 23:09:29 2003
@@ -11,7 +11,7 @@
 
 # Where to put man pages?
 
-MAN1= 		cal.1 chkdupexe.1 ddate.1 logger.1 look.1 mcookie.1 \
+MAN1= 		cal.1 ddate.1 logger.1 look.1 mcookie.1 \
 		namei.1 rename.1 script.1 whereis.1
 
 # Where to put binaries?
@@ -19,7 +19,7 @@
 
 BIN=
 
-USRBIN=		cal chkdupexe ddate logger look mcookie \
+USRBIN=		cal ddate logger look mcookie \
 		namei rename script whereis
 
 MAYBE=		reset setterm
@@ -79,7 +79,6 @@
 
 cal.o: $(LIB)/errs.h
 cal: cal.o $(ERR_O)
-chkdupexe: chkdupexe.pl
 ddate: ddate.o
 kill: kill.o procs.o
 logger: logger.o
@@ -104,8 +103,7 @@
 	$(INSTALLMAN) $(MAN1) $(MAN1DIR)
 ifeq "$(HAVE_WRITE)" "no"
 ifeq "$(USE_TTY_GROUP)" "yes"
-	chgrp tty $(USRBINDIR)/write
-	chmod g+s $(USRBINDIR)/write
+	chmod 2711 $(USRBINDIR)/write
 endif
 endif
 
