--- util-linux-2.11z/MCONFIG.orig	Mon Nov 25 12:29:48 2002
+++ util-linux-2.11z/MCONFIG	Mon Apr  7 21:35:09 2003
@@ -16,17 +16,17 @@
 # If HAVE_PAM is set to "yes", then login, chfn, chsh, and newgrp
 # will use PAM for authentication. Additionally, passwd will not be
 # installed as it is not PAM aware.
-HAVE_PAM=no
+HAVE_PAM=yes
 
 # If HAVE_SHADOW is set to "yes", then login, chfn, chsh, newgrp, passwd,
 # and vipw will not be built or installed from the login-utils
 # subdirectory.  
-HAVE_SHADOW=yes
+HAVE_SHADOW=no
 
 # If HAVE_PASSWD is set to "yes", then passwd will not be built or
 # installed from the login-utils subdirectory (but login, chfn, chsh,
 # newgrp, and vipw *will* be installed).
-HAVE_PASSWD=no
+HAVE_PASSWD=yes
 
 # If you use chfn and chsh from this package, REQUIRE_PASSWORD will require
 # non-root users to enter the account password before updating /etc/passwd.
@@ -88,7 +88,7 @@
 # If HAVE_SLN is set to "yes", then sln won't be installed
 # (but the man page sln.8 will be installed anyway).
 # sln also comes with libc and glibc.
-HAVE_SLN=no
+HAVE_SLN=yes
 
 # If HAVE_FDUTILS is set to "yes", then setfdprm won't be installed.
 HAVE_FDUTILS=no
@@ -102,28 +102,7 @@
 # DISABLE_NLS to "yes".
 DISABLE_NLS=no
 
-# Different optimizations for different cpus.
-# gcc 3.0 likes options -mcpu=i486 instead of -m486
-ifeq "$(ARCH)" "intel"
-  ifeq "$(HAVE_OLD_GCC)" "yes"
-    CPUHEAD=-m
-  else
-    CPUHEAD=-mcpu=i
-  endif
-  ifeq "$(CPU)" "i386"
-    CPUTAIL=386
-  else
-    CPUTAIL=486
-  endif
-  CPUOPT= $(CPUHEAD)$(CPUTAIL)
-  OPT=          -pipe -O2 $(CPUOPT) -fomit-frame-pointer
-else 
-  ifeq "$(ARCH)" "arm"
-    OPT=	-pipe -O2 -fsigned-char -fomit-frame-pointer
-  else
-    OPT=	-O2 -fomit-frame-pointer
-  endif
-endif
+OPT=$(RPM_OPT_FLAGS)
 
 WARNFLAGS = -Wall -Wmissing-prototypes -Wstrict-prototypes
 
@@ -138,7 +117,7 @@
 # SLANGFLAGS=-I/usr/include/slang
 # No such subdirectory - slcurses.h lives in /usr/include
 #   (no extra definition required).
-LIBSLANG=-lslang
+HAVE_NCURSES=yes
 
 #
 # Paths used for compilation (not all are actually used, see CFLAGS below)
@@ -190,7 +169,7 @@
 MANMODE=	644
 DATMODE=	644
 INFOMODE=	644
-SUIDMODE=      4755
+SUIDMODE=	700
 
 CHMOD=		chmod
 INSTALL=	install
@@ -198,7 +177,7 @@
 INSTALLBIN=	$(INSTALL) -m $(BINMODE)
 INSTALLMAN=	$(INSTALL) -m $(MANMODE)
 INSTALLDAT=	$(INSTALL) -m $(DATMODE)
-INSTALLSUID=    $(INSTALL) -m $(SUIDMODE) -o root
+INSTALLSUID=    $(INSTALL) -m $(SUIDMODE)
 
 ifeq "$(DISABLE_NLS)" "yes"
 NLSFLAGS = -DDISABLE_NLS
