diff -ur util-linux-2.10r.orig/clock/hwclock.c util-linux-2.10r/clock/hwclock.c
--- util-linux-2.10r.orig/clock/hwclock.c	Tue Oct  3 04:26:25 2000
+++ util-linux-2.10r/clock/hwclock.c	Wed Apr 18 08:25:26 2001
@@ -1126,7 +1126,7 @@
     ),RTC_DEV);
 #ifdef __alpha__
   fprintf( usageto, _(
-    "  --jensen, --arc, --srm, --funky-toy\n"
+    "  --Jensen, --ARC, --SRM, --funky-toy\n"
     "                tell hwclock the type of alpha you have (see hwclock(8))\n"
     ) );
 #endif
