diff -ur util-linux-2.10h.orig/mount/nfs_mount3.h util-linux-2.10h/mount/nfs_mount3.h
--- util-linux-2.10h.orig/mount/nfs_mount3.h	Sun Oct 31 03:45:23 1999
+++ util-linux-2.10h/mount/nfs_mount3.h	Fri Dec 29 10:55:37 2000
@@ -6,7 +6,7 @@
  */
 #include "nfs_mountversion.h"
 
-#if KERNEL_NFS_MOUNT_VERSION >= 3
+#if defined(NFS_VERSION) && KERNEL_NFS_MOUNT_VERSION >= 3
 
 /*
  * The kernel includes are at least as good as this file.
@@ -21,6 +21,12 @@
  * Check at runtime whether the running kernel can handle the new stuff.
  */
 #define NFS_MOUNT_VERSION	3
+
+#ifndef NFS_VERSION
+/* It used to be set for NFSv2 with Linux 2.2.17 and below */
+#define NFS_VERSION		2
+#define NFS_PORT		2049
+#endif
 
 struct nfs_mount_data {
 	int		version;		/* 1 */
diff -ur util-linux-2.10h.orig/mount/nfsmount.c util-linux-2.10h/mount/nfsmount.c
--- util-linux-2.10h.orig/mount/nfsmount.c	Wed Nov  3 01:18:34 1999
+++ util-linux-2.10h/mount/nfsmount.c	Fri Dec 29 10:54:23 2000
@@ -48,6 +48,8 @@
 #include "sundries.h"
 #include "nfsmount.h"
 
+#define NFS_NEED_KERNEL_TYPES
+#include <linux/types.h>
 #include <linux/nfs.h>
 #include "mount_constants.h"
 #include "nfs_mount3.h"
