--- util-linux-2.10r.orig/MCONFIG	Thu Nov  9 22:41:13 2000
+++ util-linux-2.10r/MCONFIG	Sun Jan 28 06:45:07 2001
@@ -16,17 +16,17 @@
 # If HAVE_PAM is set to "yes", then login, chfn, chsh, and newgrp
 # will use PAM for authentication. Additionally, passwd will not be
 # installed as it is not PAM aware.
-HAVE_PAM=no
+HAVE_PAM=yes
 
 # If HAVE_SHADOW is set to "yes", then login, chfn, chsh, newgrp, passwd,
 # and vipw will not be built or installed from the login-utils
 # subdirectory.  
-HAVE_SHADOW=yes
+HAVE_SHADOW=no
 
 # If HAVE_PASSWD is set to "yes", then passwd will not be built or
 # installed from the login-utils subdirectory (but login, chfn, chsh,
 # newgrp, and vipw *will* be installed).
-HAVE_PASSWD=no
+HAVE_PASSWD=yes
 
 # If you use chfn and chsh from this package, REQUIRE_PASSWORD will require
 # non-root users to enter the account password before updating /etc/passwd.
@@ -72,29 +72,17 @@
 # If HAVE_SLN is set to "yes", then sln won't be installed
 # (but the man page sln.8 will be installed anyway).
 # sln also comes with libc and glibc.
-HAVE_SLN=no
+HAVE_SLN=yes
 
 # If HAVE_TSORT is set to "yes", then tsort won't be installed.
 # GNU textutils 2.0 includes tsort.
-HAVE_TSORT=no
+HAVE_TSORT=yes
 
 # If HAVE_FDUTILS is set to "yes", then setfdprm won't be installed.
 HAVE_FDUTILS=no
 
 # Different optimizations for different cpus.
-ifeq "$(ARCH)" "intel"
-  ifeq "$(CPU)" "i386"
-    OPT=	-pipe -O2 -m386 -fomit-frame-pointer
-  else
-    OPT=	-pipe -O2 -m486 -fomit-frame-pointer
-  endif
-else 
-  ifeq "$(ARCH)" "arm"
-    OPT=	-pipe -O2 -fsigned-char -fomit-frame-pointer
-  else
-    OPT=	-O2 -fomit-frame-pointer
-  endif
-endif
+OPT=$(RPM_OPT_FLAGS)
 
 WARNFLAGS = -Wall -Wmissing-prototypes -Wstrict-prototypes
 
@@ -105,6 +93,7 @@
 # Set HAVE_SLANG to yes if you have slang (and prefer to use that for cfdisk)
 # (If neither HAVE_NCURSES nor HAVE_SLANG is defined, cfdisk is not made.)
 # HAVE_SLANG=yes
+HAVE_NCURSES=yes
 
 # There is a subdirectory /usr/include/slang containing slcurses.h
 # SLANGFLAGS=-I/usr/include/slang
@@ -150,7 +139,7 @@
 MANMODE=	644
 DATMODE=	644
 INFOMODE=	644
-SUIDMODE=      4755
+SUIDMODE=	700
 
 CHMOD=		chmod
 INSTALL=	install
@@ -158,7 +147,7 @@
 INSTALLBIN=	$(INSTALL) -m $(BINMODE)
 INSTALLMAN=	$(INSTALL) -m $(MANMODE)
 INSTALLDAT=	$(INSTALL) -m $(DATMODE)
-INSTALLSUID=    $(INSTALL) -m $(SUIDMODE) -o root
+INSTALLSUID=    $(INSTALL) -m $(SUIDMODE)
 
 CFLAGS    := $(CFLAGS) $(OPT) -I$(LIB) $(WARNFLAGS) \
 			$(CURSESFLAGS) $(SLANGFLAGS) \
