--- util-linux-2.11z/mount/mntent.c.orig	Thu Mar 15 12:09:58 2001
+++ util-linux-2.11z/mount/mntent.c	Mon Apr  7 22:39:41 2003
@@ -8,6 +8,7 @@
 #include <stdio.h>
 #include <string.h>		/* for index */
 #include <ctype.h>		/* for isdigit */
+#include <sys/stat.h>		/* for umask */
 #include "mntent.h"
 #include "sundries.h"		/* for xmalloc */
 #include "nls.h"
@@ -95,8 +96,11 @@
 mntFILE *
 my_setmntent (const char *file, char *mode) {
 	mntFILE *mfp = xmalloc(sizeof(*mfp));
+	mode_t saved_umask;
 
+	saved_umask = umask(077);
 	mfp->mntent_fp = fopen (file, mode);
+	umask(saved_umask);
 	mfp->mntent_file = xstrdup(file);
 	mfp->mntent_errs = (mfp->mntent_fp == NULL);
 	mfp->mntent_softerrs = 0;
--- util-linux-2.11z/mount/mount.c.orig	Thu Jan  2 21:38:44 2003
+++ util-linux-2.11z/mount/mount.c	Mon Apr  7 22:41:05 2003
@@ -1415,7 +1415,7 @@
 	bindtextdomain(PACKAGE, LOCALEDIR);
 	textdomain(PACKAGE);
 
-	umask(033);
+	umask(077);
 
 	/* People report that a mount called from init without console
 	   writes error messages to /etc/mtab
--- util-linux-2.11z/mount/umount.c.orig	Fri Nov  1 03:00:50 2002
+++ util-linux-2.11z/mount/umount.c	Mon Apr  7 22:41:34 2003
@@ -617,7 +617,7 @@
 	bindtextdomain(PACKAGE, LOCALEDIR);
 	textdomain(PACKAGE);
 
-	umask(033);
+	umask(077);
 
 	while ((c = getopt_long (argc, argv, "adfhlnrt:O:vV",
 				 longopts, NULL)) != -1)
