--- util-linux-2.11z/mount/lomount.c.orig	Mon Apr  7 20:22:01 2003
+++ util-linux-2.11z/mount/lomount.c	Mon Apr  7 20:44:28 2003
@@ -37,6 +37,14 @@
 extern void error (const char *fmt, ...);	/* idem */
 
 #ifdef LOOP_SET_FD
+#ifndef LO_CRYPT_SERPENT
+/* These aren't defined with unpatched Linux 2.2.19 headers */
+#define LO_CRYPT_SERPENT  7
+#define LO_CRYPT_MARS	  8
+#define LO_CRYPT_RC6      11
+#define LO_CRYPT_DES_EDE3 12
+#define LO_CRYPT_DFC      15
+#endif
 struct crypt_type_struct {
 	int id;
 	char *name;
@@ -268,7 +276,7 @@
 		loopinfo.lo_encrypt_key_size = strlen(loopinfo.lo_encrypt_key);
 		break;
 	case LO_CRYPT_DES:
-		printf(_("WARNING: Use of DES is depreciated.\n"));
+		printf(_("WARNING: Use of DES is deprecated.\n"));
 		pass = getpass (_("Password: "));
 		strncpy (loopinfo.lo_encrypt_key, pass, 8);
 		loopinfo.lo_encrypt_key[8] = 0;
@@ -294,7 +302,7 @@
 	case LO_CRYPT_RC6:
 	case LO_CRYPT_DES_EDE3:
 	case LO_CRYPT_DFC:
-		pass = getpass("Password :");
+		pass = getpass("Password: ");
 		strncpy(passwdbuff+1,pass,PASSWDBUFFLEN-1);
 		passwdbuff[0] = 'A';
 		rmd160_hash_buffer(keybits,pass,strlen(pass));
--- util-linux-2.11z/mount/losetup.8.orig	Mon Apr  7 18:58:14 2003
+++ util-linux-2.11z/mount/losetup.8	Mon Apr  7 20:44:28 2003
@@ -107,7 +107,7 @@
 .IP
 dd if=/dev/zero of=/file bs=1k count=100
 losetup -e blowfish /dev/loop0 /file
-Password :
+Password:
 mkfs -t ext2 /dev/loop0 100
 mount -t ext2 /dev/loop0 /mnt
  ...
