diff -urN util-linux-2.10h.orig/login-utils/chfn.c util-linux-2.10h/login-utils/chfn.c
--- util-linux-2.10h.orig/login-utils/chfn.c	Tue Mar 21 02:17:55 2000
+++ util-linux-2.10h/login-utils/chfn.c	Thu Jul 13 12:17:55 2000
@@ -83,6 +83,9 @@
 
 #define memzero(ptr, size) memset((char *) ptr, 0, size)
 
+#define fixed_iscntrl(c) \
+	(((c) & 0x7f) < 0x20 || (c) == 0x7f)
+
 /* we do not accept gecos field sizes longer than MAX_FIELD_SIZE */
 #define MAX_FIELD_SIZE		256
 
@@ -102,9 +105,11 @@
 #endif
 
     sanitize_env();
-    setlocale(LC_ALL, "");	/* both for messages and for iscntrl() below */
-    bindtextdomain(PACKAGE, LOCALEDIR);
-    textdomain(PACKAGE);
+    if (!getuid()) {
+	setlocale(LC_ALL, "");	/* both for messages and for iscntrl() below */
+	bindtextdomain(PACKAGE, LOCALEDIR);
+	textdomain(PACKAGE);
+    }
 
     /* whoami is the program name for error messages */
     whoami = argv[0];
@@ -416,7 +421,7 @@
 	    printf (_("'%c' is not allowed.\n"), c);
 	    return (-1);
 	}
-	if (iscntrl (c)) {
+	if (iscntrl (c) || fixed_iscntrl (c)) {
 	    if (msg) printf ("%s: ", msg);
 	    printf (_("Control characters are not allowed.\n"));
 	    return (-1);
diff -urN util-linux-2.10h.orig/login-utils/chsh.c util-linux-2.10h/login-utils/chsh.c
--- util-linux-2.10h.orig/login-utils/chsh.c	Tue Mar 21 02:17:55 2000
+++ util-linux-2.10h/login-utils/chsh.c	Thu Jul 13 12:16:30 2000
@@ -80,6 +80,9 @@
 extern int setpwnam P((struct passwd *pwd));
 #define memzero(ptr, size) memset((char *) ptr, 0, size)
 
+#define fixed_iscntrl(c) \
+	(((c) & 0x7f) < 0x20 || (c) == 0x7f)
+
 int main (argc, argv)
     int argc;
     char *argv[];
@@ -95,9 +98,12 @@
 #endif
 
     sanitize_env();
-    setlocale(LC_ALL, "");
-    bindtextdomain(PACKAGE, LOCALEDIR);
-    textdomain(PACKAGE);
+
+    if (!getuid()) {
+	setlocale(LC_ALL, "");
+	bindtextdomain(PACKAGE, LOCALEDIR);
+	textdomain(PACKAGE);
+    }
 
     /* whoami is the program name for error messages */
     whoami = argv[0];
@@ -333,7 +339,7 @@
 	    printf (_("%s: '%c' is not allowed.\n"), whoami, c);
 	    return (-1);
 	}
-	if (iscntrl (c)) {
+	if (iscntrl (c) || fixed_iscntrl (c)) {
 	    printf (_("%s: Control characters are not allowed.\n"), whoami);
 	    return (-1);
 	}
diff -urN util-linux-2.10h.orig/login-utils/newgrp.c util-linux-2.10h/login-utils/newgrp.c
--- util-linux-2.10h.orig/login-utils/newgrp.c	Fri Jul  9 06:56:38 1999
+++ util-linux-2.10h/login-utils/newgrp.c	Thu Jul 13 12:15:04 2000
@@ -59,10 +59,12 @@
     struct group *gr_entry;
     char *shell;
 
-    setlocale(LC_ALL, "");
-    bindtextdomain(PACKAGE, LOCALEDIR);
-    textdomain(PACKAGE);
-    
+    if (!getuid()) {
+	setlocale(LC_ALL, "");
+	bindtextdomain(PACKAGE, LOCALEDIR);
+	textdomain(PACKAGE);
+    }
+
     if (!(pw_entry = getpwuid(getuid()))) {
 	perror(_("newgrp: Who are you?"));
 	exit(1);
