diff -ur util-linux-2.10r.orig/mount/mntent.c util-linux-2.10r/mount/mntent.c
--- util-linux-2.10r.orig/mount/mntent.c	Mon Aug 23 16:05:31 1999
+++ util-linux-2.10r/mount/mntent.c	Sat Oct 12 10:06:43 2002
@@ -8,6 +8,7 @@
 #include <stdio.h>
 #include <string.h>		/* for index */
 #include <ctype.h>		/* for isdigit */
+#include <sys/stat.h>		/* for umask */
 #include "mntent.h"
 #include "sundries.h"		/* for xmalloc */
 #include "nls.h"
@@ -95,8 +96,11 @@
 mntFILE *
 my_setmntent (const char *file, char *mode) {
 	mntFILE *mfp = xmalloc(sizeof(*mfp));
+	mode_t saved_umask;
 
+	saved_umask = umask(077);
 	mfp->mntent_fp = fopen (file, mode);
+	umask(saved_umask);
 	mfp->mntent_file = xstrdup(file);
 	mfp->mntent_errs = (mfp->mntent_fp == NULL);
 	mfp->mntent_softerrs = 0;
diff -ur util-linux-2.10r.orig/mount/mount.c util-linux-2.10r/mount/mount.c
--- util-linux-2.10r.orig/mount/mount.c	Sun Dec  3 02:52:34 2000
+++ util-linux-2.10r/mount/mount.c	Sat Oct 12 09:50:44 2002
@@ -1336,6 +1336,8 @@
 	struct mntentchn *mc;
 	int fd;
 
+	umask(077);
+
 	setlocale(LC_ALL, "");
 	bindtextdomain(PACKAGE, LOCALEDIR);
 	textdomain(PACKAGE);
diff -ur util-linux-2.10r.orig/mount/umount.c util-linux-2.10r/mount/umount.c
--- util-linux-2.10r.orig/mount/umount.c	Sun Dec  3 02:16:32 2000
+++ util-linux-2.10r/mount/umount.c	Sat Oct 12 09:50:57 2002
@@ -426,6 +426,8 @@
   char *file;
   int result = 0;
 
+  umask(077);
+
   setlocale(LC_ALL, "");
   bindtextdomain(PACKAGE, LOCALEDIR);
   textdomain(PACKAGE);
