commit 4a64de19421cb79fa95ce5814e095b7dfebb7995
Author: geoffk1 <geoffk1>
Date:   Fri Jun 9 19:06:14 2006 +0000

    Added more Seagate Momentus 5400.2 drives.

diff --git a/knowndrives.c b/knowndrives.c
index 9494300..795504d 100644
--- a/knowndrives.c
+++ b/knowndrives.c
@@ -26,7 +26,7 @@ #include "extern.h"
 #include "knowndrives.h"
 #include "utility.h" // includes <regex.h>
 
-const char *knowndrives_c_cvsid="$Id: knowndrives.c,v 1.139 2006/04/05 19:50:07 chrfranke Exp $"
+const char *knowndrives_c_cvsid="$Id: knowndrives.c,v 1.140 2006/06/09 19:06:14 geoffk1 Exp $"
 ATACMDS_H_CVSID ATAPRINT_H_CVSID CONFIG_H_CVSID EXTERN_H_CVSID INT64_H_CVSID KNOWNDRIVES_H_CVSID UTILITY_H_CVSID;
 
 #define MODEL_STRING_LENGTH                         40
@@ -758,7 +758,7 @@ const drivesettings knowndrives[] = {
     NULL, NULL, NULL, NULL
   },
   { "Seagate Momentus 5400.2 series",
-    "^ST9(100823|808211|60822|408114|308110)A$", 
+    "^ST9(9808211|960822|808211|408114|308110|120821|10082[34]|98823|96812|94813|93811|60822)AS?$", 
     ".*",
     NULL, NULL, NULL, NULL
   },
