--- indent-2.2.9/man/texinfo2man.c.orig	2003-09-09 14:14:55 +0000
+++ indent-2.2.9/man/texinfo2man.c	2005-07-12 15:45:24 +0000
@@ -143,13 +143,23 @@ putone (char c)
 	printf (" .");
       return;
     }
+  if (c == '\'')
+    {
+      printf("\\(cq");
+      return;
+    }
+  if (c == '`')
+    {
+      printf("\\(oq");
+      return;
+    }
   if (no_break && c == '\n')
     c = ' ';
   if (no_break && c == ' ')
     putchar ('\\');
   if (to_upper)
     c = toupper (c);
-  if (c == '\'' || c == '`' || c == '\\')
+  if (c == '\\')
     putchar ('\\');
   putchar (c);
   at_start_of_line = at_start_of_line_for_dot = (c == '\n');
