--- hdparm-5.2/hdparm.c.orig	Mon Nov  4 11:36:22 2002
+++ hdparm-5.2/hdparm.c	Mon Nov  4 12:37:56 2002
@@ -18,7 +18,6 @@
 #include <sys/types.h>
 #include <sys/mount.h>
 #include <linux/types.h>
-#include <linux/fs.h>
 #include <linux/hdreg.h>
 #include <linux/major.h>
 
@@ -1496,7 +1495,6 @@
 	char c, *p;
 	char *tmpstr;
 	char name[32];
-	int neg;
 
 	if  ((progname = (char *) strrchr(*argv, '/')) == NULL)
 		progname = *argv;
