--- bind-9.3.1/bin/dig/dighost.c.reject_resolv_conf_errors	2004-11-22 18:30:31.000000000 -0500
+++ bind-9.3.1/bin/dig/dighost.c	2005-05-31 14:07:08.000000000 -0400
@@ -944,7 +944,10 @@
 	if (lwresult != LWRES_R_SUCCESS)
 		fatal("lwres_context_create failed");
 
-	(void)lwres_conf_parse(lwctx, RESOLV_CONF);
+	lwresult = lwres_conf_parse(lwctx, RESOLV_CONF);
+	if (lwresult != LWRES_R_SUCCESS)
+	        fatal("parse of %s failed", RESOLV_CONF);	    
+
 	lwconf = lwres_conf_get(lwctx);
 
 	/* Make the search list */
