--- bind-9.3.1/lib/lwres/getipnode.c.orig	2004-03-06 08:15:31 +0000
+++ bind-9.3.1/lib/lwres/getipnode.c	2005-09-20 21:35:31 +0000
@@ -331,6 +331,8 @@ lwres_getipnodebyaddr(const void *src, s
 		n = lwres_getnamebyaddr(lwrctx, LWRES_ADDRTYPE_V6, IN6ADDRSZ,
 					src, &by);
 	if (n != 0) {
+		lwres_conf_clear(lwrctx);
+		lwres_context_destroy(&lwrctx);
 		*error_num = HOST_NOT_FOUND;
 		return (NULL);
 	}
@@ -338,6 +340,7 @@ lwres_getipnodebyaddr(const void *src, s
 	lwres_gnbaresponse_free(lwrctx, &by);
 	if (he1 == NULL)
 		*error_num = NO_RECOVERY;
+	lwres_conf_clear(lwrctx);
 	lwres_context_destroy(&lwrctx);
 	return (he1);
 }
