diff -urp syslinux-4.05.orig/core/fs/iso9660/iso9660.c syslinux-4.05/core/fs/iso9660/iso9660.c
--- syslinux-4.05.orig/core/fs/iso9660/iso9660.c	2011-12-09 18:28:17 +0000
+++ syslinux-4.05/core/fs/iso9660/iso9660.c	2012-05-08 19:52:01 +0000
@@ -80,6 +80,8 @@ static bool iso_compare_name(const char
     do {
 	c1 = *p++;
 	c2 = iso_tolower(*file_name++);
+	if (c2 == '-' && c1 != c2)
+	    c2 = '_';
 
 	/* compare equal except for case? */
 	if (c1 != c2)
