--- screen-3.9.8/Makefile.in.orig	Fri Sep  1 19:07:26 2000
+++ screen-3.9.8/Makefile.in	Tue Mar  6 17:42:53 2001
@@ -50,12 +50,12 @@
 
 CFILES=	screen.c ansi.c fileio.c mark.c misc.c resize.c socket.c \
 	search.c tty.c term.c window.c utmp.c loadav.c putenv.c help.c \
-	termcap.c input.c attacher.c pty.c process.c display.c comm.c \
+	termcap.c input.c attacher.c appl_userpass.c pty.c process.c display.c comm.c \
 	kmapdef.c acls.c braille.c braille_tsi.c logfile.c layer.c \
 	sched.c teln.c nethack.c
 OFILES=	screen.o ansi.o fileio.o mark.o misc.o resize.o socket.o \
 	search.o tty.o term.o window.o utmp.o loadav.o putenv.o help.o \
-	termcap.o input.o attacher.o pty.o process.o display.o comm.o \
+	termcap.o input.o attacher.o appl_userpass.o pty.o process.o display.o comm.o \
 	kmapdef.o acls.o braille.o braille_tsi.o logfile.o layer.o \
 	sched.o teln.o nethack.o
 
--- screen-3.9.8.orig/_pam_userpass.h	Thu Jan  1 03:00:00 1970
+++ screen-3.9.8/_pam_userpass.h	Sat Mar  3 13:12:49 2001
@@ -0,0 +1,12 @@
+#ifndef __PAM_USERPASS_H
+#define __PAM_USERPASS_H
+
+#define USERPASS_AGENT_ID		"userpass"
+#define USERPASS_AGENT_ID_LENGTH	8
+
+#define USERPASS_USER_MASK		0x03
+#define USERPASS_USER_REQUIRED		1
+#define USERPASS_USER_KNOWN		2
+#define USERPASS_USER_FIXED		3
+
+#endif
--- screen-3.9.8.orig/appl_userpass.c	Thu Jan  1 03:00:00 1970
+++ screen-3.9.8/appl_userpass.c	Sat Mar  3 13:12:49 2001
@@ -0,0 +1,51 @@
+#include <string.h>
+#include <stdlib.h>
+
+#include <security/pam_appl.h>
+#include <security/pam_client.h>
+
+#include "_pam_userpass.h"
+#include "pam_userpass.h"
+
+int pam_userpass_conv(int num_msg, const struct pam_message **msg,
+	struct pam_response **resp, void *appdata_ptr)
+{
+	pam_userpass_t *userpass = (pam_userpass_t *)appdata_ptr;
+	pamc_bp_t prompt;
+	char *data;
+	char flags;
+
+	if (num_msg != 1 || msg[0]->msg_style != PAM_BINARY_PROMPT)
+		return PAM_CONV_ERR;
+
+	prompt = (pamc_bp_t)msg[0]->msg;
+	data = PAM_BP_DATA(prompt);
+
+	if (PAM_BP_CONTROL(prompt) != PAM_BPC_SELECT ||
+	    strncmp(data, USERPASS_AGENT_ID "/", USERPASS_AGENT_ID_LENGTH + 1))
+		return PAM_CONV_ERR;
+
+	flags = data[USERPASS_AGENT_ID_LENGTH + 1];
+	data += USERPASS_AGENT_ID_LENGTH + 1 + 1;
+
+	if ((flags & USERPASS_USER_MASK) == USERPASS_USER_FIXED &&
+	    strcmp(data, userpass->user))
+		return PAM_CONV_AGAIN;
+
+	if (!(*resp = malloc(sizeof(struct pam_response))))
+		return PAM_CONV_ERR;
+
+	prompt = NULL;
+	PAM_BP_RENEW(&prompt, PAM_BPC_DONE,
+		strlen(userpass->user) + 1 + strlen(userpass->pass));
+	data = PAM_BP_DATA(prompt);
+
+	strcpy(data, userpass->user);
+	data += strlen(data) + 1;
+	memcpy(data, userpass->pass, strlen(userpass->pass));
+
+	(*resp)[0].resp_retcode = 0;
+	(*resp)[0].resp = (char *)prompt;
+
+	return PAM_SUCCESS;
+}
--- screen-3.9.8/attacher.c.orig	Sun Jul  2 20:30:16 2000
+++ screen-3.9.8/attacher.c	Thu Mar  8 00:04:54 2001
@@ -33,6 +33,9 @@
 #include "screen.h"
 #include "extern.h"
 
+#include <security/pam_misc.h>
+#include "pam_userpass.h"
+
 #include <pwd.h>
 
 static sigret_t AttacherSigInt __P(SIGPROTOARG);
@@ -350,7 +353,7 @@
 
 
 #if defined(DEBUG) || !defined(DO_NOT_POLL_MASTER)
-static int AttacherPanic;
+static int AttacherPanic = 0;
 #endif
 
 #ifdef DEBUG
@@ -718,40 +721,20 @@
     }
 }				/* LockTerminal */
 
+static pam_userpass_t userpass;
+static struct pam_conv conv = {
+    &pam_userpass_conv,
+    &userpass
+};
+
+
 /* -- original copyright by Luigi Cannelloni 1985 (luigi@faui70.UUCP) -- */
 static void
 screen_builtin_lck()
 {
   char fullname[100], *cp1, message[100 + 100];
-  char *pass, mypass[9];
-
-  pass = ppp->pw_passwd;
-  if (pass == 0 || *pass == 0)
-    {
-      if ((pass = getpass("Key:   ")))
-        {
-          strncpy(mypass, pass, 8);
-          mypass[8] = 0;
-          if (*mypass == 0)
-            return;
-          if ((pass = getpass("Again: ")))
-            {
-              if (strcmp(mypass, pass))
-                {
-                  fprintf(stderr, "Passwords don't match.\007\n");
-                  sleep(2);
-                  return;
-                }
-            }
-        }
-      if (pass == 0)
-        {
-          fprintf(stderr, "Getpass error.\007\n");
-          sleep(2);
-          return;
-        }
-      pass = 0;
-    }
+  int status;
+  pam_handle_t *pamh = NULL;
 
   debug("screen_builtin_lck looking in gcos field\n");
   strncpy(fullname, ppp->pw_gecos, sizeof(fullname) - 9);
@@ -780,16 +763,28 @@
           AttacherFinit(SIGARG);
           /* NOTREACHED */
         }
-      if (pass)
-        {
-          if (!strncmp(crypt(cp1, pass), pass, strlen(pass)))
-            break;
-        }
-      else
-        {
-          if (!strcmp(cp1, mypass))
-            break;
-        }
+
+      userpass.user=ppp->pw_name;
+      userpass.pass=cp1;
+
+      status = pam_start("screen", ppp->pw_name, &conv, &pamh);
+
+      if (status == PAM_SUCCESS) 
+	{
+	  if ((status = pam_authenticate(pamh, 0)) != PAM_SUCCESS) 
+            {
+	       pam_end(pamh, status);
+	       continue;	       
+	    }
+
+	  status = pam_end(pamh, PAM_SUCCESS);
+	  if (status == PAM_SUCCESS) 
+	    {
+              memset(cp1,0,strlen(cp1));
+              userpass.pass = NULL; 
+	       break;
+	    }
+	}
       debug("screen_builtin_lck: NO!!!!!\n");
     }
   debug("password ok.\n");

--- screen-3.9.8/configure.in.orig	Wed Jun 14 17:04:17 2000
+++ screen-3.9.8/configure.in	Wed Mar  7 19:33:51 2001
@@ -1042,6 +1042,8 @@
 )
 fi
 
+AC_CHECK_LIB(pam, pam_start, LIBS="$LIBS -lpam -lpam_misc -ldl",AC_MSG_ERROR(pam required!!!))
+
 dnl AC_CHECK_HEADER(shadow.h, AC_DEFINE(SHADOWPW))
 AC_CHECKING(getspnam)
 AC_TRY_LINK([#include <shadow.h>], [getspnam("x");],AC_DEFINE(SHADOWPW))
--- screen-3.9.8.orig/pam_userpass.h	Thu Jan  1 03:00:00 1970
+++ screen-3.9.8/pam_userpass.h	Sat Mar  3 13:12:51 2001
@@ -0,0 +1,14 @@
+#ifndef _PAM_USERPASS_H
+#define _PAM_USERPASS_H
+
+#include <security/pam_appl.h>
+
+typedef struct {
+	char *user;
+	char *pass;
+} pam_userpass_t;
+
+extern int pam_userpass_conv(int num_msg, const struct pam_message **msg,
+	struct pam_response **resp, void *appdata_ptr);
+
+#endif
