--- screen-4.0.2/os.h.orig	Tue Jan  8 15:42:33 2002
+++ screen-4.0.2/os.h	Tue Dec  9 14:00:28 2003
@@ -53,6 +53,7 @@
 #endif
 #ifdef POSIX
 # include <unistd.h>
+# include <crypt.h>
 # if defined(__STDC__)
 #  include <stdlib.h>
 # endif /* __STDC__ */
@@ -270,7 +271,7 @@
 #  define pututline	pututxline
 #  define setutent	setutxent
 #  define endutent	endutxent
-#  define ut_time	ut_xtime
+#  define ut_time	ut_tv.tv_sec
 # else /* SVR4 */
 #  include <utmp.h>
 # endif /* SVR4 */
@@ -291,7 +292,7 @@
 #   ifdef _PATH_UTMP
 #    define UTMPFILE	_PATH_UTMP
 #   else
-#    define UTMPFILE	"/etc/utmp"
+#    define UTMPFILE	"/var/run/utmp"
 #   endif /* _PATH_UTMP */
 #  endif
 # endif
--- screen-4.0.2/osdef.h.in.orig	Sat Apr 28 13:26:43 2001
+++ screen-4.0.2/osdef.h.in	Tue Dec  9 14:00:28 2003
@@ -96,7 +96,6 @@
 extern int   setegid __P((int));
 #endif
 
-extern char *crypt __P((char *, char *));
 extern int   putenv __P((char *));
 
 extern int   tgetent __P((char *, char *));
