--- screen-3.9.9/os.h.orig	Sat Apr 28 16:26:43 2001
+++ screen-3.9.9/os.h	Wed Jun 13 09:43:25 2001
@@ -53,6 +53,7 @@
 #endif
 #ifdef POSIX
 # include <unistd.h>
+# include <crypt.h>
 # if defined(__STDC__)
 #  include <stdlib.h>
 # endif /* __STDC__ */
@@ -266,7 +268,7 @@
 #  define pututline	pututxline
 #  define setutent	setutxent
 #  define endutent	endutxent
-#  define ut_time	ut_xtime
+#  define ut_time	ut_tv.tv_sec
 # else /* SVR4 */
 #  include <utmp.h>
 # endif /* SVR4 */
@@ -287,7 +289,7 @@
 #   ifdef _PATH_UTMP
 #    define UTMPFILE	_PATH_UTMP
 #   else
-#    define UTMPFILE	"/etc/utmp"
+#    define UTMPFILE	"/var/run/utmp"
 #   endif /* _PATH_UTMP */
 #  endif
 # endif
--- screen-3.9.9/osdef.h.in.orig	Wed Jun 13 09:43:34 2001
+++ screen-3.9.9/osdef.h.in	Wed Jun 13 09:43:43 2001
@@ -96,7 +96,6 @@
 extern int   setegid __P((int));
 #endif
 
-extern char *crypt __P((char *, char *));
 extern int   putenv __P((char *));
 
 extern int   tgetent __P((char *, char *));
