diff -urN screen-3.9.10.orig/attacher.c screen-3.9.10/attacher.c
--- screen-3.9.10.orig/attacher.c	Sat Apr 28 17:26:42 2001
+++ screen-3.9.10/attacher.c	Sun May 19 06:59:28 2002
@@ -33,6 +33,8 @@
 #include "screen.h"
 #include "extern.h"
 
+#include <security/pam_userpass.h>
+
 #include <pwd.h>
 
 static sigret_t AttacherSigInt __P(SIGPROTOARG);
@@ -52,7 +54,7 @@
 static sigret_t AttachSigCont __P(SIGPROTOARG);
 #endif
 
-extern int real_uid, real_gid, eff_uid, eff_gid;
+extern int real_uid, real_gid, eff_uid, eff_gid, init_eff_gid;
 extern char *SockName, *SockMatch, SockPath[];
 extern struct passwd *ppp;
 extern char *attach_tty, *attach_term, *LoginName, *preselect;
@@ -720,40 +722,19 @@
     }
 }				/* LockTerminal */
 
+static pam_userpass_t userpass;
+static struct pam_conv conv = {
+    &pam_userpass_conv,
+    &userpass
+};
+
 /* -- original copyright by Luigi Cannelloni 1985 (luigi@faui70.UUCP) -- */
 static void
 screen_builtin_lck()
 {
   char fullname[100], *cp1, message[100 + 100];
-  char *pass, mypass[9];
-
-  pass = ppp->pw_passwd;
-  if (pass == 0 || *pass == 0)
-    {
-      if ((pass = getpass("Key:   ")))
-        {
-          strncpy(mypass, pass, 8);
-          mypass[8] = 0;
-          if (*mypass == 0)
-            return;
-          if ((pass = getpass("Again: ")))
-            {
-              if (strcmp(mypass, pass))
-                {
-                  fprintf(stderr, "Passwords don't match.\007\n");
-                  sleep(2);
-                  return;
-                }
-            }
-        }
-      if (pass == 0)
-        {
-          fprintf(stderr, "Getpass error.\007\n");
-          sleep(2);
-          return;
-        }
-      pass = 0;
-    }
+  int status;
+  pam_handle_t *pamh = NULL;
 
   debug("screen_builtin_lck looking in gcos field\n");
   strncpy(fullname, ppp->pw_gecos, sizeof(fullname) - 9);
@@ -782,16 +763,32 @@
           AttacherFinit(SIGARG);
           /* NOTREACHED */
         }
-      if (pass)
-        {
-          if (!strncmp(crypt(cp1, pass), pass, strlen(pass)))
-            break;
-        }
-      else
-        {
-          if (!strcmp(cp1, mypass))
-            break;
-        }
+
+      userpass.user = ppp->pw_name;
+      userpass.pass = cp1;
+
+      status = pam_start("screen", ppp->pw_name, &conv, &pamh);
+
+      if (status == PAM_SUCCESS)
+	{
+	  /* Let the PAM modules make use of our SGID privileges. */
+	  xsetegid(init_eff_gid);
+	  if ((status = pam_authenticate(pamh, 0)) != PAM_SUCCESS)
+            {
+	       xsetegid(real_gid);
+	       pam_end(pamh, status);
+	       continue;
+	    }
+	  xsetegid(real_gid);
+
+	  status = pam_end(pamh, PAM_SUCCESS);
+	  if (status == PAM_SUCCESS)
+	    {
+	       memset(cp1, 0, strlen(cp1));
+	       userpass.pass = NULL;
+	       break;
+	    }
+	}
       debug("screen_builtin_lck: NO!!!!!\n");
     }
   debug("password ok.\n");
diff -urN screen-3.9.10.orig/configure.in screen-3.9.10/configure.in
--- screen-3.9.10.orig/configure.in	Thu May 17 21:00:19 2001
+++ screen-3.9.10/configure.in	Sun May 19 06:06:26 2002
@@ -1080,6 +1080,9 @@
 )
 fi
 
+AC_CHECK_LIB(pam, pam_start, LIBS="$LIBS -lpam -ldl", AC_MSG_ERROR(PAM required!))
+AC_CHECK_LIB(pam_userpass, pam_userpass_conv, LIBS="$LIBS -lpam_userpass",AC_MSG_ERROR(pam_userpass required!))
+
 dnl AC_CHECK_HEADER(shadow.h, AC_DEFINE(SHADOWPW))
 AC_CHECKING(getspnam)
 AC_TRY_LINK([#include <shadow.h>], [getspnam("x");],AC_DEFINE(SHADOWPW))
diff -urN screen-3.9.10.orig/screen.c screen-3.9.10/screen.c
--- screen-3.9.10.orig/screen.c	Tue Sep  4 17:14:51 2001
+++ screen-3.9.10/screen.c	Sun May 19 06:57:53 2002
@@ -203,7 +203,7 @@
 
 char HostName[MAXSTR];
 int MasterPid;
-int real_uid, real_gid, eff_uid, eff_gid;
+int real_uid, real_gid, eff_uid, eff_gid, init_eff_gid;
 int default_startup;
 int ZombieKey_destroy, ZombieKey_resurrect;
 char *preselect = NULL;		/* only used in Attach() */
@@ -729,7 +729,7 @@
   real_uid = getuid();
   real_gid = getgid();
   eff_uid = geteuid();
-  eff_gid = getegid();
+  init_eff_gid = eff_gid = getegid();
   if (eff_uid != real_uid)
     {		
       /* if running with s-bit, we must install a special signal
