--- screen-3.9.9/screen.h.orig	Sat Apr 28 16:26:43 2001
+++ screen-3.9.9/screen.h	Mon Jun 11 00:54:55 2001
@@ -44,6 +44,7 @@
 #include "layer.h"
 #include "term.h"
 
+extern char bufferfile[MAXPATHLEN];
 
 #ifdef DEBUG
 # define STATIC		/* a function that the debugger should see */
--- screen-3.9.9/screen.c.orig	Fri May 25 17:40:10 2001
+++ screen-3.9.9/screen.c	Mon Jun 11 01:06:41 2001
@@ -107,6 +107,7 @@
 FILE *dfp;
 #endif
 
+char bufferfile[MAXPATHLEN];
 
 extern char *blank, *null, Term[], screenterm[], **environ, Termcap[];
 int force_vt = 1;
@@ -426,9 +427,6 @@
   logtstamp_string = SaveStr("-- %n:%t -- time-stamp -- %M/%d/%y %c:%s --\n");
   hstatusstring = SaveStr("%h");
   captionstring = SaveStr("%3n %t");
-#ifdef COPY_PASTE
-  BufferFile = SaveStr(DEFAULT_BUFFERFILE);
-#endif
   ShellProg = NULL;
 #ifdef POW_DETACH
   PowDetachString = 0;
@@ -878,6 +876,11 @@
       DebugTTY(&attach_Mode);
 #endif /* DEBUG */
     }
+
+#ifdef COPY_PASTE
+    snprintf(bufferfile,sizeof(bufferfile),"%s/.screen-exchange", home);
+    BufferFile = SaveStr(bufferfile);
+#endif
 
 #ifdef _MODE_T
   oumask = umask(0);		/* well, unsigned never fails? jw. */
--- screen-3.9.9/fileio.c.orig	Thu May  3 17:43:24 2001
+++ screen-3.9.9/fileio.c	Mon Jun 11 00:48:46 2001
@@ -377,7 +377,7 @@
 	  fnbuf[sizeof(fnbuf) - 1] = 0;
 	  fn = fnbuf;
 	}
-      public = !strcmp(fn, DEFAULT_BUFFERFILE);
+      public = !strcmp(fn, bufferfile);
 # ifdef HAVE_LSTAT
       exists = !lstat(fn, &stb);
       if (public && exists && (S_ISLNK(stb.st_mode) || stb.st_nlink > 1))
--- screen-3.9.9/process.c.orig	Tue May 15 19:42:36 2001
+++ screen-3.9.9/process.c	Mon Jun 11 00:48:46 2001
@@ -2019,7 +2019,7 @@
 #ifdef COPY_PASTE
     case RC_BUFFERFILE:
       if (*args == 0)
-	BufferFile = SaveStr(DEFAULT_BUFFERFILE);
+	BufferFile = SaveStr(bufferfile);
       else if (ParseSaveStr(act, &BufferFile))
         break;
       if (msgok)
