--- screen-4.0.2/utmp.c.orig	Mon Sep  8 14:27:17 2003
+++ screen-4.0.2/utmp.c	Wed Jan  7 22:57:45 2004
@@ -572,12 +572,11 @@
 #ifdef HAVE_UTEMPTER
   if (eff_uid && wi->w_ptyfd != -1)
     {
-      /* sigh, linux hackers made the helper functions void */
+      utempter_set_helper ("/usr/libexec/screen/utempter");
       if (SLOT_USED(u))
-	addToUtmp(wi->w_tty, host, wi->w_ptyfd);
+	return utempter_add_record (wi->w_ptyfd, host);
       else
-	removeLineFromUtmp(wi->w_tty, wi->w_ptyfd);
-      return 1;	/* pray for success */
+	return utempter_remove_record (wi->w_ptyfd);
     }
 #endif
   setutent();
