--- screen-3.9.4/doc/screen.texinfo.makeinfo	Wed Aug 18 08:13:36 1999
+++ screen-3.9.4/doc/screen.texinfo	Wed Aug 18 08:33:40 1999
@@ -1726,7 +1726,7 @@
 is @samp{splitonly}.
 
 The second form changes the text used for the caption. You can use
-all string escapes (@xref{String Escapes}). @code{Screen} uses
+all string escapes (@pxref{String Escapes}). @code{Screen} uses
 a default of @samp{%3n %t}.
 
 You can mix both forms by providing the string as an additional
@@ -2227,7 +2227,7 @@
 change after a reattach from a different terminal type, it is
 possible to bind commands to the termcap name of the keys.
 @code{Screen} will insert the correct binding after each
-reattach. @xref{Bindkey} for further details on the syntax and examples.
+reattach. (@pxref{Bindkey} for further details on the syntax and examples.)
 
 Here is the table of the default key bindings. (A) means that the
 command is executed if the keyboard is switched into application
@@ -2391,7 +2391,7 @@
 starting with @samp{(1,1)}, the terminal width and height plus the size
 of the scrollback buffer in lines, like in @samp{(80,24)+50}, 
 the current state of window XON/XOFF flow control is shown like this
-(see also @xref{Flow Control}):
+(@pxref{Flow Control}):
 @example
   +flow     automatic flow control, currently on.
   -flow     automatic flow control, currently off.
@@ -2555,7 +2555,7 @@
 @samp{.} chars). New windows have @samp{BBBB02} as default
 charset, unless a @samp{kanji} command is active.
 
-The current setting can be viewed with the @xref{Info} command.
+The current setting can be viewed with the @xref{Info}, command.
 @end deffn
 
 @deffn Command defc1 state
@@ -2910,7 +2910,7 @@
 Stuff the string @var{string} in the input buffer of the current window.
 This is like the @code{paste} command, but with much less overhead.
 You cannot paste large buffers with the @code{stuff} command. It is most
-useful for key bindings. @xref{Bindkey}
+useful for key bindings (@pxref{Bindkey}).
 
 @end deffn
 
@@ -3198,8 +3198,8 @@
 sequence of characters is encountered. There are three tables:
 one that should contain actions programmed by the user, one for
 the default actions used for terminal emulation and one for
-screen's copy mode to do cursor movement. @xref{Input Translation}
-for a list of default key bindings.
+screen's copy mode to do cursor movement. (@xref{Input Translation},
+for a list of default key bindings.)
 
 If the @samp{-d}
 option is given, bindkey modifies the default table, @samp{-m}
@@ -3878,7 +3878,7 @@
 current window (settable via @samp{\E]0;^G} or @samp{\E_\\}) is
 displayed.
 You can customize this to any string you like including
-string escapes (@xref{String Escapes}).
+string escapes (@pxref{String Escapes}).
 If you leave
 out the argument @var{string}, the current string is displayed.
 
