--- screen-4.0.2/fileio.c.orig	Mon Sep  8 14:25:28 2003
+++ screen-4.0.2/fileio.c	Wed Jan  7 23:27:03 2004
@@ -412,7 +412,7 @@
 	  fnbuf[sizeof(fnbuf) - 1] = 0;
 	  fn = fnbuf;
 	}
-      public = !strcmp(fn, DEFAULT_BUFFERFILE);
+      public = !strcmp(fn, bufferfile);
 # ifdef HAVE_LSTAT
       exists = !lstat(fn, &stb);
       if (public && exists && (S_ISLNK(stb.st_mode) || stb.st_nlink > 1))
--- screen-4.0.2/process.c.orig	Thu Sep 18 12:53:54 2003
+++ screen-4.0.2/process.c	Wed Jan  7 23:27:03 2004
@@ -2494,7 +2494,7 @@
 #ifdef COPY_PASTE
     case RC_BUFFERFILE:
       if (*args == 0)
-	BufferFile = SaveStr(DEFAULT_BUFFERFILE);
+	BufferFile = SaveStr(bufferfile);
       else if (ParseSaveStr(act, &BufferFile))
         break;
       if (msgok)
--- screen-4.0.2/screen.c.orig	Mon Sep  8 14:26:41 2003
+++ screen-4.0.2/screen.c	Wed Jan  7 23:28:18 2004
@@ -105,6 +105,7 @@
 FILE *dfp;
 #endif
 
+char bufferfile[MAXPATHLEN];
 
 extern char Term[], screenterm[], **environ, Termcap[];
 int force_vt = 1;
@@ -440,9 +441,6 @@
   timestring = SaveStr("%c:%s %M %d %H%? %l%?");
   wlisttit = SaveStr("Num Name%=Flags");
   wliststr = SaveStr("%3n %t%=%f");
-#ifdef COPY_PASTE
-  BufferFile = SaveStr(DEFAULT_BUFFERFILE);
-#endif
   ShellProg = NULL;
 #ifdef POW_DETACH
   PowDetachString = 0;
@@ -912,6 +910,11 @@
       DebugTTY(&attach_Mode);
 #endif /* DEBUG */
     }
+
+#ifdef COPY_PASTE
+    snprintf(bufferfile, sizeof(bufferfile), "%s/.screen-exchange", home);
+    BufferFile = SaveStr(bufferfile);
+#endif
 
 #ifdef _MODE_T
   oumask = umask(0);		/* well, unsigned never fails? jw. */
--- screen-4.0.2/screen.h.orig	Fri Aug 22 12:28:43 2003
+++ screen-4.0.2/screen.h	Wed Jan  7 23:27:03 2004
@@ -44,6 +44,7 @@
 #include "layer.h"
 #include "term.h"
 
+extern char bufferfile[MAXPATHLEN];
 
 #ifdef DEBUG
 # define STATIC		/* a function that the debugger should see */
