--- screen-4.0.2/etc/etcscreenrc.orig	Fri Dec  5 13:46:13 2003
+++ screen-4.0.2/etc/etcscreenrc	Tue Dec  9 14:02:31 2003
@@ -1,6 +1,6 @@
 #
 # This is an example for the global screenrc file.
-# You may want to install this file as /usr/local/etc/screenrc.
+# You may want to install this file as /etc/screenrc.
 # Check config.h for the exact location.
 #
 # Flaws of termcap and standard settings are done here.
--- screen-4.0.2/config.h.in.orig	Fri Dec  5 13:59:39 2003
+++ screen-4.0.2/config.h.in	Tue Dec  9 14:29:39 2003
@@ -64,7 +64,7 @@
  * Don't define this, if you don't want it.
  */
 #ifndef ETCSCREENRC
-# define ETCSCREENRC "/usr/local/etc/screenrc"
+# define ETCSCREENRC "/etc/screenrc"
 #endif
 
 /*
@@ -105,8 +105,8 @@
  * filesystem so screen should not even attempt to set mode or group  
  * even if running as root (e.g. on TiVo).
  */
-#undef PTYMODE
-#undef PTYGROUP
+#define PTYMODE 0620
+#define PTYGROUP 5
 #undef PTYROFS
 
 /*
@@ -180,7 +180,7 @@
 # define BLANKER_PRG
 #endif /* SIMPLESCREEN */
 
-#undef BUILTIN_TELNET
+#define BUILTIN_TELNET 1
 #undef RXVT_OSC
 #undef COLORS256
 
@@ -639,7 +639,12 @@
  * define HAVE_SVR4_PTYS if you have a /dev/ptmx character special
  * device and support the ptsname(), grantpt(), unlockpt() functions.
  */
-#undef HAVE_SVR4_PTYS
+#define HAVE_SVR4_PTYS 1
+
+/*
+ * define HAVE_GETPT if you have the getpt() function.
+ */
+#define HAVE_GETPT 1
 
 /*
  * define HAVE_GETPT if you have the getpt() function.
--- screen-4.0.2/configure.in.orig	Tue Dec  9 14:17:14 2003
+++ screen-4.0.2/configure.in	Tue Dec  9 14:18:43 2003
@@ -680,13 +680,6 @@
 AC_DEFINE(HAVE_DEV_PTC)
 fi
 
-AC_CHECKING(for SVR4 ptys)
-sysvr4ptys=
-if test -c /dev/ptmx ; then
-AC_TRY_LINK([],[ptsname(0);grantpt(0);unlockpt(0);],[AC_DEFINE(HAVE_SVR4_PTYS)
-sysvr4ptys=1])
-fi
-
 AC_CHECK_FUNCS(getpt)
 
 dnl check for openpty()
@@ -1268,10 +1261,7 @@
 
 AC_TRY_RUN(main(){exit(0);},,AC_MSG_ERROR(Can't run the compiler - internal error. Sorry.))
 
-ETCSCREENRC="\"/usr/local/etc/screenrc\""
-if test -n "$prefix"; then
-ETCSCREENRC="\"$prefix/etc/screenrc\""
-fi
+ETCSCREENRC="\"/etc/screenrc\""
 AC_MSG_CHECKING(for the global screenrc file)
 AC_ARG_WITH(sys-screenrc, [  --with-sys-screenrc=path  where to put the global screenrc file], [ ETCSCREENRC="\"${withval}\"" ])
 AC_DEFINE_UNQUOTED(ETCSCREENRC,$ETCSCREENRC)
