diff -ur screen-3.9.10.orig/pty.c screen-3.9.10/pty.c
--- screen-3.9.10.orig/pty.c	Sat Apr 28 17:26:43 2001
+++ screen-3.9.10/pty.c	Sun May 19 05:59:32 2002
@@ -75,7 +75,10 @@
 extern int eff_uid;
 
 /* used for opening a new pty-pair: */
-static char PtyName[32], TtyName[32];
+#ifndef HAVE_GETPT
+static char PtyName[32];
+#endif
+static char TtyName[32];
 
 #if !(defined(sequent) || defined(_SEQUENT_) || defined(HAVE_SVR4_PTYS))
 # ifdef hpux
@@ -243,11 +246,15 @@
 {
   register int f;
   char *m, *ptsname();
-  int unlockpt __P((int)), grantpt __P((int));
+  int unlockpt __P((int)), grantpt __P((int)), getpt __P((void));
   sigret_t (*sigcld)__P(SIGPROTOARG);
 
+#ifdef HAVE_GETPT
+  if ((f = getpt()) == -1)
+#else
   strcpy(PtyName, "/dev/ptmx");
   if ((f = open(PtyName, O_RDWR | O_NOCTTY)) == -1)
+#endif
     return -1;
 
   /*
