--- screen-3.9.9/etc/etcscreenrc.orig	Sat Apr 28 18:25:34 2001
+++ screen-3.9.9/etc/etcscreenrc	Wed Jun 13 01:04:44 2001
@@ -92,3 +92,10 @@
 bind 'O' login off
 bind '}' history
 
+# Red Hat Hack^H^H^H^HPatch
+# This makes screen treat backspaces '^?' as
+# deletes. There should be a fix in the code
+# for the way termcap inheritance works,
+# but I dont know where to put it, and this works.
+bindkey -d -k kb stuff \010
+# /Red Hat Patch
--- screen-3.9.9/etc/screenrc.orig	Fri May 25 17:50:55 2001
+++ screen-3.9.9/etc/screenrc	Wed Jun 13 01:01:13 2001
@@ -123,6 +123,15 @@
 register ] "\033:se ai\015a"
 bind ^] paste [.]
 
+# Red Hat Hack^H^H^H^HPatch
+# This makes screen treat backspaces '^?' as
+# deletes. There should be a fix in the code
+# for the way termcap inheritance works,
+# but I dont know where to put it, and this works.
+bindkey -d -k kb stuff \010
+# /Red Hat Patch
+
+
 ################
 #
 # default windows
