diff -uprk.orig screen-3.9.11.orig/utmp.c screen-3.9.11/utmp.c
--- screen-3.9.11.orig/utmp.c	2002-01-08 18:44:37 +0300
+++ screen-3.9.11/utmp.c	2002-05-26 13:45:45 +0300
@@ -576,12 +576,11 @@ struct win *wi;
 #ifdef HAVE_UTEMPTER
   if (eff_uid && wi->w_ptyfd != -1)
     {
-      /* sigh, linux hackers made the helper functions void */
+      utempter_set_helper ("/usr/libexec/screen/utempter");
       if (SLOT_USED(u))
-	addToUtmp(wi->w_tty, host, wi->w_ptyfd);
+	return utempter_add_record (wi->w_ptyfd, host);
       else
-	removeLineFromUtmp(wi->w_tty, wi->w_ptyfd);
-      return 1;	/* pray for success */
+	return utempter_remove_record (wi->w_ptyfd);
     }
 #endif
   setutent();
