--- screen-3.9.10/ansi.c.orig	Tue Sep  4 13:13:37 2001
+++ screen-3.9.10/ansi.c	Sat Dec 13 23:40:27 2003
@@ -596,14 +596,16 @@
 	      break;
 	    case ';':
 	    case ':':
-	      curr->w_NumArgs++;
+	      if (curr->w_NumArgs < MAXARGS)
+	        curr->w_NumArgs++;
 	      break;
 	    default:
 	      if (Special(c))
 		break;
 	      if (c >= '@' && c <= '~')
 		{
-		  curr->w_NumArgs++;
+		  if (curr->w_NumArgs < MAXARGS)
+		    curr->w_NumArgs++;
 		  DoCSI(c, curr->w_intermediate);
 		  if (curr->w_state != PRIN)
 		    curr->w_state = LIT;
@@ -1163,6 +1165,8 @@
 	    a1 = curr->w_width;
 	  if (a2 < 1)
 	    a2 = curr->w_height;
+	  if (a1 > 10000 || a2 > 10000)
+	    break;
 	  WChangeSize(curr, a1, a2);
 	  cols = curr->w_width;
 	  rows = curr->w_height;
