--- screen-4.0.2/doc/screen.1.orig	Fri Dec  5 13:51:57 2003
+++ screen-4.0.2/doc/screen.1	Wed Jan  7 23:22:09 2004
@@ -480,7 +480,7 @@
 Show the number (and title) of the current window.
 .IP "\fBC-a backspace\fP"
 .PD 0
-.IP "\fBC-a h\fP"
+.IP "\fBC-a C-h\fP"
 .IP "\fBC-a p\fP"
 .IP "\fBC-a C-p\fP	(prev)"
 .PD
@@ -543,7 +543,7 @@
 Write out a \*Q.termcap\*U file.
 .IP "\fBC-a ?\fP	(help)"
 Show key bindings.
-.IP "\fBC-a C-\e\fP	(quit)"
+.IP "\fBC-a \e\fP	(quit)"
 Kill all windows and terminate
 .IR screen .
 .IP "\fBC-a :\fP	(colon)"
--- screen-4.0.2/doc/screen.texinfo.orig	Fri Dec  5 13:51:46 2003
+++ screen-4.0.2/doc/screen.texinfo	Wed Jan  7 23:24:08 2004
@@ -753,7 +753,7 @@
 (help)@*
 Show key bindings.  @xref{Help}.
 
-@item @kbd{C-a C-\}
+@item @kbd{C-a \}
 (quit)@*
 Kill all windows and terminate @code{screen}.  @xref{Quit}.
 
@@ -1772,9 +1772,9 @@
 
 @node Quit,  , Suspend, Session Management
 @section Quit
-@kindex C-\
+@kindex \
 @deffn Command quit
-(@kbd{C-a C-\})@*
+(@kbd{C-a \})@*
 Kill all windows and terminate @code{screen}.  Note that on VT100-style
 terminals the keys @kbd{C-4} and @kbd{C-\} are identical.  So be careful
 not to type @kbd{C-a C-4} when selecting window no. 4.  Use the empty
@@ -4197,7 +4197,7 @@
 @kindex h
 @kindex C-h
 @deffn Command hardcopy [-h] [@var{file}]
-(@kbd{C-a h}, @kbd{C-a C-h})@*
+(@kbd{C-a h})@*
 Writes out the currently displayed image to the file @var{file}, or,
 if no filename is specified, to @file{hardcopy.@var{n}}
 in the default directory, where @var{n} is the number of the
