diff -puNr texinfo-4.8.orig/util/install-info.c texinfo-4.8/util/install-info.c
--- texinfo-4.8.orig/util/install-info.c	Sun Apr 11 17:56:47 2004
+++ texinfo-4.8/util/install-info.c	Wed Mar 30 02:20:26 2005
@@ -399,7 +399,7 @@ menu_item_equal (const char *item, char 
          ITEM.  The various suffixes should never actually appear in the
          dir file, but sometimes people put them in.  */
       static char *suffixes[]
-        = { "", ".info.gz", ".info", ".inf", ".gz",
+        = { "", ".info.gz", ".info.bz2", ".info", ".inf", ".gz", ".bz2",
 #ifdef __MSDOS__
             ".inz", ".igz",
 #endif
@@ -605,31 +605,15 @@ open_possibly_compressed_file (char *fil
     compression_program = &local_compression_program;
 
   if (data[0] == '\x1f' && data[1] == '\x8b')
-#if STRIP_DOT_EXE
-    /* An explicit .exe yields a better diagnostics from popen below
-       if they don't have gzip installed.  */
-    *compression_program = "gzip.exe";
-#else
-    *compression_program = "gzip";
-#endif
-  else if(data[0] == 'B' && data[1] == 'Z' && data[2] == 'h')
-#ifndef STRIP_DOT_EXE
-    *compression_program = "bzip2.exe";
-#else
-    *compression_program = "bzip2";
-#endif
-  else if(data[0] == 'B' && data[1] == 'Z' && data[2] == '0')
-#ifndef STRIP_DOT_EXE
-    *compression_program = "bzip.exe";
-#else
-    *compression_program = "bzip";
-#endif
+    *compression_program = "/bin/zcat";
+  else if(data[0] == 'B' && data[1] == 'Z' && (data[2] == 'h' || data[2] == '0'))
+    *compression_program = "/usr/bin/bzcat";
   else
     *compression_program = NULL;
 
   if (*compression_program)
     { /* It's compressed, so fclose the file and then open a pipe.  */
-      char *command = concat (*compression_program," -cd <", *opened_filename);
+      char *command = concat (*compression_program," ", *opened_filename);
       if (fclose (f) < 0)
         pfatal_with_name (*opened_filename);
       f = popen (command, "r");
