--- texinfo-4.8.orig/util/texindex.c
+++ texinfo-4.8/util/texindex.c
@@ -837,7 +837,7 @@ readline (struct linebuffer *linebuffer,
         {
           buffer = (char *) xrealloc (buffer, linebuffer->size *= 2);
           p += buffer - linebuffer->buffer;
-          end += buffer - linebuffer->buffer;
+          end = buffer + linebuffer->size;
           linebuffer->buffer = buffer;
         }
       if (c < 0 || c == '\n')
