--- texinfo-4.2/info/infodoc.c.orig	Wed Feb 27 15:37:33 2002
+++ texinfo-4.2/info/infodoc.c	Thu Jun 20 15:23:40 2002
@@ -184,6 +184,10 @@
 
 #endif /* !INFOKEY */
 
+static char *where_is_rep = (char *)NULL;
+static int where_is_rep_index = 0;
+static int where_is_rep_size = 0;
+
 static char *where_is_internal ();
 
 void
@@ -328,6 +332,8 @@
         {
           InfoCommand *cmd = DocInfoCmd(&function_doc_array[i]);
 
+          where_is_rep_index = 0;
+
           if (InfoFunction(cmd) != info_do_lowercase_version
 	      && !where_is_internal (info_keymap, cmd)
               && !where_is_internal (echo_area_keymap, cmd))
@@ -1052,10 +1058,6 @@
 
 /* Return a string of characters which could be typed from the keymap
    MAP to invoke FUNCTION. */
-static char *where_is_rep = (char *)NULL;
-static int where_is_rep_index = 0;
-static int where_is_rep_size = 0;
-
 char *
 where_is (map, cmd)
      Keymap map;
--- texinfo-4.2/info/infomap.c.orig	Sat Feb  9 01:02:53 2002
+++ texinfo-4.2/info/infomap.c	Thu Jun 20 15:23:40 2002
@@ -505,8 +505,8 @@
   map['b'].function = ea_backward_word;
   map['d'].function = ea_kill_word;
   map['f'].function = ea_forward_word;
-  map['h'].function = ea_forward;
-  map['l'].function = ea_backward;
+  map['h'].function = ea_backward;
+  map['l'].function = ea_forward;
   map['w'].function = ea_forward_word;
   map['x'].function = ea_delete;
   map['X'].function = ea_kill_word;
--- texinfo-4.2/info/session.c.orig	Sat Mar  2 17:05:07 2002
+++ texinfo-4.2/info/session.c	Thu Jun 20 15:23:40 2002
@@ -2110,10 +2110,13 @@
                             break;
                           }
                       }
-                    if (closest == -1)
-                      which--;
-                    else
-                      which = closest;
+                   if (which > 0)
+                     {
+                       if (closest == -1)
+                         which--;
+                       else
+                         which = closest;
+                     }
                   }
 
                 defentry = (REFERENCE *)xmalloc (sizeof (REFERENCE));
