--- texinfo-4.2/util/texi2dvi.orig	Mon Apr  1 17:20:59 2002
+++ texinfo-4.2/util/texi2dvi	Wed Jun 26 16:34:05 2002
@@ -90,7 +90,6 @@
 quiet=          # by default let the tools' message be displayed
 set_language=
 textra=
-tmpdir=${TMPDIR:-/tmp}/t2d$$  # avoid collisions on 8.3 filesystems.
 txincludes=     # TEXINPUTS extensions
 txiprereq=19990129 # minimum texinfo.tex version to have macro expansion
 verbose=false   # echo for verbose mode
@@ -198,12 +197,11 @@
 esac
 
 # Prepare the temporary directory.  Remove it at exit, unless debugging.
+tmpdir=`mktemp -dt texi2dvi.XXXXXXXXXX` || exit 1
+
 if test -z "$debug"; then
-  trap "cd / && rm -rf $tmpdir" 0 1 2 15
+  trap "rm -rf $tmpdir" EXIT HUP INT TERM
 fi
-
-# Create the temporary directory with strict rights
-(umask 077 && mkdir $tmpdir) || exit 1
 
 # Prepare the tools we might need.  This may be extra work in some
 # cases, but improves the readibility of the script.
