diff -uNp -r flex-2.5.35.old/tests/test-c++-basic/scanner.l flex-2.5.35/tests/test-c++-basic/scanner.l
--- flex-2.5.35.old/tests/test-c++-basic/scanner.l	2002-08-22 01:11:07 +0000
+++ flex-2.5.35/tests/test-c++-basic/scanner.l	2011-09-13 14:12:28.781215346 +0000
@@ -24,6 +24,7 @@
 %{
 
 #include "config.h"
+#include <stdio.h>
 
 %}
 
diff -uNp -r flex-2.5.35.old/tests/test-c++-multiple-scanners/scanner-1.l flex-2.5.35/tests/test-c++-multiple-scanners/scanner-1.l
--- flex-2.5.35.old/tests/test-c++-multiple-scanners/scanner-1.l	2002-10-01 15:10:12 +0000
+++ flex-2.5.35/tests/test-c++-multiple-scanners/scanner-1.l	2011-09-13 14:14:57.631433667 +0000
@@ -21,6 +21,7 @@
 
 %{
 #include "config.h"
+#include <stdio.h>
 
 %}
 
diff -uNp -r flex-2.5.35.old/tests/test-c++-multiple-scanners/scanner-2.l flex-2.5.35/tests/test-c++-multiple-scanners/scanner-2.l
--- flex-2.5.35.old/tests/test-c++-multiple-scanners/scanner-2.l	2006-10-22 22:17:38 +0000
+++ flex-2.5.35/tests/test-c++-multiple-scanners/scanner-2.l	2011-09-13 14:15:07.251447779 +0000
@@ -21,6 +21,7 @@
 
 %{
 #include "config.h"
+#include <stdio.h>
 
 %}
 
