diff -urp diffstat-1.51.orig/diffstat.c diffstat-1.51/diffstat.c
--- diffstat-1.51.orig/diffstat.c	2009-11-08 01:59:15 +0000
+++ diffstat-1.51/diffstat.c	2009-11-18 04:47:12 +0000
@@ -1946,7 +1946,13 @@ copy_stdin(char **dirpath)
 	} else {
 	    free(result);
 	    result = 0;
+	    rmdir(*dirpath); /* Assume that the /stdin file was not created */
+	    free(*dirpath);
+	    *dirpath = 0;
 	}
+    } else {
+	free(*dirpath);
+	*dirpath = 0;
     }
     return result;
 }
@@ -2251,7 +2257,11 @@ main(int argc, char *argv[])
 	    free(command);
 
 	    unlink(myfile);
+	    free(myfile);
+	    myfile = 0;
 	    rmdir(stdin_dir);
+	    free(stdin_dir);
+	    stdin_dir = 0;
 	} else if (which != dcEmpty)
 #endif
 	    do_file(stdin, "stdin");
