diff -ur traceroute-1.4a12.orig/traceroute.c traceroute-1.4a12/traceroute.c
--- traceroute-1.4a12.orig/traceroute.c	Thu Dec 14 11:04:50 2000
+++ traceroute-1.4a12/traceroute.c	Sun Jul  7 22:14:04 2002
@@ -296,8 +296,8 @@
 int s;				/* receive (icmp) socket file descriptor */
 int sndsock;			/* send (udp/icmp) socket file descriptor */
 
-struct sockaddr whereto;	/* Who to try to reach */
-struct sockaddr wherefrom;	/* Who we are */
+struct sockaddr_in whereto;	/* Who to try to reach */
+struct sockaddr_in wherefrom;	/* Who we are */
 int packlen;			/* total length of packet */
 int minpacket;			/* min ip packet size */
 int maxpacket = 32 * 1024;	/* max ip packet size */
