diff -uprk.orig bzip2-1.0.3.orig/bzip2.c bzip2-1.0.3/bzip2.c
--- bzip2-1.0.3.orig/bzip2.c	2005-05-05 10:04:37 +0000
+++ bzip2-1.0.3/bzip2.c	2005-05-05 10:33:08 +0000
@@ -1725,10 +1725,10 @@ void testf ( Char *name )
 
 /*---------------------------------------------*/
 static 
+__attribute__ ((__noreturn__))
 void license ( void )
 {
-   fprintf ( stderr,
-
+   printf (
     "bzip2, a block-sorting file compressor.  "
     "Version %s.\n"
     "   \n"
@@ -1745,15 +1745,17 @@ void license ( void )
     "   \n",
     BZ2_bzlibVersion()
    );
+   exit ( 0 );
 }
 
 
 /*---------------------------------------------*/
 static 
-void usage ( const char *fullProgName )
+__attribute__ ((__noreturn__))
+void usage ( const char *fullProgName, int code )
 {
    fprintf (
-      stderr,
+      code ? stderr : stdout,
       "bzip2, a block-sorting file compressor.  "
       "Version %s.\n"
       "\n   usage: %s [flags and input files in any order]\n"
@@ -1789,6 +1791,7 @@ void usage ( const char *fullProgName )
       BZ2_bzlibVersion(),
       fullProgName
    );
+   exit ( code );
 }
 
 
@@ -2016,13 +2019,11 @@ IntNative main ( IntNative argc, Char *a
                case 'V':
                case 'L': license();            break;
                case 'v': verbosity++; break;
-               case 'h': usage ( progName );
-                         exit ( 0 );
+               case 'h': usage ( progName, 0 );
                          break;
                default:  fprintf ( stderr, "%s: Bad flag `%s'\n",
                                    progName, aa->name );
-                         usage ( progName );
-                         exit ( 1 );
+                         usage ( progName, 1 );
                          break;
             }
          }
@@ -2049,12 +2050,11 @@ IntNative main ( IntNative argc, Char *a
       if (ISFLAG("--fast"))              blockSize100k = 1;          else
       if (ISFLAG("--best"))              blockSize100k = 9;          else
       if (ISFLAG("--verbose"))           verbosity++;                else
-      if (ISFLAG("--help"))              { usage ( progName ); exit ( 0 ); }
+      if (ISFLAG("--help"))              { usage ( progName, 0 ); }
          else
          if (strncmp ( aa->name, "--", 2) == 0) {
             fprintf ( stderr, "%s: Bad flag `%s'\n", progName, aa->name );
-            usage ( progName );
-            exit ( 1 );
+            usage ( progName , 1 );
          }
    }
 
