diff -ur bzip2-1.0.1.orig/bzip2.c bzip2-1.0.1/bzip2.c
--- bzip2-1.0.1.orig/bzip2.c	Wed Apr 26 01:34:06 2000
+++ bzip2-1.0.1/bzip2.c	Thu Jan 24 18:45:29 2002
@@ -566,6 +566,8 @@
    if (ret == EOF) goto errhandler_io;
    if (zStream != stdout) {
       ret = fclose ( zStream );
+      if (zStream == outputHandleJustInCase)
+	outputHandleJustInCase = NULL;
       if (ret == EOF) goto errhandler_io;
    }
    if (ferror(stream)) goto errhandler_io;
