diff -ur bzip2-1.0.3.orig/Makefile bzip2-1.0.3/Makefile
--- bzip2-1.0.3.orig/Makefile	Thu Feb 17 11:28:24 2005
+++ bzip2-1.0.3/Makefile	Thu May  5 21:55:40 2005
@@ -12,6 +12,7 @@
 
 # Where you want it installed when you do 'make install'
 PREFIX=/usr
+MANDIR=$(PREFIX)/man
 
 
 OBJS= blocksort.o  \
@@ -59,19 +60,17 @@
 install: bzip2 bzip2recover
 	if ( test ! -d $(PREFIX)/bin ) ; then mkdir -p $(PREFIX)/bin ; fi
 	if ( test ! -d $(PREFIX)/lib ) ; then mkdir -p $(PREFIX)/lib ; fi
-	if ( test ! -d $(PREFIX)/man ) ; then mkdir -p $(PREFIX)/man ; fi
-	if ( test ! -d $(PREFIX)/man/man1 ) ; then mkdir -p $(PREFIX)/man/man1 ; fi
+	if ( test ! -d $(MANDIR) ) ; then mkdir -p $(MANDIR) ; fi
+	if ( test ! -d $(MANDIR)/man1 ) ; then mkdir -p $(MANDIR)/man1 ; fi
 	if ( test ! -d $(PREFIX)/include ) ; then mkdir -p $(PREFIX)/include ; fi
 	cp -f bzip2 $(PREFIX)/bin/bzip2
-	cp -f bzip2 $(PREFIX)/bin/bunzip2
-	cp -f bzip2 $(PREFIX)/bin/bzcat
+	ln $(PREFIX)/bin/bzip2 $(PREFIX)/bin/bunzip2
+	ln $(PREFIX)/bin/bzip2 $(PREFIX)/bin/bzcat
 	cp -f bzip2recover $(PREFIX)/bin/bzip2recover
 	chmod a+x $(PREFIX)/bin/bzip2
-	chmod a+x $(PREFIX)/bin/bunzip2
-	chmod a+x $(PREFIX)/bin/bzcat
 	chmod a+x $(PREFIX)/bin/bzip2recover
-	cp -f bzip2.1 $(PREFIX)/man/man1
-	chmod a+r $(PREFIX)/man/man1/bzip2.1
+	cp -f bzip2.1 $(MANDIR)/man1
+	chmod a+r $(MANDIR)/man1/bzip2.1
 	cp -f bzlib.h $(PREFIX)/include
 	chmod a+r $(PREFIX)/include/bzlib.h
 	cp -f libbz2.a $(PREFIX)/lib
@@ -86,14 +85,14 @@
 	cp -f bzdiff $(PREFIX)/bin/bzdiff
 	ln $(PREFIX)/bin/bzdiff $(PREFIX)/bin/bzcmp
 	chmod a+x $(PREFIX)/bin/bzdiff
-	cp -f bzgrep.1 bzmore.1 bzdiff.1 $(PREFIX)/man/man1
-	chmod a+r $(PREFIX)/man/man1/bzgrep.1
-	chmod a+r $(PREFIX)/man/man1/bzmore.1
-	chmod a+r $(PREFIX)/man/man1/bzdiff.1
-	echo ".so man1/bzgrep.1" > $(PREFIX)/man/man1/bzegrep.1
-	echo ".so man1/bzgrep.1" > $(PREFIX)/man/man1/bzfgrep.1
-	echo ".so man1/bzmore.1" > $(PREFIX)/man/man1/bzless.1
-	echo ".so man1/bzdiff.1" > $(PREFIX)/man/man1/bzcmp.1
+	cp -f bzgrep.1 bzmore.1 bzdiff.1 $(MANDIR)/man1
+	chmod a+r $(MANDIR)/man1/bzgrep.1
+	chmod a+r $(MANDIR)/man1/bzmore.1
+	chmod a+r $(MANDIR)/man1/bzdiff.1
+	echo ".so man1/bzgrep.1" > $(MANDIR)/man1/bzegrep.1
+	echo ".so man1/bzgrep.1" > $(MANDIR)/man1/bzfgrep.1
+	echo ".so man1/bzmore.1" > $(MANDIR)/man1/bzless.1
+	echo ".so man1/bzdiff.1" > $(MANDIR)/man1/bzcmp.1
 
 clean: 
 	rm -f *.o libbz2.a bzip2 bzip2recover \
diff -ur bzip2-1.0.3.orig/Makefile-libbz2_so bzip2-1.0.3/Makefile-libbz2_so
--- bzip2-1.0.3.orig/Makefile-libbz2_so	Sun Oct 17 17:32:13 2004
+++ bzip2-1.0.3/Makefile-libbz2_so	Thu May  5 21:55:40 2005
@@ -20,13 +20,15 @@
       bzlib.o
 
 all: $(OBJS)
-	$(CC) -shared -Wl,-soname -Wl,libbz2.so.1.0 -o libbz2.so.1.0.3 $(OBJS)
-	$(CC) $(CFLAGS) -o bzip2-shared bzip2.c libbz2.so.1.0.3
-	rm -f libbz2.so.1.0
-	ln -s libbz2.so.1.0.3 libbz2.so.1.0
+	$(CC) -shared -Wl,-soname -Wl,libbz2.so.1 -o libbz2.so.1.0.3 $(OBJS)
 
-clean: 
-	rm -f $(OBJS) bzip2.o libbz2.so.1.0.3 libbz2.so.1.0 bzip2-shared
+install: libbz2.so.1.0.3
+	install -m 755 libbz2.so.1.0.3 $(PREFIX)/lib/
+	ln -sf libbz2.so.1.0.3 $(PREFIX)/lib/libbz2.so.1
+	ln -sf libbz2.so.1 $(PREFIX)/lib/libbz2.so
+
+clean:
+	rm -f $(OBJS) bzip2.o libbz2.so.1.0.3 bzip2-shared
 
 blocksort.o: blocksort.c
 	$(CC) $(CFLAGS) -c blocksort.c
