--- bzip2/Makefile.am
+++ bzip2/Makefile.am
@@ -21,7 +21,11 @@ libbz2_la_SOURCES = \
 	compress.c \
 	decompress.c \
 	bzlib.c
-libbz2_la_LDFLAGS = -version-info 1:6:0
+libbz2_la_LDFLAGS = -version-info 1:6:0 -Wl,--version-script=libbz2.map
+libbz2_la_DEPENDENCIES = libbz2.map
+
+libbz2.map: libbz2.map.sh
+	sh $^ >$@
 
 check-local: samples.sh bzip2
 	./samples.sh
--- /dev/null
+++ bzip2/libbz2.map.sh
@@ -0,0 +1,11 @@
+#!/bin/sh
+
+cat <<__EOF__
+
+{
+  global:
+$(sed -ne 's/^BZ_EXTERN .*BZ_API(\([^)]\+\)).*/    \1;/pg' bzlib.h)
+  local:
+    *;
+};
+__EOF__
