diff -ur bash-2.05-cwru-fixes/config-top.h bash-2.05/config-top.h
--- bash-2.05-cwru-fixes/config-top.h	Wed Feb  7 21:28:14 2001
+++ bash-2.05/config-top.h	Thu Oct 25 21:59:55 2001
@@ -34,14 +34,14 @@
 /* The default value of the PATH variable. */
 #ifndef DEFAULT_PATH_VALUE
 #define DEFAULT_PATH_VALUE \
-  "/usr/gnu/bin:/usr/local/bin:/usr/ucb:/bin:/usr/bin:."
+  "/bin:/usr/bin:/usr/local/bin:/sbin:/usr/sbin:/usr/local/sbin"
 #endif
 
 /* The value for PATH when invoking `command -p'.  This is only used when
    the Posix.2 confstr () function, or CS_PATH define are not present. */
 #ifndef STANDARD_UTILS_PATH
 #define STANDARD_UTILS_PATH \
-  "/bin:/usr/bin:/usr/ucb:/sbin:/usr/sbin:/etc:/usr/etc"
+  "/bin:/usr/bin:/usr/sbin:/sbin"
 #endif
 
 /* Default primary and secondary prompt strings. */
