diff -upk.orig bash-3.1.orig/bashline.c bash-3.1/bashline.c
--- bash-3.1.orig/bashline.c	2006-01-05 20:37:30 +0000
+++ bash-3.1/bashline.c	2006-01-05 20:38:16 +0000
@@ -800,7 +800,7 @@ operate_and_get_next (count, c)
    command being entered (if no explicit argument is given), otherwise on
    a command from the history file. */
 
-#define VI_EDIT_COMMAND		"fc -e \"${VISUAL:-${EDITOR:-vi}}\""
+#define VI_EDIT_COMMAND		"fc -e \"${VISUAL:-${EDITOR:-vitmp}}\""
 #define EMACS_EDIT_COMMAND	"fc -e \"${VISUAL:-${EDITOR:-emacs}}\""
 #define POSIX_VI_EDIT_COMMAND	"fc -e vi"
 
diff -upk.orig bash-3.1.orig/builtins/fc.def bash-3.1/builtins/fc.def
--- bash-3.1.orig/builtins/fc.def	2006-01-05 20:37:31 +0000
+++ bash-3.1/builtins/fc.def	2006-01-05 20:39:01 +0000
@@ -31,7 +31,7 @@ string, which means the most recent comm
 string.
 
    -e ENAME selects which editor to use.  Default is FCEDIT, then EDITOR,
-      then vi.
+      then vitmp.
 
    -l means list lines instead of editing.
    -n means no line numbers listed.
@@ -108,7 +108,7 @@ extern int delete_last_history __P((void
 
    -e ENAME selects which editor to use.  Default is FCEDIT, then EDITOR,
       then the editor which corresponds to the current readline editing
-      mode, then vi.
+      mode, then vitmp.
 
    -l means list lines instead of editing.
    -n means no line numbers listed.
@@ -156,7 +156,7 @@ static void fc_addhist __P((char *));
 #endif
 
 /* String to execute on a file that we want to edit. */
-#define FC_EDIT_COMMAND "${FCEDIT:-${EDITOR:-vi}}"
+#define FC_EDIT_COMMAND "${FCEDIT:-${EDITOR:-vitmp}}"
 #if defined (STRICT_POSIX)
 #  define POSIX_FC_EDIT_COMMAND "${FCEDIT:-ed}"
 #else
diff -upk.orig bash-3.1.orig/doc/bash.1 bash-3.1/doc/bash.1
--- bash-3.1.orig/doc/bash.1	2005-10-12 15:40:52 +0000
+++ bash-3.1/doc/bash.1	2006-01-05 20:38:16 +0000
@@ -6834,7 +6834,7 @@ if
 .SM
 .B FCEDIT
 is not set.  If neither variable is set,
-.FN vi
+.FN vitmp
 is used.  When editing is complete, the edited commands are
 echoed and executed.
 .sp 1
diff -upk.orig bash-3.1.orig/doc/bashbug.1 bash-3.1/doc/bashbug.1
--- bash-3.1.orig/doc/bashbug.1	1998-07-30 20:22:49 +0000
+++ bash-3.1/doc/bashbug.1	2005-05-07 18:15:30 +0000
@@ -35,7 +35,8 @@ Specifies the preferred editor. If
 is not set,
 .B bashbug
 defaults to
-.BR emacs .
+an instance of
+.BR vi .
 .TP
 .B HOME
 Directory in which the failed bug report is saved if the mail fails.
diff -upk.orig bash-3.1.orig/support/bashbug.sh bash-3.1/support/bashbug.sh
--- bash-3.1.orig/support/bashbug.sh	2006-01-05 20:37:31 +0000
+++ bash-3.1/support/bashbug.sh	2006-01-05 20:40:23 +0000
@@ -117,24 +117,8 @@ esac
 BUGADDR="${1-$BUGBASH}"
 
 if [ -z "$DEFEDITOR" ] && [ -z "$EDITOR" ]; then
-	if [ -x /usr/bin/editor ]; then
-		DEFEDITOR=editor
-	elif [ -x /usr/local/bin/ce ]; then
-		DEFEDITOR=ce
-	elif [ -x /usr/local/bin/emacs ]; then
-		DEFEDITOR=emacs
-	elif [ -x /usr/contrib/bin/emacs ]; then
-		DEFEDITOR=emacs
-	elif [ -x /usr/bin/emacs ]; then
-		DEFEDITOR=emacs
-	elif [ -x /usr/bin/xemacs ]; then
-		DEFEDITOR=xemacs
-	elif [ -x /usr/contrib/bin/jove ]; then
-		DEFEDITOR=jove
-	elif [ -x /usr/local/bin/jove ]; then
-		DEFEDITOR=jove
-	elif [ -x /usr/bin/vi ]; then
-		DEFEDITOR=vi
+	if [ -x /bin/vitmp ]; then
+		DEFEDITOR=vitmp
 	else
 		echo "$0: No default editor found: attempting to use vi" >&2
 		DEFEDITOR=vi
