--- bash-2.03/config-top.h.profile	Mon Feb 22 14:37:17 1999
+++ bash-2.03/config-top.h	Mon Feb 22 14:37:29 1999
@@ -3,6 +3,8 @@
 /* This contains various user-settable options not under the control of
    autoconf. */
 
+#define NON_INTERACTIVE_LOGIN_SHELLS
+
 /* Define CONTINUE_AFTER_KILL_ERROR if you want the kill command to
    continue processing arguments after one of them fails.  This is
    what POSIX.2 specifies. */
